{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5c0cea95",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:52:39.248675Z",
     "iopub.status.busy": "2022-09-26T14:52:39.248188Z",
     "iopub.status.idle": "2022-09-26T14:52:51.158211Z",
     "shell.execute_reply": "2022-09-26T14:52:51.157462Z"
    },
    "papermill": {
     "duration": 11.972936,
     "end_time": "2022-09-26T14:52:51.215171",
     "exception": false,
     "start_time": "2022-09-26T14:52:39.242235",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pycisTopic\n",
    "import glob\n",
    "import os\n",
    "import pybiomart as pbm\n",
    "import pandas as pd\n",
    "import pickle\n",
    "from pycisTopic.qc import *\n",
    "from IPython.display import Image, display\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "\n",
    "import multiprocess as mp  # for kde multithreading calculation\n",
    "from multiprocess import Pool\n",
    "\n",
    "%matplotlib inline\n",
    "%load_ext lab_black"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c7a43c8",
   "metadata": {
    "papermill": {
     "duration": 0.003189,
     "end_time": "2022-09-26T14:52:51.222790",
     "exception": false,
     "start_time": "2022-09-26T14:52:51.219601",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Download annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8d799645-cbe0-43be-b2af-cb773840f783",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:52:51.230783Z",
     "iopub.status.busy": "2022-09-26T14:52:51.229877Z",
     "iopub.status.idle": "2022-09-26T14:52:51.467719Z",
     "shell.execute_reply": "2022-09-26T14:52:51.466905Z"
    },
    "papermill": {
     "duration": 0.244438,
     "end_time": "2022-09-26T14:52:51.470420",
     "exception": false,
     "start_time": "2022-09-26T14:52:51.225982",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f587154d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:52:51.479411Z",
     "iopub.status.busy": "2022-09-26T14:52:51.478861Z",
     "iopub.status.idle": "2022-09-26T14:52:51.483989Z",
     "shell.execute_reply": "2022-09-26T14:52:51.483390Z"
    },
    "papermill": {
     "duration": 0.010329,
     "end_time": "2022-09-26T14:52:51.484887",
     "exception": false,
     "start_time": "2022-09-26T14:52:51.474558",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "wdir = \"/dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus\"\n",
    "os.chdir(wdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8abb4492",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:52:51.492686Z",
     "iopub.status.busy": "2022-09-26T14:52:51.492127Z",
     "iopub.status.idle": "2022-09-26T14:52:51.584747Z",
     "shell.execute_reply": "2022-09-26T14:52:51.583950Z"
    },
    "papermill": {
     "duration": 0.097508,
     "end_time": "2022-09-26T14:52:51.585772",
     "exception": false,
     "start_time": "2022-09-26T14:52:51.488264",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading cached genome annotation...\n"
     ]
    }
   ],
   "source": [
    "genome = \"hg38\"\n",
    "\n",
    "pbm_genome_name_dict = {\n",
    "    \"hg38\": \"hsapiens_gene_ensembl\",\n",
    "    \"hg37\": \"hsapiens_gene_ensembl\",\n",
    "    \"mm10\": \"mmusculus_gene_ensembl\",\n",
    "    \"dm6\": \"dmelanogaster_gene_ensembl\",\n",
    "}\n",
    "\n",
    "pbm_host_dict = {\n",
    "    \"hg38\": \"http://www.ensembl.org\",\n",
    "    \"hg37\": \"http://grch37.ensembl.org/\",\n",
    "    \"mm10\": \"http://nov2020.archive.ensembl.org/\",\n",
    "    \"dm6\": \"http://www.ensembl.org\",\n",
    "}\n",
    "\n",
    "if os.path.exists(f\"annotation.tsv\"):\n",
    "    print(f\"Loading cached genome annotation...\")\n",
    "    annotation = pd.read_csv(\"annotation.tsv\", sep=\"\\t\", header=0, index_col=0)\n",
    "else:\n",
    "    dataset = pbm.Dataset(name=pbm_genome_name_dict[genome], host=pbm_host_dict[genome])\n",
    "\n",
    "    annotation = dataset.query(\n",
    "        attributes=[\n",
    "            \"chromosome_name\",\n",
    "            \"transcription_start_site\",\n",
    "            \"strand\",\n",
    "            \"external_gene_name\",\n",
    "            \"transcript_biotype\",\n",
    "        ]\n",
    "    )\n",
    "    filter = annotation[\"Chromosome/scaffold name\"].str.contains(\"CHR|GL|JH|MT\")\n",
    "    annotation = annotation[~filter]\n",
    "    annotation[\"Chromosome/scaffold name\"] = annotation[\n",
    "        \"Chromosome/scaffold name\"\n",
    "    ].str.replace(r\"(\\b\\S)\", r\"chr\\1\")\n",
    "    annotation.columns = [\"Chromosome\", \"Start\", \"Strand\", \"Gene\", \"Transcript_type\"]\n",
    "    annotation = annotation[annotation.Transcript_type == \"protein_coding\"]\n",
    "    annotation.to_csv(\"annotation.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "79a968e4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:52:51.593967Z",
     "iopub.status.busy": "2022-09-26T14:52:51.593493Z",
     "iopub.status.idle": "2022-09-26T14:52:51.600614Z",
     "shell.execute_reply": "2022-09-26T14:52:51.600002Z"
    },
    "papermill": {
     "duration": 0.033645,
     "end_time": "2022-09-26T14:52:51.622924",
     "exception": false,
     "start_time": "2022-09-26T14:52:51.589279",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "fragments_list = sorted(glob.glob(\"../1_data_repository/full_fragments/*.tsv.gz\"))\n",
    "fragments_dict = {}\n",
    "for fragments_file in fragments_list:\n",
    "    sample = fragments_file.split(\"/\")[-1].split(\".fragments.tsv.gz\")[0]\n",
    "    fragments_dict[sample] = fragments_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9a441236",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:52:51.630894Z",
     "iopub.status.busy": "2022-09-26T14:52:51.630466Z",
     "iopub.status.idle": "2022-09-26T14:52:51.637987Z",
     "shell.execute_reply": "2022-09-26T14:52:51.637385Z"
    },
    "papermill": {
     "duration": 0.012419,
     "end_time": "2022-09-26T14:52:51.638884",
     "exception": false,
     "start_time": "2022-09-26T14:52:51.626465",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FULL': '../1_data_repository/full_fragments/BIO_ddseq_1.FULL.fragments.tsv.gz',\n",
       " 'BIO_ddseq_2.FULL': '../1_data_repository/full_fragments/BIO_ddseq_2.FULL.fragments.tsv.gz',\n",
       " 'BIO_ddseq_3.FULL': '../1_data_repository/full_fragments/BIO_ddseq_3.FULL.fragments.tsv.gz',\n",
       " 'BIO_ddseq_4.FULL': '../1_data_repository/full_fragments/BIO_ddseq_4.FULL.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_1.FULL': '../1_data_repository/full_fragments/BRO_mtscatac_1.FULL.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_2.FULL': '../1_data_repository/full_fragments/BRO_mtscatac_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_1.FULL': '../1_data_repository/full_fragments/CNA_10xmultiome_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_2.FULL': '../1_data_repository/full_fragments/CNA_10xmultiome_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_1.FULL': '../1_data_repository/full_fragments/CNA_10xv11_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_2.FULL': '../1_data_repository/full_fragments/CNA_10xv11_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_3.FULL': '../1_data_repository/full_fragments/CNA_10xv11_3.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_4.FULL': '../1_data_repository/full_fragments/CNA_10xv11_4.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_5.FULL': '../1_data_repository/full_fragments/CNA_10xv11_5.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv2_1.FULL': '../1_data_repository/full_fragments/CNA_10xv2_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv2_2.FULL': '../1_data_repository/full_fragments/CNA_10xv2_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_hydrop_1.FULL': '../1_data_repository/full_fragments/CNA_hydrop_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_hydrop_2.FULL': '../1_data_repository/full_fragments/CNA_hydrop_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_hydrop_3.FULL': '../1_data_repository/full_fragments/CNA_hydrop_3.FULL.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_1.FULL': '../1_data_repository/full_fragments/CNA_mtscatac_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_2.FULL': '../1_data_repository/full_fragments/CNA_mtscatac_2.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_1.FULL': '../1_data_repository/full_fragments/EPF_hydrop_1.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_2.FULL': '../1_data_repository/full_fragments/EPF_hydrop_2.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_3.FULL': '../1_data_repository/full_fragments/EPF_hydrop_3.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_4.FULL': '../1_data_repository/full_fragments/EPF_hydrop_4.FULL.fragments.tsv.gz',\n",
       " 'HAR_ddseq_1.FULL': '../1_data_repository/full_fragments/HAR_ddseq_1.FULL.fragments.tsv.gz',\n",
       " 'HAR_ddseq_2.FULL': '../1_data_repository/full_fragments/HAR_ddseq_2.FULL.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_1.FULL': '../1_data_repository/full_fragments/MDC_mtscatac_1.FULL.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_2.FULL': '../1_data_repository/full_fragments/MDC_mtscatac_2.FULL.fragments.tsv.gz',\n",
       " 'OHS_s3atac_1.FULL': '../1_data_repository/full_fragments/OHS_s3atac_1.FULL.fragments.tsv.gz',\n",
       " 'OHS_s3atac_2.FULL': '../1_data_repository/full_fragments/OHS_s3atac_2.FULL.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_1.FULL': '../1_data_repository/full_fragments/SAN_10xmultiome_1.FULL.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_2.FULL': '../1_data_repository/full_fragments/SAN_10xmultiome_2.FULL.fragments.tsv.gz',\n",
       " 'STA_10xv11_1.FULL': '../1_data_repository/full_fragments/STA_10xv11_1.FULL.fragments.tsv.gz',\n",
       " 'STA_10xv11_2.FULL': '../1_data_repository/full_fragments/STA_10xv11_2.FULL.fragments.tsv.gz',\n",
       " 'TXG_10xv11_1.FULL': '../1_data_repository/full_fragments/TXG_10xv11_1.FULL.fragments.tsv.gz',\n",
       " 'TXG_10xv2_1.FULL': '../1_data_repository/full_fragments/TXG_10xv2_1.FULL.fragments.tsv.gz',\n",
       " 'TXG_10xv2_2.FULL': '../1_data_repository/full_fragments/TXG_10xv2_2.FULL.fragments.tsv.gz',\n",
       " 'UCS_ddseq_1.FULL': '../1_data_repository/full_fragments/UCS_ddseq_1.FULL.fragments.tsv.gz',\n",
       " 'UCS_ddseq_2.FULL': '../1_data_repository/full_fragments/UCS_ddseq_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_1.FULL': '../1_data_repository/full_fragments/VIB_10xmultiome_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_2.FULL': '../1_data_repository/full_fragments/VIB_10xmultiome_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv1_1.FULL': '../1_data_repository/full_fragments/VIB_10xv1_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv1_2.FULL': '../1_data_repository/full_fragments/VIB_10xv1_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv2_1.FULL': '../1_data_repository/full_fragments/VIB_10xv2_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv2_2.FULL': '../1_data_repository/full_fragments/VIB_10xv2_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_1.FULL': '../1_data_repository/full_fragments/VIB_hydrop_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_11.FULL': '../1_data_repository/full_fragments/VIB_hydrop_11.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_12.FULL': '../1_data_repository/full_fragments/VIB_hydrop_12.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_2.FULL': '../1_data_repository/full_fragments/VIB_hydrop_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_21.FULL': '../1_data_repository/full_fragments/VIB_hydrop_21.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_22.FULL': '../1_data_repository/full_fragments/VIB_hydrop_22.FULL.fragments.tsv.gz'}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fragments_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f43fd7e7-21a8-4798-b7bd-c236412e02bf",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:52:51.647087Z",
     "iopub.status.busy": "2022-09-26T14:52:51.646654Z",
     "iopub.status.idle": "2022-09-26T14:52:51.655755Z",
     "shell.execute_reply": "2022-09-26T14:52:51.655043Z"
    },
    "papermill": {
     "duration": 0.01427,
     "end_time": "2022-09-26T14:52:51.656755",
     "exception": false,
     "start_time": "2022-09-26T14:52:51.642485",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FULL': 'SCREEN_peaks/BIO_ddseq_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'BIO_ddseq_2.FULL': 'SCREEN_peaks/BIO_ddseq_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'BIO_ddseq_3.FULL': 'SCREEN_peaks/BIO_ddseq_3.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'BIO_ddseq_4.FULL': 'SCREEN_peaks/BIO_ddseq_4.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'BRO_mtscatac_1.FULL': 'SCREEN_peaks/BRO_mtscatac_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'BRO_mtscatac_2.FULL': 'SCREEN_peaks/BRO_mtscatac_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xmultiome_1.FULL': 'SCREEN_peaks/CNA_10xmultiome_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xmultiome_2.FULL': 'SCREEN_peaks/CNA_10xmultiome_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_1.FULL': 'SCREEN_peaks/CNA_10xv11_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_2.FULL': 'SCREEN_peaks/CNA_10xv11_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_3.FULL': 'SCREEN_peaks/CNA_10xv11_3.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_4.FULL': 'SCREEN_peaks/CNA_10xv11_4.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_5.FULL': 'SCREEN_peaks/CNA_10xv11_5.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv2_1.FULL': 'SCREEN_peaks/CNA_10xv2_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv2_2.FULL': 'SCREEN_peaks/CNA_10xv2_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_hydrop_1.FULL': 'SCREEN_peaks/CNA_hydrop_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_hydrop_2.FULL': 'SCREEN_peaks/CNA_hydrop_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_hydrop_3.FULL': 'SCREEN_peaks/CNA_hydrop_3.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_mtscatac_1.FULL': 'SCREEN_peaks/CNA_mtscatac_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_mtscatac_2.FULL': 'SCREEN_peaks/CNA_mtscatac_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_1.FULL': 'SCREEN_peaks/EPF_hydrop_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_2.FULL': 'SCREEN_peaks/EPF_hydrop_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_3.FULL': 'SCREEN_peaks/EPF_hydrop_3.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_4.FULL': 'SCREEN_peaks/EPF_hydrop_4.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'HAR_ddseq_1.FULL': 'SCREEN_peaks/HAR_ddseq_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'HAR_ddseq_2.FULL': 'SCREEN_peaks/HAR_ddseq_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'MDC_mtscatac_1.FULL': 'SCREEN_peaks/MDC_mtscatac_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'MDC_mtscatac_2.FULL': 'SCREEN_peaks/MDC_mtscatac_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'OHS_s3atac_1.FULL': 'SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'OHS_s3atac_2.FULL': 'SCREEN_peaks/OHS_s3atac_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'SAN_10xmultiome_1.FULL': 'SCREEN_peaks/SAN_10xmultiome_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'SAN_10xmultiome_2.FULL': 'SCREEN_peaks/SAN_10xmultiome_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'STA_10xv11_1.FULL': 'SCREEN_peaks/STA_10xv11_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'STA_10xv11_2.FULL': 'SCREEN_peaks/STA_10xv11_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'TXG_10xv11_1.FULL': 'SCREEN_peaks/TXG_10xv11_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'TXG_10xv2_1.FULL': 'SCREEN_peaks/TXG_10xv2_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'TXG_10xv2_2.FULL': 'SCREEN_peaks/TXG_10xv2_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'UCS_ddseq_1.FULL': 'SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'UCS_ddseq_2.FULL': 'SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xmultiome_1.FULL': 'SCREEN_peaks/VIB_10xmultiome_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xmultiome_2.FULL': 'SCREEN_peaks/VIB_10xmultiome_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xv1_1.FULL': 'SCREEN_peaks/VIB_10xv1_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xv1_2.FULL': 'SCREEN_peaks/VIB_10xv1_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xv2_1.FULL': 'SCREEN_peaks/VIB_10xv2_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xv2_2.FULL': 'SCREEN_peaks/VIB_10xv2_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_1.FULL': 'SCREEN_peaks/VIB_hydrop_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_11.FULL': 'SCREEN_peaks/VIB_hydrop_11.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_12.FULL': 'SCREEN_peaks/VIB_hydrop_12.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_2.FULL': 'SCREEN_peaks/VIB_hydrop_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_21.FULL': 'SCREEN_peaks/VIB_hydrop_21.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_22.FULL': 'SCREEN_peaks/VIB_hydrop_22.FULL__SCREEN_consensus_peaks.bed'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions_paths_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"SCREEN_peaks/*consensus_peaks.bed\"))\n",
    "}\n",
    "regions_paths_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a58786aa",
   "metadata": {
    "papermill": {
     "duration": 0.003612,
     "end_time": "2022-09-26T14:52:51.664053",
     "exception": false,
     "start_time": "2022-09-26T14:52:51.660441",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Now, make a sub dictionary of all samples within the fragments dict that have not been run yet (good for resuming a stopped cistopic run):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "13ad51aa",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:52:51.672005Z",
     "iopub.status.busy": "2022-09-26T14:52:51.671831Z",
     "iopub.status.idle": "2022-09-26T14:52:51.676458Z",
     "shell.execute_reply": "2022-09-26T14:52:51.675843Z"
    },
    "papermill": {
     "duration": 0.00969,
     "end_time": "2022-09-26T14:52:51.677355",
     "exception": false,
     "start_time": "2022-09-26T14:52:51.667665",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "cistopic_qc_out = os.path.join(wdir, \"cistopic_qc_out_CONSENSUS\")\n",
    "if not os.path.exists(cistopic_qc_out):\n",
    "    os.makedirs(cistopic_qc_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "161f54e5-1e0b-4277-b9ca-f0247df0e4e9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:52:51.685648Z",
     "iopub.status.busy": "2022-09-26T14:52:51.685214Z",
     "iopub.status.idle": "2022-09-26T14:52:51.703492Z",
     "shell.execute_reply": "2022-09-26T14:52:51.702857Z"
    },
    "papermill": {
     "duration": 0.023448,
     "end_time": "2022-09-26T14:52:51.704481",
     "exception": false,
     "start_time": "2022-09-26T14:52:51.681033",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/BIO_ddseq_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/BIO_ddseq_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/BIO_ddseq_3.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/BIO_ddseq_4.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/BRO_mtscatac_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/BRO_mtscatac_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xmultiome_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xmultiome_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xv11_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xv11_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xv11_3.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xv11_4.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xv11_5.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xv2_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xv2_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_hydrop_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_hydrop_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_hydrop_3.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_mtscatac_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_mtscatac_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/EPF_hydrop_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/EPF_hydrop_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/EPF_hydrop_3.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/EPF_hydrop_4.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/HAR_ddseq_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/HAR_ddseq_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/MDC_mtscatac_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/MDC_mtscatac_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/OHS_s3atac_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata does not exist, adding to subdict to generate\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/OHS_s3atac_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata does not exist, adding to subdict to generate\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/SAN_10xmultiome_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/SAN_10xmultiome_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/STA_10xv11_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/STA_10xv11_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/TXG_10xv11_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/TXG_10xv2_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/TXG_10xv2_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/UCS_ddseq_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata does not exist, adding to subdict to generate\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/UCS_ddseq_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata does not exist, adding to subdict to generate\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_10xmultiome_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_10xmultiome_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_10xv1_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_10xv1_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_10xv2_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_10xv2_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_hydrop_1.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_hydrop_11.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_hydrop_12.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_hydrop_2.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_hydrop_21.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_hydrop_22.FULL__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n"
     ]
    }
   ],
   "source": [
    "fragments_sub_dict = {}\n",
    "regions_sub_dict = {}\n",
    "for sample in regions_paths_dict.keys():\n",
    "    metadata_file = os.path.join(cistopic_qc_out, sample + \"__metadata_bc.pkl\")\n",
    "    print(f\"Checking if {metadata_file} exist...\")\n",
    "    if os.path.exists(metadata_file):\n",
    "        print(\"\\tMetadata exists! Skipping...\")\n",
    "    else:\n",
    "        fragments_sub_dict[sample] = fragments_dict[sample]\n",
    "        print(\"\\tMetadata does not exist, adding to subdict to generate\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dcdbc9cb-050a-43ee-8160-bc799b872883",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:52:51.714836Z",
     "iopub.status.busy": "2022-09-26T14:52:51.714385Z",
     "iopub.status.idle": "2022-09-26T14:52:51.718702Z",
     "shell.execute_reply": "2022-09-26T14:52:51.718013Z"
    },
    "papermill": {
     "duration": 0.009693,
     "end_time": "2022-09-26T14:52:51.719686",
     "exception": false,
     "start_time": "2022-09-26T14:52:51.709993",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "regions_sub_dict = {x: regions_paths_dict[x] for x in sorted(fragments_sub_dict.keys())}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "92ba8eb2-3d7d-4d66-9eed-a92a54433d2e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:52:51.728464Z",
     "iopub.status.busy": "2022-09-26T14:52:51.728013Z",
     "iopub.status.idle": "2022-09-26T14:52:51.731219Z",
     "shell.execute_reply": "2022-09-26T14:52:51.730762Z"
    },
    "papermill": {
     "duration": 0.00851,
     "end_time": "2022-09-26T14:52:51.732079",
     "exception": false,
     "start_time": "2022-09-26T14:52:51.723569",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "ray.shutdown()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "0a585a26",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:52:51.740775Z",
     "iopub.status.busy": "2022-09-26T14:52:51.740324Z",
     "iopub.status.idle": "2022-09-26T15:28:46.883955Z",
     "shell.execute_reply": "2022-09-26T15:28:46.883368Z"
    },
    "papermill": {
     "duration": 2155.22384,
     "end_time": "2022-09-26T15:28:46.959788",
     "exception": false,
     "start_time": "2022-09-26T14:52:51.735948",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-09-26 14:52:51,747 cisTopic     INFO     n_cpu is larger than the number of samples. Setting n_cpu to the number of samples\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-26 14:52:55,943\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 14:53:09,492 cisTopic     INFO     Reading OHS_s3atac_1.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 14:53:09,562 cisTopic     INFO     Reading UCS_ddseq_2.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 14:53:09,532 cisTopic     INFO     Reading UCS_ddseq_1.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 14:53:09,526 cisTopic     INFO     Reading OHS_s3atac_2.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:01:20,682 cisTopic     INFO     Computing barcode rank plot for UCS_ddseq_2.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:01:20,682 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:02:06,106 cisTopic     INFO     Marking barcodes with more than 10\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:02:08,156 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:02:08,194 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:02:28,012 cisTopic     INFO     Computing duplicate rate plot for UCS_ddseq_2.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:02:40,721 cisTopic     INFO     Computing barcode rank plot for OHS_s3atac_1.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:02:40,721 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:03:00,103 cisTopic     INFO     Marking barcodes with more than 10\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:03:00,141 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:03:00,143 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:03:24,858 cisTopic     INFO     Computing duplicate rate plot for OHS_s3atac_1.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:03:54,117 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:03:54,406 cisTopic     INFO     Computing insert size distribution for OHS_s3atac_1.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:03:54,406 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:04:00,837 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:04:01,263 cisTopic     INFO     Computing insert size distribution for UCS_ddseq_2.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:04:01,264 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:04:07,893 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:04:08,104 cisTopic     INFO     Computing barcode rank plot for UCS_ddseq_1.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:04:08,104 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:04:08,196 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:05:16,060 cisTopic     INFO     Marking barcodes with more than 10\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:05:18,691 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:05:18,732 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:05:40,394 cisTopic     INFO     Computing duplicate rate plot for UCS_ddseq_1.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:06:27,668 cisTopic     INFO     Computing TSS profile for UCS_ddseq_2.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:07:24,585 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:07:24,830 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:07:32,500 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:07:33,000 cisTopic     INFO     Computing insert size distribution for UCS_ddseq_1.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:07:33,000 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:07:41,602 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:08:32,924 cisTopic     INFO     Computing TSS profile for OHS_s3atac_1.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:09:12,522 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:09:12,799 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:10:33,119 cisTopic     INFO     Computing TSS profile for UCS_ddseq_1.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:12:00,611 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:12:00,843 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:13:25,398 cisTopic     INFO     Computing barcode rank plot for OHS_s3atac_2.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:13:25,399 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:13:35,240 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:14:08,709 cisTopic     INFO     Marking barcodes with more than 10\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:14:08,730 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:14:08,732 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:14:46,138 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:14:55,048 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:15:03,759 cisTopic     INFO     Computing duplicate rate plot for OHS_s3atac_2.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:15:48,818 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:15:49,124 cisTopic     INFO     Computing FRIP profile for OHS_s3atac_1.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:15:51,844 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:16:07,511 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:16:07,831 cisTopic     INFO     Computing insert size distribution for OHS_s3atac_2.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:16:07,832 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:16:29,799 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:16:36,929 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449455)\u001b[0m 2022-09-26 15:17:43,627 cisTopic     INFO     Sample OHS_s3atac_1.FULL done!\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:18:04,704 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:19:47,690 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:19:47,940 cisTopic     INFO     Computing FRIP profile for UCS_ddseq_2.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:19:49,723 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:20:20,896 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:20:45,641 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449453)\u001b[0m 2022-09-26 15:21:32,417 cisTopic     INFO     Sample UCS_ddseq_2.FULL done!\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:23:09,668 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:23:26,227 cisTopic     INFO     Computing TSS profile for OHS_s3atac_2.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:24:02,192 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:24:02,304 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:24:31,907 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:24:32,075 cisTopic     INFO     Computing FRIP profile for UCS_ddseq_1.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:24:33,092 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:24:58,010 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449454)\u001b[0m 2022-09-26 15:25:53,038 cisTopic     INFO     Sample UCS_ddseq_1.FULL done!\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:26:13,538 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:26:27,359 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:26:37,204 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:26:37,279 cisTopic     INFO     Computing FRIP profile for OHS_s3atac_2.FULL\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:26:38,625 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:26:59,384 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=2449452)\u001b[0m 2022-09-26 15:28:17,342 cisTopic     INFO     Sample OHS_s3atac_2.FULL done!\n",
      "Dumping files in /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_3_cistopic_consensus/cistopic_qc_out_CONSENSUS...\n"
     ]
    }
   ],
   "source": [
    "n_cores = 16\n",
    "if regions_sub_dict != {}:\n",
    "    samples_sub = list(regions_sub_dict.keys())\n",
    "    blocks = [samples_sub[i : i + n_cores] for i in range(0, len(samples_sub), n_cores)]\n",
    "    for samples_torun_in_block in blocks:\n",
    "        fragments_sub_dict_block = {\n",
    "            key: fragments_sub_dict[key] for key in samples_torun_in_block\n",
    "        }\n",
    "        regions_sub_dict_block = {\n",
    "            key: regions_sub_dict[key] for key in samples_torun_in_block\n",
    "        }\n",
    "\n",
    "        metadata_bc_dict, profile_data_dict = compute_qc_stats(\n",
    "            fragments_dict=fragments_sub_dict_block,\n",
    "            tss_annotation=annotation,\n",
    "            stats=[\n",
    "                \"barcode_rank_plot\",\n",
    "                \"duplicate_rate\",\n",
    "                \"insert_size_distribution\",\n",
    "                \"profile_tss\",\n",
    "                \"frip\",\n",
    "            ],\n",
    "            label_list=None,\n",
    "            path_to_regions=regions_sub_dict_block,\n",
    "            n_cpu=n_cores,\n",
    "            valid_bc=None,\n",
    "            n_frag=10,\n",
    "            n_bc=None,\n",
    "            tss_flank_window=2000,\n",
    "            tss_window=50,\n",
    "            tss_minimum_signal_window=100,\n",
    "            tss_rolling_window=10,\n",
    "            # min_norm=0.2,\n",
    "            remove_duplicates=True,\n",
    "        )\n",
    "\n",
    "        ray.shutdown()\n",
    "        print(f\"Dumping files in {cistopic_qc_out}...\")\n",
    "        for sample in sorted(metadata_bc_dict.keys()):\n",
    "            metadata_bc_dict[sample][\"sample_id\"] = sample\n",
    "            metadata_bc_dict[sample].index = [\n",
    "                x + \"___\" + sample for x in list(metadata_bc_dict[sample].index)\n",
    "            ]\n",
    "            with open(\n",
    "                os.path.join(cistopic_qc_out, f\"{sample}__metadata_bc.pkl\"), \"wb\"\n",
    "            ) as f:\n",
    "                pickle.dump(metadata_bc_dict[sample], f, protocol=4)\n",
    "\n",
    "            with open(\n",
    "                os.path.join(cistopic_qc_out, f\"{sample}__profile_data.pkl\"), \"wb\"\n",
    "            ) as f:\n",
    "                pickle.dump(profile_data_dict[sample], f, protocol=4)\n",
    "else:\n",
    "    print(\"All samples already processed.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a1ad8e2",
   "metadata": {
    "papermill": {
     "duration": 0.007253,
     "end_time": "2022-09-26T15:28:47.036498",
     "exception": false,
     "start_time": "2022-09-26T15:28:47.029245",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76d9a6eb-c74e-41bb-b05b-0c2c36a0a2f9",
   "metadata": {
    "papermill": {
     "duration": 0.006475,
     "end_time": "2022-09-26T15:28:47.049698",
     "exception": false,
     "start_time": "2022-09-26T15:28:47.043223",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Calculating a KDE is simultaneously expensive and scales poorly with increasing n. Therefore, I wrote a multithreaded script that divides the QC array into equal parts (interleaved to avoid biases in the order!) and performs a KDE calculation on each part. Here, Otsu thresholding is used to find the right threshold for minimum fragments and minimum TSS enrichment. ddseq samples have a significantly higher noise floor than the other samples when it comes to fragment distribution. Therefore, the otsu algorithm is performed on all barcodes with a minimum of 300 fragments for the ddseq samples, and a minimum of 100 fragments for all the other samples. I tried to perform this filtering completely independent of sample/technique (e.g. using gaussian mixture modeling, Jenks natural breaks, or multiple step Otsu thresholding) but found that no solution worked perfectly for all samples."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0332bf6a-0bd8-41d2-9f12-634b2249af11",
   "metadata": {
    "papermill": {
     "duration": 0.006452,
     "end_time": "2022-09-26T15:28:47.062737",
     "exception": false,
     "start_time": "2022-09-26T15:28:47.056285",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "This is regulated by the code below in qc_plots.py:\n",
    "```\n",
    "min_otsu_frags_dict = {}\n",
    "for fragments_file in fragments_list:\n",
    "    sample = fragments_file.split(\"/\")[-1].split(\".\")[0]\n",
    "    tech = sample.split('_')[1]\n",
    "    if tech == \"ddseq\":\n",
    "        if sample == \"BIO_ddseq_1\":\n",
    "            min_otsu_frags_dict[sample] = 600\n",
    "        else:\n",
    "            min_otsu_frags_dict[sample] = 300\n",
    "    elif tech == \"hydrop\":\n",
    "        min_otsu_frags_dict[sample] = 300\n",
    "    else:\n",
    "        min_otsu_frags_dict[sample] = 100\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0b8bcb35-3db0-4d9c-96d0-c73d04a0d5cf",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T15:28:47.077489Z",
     "iopub.status.busy": "2022-09-26T15:28:47.076684Z",
     "iopub.status.idle": "2022-09-26T15:28:47.527302Z",
     "shell.execute_reply": "2022-09-26T15:28:47.526750Z"
    },
    "papermill": {
     "duration": 0.459421,
     "end_time": "2022-09-26T15:28:47.528659",
     "exception": false,
     "start_time": "2022-09-26T15:28:47.069238",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "import kde\n",
      "import pycisTopic\n",
      "import glob\n",
      "import os\n",
      "import pandas as pd\n",
      "import pickle\n",
      "import matplotlib.pyplot as plt\n",
      "import numpy as np\n",
      "import seaborn as sns\n",
      "import multiprocess as mp\n",
      "from multiprocess import Pool\n",
      "import pprint as pp\n",
      "\n",
      "def histogram(array, nbins=100):\n",
      "    \"\"\"\n",
      "    Draw histogram from distribution and identify centers.\n",
      "    Parameters\n",
      "    ---------\n",
      "    array: `class::np.array`\n",
      "            Scores distribution\n",
      "    nbins: int\n",
      "            Number of bins to use in the histogram\n",
      "    Return\n",
      "    ---------\n",
      "    float\n",
      "            Histogram values and bin centers.\n",
      "    \"\"\"\n",
      "    array = array.ravel().flatten()\n",
      "    hist, bin_edges = np.histogram(array, bins=nbins, range=None)\n",
      "    bin_centers = (bin_edges[:-1] + bin_edges[1:]) / 2.0\n",
      "    return hist, bin_centers\n",
      "\n",
      "\n",
      "def threshold_otsu(array, nbins=100, min_value=100):\n",
      "    \"\"\"\n",
      "    Apply Otsu threshold on topic-region distributions [Otsu, 1979].\n",
      "    Parameters\n",
      "    ---------\n",
      "    array: `class::np.array`\n",
      "            Array containing the region values for the topic to be binarized.\n",
      "    nbins: int\n",
      "            Number of bins to use in the binarization histogram\n",
      "    Return\n",
      "    ---------\n",
      "    float\n",
      "            Binarization threshold.\n",
      "    Reference\n",
      "    ---------\n",
      "    Otsu, N., 1979. A threshold selection method from gray-level histograms. IEEE transactions on systems, man, and\n",
      "    cybernetics, 9(1), pp.62-66.\n",
      "    \"\"\"\n",
      "    array = array[(array >= min_value)]\n",
      "    hist, bin_centers = histogram(array, nbins)\n",
      "    hist = hist.astype(float)\n",
      "    # Class probabilities for all possible thresholds\n",
      "    weight1 = np.cumsum(hist)\n",
      "    weight2 = np.cumsum(hist[::-1])[::-1]\n",
      "    # Class means for all possible thresholds\n",
      "    mean1 = np.cumsum(hist * bin_centers) / weight1\n",
      "    mean2 = (np.cumsum((hist * bin_centers)[::-1]) / weight2[::-1])[::-1]\n",
      "    # Clip ends to align class 1 and class 2 variables:\n",
      "    # The last value of ``weight1``/``mean1`` should pair with zero values in\n",
      "    # ``weight2``/``mean2``, which do not exist.\n",
      "    variance12 = weight1[:-1] * weight2[1:] * (mean1[:-1] - mean2[1:]) ** 2\n",
      "    idx = np.argmax(variance12)\n",
      "    threshold = bin_centers[:-1][idx]\n",
      "    return threshold\n",
      "\n",
      "\n",
      "# def calc_kde(xy):\n",
      "#     return gaussian_kde(xy)(xy)\n",
      "\n",
      "\n",
      "def plot_frag_qc(\n",
      "    x,\n",
      "    y,\n",
      "    ax,\n",
      "    x_thr_min=None,\n",
      "    x_thr_max=None,\n",
      "    y_thr_min=None,\n",
      "    y_thr_max=None,\n",
      "    ylab=None,\n",
      "    xlab=\"Number of (unique) fragments\",\n",
      "    cmap=\"viridis\",\n",
      "    density_overlay=False,\n",
      "    s=10,\n",
      "    marker=\"+\",\n",
      "    c=\"#343434\",\n",
      "    xlim=None,\n",
      "    ylim=None,\n",
      "    **kwargs\n",
      "):\n",
      "    from scipy.stats import gaussian_kde\n",
      "\n",
      "    assert all(x.index == y.index)\n",
      "    barcodes = x.index.values\n",
      "\n",
      "    if density_overlay:\n",
      "        cores = 8\n",
      "\n",
      "        x_log = np.log(x)\n",
      "\n",
      "        # Split input array for KDE [log(x), y] array in\n",
      "        # equaly spaced parts (start_offset + n * nbr_cores).\n",
      "        kde_parts = [np.vstack([x_log[i::cores], y[i::cores]]) for i in range(cores)]\n",
      "\n",
      "        # Get nultiprocess context object to spawn processes.\n",
      "        # mp_ctx = mp.get_context(\"spawn\")\n",
      "\n",
      "        # Calculate KDE in parallel.\n",
      "        with Pool(processes=cores) as pool:\n",
      "            results = pool.map(kde.calc_kde, kde_parts)\n",
      "\n",
      "        z = np.concatenate(results)\n",
      "\n",
      "        # now order x and y in the same way that z was ordered, otherwise random z value is assigned to barcode:\n",
      "        x_ordered = np.concatenate([x[i::cores] for i in range(cores)])\n",
      "        y_ordered = np.concatenate([y[i::cores] for i in range(cores)])\n",
      "\n",
      "        idx = (\n",
      "            z.argsort()\n",
      "        )  # order based on z value so that highest value is plotted on top, and not hidden by lower values\n",
      "        x, y, z, barcodes = x_ordered[idx], y_ordered[idx], z[idx], barcodes[idx]\n",
      "    else:\n",
      "        z = c\n",
      "\n",
      "    sp = ax.scatter(x, y, c=z, s=s, edgecolors=None, marker=marker, cmap=cmap, **kwargs)\n",
      "    if ylim is not None:\n",
      "        ax.set_ylim(ylim[0], ylim[1])\n",
      "    if xlim is not None:\n",
      "        ax.set_xlim(xlim[0], xlim[1])\n",
      "\n",
      "    # Start with keeping all barcodes.\n",
      "    barcodes_to_keep = np.full(x.shape[0], True)\n",
      "\n",
      "    # Filter barcodes out if needed based on thresholds:\n",
      "    if x_thr_min is not None:\n",
      "        ax.axvline(x=x_thr_min, color=\"r\", linestyle=\"--\")\n",
      "\n",
      "    if x_thr_max is not None:\n",
      "        ax.axvline(x=x_thr_max, color=\"r\", linestyle=\"--\")\n",
      "\n",
      "    if y_thr_min is not None:\n",
      "        ax.axhline(y=y_thr_min, color=\"r\", linestyle=\"--\")\n",
      "\n",
      "    if y_thr_max is not None:\n",
      "        ax.axhline(y=y_thr_max, color=\"r\", linestyle=\"--\")\n",
      "\n",
      "    ax.set_xscale(\"log\")\n",
      "    ax.set_xmargin(0.01)\n",
      "    ax.set_ymargin(0.01)\n",
      "    ax.set_xlabel(xlab, fontsize=10)\n",
      "    ax.set_ylabel(ylab, fontsize=10)\n",
      "\n",
      "def plot_qc(\n",
      "    sample,\n",
      "    min_dict,\n",
      "    max_dict,\n",
      "    metadata_bc_df,\n",
      "    include_kde=False,\n",
      "    detailed_title=True,\n",
      "    s=4,\n",
      "    min_x_val=100,\n",
      "    min_y_val=1\n",
      "):\n",
      "    fig, (ax1, ax2, ax3) = plt.subplots(1, 3, figsize=(12, 4), dpi=150)\n",
      "\n",
      "    # calculate thresholds using a double otsu strategy:\n",
      "    # first we calculate an otsu threshold using a minimum of 10 fragments\n",
      "    # then, this otsu threshold is used as the minimum for the second iteration\n",
      "\n",
      "    x_arr = np.log10(metadata_bc_df[\"Unique_nr_frag\"])\n",
      "    x_threshold_log = threshold_otsu(x_arr, nbins=5000, min_value=np.log10(min_x_val))\n",
      "    x_threshold = 10**x_threshold_log\n",
      "\n",
      "    y_arr = metadata_bc_df[\"TSS_enrichment\"]\n",
      "    y_threshold = threshold_otsu(y_arr, nbins=5000, min_value=min_y_val)\n",
      "\n",
      "    # calculate cells passing filter\n",
      "    metadata_bc_df_passing_filters = metadata_bc_df.loc[\n",
      "        (metadata_bc_df.Unique_nr_frag > x_threshold)\n",
      "        & (metadata_bc_df.TSS_enrichment > y_threshold)\n",
      "    ]\n",
      "    bc_passing_filters = metadata_bc_df_passing_filters.index\n",
      "\n",
      "    # plot everything\n",
      "    plot_frag_qc(\n",
      "        x=metadata_bc_df[\"Unique_nr_frag\"],\n",
      "        y=metadata_bc_df[\"TSS_enrichment\"],\n",
      "        ylab=\"TSS Enrichment\",\n",
      "        s=s,\n",
      "        x_thr_min=x_threshold,\n",
      "        y_thr_min=y_threshold,\n",
      "        xlim=[10, max_dict[\"Unique_nr_frag\"]],\n",
      "        ylim=[0, max_dict[\"TSS_enrichment\"]],\n",
      "        density_overlay=include_kde,\n",
      "        ax=ax1,\n",
      "    )\n",
      "    plot_frag_qc(\n",
      "        x=metadata_bc_df[\"Unique_nr_frag\"],\n",
      "        y=metadata_bc_df[\"FRIP\"],\n",
      "        x_thr_min=x_threshold,\n",
      "        ylab=\"FRIP\",\n",
      "        s=s,\n",
      "        xlim=[10, max_dict[\"Unique_nr_frag\"]],\n",
      "        ylim=[0, 1],\n",
      "        density_overlay=include_kde,\n",
      "        ax=ax2,\n",
      "    )\n",
      "    plot_frag_qc(\n",
      "        x=metadata_bc_df[\"Unique_nr_frag\"],\n",
      "        y=metadata_bc_df[\"Dupl_rate\"],\n",
      "        x_thr_min=x_threshold,\n",
      "        ylab=\"Duplicate rate per cell\",\n",
      "        s=s,\n",
      "        xlim=[10, max_dict[\"Unique_nr_frag\"]],\n",
      "        ylim=[0, 1],\n",
      "        density_overlay=include_kde,\n",
      "        ax=ax3,\n",
      "    )\n",
      "\n",
      "    if detailed_title:\n",
      "        med_nf = round(\n",
      "            metadata_bc_df.loc[bc_passing_filters, \"Unique_nr_frag\"].median(), 2\n",
      "        )\n",
      "        med_tss = round(\n",
      "            metadata_bc_df.loc[bc_passing_filters, \"TSS_enrichment\"].median(), 2\n",
      "        )\n",
      "        med_frip = round(metadata_bc_df.loc[bc_passing_filters, \"FRIP\"].median(), 2)\n",
      "        title = f\"{sample}: Kept {len(bc_passing_filters)} cells using Otsu filtering. Median Unique Fragments: {med_nf:.0f}. Median TSS Enrichment: {med_tss:.2f}. Median FRIP: {med_frip:.2f}\\nUsed a minimum of {x_threshold:.2f} fragments and TSS enrichment of {y_threshold:.2f})\"\n",
      "    else:\n",
      "        title = sample\n",
      "\n",
      "    fig.suptitle(title, x=0.5, y=0.95, fontsize=10)\n",
      "    return bc_passing_filters, fig\n",
      "\n",
      "\n",
      "###################\n",
      "# Run the loop here\n",
      "###################\n",
      "\n",
      "metadata_bc_pkl_list = sorted(glob.glob(\"cistopic_qc_out/*metadata_bc.pkl\"))\n",
      "metadata_bc_pkl_path_dict = {}\n",
      "for metadata_bc_pkl_path in metadata_bc_pkl_list:\n",
      "    sample = metadata_bc_pkl_path.split(\"/\")[-1].split(\"__metadata_bc.pkl\")[0]\n",
      "    metadata_bc_pkl_path_dict[sample] = metadata_bc_pkl_path\n",
      "\n",
      "# calculate global maxima to make nice equal plots:\n",
      "metadata_bc_df_all = pd.DataFrame()\n",
      "for sample in metadata_bc_pkl_path_dict.keys():\n",
      "    with open(metadata_bc_pkl_path_dict[sample], \"rb\") as fh:\n",
      "        metadata_bc_df = pickle.load(fh)\n",
      "        metadata_bc_df_all = pd.concat([metadata_bc_df_all, metadata_bc_df])\n",
      "        print(f\"added {sample} metadata_bc_df\")\n",
      "\n",
      "max_dict = {}\n",
      "min_dict = {}\n",
      "for stat in metadata_bc_df_all.columns:\n",
      "    max_dict[stat] = metadata_bc_df_all[stat].max()\n",
      "    min_dict[stat] = metadata_bc_df_all[stat].min()\n",
      "\n",
      "pp.pprint(min_dict)\n",
      "pp.pprint(max_dict)\n",
      "\n",
      "standard_min_x_val = 100\n",
      "standard_min_y_val = 1\n",
      "min_otsu_frags_dict = {}\n",
      "min_otsu_tss_dict = {}\n",
      "for metadata_bc_pkl_path in sorted(metadata_bc_pkl_path_dict.keys()):\n",
      "    sample = metadata_bc_pkl_path.split(\"/\")[-1].split(\"__metadata_bc.pkl\")[0]\n",
      "    tech = sample.split('_')[1]\n",
      "    if tech == \"ddseq\":\n",
      "        if \"BIO_ddseq_1.FULL\" in sample:\n",
      "            min_otsu_frags_dict[sample] = 600\n",
      "        else:\n",
      "            min_otsu_frags_dict[sample] = 300\n",
      "    elif tech == \"hydrop\":\n",
      "        min_otsu_frags_dict[sample] = 300\n",
      "    else:\n",
      "        min_otsu_frags_dict[sample] = standard_min_x_val\n",
      "\n",
      "    if tech == \"s3atac\":\n",
      "        min_otsu_tss_dict[sample] = 0\n",
      "    else:\n",
      "        min_otsu_tss_dict[sample] = standard_min_y_val\n",
      "\n",
      "pp.pprint(min_otsu_frags_dict)\n",
      "\n",
      "for sample in metadata_bc_pkl_path_dict.keys():\n",
      "    pkl_path = f\"selected_barcodes/{sample}_bc_passing_filters_otsu.pkl\"\n",
      "    if os.path.exists(pkl_path):\n",
      "        print(f\"{pkl_path} bc passing filters exists, skipping...\")\n",
      "    else:\n",
      "        print(f\"{pkl_path} bc passing filters does not exist yet, generating...\")\n",
      "        print(f\"\\tLoading {metadata_bc_pkl_path_dict[sample]}\")\n",
      "        with open(metadata_bc_pkl_path_dict[sample], \"rb\") as fh:\n",
      "            metadata_bc_df = pickle.load(fh)\n",
      "\n",
      "        print(f\"\\tFiltering cells and generating QC plots.\")\n",
      "        if not sample in min_otsu_frags_dict.keys():\n",
      "            print(f\"\\t{sample} not in minimum dict! Using standard value of {standard_min_x_val}\")\n",
      "            min_x_val = standard_min_x_val\n",
      "            min_y_val = standard_min_y_val\n",
      "        else:\n",
      "            min_x_val = min_otsu_frags_dict[sample]\n",
      "            min_y_val = min_otsu_tss_dict[sample]\n",
      "\n",
      "        bc_passing_filters, fig = plot_qc(sample=sample, min_dict=min_dict, max_dict=max_dict, metadata_bc_df=metadata_bc_df, include_kde=kde, min_x_val=min_x_val, min_y_val=min_y_val)\n",
      "\n",
      "        plt.tight_layout()\n",
      "        plt.savefig(f\"plots_qc/{sample}_qc_otsu.png\", dpi=300, facecolor=\"white\")\n",
      "        plt.show()\n",
      "        plt.close()\n",
      "\n",
      "        print(f\"\\tSaving...\")\n",
      "        with open(\n",
      "            f\"selected_barcodes/{sample}_bc_passing_filters_otsu.pkl\", \"wb\"\n",
      "        ) as fh:\n",
      "            pickle.dump(bc_passing_filters, fh)\n",
      "        fh.close()\n",
      "\n",
      "        fh = open(f\"selected_barcodes/{sample}_bc_passing_filters_otsu.txt\", \"w\")\n",
      "        for bc in list(bc_passing_filters):\n",
      "            fh.write(bc + \"\\n\")\n",
      "        fh.close()\n",
      "\n",
      "        metadata_bc_df.loc[bc_passing_filters].to_csv(f'selected_barcodes/{sample}_metadata_bc_df.tsv', sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "!cat ../0_resources/scripts/qc_plots.py"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4019bde",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-07-15T09:42:06.802579Z",
     "iopub.status.busy": "2022-07-15T09:42:06.801860Z",
     "iopub.status.idle": "2022-07-15T09:42:06.814527Z",
     "shell.execute_reply": "2022-07-15T09:42:06.813708Z",
     "shell.execute_reply.started": "2022-07-15T09:42:06.802476Z"
    },
    "papermill": {
     "duration": 0.006696,
     "end_time": "2022-09-26T15:28:47.542939",
     "exception": false,
     "start_time": "2022-09-26T15:28:47.536243",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Since multiprocessing does not work with jupyter notebooks, run the following code in terminal:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d53e61b",
   "metadata": {
    "papermill": {
     "duration": 0.006634,
     "end_time": "2022-09-26T15:28:47.556252",
     "exception": false,
     "start_time": "2022-09-26T15:28:47.549618",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "```\n",
    "mkdir plots_qc\n",
    "mkdir selected_barcodes\n",
    "SIF=../0_resources/cistopic_image/20220722_pycistopic.sif\n",
    "singularity exec \\\n",
    "    --cleanenv \\\n",
    "    -H $PWD:/home \\\n",
    "    $SIF \\\n",
    "    python ../0_resources/scripts/qc_plots.py\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6de4c150-e92a-4060-b992-ae7babd1625a",
   "metadata": {
    "papermill": {
     "duration": 0.006716,
     "end_time": "2022-09-26T15:28:47.569698",
     "exception": false,
     "start_time": "2022-09-26T15:28:47.562982",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "And then open the plots:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "68739afd-bc2c-4fab-ad96-25edf95b163c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T15:28:47.584459Z",
     "iopub.status.busy": "2022-09-26T15:28:47.583956Z",
     "iopub.status.idle": "2022-09-26T15:28:47.592368Z",
     "shell.execute_reply": "2022-09-26T15:28:47.591928Z"
    },
    "papermill": {
     "duration": 0.016706,
     "end_time": "2022-09-26T15:28:47.593273",
     "exception": false,
     "start_time": "2022-09-26T15:28:47.576567",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "metadata_bc_pkl_list = sorted(glob.glob(\"cistopic_qc_out/*metadata_bc.pkl\"))\n",
    "metadata_bc_pkl_path_dict = {}\n",
    "for metadata_bc_pkl_path in metadata_bc_pkl_list:\n",
    "    sample = metadata_bc_pkl_path.split(\"/\")[-1].split(\"__\")[0]\n",
    "    metadata_bc_pkl_path_dict[sample] = metadata_bc_pkl_path\n",
    "\n",
    "for sample in metadata_bc_pkl_path_dict.keys():\n",
    "    if os.path.exists(f\"selected_barcodes/{sample}_bc_passing_filters_otsu.pkl\"):\n",
    "        print(f\"{sample} bc passing filters exists, printing img and skipping\")\n",
    "        display(Image(f\"plots_qc/{sample}_qc_otsu.png\"))\n",
    "    else:\n",
    "        print(\n",
    "            f\"{sample} bc passing filters does not exist yet, generate using qc_plots.py script!\"\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa5a05ab-0ae5-4804-95d8-b280b86bbedf",
   "metadata": {
    "papermill": {
     "duration": 0.006619,
     "end_time": "2022-09-26T15:28:47.606748",
     "exception": false,
     "start_time": "2022-09-26T15:28:47.600129",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Write raw barcode file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed437d13-e63c-4fd5-a19b-549c83c3c3f6",
   "metadata": {
    "papermill": {
     "duration": 0.006628,
     "end_time": "2022-09-26T15:28:47.619999",
     "exception": false,
     "start_time": "2022-09-26T15:28:47.613371",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Freemuxlet reads barcodes in the bam file. These barcodes (tag: DB) do not contain the `sample` suffix, which is included in the `*_bc_passing_filters_otsu.txt` files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "df7588f2-5917-4e37-98cb-7fd0d2c83590",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T15:28:47.634042Z",
     "iopub.status.busy": "2022-09-26T15:28:47.633876Z",
     "iopub.status.idle": "2022-09-26T15:28:48.786654Z",
     "shell.execute_reply": "2022-09-26T15:28:48.786160Z"
    },
    "papermill": {
     "duration": 1.160882,
     "end_time": "2022-09-26T15:28:48.787569",
     "exception": false,
     "start_time": "2022-09-26T15:28:47.626687",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "selected_barcodes/CNA_10xv2_1.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_10xv2_1.FULL\n",
      "Done\n",
      "selected_barcodes/VIB_10xv2_2.FULL_bc_passing_filters_otsu.txt\n",
      "VIB_10xv2_2.FULL\n",
      "Done\n",
      "selected_barcodes/VIB_10xv1_2.FULL_bc_passing_filters_otsu.txt\n",
      "VIB_10xv1_2.FULL\n",
      "Done\n",
      "selected_barcodes/VIB_hydrop_11.FULL_bc_passing_filters_otsu.txt\n",
      "VIB_hydrop_11.FULL\n",
      "Done\n",
      "selected_barcodes/CNA_10xmultiome_2.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_10xmultiome_2.FULL\n",
      "Done\n",
      "selected_barcodes/VIB_mtscatac_1.FULL_bc_passing_filters_otsu.txt\n",
      "VIB_mtscatac_1.FULL\n",
      "Done\n",
      "selected_barcodes/CNA_hydrop_3.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_hydrop_3.FULL\n",
      "Done\n",
      "selected_barcodes/SAN_10xmultiome_1.FULL_bc_passing_filters_otsu.txt\n",
      "SAN_10xmultiome_1.FULL\n",
      "Done\n",
      "selected_barcodes/BIO_ddseq_4.FULL_bc_passing_filters_otsu.txt\n",
      "BIO_ddseq_4.FULL\n",
      "Done\n",
      "selected_barcodes/BRO_mtscatac_2.FULL_bc_passing_filters_otsu.txt\n",
      "BRO_mtscatac_2.FULL\n",
      "Done\n",
      "selected_barcodes/STA_10xv11_2.FULL_bc_passing_filters_otsu.txt\n",
      "STA_10xv11_2.FULL\n",
      "Done\n",
      "selected_barcodes/CNA_ddseq_1.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_ddseq_1.FULL\n",
      "Done\n",
      "selected_barcodes/VIB_10xmultiome_1.FULL_bc_passing_filters_otsu.txt\n",
      "VIB_10xmultiome_1.FULL\n",
      "Done\n",
      "selected_barcodes/TXG_10xv2_1.FULL_bc_passing_filters_otsu.txt\n",
      "TXG_10xv2_1.FULL\n",
      "Done\n",
      "selected_barcodes/UCS_ddseq_1.FULL_bc_passing_filters_otsu.txt\n",
      "UCS_ddseq_1.FULL\n",
      "Done\n",
      "selected_barcodes/VIB_mtscatac_2.FULL_bc_passing_filters_otsu.txt\n",
      "VIB_mtscatac_2.FULL\n",
      "Done\n",
      "selected_barcodes/CNA_10xv11_2.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_10xv11_2.FULL\n",
      "Done\n",
      "selected_barcodes/CNA_hydrop_1.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_hydrop_1.FULL\n",
      "Done\n",
      "selected_barcodes/CNA_10xv11_1.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_10xv11_1.FULL\n",
      "Done\n",
      "selected_barcodes/BIO_ddseq_2.FULL_bc_passing_filters_otsu.txt\n",
      "BIO_ddseq_2.FULL\n",
      "Done\n",
      "selected_barcodes/BRO_mtscatac_1.FULL_bc_passing_filters_otsu.txt\n",
      "BRO_mtscatac_1.FULL\n",
      "Done\n",
      "selected_barcodes/CNA_10xv11_3.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_10xv11_3.FULL\n",
      "Done\n",
      "selected_barcodes/CNA_hydrop_2.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_hydrop_2.FULL\n",
      "Done\n",
      "selected_barcodes/CNA_10xmultiome_1.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_10xmultiome_1.FULL\n",
      "Done\n",
      "selected_barcodes/VIB_10xv1_1.FULL_bc_passing_filters_otsu.txt\n",
      "VIB_10xv1_1.FULL\n",
      "Done\n",
      "selected_barcodes/CNA_mtscatac_1.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_mtscatac_1.FULL\n",
      "Done\n",
      "selected_barcodes/UCS_ddseq_2.FULL_bc_passing_filters_otsu.txt\n",
      "UCS_ddseq_2.FULL\n",
      "Done\n",
      "selected_barcodes/CNA_ddseq_2.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_ddseq_2.FULL\n",
      "Done\n",
      "selected_barcodes/EPF_hydrop_1.FULL_bc_passing_filters_otsu.txt\n",
      "EPF_hydrop_1.FULL\n",
      "Done\n",
      "selected_barcodes/STA_10xv11_1.FULL_bc_passing_filters_otsu.txt\n",
      "STA_10xv11_1.FULL\n",
      "Done\n",
      "selected_barcodes/EPF_hydrop_2.FULL_bc_passing_filters_otsu.txt\n",
      "EPF_hydrop_2.FULL\n",
      "Done\n",
      "selected_barcodes/TXG_10xv2_2.FULL_bc_passing_filters_otsu.txt\n",
      "TXG_10xv2_2.FULL\n",
      "Done\n",
      "selected_barcodes/TXG_10xv11_1.FULL_bc_passing_filters_otsu.txt\n",
      "TXG_10xv11_1.FULL\n",
      "Done\n",
      "selected_barcodes/VIB_hydrop_21.FULL_bc_passing_filters_otsu.txt\n",
      "VIB_hydrop_21.FULL\n",
      "Done\n",
      "selected_barcodes/BIO_ddseq_1.FULL_bc_passing_filters_otsu.txt\n",
      "BIO_ddseq_1.FULL\n",
      "Done\n",
      "selected_barcodes/VIB_10xv2_1.FULL_bc_passing_filters_otsu.txt\n",
      "VIB_10xv2_1.FULL\n",
      "Done\n",
      "selected_barcodes/MDC_mtscatac_2.FULL_bc_passing_filters_otsu.txt\n",
      "MDC_mtscatac_2.FULL\n",
      "Done\n",
      "selected_barcodes/CNA_10xv2_2.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_10xv2_2.FULL\n",
      "Done\n",
      "selected_barcodes/MDC_mtscatac_1.FULL_bc_passing_filters_otsu.txt\n",
      "MDC_mtscatac_1.FULL\n",
      "Done\n",
      "selected_barcodes/VIB_hydrop_12.FULL_bc_passing_filters_otsu.txt\n",
      "VIB_hydrop_12.FULL\n",
      "Done\n",
      "selected_barcodes/CNA_10xv11_4.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_10xv11_4.FULL\n",
      "Done\n",
      "selected_barcodes/VIB_10xmultiome_2.FULL_bc_passing_filters_otsu.txt\n",
      "VIB_10xmultiome_2.FULL\n",
      "Done\n",
      "selected_barcodes/HAR_ddseq_1.FULL_bc_passing_filters_otsu.txt\n",
      "HAR_ddseq_1.FULL\n",
      "Done\n",
      "selected_barcodes/CNA_mtscatac_2.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_mtscatac_2.FULL\n",
      "Done\n",
      "selected_barcodes/HAR_ddseq_2.FULL_bc_passing_filters_otsu.txt\n",
      "HAR_ddseq_2.FULL\n",
      "Done\n",
      "selected_barcodes/VIB_hydrop_22.FULL_bc_passing_filters_otsu.txt\n",
      "VIB_hydrop_22.FULL\n",
      "Done\n",
      "selected_barcodes/EPF_hydrop_3.FULL_bc_passing_filters_otsu.txt\n",
      "EPF_hydrop_3.FULL\n",
      "Done\n",
      "selected_barcodes/CNA_10xv11_5.FULL_bc_passing_filters_otsu.txt\n",
      "CNA_10xv11_5.FULL\n",
      "Done\n",
      "selected_barcodes/SAN_10xmultiome_2.FULL_bc_passing_filters_otsu.txt\n",
      "SAN_10xmultiome_2.FULL\n",
      "Done\n",
      "selected_barcodes/OHS_s3atac_2.FULL_bc_passing_filters_otsu.txt\n",
      "OHS_s3atac_2.FULL\n",
      "Done\n",
      "selected_barcodes/OHS_s3atac_1.FULL_bc_passing_filters_otsu.txt\n",
      "OHS_s3atac_1.FULL\n",
      "Done\n",
      "selected_barcodes/BIO_ddseq_3.FULL_bc_passing_filters_otsu.txt\n",
      "BIO_ddseq_3.FULL\n",
      "Done\n",
      "selected_barcodes/EPF_hydrop_4.FULL_bc_passing_filters_otsu.txt\n",
      "EPF_hydrop_4.FULL\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "for file in glob.glob(\"selected_barcodes/*_bc_passing_filters_otsu.txt\"):\n",
    "    print(file)\n",
    "    sample = file.split(\"/\")[-1].split(\"_bc_passing_filters_otsu.txt\")[0]\n",
    "    print(sample)\n",
    "    df = pd.read_csv(file, header=None, index_col=None)\n",
    "\n",
    "    with open(f\"selected_barcodes/{sample}_bc_passing_filters_otsu.RAW.txt\", \"w\") as fp:\n",
    "        for x in df[0]:\n",
    "            # write each item on a new line\n",
    "            fp.write(x.split(\"___\")[0] + \"\\n\")\n",
    "        print(\"Done\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "20220628_pycistopic.sif",
   "language": "python",
   "name": "cistopic_20220628"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 2184.339725,
   "end_time": "2022-09-26T15:28:53.425907",
   "environment_variables": {},
   "exception": null,
   "input_path": "1_run_qc_consensus.ipynb",
   "output_path": "1_run_qc_consensus.OUT.ipynb",
   "parameters": {},
   "start_time": "2022-09-26T14:52:29.086182",
   "version": "2.3.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
