{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0c3e46cd",
   "metadata": {},
   "source": [
    "# Running vsn"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf019b98",
   "metadata": {},
   "source": [
    "First, generate a metadata file as described here: https://vsn-pipelines.readthedocs.io/en/latest/scatac-seq.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "841beb42",
   "metadata": {},
   "source": [
    "You can do this manually, but I did it automatically using some python string manipulations (see notebook 1b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "17a906c9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-10T17:34:33.505605Z",
     "iopub.status.busy": "2022-10-10T17:34:33.504926Z",
     "iopub.status.idle": "2022-10-10T17:34:33.661499Z",
     "shell.execute_reply": "2022-10-10T17:34:33.660042Z",
     "shell.execute_reply.started": "2022-10-10T17:34:33.505555Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "sample_name\ttechnology\tfastq_PE1_path\tfastq_barcode_path\tfastq_PE2_path\n",
      "TXG_10xmultiome_e18mousebrainfresh.FULL\tmultiome_revcomp\t../1_data_repository/publicdata_full_fastq/all_merged/TXG_10xmultiome_e18mousebrainfresh__R1.FULL.fastq.gz\t../1_data_repository/publicdata_full_fastq/all_merged/TXG_10xmultiome_e18mousebrainfresh__R2.FULL.fastq.gz\t../1_data_repository/publicdata_full_fastq/all_merged/TXG_10xmultiome_e18mousebrainfresh__R3.FULL.fastq.gz\n",
      "TXG_10xv11_adultmousecortexchromiumx.FULL\tatac_revcomp\t../1_data_repository/publicdata_full_fastq/all_merged/TXG_10xv11_adultmousecortexchromiumx__R1.FULL.fastq.gz\t../1_data_repository/publicdata_full_fastq/all_merged/TXG_10xv11_adultmousecortexchromiumx__R2.FULL.fastq.gz\t../1_data_repository/publicdata_full_fastq/all_merged/TXG_10xv11_adultmousecortexchromiumx__R3.FULL.fastq.gz\n",
      "TXG_10xv1_adultmousefresh.FULL\tatac_revcomp\t../1_data_repository/publicdata_full_fastq/all_merged/TXG_10xv1_adultmousefresh__R1.FULL.fastq.gz\t../1_data_repository/publicdata_full_fastq/all_merged/TXG_10xv1_adultmousefresh__R2.FULL.fastq.gz\t../1_data_repository/publicdata_full_fastq/all_merged/TXG_10xv1_adultmousefresh__R3.FULL.fastq.gz\n",
      "TXG_10xv2_adultmousecortex.FULL\tatac_revcomp\t../1_data_repository/publicdata_full_fastq/all_merged/TXG_10xv2_adultmousecortex__R1.FULL.fastq.gz\t../1_data_repository/publicdata_full_fastq/all_merged/TXG_10xv2_adultmousecortex__R2.FULL.fastq.gz\t../1_data_repository/publicdata_full_fastq/all_merged/TXG_10xv2_adultmousecortex__R3.FULL.fastq.gz\n",
      "TXG_10xv2_adultmousecortexchromiumx.FULL\tatac_revcomp\t../1_data_repository/publicdata_full_fastq/all_merged/TXG_10xv2_adultmousecortexchromiumx__R1.FULL.fastq.gz\t../1_data_repository/publicdata_full_fastq/all_merged/TXG_10xv2_adultmousecortexchromiumx__R2.FULL.fastq.gz\t../1_data_repository/publicdata_full_fastq/all_merged/TXG_10xv2_adultmousecortexchromiumx__R3.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310661.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310661__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310661__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310662.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310662__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310662__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310663.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310663__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310663__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310664.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310664__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310664__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310665.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310665__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310665__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310666.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310666__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310666__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310667.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310667__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310667__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310668.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310668__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310668__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310669.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310669__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310669__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310670.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310670__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310670__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310671.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310671__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310671__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310672.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310672__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310672__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310673.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310673__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310673__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310674.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310674__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310674__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310675.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310675__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310675__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310676.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310676__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310676__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310677.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310677__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310677__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310678.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310678__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310678__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310679.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310679__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310679__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310680.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310680__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310680__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310681.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310681__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310681__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310682.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310682__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310682__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310683.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310683__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310683__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310684.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310684__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310684__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310685.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310685__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310685__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310686.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310686__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310686__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310687.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310687__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310687__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310688.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310688__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310688__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310689.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310689__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310689__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310690.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310690__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310690__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310691.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310691__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310691__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310692.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310692__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310692__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310693.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310693__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310693__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310694.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310694__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310694__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310695.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310695__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310695__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310696.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310696__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310696__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310697.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310697__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310697__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310698.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310698__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310698__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310699.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310699__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310699__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310700.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310700__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310700__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310701.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310701__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310701__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310702.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310702__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310702__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310703.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310703__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310703__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310704.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310704__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310704__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310705.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310705__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310705__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310706.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310706__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310706__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310707.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310707__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310707__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310708.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310708__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310708__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310709.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310709__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310709__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310710.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310710__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310710__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310711.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310711__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310711__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310712.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310712__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310712__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310713.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310713__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310713__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310714.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310714__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310714__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310715.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310715__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310715__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310716.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310716__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310716__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310717.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310717__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310717__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310718.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310718__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310718__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310719.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310719__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310719__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310720.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310720__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310720__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310721.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310721__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310721__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310722.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310722__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310722__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310723.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310723__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310723__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310724.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310724__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310724__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310725.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310725__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310725__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310726.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310726__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310726__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310727.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310727__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310727__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310728.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310728__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310728__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310729.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310729__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310729__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310730.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310730__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310730__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310731.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310731__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310731__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8310732.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310732__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8310732__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8994079.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994079__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994079__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8994080.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994080__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994080__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8994081.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994081__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994081__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8994082.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994082__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994082__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8994083.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994083__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994083__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8994084.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994084__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994084__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8994085.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994085__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994085__R2.FULL.fastq.gz\n",
      "BIO_ddseq_SRR8994086.FULL\tbiorad\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994086__R1.FULL.fastq.gz\t\t../1_data_repository/publicdata_full_fastq/all_merged/BIO_ddseq_SRR8994086__R2.FULL.fastq.gz\n",
      "OHS_s3atac_s3ATAC_Mouse.FULL\ts3atac\t../1_data_repository/publicdata_full_fastq/all_merged/OHS_s3atac_s3ATAC_Mouse__R1.FULL.fastq.gz\t../1_data_repository/publicdata_full_fastq/all_merged/OHS_s3atac_s3ATAC_Mouse__R2.FULL.fastq.gz\t../1_data_repository/publicdata_full_fastq/all_merged/OHS_s3atac_s3ATAC_Mouse__R3.FULL.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "cat metadata.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed03a969",
   "metadata": {},
   "source": [
    "Generate a config file read by the vsn pipeline. $nwork is a temporary file destination."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3f43af8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "VSN=../0_resources/vsn-pipelines/main_atac.nf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "9511f16a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-08T18:22:19.931505Z",
     "iopub.status.busy": "2022-06-08T18:22:19.930474Z",
     "iopub.status.idle": "2022-06-08T18:22:23.634196Z",
     "shell.execute_reply": "2022-06-08T18:22:23.630814Z",
     "shell.execute_reply.started": "2022-06-08T18:22:19.931146Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘/scratch/leuven/330/vsc33042/ATAC_BENCHMARK_FULL/’: File exists\n"
     ]
    }
   ],
   "source": [
    "nwork=${VSC_SCRATCH}/ATAC_BENCHMARK_FULL/\n",
    "mkdir $nwork\n",
    "export NXF_WORK=$nwork\n",
    "\n",
    "\n",
    "../0_resources/nextflow/nextflow-21.04.3-all config $VSN \\\n",
    "    -profile atac_preprocess_bap,vsc \\\n",
    "    > atac_preprocess.config"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e115b776",
   "metadata": {},
   "source": [
    "### now make some changes to the config:\n",
    "* redirect to correct metadata_auto.tsv file\n",
    "* fix BWA parameters\n",
    "    * change the bwa index directory to the right, in this case hg39\n",
    "    * change the bwa executor to local to run on current node\n",
    "    * number of bwa CPUs: better to have 2 forks running with 17 threads than to have 1 fork with 36 threads due to I/O overhead\n",
    "* add whitelists for each sample\n",
    "* check if bap parameters are correct\n",
    "* make sure all output dirs etc. exist\n",
    "* Increase the time limit"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b8fc045",
   "metadata": {},
   "source": [
    "Here is a functional config file tailored to our computing environment:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bbf69628",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-10-10T17:44:37.601839Z",
     "iopub.status.busy": "2022-10-10T17:44:37.601236Z",
     "iopub.status.idle": "2022-10-10T17:44:37.786157Z",
     "shell.execute_reply": "2022-10-10T17:44:37.784841Z",
     "shell.execute_reply.started": "2022-10-10T17:44:37.601791Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "manifest {\n",
      "   name = 'vib-singlecell-nf/vsn-pipelines'\n",
      "   description = 'A repository of pipelines for single-cell data in Nextflow DSL2'\n",
      "   homePage = 'https://github.com/vib-singlecell-nf/vsn-pipelines'\n",
      "   version = '0.27.0'\n",
      "   mainScript = 'main.nf'\n",
      "   defaultBranch = 'master'\n",
      "   nextflowVersion = '!21.04.3'\n",
      "}\n",
      "\n",
      "params {\n",
      "   global {\n",
      "      project_name = 'benchmark_atac'\n",
      "      outdir = 'public_preprocessing_output'\n",
      "   }\n",
      "   misc {\n",
      "      test {\n",
      "         enabled = false\n",
      "      }\n",
      "   }\n",
      "   utils {\n",
      "      container = 'vibsinglecellnf/utils:0.4.0'\n",
      "      file_converter {\n",
      "         off = 'h5ad'\n",
      "         tagCellWithSampleId = true\n",
      "         remove10xGEMWell = false\n",
      "         useFilteredMatrix = true\n",
      "         makeVarIndexUnique = false\n",
      "      }\n",
      "      publish {\n",
      "         compressionLevel = 6\n",
      "         annotateWithBatchVariableName = false\n",
      "         mode = 'copy'\n",
      "      }\n",
      "   }\n",
      "   atac_preprocess_tools {\n",
      "      mark_duplicates_method = 'MarkDuplicates'\n",
      "      adapter_trimming_method = 'Trim_Galore'\n",
      "   }\n",
      "   data {\n",
      "      atac_preprocess {\n",
      "         metadata = 'metadata.tsv'\n",
      "      }\n",
      "   }\n",
      "   tools {\n",
      "      singlecelltoolkit {\n",
      "         container = 'vibsinglecellnf/singlecelltoolkit:2022-04-15-16314db'\n",
      "         barcode_correction {\n",
      "            max_mismatches = 1\n",
      "            min_frac_bcs_to_find = 0.5\n",
      "            whitelist {\n",
      "                atac = '../0_resources/whitelists/737K-cratac-v1.txt.gz'\n",
      "                atac_revcomp = '../0_resources/whitelists/737K-cratac-v1.REVCOMP.txt.gz'\n",
      "                multiome = '../0_resources/whitelists/737K-arc-v1.txt.gz'\n",
      "                multiome_revcomp = '../0_resources/whitelists/737K-arc-v1.REVCOMP.txt.gz'\n",
      "                hydrop_2x384 = '../0_resources/whitelists/hydrop_384x384.REVCOMP.txt.gz'\n",
      "                OHS_s3atac_1 = '../1_data_repository/s3_atac/whitelist/OHS_s3atac_1.whitelist.txt'\n",
      "                OHS_s3atac_2 = '../1_data_repository/s3_atac/whitelist/OHS_s3atac_2.whitelist.txt'\n",
      "                s3atac = '../1_data_repository/publicdata_full_fastq/s3atac_publicdata/s3ATAC_Mouse_whitelist.txt'\n",
      "            }\n",
      "         }\n",
      "         barcode_10x_scatac_fastqs {\n",
      "            uncorrected_bc_tag = 'CR'\n",
      "            barcode_quality_tag = 'CY'\n",
      "         }\n",
      "      }\n",
      "      trimgalore {\n",
      "         container = 'vibsinglecellnf/trimgalore:0.6.6'\n",
      "         trim {\n",
      "            paired = 'true'\n",
      "         }\n",
      "      }\n",
      "      fastp {\n",
      "         container = 'vibsinglecellnf/fastp:0.20.0'\n",
      "         thread = 1\n",
      "         clean_and_fastqc {\n",
      "            length_required = 20\n",
      "            adapter_fasta = '/home/luna.kuleuven.be/u0117999/.nextflow/assets/vib-singlecell-nf/vsn-pipelines/src/fastp/assets/fastp.adapters'\n",
      "         }\n",
      "      }\n",
      "      bwamaptools {\n",
      "         container = 'vibsinglecellnf/bwamaptools:bwa-mem2-2.2.1-zlibng-2.0.6'\n",
      "         bwa_fasta = '/dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/0_resources/genomes/mm10_bwamem2/genome.fa'\n",
      "         bwa_version = 'bwa-mem2'\n",
      "      }\n",
      "      gatk {\n",
      "         container = 'broadinstitute/gatk:4.2.0.0'\n",
      "         estimate_library_complexity {\n",
      "            barcode_tag = 'CB'\n",
      "         }\n",
      "      }\n",
      "      sinto {\n",
      "         container = 'vibsinglecellnf/sinto:0.7.3.1'\n",
      "         fragments {\n",
      "            min_mapq = 30\n",
      "            barcodetag = 'CB'\n",
      "            barcode_regex = ''\n",
      "            use_chrom = '\"^(chr|)([0-9]{1,2}|[XY]|[23][LR])$\"'\n",
      "            min_distance = 10\n",
      "            max_distance = 5000\n",
      "            chunksize = 5000000\n",
      "         }\n",
      "      }\n",
      "      barcard {\n",
      "         barcode_multiplet {\n",
      "            report_ipynb = '/src/barcard/bin/barcard_otsu_filtering.ipynb'\n",
      "         }\n",
      "      }\n",
      "      bap {\n",
      "         container = 'vibsinglecellnf/bap:2021-04-27-3b48f4b'\n",
      "         barcode_multiplet {\n",
      "            report_ipynb = '/src/bap/bin/bap_barcode_multiplet_report.ipynb'\n",
      "            bead_tag = 'CB'\n",
      "            drop_tag = 'DB'\n",
      "            minimum_barcode_fragments = 1000\n",
      "            barcode_whitelist = ''\n",
      "            minimum_jaccard_index = [\n",
      "                default: 0.02,\n",
      "                BIO_ddseq_3: 0.06502,\n",
      "                BRO_mtscatac_1: 0.03449,\n",
      "                BRO_mtscatac_2: 0.07238,\n",
      "                CNA_10xmultiome_1: 0.02,\n",
      "                CNA_10xmultiome_2: 0.02,\n",
      "                CNA_10xv11_4: 0.04732,\n",
      "                CNA_ddseq_1: 0.0715,\n",
      "                CNA_mtscatac_2: 0.02631,\n",
      "                HAR_ddseq_1: 0.03247,\n",
      "                VIB_10xmultiome_1: 0.11137,\n",
      "                VIB_10xmultiome_2: 0.03885,\n",
      "                VIB_hydrop_11: 0.02164,\n",
      "                VIB_hydrop_21: 0.04463\n",
      "            ]\n",
      "            nc_threshold = 6\n",
      "            regularize_threshold = 4\n",
      "            mapq = 30\n",
      "            max_insert = 2000\n",
      "            reference_genome = 'hg38'\n",
      "            bedtools_genome = ''\n",
      "            blacklist_file = ''\n",
      "            tss_file = ''\n",
      "            mito_chromosome = 'default'\n",
      "         }\n",
      "      }\n",
      "   }\n",
      "}\n",
      "\n",
      "process {\n",
      "    executor = 'local'\n",
      "    cpus = 4\n",
      "    memory = '20 GB'\n",
      "    disk = '100 GB'\n",
      "    clusterOptions = '-A cluster_account'\n",
      "    time = '24h'\n",
      "    maxForks = 4\n",
      "\n",
      "    withLabel:compute_resources__default {\n",
      "       time = '24h'\n",
      "    }\n",
      "    withLabel:compute_resources__minimal {\n",
      "       cpus = 1\n",
      "       memory = '1 GB'\n",
      "    }\n",
      "    withLabel:compute_resources__sctk_barcode {\n",
      "       cpus = 2\n",
      "       memory = '3 GB'\n",
      "    }\n",
      "\n",
      "    withLabel:compute_resources__mem {\n",
      "       cpus = 1\n",
      "       memory = '40 GB'\n",
      "    }\n",
      "    withLabel:compute_resources__cpu {\n",
      "       cpus = 5\n",
      "       memory = '4 GB'\n",
      "    }\n",
      "    withLabel:compute_resources__report {\n",
      "       cpus = 1\n",
      "       memory = '2 GB'\n",
      "    }\n",
      "    withLabel:compute_resources__24hqueue {\n",
      "       time = '99h'\n",
      "    }\n",
      "    withLabel:compute_resources__bwa_mem {\n",
      "       executor = 'local'\n",
      "       cpus = 8\n",
      "       memory = '40 GB'\n",
      "       time = '999h'\n",
      "    }\n",
      "    withLabel:compute_resources__sctk_barcode {\n",
      "       cpus = 5\n",
      "       memory = '5 GB'\n",
      "    }\n",
      "\n",
      "    withLabel:compute_resources__barcode_10x_scatac_fastq_5cpus {\n",
      "       cpus = 4\n",
      "       memory = '5 GB'\n",
      "    }\n",
      "\n",
      "    withLabel:compute_resources__barcard__create_fragments_from_bam {\n",
      "       cpus = 4\n",
      "       memory = '5 GB'\n",
      "    }\n",
      "\n",
      "    withLabel:compute_resources__barcard__detect_barcode_multiplets {\n",
      "       cpus = 4\n",
      "       memory = '5 GB'\n",
      "    }\n",
      "\n",
      "    withLabel:compute_resources__samtools__write_fragments {\n",
      "       cpus = 5\n",
      "       memory = '5 GB'\n",
      "    }\n",
      "\n",
      "    withLabel:compute_resources__trimgalore__trim_5cpus {\n",
      "       cpus = 5\n",
      "       memory = '10 GB'\n",
      "    }\n",
      "\n",
      "    withLabel:compute_resources__picard__merge_sam_files_and_sort {\n",
      "       cpus = 4\n",
      "       memory = '5 GB'\n",
      "    }\n",
      "    withLabel:compute_resources__bap_barcode_multiplet_pipeline_8cpus {\n",
      "       cpus = 4\n",
      "       memory = '5 GB'\n",
      "    }\n",
      "}\n",
      "\n",
      "timeline {\n",
      "   enabled = true\n",
      "   file = 'out/nextflow_reports/execution_timeline.html'\n",
      "}\n",
      "\n",
      "report {\n",
      "   enabled = true\n",
      "   file = 'out/nextflow_reports/execution_report.html'\n",
      "}\n",
      "\n",
      "trace {\n",
      "   enabled = true\n",
      "   file = 'out/nextflow_reports/execution_trace.txt'\n",
      "}\n",
      "\n",
      "dag {\n",
      "   enabled = true\n",
      "   file = 'out/nextflow_reports/pipeline_dag.svg'\n",
      "}\n",
      "\n",
      "min {\n",
      "   enabled = false\n",
      "}\n",
      "\n",
      "vsc {\n",
      "   enabled = true\n",
      "}\n",
      "\n",
      "singularity {\n",
      "   enabled = true\n",
      "   autoMounts = true\n",
      "   runOptions = '--cleanenv -H $PWD -B /lustre1,/staging,/data,${VSC_SCRATCH},${VSC_SCRATCH}/tmp:/tmp,${HOME}/.nextflow/assets/'\n",
      "   cacheDir = '/lustre1/project/stg_00090/scatac_benchmark/0_resources/vsn_cache'\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "cat atac_preprocess_rapid.config"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43dd49b5",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-06-09T08:56:14.239433Z",
     "iopub.status.busy": "2022-06-09T08:56:14.239085Z",
     "iopub.status.idle": "2022-06-09T08:56:14.493186Z",
     "shell.execute_reply": "2022-06-09T08:56:14.490236Z",
     "shell.execute_reply.started": "2022-06-09T08:56:14.239394Z"
    }
   },
   "source": [
    "I suggest you run a difftool on this to compare this to your config."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b62325f",
   "metadata": {},
   "source": [
    "Then, in a tmux session to avoid interruption:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7cf4d197",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "sleep 2h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96c11b8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "VSN=../0_resources/vsn-pipelines_/main_atac.nf\n",
    "NXF=/readonly/dodrio/scratch/projects/starting_2022_023/software/nextflow/nextflow-21.04.3/nextflow\n",
    "$NXF -C atac_preprocess_rapid.config run $VSN -entry atac_preprocess_rapid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6adccf0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "commented some shit "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
