{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Generate the main reference (R):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# RNA-ATAC integration\n",
    "################################################################################\n",
    "# https://satijalab.org/seurat/articles/atacseq_integration_vignette.html\n",
    "\n",
    "library(Seurat)\n",
    "library(SeuratDisk)\n",
    "library(Signac)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "library(ggplot2)\n",
    "\n",
    "################################################################################\n",
    "################################################################################\n",
    "# reference data:\n",
    "################################################################################\n",
    "\n",
    "f_ref = '/lustre1/project/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/scRNA/cell_type_classification/seurat/reference/pbmc_ssc_mat.rds'\n",
    "f_meta = '/lustre1/project/stg_00002/lcb/cflerin/analysis/pbmc_atac/analysis2/scRNA/cell_type_classification/seurat/reference/pbmc_ssc_metadata.rds'\n",
    "\n",
    "pbmc.data <- readRDS(f_ref)\n",
    "pbmc.metadata <- readRDS(f_meta)\n",
    "\n",
    "pbmc <- CreateSeuratObject(counts = pbmc.data, meta.data = pbmc.metadata)\n",
    "pbmc <- subset(pbmc, subset = nFeature_RNA > 200)\n",
    "\n",
    "pbmc.list <- SplitObject(pbmc, split.by = \"Method\")\n",
    "#pbmc.list$bench = qdata\n",
    "\n",
    "for (i in names(pbmc.list)) {\n",
    "    pbmc.list[[i]] = NormalizeData(pbmc.list[[i]])\n",
    "    pbmc.list[[i]] = FindVariableFeatures(pbmc.list[[i]], selection.method = \"vst\", nfeatures = 2000, verbose = FALSE)\n",
    "    pbmc.list[[i]] = ScaleData(pbmc.list[[i]])\n",
    "    pbmc.list[[i]] = RunPCA(pbmc.list[[i]])\n",
    "    pbmc.list[[i]] = RunUMAP(pbmc.list[[i]], dims = 1:30)\n",
    "}\n",
    "\n",
    "##################################################\n",
    "### integrate reference:\n",
    "\n",
    "reference.list <- pbmc.list[\n",
    "    c(\"10x Chromium (v2) A\", \"10x Chromium (v2) B\", \"10x Chromium (v3)\", \"10x Chromium (v2)\")\n",
    "]\n",
    "#reference.list$query = qdata\n",
    "#pbmc.ref = pbmc.list$`10x Chromium (v3)`\n",
    "pbmc.anchors <- FindIntegrationAnchors(object.list = reference.list, dims = 1:30)\n",
    "\n",
    "pbmc.integrated <- IntegrateData(anchorset = pbmc.anchors, dims = 1:30)\n",
    "DefaultAssay(pbmc.integrated) <- \"integrated\"\n",
    "\n",
    "pbmc.integrated <- ScaleData(pbmc.integrated, verbose = FALSE)\n",
    "pbmc.integrated <- RunPCA(pbmc.integrated, npcs = 30, verbose = FALSE)\n",
    "pbmc.integrated <- RunUMAP(pbmc.integrated, reduction = \"pca\", dims = 1:30, verbose = FALSE)\n",
    "\n",
    "##################################################\n",
    "\n",
    "pbmc.rna = pbmc.list$`10x Chromium (v3)`\n",
    "\n",
    "saveRDS(pbmc.rna, file='pbmc_RNA_reference_10xv3.rds')\n",
    "\n",
    "# pbmc.rna <- NormalizeData(pbmc.rna)\n",
    "# pbmc.rna <- FindVariableFeatures(pbmc.rna)\n",
    "# pbmc.rna <- ScaleData(pbmc.rna)\n",
    "# pbmc.rna <- RunPCA(pbmc.rna)\n",
    "# pbmc.rna <- RunUMAP(pbmc.rna, dims = 1:30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remove megakaryocytes manually"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# RNA-ATAC integration\n",
    "################################################################################\n",
    "# https://satijalab.org/seurat/articles/atacseq_integration_vignette.html\n",
    "\n",
    "library(Seurat)\n",
    "library(SeuratDisk)\n",
    "library(Signac)\n",
    "library(EnsDb.Hsapiens.v86)\n",
    "library(ggplot2)\n",
    "\n",
    "\n",
    "################################################################################\n",
    "# reference data:\n",
    "################################################################################\n",
    "\n",
    "f_ref = '../0_resources/seurat_references/pbmc_ssc_mat.rds'\n",
    "f_meta = '../0_resources/seurat_references/pbmc_ssc_metadata.rds'\n",
    "\n",
    "\n",
    "pbmc.data <- readRDS(f_ref)\n",
    "pbmc.metadata <- readRDS(f_meta)\n",
    "\n",
    "pbmc <- CreateSeuratObject(counts = pbmc.data, meta.data = pbmc.metadata)\n",
    "pbmc <- subset(pbmc, subset = nFeature_RNA > 200)\n",
    "\n",
    "pbmc.list <- SplitObject(pbmc, split.by = \"Method\")\n",
    "#pbmc.list$bench = qdata\n",
    "\n",
    "for (i in names(pbmc.list)) {\n",
    "    pbmc.list[[i]] = NormalizeData(pbmc.list[[i]])\n",
    "    pbmc.list[[i]] = FindVariableFeatures(pbmc.list[[i]], selection.method = \"vst\", nfeatures = 2000, verbose = FALSE)\n",
    "    pbmc.list[[i]] = ScaleData(pbmc.list[[i]])\n",
    "    pbmc.list[[i]] = RunPCA(pbmc.list[[i]])\n",
    "    pbmc.list[[i]] = RunUMAP(pbmc.list[[i]], dims = 1:30)\n",
    "}\n",
    "\n",
    "##################################################\n",
    "### integrate reference:\n",
    "\n",
    "reference.list <- pbmc.list[\n",
    "    c(\"10x Chromium (v2) A\", \"10x Chromium (v2) B\", \"10x Chromium (v3)\", \"10x Chromium (v2)\")\n",
    "]\n",
    "#reference.list$query = qdata\n",
    "#pbmc.ref = pbmc.list$`10x Chromium (v3)`\n",
    "pbmc.anchors <- FindIntegrationAnchors(object.list = reference.list, dims = 1:30)\n",
    "\n",
    "pbmc.integrated <- IntegrateData(anchorset = pbmc.anchors, dims = 1:30)\n",
    "DefaultAssay(pbmc.integrated) <- \"integrated\"\n",
    "\n",
    "pbmc.integrated <- ScaleData(pbmc.integrated, verbose = FALSE)\n",
    "pbmc.integrated <- RunPCA(pbmc.integrated, npcs = 30, verbose = FALSE)\n",
    "pbmc.integrated <- RunUMAP(pbmc.integrated, reduction = \"pca\", dims = 1:30, verbose = FALSE)\n",
    "\n",
    "##################################################\n",
    "\n",
    "# pbmc.rna = pbmc.list$`10x Chromium (v3)`\n",
    "\n",
    "pbmc.rna = readRDS('../0_resources/seurat_references/pbmc_ssc_mat__integrated.rds')\n",
    "\n",
    "# pbmc.rna <- NormalizeData(pbmc.rna)\n",
    "# pbmc.rna <- FindVariableFeatures(pbmc.rna)\n",
    "# pbmc.rna <- ScaleData(pbmc.rna)\n",
    "# pbmc.rna <- RunPCA(pbmc.rna)\n",
    "# pbmc.rna <- RunUMAP(pbmc.rna, dims = 1:30)\n",
    "\n",
    "saveRDS(pbmc, file=\"../0_resources/seurat_references/pbmc_ref.rds\")\n",
    "\n",
    "\n",
    "# ATAC analysis add gene annotation information\n",
    "annotations <- GetGRangesFromEnsDb(ensdb = EnsDb.Hsapiens.v86)\n",
    "\n",
    "# use chr prefix:\n",
    "ucsc.levels <- str_replace(string=paste(\"chr\",seqlevels(annotations),sep=\"\"), pattern=\"chrMT\", replacement=\"chrM\")\n",
    "seqlevels(annotations) <- ucsc.levels\n",
    "\n",
    "genome(annotations) <- \"hg38\"\n",
    "saveRDS(annotations, file=\"../0_resources/seurat_references/granges_annotation.rds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Then, invoke the following R script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/usr/bin/env Rscript\n",
      "library(Seurat)\n",
      "library(SeuratDisk)\n",
      "library(Signac)\n",
      "library(EnsDb.Hsapiens.v86)\n",
      "library(ggplot2)\n",
      "library(stringr)\n",
      "\n",
      "args = commandArgs(trailingOnly=TRUE)\n",
      "sample_id = args[1]\n",
      "f_loom = args[2]\n",
      "f_frag = args[3]\n",
      "f_reference = args[4]\n",
      "f_annotation = args[5]\n",
      "f_out = args[6]\n",
      "\n",
      "print(paste0(\"Processing sample \", args[1]))\n",
      "\n",
      "# load pbmc object\n",
      "# pbmc.integrated <- readRDS(\"../0_resources/seurat_references/pbmc_integrated.RDS\")\n",
      "pbmc.rna <- readRDS('../0_resources/seurat_references/pbmc_ssc_mat__integrated.rds')\n",
      "\n",
      "################################################################################\n",
      "# ATAC\n",
      "################################################################################\n",
      "\n",
      "### get data from loom:\n",
      "atacloomcon <- Connect(filename = f_loom, mode = \"r\")\n",
      "atacloomcon\n",
      "atac_tmp <- as.Seurat(atacloomcon, assay='ATAC')\n",
      "atacloomcon$close_all()\n",
      "\n",
      "# correctly parse regions (default delims are '-','-')\n",
      "regions = StringToGRanges(\n",
      "    rownames(GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC')),\n",
      "    sep=c(':','-')\n",
      "    )\n",
      "\n",
      "# create chromatin assay\n",
      "chromatinassay = CreateChromatinAssay(\n",
      "    counts=GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC'),\n",
      "    # genome='hg38',\n",
      "    fragments = f_frag,\n",
      "    ranges=regions\n",
      "    )\n",
      "    #annotation=annotation)\n",
      "\n",
      "atac <- CreateSeuratObject(counts = chromatinassay, assay='ATAC')\n",
      "\n",
      "annotations <- readRDS(f_annotation)\n",
      "Annotation(atac) <- annotations\n",
      "\n",
      "# We exclude the first dimension as this is typically correlated with sequencing depth\n",
      "atac <- RunTFIDF(atac)\n",
      "atac <- FindTopFeatures(atac, min.cutoff = \"q0\")\n",
      "atac <- RunSVD(atac)\n",
      "atac <- RunUMAP(atac, reduction = \"lsi\", dims = 2:30, reduction.name = \"umap.atac\", reduction.key = \"atacUMAP_\")\n",
      "\n",
      "##################################################\n",
      "# Identify RNA-ATAC anchors\n",
      "\n",
      "# quantify gene activity\n",
      "gene.activities <- GeneActivity(atac, features = VariableFeatures(pbmc.rna))\n",
      "# gene.activities <- gene.activities[rownames(gene.activities)!=\"\",]\n",
      "\n",
      "# add gene activities as a new assay\n",
      "atac[[\"ACTIVITY\"]] <- CreateAssayObject(counts = gene.activities)\n",
      "\n",
      "# normalize gene activities\n",
      "DefaultAssay(atac) <- \"ACTIVITY\"\n",
      "atac <- NormalizeData(atac)\n",
      "atac <- ScaleData(atac, features = rownames(atac))\n",
      "\n",
      "# Identify anchors\n",
      "transfer.anchors <- FindTransferAnchors(reference = pbmc.rna,\n",
      "    query = atac,\n",
      "    features = VariableFeatures(object = pbmc.rna),\n",
      "    reference.assay = \"RNA\",\n",
      "    query.assay = \"ACTIVITY\",\n",
      "    reduction = \"cca\")\n",
      "\n",
      "# predict celltype\n",
      "celltype.predictions <- TransferData(\n",
      "    anchorset = transfer.anchors,\n",
      "    refdata = pbmc.rna$CellType,\n",
      "    weight.reduction = atac[[\"lsi\"]],\n",
      "    dims = 2:30)\n",
      "\n",
      "pbmc.atac <- AddMetaData(atac, metadata = celltype.predictions)\n",
      "\n",
      "md = celltype.predictions\n",
      "pred_thr = 0.7\n",
      "\n",
      "tmp = data.frame(\n",
      "      composite_sample_id = paste0(rownames(md),'-',sample_id),\n",
      "      barcode = rownames(md),\n",
      "      sample_id = sample_id,\n",
      "      cell_type = md$predicted.id,\n",
      "      cell_type_pred_score = md$prediction.score.max\n",
      "      )\n",
      "\n",
      "tmp$cell_type_hiconf_70 = tmp$cell_type\n",
      "tmp$cell_type_hiconf_70[tmp$cell_type_pred_score<pred_thr] = 'Unknown'\n",
      "\n",
      "table(tmp$cell_type)\n",
      "table(tmp$cell_type_hiconf_70)\n",
      "\n",
      "write.table(tmp, file=f_out, sep='\\t', row.names=FALSE, quote=FALSE)\n",
      "print(paste0('Wrote file ', f_out))\n"
     ]
    }
   ],
   "source": [
    "!cat ../0_resources/scripts/seurat_label_transfer.R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/usr/bin/env Rscript\n",
      "library(Seurat)\n",
      "library(SeuratDisk)\n",
      "library(Signac)\n",
      "library(EnsDb.Hsapiens.v86)\n",
      "library(ggplot2)\n",
      "library(stringr)\n",
      "\n",
      "args = commandArgs(trailingOnly=TRUE)\n",
      "sample_id = args[1]\n",
      "f_loom = args[2]\n",
      "f_frag = args[3]\n",
      "f_reference = args[4]\n",
      "f_annotation = args[5]\n",
      "f_out = args[6]\n",
      "\n",
      "print(paste0(\"Processing sample \", args[1]))\n",
      "\n",
      "# load pbmc object\n",
      "# pbmc.integrated <- readRDS(\"../0_resources/seurat_references/pbmc_integrated.RDS\")\n",
      "pbmc.rna <- readRDS('../0_resources/seurat_references/pbmc_ssc_mat__integrated.rds')\n",
      "\n",
      "################################################################################\n",
      "# ATAC\n",
      "################################################################################\n",
      "\n",
      "### get data from loom:\n",
      "atacloomcon <- Connect(filename = f_loom, mode = \"r\")\n",
      "atacloomcon\n",
      "atac_tmp <- as.Seurat(atacloomcon, assay='ATAC')\n",
      "atacloomcon$close_all()\n",
      "\n",
      "# correctly parse regions (default delims are '-','-')\n",
      "regions = StringToGRanges(\n",
      "    rownames(GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC')),\n",
      "    sep=c(':','-')\n",
      "    )\n",
      "\n",
      "# create chromatin assay\n",
      "chromatinassay = CreateChromatinAssay(\n",
      "    counts=GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC'),\n",
      "    # genome='hg38',\n",
      "    fragments = f_frag,\n",
      "    ranges=regions\n",
      "    )\n",
      "    #annotation=annotation)\n",
      "\n",
      "atac <- CreateSeuratObject(counts = chromatinassay, assay='ATAC')\n",
      "\n",
      "annotations <- readRDS(f_annotation)\n",
      "Annotation(atac) <- annotations\n",
      "\n",
      "# We exclude the first dimension as this is typically correlated with sequencing depth\n",
      "atac <- RunTFIDF(atac)\n",
      "atac <- FindTopFeatures(atac, min.cutoff = \"q0\")\n",
      "atac <- RunSVD(atac)\n",
      "atac <- RunUMAP(atac, reduction = \"lsi\", dims = 2:30, reduction.name = \"umap.atac\", reduction.key = \"atacUMAP_\")\n",
      "\n",
      "##################################################\n",
      "# Identify RNA-ATAC anchors\n",
      "\n",
      "# quantify gene activity\n",
      "gene.activities <- GeneActivity(atac, features = VariableFeatures(pbmc.rna))\n",
      "# gene.activities <- gene.activities[rownames(gene.activities)!=\"\",]\n",
      "\n",
      "# add gene activities as a new assay\n",
      "atac[[\"ACTIVITY\"]] <- CreateAssayObject(counts = gene.activities)\n",
      "\n",
      "# normalize gene activities\n",
      "DefaultAssay(atac) <- \"ACTIVITY\"\n",
      "atac <- NormalizeData(atac)\n",
      "atac <- ScaleData(atac, features = rownames(atac))\n",
      "\n",
      "# Identify anchors\n",
      "transfer.anchors <- FindTransferAnchors(reference = pbmc.rna,\n",
      "    query = atac,\n",
      "    features = VariableFeatures(object = pbmc.rna),\n",
      "    reference.assay = \"RNA\",\n",
      "    query.assay = \"ACTIVITY\",\n",
      "    reduction = \"cca\")\n",
      "\n",
      "# predict celltype\n",
      "celltype.predictions <- TransferData(\n",
      "    anchorset = transfer.anchors,\n",
      "    refdata = pbmc.rna$CellType,\n",
      "    weight.reduction = atac[[\"lsi\"]],\n",
      "    dims = 2:30)\n",
      "\n",
      "pbmc.atac <- AddMetaData(atac, metadata = celltype.predictions)\n",
      "\n",
      "md = celltype.predictions\n",
      "pred_thr = 0.7\n",
      "\n",
      "tmp = data.frame(\n",
      "      composite_sample_id = paste0(rownames(md),'-',sample_id),\n",
      "      barcode = rownames(md),\n",
      "      sample_id = sample_id,\n",
      "      cell_type = md$predicted.id,\n",
      "      cell_type_pred_score = md$prediction.score.max\n",
      "      )\n",
      "\n",
      "tmp$cell_type_hiconf_70 = tmp$cell_type\n",
      "tmp$cell_type_hiconf_70[tmp$cell_type_pred_score<pred_thr] = 'Unknown'\n",
      "\n",
      "table(tmp$cell_type)\n",
      "table(tmp$cell_type_hiconf_70)\n",
      "\n",
      "write.table(tmp, file=f_out, sep='\\t', row.names=FALSE, quote=FALSE)\n",
      "print(paste0('Wrote file ', f_out))\n"
     ]
    }
   ],
   "source": [
    "!cat /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/0_resources/scripts/seurat_label_transfer.R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import glob\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/full_2_cistopic\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FULL': '../1_data_repository/full_fragments/BIO_ddseq_1.FULL.fragments.tsv.gz',\n",
       " 'BIO_ddseq_2.FULL': '../1_data_repository/full_fragments/BIO_ddseq_2.FULL.fragments.tsv.gz',\n",
       " 'BIO_ddseq_3.FULL': '../1_data_repository/full_fragments/BIO_ddseq_3.FULL.fragments.tsv.gz',\n",
       " 'BIO_ddseq_4.FULL': '../1_data_repository/full_fragments/BIO_ddseq_4.FULL.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_1.FULL': '../1_data_repository/full_fragments/BRO_mtscatac_1.FULL.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_2.FULL': '../1_data_repository/full_fragments/BRO_mtscatac_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_1.FULL': '../1_data_repository/full_fragments/CNA_10xmultiome_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_2.FULL': '../1_data_repository/full_fragments/CNA_10xmultiome_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_1.FULL': '../1_data_repository/full_fragments/CNA_10xv11_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_2.FULL': '../1_data_repository/full_fragments/CNA_10xv11_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_3.FULL': '../1_data_repository/full_fragments/CNA_10xv11_3.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_4.FULL': '../1_data_repository/full_fragments/CNA_10xv11_4.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_5.FULL': '../1_data_repository/full_fragments/CNA_10xv11_5.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv2_1.FULL': '../1_data_repository/full_fragments/CNA_10xv2_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv2_2.FULL': '../1_data_repository/full_fragments/CNA_10xv2_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_hydrop_1.FULL': '../1_data_repository/full_fragments/CNA_hydrop_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_hydrop_2.FULL': '../1_data_repository/full_fragments/CNA_hydrop_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_hydrop_3.FULL': '../1_data_repository/full_fragments/CNA_hydrop_3.FULL.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_1.FULL': '../1_data_repository/full_fragments/CNA_mtscatac_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_2.FULL': '../1_data_repository/full_fragments/CNA_mtscatac_2.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_1.FULL': '../1_data_repository/full_fragments/EPF_hydrop_1.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_2.FULL': '../1_data_repository/full_fragments/EPF_hydrop_2.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_3.FULL': '../1_data_repository/full_fragments/EPF_hydrop_3.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_4.FULL': '../1_data_repository/full_fragments/EPF_hydrop_4.FULL.fragments.tsv.gz',\n",
       " 'HAR_ddseq_1.FULL': '../1_data_repository/full_fragments/HAR_ddseq_1.FULL.fragments.tsv.gz',\n",
       " 'HAR_ddseq_2.FULL': '../1_data_repository/full_fragments/HAR_ddseq_2.FULL.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_1.FULL': '../1_data_repository/full_fragments/MDC_mtscatac_1.FULL.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_2.FULL': '../1_data_repository/full_fragments/MDC_mtscatac_2.FULL.fragments.tsv.gz',\n",
       " 'OHS_s3atac_1.FULL': '../1_data_repository/full_fragments/OHS_s3atac_1.FULL.fragments.tsv.gz',\n",
       " 'OHS_s3atac_2.FULL': '../1_data_repository/full_fragments/OHS_s3atac_2.FULL.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_1.FULL': '../1_data_repository/full_fragments/SAN_10xmultiome_1.FULL.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_2.FULL': '../1_data_repository/full_fragments/SAN_10xmultiome_2.FULL.fragments.tsv.gz',\n",
       " 'STA_10xv11_1.FULL': '../1_data_repository/full_fragments/STA_10xv11_1.FULL.fragments.tsv.gz',\n",
       " 'STA_10xv11_2.FULL': '../1_data_repository/full_fragments/STA_10xv11_2.FULL.fragments.tsv.gz',\n",
       " 'TXG_10xv11_1.FULL': '../1_data_repository/full_fragments/TXG_10xv11_1.FULL.fragments.tsv.gz',\n",
       " 'TXG_10xv2_1.FULL': '../1_data_repository/full_fragments/TXG_10xv2_1.FULL.fragments.tsv.gz',\n",
       " 'TXG_10xv2_2.FULL': '../1_data_repository/full_fragments/TXG_10xv2_2.FULL.fragments.tsv.gz',\n",
       " 'UCS_ddseq_1.FULL': '../1_data_repository/full_fragments/UCS_ddseq_1.FULL.fragments.tsv.gz',\n",
       " 'UCS_ddseq_2.FULL': '../1_data_repository/full_fragments/UCS_ddseq_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_1.FULL': '../1_data_repository/full_fragments/VIB_10xmultiome_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_2.FULL': '../1_data_repository/full_fragments/VIB_10xmultiome_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv1_1.FULL': '../1_data_repository/full_fragments/VIB_10xv1_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv1_2.FULL': '../1_data_repository/full_fragments/VIB_10xv1_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv2_1.FULL': '../1_data_repository/full_fragments/VIB_10xv2_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv2_2.FULL': '../1_data_repository/full_fragments/VIB_10xv2_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_1.FULL': '../1_data_repository/full_fragments/VIB_hydrop_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_11.FULL': '../1_data_repository/full_fragments/VIB_hydrop_11.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_12.FULL': '../1_data_repository/full_fragments/VIB_hydrop_12.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_2.FULL': '../1_data_repository/full_fragments/VIB_hydrop_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_21.FULL': '../1_data_repository/full_fragments/VIB_hydrop_21.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_22.FULL': '../1_data_repository/full_fragments/VIB_hydrop_22.FULL.fragments.tsv.gz'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scrub_name_suffix = \"0-4\"\n",
    "frags_path_dict  = {x.split('/')[-1].split(f'.fragments.tsv.gz')[0]:x for x in sorted(glob.glob(f'../1_data_repository/full_fragments/*.fragments.tsv.gz'))}\n",
    "frags_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FULL': '../1_data_repository/full_fragments/BIO_ddseq_1.FULL.fragments.tsv.gz',\n",
       " 'BIO_ddseq_2.FULL': '../1_data_repository/full_fragments/BIO_ddseq_2.FULL.fragments.tsv.gz',\n",
       " 'BIO_ddseq_3.FULL': '../1_data_repository/full_fragments/BIO_ddseq_3.FULL.fragments.tsv.gz',\n",
       " 'BIO_ddseq_4.FULL': '../1_data_repository/full_fragments/BIO_ddseq_4.FULL.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_1.FULL': '../1_data_repository/full_fragments/BRO_mtscatac_1.FULL.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_2.FULL': '../1_data_repository/full_fragments/BRO_mtscatac_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_1.FULL': '../1_data_repository/full_fragments/CNA_10xmultiome_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_2.FULL': '../1_data_repository/full_fragments/CNA_10xmultiome_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_1.FULL': '../1_data_repository/full_fragments/CNA_10xv11_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_2.FULL': '../1_data_repository/full_fragments/CNA_10xv11_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_3.FULL': '../1_data_repository/full_fragments/CNA_10xv11_3.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_4.FULL': '../1_data_repository/full_fragments/CNA_10xv11_4.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_5.FULL': '../1_data_repository/full_fragments/CNA_10xv11_5.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv2_1.FULL': '../1_data_repository/full_fragments/CNA_10xv2_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv2_2.FULL': '../1_data_repository/full_fragments/CNA_10xv2_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_hydrop_1.FULL': '../1_data_repository/full_fragments/CNA_hydrop_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_hydrop_2.FULL': '../1_data_repository/full_fragments/CNA_hydrop_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_hydrop_3.FULL': '../1_data_repository/full_fragments/CNA_hydrop_3.FULL.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_1.FULL': '../1_data_repository/full_fragments/CNA_mtscatac_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_2.FULL': '../1_data_repository/full_fragments/CNA_mtscatac_2.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_1.FULL': '../1_data_repository/full_fragments/EPF_hydrop_1.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_2.FULL': '../1_data_repository/full_fragments/EPF_hydrop_2.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_3.FULL': '../1_data_repository/full_fragments/EPF_hydrop_3.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_4.FULL': '../1_data_repository/full_fragments/EPF_hydrop_4.FULL.fragments.tsv.gz',\n",
       " 'HAR_ddseq_1.FULL': '../1_data_repository/full_fragments/HAR_ddseq_1.FULL.fragments.tsv.gz',\n",
       " 'HAR_ddseq_2.FULL': '../1_data_repository/full_fragments/HAR_ddseq_2.FULL.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_1.FULL': '../1_data_repository/full_fragments/MDC_mtscatac_1.FULL.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_2.FULL': '../1_data_repository/full_fragments/MDC_mtscatac_2.FULL.fragments.tsv.gz',\n",
       " 'OHS_s3atac_1.FULL': '../1_data_repository/full_fragments/OHS_s3atac_1.FULL.fragments.tsv.gz',\n",
       " 'OHS_s3atac_2.FULL': '../1_data_repository/full_fragments/OHS_s3atac_2.FULL.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_1.FULL': '../1_data_repository/full_fragments/SAN_10xmultiome_1.FULL.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_2.FULL': '../1_data_repository/full_fragments/SAN_10xmultiome_2.FULL.fragments.tsv.gz',\n",
       " 'STA_10xv11_1.FULL': '../1_data_repository/full_fragments/STA_10xv11_1.FULL.fragments.tsv.gz',\n",
       " 'STA_10xv11_2.FULL': '../1_data_repository/full_fragments/STA_10xv11_2.FULL.fragments.tsv.gz',\n",
       " 'TXG_10xv11_1.FULL': '../1_data_repository/full_fragments/TXG_10xv11_1.FULL.fragments.tsv.gz',\n",
       " 'TXG_10xv2_1.FULL': '../1_data_repository/full_fragments/TXG_10xv2_1.FULL.fragments.tsv.gz',\n",
       " 'TXG_10xv2_2.FULL': '../1_data_repository/full_fragments/TXG_10xv2_2.FULL.fragments.tsv.gz',\n",
       " 'UCS_ddseq_1.FULL': '../1_data_repository/full_fragments/UCS_ddseq_1.FULL.fragments.tsv.gz',\n",
       " 'UCS_ddseq_2.FULL': '../1_data_repository/full_fragments/UCS_ddseq_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_1.FULL': '../1_data_repository/full_fragments/VIB_10xmultiome_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_2.FULL': '../1_data_repository/full_fragments/VIB_10xmultiome_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv1_1.FULL': '../1_data_repository/full_fragments/VIB_10xv1_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv1_2.FULL': '../1_data_repository/full_fragments/VIB_10xv1_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv2_1.FULL': '../1_data_repository/full_fragments/VIB_10xv2_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv2_2.FULL': '../1_data_repository/full_fragments/VIB_10xv2_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_1.FULL': '../1_data_repository/full_fragments/VIB_hydrop_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_11.FULL': '../1_data_repository/full_fragments/VIB_hydrop_11.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_12.FULL': '../1_data_repository/full_fragments/VIB_hydrop_12.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_2.FULL': '../1_data_repository/full_fragments/VIB_hydrop_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_21.FULL': '../1_data_repository/full_fragments/VIB_hydrop_21.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_22.FULL': '../1_data_repository/full_fragments/VIB_hydrop_22.FULL.fragments.tsv.gz'}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scrub_name_suffix = \"0-4\"\n",
    "frags_path_dict  = {x.split('/')[-1].split(f'.fragments.tsv.gz')[0] :x for x in sorted(glob.glob(f'../1_data_repository/full_fragments/*.fragments.tsv.gz'))}\n",
    "frags_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FULL': 'cell_region_looms/BIO_ddseq_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'BIO_ddseq_2.FULL': 'cell_region_looms/BIO_ddseq_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'BIO_ddseq_3.FULL': 'cell_region_looms/BIO_ddseq_3.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'BIO_ddseq_4.FULL': 'cell_region_looms/BIO_ddseq_4.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'BRO_mtscatac_1.FULL': 'cell_region_looms/BRO_mtscatac_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'BRO_mtscatac_2.FULL': 'cell_region_looms/BRO_mtscatac_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xmultiome_1.FULL': 'cell_region_looms/CNA_10xmultiome_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xmultiome_2.FULL': 'cell_region_looms/CNA_10xmultiome_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xv11_1.FULL': 'cell_region_looms/CNA_10xv11_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xv11_2.FULL': 'cell_region_looms/CNA_10xv11_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xv11_3.FULL': 'cell_region_looms/CNA_10xv11_3.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xv11_4.FULL': 'cell_region_looms/CNA_10xv11_4.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xv11_5.FULL': 'cell_region_looms/CNA_10xv11_5.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xv2_1.FULL': 'cell_region_looms/CNA_10xv2_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xv2_2.FULL': 'cell_region_looms/CNA_10xv2_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_hydrop_1.FULL': 'cell_region_looms/CNA_hydrop_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_hydrop_2.FULL': 'cell_region_looms/CNA_hydrop_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_hydrop_3.FULL': 'cell_region_looms/CNA_hydrop_3.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_mtscatac_1.FULL': 'cell_region_looms/CNA_mtscatac_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_mtscatac_2.FULL': 'cell_region_looms/CNA_mtscatac_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'EPF_hydrop_1.FULL': 'cell_region_looms/EPF_hydrop_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'EPF_hydrop_2.FULL': 'cell_region_looms/EPF_hydrop_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'EPF_hydrop_3.FULL': 'cell_region_looms/EPF_hydrop_3.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'EPF_hydrop_4.FULL': 'cell_region_looms/EPF_hydrop_4.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'HAR_ddseq_1.FULL': 'cell_region_looms/HAR_ddseq_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'HAR_ddseq_2.FULL': 'cell_region_looms/HAR_ddseq_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'MDC_mtscatac_1.FULL': 'cell_region_looms/MDC_mtscatac_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'MDC_mtscatac_2.FULL': 'cell_region_looms/MDC_mtscatac_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'OHS_s3atac_1.FULL': 'cell_region_looms/OHS_s3atac_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'OHS_s3atac_2.FULL': 'cell_region_looms/OHS_s3atac_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'SAN_10xmultiome_1.FULL': 'cell_region_looms/SAN_10xmultiome_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'SAN_10xmultiome_2.FULL': 'cell_region_looms/SAN_10xmultiome_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'STA_10xv11_1.FULL': 'cell_region_looms/STA_10xv11_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'STA_10xv11_2.FULL': 'cell_region_looms/STA_10xv11_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'TXG_10xv11_1.FULL': 'cell_region_looms/TXG_10xv11_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'TXG_10xv2_1.FULL': 'cell_region_looms/TXG_10xv2_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'TXG_10xv2_2.FULL': 'cell_region_looms/TXG_10xv2_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'UCS_ddseq_1.FULL': 'cell_region_looms/UCS_ddseq_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'UCS_ddseq_2.FULL': 'cell_region_looms/UCS_ddseq_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_10xmultiome_1.FULL': 'cell_region_looms/VIB_10xmultiome_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_10xmultiome_2.FULL': 'cell_region_looms/VIB_10xmultiome_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_10xv1_1.FULL': 'cell_region_looms/VIB_10xv1_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_10xv1_2.FULL': 'cell_region_looms/VIB_10xv1_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_10xv2_1.FULL': 'cell_region_looms/VIB_10xv2_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_10xv2_2.FULL': 'cell_region_looms/VIB_10xv2_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_hydrop_1.FULL': 'cell_region_looms/VIB_hydrop_1.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_hydrop_11.FULL': 'cell_region_looms/VIB_hydrop_11.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_hydrop_12.FULL': 'cell_region_looms/VIB_hydrop_12.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_hydrop_2.FULL': 'cell_region_looms/VIB_hydrop_2.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_hydrop_21.FULL': 'cell_region_looms/VIB_hydrop_21.FULL__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_hydrop_22.FULL': 'cell_region_looms/VIB_hydrop_22.FULL__cto.scrublet0-4.fmx.singlets.loom'}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scrub_name_suffix = \"0-4\"\n",
    "loom_path_dict  = {x.split('/')[-1].split(f'__')[0]:x for x in sorted(glob.glob(f'cell_region_looms/*singlets.loom'))}\n",
    "loom_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/usr/bin/env Rscript\n",
      "library(Seurat)\n",
      "library(SeuratDisk)\n",
      "library(Signac)\n",
      "# library(EnsDb.Hsapiens.v86)\n",
      "library(ggplot2)\n",
      "library(stringr)\n",
      "\n",
      "args = commandArgs(trailingOnly=TRUE)\n",
      "sample_id = args[1]\n",
      "f_loom = args[2]\n",
      "f_frag = args[3]\n",
      "f_reference = args[4]\n",
      "f_annotation = args[5]\n",
      "f_out = args[6]\n",
      "\n",
      "print(paste0(\"Processing sample \", args[1]))\n",
      "\n",
      "# load pbmc object\n",
      "pbmc.rna <- readRDS(\"../0_resources/seurat_references/pbmc_integrated.RDS\")\n",
      "#pbmc.rna <- readRDS('/lustre1/project/stg_00090/scatac_benchmark/0_resources/seurat_references/pbmc_ssc_mat__integrated.rds')\n",
      "\n",
      "################################################################################\n",
      "# ATAC\n",
      "################################################################################\n",
      "\n",
      "### get data from loom:\n",
      "atacloomcon <- Connect(filename = f_loom, mode = \"r\")\n",
      "atacloomcon\n",
      "atac_tmp <- as.Seurat(atacloomcon, assay='ATAC')\n",
      "atacloomcon$close_all()\n",
      "\n",
      "# subset by removing contig chromosomes\n",
      "rawregions = rownames(GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC'))\n",
      "standard_regions = rawregions[!grepl('random|KI|Un', rawregions)]\n",
      "\n",
      "atac_tmp_data_subset = GetAssayData(atac_tmp[standard_regions], slot = \"counts\", assay='ATAC')\n",
      "atac_tmp <- CreateSeuratObject(atac_tmp_data_subset, assay='ATAC')\n",
      "\n",
      "\n",
      "# subset by removing contig chromosomes\n",
      "rawregions = rownames(GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC'))\n",
      "# standard_regions = rawregions[!grepl('random|KIe\n",
      "pbmc.rna <- readRDS('../0_resources/seurat_references/pbmc_ssc_mat__integrated.rds')\n",
      "\n",
      "\n",
      "\n",
      "# correctly parse regions (default delims are '-','-')\n",
      "regions = StringToGRanges(\n",
      "    rownames(GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC')),\n",
      "    sep=c(':','-')\n",
      "    )\n",
      "\n",
      "# create chromatin assay\n",
      "chromatinassay = CreateChromatinAssay(\n",
      "    counts=GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC'),\n",
      "    # genome='hg38',\n",
      "    fragments = f_frag,\n",
      "    ranges=regions\n",
      "    )\n",
      "    #annotation=annotation)\n",
      "\n",
      "atac <- CreateSeuratObject(counts = chromatinassay, assay='ATAC')\n",
      "\n",
      "annotations <- readRDS(f_annotation)\n",
      "Annotation(atac) <- annotations\n",
      "\n",
      "# We exclude the first dimension as this is typically correlated with sequencing depth\n",
      "atac <- RunTFIDF(atac)\n",
      "atac <- FindTopFeatures(atac, min.cutoff = \"q0\")\n",
      "atac <- RunSVD(atac)\n",
      "atac <- RunUMAP(atac, reduction = \"lsi\", dims = 2:30, reduction.name = \"umap.atac\", reduction.key = \"atacUMAP_\")\n",
      "\n",
      "##################################################\n",
      "# Identify RNA-ATAC anchors\n",
      "\n",
      "# quantify gene activity\n",
      "gene.activities <- GeneActivity(atac, features = VariableFeatures(pbmc.rna))\n",
      "# gene.activities <- gene.activities[rownames(gene.activities)!=\"\",]\n",
      "\n",
      "# add gene activities as a new assay\n",
      "atac[[\"ACTIVITY\"]] <- CreateAssayObject(counts = gene.activities)\n",
      "\n",
      "# normalize gene activities\n",
      "DefaultAssay(atac) <- \"ACTIVITY\"\n",
      "atac <- NormalizeData(atac)\n",
      "atac <- ScaleData(atac, features = rownames(atac))\n",
      "\n",
      "# Identify anchors\n",
      "transfer.anchors <- FindTransferAnchors(reference = pbmc.rna,\n",
      "    query = atac,\n",
      "    features = VariableFeatures(object = pbmc.rna),\n",
      "    reference.assay = \"RNA\",\n",
      "    query.assay = \"ACTIVITY\",\n",
      "    reduction = \"cca\")\n",
      "\n",
      "# predict celltype\n",
      "celltype.predictions <- TransferData(\n",
      "    anchorset = transfer.anchors,\n",
      "    refdata = pbmc.rna$CellType,\n",
      "    weight.reduction = atac[[\"lsi\"]],\n",
      "    dims = 2:30)\n",
      "\n",
      "pbmc.atac <- AddMetaData(atac, metadata = celltype.predictions)\n",
      "\n",
      "md = celltype.predictions\n",
      "pred_thr = 0.7\n",
      "\n",
      "tmp = data.frame(\n",
      "      composite_sample_id = paste0(rownames(md),'-',sample_id),\n",
      "      barcode = rownames(md),\n",
      "      sample_id = sample_id,\n",
      "      cell_type = md$predicted.id,\n",
      "      cell_type_pred_score = md$prediction.score.max\n",
      "      )\n",
      "\n",
      "tmp$cell_type_hiconf_70 = tmp$cell_type\n",
      "tmp$cell_type_hiconf_70[tmp$cell_type_pred_score<pred_thr] = 'Unknown'\n",
      "\n",
      "table(tmp$cell_type)\n",
      "table(tmp$cell_type_hiconf_70)\n",
      "\n",
      "write.table(tmp, file=f_out, sep='\\t', row.names=FALSE, quote=FALSE)\n",
      "print(paste0('Wrote file ', f_out))\n"
     ]
    }
   ],
   "source": [
    "!cat ../0_resources/scripts/seurat_label_transfer_consensus.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#cell_type_classification/BIO_ddseq_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_3.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_4.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BRO_mtscatac_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BRO_mtscatac_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xmultiome_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xmultiome_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_3.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_4.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_5.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv2_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv2_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_3.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_mtscatac_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_mtscatac_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_3.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_4.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/HAR_ddseq_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/HAR_ddseq_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/MDC_mtscatac_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/MDC_mtscatac_2.FULL__cell_type_seurat.txt already exists!\n",
      "img_path=../0_resources/vsn_cache/cflerin-seurat-4.0.3-plus.sif && singularity exec --cleanenv -H $PWD -B /dodrio,/readonly/dodrio,/tmp $img_path Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R OHS_s3atac_1.FULL cell_region_looms/OHS_s3atac_1.FULL__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/full_fragments/OHS_s3atac_1.FULL.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/OHS_s3atac_1.FULL__cell_type_seurat.txt\n",
      "img_path=../0_resources/vsn_cache/cflerin-seurat-4.0.3-plus.sif && singularity exec --cleanenv -H $PWD -B /dodrio,/readonly/dodrio,/tmp $img_path Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R OHS_s3atac_2.FULL cell_region_looms/OHS_s3atac_2.FULL__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/full_fragments/OHS_s3atac_2.FULL.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/OHS_s3atac_2.FULL__cell_type_seurat.txt\n",
      "#cell_type_classification/SAN_10xmultiome_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/SAN_10xmultiome_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/STA_10xv11_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/STA_10xv11_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/TXG_10xv11_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/TXG_10xv2_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/TXG_10xv2_2.FULL__cell_type_seurat.txt already exists!\n",
      "img_path=../0_resources/vsn_cache/cflerin-seurat-4.0.3-plus.sif && singularity exec --cleanenv -H $PWD -B /dodrio,/readonly/dodrio,/tmp $img_path Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R UCS_ddseq_1.FULL cell_region_looms/UCS_ddseq_1.FULL__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/full_fragments/UCS_ddseq_1.FULL.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/UCS_ddseq_1.FULL__cell_type_seurat.txt\n",
      "img_path=../0_resources/vsn_cache/cflerin-seurat-4.0.3-plus.sif && singularity exec --cleanenv -H $PWD -B /dodrio,/readonly/dodrio,/tmp $img_path Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R UCS_ddseq_2.FULL cell_region_looms/UCS_ddseq_2.FULL__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/full_fragments/UCS_ddseq_2.FULL.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/UCS_ddseq_2.FULL__cell_type_seurat.txt\n",
      "#cell_type_classification/VIB_10xmultiome_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xmultiome_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv1_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv1_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv2_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv2_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_11.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_12.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_21.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_22.FULL__cell_type_seurat.txt already exists!\n"
     ]
    }
   ],
   "source": [
    "parallel_filename = \"seurat_label_transfer.parallel\"\n",
    "script_path = \"../0_resources/scripts/seurat_label_transfer_consensus.R\"\n",
    "img_path = \"../0_resources/vsn_cache/cflerin-seurat-4.0.3-plus.sif\"\n",
    "reference_path = \"../0_resources/seurat_references/pbmc_ref.rds\"\n",
    "annotation_path = \"../0_resources/seurat_references/granges_annotation.rds\"\n",
    "bind_mounts = \"/dodrio,/readonly/dodrio,/tmp\"\n",
    "#bind_mounts = \"/lustre1,/staging,${VSC_SCRATCH}/tmp:/tmp\"\n",
    "\n",
    "with open(parallel_filename, 'w') as f:\n",
    "    for sample in loom_path_dict.keys():\n",
    "        outfile = f\"cell_type_classification/{sample}__cell_type_seurat.txt\"\n",
    "        if not os.path.exists(outfile):\n",
    "            loomfile = loom_path_dict[sample]\n",
    "            fragfile = frags_path_dict[sample]\n",
    "            command=f\"img_path={img_path} && singularity exec --cleanenv -H $PWD -B {bind_mounts} $img_path Rscript {script_path} {sample} {loomfile} {fragfile} {reference_path} {annotation_path} {outfile}\"\n",
    "            f.write(f\"{command}\\n\")\n",
    "            print(command)\n",
    "\n",
    "        else:\n",
    "            f.write(f\"#{outfile} already exists!\\n\")\n",
    "            print(f\"#{outfile} already exists!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#cell_type_classification/BIO_ddseq_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_3.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_4.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BRO_mtscatac_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BRO_mtscatac_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xmultiome_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xmultiome_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_3.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_4.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_5.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv2_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv2_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_3.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_mtscatac_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_mtscatac_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_3.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_4.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/HAR_ddseq_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/HAR_ddseq_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/MDC_mtscatac_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/MDC_mtscatac_2.FULL__cell_type_seurat.txt already exists!\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R OHS_s3atac_1.FULL cell_region_looms/OHS_s3atac_1.FULL__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/full_fragments/OHS_s3atac_1.FULL.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/OHS_s3atac_1.FULL__cell_type_seurat.txt\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R OHS_s3atac_2.FULL cell_region_looms/OHS_s3atac_2.FULL__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/full_fragments/OHS_s3atac_2.FULL.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/OHS_s3atac_2.FULL__cell_type_seurat.txt\n",
      "#cell_type_classification/SAN_10xmultiome_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/SAN_10xmultiome_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/STA_10xv11_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/STA_10xv11_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/TXG_10xv11_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/TXG_10xv2_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/TXG_10xv2_2.FULL__cell_type_seurat.txt already exists!\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R UCS_ddseq_1.FULL cell_region_looms/UCS_ddseq_1.FULL__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/full_fragments/UCS_ddseq_1.FULL.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/UCS_ddseq_1.FULL__cell_type_seurat.txt\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R UCS_ddseq_2.FULL cell_region_looms/UCS_ddseq_2.FULL__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/full_fragments/UCS_ddseq_2.FULL.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/UCS_ddseq_2.FULL__cell_type_seurat.txt\n",
      "#cell_type_classification/VIB_10xmultiome_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xmultiome_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv1_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv1_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv2_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv2_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_11.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_12.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_21.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_22.FULL__cell_type_seurat.txt already exists!\n"
     ]
    }
   ],
   "source": [
    "parallel_filename = \"seurat_label_transfer.parallel\"\n",
    "script_path = \"../0_resources/scripts/seurat_label_transfer_consensus.R\"\n",
    "img_path = \"../0_resources/vsn_cache/cflerin-seurat-4.0.3-plus.sif\"\n",
    "reference_path = \"../0_resources/seurat_references/pbmc_ref.rds\"\n",
    "annotation_path = \"../0_resources/seurat_references/granges_annotation.rds\"\n",
    "bind_mounts = \"/dodrio,/readonly/dodrio,/tmp\"\n",
    "#bind_mounts = \"/lustre1,/staging,${VSC_SCRATCH}/tmp:/tmp\"\n",
    "\n",
    "with open(parallel_filename, 'w') as f:\n",
    "    for sample in loom_path_dict.keys():\n",
    "        outfile = f\"cell_type_classification/{sample}__cell_type_seurat.txt\"\n",
    "        if not os.path.exists(outfile):\n",
    "            loomfile = loom_path_dict[sample]\n",
    "            fragfile = frags_path_dict[sample]\n",
    "            command=f\"Rscript {script_path} {sample} {loomfile} {fragfile} {reference_path} {annotation_path} {outfile}\"\n",
    "            f.write(f\"{command}\\n\")\n",
    "            print(command)\n",
    "\n",
    "        else:\n",
    "            f.write(f\"#{outfile} already exists!\\n\")\n",
    "            print(f\"#{outfile} already exists!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#cell_type_classification/BIO_ddseq_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_3.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_4.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BRO_mtscatac_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BRO_mtscatac_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xmultiome_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xmultiome_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_3.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_4.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_5.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv2_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv2_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_3.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_mtscatac_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_mtscatac_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_3.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_4.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/HAR_ddseq_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/HAR_ddseq_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/MDC_mtscatac_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/MDC_mtscatac_2.FULL__cell_type_seurat.txt already exists!\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R OHS_s3atac_1.FULL cell_region_looms/OHS_s3atac_1.FULL__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/full_fragments/OHS_s3atac_1.FULL.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/OHS_s3atac_1.FULL__cell_type_seurat.txt\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R OHS_s3atac_2.FULL cell_region_looms/OHS_s3atac_2.FULL__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/full_fragments/OHS_s3atac_2.FULL.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/OHS_s3atac_2.FULL__cell_type_seurat.txt\n",
      "#cell_type_classification/SAN_10xmultiome_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/SAN_10xmultiome_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/STA_10xv11_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/STA_10xv11_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/TXG_10xv11_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/TXG_10xv2_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/TXG_10xv2_2.FULL__cell_type_seurat.txt already exists!\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R UCS_ddseq_1.FULL cell_region_looms/UCS_ddseq_1.FULL__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/full_fragments/UCS_ddseq_1.FULL.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/UCS_ddseq_1.FULL__cell_type_seurat.txt\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R UCS_ddseq_2.FULL cell_region_looms/UCS_ddseq_2.FULL__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/full_fragments/UCS_ddseq_2.FULL.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/UCS_ddseq_2.FULL__cell_type_seurat.txt\n",
      "#cell_type_classification/VIB_10xmultiome_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xmultiome_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv1_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv1_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv2_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv2_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_1.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_11.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_12.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_2.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_21.FULL__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_hydrop_22.FULL__cell_type_seurat.txt already exists!\n"
     ]
    }
   ],
   "source": [
    "parallel_filename = \"seurat_label_transfer.parallel\"\n",
    "script_path = \"../0_resources/scripts/seurat_label_transfer_consensus.R\"\n",
    "img_path = \"../0_resources/vsn_cache/cflerin-seurat-4.0.3-plus.sif\"\n",
    "reference_path = \"../0_resources/seurat_references/pbmc_ref.rds\"\n",
    "annotation_path = \"../0_resources/seurat_references/granges_annotation.rds\"\n",
    "bind_mounts = \"/dodrio,/readonly/dodrio,/tmp\"\n",
    "#bind_mounts = \"/lustre1,/staging,${VSC_SCRATCH}/tmp:/tmp\"\n",
    "\n",
    "with open(parallel_filename, 'w') as f:\n",
    "    for sample in loom_path_dict.keys():\n",
    "        outfile = f\"cell_type_classification/{sample}__cell_type_seurat.txt\"\n",
    "        if not os.path.exists(outfile):\n",
    "            loomfile = loom_path_dict[sample]\n",
    "            fragfile = frags_path_dict[sample]\n",
    "            command=f\"Rscript {script_path} {sample} {loomfile} {fragfile} {reference_path} {annotation_path} {outfile}\"\n",
    "            f.write(f\"{command}\\n\")\n",
    "            print(command)\n",
    "\n",
    "        else:\n",
    "            f.write(f\"#{outfile} already exists!\\n\")\n",
    "            print(f\"#{outfile} already exists!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run the Rscript"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "mkdir cell_type_classification\n",
    "export OMP_THREAD_LIMIT=80\n",
    "cat seurat_label_transfer.parallel | parallel -j 2 --progress\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Troubleshooting"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "when running\n",
    "```\n",
    "> chromatinassay = CreateChromatinAssay(\n",
    "+     counts=GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC'),\n",
    "+     genome='hg38',\n",
    "+     fragments = f_frag,\n",
    "+     ranges=regions\n",
    "+     )\n",
    "```\n",
    "i get the error\n",
    "```\n",
    "Computing hash\n",
    "Checking for 2208 cell barcodes\n",
    "Error in CreateFragmentObject(path = fragments, cells = cells, validate.fragments = validate.fragments,  :\n",
    "  Not all cells requested could be found in the fragment file.\n",
    "In addition: Warning message:\n",
    "In CreateChromatinAssay(counts = GetAssayData(atac_tmp, slot = \"counts\",  :\n",
    "  Overlapping ranges supplied. Ranges should be non-overlapping.\n",
    "```\n",
    "and `chromatinassay` is empty. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t10151\t10183\tGCGTTGACTTCGTCGACGTA-12\t1\n",
      "chr1\t16228\t16272\tATTGTAGAGGTAGGAGTCAA-12\t6\n",
      "chr1\t17490\t17527\tTTGTACTGCAAGGTGCCAGA-12\t3\n",
      "chr1\t30857\t31030\tCATGAACGTTGTGACGTGGA-12\t1\n",
      "chr1\t68280\t68311\tCATTGGTGCAGGCGAGATTC-12\t1\n",
      "chr1\t74871\t74900\tATCATCCGTTAATAGGCAGG-11\t11\n",
      "chr1\t79643\t79855\tGGAGCGGATTCAAGGTCGAT-11\t2\n",
      "chr1\t88228\t88302\tAAGACCAAGCGCTGATGGTA-12\t4\n",
      "chr1\t89893\t89944\tTCCGATTCAGCACGTTAAGA-12\t3\n",
      "chr1\t91177\t91237\tCTCCTCACTACTTGAAGAAG-12\t4\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "zcat fragments_postbap/VIB_Hydrop_1.sinto.mm.fragments.tsv.gz | head"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "this error doesn't make sense, because the barcodes are present in the fragments file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "the output of the above code can be read in the cell_type_redictions.ipynb notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "it turns out the loom file writing had a bug in it that did not add the `-1?` do the barcode. i fixed this in the loom writing."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "now that this is fixed, i got an error\n",
    "\n",
    "```\n",
    "> chromatinassay = CreateChromatinAssay(\n",
    "+     counts=GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC'),\n",
    "+     genome='hg38',\n",
    "+     fragments = f_frag,\n",
    "+     ranges=regions\n",
    "+     )\n",
    "Computing hash\n",
    "Checking for 2208 cell barcodes\n",
    "Error in .order_seqlevels(chrom_sizes[, \"chrom\"]) :\n",
    "  !anyNA(m32) is not TRUE\n",
    "In addition: Warning message:\n",
    "In CreateChromatinAssay(counts = GetAssayData(atac_tmp, slot = \"counts\",  :\n",
    "  Overlapping ranges supplied. Ranges should be non-overlapping.\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "it turns out that this is a known bug (https://github.com/timoast/signac/issues/687) and that commenting out the line `genome='hg38'`"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "20220628_pycistopic.sif",
   "language": "python",
   "name": "cistopic_20220628"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
