{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d72f1205",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:11:45.074724Z",
     "iopub.status.busy": "2022-09-26T14:11:45.074223Z",
     "iopub.status.idle": "2022-09-26T14:11:54.791082Z",
     "shell.execute_reply": "2022-09-26T14:11:54.790367Z"
    },
    "papermill": {
     "duration": 10.067836,
     "end_time": "2022-09-26T14:11:55.136300",
     "exception": false,
     "start_time": "2022-09-26T14:11:45.068464",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pycisTopic.pseudobulk_peak_calling import export_pseudobulk, peak_calling\n",
    "import pyranges as pr\n",
    "import requests\n",
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e38440ab",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:11:55.339809Z",
     "iopub.status.busy": "2022-09-26T14:11:55.339131Z",
     "iopub.status.idle": "2022-09-26T14:11:55.597245Z",
     "shell.execute_reply": "2022-09-26T14:11:55.596570Z"
    },
    "papermill": {
     "duration": 0.278507,
     "end_time": "2022-09-26T14:11:55.598465",
     "exception": false,
     "start_time": "2022-09-26T14:11:55.319958",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%load_ext lab_black"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "938a7dbe",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:11:55.608089Z",
     "iopub.status.busy": "2022-09-26T14:11:55.607756Z",
     "iopub.status.idle": "2022-09-26T14:11:55.628630Z",
     "shell.execute_reply": "2022-09-26T14:11:55.628081Z"
    },
    "papermill": {
     "duration": 0.026463,
     "end_time": "2022-09-26T14:11:55.629633",
     "exception": false,
     "start_time": "2022-09-26T14:11:55.603170",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# get chromosome sizes (hg38)\n",
    "if not os.path.exists(\"chromsizes.txt\"):\n",
    "    target_url = (\n",
    "        \"http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes\"\n",
    "    )\n",
    "    chromsizes = pd.read_csv(target_url, sep=\"\\t\", header=None)\n",
    "    chromsizes.columns = [\"Chromosome\", \"End\"]\n",
    "    chromsizes[\"Start\"] = [0] * chromsizes.shape[0]\n",
    "    chromsizes = chromsizes.loc[:, [\"Chromosome\", \"Start\", \"End\"]]\n",
    "    chromsizes = pr.PyRanges(chromsizes)\n",
    "    chromsizes.to_csv(\"chromsizes.txt\")\n",
    "    chromsizes\n",
    "else:\n",
    "    chromsizes = pd.read_csv(\"chromsizes.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6dd8a2fc",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:11:55.638268Z",
     "iopub.status.busy": "2022-09-26T14:11:55.637940Z",
     "iopub.status.idle": "2022-09-26T14:11:55.652841Z",
     "shell.execute_reply": "2022-09-26T14:11:55.652326Z"
    },
    "papermill": {
     "duration": 0.082178,
     "end_time": "2022-09-26T14:11:55.715759",
     "exception": false,
     "start_time": "2022-09-26T14:11:55.633581",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FULL': 'cistopic_objects/BIO_ddseq_1.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'BIO_ddseq_2.FULL': 'cistopic_objects/BIO_ddseq_2.FULL__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'BIO_ddseq_3.FULL': 'cistopic_objects/BIO_ddseq_3.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'BIO_ddseq_4.FULL': 'cistopic_objects/BIO_ddseq_4.FULL__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'BRO_mtscatac_1.FULL': 'cistopic_objects/BRO_mtscatac_1.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'BRO_mtscatac_2.FULL': 'cistopic_objects/BRO_mtscatac_2.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xmultiome_1.FULL': 'cistopic_objects/CNA_10xmultiome_1.FULL__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xmultiome_2.FULL': 'cistopic_objects/CNA_10xmultiome_2.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv11_1.FULL': 'cistopic_objects/CNA_10xv11_1.FULL__cto.scrublet0-4.fmx.singlets.model_8topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv11_2.FULL': 'cistopic_objects/CNA_10xv11_2.FULL__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv11_3.FULL': 'cistopic_objects/CNA_10xv11_3.FULL__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv11_4.FULL': 'cistopic_objects/CNA_10xv11_4.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv11_5.FULL': 'cistopic_objects/CNA_10xv11_5.FULL__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv2_1.FULL': 'cistopic_objects/CNA_10xv2_1.FULL__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv2_2.FULL': 'cistopic_objects/CNA_10xv2_2.FULL__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'CNA_hydrop_1.FULL': 'cistopic_objects/CNA_hydrop_1.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_hydrop_2.FULL': 'cistopic_objects/CNA_hydrop_2.FULL__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'CNA_hydrop_3.FULL': 'cistopic_objects/CNA_hydrop_3.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_mtscatac_1.FULL': 'cistopic_objects/CNA_mtscatac_1.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'CNA_mtscatac_2.FULL': 'cistopic_objects/CNA_mtscatac_2.FULL__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl',\n",
       " 'EPF_hydrop_1.FULL': 'cistopic_objects/EPF_hydrop_1.FULL__cto.scrublet0-4.fmx.singlets.model_8topics.dimreduc.consensus.pkl',\n",
       " 'EPF_hydrop_2.FULL': 'cistopic_objects/EPF_hydrop_2.FULL__cto.scrublet0-4.fmx.singlets.model_8topics.dimreduc.consensus.pkl',\n",
       " 'EPF_hydrop_3.FULL': 'cistopic_objects/EPF_hydrop_3.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'EPF_hydrop_4.FULL': 'cistopic_objects/EPF_hydrop_4.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'HAR_ddseq_1.FULL': 'cistopic_objects/HAR_ddseq_1.FULL__cto.scrublet0-4.fmx.singlets.model_8topics.dimreduc.consensus.pkl',\n",
       " 'HAR_ddseq_2.FULL': 'cistopic_objects/HAR_ddseq_2.FULL__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl',\n",
       " 'MDC_mtscatac_1.FULL': 'cistopic_objects/MDC_mtscatac_1.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'MDC_mtscatac_2.FULL': 'cistopic_objects/MDC_mtscatac_2.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'OHS_s3atac_1.FULL': 'cistopic_objects/OHS_s3atac_1.FULL__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'OHS_s3atac_2.FULL': 'cistopic_objects/OHS_s3atac_2.FULL__cto.scrublet0-4.fmx.singlets.model_7topics.dimreduc.consensus.pkl',\n",
       " 'SAN_10xmultiome_1.FULL': 'cistopic_objects/SAN_10xmultiome_1.FULL__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl',\n",
       " 'SAN_10xmultiome_2.FULL': 'cistopic_objects/SAN_10xmultiome_2.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'STA_10xv11_1.FULL': 'cistopic_objects/STA_10xv11_1.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'STA_10xv11_2.FULL': 'cistopic_objects/STA_10xv11_2.FULL__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl',\n",
       " 'TXG_10xv11_1.FULL': 'cistopic_objects/TXG_10xv11_1.FULL__cto.scrublet0-4.fmx.singlets.model_16topics.dimreduc.consensus.pkl',\n",
       " 'TXG_10xv2_1.FULL': 'cistopic_objects/TXG_10xv2_1.FULL__cto.scrublet0-4.fmx.singlets.model_18topics.dimreduc.consensus.pkl',\n",
       " 'TXG_10xv2_2.FULL': 'cistopic_objects/TXG_10xv2_2.FULL__cto.scrublet0-4.fmx.singlets.model_16topics.dimreduc.consensus.pkl',\n",
       " 'UCS_ddseq_1.FULL': 'cistopic_objects/UCS_ddseq_1.FULL__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl',\n",
       " 'UCS_ddseq_2.FULL': 'cistopic_objects/UCS_ddseq_2.FULL__cto.scrublet0-4.fmx.singlets.model_17topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xmultiome_1.FULL': 'cistopic_objects/VIB_10xmultiome_1.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xmultiome_2.FULL': 'cistopic_objects/VIB_10xmultiome_2.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xv1_1.FULL': 'cistopic_objects/VIB_10xv1_1.FULL__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xv1_2.FULL': 'cistopic_objects/VIB_10xv1_2.FULL__cto.scrublet0-4.fmx.singlets.model_8topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xv2_1.FULL': 'cistopic_objects/VIB_10xv2_1.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xv2_2.FULL': 'cistopic_objects/VIB_10xv2_2.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'VIB_hydrop_1.FULL': 'cistopic_objects/VIB_hydrop_1.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'VIB_hydrop_11.FULL': 'cistopic_objects/VIB_hydrop_11.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'VIB_hydrop_12.FULL': 'cistopic_objects/VIB_hydrop_12.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'VIB_hydrop_2.FULL': 'cistopic_objects/VIB_hydrop_2.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'VIB_hydrop_21.FULL': 'cistopic_objects/VIB_hydrop_21.FULL__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'VIB_hydrop_22.FULL': 'cistopic_objects/VIB_hydrop_22.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cto_consensus_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"cistopic_objects/*singlets.model*consensus.pkl\"))\n",
    "}\n",
    "cto_consensus_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "770b87a2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:11:55.724829Z",
     "iopub.status.busy": "2022-09-26T14:11:55.724652Z",
     "iopub.status.idle": "2022-09-26T14:11:55.732504Z",
     "shell.execute_reply": "2022-09-26T14:11:55.731993Z"
    },
    "papermill": {
     "duration": 0.013276,
     "end_time": "2022-09-26T14:11:55.733424",
     "exception": false,
     "start_time": "2022-09-26T14:11:55.720148",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FULL': '../1_data_repository/full_fragments/BIO_ddseq_1.FULL.fragments.tsv.gz',\n",
       " 'BIO_ddseq_2.FULL': '../1_data_repository/full_fragments/BIO_ddseq_2.FULL.fragments.tsv.gz',\n",
       " 'BIO_ddseq_3.FULL': '../1_data_repository/full_fragments/BIO_ddseq_3.FULL.fragments.tsv.gz',\n",
       " 'BIO_ddseq_4.FULL': '../1_data_repository/full_fragments/BIO_ddseq_4.FULL.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_1.FULL': '../1_data_repository/full_fragments/BRO_mtscatac_1.FULL.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_2.FULL': '../1_data_repository/full_fragments/BRO_mtscatac_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_1.FULL': '../1_data_repository/full_fragments/CNA_10xmultiome_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_2.FULL': '../1_data_repository/full_fragments/CNA_10xmultiome_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_1.FULL': '../1_data_repository/full_fragments/CNA_10xv11_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_2.FULL': '../1_data_repository/full_fragments/CNA_10xv11_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_3.FULL': '../1_data_repository/full_fragments/CNA_10xv11_3.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_4.FULL': '../1_data_repository/full_fragments/CNA_10xv11_4.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv11_5.FULL': '../1_data_repository/full_fragments/CNA_10xv11_5.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv2_1.FULL': '../1_data_repository/full_fragments/CNA_10xv2_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_10xv2_2.FULL': '../1_data_repository/full_fragments/CNA_10xv2_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_hydrop_1.FULL': '../1_data_repository/full_fragments/CNA_hydrop_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_hydrop_2.FULL': '../1_data_repository/full_fragments/CNA_hydrop_2.FULL.fragments.tsv.gz',\n",
       " 'CNA_hydrop_3.FULL': '../1_data_repository/full_fragments/CNA_hydrop_3.FULL.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_1.FULL': '../1_data_repository/full_fragments/CNA_mtscatac_1.FULL.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_2.FULL': '../1_data_repository/full_fragments/CNA_mtscatac_2.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_1.FULL': '../1_data_repository/full_fragments/EPF_hydrop_1.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_2.FULL': '../1_data_repository/full_fragments/EPF_hydrop_2.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_3.FULL': '../1_data_repository/full_fragments/EPF_hydrop_3.FULL.fragments.tsv.gz',\n",
       " 'EPF_hydrop_4.FULL': '../1_data_repository/full_fragments/EPF_hydrop_4.FULL.fragments.tsv.gz',\n",
       " 'HAR_ddseq_1.FULL': '../1_data_repository/full_fragments/HAR_ddseq_1.FULL.fragments.tsv.gz',\n",
       " 'HAR_ddseq_2.FULL': '../1_data_repository/full_fragments/HAR_ddseq_2.FULL.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_1.FULL': '../1_data_repository/full_fragments/MDC_mtscatac_1.FULL.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_2.FULL': '../1_data_repository/full_fragments/MDC_mtscatac_2.FULL.fragments.tsv.gz',\n",
       " 'OHS_s3atac_1.FULL': '../1_data_repository/full_fragments/OHS_s3atac_1.FULL.fragments.tsv.gz',\n",
       " 'OHS_s3atac_2.FULL': '../1_data_repository/full_fragments/OHS_s3atac_2.FULL.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_1.FULL': '../1_data_repository/full_fragments/SAN_10xmultiome_1.FULL.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_2.FULL': '../1_data_repository/full_fragments/SAN_10xmultiome_2.FULL.fragments.tsv.gz',\n",
       " 'STA_10xv11_1.FULL': '../1_data_repository/full_fragments/STA_10xv11_1.FULL.fragments.tsv.gz',\n",
       " 'STA_10xv11_2.FULL': '../1_data_repository/full_fragments/STA_10xv11_2.FULL.fragments.tsv.gz',\n",
       " 'TXG_10xv11_1.FULL': '../1_data_repository/full_fragments/TXG_10xv11_1.FULL.fragments.tsv.gz',\n",
       " 'TXG_10xv2_1.FULL': '../1_data_repository/full_fragments/TXG_10xv2_1.FULL.fragments.tsv.gz',\n",
       " 'TXG_10xv2_2.FULL': '../1_data_repository/full_fragments/TXG_10xv2_2.FULL.fragments.tsv.gz',\n",
       " 'UCS_ddseq_1.FULL': '../1_data_repository/full_fragments/UCS_ddseq_1.FULL.fragments.tsv.gz',\n",
       " 'UCS_ddseq_2.FULL': '../1_data_repository/full_fragments/UCS_ddseq_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_1.FULL': '../1_data_repository/full_fragments/VIB_10xmultiome_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_2.FULL': '../1_data_repository/full_fragments/VIB_10xmultiome_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv1_1.FULL': '../1_data_repository/full_fragments/VIB_10xv1_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv1_2.FULL': '../1_data_repository/full_fragments/VIB_10xv1_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv2_1.FULL': '../1_data_repository/full_fragments/VIB_10xv2_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_10xv2_2.FULL': '../1_data_repository/full_fragments/VIB_10xv2_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_1.FULL': '../1_data_repository/full_fragments/VIB_hydrop_1.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_11.FULL': '../1_data_repository/full_fragments/VIB_hydrop_11.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_12.FULL': '../1_data_repository/full_fragments/VIB_hydrop_12.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_2.FULL': '../1_data_repository/full_fragments/VIB_hydrop_2.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_21.FULL': '../1_data_repository/full_fragments/VIB_hydrop_21.FULL.fragments.tsv.gz',\n",
       " 'VIB_hydrop_22.FULL': '../1_data_repository/full_fragments/VIB_hydrop_22.FULL.fragments.tsv.gz'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fragments_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\".fragments.tsv.gz\")[0]: x\n",
    "    for x in sorted(glob.glob(\"../1_data_repository/full_fragments/*.tsv.gz\"))\n",
    "}\n",
    "fragments_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1832cf0e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:11:55.742915Z",
     "iopub.status.busy": "2022-09-26T14:11:55.742538Z",
     "iopub.status.idle": "2022-09-26T14:11:55.907283Z",
     "shell.execute_reply": "2022-09-26T14:11:55.906495Z"
    },
    "papermill": {
     "duration": 0.173155,
     "end_time": "2022-09-26T14:11:55.910859",
     "exception": false,
     "start_time": "2022-09-26T14:11:55.737704",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import gc\n",
    "import logging\n",
    "import os\n",
    "import re\n",
    "import subprocess\n",
    "import sys\n",
    "from typing import Dict, List, Optional, Union\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyBigWig\n",
    "import pyranges as pr\n",
    "import ray\n",
    "\n",
    "from pycisTopic.cistopic_class import *\n",
    "from pycisTopic.utils import *\n",
    "\n",
    "\n",
    "def export_pseudobulk(\n",
    "    input_data: Union[\"CistopicObject\", pd.DataFrame, Dict[str, pd.DataFrame]],\n",
    "    variable: str,\n",
    "    chromsizes: Union[pd.DataFrame, pr.PyRanges],\n",
    "    bed_path: str,\n",
    "    bigwig_path: str,\n",
    "    path_to_fragments: Optional[Dict[str, str]] = None,\n",
    "    sample_id_col: Optional[str] = \"sample_id\",\n",
    "    n_cpu: Optional[int] = 1,\n",
    "    normalize_bigwig: Optional[bool] = True,\n",
    "    remove_duplicates: Optional[bool] = True,\n",
    "    split_pattern: Optional[str] = \"___\",\n",
    "    use_polars: Optional[bool] = True,\n",
    "    **kwargs\n",
    "):\n",
    "    \"\"\"\n",
    "    Create pseudobulks as bed and bigwig from single cell fragments file given a barcode annotation.\n",
    "    Parameters\n",
    "    ---------\n",
    "    input_data: CistopicObject or pd.DataFrame\n",
    "            A :class:`CistopicObject` containing the specified `variable` as a column in :class:`CistopicObject.cell_data` or a cell metadata\n",
    "            :class:`pd.DataFrame` containing barcode as rows, containing the specified `variable` as a column (additional columns are\n",
    "            possible) and a `sample_id` column. Index names must contain the BARCODE (e.g. ATGTCGTC-1), additional tags are possible separating with -\n",
    "            (e.g. ATGCTGTGCG-1-Sample_1). The levels in the sample_id column must agree with the keys in the path_to_fragments dictionary.\n",
    "            Alternatively, if the cell metadata contains a column named barcode it will be used instead of the index names.\n",
    "    variable: str\n",
    "            A character string indicating the column that will be used to create the different group pseudobulk. It must be included in\n",
    "            the cell metadata provided as input_data.\n",
    "    chromsizes: pd.DataFrame or pr.PyRanges\n",
    "            A data frame or :class:`pr.PyRanges` containing size of each chromosome, containing 'Chromosome', 'Start' and 'End' columns.\n",
    "    bed_path: str\n",
    "            Path to folder where the fragments bed files per group will be saved. If None, files will not be generated.\n",
    "    bigwig_path: str\n",
    "            Path to folder where the bigwig files per group will be saved. If None, files will not be generated.\n",
    "    path_to_fragments: str or dict, optional\n",
    "            A dictionary of character strings, with sample name as names indicating the path to the fragments file/s from which pseudobulk profiles have to\n",
    "            be created. If a :class:`CistopicObject` is provided as input it will be ignored, but if a cell metadata :class:`pd.DataFrame` is provided it\n",
    "            is necessary to provide it. The keys of the dictionary need to match with the sample_id tag added to the index names of the input data frame.\n",
    "    sample_id_col: str, optional\n",
    "            Name of the column containing the sample name per barcode in the input :class:`CistopicObject.cell_data` or class:`pd.DataFrame`. Default: 'sample_id'.\n",
    "    n_cpu: int, optional\n",
    "            Number of cores to use. Default: 1.\n",
    "    normalize_bigwig: bool, optional\n",
    "            Whether bigwig files should be CPM normalized. Default: True.\n",
    "    remove_duplicates: bool, optional\n",
    "            Whether duplicates should be removed before converting the data to bigwig.\n",
    "    split_pattern: str, optional\n",
    "            Pattern to split cell barcode from sample id. Default: ___ .\n",
    "    use_polars: bool, optional\n",
    "            Whether to use polars to read fragments files. Default: True.\n",
    "    **kwargs\n",
    "            Additional parameters for ray.init()\n",
    "    Return\n",
    "    ------\n",
    "    dict\n",
    "            A dictionary containing the paths to the newly created bed fragments files per group a dictionary containing the paths to the\n",
    "            newly created bigwig files per group.\n",
    "    \"\"\"\n",
    "    # Create logger\n",
    "    level = logging.INFO\n",
    "    log_format = \"%(asctime)s %(name)-12s %(levelname)-8s %(message)s\"\n",
    "    handlers = [logging.StreamHandler(stream=sys.stdout)]\n",
    "    logging.basicConfig(level=level, format=log_format, handlers=handlers)\n",
    "    log = logging.getLogger(\"cisTopic\")\n",
    "\n",
    "    # Get fragments file\n",
    "    if isinstance(input_data, CistopicObject):\n",
    "        path_to_fragments = input_data.path_to_fragments\n",
    "        if path_to_fragments is None:\n",
    "            log.error(\"No path_to_fragments in this cisTopic object.\")\n",
    "        cell_data = input_data.cell_data\n",
    "    elif isinstance(input_data, pd.DataFrame):\n",
    "        if path_to_fragments is None:\n",
    "            log.error(\"Please, provide path_to_fragments.\")\n",
    "        cell_data = input_data\n",
    "    # Check for sample_id column\n",
    "    try:\n",
    "        sample_ids = list(set(cell_data[sample_id_col]))\n",
    "    except ValueError:\n",
    "        print(\n",
    "            'Please, include a sample identification column (e.g. \"sample_id\") in your cell metadata!'\n",
    "        )\n",
    "\n",
    "    # Get fragments\n",
    "    fragments_df_dict = {}\n",
    "    for sample_id in path_to_fragments.keys():\n",
    "        if sample_id not in sample_ids:\n",
    "            log.info(\n",
    "                \"The following path_to_fragments entry is not found in the cell metadata sample_id_col: \",\n",
    "                sample_id,\n",
    "                \". It will be ignored.\",\n",
    "            )\n",
    "        else:\n",
    "            log.info(\"Reading fragments from \" + path_to_fragments[sample_id])\n",
    "            fragments_df = read_fragments_from_file(\n",
    "                path_to_fragments[sample_id], use_polars=use_polars\n",
    "            ).df\n",
    "            # Convert to int32 for memory efficiency\n",
    "            fragments_df.Start = np.int32(fragments_df.Start)\n",
    "            fragments_df.End = np.int32(fragments_df.End)\n",
    "            if \"Score\" in fragments_df:\n",
    "                fragments_df.Score = np.int32(fragments_df.Score)\n",
    "            if \"barcode\" in cell_data:\n",
    "                fragments_df = fragments_df.loc[\n",
    "                    fragments_df[\"Name\"].isin(cell_data[\"barcode\"].tolist())\n",
    "                ]\n",
    "            else:\n",
    "                fragments_df = fragments_df.loc[\n",
    "                    fragments_df[\"Name\"].isin(\n",
    "                        prepare_tag_cells(cell_data.index.tolist(), split_pattern)\n",
    "                    )\n",
    "                ]\n",
    "            fragments_df_dict[sample_id] = fragments_df\n",
    "\n",
    "    # Set groups\n",
    "    if \"barcode\" in cell_data:\n",
    "        cell_data = cell_data.loc[:, [variable, sample_id_col, \"barcode\"]]\n",
    "    else:\n",
    "        cell_data = cell_data.loc[:, [variable, sample_id_col]]\n",
    "    cell_data[variable] = cell_data[variable].replace(\" \", \"\", regex=True)\n",
    "    cell_data[variable] = cell_data[variable].replace(\"[^A-Za-z0-9]+\", \"_\", regex=True)\n",
    "    groups = sorted(list(set(cell_data[variable])))\n",
    "    # Check chromosome sizes\n",
    "    if isinstance(chromsizes, pd.DataFrame):\n",
    "        chromsizes = chromsizes.loc[:, [\"Chromosome\", \"Start\", \"End\"]]\n",
    "        chromsizes = pr.PyRanges(chromsizes)\n",
    "    # Check that output dir exist and generate output paths\n",
    "    if isinstance(bed_path, str):\n",
    "        if not os.path.exists(bed_path):\n",
    "            os.makedirs(bed_path)\n",
    "        bed_paths = {\n",
    "            group: os.path.join(bed_path, str(group) + \".bed.gz\") for group in groups\n",
    "        }\n",
    "    else:\n",
    "        bed_paths = {}\n",
    "    if isinstance(bigwig_path, str):\n",
    "        if not os.path.exists(bigwig_path):\n",
    "            os.makedirs(bigwig_path)\n",
    "        bw_paths = {\n",
    "            group: os.path.join(bigwig_path, str(group) + \".bw\") for group in groups\n",
    "        }\n",
    "    else:\n",
    "        bw_paths = {}\n",
    "    # Create pseudobulks\n",
    "    if n_cpu > 1:\n",
    "        ray.init(num_cpus=n_cpu, **kwargs)\n",
    "        ray_handle = ray.wait(\n",
    "            [\n",
    "                export_pseudobulk_ray.remote(\n",
    "                    cell_data,\n",
    "                    group,\n",
    "                    fragments_df_dict,\n",
    "                    chromsizes,\n",
    "                    bigwig_path,\n",
    "                    bed_path,\n",
    "                    sample_id_col,\n",
    "                    normalize_bigwig,\n",
    "                    remove_duplicates,\n",
    "                    split_pattern,\n",
    "                )\n",
    "                for group in groups\n",
    "            ],\n",
    "            num_returns=len(groups),\n",
    "        )\n",
    "        ray.shutdown()\n",
    "    else:\n",
    "        [\n",
    "            export_pseudobulk_one_sample(\n",
    "                cell_data,\n",
    "                group,\n",
    "                fragments_df_dict,\n",
    "                chromsizes,\n",
    "                bigwig_path,\n",
    "                bed_path,\n",
    "                sample_id_col,\n",
    "                normalize_bigwig,\n",
    "                remove_duplicates,\n",
    "                split_pattern,\n",
    "            )\n",
    "            for group in groups\n",
    "        ]\n",
    "\n",
    "    return bw_paths, bed_paths\n",
    "\n",
    "\n",
    "def export_pseudobulk_one_sample(\n",
    "    cell_data: pd.DataFrame,\n",
    "    group: str,\n",
    "    fragments_df_dict: Dict[str, pd.DataFrame],\n",
    "    chromsizes: pr.PyRanges,\n",
    "    bigwig_path: str,\n",
    "    bed_path: str,\n",
    "    sample_id_col: Optional[str] = \"sample_id\",\n",
    "    normalize_bigwig: Optional[bool] = True,\n",
    "    remove_duplicates: Optional[bool] = True,\n",
    "    split_pattern: Optional[str] = \"___\",\n",
    "):\n",
    "    \"\"\"\n",
    "    Create pseudobulk as bed and bigwig from single cell fragments file given a barcode annotation and a group.\n",
    "    Parameters\n",
    "    ---------\n",
    "    cell_data: pd.DataFrame\n",
    "            A cell metadata :class:`pd.Dataframe` containing barcodes, their annotation and their sample of origin.\n",
    "    group: str\n",
    "            A character string indicating the group for which pseudobulks will be created.\n",
    "    fragments_df_dict: dict\n",
    "            A dictionary containing data frames as values with 'Chromosome', 'Start', 'End', 'Name', and 'Score' as columns; and sample label\n",
    "            as keys. 'Score' indicates the number of times that a fragments is found assigned to that barcode.\n",
    "    chromsizes: pr.PyRanges\n",
    "            A :class:`pr.PyRanges` containing size of each column, containing 'Chromosome', 'Start' and 'End' columns.\n",
    "    bigwig_path: str\n",
    "            Path to folder where the bigwig file will be saved.\n",
    "    bed_path: str\n",
    "            Path to folder where the fragments bed file will be saved.\n",
    "    sample_id_col: str, optional\n",
    "            Name of the column containing the sample name per barcode in the input :class:`CistopicObject.cell_data` or class:`pd.DataFrame`. Default: 'sample_id'.\n",
    "    normalize_bigwig: bool, optional\n",
    "            Whether bigwig files should be CPM normalized. Default: True.\n",
    "    remove_duplicates: bool, optional\n",
    "            Whether duplicates should be removed before converting the data to bigwig.\n",
    "    split_pattern: str\n",
    "            Pattern to split cell barcode from sample id. Default: ___ .\n",
    "    \"\"\"\n",
    "    # Create logger\n",
    "    level = logging.INFO\n",
    "    log_format = \"%(asctime)s %(name)-12s %(levelname)-8s %(message)s\"\n",
    "    handlers = [logging.StreamHandler(stream=sys.stdout)]\n",
    "    logging.basicConfig(level=level, format=log_format, handlers=handlers)\n",
    "    log = logging.getLogger(\"cisTopic\")\n",
    "\n",
    "    log.info(\"Creating pseudobulk for \" + str(group))\n",
    "    group_fragments_list = []\n",
    "    group_fragments_dict = {}\n",
    "    for sample_id in fragments_df_dict:\n",
    "        sample_data = cell_data[cell_data.loc[:, sample_id_col].isin([sample_id])]\n",
    "        if \"barcode\" in sample_data:\n",
    "            sample_data.index = sample_data[\"barcode\"].tolist()\n",
    "        else:\n",
    "            sample_data.index = prepare_tag_cells(\n",
    "                sample_data.index.tolist(), split_pattern\n",
    "            )\n",
    "        group_var = sample_data.iloc[:, 0]\n",
    "        barcodes = group_var[group_var.isin([group])].index.tolist()\n",
    "        fragments_df = fragments_df_dict[sample_id]\n",
    "        group_fragments = fragments_df.loc[fragments_df[\"Name\"].isin(barcodes)]\n",
    "        if len(fragments_df_dict) > 1:\n",
    "            group_fragments_dict[sample_id] = group_fragments\n",
    "\n",
    "    if len(fragments_df_dict) > 1:\n",
    "        group_fragments_list = [\n",
    "            group_fragments_dict[list(group_fragments_dict.keys())[x]]\n",
    "            for x in range(len(fragments_df_dict))\n",
    "        ]\n",
    "        group_fragments = group_fragments_list[0].append(group_fragments_list[1:])\n",
    "\n",
    "    group_fragments = group_fragments[\n",
    "        group_fragments[\"Chromosome\"].isin(chromsizes.Chromosome)\n",
    "    ]\n",
    "\n",
    "    del group_fragments_dict\n",
    "    del group_fragments_list\n",
    "    del fragments_df\n",
    "    gc.collect()\n",
    "\n",
    "    group_pr = pr.PyRanges(group_fragments)\n",
    "    if isinstance(bigwig_path, str):\n",
    "        bigwig_path_group = os.path.join(bigwig_path, str(group) + \".bw\")\n",
    "        if remove_duplicates:\n",
    "            group_pr.to_bigwig(\n",
    "                path=bigwig_path_group,\n",
    "                chromosome_sizes=chromsizes,\n",
    "                rpm=normalize_bigwig,\n",
    "            )\n",
    "        else:\n",
    "            group_pr.to_bigwig(\n",
    "                path=bigwig_path_group,\n",
    "                chromosome_sizes=chromsizes,\n",
    "                rpm=normalize_bigwig,\n",
    "                value_col=\"Score\",\n",
    "            )\n",
    "    if isinstance(bed_path, str):\n",
    "        bed_path_group = os.path.join(bed_path, str(group) + \".bed.gz\")\n",
    "        group_pr.to_bed(\n",
    "            path=bed_path_group, keep=False, compression=\"infer\", chain=False\n",
    "        )\n",
    "\n",
    "    log.info(str(group) + \" done!\")\n",
    "\n",
    "\n",
    "@ray.remote\n",
    "def export_pseudobulk_ray(\n",
    "    cell_data: pd.DataFrame,\n",
    "    group: str,\n",
    "    fragments_df_dict: Dict[str, pd.DataFrame],\n",
    "    chromsizes: pr.PyRanges,\n",
    "    bigwig_path: str,\n",
    "    bed_path: str,\n",
    "    sample_id_col: Optional[str] = \"sample_id\",\n",
    "    normalize_bigwig: Optional[bool] = True,\n",
    "    remove_duplicates: Optional[bool] = True,\n",
    "    split_pattern: Optional[str] = \"___\",\n",
    "):\n",
    "    \"\"\"\n",
    "    Create pseudobulk as bed and bigwig from single cell fragments file given a barcode annotation and a group.\n",
    "    Parameters\n",
    "    ---------\n",
    "    cell_data: pd.DataFrame\n",
    "            A cell metadata :class:`pd.Dataframe` containing barcodes, their annotation and their sample of origin.\n",
    "    group: str\n",
    "            A character string indicating the group for which pseudobulks will be created.\n",
    "    fragments_df_dict: dict\n",
    "            A dictionary containing data frames as values with 'Chromosome', 'Start', 'End', 'Name', and 'Score' as columns; and sample label\n",
    "            as keys. 'Score' indicates the number of times that a fragments is found assigned to that barcode.\n",
    "    chromsizes: pr.PyRanges\n",
    "            A :class:`pr.PyRanges` containing size of each column, containing 'Chromosome', 'Start' and 'End' columns.\n",
    "    bed_path: str\n",
    "            Path to folder where the fragments bed file will be saved.\n",
    "    bigwig_path: str\n",
    "            Path to folder where the bigwig file will be saved.\n",
    "    sample_id_col: str, optional\n",
    "            Name of the column containing the sample name per barcode in the input :class:`CistopicObject.cell_data` or class:`pd.DataFrame`. Default: 'sample_id'.\n",
    "    normalize_bigwig: bool, optional\n",
    "            Whether bigwig files should be CPM normalized. Default: True.\n",
    "    remove_duplicates: bool, optional\n",
    "            Whether duplicates should be removed before converting the data to bigwig.\n",
    "    split_pattern: str\n",
    "            Pattern to split cell barcode from sample id. Default: ___ .\n",
    "    \"\"\"\n",
    "    export_pseudobulk_one_sample(\n",
    "        cell_data,\n",
    "        group,\n",
    "        fragments_df_dict,\n",
    "        chromsizes,\n",
    "        bigwig_path,\n",
    "        bed_path,\n",
    "        sample_id_col,\n",
    "        normalize_bigwig,\n",
    "        remove_duplicates,\n",
    "        split_pattern,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "15d0893a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:11:55.920125Z",
     "iopub.status.busy": "2022-09-26T14:11:55.919950Z",
     "iopub.status.idle": "2022-09-26T14:11:55.942832Z",
     "shell.execute_reply": "2022-09-26T14:11:55.942274Z"
    },
    "papermill": {
     "duration": 0.028476,
     "end_time": "2022-09-26T14:11:55.943798",
     "exception": false,
     "start_time": "2022-09-26T14:11:55.915322",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SCREEN_peaks/BIO_ddseq_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/BIO_ddseq_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/BIO_ddseq_3.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/BIO_ddseq_4.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/BRO_mtscatac_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/BRO_mtscatac_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/CNA_10xmultiome_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/CNA_10xmultiome_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/CNA_10xv11_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/CNA_10xv11_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/CNA_10xv11_3.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/CNA_10xv11_4.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/CNA_10xv11_5.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/CNA_10xv2_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/CNA_10xv2_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/CNA_hydrop_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/CNA_hydrop_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/CNA_hydrop_3.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/CNA_mtscatac_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/CNA_mtscatac_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/EPF_hydrop_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/EPF_hydrop_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/EPF_hydrop_3.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/EPF_hydrop_4.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/HAR_ddseq_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/HAR_ddseq_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/MDC_mtscatac_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/MDC_mtscatac_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/OHS_s3atac_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/SAN_10xmultiome_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/SAN_10xmultiome_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/STA_10xv11_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/STA_10xv11_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/TXG_10xv11_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/TXG_10xv2_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/TXG_10xv2_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/VIB_10xmultiome_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/VIB_10xmultiome_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/VIB_10xv1_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/VIB_10xv1_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/VIB_10xv2_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/VIB_10xv2_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/VIB_hydrop_1.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/VIB_hydrop_11.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/VIB_hydrop_12.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/VIB_hydrop_2.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/VIB_hydrop_21.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n",
      "SCREEN_peaks/VIB_hydrop_22.FULL__SCREEN_pseudobulk_bed_files exists, skipping...\n"
     ]
    }
   ],
   "source": [
    "# for sample in [\"BRO_mtscatac_1.LIBDS\"]:\n",
    "bw_paths_dict = {}\n",
    "bed_paths_dict = {}\n",
    "\n",
    "import ray\n",
    "\n",
    "if ray.is_initialized():\n",
    "    print(\"Shutting down Ray\")\n",
    "    ray.shutdown()\n",
    "\n",
    "for sample in cto_consensus_path_dict.keys():\n",
    "    # for sample in [\"BRO_mtscatac_1.LIBDS\"]:\n",
    "    bed_path = os.path.join(\n",
    "        \"SCREEN_peaks\", f\"{sample}__SCREEN_pseudobulk_bed_files\"\n",
    "    )\n",
    "    bw_path = os.path.join(\n",
    "        \"SCREEN_peaks\", f\"{sample}__SCREEN_pseudobulk_bw_files\"\n",
    "    )\n",
    "    if not os.path.exists(bed_path):\n",
    "        cto_path = cto_consensus_path_dict[sample]\n",
    "        with open(cto_path, \"rb\") as f:\n",
    "            cto = pickle.load(f)\n",
    "\n",
    "            bw_paths, bed_paths = export_pseudobulk(\n",
    "                input_data=cto,\n",
    "                variable=\"consensus_cell_type\",\n",
    "                sample_id_col=\"sample_id\",\n",
    "                chromsizes=chromsizes,\n",
    "                bed_path=bed_path,\n",
    "                bigwig_path=bw_path,\n",
    "                path_to_fragments=fragments_path_dict[sample],\n",
    "                n_cpu=16,\n",
    "                normalize_bigwig=True,\n",
    "                remove_duplicates=True,\n",
    "            )\n",
    "\n",
    "            if ray.is_initialized():\n",
    "                print(\"Shutting down Ray\")\n",
    "                ray.shutdown()\n",
    "    else:\n",
    "        print(f\"{bed_path} exists, skipping...\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e20c3bc6",
   "metadata": {
    "papermill": {
     "duration": 0.004199,
     "end_time": "2022-09-26T14:11:55.952353",
     "exception": false,
     "start_time": "2022-09-26T14:11:55.948154",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "941734a4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:11:55.961838Z",
     "iopub.status.busy": "2022-09-26T14:11:55.961359Z",
     "iopub.status.idle": "2022-09-26T14:11:55.969727Z",
     "shell.execute_reply": "2022-09-26T14:11:55.969218Z"
    },
    "papermill": {
     "duration": 0.014072,
     "end_time": "2022-09-26T14:11:55.970701",
     "exception": false,
     "start_time": "2022-09-26T14:11:55.956629",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FULL': 'cistopic_objects/BIO_ddseq_1.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'BIO_ddseq_2.FULL': 'cistopic_objects/BIO_ddseq_2.FULL__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'BIO_ddseq_3.FULL': 'cistopic_objects/BIO_ddseq_3.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'BIO_ddseq_4.FULL': 'cistopic_objects/BIO_ddseq_4.FULL__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'BRO_mtscatac_1.FULL': 'cistopic_objects/BRO_mtscatac_1.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'BRO_mtscatac_2.FULL': 'cistopic_objects/BRO_mtscatac_2.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xmultiome_1.FULL': 'cistopic_objects/CNA_10xmultiome_1.FULL__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xmultiome_2.FULL': 'cistopic_objects/CNA_10xmultiome_2.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xv11_1.FULL': 'cistopic_objects/CNA_10xv11_1.FULL__cto.scrublet0-4.fmx.singlets.model_8topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xv11_2.FULL': 'cistopic_objects/CNA_10xv11_2.FULL__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xv11_3.FULL': 'cistopic_objects/CNA_10xv11_3.FULL__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xv11_4.FULL': 'cistopic_objects/CNA_10xv11_4.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xv11_5.FULL': 'cistopic_objects/CNA_10xv11_5.FULL__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xv2_1.FULL': 'cistopic_objects/CNA_10xv2_1.FULL__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xv2_2.FULL': 'cistopic_objects/CNA_10xv2_2.FULL__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_hydrop_1.FULL': 'cistopic_objects/CNA_hydrop_1.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_hydrop_2.FULL': 'cistopic_objects/CNA_hydrop_2.FULL__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_hydrop_3.FULL': 'cistopic_objects/CNA_hydrop_3.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_mtscatac_1.FULL': 'cistopic_objects/CNA_mtscatac_1.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_mtscatac_2.FULL': 'cistopic_objects/CNA_mtscatac_2.FULL__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'EPF_hydrop_1.FULL': 'cistopic_objects/EPF_hydrop_1.FULL__cto.scrublet0-4.fmx.singlets.model_8topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'EPF_hydrop_2.FULL': 'cistopic_objects/EPF_hydrop_2.FULL__cto.scrublet0-4.fmx.singlets.model_8topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'EPF_hydrop_3.FULL': 'cistopic_objects/EPF_hydrop_3.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'EPF_hydrop_4.FULL': 'cistopic_objects/EPF_hydrop_4.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'HAR_ddseq_1.FULL': 'cistopic_objects/HAR_ddseq_1.FULL__cto.scrublet0-4.fmx.singlets.model_8topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'HAR_ddseq_2.FULL': 'cistopic_objects/HAR_ddseq_2.FULL__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'MDC_mtscatac_1.FULL': 'cistopic_objects/MDC_mtscatac_1.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'MDC_mtscatac_2.FULL': 'cistopic_objects/MDC_mtscatac_2.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'OHS_s3atac_1.FULL': 'cistopic_objects/OHS_s3atac_1.FULL__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'OHS_s3atac_2.FULL': 'cistopic_objects/OHS_s3atac_2.FULL__cto.scrublet0-4.fmx.singlets.model_7topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'SAN_10xmultiome_1.FULL': 'cistopic_objects/SAN_10xmultiome_1.FULL__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'SAN_10xmultiome_2.FULL': 'cistopic_objects/SAN_10xmultiome_2.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'STA_10xv11_1.FULL': 'cistopic_objects/STA_10xv11_1.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'STA_10xv11_2.FULL': 'cistopic_objects/STA_10xv11_2.FULL__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'TXG_10xv11_1.FULL': 'cistopic_objects/TXG_10xv11_1.FULL__cto.scrublet0-4.fmx.singlets.model_16topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'TXG_10xv2_1.FULL': 'cistopic_objects/TXG_10xv2_1.FULL__cto.scrublet0-4.fmx.singlets.model_18topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'TXG_10xv2_2.FULL': 'cistopic_objects/TXG_10xv2_2.FULL__cto.scrublet0-4.fmx.singlets.model_16topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'UCS_ddseq_1.FULL': 'cistopic_objects/UCS_ddseq_1.FULL__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'UCS_ddseq_2.FULL': 'cistopic_objects/UCS_ddseq_2.FULL__cto.scrublet0-4.fmx.singlets.model_17topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_10xmultiome_1.FULL': 'cistopic_objects/VIB_10xmultiome_1.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_10xmultiome_2.FULL': 'cistopic_objects/VIB_10xmultiome_2.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_10xv1_1.FULL': 'cistopic_objects/VIB_10xv1_1.FULL__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_10xv1_2.FULL': 'cistopic_objects/VIB_10xv1_2.FULL__cto.scrublet0-4.fmx.singlets.model_8topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_10xv2_1.FULL': 'cistopic_objects/VIB_10xv2_1.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_10xv2_2.FULL': 'cistopic_objects/VIB_10xv2_2.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_hydrop_1.FULL': 'cistopic_objects/VIB_hydrop_1.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_hydrop_11.FULL': 'cistopic_objects/VIB_hydrop_11.FULL__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_hydrop_12.FULL': 'cistopic_objects/VIB_hydrop_12.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_hydrop_2.FULL': 'cistopic_objects/VIB_hydrop_2.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_hydrop_21.FULL': 'cistopic_objects/VIB_hydrop_21.FULL__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_hydrop_22.FULL': 'cistopic_objects/VIB_hydrop_22.FULL__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_data_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"cistopic_objects/*consensus.cell_data.tsv\"))\n",
    "}\n",
    "cell_data_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9dd390ee",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:11:55.980354Z",
     "iopub.status.busy": "2022-09-26T14:11:55.980027Z",
     "iopub.status.idle": "2022-09-26T14:11:55.987914Z",
     "shell.execute_reply": "2022-09-26T14:11:55.987404Z"
    },
    "papermill": {
     "duration": 0.013668,
     "end_time": "2022-09-26T14:11:55.988877",
     "exception": false,
     "start_time": "2022-09-26T14:11:55.975209",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FULL': 'SCREEN_peaks/BIO_ddseq_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'BIO_ddseq_2.FULL': 'SCREEN_peaks/BIO_ddseq_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'BIO_ddseq_3.FULL': 'SCREEN_peaks/BIO_ddseq_3.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'BIO_ddseq_4.FULL': 'SCREEN_peaks/BIO_ddseq_4.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'BRO_mtscatac_1.FULL': 'SCREEN_peaks/BRO_mtscatac_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'BRO_mtscatac_2.FULL': 'SCREEN_peaks/BRO_mtscatac_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xmultiome_1.FULL': 'SCREEN_peaks/CNA_10xmultiome_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xmultiome_2.FULL': 'SCREEN_peaks/CNA_10xmultiome_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xv11_1.FULL': 'SCREEN_peaks/CNA_10xv11_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xv11_2.FULL': 'SCREEN_peaks/CNA_10xv11_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xv11_3.FULL': 'SCREEN_peaks/CNA_10xv11_3.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xv11_4.FULL': 'SCREEN_peaks/CNA_10xv11_4.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xv11_5.FULL': 'SCREEN_peaks/CNA_10xv11_5.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xv2_1.FULL': 'SCREEN_peaks/CNA_10xv2_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xv2_2.FULL': 'SCREEN_peaks/CNA_10xv2_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_hydrop_1.FULL': 'SCREEN_peaks/CNA_hydrop_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_hydrop_2.FULL': 'SCREEN_peaks/CNA_hydrop_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_hydrop_3.FULL': 'SCREEN_peaks/CNA_hydrop_3.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_mtscatac_1.FULL': 'SCREEN_peaks/CNA_mtscatac_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_mtscatac_2.FULL': 'SCREEN_peaks/CNA_mtscatac_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'EPF_hydrop_1.FULL': 'SCREEN_peaks/EPF_hydrop_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'EPF_hydrop_2.FULL': 'SCREEN_peaks/EPF_hydrop_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'EPF_hydrop_3.FULL': 'SCREEN_peaks/EPF_hydrop_3.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'EPF_hydrop_4.FULL': 'SCREEN_peaks/EPF_hydrop_4.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'HAR_ddseq_1.FULL': 'SCREEN_peaks/HAR_ddseq_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'HAR_ddseq_2.FULL': 'SCREEN_peaks/HAR_ddseq_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'MDC_mtscatac_1.FULL': 'SCREEN_peaks/MDC_mtscatac_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'MDC_mtscatac_2.FULL': 'SCREEN_peaks/MDC_mtscatac_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'OHS_s3atac_1.FULL': 'SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'OHS_s3atac_2.FULL': 'SCREEN_peaks/OHS_s3atac_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'SAN_10xmultiome_1.FULL': 'SCREEN_peaks/SAN_10xmultiome_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'SAN_10xmultiome_2.FULL': 'SCREEN_peaks/SAN_10xmultiome_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'STA_10xv11_1.FULL': 'SCREEN_peaks/STA_10xv11_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'STA_10xv11_2.FULL': 'SCREEN_peaks/STA_10xv11_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'TXG_10xv11_1.FULL': 'SCREEN_peaks/TXG_10xv11_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'TXG_10xv2_1.FULL': 'SCREEN_peaks/TXG_10xv2_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'TXG_10xv2_2.FULL': 'SCREEN_peaks/TXG_10xv2_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'UCS_ddseq_1.FULL': 'SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'UCS_ddseq_2.FULL': 'SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_10xmultiome_1.FULL': 'SCREEN_peaks/VIB_10xmultiome_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_10xmultiome_2.FULL': 'SCREEN_peaks/VIB_10xmultiome_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_10xv1_1.FULL': 'SCREEN_peaks/VIB_10xv1_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_10xv1_2.FULL': 'SCREEN_peaks/VIB_10xv1_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_10xv2_1.FULL': 'SCREEN_peaks/VIB_10xv2_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_10xv2_2.FULL': 'SCREEN_peaks/VIB_10xv2_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_hydrop_1.FULL': 'SCREEN_peaks/VIB_hydrop_1.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_hydrop_11.FULL': 'SCREEN_peaks/VIB_hydrop_11.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_hydrop_12.FULL': 'SCREEN_peaks/VIB_hydrop_12.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_hydrop_2.FULL': 'SCREEN_peaks/VIB_hydrop_2.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_hydrop_21.FULL': 'SCREEN_peaks/VIB_hydrop_21.FULL__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_hydrop_22.FULL': 'SCREEN_peaks/VIB_hydrop_22.FULL__SCREEN_pseudobulk_bw_files'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bw_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"SCREEN_peaks/*_pseudobulk_bw_files\"))\n",
    "}\n",
    "bw_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d57ef8d8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:11:55.998943Z",
     "iopub.status.busy": "2022-09-26T14:11:55.998469Z",
     "iopub.status.idle": "2022-09-26T14:11:56.005997Z",
     "shell.execute_reply": "2022-09-26T14:11:56.005485Z"
    },
    "papermill": {
     "duration": 0.013443,
     "end_time": "2022-09-26T14:11:56.006953",
     "exception": false,
     "start_time": "2022-09-26T14:11:55.993510",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FULL': 'SCREEN_peaks/BIO_ddseq_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'BIO_ddseq_2.FULL': 'SCREEN_peaks/BIO_ddseq_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'BIO_ddseq_3.FULL': 'SCREEN_peaks/BIO_ddseq_3.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'BIO_ddseq_4.FULL': 'SCREEN_peaks/BIO_ddseq_4.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'BRO_mtscatac_1.FULL': 'SCREEN_peaks/BRO_mtscatac_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'BRO_mtscatac_2.FULL': 'SCREEN_peaks/BRO_mtscatac_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xmultiome_1.FULL': 'SCREEN_peaks/CNA_10xmultiome_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xmultiome_2.FULL': 'SCREEN_peaks/CNA_10xmultiome_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xv11_1.FULL': 'SCREEN_peaks/CNA_10xv11_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xv11_2.FULL': 'SCREEN_peaks/CNA_10xv11_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xv11_3.FULL': 'SCREEN_peaks/CNA_10xv11_3.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xv11_4.FULL': 'SCREEN_peaks/CNA_10xv11_4.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xv11_5.FULL': 'SCREEN_peaks/CNA_10xv11_5.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xv2_1.FULL': 'SCREEN_peaks/CNA_10xv2_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xv2_2.FULL': 'SCREEN_peaks/CNA_10xv2_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_hydrop_1.FULL': 'SCREEN_peaks/CNA_hydrop_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_hydrop_2.FULL': 'SCREEN_peaks/CNA_hydrop_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_hydrop_3.FULL': 'SCREEN_peaks/CNA_hydrop_3.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_mtscatac_1.FULL': 'SCREEN_peaks/CNA_mtscatac_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_mtscatac_2.FULL': 'SCREEN_peaks/CNA_mtscatac_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'EPF_hydrop_1.FULL': 'SCREEN_peaks/EPF_hydrop_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'EPF_hydrop_2.FULL': 'SCREEN_peaks/EPF_hydrop_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'EPF_hydrop_3.FULL': 'SCREEN_peaks/EPF_hydrop_3.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'EPF_hydrop_4.FULL': 'SCREEN_peaks/EPF_hydrop_4.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'HAR_ddseq_1.FULL': 'SCREEN_peaks/HAR_ddseq_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'HAR_ddseq_2.FULL': 'SCREEN_peaks/HAR_ddseq_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'MDC_mtscatac_1.FULL': 'SCREEN_peaks/MDC_mtscatac_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'MDC_mtscatac_2.FULL': 'SCREEN_peaks/MDC_mtscatac_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'OHS_s3atac_1.FULL': 'SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'OHS_s3atac_2.FULL': 'SCREEN_peaks/OHS_s3atac_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'SAN_10xmultiome_1.FULL': 'SCREEN_peaks/SAN_10xmultiome_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'SAN_10xmultiome_2.FULL': 'SCREEN_peaks/SAN_10xmultiome_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'STA_10xv11_1.FULL': 'SCREEN_peaks/STA_10xv11_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'STA_10xv11_2.FULL': 'SCREEN_peaks/STA_10xv11_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'TXG_10xv11_1.FULL': 'SCREEN_peaks/TXG_10xv11_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'TXG_10xv2_1.FULL': 'SCREEN_peaks/TXG_10xv2_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'TXG_10xv2_2.FULL': 'SCREEN_peaks/TXG_10xv2_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'UCS_ddseq_1.FULL': 'SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'UCS_ddseq_2.FULL': 'SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_10xmultiome_1.FULL': 'SCREEN_peaks/VIB_10xmultiome_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_10xmultiome_2.FULL': 'SCREEN_peaks/VIB_10xmultiome_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_10xv1_1.FULL': 'SCREEN_peaks/VIB_10xv1_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_10xv1_2.FULL': 'SCREEN_peaks/VIB_10xv1_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_10xv2_1.FULL': 'SCREEN_peaks/VIB_10xv2_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_10xv2_2.FULL': 'SCREEN_peaks/VIB_10xv2_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_hydrop_1.FULL': 'SCREEN_peaks/VIB_hydrop_1.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_hydrop_11.FULL': 'SCREEN_peaks/VIB_hydrop_11.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_hydrop_12.FULL': 'SCREEN_peaks/VIB_hydrop_12.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_hydrop_2.FULL': 'SCREEN_peaks/VIB_hydrop_2.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_hydrop_21.FULL': 'SCREEN_peaks/VIB_hydrop_21.FULL__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_hydrop_22.FULL': 'SCREEN_peaks/VIB_hydrop_22.FULL__SCREEN_pseudobulk_bed_files'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"SCREEN_peaks/*_pseudobulk_bed_files\"))\n",
    "}\n",
    "bed_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "40d98405",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:11:56.017276Z",
     "iopub.status.busy": "2022-09-26T14:11:56.016849Z",
     "iopub.status.idle": "2022-09-26T14:11:56.046116Z",
     "shell.execute_reply": "2022-09-26T14:11:56.045592Z"
    },
    "papermill": {
     "duration": 0.035334,
     "end_time": "2022-09-26T14:11:56.047089",
     "exception": false,
     "start_time": "2022-09-26T14:11:56.011755",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "for sample in bed_path_dict.keys():\n",
    "    bed_paths = {\n",
    "        x.split(\"/\")[-1].split(\"__\")[0].split(\".bed.gz\")[0]: x\n",
    "        for x in glob.glob(bed_path_dict[sample] + \"/*\")\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "211a165d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:11:56.057523Z",
     "iopub.status.busy": "2022-09-26T14:11:56.057106Z",
     "iopub.status.idle": "2022-09-26T14:11:56.060300Z",
     "shell.execute_reply": "2022-09-26T14:11:56.059781Z"
    },
    "papermill": {
     "duration": 0.009309,
     "end_time": "2022-09-26T14:11:56.061260",
     "exception": false,
     "start_time": "2022-09-26T14:11:56.051951",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pycisTopic.pseudobulk_peak_calling import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5100f7a1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:11:56.071810Z",
     "iopub.status.busy": "2022-09-26T14:11:56.071385Z",
     "iopub.status.idle": "2022-09-26T14:11:56.074458Z",
     "shell.execute_reply": "2022-09-26T14:11:56.073931Z"
    },
    "papermill": {
     "duration": 0.009342,
     "end_time": "2022-09-26T14:11:56.075450",
     "exception": false,
     "start_time": "2022-09-26T14:11:56.066108",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import ray"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ff106404",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:11:56.085731Z",
     "iopub.status.busy": "2022-09-26T14:11:56.085402Z",
     "iopub.status.idle": "2022-09-26T14:43:48.332067Z",
     "shell.execute_reply": "2022-09-26T14:43:48.331288Z"
    },
    "papermill": {
     "duration": 1912.254661,
     "end_time": "2022-09-26T14:43:48.334810",
     "exception": false,
     "start_time": "2022-09-26T14:11:56.080149",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SCREEN_peaks/BIO_ddseq_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/BIO_ddseq_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/BIO_ddseq_3.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/BIO_ddseq_4.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/BRO_mtscatac_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/BRO_mtscatac_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/CNA_10xmultiome_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/CNA_10xmultiome_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/CNA_10xv11_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/CNA_10xv11_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/CNA_10xv11_3.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/CNA_10xv11_4.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/CNA_10xv11_5.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/CNA_10xv2_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/CNA_10xv2_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/CNA_hydrop_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/CNA_hydrop_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/CNA_hydrop_3.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/CNA_mtscatac_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/CNA_mtscatac_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/EPF_hydrop_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/EPF_hydrop_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/EPF_hydrop_3.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/EPF_hydrop_4.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/HAR_ddseq_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/HAR_ddseq_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/MDC_mtscatac_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/MDC_mtscatac_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "Starting SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_narrow_peaks_dict.pkl\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-26 14:12:00,575\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2444689)\u001b[0m 2022-09-26 14:12:28,651 cisTopic     INFO     Calling peaks for CytotoxicTcell with macs2 callpeak --treatment SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_pseudobulk_bed_files/CytotoxicTcell.bed.gz --name CytotoxicTcell  --outdir SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2444688)\u001b[0m 2022-09-26 14:12:28,654 cisTopic     INFO     Calling peaks for CD14_monocyte with macs2 callpeak --treatment SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_pseudobulk_bed_files/CD14_monocyte.bed.gz --name CD14_monocyte  --outdir SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2444690)\u001b[0m 2022-09-26 14:12:28,673 cisTopic     INFO     Calling peaks for Naturalkillercell with macs2 callpeak --treatment SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_pseudobulk_bed_files/Naturalkillercell.bed.gz --name Naturalkillercell  --outdir SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2444691)\u001b[0m 2022-09-26 14:12:28,679 cisTopic     INFO     Calling peaks for CD4_Tcell with macs2 callpeak --treatment SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_pseudobulk_bed_files/CD4_Tcell.bed.gz --name CD4_Tcell  --outdir SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2444692)\u001b[0m 2022-09-26 14:12:28,698 cisTopic     INFO     Calling peaks for Bcell with macs2 callpeak --treatment SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_pseudobulk_bed_files/Bcell.bed.gz --name Bcell  --outdir SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2444690)\u001b[0m 2022-09-26 14:15:30,346 cisTopic     INFO     Naturalkillercell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2444689)\u001b[0m 2022-09-26 14:16:26,039 cisTopic     INFO     CytotoxicTcell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2444692)\u001b[0m 2022-09-26 14:17:05,330 cisTopic     INFO     Bcell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2444691)\u001b[0m 2022-09-26 14:19:53,955 cisTopic     INFO     CD4_Tcell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2444688)\u001b[0m 2022-09-26 14:20:20,526 cisTopic     INFO     CD14_monocyte done!\n",
      "Starting SCREEN_peaks/OHS_s3atac_2.FULL__SCREEN_narrow_peaks_dict.pkl\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-26 14:20:28,433\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2445776)\u001b[0m 2022-09-26 14:20:54,580 cisTopic     INFO     Calling peaks for CD4_Tcell with macs2 callpeak --treatment SCREEN_peaks/OHS_s3atac_2.FULL__SCREEN_pseudobulk_bed_files/CD4_Tcell.bed.gz --name CD4_Tcell  --outdir SCREEN_peaks/OHS_s3atac_2.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2445774)\u001b[0m 2022-09-26 14:20:54,594 cisTopic     INFO     Calling peaks for CD14_monocyte with macs2 callpeak --treatment SCREEN_peaks/OHS_s3atac_2.FULL__SCREEN_pseudobulk_bed_files/CD14_monocyte.bed.gz --name CD14_monocyte  --outdir SCREEN_peaks/OHS_s3atac_2.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2445772)\u001b[0m 2022-09-26 14:20:54,590 cisTopic     INFO     Calling peaks for Naturalkillercell with macs2 callpeak --treatment SCREEN_peaks/OHS_s3atac_2.FULL__SCREEN_pseudobulk_bed_files/Naturalkillercell.bed.gz --name Naturalkillercell  --outdir SCREEN_peaks/OHS_s3atac_2.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2445775)\u001b[0m 2022-09-26 14:20:54,577 cisTopic     INFO     Calling peaks for Bcell with macs2 callpeak --treatment SCREEN_peaks/OHS_s3atac_2.FULL__SCREEN_pseudobulk_bed_files/Bcell.bed.gz --name Bcell  --outdir SCREEN_peaks/OHS_s3atac_2.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2445772)\u001b[0m 2022-09-26 14:21:17,038 cisTopic     INFO     Naturalkillercell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2445775)\u001b[0m 2022-09-26 14:25:06,611 cisTopic     INFO     Bcell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2445774)\u001b[0m 2022-09-26 14:31:27,141 cisTopic     INFO     CD14_monocyte done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2445776)\u001b[0m 2022-09-26 14:32:20,177 cisTopic     INFO     CD4_Tcell done!\n",
      "SCREEN_peaks/SAN_10xmultiome_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/SAN_10xmultiome_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/STA_10xv11_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/STA_10xv11_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/TXG_10xv11_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/TXG_10xv2_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/TXG_10xv2_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "Starting SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_narrow_peaks_dict.pkl\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-26 14:32:28,012\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2446976)\u001b[0m 2022-09-26 14:32:55,933 cisTopic     INFO     Calling peaks for CytotoxicTcell with macs2 callpeak --treatment SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_pseudobulk_bed_files/CytotoxicTcell.bed.gz --name CytotoxicTcell  --outdir SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2446979)\u001b[0m 2022-09-26 14:32:55,919 cisTopic     INFO     Calling peaks for CD4_Tcell with macs2 callpeak --treatment SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_pseudobulk_bed_files/CD4_Tcell.bed.gz --name CD4_Tcell  --outdir SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2446980)\u001b[0m 2022-09-26 14:32:55,923 cisTopic     INFO     Calling peaks for Bcell with macs2 callpeak --treatment SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_pseudobulk_bed_files/Bcell.bed.gz --name Bcell  --outdir SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2446977)\u001b[0m 2022-09-26 14:32:55,921 cisTopic     INFO     Calling peaks for CD14_monocyte with macs2 callpeak --treatment SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_pseudobulk_bed_files/CD14_monocyte.bed.gz --name CD14_monocyte  --outdir SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2446978)\u001b[0m 2022-09-26 14:32:55,924 cisTopic     INFO     Calling peaks for Naturalkillercell with macs2 callpeak --treatment SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_pseudobulk_bed_files/Naturalkillercell.bed.gz --name Naturalkillercell  --outdir SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2446978)\u001b[0m 2022-09-26 14:34:40,105 cisTopic     INFO     Naturalkillercell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2446980)\u001b[0m 2022-09-26 14:36:10,657 cisTopic     INFO     Bcell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2446976)\u001b[0m 2022-09-26 14:37:23,839 cisTopic     INFO     CytotoxicTcell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2446977)\u001b[0m 2022-09-26 14:37:28,934 cisTopic     INFO     CD14_monocyte done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2446979)\u001b[0m 2022-09-26 14:38:04,004 cisTopic     INFO     CD4_Tcell done!\n",
      "Starting SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_narrow_peaks_dict.pkl\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-26 14:38:11,564\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2448007)\u001b[0m 2022-09-26 14:38:39,414 cisTopic     INFO     Calling peaks for CD14_monocyte with macs2 callpeak --treatment SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_pseudobulk_bed_files/CD14_monocyte.bed.gz --name CD14_monocyte  --outdir SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2448010)\u001b[0m 2022-09-26 14:38:39,428 cisTopic     INFO     Calling peaks for CD4_Tcell with macs2 callpeak --treatment SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_pseudobulk_bed_files/CD4_Tcell.bed.gz --name CD4_Tcell  --outdir SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2448009)\u001b[0m 2022-09-26 14:38:39,427 cisTopic     INFO     Calling peaks for Naturalkillercell with macs2 callpeak --treatment SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_pseudobulk_bed_files/Naturalkillercell.bed.gz --name Naturalkillercell  --outdir SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2448008)\u001b[0m 2022-09-26 14:38:39,439 cisTopic     INFO     Calling peaks for CytotoxicTcell with macs2 callpeak --treatment SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_pseudobulk_bed_files/CytotoxicTcell.bed.gz --name CytotoxicTcell  --outdir SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2448011)\u001b[0m 2022-09-26 14:38:39,420 cisTopic     INFO     Calling peaks for Bcell with macs2 callpeak --treatment SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_pseudobulk_bed_files/Bcell.bed.gz --name Bcell  --outdir SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2448009)\u001b[0m 2022-09-26 14:41:15,615 cisTopic     INFO     Naturalkillercell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2448011)\u001b[0m 2022-09-26 14:42:01,541 cisTopic     INFO     Bcell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2448007)\u001b[0m 2022-09-26 14:42:57,767 cisTopic     INFO     CD14_monocyte done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2448008)\u001b[0m 2022-09-26 14:43:10,068 cisTopic     INFO     CytotoxicTcell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=2448010)\u001b[0m 2022-09-26 14:43:43,961 cisTopic     INFO     CD4_Tcell done!\n",
      "SCREEN_peaks/VIB_10xmultiome_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/VIB_10xmultiome_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/VIB_10xv1_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/VIB_10xv1_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/VIB_10xv2_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/VIB_10xv2_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/VIB_hydrop_1.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/VIB_hydrop_11.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/VIB_hydrop_12.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/VIB_hydrop_2.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/VIB_hydrop_21.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n",
      "SCREEN_peaks/VIB_hydrop_22.FULL__SCREEN_narrow_peaks_dict.pkl already exists\n"
     ]
    }
   ],
   "source": [
    "narrow_peaks_dict = {}\n",
    "ray.shutdown()\n",
    "for sample in bed_path_dict.keys():\n",
    "    narrow_peaks_dict_path = bed_path_dict[sample].replace(\n",
    "        \"_pseudobulk_bed_files\", \"_narrow_peaks_dict.pkl\"\n",
    "    )\n",
    "    peak_path = os.path.join(\"SCREEN_peaks\", f\"{sample}__SCREEN_consensus_peaks\")\n",
    "    if not os.path.exists(peak_path):\n",
    "        os.mkdir(peak_path)\n",
    "\n",
    "    if not os.path.exists(narrow_peaks_dict_path):\n",
    "        cell_data = pd.read_csv(cell_data_path_dict[sample])\n",
    "        cto_celltypes = set(\n",
    "            [\n",
    "                x.replace(\" \", \"\").replace(\"+\", \"\").replace(\"_\", \"\")\n",
    "                for x in set(cell_data[\"consensus_cell_type\"].unique())\n",
    "            ]\n",
    "        )\n",
    "\n",
    "        bed_celltypes = set(\n",
    "            [\n",
    "                x.split(\".\")[0].replace(\"+\", \"\").replace(\"_\", \"\")\n",
    "                for x in os.listdir(bed_path_dict[sample])\n",
    "            ]\n",
    "        )\n",
    "\n",
    "        if cto_celltypes == bed_celltypes:\n",
    "            print(f\"Starting {narrow_peaks_dict_path}\")\n",
    "            bed_paths = {\n",
    "                x.split(\"/\")[-1].split(\"__\")[0].split(\".bed.gz\")[0]: x\n",
    "                for x in glob.glob(bed_path_dict[sample] + \"/*\")\n",
    "            }\n",
    "            narrow_peaks_dict = peak_calling(\n",
    "                macs_path=\"macs2\",\n",
    "                bed_paths=bed_paths,\n",
    "                outdir=peak_path,\n",
    "                genome_size=\"hs\",\n",
    "                n_cpu=20,\n",
    "                input_format=\"BEDPE\",\n",
    "                shift=73,\n",
    "                ext_size=146,\n",
    "                keep_dup=\"all\",\n",
    "                q_value=0.05,\n",
    "            )\n",
    "            with open(narrow_peaks_dict_path, \"wb\") as f:\n",
    "                pickle.dump(narrow_peaks_dict, f)\n",
    "        else:\n",
    "            print(f\"{sample} cell types not matching!! Rerun bed file writing.\")\n",
    "            print(f\"\\t{bed_celltypes}\")\n",
    "            print(f\"\\t{cto_celltypes}\")\n",
    "    else:\n",
    "        print(f\"{narrow_peaks_dict_path} already exists\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e120de6d",
   "metadata": {
    "papermill": {
     "duration": 0.015003,
     "end_time": "2022-09-26T14:43:48.383511",
     "exception": false,
     "start_time": "2022-09-26T14:43:48.368508",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# call consensus peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "fd8faf97",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:43:48.396546Z",
     "iopub.status.busy": "2022-09-26T14:43:48.396250Z",
     "iopub.status.idle": "2022-09-26T14:43:48.401140Z",
     "shell.execute_reply": "2022-09-26T14:43:48.400615Z"
    },
    "papermill": {
     "duration": 0.012753,
     "end_time": "2022-09-26T14:43:48.402097",
     "exception": false,
     "start_time": "2022-09-26T14:43:48.389344",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pycisTopic.iterative_peak_calling import get_consensus_peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "dfbd901f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:43:48.414888Z",
     "iopub.status.busy": "2022-09-26T14:43:48.414715Z",
     "iopub.status.idle": "2022-09-26T14:43:48.620114Z",
     "shell.execute_reply": "2022-09-26T14:43:48.619582Z"
    },
    "papermill": {
     "duration": 0.213139,
     "end_time": "2022-09-26T14:43:48.621111",
     "exception": false,
     "start_time": "2022-09-26T14:43:48.407972",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FULL': 'SCREEN_peaks/BIO_ddseq_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'BIO_ddseq_2.FULL': 'SCREEN_peaks/BIO_ddseq_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'BIO_ddseq_3.FULL': 'SCREEN_peaks/BIO_ddseq_3.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'BIO_ddseq_4.FULL': 'SCREEN_peaks/BIO_ddseq_4.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'BRO_mtscatac_1.FULL': 'SCREEN_peaks/BRO_mtscatac_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'BRO_mtscatac_2.FULL': 'SCREEN_peaks/BRO_mtscatac_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xmultiome_1.FULL': 'SCREEN_peaks/CNA_10xmultiome_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xmultiome_2.FULL': 'SCREEN_peaks/CNA_10xmultiome_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xv11_1.FULL': 'SCREEN_peaks/CNA_10xv11_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xv11_2.FULL': 'SCREEN_peaks/CNA_10xv11_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xv11_3.FULL': 'SCREEN_peaks/CNA_10xv11_3.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xv11_4.FULL': 'SCREEN_peaks/CNA_10xv11_4.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xv11_5.FULL': 'SCREEN_peaks/CNA_10xv11_5.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xv2_1.FULL': 'SCREEN_peaks/CNA_10xv2_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xv2_2.FULL': 'SCREEN_peaks/CNA_10xv2_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_hydrop_1.FULL': 'SCREEN_peaks/CNA_hydrop_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_hydrop_2.FULL': 'SCREEN_peaks/CNA_hydrop_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_hydrop_3.FULL': 'SCREEN_peaks/CNA_hydrop_3.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_mtscatac_1.FULL': 'SCREEN_peaks/CNA_mtscatac_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_mtscatac_2.FULL': 'SCREEN_peaks/CNA_mtscatac_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'EPF_hydrop_1.FULL': 'SCREEN_peaks/EPF_hydrop_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'EPF_hydrop_2.FULL': 'SCREEN_peaks/EPF_hydrop_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'EPF_hydrop_3.FULL': 'SCREEN_peaks/EPF_hydrop_3.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'EPF_hydrop_4.FULL': 'SCREEN_peaks/EPF_hydrop_4.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'HAR_ddseq_1.FULL': 'SCREEN_peaks/HAR_ddseq_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'HAR_ddseq_2.FULL': 'SCREEN_peaks/HAR_ddseq_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'MDC_mtscatac_1.FULL': 'SCREEN_peaks/MDC_mtscatac_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'MDC_mtscatac_2.FULL': 'SCREEN_peaks/MDC_mtscatac_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'OHS_s3atac_1.FULL': 'SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'OHS_s3atac_2.FULL': 'SCREEN_peaks/OHS_s3atac_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'SAN_10xmultiome_1.FULL': 'SCREEN_peaks/SAN_10xmultiome_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'SAN_10xmultiome_2.FULL': 'SCREEN_peaks/SAN_10xmultiome_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'STA_10xv11_1.FULL': 'SCREEN_peaks/STA_10xv11_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'STA_10xv11_2.FULL': 'SCREEN_peaks/STA_10xv11_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'TXG_10xv11_1.FULL': 'SCREEN_peaks/TXG_10xv11_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'TXG_10xv2_1.FULL': 'SCREEN_peaks/TXG_10xv2_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'TXG_10xv2_2.FULL': 'SCREEN_peaks/TXG_10xv2_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'UCS_ddseq_1.FULL': 'SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'UCS_ddseq_2.FULL': 'SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_10xmultiome_1.FULL': 'SCREEN_peaks/VIB_10xmultiome_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_10xmultiome_2.FULL': 'SCREEN_peaks/VIB_10xmultiome_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_10xv1_1.FULL': 'SCREEN_peaks/VIB_10xv1_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_10xv1_2.FULL': 'SCREEN_peaks/VIB_10xv1_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_10xv2_1.FULL': 'SCREEN_peaks/VIB_10xv2_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_10xv2_2.FULL': 'SCREEN_peaks/VIB_10xv2_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_hydrop_1.FULL': 'SCREEN_peaks/VIB_hydrop_1.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_hydrop_11.FULL': 'SCREEN_peaks/VIB_hydrop_11.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_hydrop_12.FULL': 'SCREEN_peaks/VIB_hydrop_12.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_hydrop_2.FULL': 'SCREEN_peaks/VIB_hydrop_2.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_hydrop_21.FULL': 'SCREEN_peaks/VIB_hydrop_21.FULL__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_hydrop_22.FULL': 'SCREEN_peaks/VIB_hydrop_22.FULL__SCREEN_narrow_peaks_dict.pkl'}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "narrow_peaks_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"SCREEN_peaks/*_narrow_peaks_dict.pkl\"))\n",
    "}\n",
    "narrow_peaks_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4a65d294",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:43:48.634593Z",
     "iopub.status.busy": "2022-09-26T14:43:48.634262Z",
     "iopub.status.idle": "2022-09-26T14:43:48.637583Z",
     "shell.execute_reply": "2022-09-26T14:43:48.637083Z"
    },
    "papermill": {
     "duration": 0.010868,
     "end_time": "2022-09-26T14:43:48.638464",
     "exception": false,
     "start_time": "2022-09-26T14:43:48.627596",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "path_to_blacklist = \"../0_resources/regions/hg38-blacklist.v2.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7817628d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:43:48.651481Z",
     "iopub.status.busy": "2022-09-26T14:43:48.651175Z",
     "iopub.status.idle": "2022-09-26T14:51:18.102835Z",
     "shell.execute_reply": "2022-09-26T14:51:18.102111Z"
    },
    "papermill": {
     "duration": 449.48892,
     "end_time": "2022-09-26T14:51:18.133543",
     "exception": false,
     "start_time": "2022-09-26T14:43:48.644623",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BIO_ddseq_1.FULL\n",
      "SCREEN_peaks/BIO_ddseq_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "BIO_ddseq_2.FULL\n",
      "SCREEN_peaks/BIO_ddseq_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "BIO_ddseq_3.FULL\n",
      "SCREEN_peaks/BIO_ddseq_3.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "BIO_ddseq_4.FULL\n",
      "SCREEN_peaks/BIO_ddseq_4.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "BRO_mtscatac_1.FULL\n",
      "SCREEN_peaks/BRO_mtscatac_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "BRO_mtscatac_2.FULL\n",
      "SCREEN_peaks/BRO_mtscatac_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xmultiome_1.FULL\n",
      "SCREEN_peaks/CNA_10xmultiome_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xmultiome_2.FULL\n",
      "SCREEN_peaks/CNA_10xmultiome_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xv11_1.FULL\n",
      "SCREEN_peaks/CNA_10xv11_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xv11_2.FULL\n",
      "SCREEN_peaks/CNA_10xv11_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xv11_3.FULL\n",
      "SCREEN_peaks/CNA_10xv11_3.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xv11_4.FULL\n",
      "SCREEN_peaks/CNA_10xv11_4.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xv11_5.FULL\n",
      "SCREEN_peaks/CNA_10xv11_5.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xv2_1.FULL\n",
      "SCREEN_peaks/CNA_10xv2_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xv2_2.FULL\n",
      "SCREEN_peaks/CNA_10xv2_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_hydrop_1.FULL\n",
      "SCREEN_peaks/CNA_hydrop_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_hydrop_2.FULL\n",
      "SCREEN_peaks/CNA_hydrop_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_hydrop_3.FULL\n",
      "SCREEN_peaks/CNA_hydrop_3.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_mtscatac_1.FULL\n",
      "SCREEN_peaks/CNA_mtscatac_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_mtscatac_2.FULL\n",
      "SCREEN_peaks/CNA_mtscatac_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "EPF_hydrop_1.FULL\n",
      "SCREEN_peaks/EPF_hydrop_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "EPF_hydrop_2.FULL\n",
      "SCREEN_peaks/EPF_hydrop_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "EPF_hydrop_3.FULL\n",
      "SCREEN_peaks/EPF_hydrop_3.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "EPF_hydrop_4.FULL\n",
      "SCREEN_peaks/EPF_hydrop_4.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "HAR_ddseq_1.FULL\n",
      "SCREEN_peaks/HAR_ddseq_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "HAR_ddseq_2.FULL\n",
      "SCREEN_peaks/HAR_ddseq_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "MDC_mtscatac_1.FULL\n",
      "SCREEN_peaks/MDC_mtscatac_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "MDC_mtscatac_2.FULL\n",
      "SCREEN_peaks/MDC_mtscatac_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "OHS_s3atac_1.FULL\n",
      "2022-09-26 14:43:49,010 cisTopic     INFO     Extending and merging peaks per class\n",
      "2022-09-26 14:44:41,076 cisTopic     INFO     Normalizing peak scores\n",
      "2022-09-26 14:44:41,252 cisTopic     INFO     Merging peaks\n",
      "Warning! Start and End columns now have different dtypes: int64 and int32\n",
      "2022-09-26 14:45:23,329 cisTopic     INFO     Done!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OHS_s3atac_2.FULL\n",
      "2022-09-26 14:45:24,541 cisTopic     INFO     Extending and merging peaks per class\n",
      "2022-09-26 14:46:57,886 cisTopic     INFO     Normalizing peak scores\n",
      "2022-09-26 14:46:58,230 cisTopic     INFO     Merging peaks\n",
      "Warning! Start and End columns now have different dtypes: int64 and int32\n",
      "2022-09-26 14:48:46,695 cisTopic     INFO     Done!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SAN_10xmultiome_1.FULL\n",
      "SCREEN_peaks/SAN_10xmultiome_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "SAN_10xmultiome_2.FULL\n",
      "SCREEN_peaks/SAN_10xmultiome_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "STA_10xv11_1.FULL\n",
      "SCREEN_peaks/STA_10xv11_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "STA_10xv11_2.FULL\n",
      "SCREEN_peaks/STA_10xv11_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "TXG_10xv11_1.FULL\n",
      "SCREEN_peaks/TXG_10xv11_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "TXG_10xv2_1.FULL\n",
      "SCREEN_peaks/TXG_10xv2_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "TXG_10xv2_2.FULL\n",
      "SCREEN_peaks/TXG_10xv2_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "UCS_ddseq_1.FULL\n",
      "2022-09-26 14:48:48,926 cisTopic     INFO     Extending and merging peaks per class\n",
      "2022-09-26 14:49:27,981 cisTopic     INFO     Normalizing peak scores\n",
      "2022-09-26 14:49:28,125 cisTopic     INFO     Merging peaks\n",
      "Warning! Start and End columns now have different dtypes: int64 and int32\n",
      "2022-09-26 14:50:00,943 cisTopic     INFO     Done!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UCS_ddseq_2.FULL\n",
      "2022-09-26 14:50:01,923 cisTopic     INFO     Extending and merging peaks per class\n",
      "2022-09-26 14:50:41,748 cisTopic     INFO     Normalizing peak scores\n",
      "2022-09-26 14:50:41,909 cisTopic     INFO     Merging peaks\n",
      "Warning! Start and End columns now have different dtypes: int64 and int32\n",
      "2022-09-26 14:51:17,313 cisTopic     INFO     Done!\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n",
      "/opt/venv/lib/python3.8/site-packages/pyranges/out.py:37: FutureWarning: Passing a set as an indexer is deprecated and will raise in a future version. Use a list instead.\n",
      "  return pd.concat([outdf, df.get(noncanonical)], axis=1)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VIB_10xmultiome_1.FULL\n",
      "SCREEN_peaks/VIB_10xmultiome_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_10xmultiome_2.FULL\n",
      "SCREEN_peaks/VIB_10xmultiome_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_10xv1_1.FULL\n",
      "SCREEN_peaks/VIB_10xv1_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_10xv1_2.FULL\n",
      "SCREEN_peaks/VIB_10xv1_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_10xv2_1.FULL\n",
      "SCREEN_peaks/VIB_10xv2_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_10xv2_2.FULL\n",
      "SCREEN_peaks/VIB_10xv2_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_hydrop_1.FULL\n",
      "SCREEN_peaks/VIB_hydrop_1.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_hydrop_11.FULL\n",
      "SCREEN_peaks/VIB_hydrop_11.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_hydrop_12.FULL\n",
      "SCREEN_peaks/VIB_hydrop_12.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_hydrop_2.FULL\n",
      "SCREEN_peaks/VIB_hydrop_2.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_hydrop_21.FULL\n",
      "SCREEN_peaks/VIB_hydrop_21.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_hydrop_22.FULL\n",
      "SCREEN_peaks/VIB_hydrop_22.FULL__SCREEN_consensus_peaks.bed already exists, skipping...\n"
     ]
    }
   ],
   "source": [
    "peak_half_width = 250\n",
    "\n",
    "# Get consensus peaks\n",
    "consensus_peaks_dict = {}\n",
    "for sample in narrow_peaks_path_dict.keys():\n",
    "    print(sample)\n",
    "    consensus_out_path = narrow_peaks_path_dict[sample].replace(\n",
    "        \"_narrow_peaks_dict.pkl\", \"_consensus_peaks.bed\"\n",
    "    )\n",
    "    if not os.path.exists(consensus_out_path):\n",
    "        cell_data = pd.read_csv(cell_data_path_dict[sample])\n",
    "        cto_celltypes = set(\n",
    "            [\n",
    "                x.replace(\" \", \"\").replace(\"+\", \"\").replace(\"_\", \"\")\n",
    "                for x in set(cell_data[\"consensus_cell_type\"].unique())\n",
    "            ]\n",
    "        )\n",
    "\n",
    "        with open(narrow_peaks_path_dict[sample], \"rb\") as f:\n",
    "            narrow_peaks_dict = pickle.load(f)\n",
    "        peaks_celltypes = set(\n",
    "            [\n",
    "                x.replace(\" \", \"\").replace(\"+\", \"\").replace(\"_\", \"\")\n",
    "                for x in set(narrow_peaks_dict.keys())\n",
    "            ]\n",
    "        )\n",
    "\n",
    "        if cto_celltypes == peaks_celltypes:\n",
    "            consensus_peaks = get_consensus_peaks(\n",
    "                narrow_peaks_dict,\n",
    "                peak_half_width,\n",
    "                chromsizes=chromsizes,\n",
    "                path_to_blacklist=path_to_blacklist,\n",
    "            )\n",
    "\n",
    "            consensus_peaks.to_bed(\n",
    "                path=consensus_out_path, keep=True, compression=\"infer\", chain=False\n",
    "            )\n",
    "        else:\n",
    "            print(\"CELL TYPE SETS NOT MATCHING! Rerun peak calling.\")\n",
    "            print(peaks_celltypes - cto_celltypes)\n",
    "            print(cto_celltypes - peaks_celltypes)\n",
    "    else:\n",
    "        print(f\"{consensus_out_path} already exists, skipping...\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c91ef55",
   "metadata": {
    "papermill": {
     "duration": 0.020846,
     "end_time": "2022-09-26T14:51:18.228524",
     "exception": false,
     "start_time": "2022-09-26T14:51:18.207678",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Check % chrM in consensus peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "86b9e268",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:51:18.245612Z",
     "iopub.status.busy": "2022-09-26T14:51:18.245051Z",
     "iopub.status.idle": "2022-09-26T14:51:18.254960Z",
     "shell.execute_reply": "2022-09-26T14:51:18.254427Z"
    },
    "papermill": {
     "duration": 0.019318,
     "end_time": "2022-09-26T14:51:18.255887",
     "exception": false,
     "start_time": "2022-09-26T14:51:18.236569",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FULL': 'SCREEN_peaks/BIO_ddseq_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'BIO_ddseq_2.FULL': 'SCREEN_peaks/BIO_ddseq_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'BIO_ddseq_3.FULL': 'SCREEN_peaks/BIO_ddseq_3.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'BIO_ddseq_4.FULL': 'SCREEN_peaks/BIO_ddseq_4.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'BRO_mtscatac_1.FULL': 'SCREEN_peaks/BRO_mtscatac_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'BRO_mtscatac_2.FULL': 'SCREEN_peaks/BRO_mtscatac_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xmultiome_1.FULL': 'SCREEN_peaks/CNA_10xmultiome_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xmultiome_2.FULL': 'SCREEN_peaks/CNA_10xmultiome_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_1.FULL': 'SCREEN_peaks/CNA_10xv11_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_2.FULL': 'SCREEN_peaks/CNA_10xv11_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_3.FULL': 'SCREEN_peaks/CNA_10xv11_3.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_4.FULL': 'SCREEN_peaks/CNA_10xv11_4.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_5.FULL': 'SCREEN_peaks/CNA_10xv11_5.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv2_1.FULL': 'SCREEN_peaks/CNA_10xv2_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv2_2.FULL': 'SCREEN_peaks/CNA_10xv2_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_hydrop_1.FULL': 'SCREEN_peaks/CNA_hydrop_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_hydrop_2.FULL': 'SCREEN_peaks/CNA_hydrop_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_hydrop_3.FULL': 'SCREEN_peaks/CNA_hydrop_3.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_mtscatac_1.FULL': 'SCREEN_peaks/CNA_mtscatac_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_mtscatac_2.FULL': 'SCREEN_peaks/CNA_mtscatac_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_1.FULL': 'SCREEN_peaks/EPF_hydrop_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_2.FULL': 'SCREEN_peaks/EPF_hydrop_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_3.FULL': 'SCREEN_peaks/EPF_hydrop_3.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_4.FULL': 'SCREEN_peaks/EPF_hydrop_4.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'HAR_ddseq_1.FULL': 'SCREEN_peaks/HAR_ddseq_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'HAR_ddseq_2.FULL': 'SCREEN_peaks/HAR_ddseq_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'MDC_mtscatac_1.FULL': 'SCREEN_peaks/MDC_mtscatac_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'MDC_mtscatac_2.FULL': 'SCREEN_peaks/MDC_mtscatac_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'OHS_s3atac_1.FULL': 'SCREEN_peaks/OHS_s3atac_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'OHS_s3atac_2.FULL': 'SCREEN_peaks/OHS_s3atac_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'SAN_10xmultiome_1.FULL': 'SCREEN_peaks/SAN_10xmultiome_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'SAN_10xmultiome_2.FULL': 'SCREEN_peaks/SAN_10xmultiome_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'STA_10xv11_1.FULL': 'SCREEN_peaks/STA_10xv11_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'STA_10xv11_2.FULL': 'SCREEN_peaks/STA_10xv11_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'TXG_10xv11_1.FULL': 'SCREEN_peaks/TXG_10xv11_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'TXG_10xv2_1.FULL': 'SCREEN_peaks/TXG_10xv2_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'TXG_10xv2_2.FULL': 'SCREEN_peaks/TXG_10xv2_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'UCS_ddseq_1.FULL': 'SCREEN_peaks/UCS_ddseq_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'UCS_ddseq_2.FULL': 'SCREEN_peaks/UCS_ddseq_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xmultiome_1.FULL': 'SCREEN_peaks/VIB_10xmultiome_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xmultiome_2.FULL': 'SCREEN_peaks/VIB_10xmultiome_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xv1_1.FULL': 'SCREEN_peaks/VIB_10xv1_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xv1_2.FULL': 'SCREEN_peaks/VIB_10xv1_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xv2_1.FULL': 'SCREEN_peaks/VIB_10xv2_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xv2_2.FULL': 'SCREEN_peaks/VIB_10xv2_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_1.FULL': 'SCREEN_peaks/VIB_hydrop_1.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_11.FULL': 'SCREEN_peaks/VIB_hydrop_11.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_12.FULL': 'SCREEN_peaks/VIB_hydrop_12.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_2.FULL': 'SCREEN_peaks/VIB_hydrop_2.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_21.FULL': 'SCREEN_peaks/VIB_hydrop_21.FULL__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_22.FULL': 'SCREEN_peaks/VIB_hydrop_22.FULL__SCREEN_consensus_peaks.bed'}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consensus_peaks_path_dict = {x.split('/')[-1].split(f'__')[0]: x for x in sorted(glob.glob(\"SCREEN_peaks/*consensus_peaks.bed\"))}\n",
    "consensus_peaks_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "51d41a9e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-26T14:51:18.272526Z",
     "iopub.status.busy": "2022-09-26T14:51:18.272208Z",
     "iopub.status.idle": "2022-09-26T14:51:34.957278Z",
     "shell.execute_reply": "2022-09-26T14:51:34.956591Z"
    },
    "papermill": {
     "duration": 16.696852,
     "end_time": "2022-09-26T14:51:34.960798",
     "exception": false,
     "start_time": "2022-09-26T14:51:18.263946",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BIO_ddseq_1.FULL\n",
      "\tpeaks on standard chromosomes: 165715\n",
      "\tpeaks on contigs: 242\n",
      "\tpeaks on chrM: 24\n",
      "\t% peaks non standard chromosomes: 0.16025930678812636%\n",
      "\tpeaks on chrY: 0\n",
      "BIO_ddseq_2.FULL\n",
      "\tpeaks on standard chromosomes: 153883\n",
      "\tpeaks on contigs: 81\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.06688919771927318%\n",
      "\tpeaks on chrY: 0\n",
      "BIO_ddseq_3.FULL\n",
      "\tpeaks on standard chromosomes: 191169\n",
      "\tpeaks on contigs: 290\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.1634618397551728%\n",
      "\tpeaks on chrY: 207\n",
      "BIO_ddseq_4.FULL\n",
      "\tpeaks on standard chromosomes: 174118\n",
      "\tpeaks on contigs: 264\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.1645595023078467%\n",
      "\tpeaks on chrY: 191\n",
      "BRO_mtscatac_1.FULL\n",
      "\tpeaks on standard chromosomes: 203455\n",
      "\tpeaks on contigs: 257\n",
      "\tpeaks on chrM: 21\n",
      "\t% peaks non standard chromosomes: 0.136453102835574%\n",
      "\tpeaks on chrY: 148\n",
      "BRO_mtscatac_2.FULL\n",
      "\tpeaks on standard chromosomes: 202202\n",
      "\tpeaks on contigs: 251\n",
      "\tpeaks on chrM: 20\n",
      "\t% peaks non standard chromosomes: 0.1338450064946931%\n",
      "\tpeaks on chrY: 146\n",
      "CNA_10xmultiome_1.FULL\n",
      "\tpeaks on standard chromosomes: 196639\n",
      "\tpeaks on contigs: 280\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.15334541817092429%\n",
      "\tpeaks on chrY: 115\n",
      "CNA_10xmultiome_2.FULL\n",
      "\tpeaks on standard chromosomes: 174907\n",
      "\tpeaks on contigs: 358\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.21735657888731688%\n",
      "\tpeaks on chrY: 90\n",
      "CNA_10xv11_1.FULL\n",
      "\tpeaks on standard chromosomes: 166939\n",
      "\tpeaks on contigs: 293\n",
      "\tpeaks on chrM: 24\n",
      "\t% peaks non standard chromosomes: 0.18952982254747214%\n",
      "\tpeaks on chrY: 111\n",
      "CNA_10xv11_2.FULL\n",
      "\tpeaks on standard chromosomes: 160448\n",
      "\tpeaks on contigs: 284\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.19097384218220273%\n",
      "\tpeaks on chrY: 111\n",
      "CNA_10xv11_3.FULL\n",
      "\tpeaks on standard chromosomes: 234595\n",
      "\tpeaks on contigs: 420\n",
      "\tpeaks on chrM: 24\n",
      "\t% peaks non standard chromosomes: 0.18890482005114045%\n",
      "\tpeaks on chrY: 149\n",
      "CNA_10xv11_4.FULL\n",
      "\tpeaks on standard chromosomes: 187553\n",
      "\tpeaks on contigs: 293\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.16767091787851046%\n",
      "\tpeaks on chrY: 92\n",
      "CNA_10xv11_5.FULL\n",
      "\tpeaks on standard chromosomes: 170523\n",
      "\tpeaks on contigs: 340\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.21183837083418675%\n",
      "\tpeaks on chrY: 97\n",
      "CNA_10xv2_1.FULL\n",
      "\tpeaks on standard chromosomes: 244617\n",
      "\tpeaks on contigs: 307\n",
      "\tpeaks on chrM: 24\n",
      "\t% peaks non standard chromosomes: 0.13513072162254844%\n",
      "\tpeaks on chrY: 179\n",
      "CNA_10xv2_2.FULL\n",
      "\tpeaks on standard chromosomes: 264416\n",
      "\tpeaks on contigs: 347\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.13973548450446777%\n",
      "\tpeaks on chrY: 180\n",
      "CNA_hydrop_1.FULL\n",
      "\tpeaks on standard chromosomes: 86842\n",
      "\tpeaks on contigs: 265\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.3293966417610669%\n",
      "\tpeaks on chrY: 74\n",
      "CNA_hydrop_2.FULL\n",
      "\tpeaks on standard chromosomes: 222480\n",
      "\tpeaks on contigs: 574\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.2671735193387007%\n",
      "\tpeaks on chrY: 163\n",
      "CNA_hydrop_3.FULL\n",
      "\tpeaks on standard chromosomes: 66251\n",
      "\tpeaks on contigs: 176\n",
      "\tpeaks on chrM: 21\n",
      "\t% peaks non standard chromosomes: 0.2964724295689863%\n",
      "\tpeaks on chrY: 49\n",
      "CNA_mtscatac_1.FULL\n",
      "\tpeaks on standard chromosomes: 144098\n",
      "\tpeaks on contigs: 225\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.17180940240810277%\n",
      "\tpeaks on chrY: 86\n",
      "CNA_mtscatac_2.FULL\n",
      "\tpeaks on standard chromosomes: 151231\n",
      "\tpeaks on contigs: 209\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.15251350173640912%\n",
      "\tpeaks on chrY: 74\n",
      "EPF_hydrop_1.FULL\n",
      "\tpeaks on standard chromosomes: 130746\n",
      "\tpeaks on contigs: 280\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.23044991148281546%\n",
      "\tpeaks on chrY: 55\n",
      "EPF_hydrop_2.FULL\n",
      "\tpeaks on standard chromosomes: 107130\n",
      "\tpeaks on contigs: 184\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.1919206976224193%\n",
      "\tpeaks on chrY: 39\n",
      "EPF_hydrop_3.FULL\n",
      "\tpeaks on standard chromosomes: 131776\n",
      "\tpeaks on contigs: 347\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.2792387150478641%\n",
      "\tpeaks on chrY: 61\n",
      "EPF_hydrop_4.FULL\n",
      "\tpeaks on standard chromosomes: 96902\n",
      "\tpeaks on contigs: 171\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.19877439620989754%\n",
      "\tpeaks on chrY: 44\n",
      "HAR_ddseq_1.FULL\n",
      "\tpeaks on standard chromosomes: 152046\n",
      "\tpeaks on contigs: 160\n",
      "\tpeaks on chrM: 24\n",
      "\t% peaks non standard chromosomes: 0.12086973658280234%\n",
      "\tpeaks on chrY: 83\n",
      "HAR_ddseq_2.FULL\n",
      "\tpeaks on standard chromosomes: 162752\n",
      "\tpeaks on contigs: 208\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.14173257333586936%\n",
      "\tpeaks on chrY: 103\n",
      "MDC_mtscatac_1.FULL\n",
      "\tpeaks on standard chromosomes: 251451\n",
      "\tpeaks on contigs: 584\n",
      "\tpeaks on chrM: 21\n",
      "\t% peaks non standard chromosomes: 0.24002602596248454%\n",
      "\tpeaks on chrY: 156\n",
      "MDC_mtscatac_2.FULL\n",
      "\tpeaks on standard chromosomes: 215826\n",
      "\tpeaks on contigs: 562\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.26985813964234556%\n",
      "\tpeaks on chrY: 119\n",
      "OHS_s3atac_1.FULL\n",
      "\tpeaks on standard chromosomes: 222028\n",
      "\tpeaks on contigs: 313\n",
      "\tpeaks on chrM: 21\n",
      "\t% peaks non standard chromosomes: 0.15020552072746243%\n",
      "\tpeaks on chrY: 61\n",
      "OHS_s3atac_2.FULL\n",
      "\tpeaks on standard chromosomes: 506393\n",
      "\tpeaks on contigs: 1315\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.2633289346700018%\n",
      "\tpeaks on chrY: 22\n",
      "SAN_10xmultiome_1.FULL\n",
      "\tpeaks on standard chromosomes: 288398\n",
      "\tpeaks on contigs: 412\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.1506060595569066%\n",
      "\tpeaks on chrY: 131\n",
      "SAN_10xmultiome_2.FULL\n",
      "\tpeaks on standard chromosomes: 286482\n",
      "\tpeaks on contigs: 439\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.16100702576112413%\n",
      "\tpeaks on chrY: 116\n",
      "STA_10xv11_1.FULL\n",
      "\tpeaks on standard chromosomes: 195960\n",
      "\tpeaks on contigs: 190\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.10857763300760044%\n",
      "\tpeaks on chrY: 109\n",
      "STA_10xv11_2.FULL\n",
      "\tpeaks on standard chromosomes: 226923\n",
      "\tpeaks on contigs: 228\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.11048799598545608%\n",
      "\tpeaks on chrY: 117\n",
      "TXG_10xv11_1.FULL\n",
      "\tpeaks on standard chromosomes: 248577\n",
      "\tpeaks on contigs: 456\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.19232622382114867%\n",
      "\tpeaks on chrY: 279\n",
      "TXG_10xv2_1.FULL\n",
      "\tpeaks on standard chromosomes: 347795\n",
      "\tpeaks on contigs: 526\n",
      "\tpeaks on chrM: 21\n",
      "\t% peaks non standard chromosomes: 0.15702958586676313%\n",
      "\tpeaks on chrY: 380\n",
      "TXG_10xv2_2.FULL\n",
      "\tpeaks on standard chromosomes: 344808\n",
      "\tpeaks on contigs: 536\n",
      "\tpeaks on chrM: 24\n",
      "\t% peaks non standard chromosomes: 0.16214588496907648%\n",
      "\tpeaks on chrY: 377\n",
      "UCS_ddseq_1.FULL\n",
      "\tpeaks on standard chromosomes: 170853\n",
      "\tpeaks on contigs: 281\n",
      "\tpeaks on chrM: 21\n",
      "\t% peaks non standard chromosomes: 0.17644824866349215%\n",
      "\tpeaks on chrY: 76\n",
      "UCS_ddseq_2.FULL\n",
      "\tpeaks on standard chromosomes: 199578\n",
      "\tpeaks on contigs: 370\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.1960294044106616%\n",
      "\tpeaks on chrY: 122\n",
      "VIB_10xmultiome_1.FULL\n",
      "\tpeaks on standard chromosomes: 122710\n",
      "\tpeaks on contigs: 106\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.10501550810410375%\n",
      "\tpeaks on chrY: 62\n",
      "VIB_10xmultiome_2.FULL\n",
      "\tpeaks on standard chromosomes: 167153\n",
      "\tpeaks on contigs: 166\n",
      "\tpeaks on chrM: 25\n",
      "\t% peaks non standard chromosomes: 0.11413615068362176%\n",
      "\tpeaks on chrY: 97\n",
      "VIB_10xv1_1.FULL\n",
      "\tpeaks on standard chromosomes: 165105\n",
      "\tpeaks on contigs: 195\n",
      "\tpeaks on chrM: 21\n",
      "\t% peaks non standard chromosomes: 0.1306549077249714%\n",
      "\tpeaks on chrY: 114\n",
      "VIB_10xv1_2.FULL\n",
      "\tpeaks on standard chromosomes: 210226\n",
      "\tpeaks on contigs: 380\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.19133167797406814%\n",
      "\tpeaks on chrY: 91\n",
      "VIB_10xv2_1.FULL\n",
      "\tpeaks on standard chromosomes: 229427\n",
      "\tpeaks on contigs: 220\n",
      "\tpeaks on chrM: 24\n",
      "\t% peaks non standard chromosomes: 0.10623892437443125%\n",
      "\tpeaks on chrY: 138\n",
      "VIB_10xv2_2.FULL\n",
      "\tpeaks on standard chromosomes: 259038\n",
      "\tpeaks on contigs: 252\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.10604944603625735%\n",
      "\tpeaks on chrY: 165\n",
      "VIB_hydrop_1.FULL\n",
      "\tpeaks on standard chromosomes: 113523\n",
      "\tpeaks on contigs: 113\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.11965616449203319%\n",
      "\tpeaks on chrY: 56\n",
      "VIB_hydrop_11.FULL\n",
      "\tpeaks on standard chromosomes: 110395\n",
      "\tpeaks on contigs: 127\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.13569134741507982%\n",
      "\tpeaks on chrY: 59\n",
      "VIB_hydrop_12.FULL\n",
      "\tpeaks on standard chromosomes: 79118\n",
      "\tpeaks on contigs: 75\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.12245155589219213%\n",
      "\tpeaks on chrY: 38\n",
      "VIB_hydrop_2.FULL\n",
      "\tpeaks on standard chromosomes: 127542\n",
      "\tpeaks on contigs: 125\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.11590570913932179%\n",
      "\tpeaks on chrY: 58\n",
      "VIB_hydrop_21.FULL\n",
      "\tpeaks on standard chromosomes: 99146\n",
      "\tpeaks on contigs: 97\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.12088731287651362%\n",
      "\tpeaks on chrY: 46\n",
      "VIB_hydrop_22.FULL\n",
      "\tpeaks on standard chromosomes: 94547\n",
      "\tpeaks on contigs: 76\n",
      "\tpeaks on chrM: 24\n",
      "\t% peaks non standard chromosomes: 0.10565575242744091%\n",
      "\tpeaks on chrY: 35\n"
     ]
    }
   ],
   "source": [
    "for sample, path in consensus_peaks_path_dict.items():\n",
    "    print(sample)\n",
    "    peaks_df = pd.read_csv(path, sep='\\t', header=None)\n",
    "    \n",
    "    chroms_in_df = list(sorted(peaks_df[0].unique()))\n",
    "    chroms_standard = [\"chr\" + str(x + 1) for x in range(22)] + ['chrX']\n",
    "    chroms_nonstandard = list(set(chroms_in_df ) - set(chroms_standard) -  set(['chrM']))\n",
    "    \n",
    "    n_standard = peaks_df[0].value_counts()[chroms_standard].sum()\n",
    "    n_contigs = peaks_df[0].value_counts()[chroms_nonstandard].sum()\n",
    "    n_chrm = peaks_df[0].value_counts()['chrM'].sum()\n",
    "    pct_nonstandard = (n_contigs + n_chrm)/len(peaks_df)*100\n",
    "        \n",
    "    print(f\"\\tpeaks on standard chromosomes: {n_standard}\")\n",
    "    print(f\"\\tpeaks on contigs: {n_contigs}\")\n",
    "    print(f\"\\tpeaks on chrM: {n_chrm}\")\n",
    "    print(f\"\\t% peaks non standard chromosomes: {pct_nonstandard}%\")\n",
    "    if 'chrY' in chroms_in_df:\n",
    "        n_chrY = peaks_df[0].value_counts()['chrY'].sum()\n",
    "        print(f\"\\tpeaks on chrY: {n_chrY}\")\n",
    "    else:\n",
    "        print(f\"\\tpeaks on chrY: 0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "51ab9f8d",
   "metadata": {
    "papermill": {
     "duration": 0.009995,
     "end_time": "2022-09-26T14:51:34.986388",
     "exception": false,
     "start_time": "2022-09-26T14:51:34.976393",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9b9503a",
   "metadata": {
    "papermill": {
     "duration": 0.009863,
     "end_time": "2022-09-26T14:51:35.008446",
     "exception": false,
     "start_time": "2022-09-26T14:51:34.998583",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5d8c51e",
   "metadata": {
    "papermill": {
     "duration": 0.009449,
     "end_time": "2022-09-26T14:51:35.027247",
     "exception": false,
     "start_time": "2022-09-26T14:51:35.017798",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "20220628_pycistopic.sif",
   "language": "python",
   "name": "cistopic_20220628"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 2403.086769,
   "end_time": "2022-09-26T14:51:38.739647",
   "environment_variables": {},
   "exception": null,
   "input_path": "8_consensus_peaks.ipynb",
   "output_path": "8_consensus_peaks.OUT.ipynb",
   "parameters": {},
   "start_time": "2022-09-26T14:11:35.652878",
   "version": "2.3.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
