{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b324373a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:06:21.249914Z",
     "iopub.status.busy": "2023-02-16T09:06:21.249004Z",
     "iopub.status.idle": "2023-02-16T09:06:21.261573Z",
     "shell.execute_reply": "2023-02-16T09:06:21.260042Z",
     "shell.execute_reply.started": "2023-02-16T09:06:21.249852Z"
    },
    "papermill": {
     "duration": 9.289086,
     "end_time": "2022-09-18T16:28:06.123108",
     "exception": false,
     "start_time": "2022-09-18T16:27:56.834022",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pycisTopic.pseudobulk_peak_calling import export_pseudobulk, peak_calling\n",
    "import pyranges as pr\n",
    "import requests\n",
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "5c7a63b8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:06:21.265551Z",
     "iopub.status.busy": "2023-02-16T09:06:21.264409Z",
     "iopub.status.idle": "2023-02-16T09:06:21.294934Z",
     "shell.execute_reply": "2023-02-16T09:06:21.293420Z",
     "shell.execute_reply.started": "2023-02-16T09:06:21.265494Z"
    },
    "papermill": {
     "duration": 0.302426,
     "end_time": "2022-09-18T16:28:06.459310",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.156884",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The lab_black extension is already loaded. To reload it, use:\n",
      "  %reload_ext lab_black\n"
     ]
    }
   ],
   "source": [
    "%load_ext lab_black"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7c2ab493",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:06:21.297859Z",
     "iopub.status.busy": "2023-02-16T09:06:21.297112Z",
     "iopub.status.idle": "2023-02-16T09:06:21.346860Z",
     "shell.execute_reply": "2023-02-16T09:06:21.345464Z",
     "shell.execute_reply.started": "2023-02-16T09:06:21.297804Z"
    },
    "papermill": {
     "duration": 0.031625,
     "end_time": "2022-09-18T16:28:06.495930",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.464305",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# get chromosome sizes (hg38)\n",
    "if not os.path.exists(\"chromsizes.txt\"):\n",
    "    target_url = (\n",
    "        \"http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes\"\n",
    "    )\n",
    "    chromsizes = pd.read_csv(target_url, sep=\"\\t\", header=None)\n",
    "    chromsizes.columns = [\"Chromosome\", \"End\"]\n",
    "    chromsizes[\"Start\"] = [0] * chromsizes.shape[0]\n",
    "    chromsizes = chromsizes.loc[:, [\"Chromosome\", \"Start\", \"End\"]]\n",
    "    chromsizes = pr.PyRanges(chromsizes)\n",
    "    chromsizes.to_csv(\"chromsizes.txt\")\n",
    "    chromsizes\n",
    "else:\n",
    "    chromsizes = pd.read_csv(\"chromsizes.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "0724339f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:06:21.350298Z",
     "iopub.status.busy": "2023-02-16T09:06:21.349740Z",
     "iopub.status.idle": "2023-02-16T09:06:21.362561Z",
     "shell.execute_reply": "2023-02-16T09:06:21.361429Z",
     "shell.execute_reply.started": "2023-02-16T09:06:21.350262Z"
    },
    "papermill": {
     "duration": 0.020844,
     "end_time": "2022-09-18T16:28:06.520860",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.500016",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'master_all_1.FIXEDCELLS': 'cistopic_objects_master/master_all_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_24topics.dimreduc.consensus_harmony.pkl'}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cto_consensus_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(\n",
    "        glob.glob(\"cistopic_objects_master/*singlets.model*consensus_harmony.pkl\")\n",
    "    )\n",
    "}\n",
    "cto_consensus_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ec49919c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:06:21.364854Z",
     "iopub.status.busy": "2023-02-16T09:06:21.364410Z",
     "iopub.status.idle": "2023-02-16T09:06:21.450521Z",
     "shell.execute_reply": "2023-02-16T09:06:21.448985Z",
     "shell.execute_reply.started": "2023-02-16T09:06:21.364824Z"
    },
    "papermill": {
     "duration": 0.197783,
     "end_time": "2022-09-18T16:28:06.740933",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.543150",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import gc\n",
    "import logging\n",
    "import os\n",
    "import re\n",
    "import subprocess\n",
    "import sys\n",
    "from typing import Dict, List, Optional, Union\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyBigWig\n",
    "import pyranges as pr\n",
    "import ray\n",
    "\n",
    "from pycisTopic.cistopic_class import *\n",
    "from pycisTopic.utils import *\n",
    "\n",
    "\n",
    "def export_pseudobulk(\n",
    "    input_data: Union[\"CistopicObject\", pd.DataFrame, Dict[str, pd.DataFrame]],\n",
    "    variable: str,\n",
    "    chromsizes: Union[pd.DataFrame, pr.PyRanges],\n",
    "    bed_path: str,\n",
    "    bigwig_path: str,\n",
    "    path_to_fragments: Optional[Dict[str, str]] = None,\n",
    "    sample_id_col: Optional[str] = \"sample_id\",\n",
    "    n_cpu: Optional[int] = 1,\n",
    "    normalize_bigwig: Optional[bool] = True,\n",
    "    remove_duplicates: Optional[bool] = True,\n",
    "    split_pattern: Optional[str] = \"___\",\n",
    "    use_polars: Optional[bool] = True,\n",
    "    **kwargs\n",
    "):\n",
    "    \"\"\"\n",
    "    Create pseudobulks as bed and bigwig from single cell fragments file given a barcode annotation.\n",
    "    Parameters\n",
    "    ---------\n",
    "    input_data: CistopicObject or pd.DataFrame\n",
    "            A :class:`CistopicObject` containing the specified `variable` as a column in :class:`CistopicObject.cell_data` or a cell metadata\n",
    "            :class:`pd.DataFrame` containing barcode as rows, containing the specified `variable` as a column (additional columns are\n",
    "            possible) and a `sample_id` column. Index names must contain the BARCODE (e.g. ATGTCGTC-1), additional tags are possible separating with -\n",
    "            (e.g. ATGCTGTGCG-1-Sample_1). The levels in the sample_id column must agree with the keys in the path_to_fragments dictionary.\n",
    "            Alternatively, if the cell metadata contains a column named barcode it will be used instead of the index names.\n",
    "    variable: str\n",
    "            A character string indicating the column that will be used to create the different group pseudobulk. It must be included in\n",
    "            the cell metadata provided as input_data.\n",
    "    chromsizes: pd.DataFrame or pr.PyRanges\n",
    "            A data frame or :class:`pr.PyRanges` containing size of each chromosome, containing 'Chromosome', 'Start' and 'End' columns.\n",
    "    bed_path: str\n",
    "            Path to folder where the fragments bed files per group will be saved. If None, files will not be generated.\n",
    "    bigwig_path: str\n",
    "            Path to folder where the bigwig files per group will be saved. If None, files will not be generated.\n",
    "    path_to_fragments: str or dict, optional\n",
    "            A dictionary of character strings, with sample name as names indicating the path to the fragments file/s from which pseudobulk profiles have to\n",
    "            be created. If a :class:`CistopicObject` is provided as input it will be ignored, but if a cell metadata :class:`pd.DataFrame` is provided it\n",
    "            is necessary to provide it. The keys of the dictionary need to match with the sample_id tag added to the index names of the input data frame.\n",
    "    sample_id_col: str, optional\n",
    "            Name of the column containing the sample name per barcode in the input :class:`CistopicObject.cell_data` or class:`pd.DataFrame`. Default: 'sample_id'.\n",
    "    n_cpu: int, optional\n",
    "            Number of cores to use. Default: 1.\n",
    "    normalize_bigwig: bool, optional\n",
    "            Whether bigwig files should be CPM normalized. Default: True.\n",
    "    remove_duplicates: bool, optional\n",
    "            Whether duplicates should be removed before converting the data to bigwig.\n",
    "    split_pattern: str, optional\n",
    "            Pattern to split cell barcode from sample id. Default: ___ .\n",
    "    use_polars: bool, optional\n",
    "            Whether to use polars to read fragments files. Default: True.\n",
    "    **kwargs\n",
    "            Additional parameters for ray.init()\n",
    "    Return\n",
    "    ------\n",
    "    dict\n",
    "            A dictionary containing the paths to the newly created bed fragments files per group a dictionary containing the paths to the\n",
    "            newly created bigwig files per group.\n",
    "    \"\"\"\n",
    "    # Create logger\n",
    "    level = logging.INFO\n",
    "    log_format = \"%(asctime)s %(name)-12s %(levelname)-8s %(message)s\"\n",
    "    handlers = [logging.StreamHandler(stream=sys.stdout)]\n",
    "    logging.basicConfig(level=level, format=log_format, handlers=handlers)\n",
    "    log = logging.getLogger(\"cisTopic\")\n",
    "\n",
    "    # Get fragments file\n",
    "    if isinstance(input_data, CistopicObject):\n",
    "        path_to_fragments = input_data.path_to_fragments\n",
    "        if path_to_fragments is None:\n",
    "            log.error(\"No path_to_fragments in this cisTopic object.\")\n",
    "        cell_data = input_data.cell_data\n",
    "    elif isinstance(input_data, pd.DataFrame):\n",
    "        if path_to_fragments is None:\n",
    "            log.error(\"Please, provide path_to_fragments.\")\n",
    "        cell_data = input_data\n",
    "    # Check for sample_id column\n",
    "    try:\n",
    "        sample_ids = list(set(cell_data[sample_id_col]))\n",
    "    except ValueError:\n",
    "        print(\n",
    "            'Please, include a sample identification column (e.g. \"sample_id\") in your cell metadata!'\n",
    "        )\n",
    "\n",
    "    # Get fragments\n",
    "    fragments_df_dict = {}\n",
    "    for sample_id in path_to_fragments.keys():\n",
    "        if sample_id not in sample_ids:\n",
    "            log.info(\n",
    "                \"The following path_to_fragments entry is not found in the cell metadata sample_id_col: \",\n",
    "                sample_id,\n",
    "                \". It will be ignored.\",\n",
    "            )\n",
    "        else:\n",
    "            log.info(\"Reading fragments from \" + path_to_fragments[sample_id])\n",
    "            fragments_df = read_fragments_from_file(\n",
    "                path_to_fragments[sample_id], use_polars=use_polars\n",
    "            ).df\n",
    "            # Convert to int32 for memory efficiency\n",
    "            fragments_df.Start = np.int32(fragments_df.Start)\n",
    "            fragments_df.End = np.int32(fragments_df.End)\n",
    "            if \"Score\" in fragments_df:\n",
    "                fragments_df.Score = np.int32(fragments_df.Score)\n",
    "            if \"barcode\" in cell_data:\n",
    "                fragments_df = fragments_df.loc[\n",
    "                    fragments_df[\"Name\"].isin(cell_data[\"barcode\"].tolist())\n",
    "                ]\n",
    "            else:\n",
    "                fragments_df = fragments_df.loc[\n",
    "                    fragments_df[\"Name\"].isin(\n",
    "                        prepare_tag_cells(cell_data.index.tolist(), split_pattern)\n",
    "                    )\n",
    "                ]\n",
    "            fragments_df_dict[sample_id] = fragments_df\n",
    "\n",
    "    # Set groups\n",
    "    if \"barcode\" in cell_data:\n",
    "        cell_data = cell_data.loc[:, [variable, sample_id_col, \"barcode\"]]\n",
    "    else:\n",
    "        cell_data = cell_data.loc[:, [variable, sample_id_col]]\n",
    "    cell_data[variable] = cell_data[variable].replace(\" \", \"\", regex=True)\n",
    "    cell_data[variable] = cell_data[variable].replace(\"[^A-Za-z0-9]+\", \"_\", regex=True)\n",
    "    groups = sorted(list(set(cell_data[variable])))\n",
    "    # Check chromosome sizes\n",
    "    if isinstance(chromsizes, pd.DataFrame):\n",
    "        chromsizes = chromsizes.loc[:, [\"Chromosome\", \"Start\", \"End\"]]\n",
    "        chromsizes = pr.PyRanges(chromsizes)\n",
    "    # Check that output dir exist and generate output paths\n",
    "    if isinstance(bed_path, str):\n",
    "        if not os.path.exists(bed_path):\n",
    "            os.makedirs(bed_path)\n",
    "        bed_paths = {\n",
    "            group: os.path.join(bed_path, str(group) + \".bed.gz\") for group in groups\n",
    "        }\n",
    "    else:\n",
    "        bed_paths = {}\n",
    "    if isinstance(bigwig_path, str):\n",
    "        if not os.path.exists(bigwig_path):\n",
    "            os.makedirs(bigwig_path)\n",
    "        bw_paths = {\n",
    "            group: os.path.join(bigwig_path, str(group) + \".bw\") for group in groups\n",
    "        }\n",
    "    else:\n",
    "        bw_paths = {}\n",
    "    # Create pseudobulks\n",
    "    if n_cpu > 1:\n",
    "        ray.init(num_cpus=n_cpu, **kwargs)\n",
    "        ray_handle = ray.wait(\n",
    "            [\n",
    "                export_pseudobulk_ray.remote(\n",
    "                    cell_data,\n",
    "                    group,\n",
    "                    fragments_df_dict,\n",
    "                    chromsizes,\n",
    "                    bigwig_path,\n",
    "                    bed_path,\n",
    "                    sample_id_col,\n",
    "                    normalize_bigwig,\n",
    "                    remove_duplicates,\n",
    "                    split_pattern,\n",
    "                )\n",
    "                for group in groups\n",
    "            ],\n",
    "            num_returns=len(groups),\n",
    "        )\n",
    "        ray.shutdown()\n",
    "    else:\n",
    "        [\n",
    "            export_pseudobulk_one_sample(\n",
    "                cell_data,\n",
    "                group,\n",
    "                fragments_df_dict,\n",
    "                chromsizes,\n",
    "                bigwig_path,\n",
    "                bed_path,\n",
    "                sample_id_col,\n",
    "                normalize_bigwig,\n",
    "                remove_duplicates,\n",
    "                split_pattern,\n",
    "            )\n",
    "            for group in groups\n",
    "        ]\n",
    "\n",
    "    return bw_paths, bed_paths\n",
    "\n",
    "\n",
    "def export_pseudobulk_one_sample(\n",
    "    cell_data: pd.DataFrame,\n",
    "    group: str,\n",
    "    fragments_df_dict: Dict[str, pd.DataFrame],\n",
    "    chromsizes: pr.PyRanges,\n",
    "    bigwig_path: str,\n",
    "    bed_path: str,\n",
    "    sample_id_col: Optional[str] = \"sample_id\",\n",
    "    normalize_bigwig: Optional[bool] = True,\n",
    "    remove_duplicates: Optional[bool] = True,\n",
    "    split_pattern: Optional[str] = \"___\",\n",
    "):\n",
    "    \"\"\"\n",
    "    Create pseudobulk as bed and bigwig from single cell fragments file given a barcode annotation and a group.\n",
    "    Parameters\n",
    "    ---------\n",
    "    cell_data: pd.DataFrame\n",
    "            A cell metadata :class:`pd.Dataframe` containing barcodes, their annotation and their sample of origin.\n",
    "    group: str\n",
    "            A character string indicating the group for which pseudobulks will be created.\n",
    "    fragments_df_dict: dict\n",
    "            A dictionary containing data frames as values with 'Chromosome', 'Start', 'End', 'Name', and 'Score' as columns; and sample label\n",
    "            as keys. 'Score' indicates the number of times that a fragments is found assigned to that barcode.\n",
    "    chromsizes: pr.PyRanges\n",
    "            A :class:`pr.PyRanges` containing size of each column, containing 'Chromosome', 'Start' and 'End' columns.\n",
    "    bigwig_path: str\n",
    "            Path to folder where the bigwig file will be saved.\n",
    "    bed_path: str\n",
    "            Path to folder where the fragments bed file will be saved.\n",
    "    sample_id_col: str, optional\n",
    "            Name of the column containing the sample name per barcode in the input :class:`CistopicObject.cell_data` or class:`pd.DataFrame`. Default: 'sample_id'.\n",
    "    normalize_bigwig: bool, optional\n",
    "            Whether bigwig files should be CPM normalized. Default: True.\n",
    "    remove_duplicates: bool, optional\n",
    "            Whether duplicates should be removed before converting the data to bigwig.\n",
    "    split_pattern: str\n",
    "            Pattern to split cell barcode from sample id. Default: ___ .\n",
    "    \"\"\"\n",
    "    # Create logger\n",
    "    level = logging.INFO\n",
    "    log_format = \"%(asctime)s %(name)-12s %(levelname)-8s %(message)s\"\n",
    "    handlers = [logging.StreamHandler(stream=sys.stdout)]\n",
    "    logging.basicConfig(level=level, format=log_format, handlers=handlers)\n",
    "    log = logging.getLogger(\"cisTopic\")\n",
    "\n",
    "    log.info(\"Creating pseudobulk for \" + str(group))\n",
    "    group_fragments_list = []\n",
    "    group_fragments_dict = {}\n",
    "    for sample_id in fragments_df_dict:\n",
    "        sample_data = cell_data[cell_data.loc[:, sample_id_col].isin([sample_id])]\n",
    "        if \"barcode\" in sample_data:\n",
    "            sample_data.index = sample_data[\"barcode\"].tolist()\n",
    "        else:\n",
    "            sample_data.index = prepare_tag_cells(\n",
    "                sample_data.index.tolist(), split_pattern\n",
    "            )\n",
    "        group_var = sample_data.iloc[:, 0]\n",
    "        barcodes = group_var[group_var.isin([group])].index.tolist()\n",
    "        fragments_df = fragments_df_dict[sample_id]\n",
    "        group_fragments = fragments_df.loc[fragments_df[\"Name\"].isin(barcodes)]\n",
    "        if len(fragments_df_dict) > 1:\n",
    "            group_fragments_dict[sample_id] = group_fragments\n",
    "\n",
    "    if len(fragments_df_dict) > 1:\n",
    "        group_fragments_list = [\n",
    "            group_fragments_dict[list(group_fragments_dict.keys())[x]]\n",
    "            for x in range(len(fragments_df_dict))\n",
    "        ]\n",
    "        group_fragments = group_fragments_list[0].append(group_fragments_list[1:])\n",
    "\n",
    "    group_fragments = group_fragments[\n",
    "        group_fragments[\"Chromosome\"].isin(chromsizes.Chromosome)\n",
    "    ]\n",
    "\n",
    "    del group_fragments_dict\n",
    "    del group_fragments_list\n",
    "    del fragments_df\n",
    "    gc.collect()\n",
    "\n",
    "    group_pr = pr.PyRanges(group_fragments)\n",
    "    if isinstance(bigwig_path, str):\n",
    "        bigwig_path_group = os.path.join(bigwig_path, str(group) + \".bw\")\n",
    "        if remove_duplicates:\n",
    "            group_pr.to_bigwig(\n",
    "                path=bigwig_path_group,\n",
    "                chromosome_sizes=chromsizes,\n",
    "                rpm=normalize_bigwig,\n",
    "            )\n",
    "        else:\n",
    "            group_pr.to_bigwig(\n",
    "                path=bigwig_path_group,\n",
    "                chromosome_sizes=chromsizes,\n",
    "                rpm=normalize_bigwig,\n",
    "                value_col=\"Score\",\n",
    "            )\n",
    "    if isinstance(bed_path, str):\n",
    "        bed_path_group = os.path.join(bed_path, str(group) + \".bed.gz\")\n",
    "        group_pr.to_bed(\n",
    "            path=bed_path_group, keep=False, compression=\"infer\", chain=False\n",
    "        )\n",
    "\n",
    "    log.info(str(group) + \" done!\")\n",
    "\n",
    "\n",
    "@ray.remote\n",
    "def export_pseudobulk_ray(\n",
    "    cell_data: pd.DataFrame,\n",
    "    group: str,\n",
    "    fragments_df_dict: Dict[str, pd.DataFrame],\n",
    "    chromsizes: pr.PyRanges,\n",
    "    bigwig_path: str,\n",
    "    bed_path: str,\n",
    "    sample_id_col: Optional[str] = \"sample_id\",\n",
    "    normalize_bigwig: Optional[bool] = True,\n",
    "    remove_duplicates: Optional[bool] = True,\n",
    "    split_pattern: Optional[str] = \"___\",\n",
    "):\n",
    "    \"\"\"\n",
    "    Create pseudobulk as bed and bigwig from single cell fragments file given a barcode annotation and a group.\n",
    "    Parameters\n",
    "    ---------\n",
    "    cell_data: pd.DataFrame\n",
    "            A cell metadata :class:`pd.Dataframe` containing barcodes, their annotation and their sample of origin.\n",
    "    group: str\n",
    "            A character string indicating the group for which pseudobulks will be created.\n",
    "    fragments_df_dict: dict\n",
    "            A dictionary containing data frames as values with 'Chromosome', 'Start', 'End', 'Name', and 'Score' as columns; and sample label\n",
    "            as keys. 'Score' indicates the number of times that a fragments is found assigned to that barcode.\n",
    "    chromsizes: pr.PyRanges\n",
    "            A :class:`pr.PyRanges` containing size of each column, containing 'Chromosome', 'Start' and 'End' columns.\n",
    "    bed_path: str\n",
    "            Path to folder where the fragments bed file will be saved.\n",
    "    bigwig_path: str\n",
    "            Path to folder where the bigwig file will be saved.\n",
    "    sample_id_col: str, optional\n",
    "            Name of the column containing the sample name per barcode in the input :class:`CistopicObject.cell_data` or class:`pd.DataFrame`. Default: 'sample_id'.\n",
    "    normalize_bigwig: bool, optional\n",
    "            Whether bigwig files should be CPM normalized. Default: True.\n",
    "    remove_duplicates: bool, optional\n",
    "            Whether duplicates should be removed before converting the data to bigwig.\n",
    "    split_pattern: str\n",
    "            Pattern to split cell barcode from sample id. Default: ___ .\n",
    "    \"\"\"\n",
    "    export_pseudobulk_one_sample(\n",
    "        cell_data,\n",
    "        group,\n",
    "        fragments_df_dict,\n",
    "        chromsizes,\n",
    "        bigwig_path,\n",
    "        bed_path,\n",
    "        sample_id_col,\n",
    "        normalize_bigwig,\n",
    "        remove_duplicates,\n",
    "        split_pattern,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "76efffdc-cd08-4bcd-a3fb-4e6d1cea7256",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:06:21.451306Z",
     "iopub.status.busy": "2023-02-16T09:06:21.451102Z",
     "iopub.status.idle": "2023-02-16T09:06:21.456332Z",
     "shell.execute_reply": "2023-02-16T09:06:21.455384Z",
     "shell.execute_reply.started": "2023-02-16T09:06:21.451288Z"
    }
   },
   "outputs": [],
   "source": [
    "sample = \"master_all_1.FIXEDCELLS\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "7aa069a8-b245-4ed5-9406-795e57b4f3a6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:06:21.457145Z",
     "iopub.status.busy": "2023-02-16T09:06:21.456944Z",
     "iopub.status.idle": "2023-02-16T09:06:56.223845Z",
     "shell.execute_reply": "2023-02-16T09:06:56.222629Z",
     "shell.execute_reply.started": "2023-02-16T09:06:21.457127Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "cto_path = cto_consensus_path_dict[sample]\n",
    "with open(cto_path, \"rb\") as f:\n",
    "    cto = pickle.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1a8bf1b5-d47f-4eee-a986-7c48846fd117",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:06:56.225221Z",
     "iopub.status.busy": "2023-02-16T09:06:56.224953Z",
     "iopub.status.idle": "2023-02-16T09:06:56.365149Z",
     "shell.execute_reply": "2023-02-16T09:06:56.364032Z",
     "shell.execute_reply.started": "2023-02-16T09:06:56.225199Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cisTopic_log_nr_frag</th>\n",
       "      <th>cisTopic_nr_frag</th>\n",
       "      <th>cisTopic_log_nr_acc</th>\n",
       "      <th>cisTopic_nr_acc</th>\n",
       "      <th>Log_total_nr_frag</th>\n",
       "      <th>Log_unique_nr_frag</th>\n",
       "      <th>Total_nr_frag</th>\n",
       "      <th>Unique_nr_frag</th>\n",
       "      <th>Dupl_nr_frag</th>\n",
       "      <th>Dupl_rate</th>\n",
       "      <th>...</th>\n",
       "      <th>centre_tech</th>\n",
       "      <th>Log_unique_nr_frag_in_regions</th>\n",
       "      <th>consensus_cell_type_individual</th>\n",
       "      <th>seurat_cell_type_pred_score</th>\n",
       "      <th>cell_type__mega</th>\n",
       "      <th>seurat_cell_type_pred_score__mega</th>\n",
       "      <th>consensus_cell_type__mega</th>\n",
       "      <th>seurat_cell_type__mega</th>\n",
       "      <th>harmony_pycisTopic_leiden_10_3.0</th>\n",
       "      <th>harmony_consensus_cell_type__mega</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXEDCELLS</th>\n",
       "      <td>3.436481</td>\n",
       "      <td>2732</td>\n",
       "      <td>3.321598</td>\n",
       "      <td>2097</td>\n",
       "      <td>4.540492</td>\n",
       "      <td>4.520903</td>\n",
       "      <td>34713</td>\n",
       "      <td>33182</td>\n",
       "      <td>1531</td>\n",
       "      <td>0.044105</td>\n",
       "      <td>...</td>\n",
       "      <td>OHS_s3atac</td>\n",
       "      <td>7.627057</td>\n",
       "      <td>B cell</td>\n",
       "      <td>0.724987</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>0.417394</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>43</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXEDCELLS</th>\n",
       "      <td>3.295567</td>\n",
       "      <td>1975</td>\n",
       "      <td>2.93852</td>\n",
       "      <td>868</td>\n",
       "      <td>4.001474</td>\n",
       "      <td>3.981003</td>\n",
       "      <td>10034</td>\n",
       "      <td>9572</td>\n",
       "      <td>462</td>\n",
       "      <td>0.046043</td>\n",
       "      <td>...</td>\n",
       "      <td>OHS_s3atac</td>\n",
       "      <td>7.431892</td>\n",
       "      <td>CD4+ T cell</td>\n",
       "      <td>0.376578</td>\n",
       "      <td>B cell</td>\n",
       "      <td>0.342953</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>B cell</td>\n",
       "      <td>20</td>\n",
       "      <td>CD4+ T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXEDCELLS</th>\n",
       "      <td>3.398287</td>\n",
       "      <td>2502</td>\n",
       "      <td>3.242541</td>\n",
       "      <td>1748</td>\n",
       "      <td>4.226136</td>\n",
       "      <td>4.202870</td>\n",
       "      <td>16832</td>\n",
       "      <td>15954</td>\n",
       "      <td>878</td>\n",
       "      <td>0.052163</td>\n",
       "      <td>...</td>\n",
       "      <td>OHS_s3atac</td>\n",
       "      <td>7.489971</td>\n",
       "      <td>CD4+ T cell</td>\n",
       "      <td>0.531326</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>0.391198</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>59</td>\n",
       "      <td>Natural killer cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXEDCELLS</th>\n",
       "      <td>3.845284</td>\n",
       "      <td>7003</td>\n",
       "      <td>3.750431</td>\n",
       "      <td>5629</td>\n",
       "      <td>4.480108</td>\n",
       "      <td>4.457746</td>\n",
       "      <td>30207</td>\n",
       "      <td>28691</td>\n",
       "      <td>1516</td>\n",
       "      <td>0.050187</td>\n",
       "      <td>...</td>\n",
       "      <td>OHS_s3atac</td>\n",
       "      <td>8.672828</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B cell</td>\n",
       "      <td>0.327305</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>B cell</td>\n",
       "      <td>76</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXEDCELLS</th>\n",
       "      <td>3.206286</td>\n",
       "      <td>1608</td>\n",
       "      <td>3.068928</td>\n",
       "      <td>1172</td>\n",
       "      <td>3.818688</td>\n",
       "      <td>3.796227</td>\n",
       "      <td>6587</td>\n",
       "      <td>6255</td>\n",
       "      <td>332</td>\n",
       "      <td>0.050402</td>\n",
       "      <td>...</td>\n",
       "      <td>OHS_s3atac</td>\n",
       "      <td>7.195937</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>0.441366</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>0.375826</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>65</td>\n",
       "      <td>Natural killer cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS</th>\n",
       "      <td>3.738543</td>\n",
       "      <td>5477</td>\n",
       "      <td>3.263873</td>\n",
       "      <td>1836</td>\n",
       "      <td>4.283573</td>\n",
       "      <td>4.080951</td>\n",
       "      <td>19212</td>\n",
       "      <td>12049</td>\n",
       "      <td>7163</td>\n",
       "      <td>0.372840</td>\n",
       "      <td>...</td>\n",
       "      <td>MDC_mtscatac</td>\n",
       "      <td>7.414573</td>\n",
       "      <td>CD4+ T cell</td>\n",
       "      <td>0.309143</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>0.388897</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>0</td>\n",
       "      <td>CD4+ T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS</th>\n",
       "      <td>3.416308</td>\n",
       "      <td>2608</td>\n",
       "      <td>2.569374</td>\n",
       "      <td>371</td>\n",
       "      <td>3.753353</td>\n",
       "      <td>3.514415</td>\n",
       "      <td>5667</td>\n",
       "      <td>3269</td>\n",
       "      <td>2398</td>\n",
       "      <td>0.423152</td>\n",
       "      <td>...</td>\n",
       "      <td>MDC_mtscatac</td>\n",
       "      <td>6.028279</td>\n",
       "      <td>CD4+ T cell</td>\n",
       "      <td>0.402943</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>0.432030</td>\n",
       "      <td>CD4+ T cell</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>43</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS</th>\n",
       "      <td>3.26031</td>\n",
       "      <td>1821</td>\n",
       "      <td>3.084934</td>\n",
       "      <td>1216</td>\n",
       "      <td>3.563006</td>\n",
       "      <td>3.408410</td>\n",
       "      <td>3656</td>\n",
       "      <td>2561</td>\n",
       "      <td>1095</td>\n",
       "      <td>0.299508</td>\n",
       "      <td>...</td>\n",
       "      <td>MDC_mtscatac</td>\n",
       "      <td>7.017506</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>0.450904</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>0.412242</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>8</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS</th>\n",
       "      <td>3.425371</td>\n",
       "      <td>2663</td>\n",
       "      <td>2.921686</td>\n",
       "      <td>835</td>\n",
       "      <td>4.046729</td>\n",
       "      <td>3.840984</td>\n",
       "      <td>11136</td>\n",
       "      <td>6934</td>\n",
       "      <td>4202</td>\n",
       "      <td>0.377335</td>\n",
       "      <td>...</td>\n",
       "      <td>MDC_mtscatac</td>\n",
       "      <td>6.751101</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>0.334340</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>0.395910</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>54</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS</th>\n",
       "      <td>3.294466</td>\n",
       "      <td>1970</td>\n",
       "      <td>2.884229</td>\n",
       "      <td>766</td>\n",
       "      <td>4.079579</td>\n",
       "      <td>3.871981</td>\n",
       "      <td>12011</td>\n",
       "      <td>7447</td>\n",
       "      <td>4564</td>\n",
       "      <td>0.379985</td>\n",
       "      <td>...</td>\n",
       "      <td>MDC_mtscatac</td>\n",
       "      <td>6.907755</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>0.390312</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>0.393253</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "      <td>43</td>\n",
       "      <td>Cytotoxic T cell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>169227 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   cisTopic_log_nr_frag  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...             3.436481   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...             3.295567   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...             3.398287   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...             3.845284   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...             3.206286   \n",
       "...                                                                 ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                   3.738543   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                   3.416308   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                    3.26031   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                   3.425371   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                   3.294466   \n",
       "\n",
       "                                                   cisTopic_nr_frag  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...             2732   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...             1975   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...             2502   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...             7003   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...             1608   \n",
       "...                                                             ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                   5477   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                   2608   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                   1821   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                   2663   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                   1970   \n",
       "\n",
       "                                                   cisTopic_log_nr_acc  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...            3.321598   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...             2.93852   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...            3.242541   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...            3.750431   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...            3.068928   \n",
       "...                                                                ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                  3.263873   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                  2.569374   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                  3.084934   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                  2.921686   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                  2.884229   \n",
       "\n",
       "                                                   cisTopic_nr_acc  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...            2097   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...             868   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...            1748   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...            5629   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...            1172   \n",
       "...                                                            ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                  1836   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                   371   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                  1216   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                   835   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                   766   \n",
       "\n",
       "                                                    Log_total_nr_frag  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...           4.540492   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...           4.001474   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...           4.226136   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...           4.480108   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...           3.818688   \n",
       "...                                                               ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                 4.283573   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                 3.753353   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                 3.563006   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                 4.046729   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                 4.079579   \n",
       "\n",
       "                                                    Log_unique_nr_frag  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...            4.520903   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...            3.981003   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...            4.202870   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...            4.457746   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...            3.796227   \n",
       "...                                                                ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                  4.080951   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                  3.514415   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                  3.408410   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                  3.840984   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                  3.871981   \n",
       "\n",
       "                                                    Total_nr_frag  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...          34713   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...          10034   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...          16832   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...          30207   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...           6587   \n",
       "...                                                           ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                19212   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                 5667   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                 3656   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                11136   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                12011   \n",
       "\n",
       "                                                    Unique_nr_frag  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...           33182   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...            9572   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...           15954   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...           28691   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...            6255   \n",
       "...                                                            ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                 12049   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                  3269   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                  2561   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                  6934   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                  7447   \n",
       "\n",
       "                                                    Dupl_nr_frag  Dupl_rate  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...          1531   0.044105   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...           462   0.046043   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...           878   0.052163   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...          1516   0.050187   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...           332   0.050402   \n",
       "...                                                          ...        ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                7163   0.372840   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                2398   0.423152   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                1095   0.299508   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                4202   0.377335   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                4564   0.379985   \n",
       "\n",
       "                                                    ...   centre_tech  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...  ...    OHS_s3atac   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...  ...    OHS_s3atac   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...  ...    OHS_s3atac   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...  ...    OHS_s3atac   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...  ...    OHS_s3atac   \n",
       "...                                                 ...           ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS        ...  MDC_mtscatac   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS        ...  MDC_mtscatac   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS        ...  MDC_mtscatac   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS        ...  MDC_mtscatac   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS        ...  MDC_mtscatac   \n",
       "\n",
       "                                                    Log_unique_nr_frag_in_regions  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...                       7.627057   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...                       7.431892   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...                       7.489971   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...                       8.672828   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...                       7.195937   \n",
       "...                                                                           ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                             7.414573   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                             6.028279   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                             7.017506   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                             6.751101   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                             6.907755   \n",
       "\n",
       "                                                    consensus_cell_type_individual  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...                          B cell   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...                     CD4+ T cell   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...                     CD4+ T cell   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...                             NaN   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...                Cytotoxic T cell   \n",
       "...                                                                            ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                           CD4+ T cell   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                           CD4+ T cell   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                      Cytotoxic T cell   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                      Cytotoxic T cell   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                      Cytotoxic T cell   \n",
       "\n",
       "                                                    seurat_cell_type_pred_score  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...                     0.724987   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...                     0.376578   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...                     0.531326   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...                          NaN   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...                     0.441366   \n",
       "...                                                                         ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                           0.309143   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                           0.402943   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                           0.450904   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                           0.334340   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                           0.390312   \n",
       "\n",
       "                                                     cell_type__mega  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...  Cytotoxic T cell   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...            B cell   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...  Cytotoxic T cell   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...            B cell   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...  Cytotoxic T cell   \n",
       "...                                                              ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS        Cytotoxic T cell   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS        Cytotoxic T cell   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS        Cytotoxic T cell   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS        Cytotoxic T cell   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS        Cytotoxic T cell   \n",
       "\n",
       "                                                   seurat_cell_type_pred_score__mega  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...                          0.417394   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...                          0.342953   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...                          0.391198   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...                          0.327305   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...                          0.375826   \n",
       "...                                                                              ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                                0.388897   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                                0.432030   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                                0.412242   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                                0.395910   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                                0.393253   \n",
       "\n",
       "                                                   consensus_cell_type__mega  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...          Cytotoxic T cell   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...          Cytotoxic T cell   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...          Cytotoxic T cell   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...          Cytotoxic T cell   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...          Cytotoxic T cell   \n",
       "...                                                                      ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                Cytotoxic T cell   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                     CD4+ T cell   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                Cytotoxic T cell   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                Cytotoxic T cell   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                Cytotoxic T cell   \n",
       "\n",
       "                                                   seurat_cell_type__mega  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...       Cytotoxic T cell   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...                 B cell   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...       Cytotoxic T cell   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...                 B cell   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...       Cytotoxic T cell   \n",
       "...                                                                   ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS             Cytotoxic T cell   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS             Cytotoxic T cell   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS             Cytotoxic T cell   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS             Cytotoxic T cell   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS             Cytotoxic T cell   \n",
       "\n",
       "                                                   harmony_pycisTopic_leiden_10_3.0  \\\n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...                               43   \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...                               20   \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...                               59   \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...                               76   \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...                               65   \n",
       "...                                                                             ...   \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                                      0   \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                                     43   \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                                      8   \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                                     54   \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                                     43   \n",
       "\n",
       "                                                   harmony_consensus_cell_type__mega  \n",
       "TCCTGAGCGAACGTCAGGGAACCGCG___OHS_s3atac_2.FIXED...                  Cytotoxic T cell  \n",
       "ACGCGACGACCGGAAGAAGCACGGAC___OHS_s3atac_2.FIXED...                       CD4+ T cell  \n",
       "GGTTAGTTCAAGTAGGACTATCGCAC___OHS_s3atac_2.FIXED...               Natural killer cell  \n",
       "CTCTCTACTGAACGGCGTCTGCTTCC___OHS_s3atac_2.FIXED...                  Cytotoxic T cell  \n",
       "ATTGAGGAAATGATGCTCCTGCTTCC___OHS_s3atac_2.FIXED...               Natural killer cell  \n",
       "...                                                                              ...  \n",
       "TTCTTAGACTTCTGAG___MDC_mtscatac_1.FIXEDCELLS                             CD4+ T cell  \n",
       "CATCAGAACCCGGACT___MDC_mtscatac_1.FIXEDCELLS                        Cytotoxic T cell  \n",
       "GAACCATACTACACTG___MDC_mtscatac_1.FIXEDCELLS                        Cytotoxic T cell  \n",
       "CAGTTGATGGCACGCT___MDC_mtscatac_1.FIXEDCELLS                        Cytotoxic T cell  \n",
       "ACCTCGTTGAGTTGAA___MDC_mtscatac_1.FIXEDCELLS                        Cytotoxic T cell  \n",
       "\n",
       "[169227 rows x 32 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cto.cell_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c39285ac",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:06:56.366545Z",
     "iopub.status.busy": "2023-02-16T09:06:56.366300Z",
     "iopub.status.idle": "2023-02-16T09:06:56.382450Z",
     "shell.execute_reply": "2023-02-16T09:06:56.381373Z",
     "shell.execute_reply.started": "2023-02-16T09:06:56.366524Z"
    },
    "papermill": {
     "duration": 0.029556,
     "end_time": "2022-09-18T16:28:06.775195",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.745639",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files exists, skipping...\n"
     ]
    }
   ],
   "source": [
    "# for sample in [\"BRO_mtscatac_1.LIBDS\"]:\n",
    "bw_paths_dict = {}\n",
    "bed_paths_dict = {}\n",
    "\n",
    "import ray\n",
    "\n",
    "if ray.is_initialized():\n",
    "    print(\"Shutting down Ray\")\n",
    "    ray.shutdown()\n",
    "\n",
    "for sample in cto_consensus_path_dict.keys():\n",
    "    # for sample in [\"BRO_mtscatac_1.LIBDS\"]:\n",
    "    bed_path = os.path.join(\n",
    "        \"final_consensus_peaks\", f\"{sample}__SCREEN_pseudobulk_bed_files\"\n",
    "    )\n",
    "    bw_path = os.path.join(\n",
    "        \"final_consensus_peaks\", f\"{sample}__SCREEN_pseudobulk_bw_files\"\n",
    "    )\n",
    "    if not os.path.exists(bed_path):\n",
    "        cto_path = cto_consensus_path_dict[sample]\n",
    "        with open(cto_path, \"rb\") as f:\n",
    "            cto = pickle.load(f)\n",
    "\n",
    "            bw_paths, bed_paths = export_pseudobulk(\n",
    "                input_data=cto,\n",
    "                variable=\"harmony_consensus_cell_type__mega\",\n",
    "                sample_id_col=\"sample_id\",\n",
    "                chromsizes=chromsizes,\n",
    "                bed_path=bed_path,\n",
    "                bigwig_path=bw_path,\n",
    "                path_to_fragments=\"merged_all_1.fragments.ID.sorted.tsv.gz\",\n",
    "                n_cpu=16,\n",
    "                normalize_bigwig=True,\n",
    "                remove_duplicates=True,\n",
    "            )\n",
    "\n",
    "            if ray.is_initialized():\n",
    "                print(\"Shutting down Ray\")\n",
    "                ray.shutdown()\n",
    "    else:\n",
    "        print(f\"{bed_path} exists, skipping...\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "e2eb9f88-85d2-4dcb-9d80-ef444f7efeda",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:08:18.564338Z",
     "iopub.status.busy": "2023-02-16T09:08:18.563374Z",
     "iopub.status.idle": "2023-02-16T09:08:22.672922Z",
     "shell.execute_reply": "2023-02-16T09:08:22.671868Z",
     "shell.execute_reply.started": "2023-02-16T09:08:18.564277Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "cto.cell_data.to_csv(\n",
    "    \"cistopic_objects_master/master_all_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_24topics.dimreduc.consensus_harmony.cell_data.csv\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8092c1f-c2d1-419a-ac24-ec6fe37beb0e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "f569dd2f",
   "metadata": {
    "papermill": {
     "duration": 0.004302,
     "end_time": "2022-09-18T16:28:06.784126",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.779824",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "7e9d717d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:08:39.979423Z",
     "iopub.status.busy": "2023-02-16T09:08:39.978611Z",
     "iopub.status.idle": "2023-02-16T09:08:39.996798Z",
     "shell.execute_reply": "2023-02-16T09:08:39.995419Z",
     "shell.execute_reply.started": "2023-02-16T09:08:39.979364Z"
    },
    "papermill": {
     "duration": 0.01414,
     "end_time": "2022-09-18T16:28:06.802426",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.788286",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'master_all_1.FIXEDCELLS': 'cistopic_objects_master/master_all_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_24topics.dimreduc.consensus_harmony.cell_data.csv'}"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_data_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(\n",
    "        glob.glob(\"cistopic_objects_master/*consensus_harmony.cell_data.csv\")\n",
    "    )\n",
    "}\n",
    "cell_data_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "e82383de",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:08:48.518601Z",
     "iopub.status.busy": "2023-02-16T09:08:48.517786Z",
     "iopub.status.idle": "2023-02-16T09:08:48.535334Z",
     "shell.execute_reply": "2023-02-16T09:08:48.533964Z",
     "shell.execute_reply.started": "2023-02-16T09:08:48.518544Z"
    },
    "papermill": {
     "duration": 0.014201,
     "end_time": "2022-09-18T16:28:06.821370",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.807169",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'master_all_1.FIXEDCELLS': 'final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files'}"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bw_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"final_consensus_peaks/*_pseudobulk_bw_files\"))\n",
    "}\n",
    "bw_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "8852d2b8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:08:48.786556Z",
     "iopub.status.busy": "2023-02-16T09:08:48.785753Z",
     "iopub.status.idle": "2023-02-16T09:08:48.802668Z",
     "shell.execute_reply": "2023-02-16T09:08:48.801391Z",
     "shell.execute_reply.started": "2023-02-16T09:08:48.786499Z"
    },
    "papermill": {
     "duration": 0.013807,
     "end_time": "2022-09-18T16:28:06.839683",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.825876",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'master_all_1.FIXEDCELLS': 'final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files'}"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"final_consensus_peaks/*_pseudobulk_bed_files\"))\n",
    "}\n",
    "bed_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "a86b4378",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:08:48.975080Z",
     "iopub.status.busy": "2023-02-16T09:08:48.974296Z",
     "iopub.status.idle": "2023-02-16T09:08:48.990528Z",
     "shell.execute_reply": "2023-02-16T09:08:48.989216Z",
     "shell.execute_reply.started": "2023-02-16T09:08:48.975023Z"
    },
    "papermill": {
     "duration": 0.031804,
     "end_time": "2022-09-18T16:28:06.876112",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.844308",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "for sample in bed_path_dict.keys():\n",
    "    bed_paths = {\n",
    "        x.split(\"/\")[-1].split(\"__\")[0].split(\".bed.gz\")[0]: x\n",
    "        for x in glob.glob(bed_path_dict[sample] + \"/*\")\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "e8d5ddd9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:08:49.087736Z",
     "iopub.status.busy": "2023-02-16T09:08:49.086940Z",
     "iopub.status.idle": "2023-02-16T09:08:49.095634Z",
     "shell.execute_reply": "2023-02-16T09:08:49.094215Z",
     "shell.execute_reply.started": "2023-02-16T09:08:49.087680Z"
    },
    "papermill": {
     "duration": 0.009395,
     "end_time": "2022-09-18T16:28:06.891214",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.881819",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pycisTopic.pseudobulk_peak_calling import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "10889f7a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:08:49.277674Z",
     "iopub.status.busy": "2023-02-16T09:08:49.276878Z",
     "iopub.status.idle": "2023-02-16T09:08:49.285034Z",
     "shell.execute_reply": "2023-02-16T09:08:49.283633Z",
     "shell.execute_reply.started": "2023-02-16T09:08:49.277618Z"
    },
    "papermill": {
     "duration": 0.009129,
     "end_time": "2022-09-18T16:28:06.905085",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.895956",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import ray"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "87fa97a4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:09:07.447650Z",
     "iopub.status.busy": "2023-02-16T09:09:07.446726Z",
     "iopub.status.idle": "2023-02-16T09:42:37.712230Z",
     "shell.execute_reply": "2023-02-16T09:42:37.711056Z",
     "shell.execute_reply.started": "2023-02-16T09:09:07.447590Z"
    },
    "papermill": {
     "duration": 0.049072,
     "end_time": "2022-09-18T16:28:06.958695",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.909623",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-02-16 10:09:12,301\tINFO worker.py:1509 -- Started a local Ray instance. View the dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265 \u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=1290657)\u001b[0m 2023-02-16 10:09:15,280 cisTopic     INFO     Calling peaks for CD14_monocyte with macs2 callpeak --treatment final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files/CD14_monocyte.bed.gz --name CD14_monocyte  --outdir final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=1290659)\u001b[0m 2023-02-16 10:09:15,285 cisTopic     INFO     Calling peaks for CytotoxicTcell with macs2 callpeak --treatment final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files/CytotoxicTcell.bed.gz --name CytotoxicTcell  --outdir final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=1290665)\u001b[0m 2023-02-16 10:09:15,301 cisTopic     INFO     Calling peaks for CD4_Tcell with macs2 callpeak --treatment final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files/CD4_Tcell.bed.gz --name CD4_Tcell  --outdir final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=1290662)\u001b[0m 2023-02-16 10:09:15,320 cisTopic     INFO     Calling peaks for Naturalkillercell with macs2 callpeak --treatment final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files/Naturalkillercell.bed.gz --name Naturalkillercell  --outdir final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=1290669)\u001b[0m 2023-02-16 10:09:15,359 cisTopic     INFO     Calling peaks for Bcell with macs2 callpeak --treatment final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files/Bcell.bed.gz --name Bcell  --outdir final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=1290661)\u001b[0m 2023-02-16 10:09:15,329 cisTopic     INFO     Calling peaks for Dendriticcell with macs2 callpeak --treatment final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files/Dendriticcell.bed.gz --name Dendriticcell  --outdir final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=1290660)\u001b[0m 2023-02-16 10:09:15,420 cisTopic     INFO     Calling peaks for CD16_monocyte with macs2 callpeak --treatment final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files/CD16_monocyte.bed.gz --name CD16_monocyte  --outdir final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_consensus_peaks --format BEDPE --gsize hs --qvalue 0.05 --nomodel --shift 73 --extsize 146 --keep-dup all --call-summits --nolambda\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=1290661)\u001b[0m 2023-02-16 10:13:34,043 cisTopic     INFO     Dendriticcell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=1290660)\u001b[0m 2023-02-16 10:14:48,119 cisTopic     INFO     CD16_monocyte done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=1290662)\u001b[0m 2023-02-16 10:17:30,440 cisTopic     INFO     Naturalkillercell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=1290669)\u001b[0m 2023-02-16 10:20:15,426 cisTopic     INFO     Bcell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=1290659)\u001b[0m 2023-02-16 10:33:03,375 cisTopic     INFO     CytotoxicTcell done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=1290657)\u001b[0m 2023-02-16 10:37:11,956 cisTopic     INFO     CD14_monocyte done!\n",
      "\u001b[2m\u001b[36m(macs_call_peak_ray pid=1290665)\u001b[0m 2023-02-16 10:42:33,810 cisTopic     INFO     CD4_Tcell done!\n"
     ]
    }
   ],
   "source": [
    "narrow_peaks_dict = {}\n",
    "ray.shutdown()\n",
    "for sample in bed_path_dict.keys():\n",
    "    narrow_peaks_dict_path = bed_path_dict[sample].replace(\n",
    "        \"_pseudobulk_bed_files\", \"_narrow_peaks_dict.pkl\"\n",
    "    )\n",
    "    peak_path = os.path.join(\n",
    "        \"final_consensus_peaks\", f\"{sample}__SCREEN_consensus_peaks\"\n",
    "    )\n",
    "    if not os.path.exists(peak_path):\n",
    "        os.mkdir(peak_path)\n",
    "\n",
    "    if not os.path.exists(narrow_peaks_dict_path):\n",
    "        cell_data = pd.read_csv(cell_data_path_dict[sample])\n",
    "        cto_celltypes = set(\n",
    "            [\n",
    "                x.replace(\" \", \"\").replace(\"+\", \"\").replace(\"_\", \"\")\n",
    "                for x in set(cell_data[\"harmony_consensus_cell_type__mega\"].unique())\n",
    "            ]\n",
    "        )\n",
    "\n",
    "        bed_celltypes = set(\n",
    "            [\n",
    "                x.split(\".\")[0].replace(\"+\", \"\").replace(\"_\", \"\")\n",
    "                for x in os.listdir(bed_path_dict[sample])\n",
    "            ]\n",
    "        )\n",
    "\n",
    "        if cto_celltypes == bed_celltypes:\n",
    "            print(f\"Starting {narrow_peaks_dict_path}\")\n",
    "            bed_paths = {\n",
    "                x.split(\"/\")[-1].split(\"__\")[0].split(\".bed.gz\")[0]: x\n",
    "                for x in glob.glob(bed_path_dict[sample] + \"/*\")\n",
    "            }\n",
    "            narrow_peaks_dict = peak_calling(\n",
    "                macs_path=\"macs2\",\n",
    "                bed_paths=bed_paths,\n",
    "                outdir=peak_path,\n",
    "                genome_size=\"hs\",\n",
    "                n_cpu=20,\n",
    "                input_format=\"BEDPE\",\n",
    "                shift=73,\n",
    "                ext_size=146,\n",
    "                keep_dup=\"all\",\n",
    "                q_value=0.05,\n",
    "            )\n",
    "            with open(narrow_peaks_dict_path, \"wb\") as f:\n",
    "                pickle.dump(narrow_peaks_dict, f)\n",
    "        else:\n",
    "            print(f\"{sample} cell types not matching!! Rerun bed file writing.\")\n",
    "            print(f\"\\t{bed_celltypes}\")\n",
    "            print(f\"\\t{cto_celltypes}\")\n",
    "    else:\n",
    "        print(f\"{narrow_peaks_dict_path} already exists\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dcd338ac",
   "metadata": {
    "papermill": {
     "duration": 0.004524,
     "end_time": "2022-09-18T16:28:06.968187",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.963663",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# call consensus peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "b6eed5af",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:42:37.713930Z",
     "iopub.status.busy": "2023-02-16T09:42:37.713699Z",
     "iopub.status.idle": "2023-02-16T09:42:37.793646Z",
     "shell.execute_reply": "2023-02-16T09:42:37.792030Z",
     "shell.execute_reply.started": "2023-02-16T09:42:37.713911Z"
    },
    "papermill": {
     "duration": 0.010761,
     "end_time": "2022-09-18T16:28:06.983492",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.972731",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pycisTopic.iterative_peak_calling import get_consensus_peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "79cf9c75",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:42:37.796553Z",
     "iopub.status.busy": "2023-02-16T09:42:37.795774Z",
     "iopub.status.idle": "2023-02-16T09:42:37.818228Z",
     "shell.execute_reply": "2023-02-16T09:42:37.816859Z",
     "shell.execute_reply.started": "2023-02-16T09:42:37.796496Z"
    },
    "papermill": {
     "duration": 0.01512,
     "end_time": "2022-09-18T16:28:07.003585",
     "exception": false,
     "start_time": "2022-09-18T16:28:06.988465",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'master_all_1.FIXEDCELLS': 'final_consensus_peaks/master_all_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl'}"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "narrow_peaks_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"final_consensus_peaks/*_narrow_peaks_dict.pkl\"))\n",
    "}\n",
    "narrow_peaks_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6fa6789f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:42:37.821432Z",
     "iopub.status.busy": "2023-02-16T09:42:37.820867Z",
     "iopub.status.idle": "2023-02-16T09:42:37.827775Z",
     "shell.execute_reply": "2023-02-16T09:42:37.826610Z",
     "shell.execute_reply.started": "2023-02-16T09:42:37.821394Z"
    },
    "papermill": {
     "duration": 0.009936,
     "end_time": "2022-09-18T16:28:07.018459",
     "exception": false,
     "start_time": "2022-09-18T16:28:07.008523",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "path_to_blacklist = \"../0_resources/regions/hg38-blacklist.v2.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "ef348f9b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-02-16T09:42:37.829536Z",
     "iopub.status.busy": "2023-02-16T09:42:37.829149Z",
     "iopub.status.idle": "2023-02-16T09:42:38.766569Z",
     "shell.execute_reply": "2023-02-16T09:42:38.765269Z",
     "shell.execute_reply.started": "2023-02-16T09:42:37.829502Z"
    },
    "papermill": {
     "duration": 0.045812,
     "end_time": "2022-09-18T16:28:07.069097",
     "exception": false,
     "start_time": "2022-09-18T16:28:07.023285",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "master_all_1.FIXEDCELLS\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'consensus_cell_type'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m/opt/venv/lib/python3.8/site-packages/pandas/core/indexes/base.py:3800\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3799\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3800\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3801\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m/opt/venv/lib/python3.8/site-packages/pandas/_libs/index.pyx:138\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m/opt/venv/lib/python3.8/site-packages/pandas/_libs/index.pyx:165\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5745\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5753\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'consensus_cell_type'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [54], line 10\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mexists(consensus_out_path):\n\u001b[1;32m      9\u001b[0m     cell_data \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mread_csv(cell_data_path_dict[sample])\n\u001b[0;32m---> 10\u001b[0m     cto_celltypes \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mset\u001b[39m([x\u001b[38;5;241m.\u001b[39mreplace(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m \u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m.\u001b[39mreplace(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m+\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m)\u001b[38;5;241m.\u001b[39mreplace(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mset\u001b[39m(\u001b[43mcell_data\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mconsensus_cell_type\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39munique())])\n\u001b[1;32m     12\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mopen\u001b[39m(narrow_peaks_path_dict[sample], \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mrb\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m f:\n\u001b[1;32m     13\u001b[0m         narrow_peaks_dict \u001b[38;5;241m=\u001b[39m pickle\u001b[38;5;241m.\u001b[39mload(f)\n",
      "File \u001b[0;32m/opt/venv/lib/python3.8/site-packages/pandas/core/frame.py:3805\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   3804\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 3805\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   3807\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m/opt/venv/lib/python3.8/site-packages/pandas/core/indexes/base.py:3802\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3800\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3801\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3802\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3804\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3805\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3806\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'consensus_cell_type'"
     ]
    }
   ],
   "source": [
    "peak_half_width = 250\n",
    "\n",
    "# Get consensus peaks\n",
    "consensus_peaks_dict = {}\n",
    "for sample in narrow_peaks_path_dict.keys():\n",
    "    print(sample)\n",
    "    consensus_out_path = narrow_peaks_path_dict[sample].replace('_narrow_peaks_dict.pkl', '_consensus_peaks.bed')\n",
    "    if not os.path.exists(consensus_out_path):\n",
    "        cell_data = pd.read_csv(cell_data_path_dict[sample])\n",
    "        cto_celltypes = set([x.replace(' ', '').replace('+', '').replace('_', '') for x in set(cell_data['consensus_cell_type'].unique())])\n",
    "        \n",
    "        with open(narrow_peaks_path_dict[sample], 'rb') as f:\n",
    "            narrow_peaks_dict = pickle.load(f)\n",
    "        peaks_celltypes = set([x.replace(' ', '').replace('+', '').replace('_', '') for x in set(narrow_peaks_dict.keys())])\n",
    "    \n",
    "        if cto_celltypes == peaks_celltypes:\n",
    "            consensus_peaks = get_consensus_peaks(\n",
    "                narrow_peaks_dict,\n",
    "                peak_half_width,\n",
    "                chromsizes = chromsizes,\n",
    "                path_to_blacklist = path_to_blacklist\n",
    "            )\n",
    "\n",
    "            consensus_peaks.to_bed(\n",
    "                path=consensus_out_path,\n",
    "                keep=True,\n",
    "                compression='infer',\n",
    "                chain=False\n",
    "            )\n",
    "        else:\n",
    "            print('CELL TYPE SETS NOT MATCHING! Rerun peak calling.')\n",
    "            print(peaks_celltypes - cto_celltypes)\n",
    "            print(cto_celltypes - peaks_celltypes)\n",
    "    else:\n",
    "        print(f\"{consensus_out_path} already exists, skipping...\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "795f9b6b",
   "metadata": {
    "papermill": {
     "duration": 0.005031,
     "end_time": "2022-09-18T16:28:07.079362",
     "exception": false,
     "start_time": "2022-09-18T16:28:07.074331",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Check % chrM in consensus peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8d4531d",
   "metadata": {
    "execution": {
     "iopub.status.busy": "2023-02-16T09:42:38.767287Z",
     "iopub.status.idle": "2023-02-16T09:42:38.767570Z",
     "shell.execute_reply": "2023-02-16T09:42:38.767442Z",
     "shell.execute_reply.started": "2023-02-16T09:42:38.767428Z"
    },
    "papermill": {
     "duration": 0.015493,
     "end_time": "2022-09-18T16:28:07.099734",
     "exception": false,
     "start_time": "2022-09-18T16:28:07.084241",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "consensus_peaks_path_dict = {x.split('/')[-1].split(f'__')[0]: x for x in sorted(glob.glob(\"final_consensus_peaks/*consensus_peaks.bed\"))}\n",
    "consensus_peaks_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45003f1b",
   "metadata": {
    "execution": {
     "iopub.status.busy": "2023-02-16T09:42:38.768972Z",
     "iopub.status.idle": "2023-02-16T09:42:38.769386Z",
     "shell.execute_reply": "2023-02-16T09:42:38.769258Z",
     "shell.execute_reply.started": "2023-02-16T09:42:38.769244Z"
    },
    "papermill": {
     "duration": 15.270725,
     "end_time": "2022-09-18T16:28:22.375682",
     "exception": false,
     "start_time": "2022-09-18T16:28:07.104957",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "for sample, path in consensus_peaks_path_dict.items():\n",
    "    print(sample)\n",
    "    peaks_df = pd.read_csv(path, sep='\\t', header=None)\n",
    "    \n",
    "    chroms_in_df = list(sorted(peaks_df[0].unique()))\n",
    "    chroms_standard = [\"chr\" + str(x + 1) for x in range(22)] + ['chrX']\n",
    "    chroms_nonstandard = list(set(chroms_in_df ) - set(chroms_standard) -  set(['chrM']))\n",
    "    \n",
    "    n_standard = peaks_df[0].value_counts()[chroms_standard].sum()\n",
    "    n_contigs = peaks_df[0].value_counts()[chroms_nonstandard].sum()\n",
    "    n_chrm = peaks_df[0].value_counts()['chrM'].sum()\n",
    "    pct_nonstandard = (n_contigs + n_chrm)/len(peaks_df)*100\n",
    "        \n",
    "    print(f\"\\tpeaks on standard chromosomes: {n_standard}\")\n",
    "    print(f\"\\tpeaks on contigs: {n_contigs}\")\n",
    "    print(f\"\\tpeaks on chrM: {n_chrm}\")\n",
    "    print(f\"\\t% peaks non standard chromosomes: {pct_nonstandard}%\")\n",
    "    if 'chrY' in chroms_in_df:\n",
    "        n_chrY = peaks_df[0].value_counts()['chrY'].sum()\n",
    "        print(f\"\\tpeaks on chrY: {n_chrY}\")\n",
    "    else:\n",
    "        print(f\"\\tpeaks on chrY: 0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "883eafc3",
   "metadata": {
    "papermill": {
     "duration": 0.007013,
     "end_time": "2022-09-18T16:28:22.395593",
     "exception": false,
     "start_time": "2022-09-18T16:28:22.388580",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff937bb6",
   "metadata": {
    "papermill": {
     "duration": 0.006579,
     "end_time": "2022-09-18T16:28:22.408851",
     "exception": false,
     "start_time": "2022-09-18T16:28:22.402272",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7df12b75",
   "metadata": {
    "papermill": {
     "duration": 0.006488,
     "end_time": "2022-09-18T16:28:22.421849",
     "exception": false,
     "start_time": "2022-09-18T16:28:22.415361",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "20221014_pycistopic.sif",
   "language": "python",
   "name": "cistopic_20221014"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 33.135392,
   "end_time": "2022-09-18T16:28:23.816937",
   "environment_variables": {},
   "exception": null,
   "input_path": "8_consensus_peaks.ipynb",
   "output_path": "8_consensus_peaks.OUT.ipynb",
   "parameters": {},
   "start_time": "2022-09-18T16:27:50.681545",
   "version": "2.3.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
