{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d6672f7e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-01-16T15:59:08.619602Z",
     "iopub.status.busy": "2023-01-16T15:59:08.618623Z",
     "iopub.status.idle": "2023-01-16T15:59:28.046550Z",
     "shell.execute_reply": "2023-01-16T15:59:28.045305Z",
     "shell.execute_reply.started": "2023-01-16T15:59:08.619444Z"
    },
    "papermill": {
     "duration": 10.33396,
     "end_time": "2022-09-16T16:02:23.635151",
     "exception": false,
     "start_time": "2022-09-16T16:02:13.301191",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pycisTopic\n",
    "import glob\n",
    "import os\n",
    "import pybiomart as pbm\n",
    "import pandas as pd\n",
    "import pickle\n",
    "from pycisTopic.qc import *\n",
    "from IPython.display import Image, display\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "\n",
    "import multiprocess as mp  # for kde multithreading calculation\n",
    "from multiprocess import Pool\n",
    "\n",
    "%matplotlib inline\n",
    "%load_ext lab_black"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90b5efc3",
   "metadata": {
    "papermill": {
     "duration": 0.004961,
     "end_time": "2022-09-16T16:02:23.654074",
     "exception": false,
     "start_time": "2022-09-16T16:02:23.649113",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Download annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "66c949f7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-01-16T15:59:28.049117Z",
     "iopub.status.busy": "2023-01-16T15:59:28.048379Z",
     "iopub.status.idle": "2023-01-16T15:59:28.296576Z",
     "shell.execute_reply": "2023-01-16T15:59:28.294641Z",
     "shell.execute_reply.started": "2023-01-16T15:59:28.049085Z"
    },
    "papermill": {
     "duration": 0.249583,
     "end_time": "2022-09-16T16:02:23.908035",
     "exception": false,
     "start_time": "2022-09-16T16:02:23.658452",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/lustre1/project/stg_00090/scatac_benchmark/fixedcells_3_cistopic_consensus\n"
     ]
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "15a5754a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-01-16T16:00:45.371161Z",
     "iopub.status.busy": "2023-01-16T16:00:45.370202Z",
     "iopub.status.idle": "2023-01-16T16:00:45.380677Z",
     "shell.execute_reply": "2023-01-16T16:00:45.379151Z",
     "shell.execute_reply.started": "2023-01-16T16:00:45.371100Z"
    },
    "papermill": {
     "duration": 0.011366,
     "end_time": "2022-09-16T16:02:23.924742",
     "exception": false,
     "start_time": "2022-09-16T16:02:23.913376",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "wdir = \"/lustre1/project/stg_00090/scatac_benchmark/fixedcells_3_cistopic_consensus\"\n",
    "os.chdir(wdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b78786dc",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-01-16T16:00:46.223132Z",
     "iopub.status.busy": "2023-01-16T16:00:46.222319Z",
     "iopub.status.idle": "2023-01-16T16:00:46.850917Z",
     "shell.execute_reply": "2023-01-16T16:00:46.849392Z",
     "shell.execute_reply.started": "2023-01-16T16:00:46.223074Z"
    },
    "papermill": {
     "duration": 0.100299,
     "end_time": "2022-09-16T16:02:24.029443",
     "exception": false,
     "start_time": "2022-09-16T16:02:23.929144",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading cached genome annotation...\n"
     ]
    }
   ],
   "source": [
    "genome = \"hg38\"\n",
    "\n",
    "pbm_genome_name_dict = {\n",
    "    \"hg38\": \"hsapiens_gene_ensembl\",\n",
    "    \"hg37\": \"hsapiens_gene_ensembl\",\n",
    "    \"mm10\": \"mmusculus_gene_ensembl\",\n",
    "    \"dm6\": \"dmelanogaster_gene_ensembl\",\n",
    "}\n",
    "\n",
    "pbm_host_dict = {\n",
    "    \"hg38\": \"http://www.ensembl.org\",\n",
    "    \"hg37\": \"http://grch37.ensembl.org/\",\n",
    "    \"mm10\": \"http://nov2020.archive.ensembl.org/\",\n",
    "    \"dm6\": \"http://www.ensembl.org\",\n",
    "}\n",
    "\n",
    "if os.path.exists(f\"annotation.tsv\"):\n",
    "    print(f\"Loading cached genome annotation...\")\n",
    "    annotation = pd.read_csv(\"annotation.tsv\", sep=\"\\t\", header=0, index_col=0)\n",
    "else:\n",
    "    dataset = pbm.Dataset(name=pbm_genome_name_dict[genome], host=pbm_host_dict[genome])\n",
    "\n",
    "    annotation = dataset.query(\n",
    "        attributes=[\n",
    "            \"chromosome_name\",\n",
    "            \"transcription_start_site\",\n",
    "            \"strand\",\n",
    "            \"external_gene_name\",\n",
    "            \"transcript_biotype\",\n",
    "        ]\n",
    "    )\n",
    "    filter = annotation[\"Chromosome/scaffold name\"].str.contains(\"CHR|GL|JH|MT\")\n",
    "    annotation = annotation[~filter]\n",
    "    annotation[\"Chromosome/scaffold name\"] = annotation[\n",
    "        \"Chromosome/scaffold name\"\n",
    "    ].str.replace(r\"(\\b\\S)\", r\"chr\\1\")\n",
    "    annotation.columns = [\"Chromosome\", \"Start\", \"Strand\", \"Gene\", \"Transcript_type\"]\n",
    "    annottation = annotation[annotation.Transcript_type == \"protein_coding\"]\n",
    "    annotation.to_csv(\"annotation.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7e7ebed9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-01-16T16:00:50.436714Z",
     "iopub.status.busy": "2023-01-16T16:00:50.435895Z",
     "iopub.status.idle": "2023-01-16T16:00:50.480293Z",
     "shell.execute_reply": "2023-01-16T16:00:50.479120Z",
     "shell.execute_reply.started": "2023-01-16T16:00:50.436646Z"
    },
    "papermill": {
     "duration": 0.021261,
     "end_time": "2022-09-16T16:02:24.055470",
     "exception": false,
     "start_time": "2022-09-16T16:02:24.034209",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chrKI270711chr.chr1</td>\n",
       "      <td>24650</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chrKI270711chr.chr1</td>\n",
       "      <td>29626</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>chrKI270734chr.chr1</td>\n",
       "      <td>161852</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629</th>\n",
       "      <td>chrKI270734chr.chr1</td>\n",
       "      <td>161750</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>713</th>\n",
       "      <td>chrKI270734chr.chr1</td>\n",
       "      <td>131494</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242597</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154257071</td>\n",
       "      <td>1</td>\n",
       "      <td>UBAP2L</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242604</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154272589</td>\n",
       "      <td>1</td>\n",
       "      <td>HAX1</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242605</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154272609</td>\n",
       "      <td>1</td>\n",
       "      <td>HAX1</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242607</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154272629</td>\n",
       "      <td>1</td>\n",
       "      <td>HAX1</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242610</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154272668</td>\n",
       "      <td>1</td>\n",
       "      <td>HAX1</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77891 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Chromosome      Start  Strand    Gene Transcript_type\n",
       "57      chrKI270711chr.chr1      24650      -1     NaN  protein_coding\n",
       "58      chrKI270711chr.chr1      29626      -1     NaN  protein_coding\n",
       "628     chrKI270734chr.chr1     161852      -1     NaN  protein_coding\n",
       "629     chrKI270734chr.chr1     161750      -1     NaN  protein_coding\n",
       "713     chrKI270734chr.chr1     131494       1     NaN  protein_coding\n",
       "...                     ...        ...     ...     ...             ...\n",
       "242597                 chr1  154257071       1  UBAP2L  protein_coding\n",
       "242604                 chr1  154272589       1    HAX1  protein_coding\n",
       "242605                 chr1  154272609       1    HAX1  protein_coding\n",
       "242607                 chr1  154272629       1    HAX1  protein_coding\n",
       "242610                 chr1  154272668       1    HAX1  protein_coding\n",
       "\n",
       "[77891 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "acd5b055-b06d-442a-bd4e-d15f0b9da986",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-01-16T16:04:33.939633Z",
     "iopub.status.busy": "2023-01-16T16:04:33.938803Z",
     "iopub.status.idle": "2023-01-16T16:04:33.965402Z",
     "shell.execute_reply": "2023-01-16T16:04:33.964229Z",
     "shell.execute_reply.started": "2023-01-16T16:04:33.939578Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Transcript_type</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>chrKI270711chr.chr1</td>\n",
       "      <td>24650</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>24651</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>chrKI270711chr.chr1</td>\n",
       "      <td>29626</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>29627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>628</th>\n",
       "      <td>chrKI270734chr.chr1</td>\n",
       "      <td>161852</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>161853</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>629</th>\n",
       "      <td>chrKI270734chr.chr1</td>\n",
       "      <td>161750</td>\n",
       "      <td>-1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>161751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>713</th>\n",
       "      <td>chrKI270734chr.chr1</td>\n",
       "      <td>131494</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>131495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242597</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154257071</td>\n",
       "      <td>1</td>\n",
       "      <td>UBAP2L</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>154257072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242604</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154272589</td>\n",
       "      <td>1</td>\n",
       "      <td>HAX1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>154272590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242605</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154272609</td>\n",
       "      <td>1</td>\n",
       "      <td>HAX1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>154272610</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242607</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154272629</td>\n",
       "      <td>1</td>\n",
       "      <td>HAX1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>154272630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>242610</th>\n",
       "      <td>chr1</td>\n",
       "      <td>154272668</td>\n",
       "      <td>1</td>\n",
       "      <td>HAX1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>154272669</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>77891 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Chromosome      Start  Strand    Gene Transcript_type  \\\n",
       "57      chrKI270711chr.chr1      24650      -1     NaN  protein_coding   \n",
       "58      chrKI270711chr.chr1      29626      -1     NaN  protein_coding   \n",
       "628     chrKI270734chr.chr1     161852      -1     NaN  protein_coding   \n",
       "629     chrKI270734chr.chr1     161750      -1     NaN  protein_coding   \n",
       "713     chrKI270734chr.chr1     131494       1     NaN  protein_coding   \n",
       "...                     ...        ...     ...     ...             ...   \n",
       "242597                 chr1  154257071       1  UBAP2L  protein_coding   \n",
       "242604                 chr1  154272589       1    HAX1  protein_coding   \n",
       "242605                 chr1  154272609       1    HAX1  protein_coding   \n",
       "242607                 chr1  154272629       1    HAX1  protein_coding   \n",
       "242610                 chr1  154272668       1    HAX1  protein_coding   \n",
       "\n",
       "              End  \n",
       "57          24651  \n",
       "58          29627  \n",
       "628        161853  \n",
       "629        161751  \n",
       "713        131495  \n",
       "...           ...  \n",
       "242597  154257072  \n",
       "242604  154272590  \n",
       "242605  154272610  \n",
       "242607  154272630  \n",
       "242610  154272669  \n",
       "\n",
       "[77891 rows x 6 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b3e8f882-9c8f-4dca-b480-4731e07fbcf0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-01-16T16:07:09.646752Z",
     "iopub.status.busy": "2023-01-16T16:07:09.645838Z",
     "iopub.status.idle": "2023-01-16T16:07:09.872296Z",
     "shell.execute_reply": "2023-01-16T16:07:09.871309Z",
     "shell.execute_reply.started": "2023-01-16T16:07:09.646691Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# save TSS_ann also to .bed\n",
    "\n",
    "annotation[\"End\"] = annotation[\"Start\"] + 1\n",
    "annotation[\"Score\"] = None\n",
    "annotation[[\"Chromosome\", \"Start\", \"End\", \"Gene\", \"Score\", \"Strand\"]].to_csv(\n",
    "    \"tss.bed\", sep=\"\\t\", index=False, header=None\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2d10f3e6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-16T16:02:24.066042Z",
     "iopub.status.busy": "2022-09-16T16:02:24.065533Z",
     "iopub.status.idle": "2022-09-16T16:02:24.074929Z",
     "shell.execute_reply": "2022-09-16T16:02:24.074264Z"
    },
    "papermill": {
     "duration": 0.015284,
     "end_time": "2022-09-16T16:02:24.075840",
     "exception": false,
     "start_time": "2022-09-16T16:02:24.060556",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BIO_ddseq_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BIO_ddseq_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BIO_ddseq_3.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BIO_ddseq_4.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BRO_mtscatac_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BRO_mtscatac_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xmultiome_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xmultiome_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv11_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv11_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv11_3.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv11_4.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv11_5.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv2_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv2_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_hydrop_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_hydrop_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_hydrop_3.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_mtscatac_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_mtscatac_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/EPF_hydrop_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/EPF_hydrop_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/EPF_hydrop_3.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/EPF_hydrop_4.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/HAR_ddseq_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/HAR_ddseq_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/MDC_mtscatac_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/MDC_mtscatac_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'OHS_s3atac_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/OHS_s3atac_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'OHS_s3atac_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/OHS_s3atac_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/SAN_10xmultiome_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/SAN_10xmultiome_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'STA_10xv11_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/STA_10xv11_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'STA_10xv11_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/STA_10xv11_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'TXG_10xv11_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/TXG_10xv11_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'TXG_10xv2_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/TXG_10xv2_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'TXG_10xv2_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/TXG_10xv2_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'UCS_ddseq_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/UCS_ddseq_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'UCS_ddseq_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/UCS_ddseq_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xmultiome_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xmultiome_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xv1_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xv1_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xv1_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xv1_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xv2_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xv2_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xv2_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xv2_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_11.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_11.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_12.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_12.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_21.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_21.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_22.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_22.FIXEDCELLS.fragments.tsv.gz'}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fragments_list = sorted(glob.glob(\"../1_data_repository/fixedcells_fragments/*.tsv.gz\"))\n",
    "fragments_dict = {}\n",
    "for fragments_file in fragments_list:\n",
    "    sample = fragments_file.split(\"/\")[-1].split(\".fragments.tsv.gz\")[0]\n",
    "    fragments_dict[sample] = fragments_file\n",
    "fragments_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "96e44e0d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-16T16:02:24.087056Z",
     "iopub.status.busy": "2022-09-16T16:02:24.086789Z",
     "iopub.status.idle": "2022-09-16T16:02:24.096496Z",
     "shell.execute_reply": "2022-09-16T16:02:24.095841Z"
    },
    "papermill": {
     "duration": 0.01676,
     "end_time": "2022-09-16T16:02:24.097416",
     "exception": false,
     "start_time": "2022-09-16T16:02:24.080656",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': 'SCREEN_peaks/BIO_ddseq_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': 'SCREEN_peaks/BIO_ddseq_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': 'SCREEN_peaks/BIO_ddseq_3.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': 'SCREEN_peaks/BIO_ddseq_4.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': 'SCREEN_peaks/BRO_mtscatac_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': 'SCREEN_peaks/BRO_mtscatac_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': 'SCREEN_peaks/CNA_10xmultiome_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': 'SCREEN_peaks/CNA_10xmultiome_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': 'SCREEN_peaks/CNA_10xv11_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': 'SCREEN_peaks/CNA_10xv11_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': 'SCREEN_peaks/CNA_10xv11_3.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': 'SCREEN_peaks/CNA_10xv11_4.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': 'SCREEN_peaks/CNA_10xv11_5.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': 'SCREEN_peaks/CNA_10xv2_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': 'SCREEN_peaks/CNA_10xv2_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': 'SCREEN_peaks/CNA_hydrop_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': 'SCREEN_peaks/CNA_hydrop_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': 'SCREEN_peaks/CNA_hydrop_3.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': 'SCREEN_peaks/CNA_mtscatac_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': 'SCREEN_peaks/CNA_mtscatac_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': 'SCREEN_peaks/EPF_hydrop_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': 'SCREEN_peaks/EPF_hydrop_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': 'SCREEN_peaks/EPF_hydrop_3.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': 'SCREEN_peaks/EPF_hydrop_4.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': 'SCREEN_peaks/HAR_ddseq_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': 'SCREEN_peaks/HAR_ddseq_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': 'SCREEN_peaks/MDC_mtscatac_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'MDC_mtscatac_2.FIXEDCELLS': 'SCREEN_peaks/MDC_mtscatac_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'OHS_s3atac_1.FIXEDCELLS': 'SCREEN_peaks/OHS_s3atac_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'OHS_s3atac_2.FIXEDCELLS': 'SCREEN_peaks/OHS_s3atac_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'SAN_10xmultiome_1.FIXEDCELLS': 'SCREEN_peaks/SAN_10xmultiome_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'SAN_10xmultiome_2.FIXEDCELLS': 'SCREEN_peaks/SAN_10xmultiome_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'STA_10xv11_1.FIXEDCELLS': 'SCREEN_peaks/STA_10xv11_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'STA_10xv11_2.FIXEDCELLS': 'SCREEN_peaks/STA_10xv11_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'TXG_10xv11_1.FIXEDCELLS': 'SCREEN_peaks/TXG_10xv11_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'TXG_10xv2_1.FIXEDCELLS': 'SCREEN_peaks/TXG_10xv2_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'TXG_10xv2_2.FIXEDCELLS': 'SCREEN_peaks/TXG_10xv2_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'UCS_ddseq_1.FIXEDCELLS': 'SCREEN_peaks/UCS_ddseq_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'UCS_ddseq_2.FIXEDCELLS': 'SCREEN_peaks/UCS_ddseq_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xmultiome_1.FIXEDCELLS': 'SCREEN_peaks/VIB_10xmultiome_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xmultiome_2.FIXEDCELLS': 'SCREEN_peaks/VIB_10xmultiome_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xv1_1.FIXEDCELLS': 'SCREEN_peaks/VIB_10xv1_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xv1_2.FIXEDCELLS': 'SCREEN_peaks/VIB_10xv1_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xv2_1.FIXEDCELLS': 'SCREEN_peaks/VIB_10xv2_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_10xv2_2.FIXEDCELLS': 'SCREEN_peaks/VIB_10xv2_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_1.FIXEDCELLS': 'SCREEN_peaks/VIB_hydrop_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_11.FIXEDCELLS': 'SCREEN_peaks/VIB_hydrop_11.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_12.FIXEDCELLS': 'SCREEN_peaks/VIB_hydrop_12.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_2.FIXEDCELLS': 'SCREEN_peaks/VIB_hydrop_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_21.FIXEDCELLS': 'SCREEN_peaks/VIB_hydrop_21.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_22.FIXEDCELLS': 'SCREEN_peaks/VIB_hydrop_22.FIXEDCELLS__SCREEN_consensus_peaks.bed'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions_paths_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"SCREEN_peaks/*consensus_peaks.bed\"))\n",
    "}\n",
    "regions_paths_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "49accb21",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-16T16:02:24.107817Z",
     "iopub.status.busy": "2022-09-16T16:02:24.107384Z",
     "iopub.status.idle": "2022-09-16T16:02:24.113967Z",
     "shell.execute_reply": "2022-09-16T16:02:24.113365Z"
    },
    "papermill": {
     "duration": 0.012785,
     "end_time": "2022-09-16T16:02:24.114945",
     "exception": false,
     "start_time": "2022-09-16T16:02:24.102160",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "regions_paths_dict[\"VIB_hydrop_11.FIXEDCELLS\"] = \"SCREEN_peaks/VIB_hydrop_1.FIXEDCELLS__SCREEN_consensus_peaks.bed\"\n",
    "regions_paths_dict[\"VIB_hydrop_12.FIXEDCELLS\"] = \"SCREEN_peaks/VIB_hydrop_1.FIXEDCELLS__SCREEN_consensus_peaks.bed\"\n",
    "regions_paths_dict[\"VIB_hydrop_21.FIXEDCELLS\"] = \"SCREEN_peaks/VIB_hydrop_2.FIXEDCELLS__SCREEN_consensus_peaks.bed\"\n",
    "regions_paths_dict[\"VIB_hydrop_22.FIXEDCELLS\"] = \"SCREEN_peaks/VIB_hydrop_2.FIXEDCELLS__SCREEN_consensus_peaks.bed\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c354a43f",
   "metadata": {
    "papermill": {
     "duration": 0.004561,
     "end_time": "2022-09-16T16:02:24.124269",
     "exception": false,
     "start_time": "2022-09-16T16:02:24.119708",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Now, make a sub dictionary of all samples within the fragments dict that have not been run yet (good for resuming a stopped cistopic run):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5cb86b44",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-16T16:02:24.134560Z",
     "iopub.status.busy": "2022-09-16T16:02:24.134076Z",
     "iopub.status.idle": "2022-09-16T16:02:24.138906Z",
     "shell.execute_reply": "2022-09-16T16:02:24.138179Z"
    },
    "papermill": {
     "duration": 0.010994,
     "end_time": "2022-09-16T16:02:24.139883",
     "exception": false,
     "start_time": "2022-09-16T16:02:24.128889",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "cistopic_qc_out = os.path.join(wdir, \"cistopic_qc_out_CONSENSUS\")\n",
    "if not os.path.exists(cistopic_qc_out):\n",
    "    os.makedirs(cistopic_qc_out)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "72462d0a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-16T16:02:24.150300Z",
     "iopub.status.busy": "2022-09-16T16:02:24.149791Z",
     "iopub.status.idle": "2022-09-16T16:02:24.171383Z",
     "shell.execute_reply": "2022-09-16T16:02:24.170766Z"
    },
    "papermill": {
     "duration": 0.027707,
     "end_time": "2022-09-16T16:02:24.172368",
     "exception": false,
     "start_time": "2022-09-16T16:02:24.144661",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/BIO_ddseq_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/BIO_ddseq_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/BIO_ddseq_3.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/BIO_ddseq_4.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/BRO_mtscatac_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/BRO_mtscatac_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xmultiome_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xmultiome_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xv11_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xv11_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xv11_3.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xv11_4.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xv11_5.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xv2_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_10xv2_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_hydrop_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_hydrop_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_hydrop_3.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_mtscatac_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/CNA_mtscatac_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/EPF_hydrop_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/EPF_hydrop_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/EPF_hydrop_3.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/EPF_hydrop_4.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/HAR_ddseq_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/HAR_ddseq_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/MDC_mtscatac_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/MDC_mtscatac_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/OHS_s3atac_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/OHS_s3atac_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/SAN_10xmultiome_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/SAN_10xmultiome_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/STA_10xv11_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/STA_10xv11_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/TXG_10xv11_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/TXG_10xv2_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/TXG_10xv2_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/UCS_ddseq_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/UCS_ddseq_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_10xmultiome_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_10xmultiome_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_10xv1_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_10xv1_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_10xv2_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_10xv2_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_hydrop_1.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_hydrop_11.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata does not exist, adding to subdict to generate\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_hydrop_12.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata does not exist, adding to subdict to generate\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_hydrop_2.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata exists! Skipping...\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_hydrop_21.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata does not exist, adding to subdict to generate\n",
      "Checking if /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS/VIB_hydrop_22.FIXEDCELLS__metadata_bc.pkl exist...\n",
      "\tMetadata does not exist, adding to subdict to generate\n"
     ]
    }
   ],
   "source": [
    "fragments_sub_dict = {}\n",
    "regions_sub_dict = {}\n",
    "for sample in regions_paths_dict.keys():\n",
    "    metadata_file = os.path.join(cistopic_qc_out, sample + \"__metadata_bc.pkl\")\n",
    "    print(f\"Checking if {metadata_file} exist...\")\n",
    "    if os.path.exists(metadata_file):\n",
    "        print(\"\\tMetadata exists! Skipping...\")\n",
    "    else:\n",
    "        fragments_sub_dict[sample] = fragments_dict[sample]\n",
    "        print(\"\\tMetadata does not exist, adding to subdict to generate\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c18c1acf",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-16T16:02:24.183234Z",
     "iopub.status.busy": "2022-09-16T16:02:24.182614Z",
     "iopub.status.idle": "2022-09-16T16:02:24.188105Z",
     "shell.execute_reply": "2022-09-16T16:02:24.187534Z"
    },
    "papermill": {
     "duration": 0.011684,
     "end_time": "2022-09-16T16:02:24.189068",
     "exception": false,
     "start_time": "2022-09-16T16:02:24.177384",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'VIB_hydrop_11.FIXEDCELLS': 'SCREEN_peaks/VIB_hydrop_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_12.FIXEDCELLS': 'SCREEN_peaks/VIB_hydrop_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_21.FIXEDCELLS': 'SCREEN_peaks/VIB_hydrop_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'VIB_hydrop_22.FIXEDCELLS': 'SCREEN_peaks/VIB_hydrop_2.FIXEDCELLS__SCREEN_consensus_peaks.bed'}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "regions_sub_dict = {x: regions_paths_dict[x] for x in sorted(fragments_sub_dict.keys())}\n",
    "regions_sub_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "184a3961",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-16T16:02:24.199730Z",
     "iopub.status.busy": "2022-09-16T16:02:24.199306Z",
     "iopub.status.idle": "2022-09-16T16:02:24.202750Z",
     "shell.execute_reply": "2022-09-16T16:02:24.202161Z"
    },
    "papermill": {
     "duration": 0.009728,
     "end_time": "2022-09-16T16:02:24.203717",
     "exception": false,
     "start_time": "2022-09-16T16:02:24.193989",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "ray.shutdown()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ad1e1c4d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-16T16:02:24.214490Z",
     "iopub.status.busy": "2022-09-16T16:02:24.214008Z",
     "iopub.status.idle": "2022-09-16T16:14:56.991547Z",
     "shell.execute_reply": "2022-09-16T16:14:56.990713Z"
    },
    "papermill": {
     "duration": 752.884288,
     "end_time": "2022-09-16T16:14:57.092887",
     "exception": false,
     "start_time": "2022-09-16T16:02:24.208599",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-09-16 16:02:24,220 cisTopic     INFO     n_cpu is larger than the number of samples. Setting n_cpu to the number of samples\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-16 16:02:28,277\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:02:42,302 cisTopic     INFO     Reading VIB_hydrop_12.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:02:42,305 cisTopic     INFO     Reading VIB_hydrop_11.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:02:42,360 cisTopic     INFO     Reading VIB_hydrop_21.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:02:42,362 cisTopic     INFO     Reading VIB_hydrop_22.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:03:21,297 cisTopic     INFO     Computing barcode rank plot for VIB_hydrop_22.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:03:21,297 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:03:23,680 cisTopic     INFO     Marking barcodes with more than 10\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:03:23,786 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:03:23,808 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:03:24,542 cisTopic     INFO     Computing barcode rank plot for VIB_hydrop_11.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:03:24,542 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:03:25,991 cisTopic     INFO     Computing duplicate rate plot for VIB_hydrop_22.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:03:27,089 cisTopic     INFO     Marking barcodes with more than 10\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:03:27,204 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:03:27,220 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:03:29,460 cisTopic     INFO     Computing duplicate rate plot for VIB_hydrop_11.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:03:29,809 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:03:30,119 cisTopic     INFO     Computing insert size distribution for VIB_hydrop_22.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:03:30,120 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:03:30,861 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:03:33,391 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:03:33,703 cisTopic     INFO     Computing insert size distribution for VIB_hydrop_11.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:03:33,703 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:03:34,489 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:03:40,803 cisTopic     INFO     Computing barcode rank plot for VIB_hydrop_12.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:03:40,804 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:03:43,866 cisTopic     INFO     Marking barcodes with more than 10\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:03:43,894 cisTopic     INFO     Computing TSS profile for VIB_hydrop_22.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:03:43,982 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:03:44,003 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:03:47,031 cisTopic     INFO     Computing duplicate rate plot for VIB_hydrop_12.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:03:49,191 cisTopic     INFO     Computing TSS profile for VIB_hydrop_11.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:03:50,059 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:03:50,301 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:03:52,107 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:03:52,417 cisTopic     INFO     Computing insert size distribution for VIB_hydrop_12.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:03:52,417 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:03:53,538 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:03:56,049 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:03:56,294 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:04:13,768 cisTopic     INFO     Computing TSS profile for VIB_hydrop_12.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:04:14,404 cisTopic     INFO     Computing barcode rank plot for VIB_hydrop_21.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:04:14,404 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:04:18,880 cisTopic     INFO     Marking barcodes with more than 10\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:04:19,025 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:04:19,086 cisTopic     INFO     Returning valid barcodes\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:04:23,238 cisTopic     INFO     Computing duplicate rate plot for VIB_hydrop_21.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:04:23,499 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:04:23,745 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:04:30,167 cisTopic     INFO     Return plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:04:30,484 cisTopic     INFO     Computing insert size distribution for VIB_hydrop_21.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:04:30,484 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:04:32,219 cisTopic     INFO     Returning plot data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:05:03,597 cisTopic     INFO     Computing TSS profile for VIB_hydrop_21.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:05:18,858 cisTopic     INFO     Formatting annnotation\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:05:19,122 cisTopic     INFO     Creating coverage matrix\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:05:27,278 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:05:34,010 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:06:25,742 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:08:24,140 cisTopic     INFO     Coverage matrix done\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:09:47,245 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:09:56,405 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:11:13,308 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:12:39,245 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:12:39,617 cisTopic     INFO     Computing FRIP profile for VIB_hydrop_22.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:12:41,018 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:12:43,644 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:12:50,836 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:12:51,174 cisTopic     INFO     Computing FRIP profile for VIB_hydrop_11.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:12:52,443 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490946)\u001b[0m 2022-09-16 16:12:54,183 cisTopic     INFO     Sample VIB_hydrop_22.FIXEDCELLS done!\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:12:55,904 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490948)\u001b[0m 2022-09-16 16:13:04,256 cisTopic     INFO     Sample VIB_hydrop_11.FIXEDCELLS done!\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:13:18,455 cisTopic     INFO     Returning normalized TSS coverage matrix per barcode\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:13:45,957 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:13:46,123 cisTopic     INFO     Computing FRIP profile for VIB_hydrop_12.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:13:46,726 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:13:48,358 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490945)\u001b[0m 2022-09-16 16:13:54,691 cisTopic     INFO     Sample VIB_hydrop_12.FIXEDCELLS done!\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:14:29,902 cisTopic     INFO     Returning normalized sample TSS enrichment data\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:14:29,988 cisTopic     INFO     Computing FRIP profile for VIB_hydrop_21.FIXEDCELLS\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:14:30,396 cisTopic     INFO     Counting fragments\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:14:31,518 cisTopic     INFO     Intersecting fragments with regions\n",
      "\u001b[2m\u001b[36m(compute_qc_stats_ray pid=3490947)\u001b[0m 2022-09-16 16:14:36,385 cisTopic     INFO     Sample VIB_hydrop_21.FIXEDCELLS done!\n",
      "Dumping files in /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus/cistopic_qc_out_CONSENSUS...\n"
     ]
    }
   ],
   "source": [
    "n_cores = 8\n",
    "if regions_sub_dict != {}:\n",
    "    samples_sub = list(regions_sub_dict.keys())\n",
    "    blocks = [samples_sub[i : i + n_cores] for i in range(0, len(samples_sub), n_cores)]\n",
    "    for samples_torun_in_block in blocks:\n",
    "        fragments_sub_dict_block = {\n",
    "            key: fragments_sub_dict[key] for key in samples_torun_in_block\n",
    "        }\n",
    "        regions_sub_dict_block = {\n",
    "            key: regions_sub_dict[key] for key in samples_torun_in_block\n",
    "        }\n",
    "\n",
    "        metadata_bc_dict, profile_data_dict = compute_qc_stats(\n",
    "            fragments_dict=fragments_sub_dict_block,\n",
    "            tss_annotation=annotation,\n",
    "            stats=[\n",
    "                \"barcode_rank_plot\",\n",
    "                \"duplicate_rate\",\n",
    "                \"insert_size_distribution\",\n",
    "                \"profile_tss\",\n",
    "                \"frip\",\n",
    "            ],\n",
    "            label_list=None,\n",
    "            path_to_regions=regions_sub_dict_block,\n",
    "            n_cpu=n_cores,\n",
    "            valid_bc=None,\n",
    "            n_frag=10,\n",
    "            n_bc=None,\n",
    "            tss_flank_window=2000,\n",
    "            tss_window=50,\n",
    "            tss_minimum_signal_window=100,\n",
    "            tss_rolling_window=10,\n",
    "            # min_norm=0.2,\n",
    "            remove_duplicates=True,\n",
    "        )\n",
    "\n",
    "        ray.shutdown()\n",
    "        print(f\"Dumping files in {cistopic_qc_out}...\")\n",
    "        for sample in sorted(metadata_bc_dict.keys()):\n",
    "            metadata_bc_dict[sample][\"sample_id\"] = sample\n",
    "            metadata_bc_dict[sample].index = [\n",
    "                x + \"___\" + sample for x in list(metadata_bc_dict[sample].index)\n",
    "            ]\n",
    "            with open(\n",
    "                os.path.join(cistopic_qc_out, f\"{sample}__metadata_bc.pkl\"), \"wb\"\n",
    "            ) as f:\n",
    "                pickle.dump(metadata_bc_dict[sample], f, protocol=4)\n",
    "\n",
    "            with open(\n",
    "                os.path.join(cistopic_qc_out, f\"{sample}__profile_data.pkl\"), \"wb\"\n",
    "            ) as f:\n",
    "                pickle.dump(profile_data_dict[sample], f, protocol=4)\n",
    "else:\n",
    "    print(\"All samples already processed.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e60f1a39",
   "metadata": {
    "papermill": {
     "duration": 0.192502,
     "end_time": "2022-09-16T16:14:57.305321",
     "exception": false,
     "start_time": "2022-09-16T16:14:57.112819",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Plot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f2c61cf",
   "metadata": {
    "papermill": {
     "duration": 0.007557,
     "end_time": "2022-09-16T16:14:57.321141",
     "exception": false,
     "start_time": "2022-09-16T16:14:57.313584",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Calculating a KDE is simultaneously expensive and scales poorly with increasing n. Therefore, I wrote a multithreaded script that divides the QC array into equal parts (interleaved to avoid biases in the order!) and performs a KDE calculation on each part. Here, Otsu thresholding is used to find the right threshold for minimum fragments and minimum TSS enrichment. ddseq samples have a significantly higher noise floor than the other samples when it comes to fragment distribution. Therefore, the otsu algorithm is performed on all barcodes with a minimum of 300 fragments for the ddseq samples, and a minimum of 100 fragments for all the other samples. I tried to perform this filtering completely independent of sample/technique (e.g. using gaussian mixture modeling, Jenks natural breaks, or multiple step Otsu thresholding) but found that no solution worked perfectly for all samples."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3848ed02",
   "metadata": {
    "papermill": {
     "duration": 0.00763,
     "end_time": "2022-09-16T16:14:57.336243",
     "exception": false,
     "start_time": "2022-09-16T16:14:57.328613",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "This is regulated by the code below in qc_plots.py:\n",
    "```\n",
    "min_otsu_frags_dict = {}\n",
    "for fragments_file in fragments_list:\n",
    "    sample = fragments_file.split(\"/\")[-1].split(\".\")[0]\n",
    "    tech = sample.split('_')[1]\n",
    "    if tech == \"ddseq\":\n",
    "        if sample == \"BIO_ddseq_1\":\n",
    "            min_otsu_frags_dict[sample] = 600\n",
    "        else:\n",
    "            min_otsu_frags_dict[sample] = 300\n",
    "    elif tech == \"hydrop\":\n",
    "        min_otsu_frags_dict[sample] = 300\n",
    "    else:\n",
    "        min_otsu_frags_dict[sample] = 100\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6af02c2a",
   "metadata": {
    "collapsed": true,
    "execution": {
     "iopub.execute_input": "2022-09-16T16:14:57.353163Z",
     "iopub.status.busy": "2022-09-16T16:14:57.352479Z",
     "iopub.status.idle": "2022-09-16T16:14:57.579371Z",
     "shell.execute_reply": "2022-09-16T16:14:57.578490Z"
    },
    "jupyter": {
     "outputs_hidden": true
    },
    "papermill": {
     "duration": 0.237042,
     "end_time": "2022-09-16T16:14:57.580765",
     "exception": false,
     "start_time": "2022-09-16T16:14:57.343723",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "import kde\n",
      "import pycisTopic\n",
      "import glob\n",
      "import os\n",
      "import pandas as pd\n",
      "import pickle\n",
      "import matplotlib.pyplot as plt\n",
      "import numpy as np\n",
      "import seaborn as sns\n",
      "import multiprocess as mp\n",
      "from multiprocess import Pool\n",
      "import pprint as pp\n",
      "\n",
      "def histogram(array, nbins=100):\n",
      "    \"\"\"\n",
      "    Draw histogram from distribution and identify centers.\n",
      "    Parameters\n",
      "    ---------\n",
      "    array: `class::np.array`\n",
      "            Scores distribution\n",
      "    nbins: int\n",
      "            Number of bins to use in the histogram\n",
      "    Return\n",
      "    ---------\n",
      "    float\n",
      "            Histogram values and bin centers.\n",
      "    \"\"\"\n",
      "    array = array.ravel().flatten()\n",
      "    hist, bin_edges = np.histogram(array, bins=nbins, range=None)\n",
      "    bin_centers = (bin_edges[:-1] + bin_edges[1:]) / 2.0\n",
      "    return hist, bin_centers\n",
      "\n",
      "\n",
      "def threshold_otsu(array, nbins=100, min_value=100):\n",
      "    \"\"\"\n",
      "    Apply Otsu threshold on topic-region distributions [Otsu, 1979].\n",
      "    Parameters\n",
      "    ---------\n",
      "    array: `class::np.array`\n",
      "            Array containing the region values for the topic to be binarized.\n",
      "    nbins: int\n",
      "            Number of bins to use in the binarization histogram\n",
      "    Return\n",
      "    ---------\n",
      "    float\n",
      "            Binarization threshold.\n",
      "    Reference\n",
      "    ---------\n",
      "    Otsu, N., 1979. A threshold selection method from gray-level histograms. IEEE transactions on systems, man, and\n",
      "    cybernetics, 9(1), pp.62-66.\n",
      "    \"\"\"\n",
      "    array = array[(array >= min_value)]\n",
      "    hist, bin_centers = histogram(array, nbins)\n",
      "    hist = hist.astype(float)\n",
      "    # Class probabilities for all possible thresholds\n",
      "    weight1 = np.cumsum(hist)\n",
      "    weight2 = np.cumsum(hist[::-1])[::-1]\n",
      "    # Class means for all possible thresholds\n",
      "    mean1 = np.cumsum(hist * bin_centers) / weight1\n",
      "    mean2 = (np.cumsum((hist * bin_centers)[::-1]) / weight2[::-1])[::-1]\n",
      "    # Clip ends to align class 1 and class 2 variables:\n",
      "    # The last value of ``weight1``/``mean1`` should pair with zero values in\n",
      "    # ``weight2``/``mean2``, which do not exist.\n",
      "    variance12 = weight1[:-1] * weight2[1:] * (mean1[:-1] - mean2[1:]) ** 2\n",
      "    idx = np.argmax(variance12)\n",
      "    threshold = bin_centers[:-1][idx]\n",
      "    return threshold\n",
      "\n",
      "\n",
      "# def calc_kde(xy):\n",
      "#     return gaussian_kde(xy)(xy)\n",
      "\n",
      "\n",
      "def plot_frag_qc(\n",
      "    x,\n",
      "    y,\n",
      "    ax,\n",
      "    x_thr_min=None,\n",
      "    x_thr_max=None,\n",
      "    y_thr_min=None,\n",
      "    y_thr_max=None,\n",
      "    ylab=None,\n",
      "    xlab=\"Number of (unique) fragments\",\n",
      "    cmap=\"viridis\",\n",
      "    density_overlay=False,\n",
      "    s=10,\n",
      "    marker=\"+\",\n",
      "    c=\"#343434\",\n",
      "    xlim=None,\n",
      "    ylim=None,\n",
      "    **kwargs\n",
      "):\n",
      "    from scipy.stats import gaussian_kde\n",
      "\n",
      "    assert all(x.index == y.index)\n",
      "    barcodes = x.index.values\n",
      "\n",
      "    if density_overlay:\n",
      "        cores = 8\n",
      "\n",
      "        x_log = np.log(x)\n",
      "\n",
      "        # Split input array for KDE [log(x), y] array in\n",
      "        # equaly spaced parts (start_offset + n * nbr_cores).\n",
      "        kde_parts = [np.vstack([x_log[i::cores], y[i::cores]]) for i in range(cores)]\n",
      "\n",
      "        # Get nultiprocess context object to spawn processes.\n",
      "        # mp_ctx = mp.get_context(\"spawn\")\n",
      "\n",
      "        # Calculate KDE in parallel.\n",
      "        with Pool(processes=cores) as pool:\n",
      "            results = pool.map(kde.calc_kde, kde_parts)\n",
      "\n",
      "        z = np.concatenate(results)\n",
      "\n",
      "        # now order x and y in the same way that z was ordered, otherwise random z value is assigned to barcode:\n",
      "        x_ordered = np.concatenate([x[i::cores] for i in range(cores)])\n",
      "        y_ordered = np.concatenate([y[i::cores] for i in range(cores)])\n",
      "\n",
      "        idx = (\n",
      "            z.argsort()\n",
      "        )  # order based on z value so that highest value is plotted on top, and not hidden by lower values\n",
      "        x, y, z, barcodes = x_ordered[idx], y_ordered[idx], z[idx], barcodes[idx]\n",
      "    else:\n",
      "        z = c\n",
      "\n",
      "    sp = ax.scatter(x, y, c=z, s=s, edgecolors=None, marker=marker, cmap=cmap, **kwargs)\n",
      "    if ylim is not None:\n",
      "        ax.set_ylim(ylim[0], ylim[1])\n",
      "    if xlim is not None:\n",
      "        ax.set_xlim(xlim[0], xlim[1])\n",
      "\n",
      "    # Start with keeping all barcodes.\n",
      "    barcodes_to_keep = np.full(x.shape[0], True)\n",
      "\n",
      "    # Filter barcodes out if needed based on thresholds:\n",
      "    if x_thr_min is not None:\n",
      "        ax.axvline(x=x_thr_min, color=\"r\", linestyle=\"--\")\n",
      "\n",
      "    if x_thr_max is not None:\n",
      "        ax.axvline(x=x_thr_max, color=\"r\", linestyle=\"--\")\n",
      "\n",
      "    if y_thr_min is not None:\n",
      "        ax.axhline(y=y_thr_min, color=\"r\", linestyle=\"--\")\n",
      "\n",
      "    if y_thr_max is not None:\n",
      "        ax.axhline(y=y_thr_max, color=\"r\", linestyle=\"--\")\n",
      "\n",
      "    ax.set_xscale(\"log\")\n",
      "    ax.set_xmargin(0.01)\n",
      "    ax.set_ymargin(0.01)\n",
      "    ax.set_xlabel(xlab, fontsize=10)\n",
      "    ax.set_ylabel(ylab, fontsize=10)\n",
      "\n",
      "def plot_qc(\n",
      "    sample,\n",
      "    min_dict,\n",
      "    max_dict,\n",
      "    metadata_bc_df,\n",
      "    include_kde=False,\n",
      "    detailed_title=True,\n",
      "    s=4,\n",
      "    min_x_val=100,\n",
      "    min_y_val=1\n",
      "):\n",
      "    fig, (ax1, ax2, ax3) = plt.subplots(1, 3, figsize=(12, 4), dpi=150)\n",
      "\n",
      "    # calculate thresholds using a double otsu strategy:\n",
      "    # first we calculate an otsu threshold using a minimum of 10 fragments\n",
      "    # then, this otsu threshold is used as the minimum for the second iteration\n",
      "\n",
      "    x_arr = np.log10(metadata_bc_df[\"Unique_nr_frag\"])\n",
      "    x_threshold_log = threshold_otsu(x_arr, nbins=5000, min_value=np.log10(min_x_val))\n",
      "    x_threshold = 10**x_threshold_log\n",
      "\n",
      "    y_arr = metadata_bc_df[\"TSS_enrichment\"]\n",
      "    y_threshold = threshold_otsu(y_arr, nbins=5000, min_value=min_y_val)\n",
      "\n",
      "    # calculate cells passing filter\n",
      "    metadata_bc_df_passing_filters = metadata_bc_df.loc[\n",
      "        (metadata_bc_df.Unique_nr_frag > x_threshold)\n",
      "        & (metadata_bc_df.TSS_enrichment > y_threshold)\n",
      "    ]\n",
      "    bc_passing_filters = metadata_bc_df_passing_filters.index\n",
      "\n",
      "    # plot everything\n",
      "    plot_frag_qc(\n",
      "        x=metadata_bc_df[\"Unique_nr_frag\"],\n",
      "        y=metadata_bc_df[\"TSS_enrichment\"],\n",
      "        ylab=\"TSS Enrichment\",\n",
      "        s=s,\n",
      "        x_thr_min=x_threshold,\n",
      "        y_thr_min=y_threshold,\n",
      "        xlim=[10, max_dict[\"Unique_nr_frag\"]],\n",
      "        ylim=[0, max_dict[\"TSS_enrichment\"]],\n",
      "        density_overlay=include_kde,\n",
      "        ax=ax1,\n",
      "    )\n",
      "    plot_frag_qc(\n",
      "        x=metadata_bc_df[\"Unique_nr_frag\"],\n",
      "        y=metadata_bc_df[\"FRIP\"],\n",
      "        x_thr_min=x_threshold,\n",
      "        ylab=\"FRIP\",\n",
      "        s=s,\n",
      "        xlim=[10, max_dict[\"Unique_nr_frag\"]],\n",
      "        ylim=[0, 1],\n",
      "        density_overlay=include_kde,\n",
      "        ax=ax2,\n",
      "    )\n",
      "    plot_frag_qc(\n",
      "        x=metadata_bc_df[\"Unique_nr_frag\"],\n",
      "        y=metadata_bc_df[\"Dupl_rate\"],\n",
      "        x_thr_min=x_threshold,\n",
      "        ylab=\"Duplicate rate per cell\",\n",
      "        s=s,\n",
      "        xlim=[10, max_dict[\"Unique_nr_frag\"]],\n",
      "        ylim=[0, 1],\n",
      "        density_overlay=include_kde,\n",
      "        ax=ax3,\n",
      "    )\n",
      "\n",
      "    if detailed_title:\n",
      "        med_nf = round(\n",
      "            metadata_bc_df.loc[bc_passing_filters, \"Unique_nr_frag\"].median(), 2\n",
      "        )\n",
      "        med_tss = round(\n",
      "            metadata_bc_df.loc[bc_passing_filters, \"TSS_enrichment\"].median(), 2\n",
      "        )\n",
      "        med_frip = round(metadata_bc_df.loc[bc_passing_filters, \"FRIP\"].median(), 2)\n",
      "        title = f\"{sample}: Kept {len(bc_passing_filters)} cells using Otsu filtering. Median Unique Fragments: {med_nf:.0f}. Median TSS Enrichment: {med_tss:.2f}. Median FRIP: {med_frip:.2f}\\nUsed a minimum of {x_threshold:.2f} fragments and TSS enrichment of {y_threshold:.2f})\"\n",
      "    else:\n",
      "        title = sample\n",
      "\n",
      "    fig.suptitle(title, x=0.5, y=0.95, fontsize=10)\n",
      "    return bc_passing_filters, fig\n",
      "\n",
      "\n",
      "###################\n",
      "# Run the loop here\n",
      "###################\n",
      "\n",
      "metadata_bc_pkl_list = sorted(glob.glob(\"cistopic_qc_out/*metadata_bc.pkl\"))\n",
      "metadata_bc_pkl_path_dict = {}\n",
      "for metadata_bc_pkl_path in metadata_bc_pkl_list:\n",
      "    sample = metadata_bc_pkl_path.split(\"/\")[-1].split(\"__metadata_bc.pkl\")[0]\n",
      "    metadata_bc_pkl_path_dict[sample] = metadata_bc_pkl_path\n",
      "\n",
      "# calculate global maxima to make nice equal plots:\n",
      "metadata_bc_df_all = pd.DataFrame()\n",
      "for sample in metadata_bc_pkl_path_dict.keys():\n",
      "    with open(metadata_bc_pkl_path_dict[sample], \"rb\") as fh:\n",
      "        metadata_bc_df = pickle.load(fh)\n",
      "        metadata_bc_df_all = pd.concat([metadata_bc_df_all, metadata_bc_df])\n",
      "        print(f\"added {sample} metadata_bc_df\")\n",
      "\n",
      "max_dict = {}\n",
      "min_dict = {}\n",
      "for stat in metadata_bc_df_all.columns:\n",
      "    max_dict[stat] = metadata_bc_df_all[stat].max()\n",
      "    min_dict[stat] = metadata_bc_df_all[stat].min()\n",
      "\n",
      "pp.pprint(min_dict)\n",
      "pp.pprint(max_dict)\n",
      "\n",
      "standard_min_x_val = 100\n",
      "standard_min_y_val = 1\n",
      "min_otsu_frags_dict = {}\n",
      "min_otsu_tss_dict = {}\n",
      "for metadata_bc_pkl_path in sorted(metadata_bc_pkl_path_dict.keys()):\n",
      "    sample = metadata_bc_pkl_path.split(\"/\")[-1].split(\"__metadata_bc.pkl\")[0]\n",
      "    tech = sample.split('_')[1]\n",
      "    if tech == \"ddseq\":\n",
      "        if \"BIO_ddseq_1.FULL\" in sample:\n",
      "            min_otsu_frags_dict[sample] = 600\n",
      "        else:\n",
      "            min_otsu_frags_dict[sample] = 300\n",
      "    elif tech == \"hydrop\":\n",
      "        min_otsu_frags_dict[sample] = 300\n",
      "    else:\n",
      "        min_otsu_frags_dict[sample] = standard_min_x_val\n",
      "\n",
      "    if tech == \"s3atac\":\n",
      "        min_otsu_tss_dict[sample] = 0\n",
      "    else:\n",
      "        min_otsu_tss_dict[sample] = standard_min_y_val\n",
      "\n",
      "pp.pprint(min_otsu_frags_dict)\n",
      "\n",
      "for sample in metadata_bc_pkl_path_dict.keys():\n",
      "    pkl_path = f\"selected_barcodes/{sample}_bc_passing_filters_otsu.pkl\"\n",
      "    if os.path.exists(pkl_path):\n",
      "        print(f\"{pkl_path} bc passing filters exists, skipping...\")\n",
      "    else:\n",
      "        print(f\"{pkl_path} bc passing filters does not exist yet, generating...\")\n",
      "        print(f\"\\tLoading {metadata_bc_pkl_path_dict[sample]}\")\n",
      "        with open(metadata_bc_pkl_path_dict[sample], \"rb\") as fh:\n",
      "            metadata_bc_df = pickle.load(fh)\n",
      "\n",
      "        print(f\"\\tFiltering cells and generating QC plots.\")\n",
      "        if not sample in min_otsu_frags_dict.keys():\n",
      "            print(f\"\\t{sample} not in minimum dict! Using standard value of {standard_min_x_val}\")\n",
      "            min_x_val = standard_min_x_val\n",
      "            min_y_val = standard_min_y_val\n",
      "        else:\n",
      "            min_x_val = min_otsu_frags_dict[sample]\n",
      "            min_y_val = min_otsu_tss_dict[sample]\n",
      "\n",
      "        bc_passing_filters, fig = plot_qc(sample=sample, min_dict=min_dict, max_dict=max_dict, metadata_bc_df=metadata_bc_df, include_kde=kde, min_x_val=min_x_val, min_y_val=min_y_val)\n",
      "\n",
      "        plt.tight_layout()\n",
      "        plt.savefig(f\"plots_qc/{sample}_qc_otsu.png\", dpi=300, facecolor=\"white\")\n",
      "        plt.show()\n",
      "        plt.close()\n",
      "\n",
      "        print(f\"\\tSaving...\")\n",
      "        with open(\n",
      "            f\"selected_barcodes/{sample}_bc_passing_filters_otsu.pkl\", \"wb\"\n",
      "        ) as fh:\n",
      "            pickle.dump(bc_passing_filters, fh)\n",
      "        fh.close()\n",
      "\n",
      "        fh = open(f\"selected_barcodes/{sample}_bc_passing_filters_otsu.txt\", \"w\")\n",
      "        for bc in list(bc_passing_filters):\n",
      "            fh.write(bc + \"\\n\")\n",
      "        fh.close()\n",
      "\n",
      "        metadata_bc_df.loc[bc_passing_filters].to_csv(f'selected_barcodes/{sample}_metadata_bc_df.tsv', sep='\\t')\n"
     ]
    }
   ],
   "source": [
    "!cat ../0_resources/scripts/qc_plots.py"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9d71135",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-07-15T09:42:06.802579Z",
     "iopub.status.busy": "2022-07-15T09:42:06.801860Z",
     "iopub.status.idle": "2022-07-15T09:42:06.814527Z",
     "shell.execute_reply": "2022-07-15T09:42:06.813708Z",
     "shell.execute_reply.started": "2022-07-15T09:42:06.802476Z"
    },
    "papermill": {
     "duration": 0.007834,
     "end_time": "2022-09-16T16:14:57.596671",
     "exception": false,
     "start_time": "2022-09-16T16:14:57.588837",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Since multiprocessing does not work with jupyter notebooks, run the following code in terminal:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2df89954",
   "metadata": {
    "papermill": {
     "duration": 0.007778,
     "end_time": "2022-09-16T16:14:57.612345",
     "exception": false,
     "start_time": "2022-09-16T16:14:57.604567",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "```\n",
    "mkdir plots_qc\n",
    "mkdir selected_barcodes\n",
    "SIF=../0_resources/cistopic_image/20220722_pycistopic.sif\n",
    "singularity exec \\\n",
    "    --cleanenv \\\n",
    "    -H $PWD:/home \\\n",
    "    $SIF \\\n",
    "    python ../0_resources/scripts/qc_plots.py\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a032507a",
   "metadata": {
    "papermill": {
     "duration": 0.007673,
     "end_time": "2022-09-16T16:14:57.627647",
     "exception": false,
     "start_time": "2022-09-16T16:14:57.619974",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "And then open the plots:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6f377694",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-16T16:14:57.644090Z",
     "iopub.status.busy": "2022-09-16T16:14:57.643776Z",
     "iopub.status.idle": "2022-09-16T16:14:57.653108Z",
     "shell.execute_reply": "2022-09-16T16:14:57.652479Z"
    },
    "papermill": {
     "duration": 0.018828,
     "end_time": "2022-09-16T16:14:57.654123",
     "exception": false,
     "start_time": "2022-09-16T16:14:57.635295",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "metadata_bc_pkl_list = sorted(glob.glob(\"cistopic_qc_out/*metadata_bc.pkl\"))\n",
    "metadata_bc_pkl_path_dict = {}\n",
    "for metadata_bc_pkl_path in metadata_bc_pkl_list:\n",
    "    sample = metadata_bc_pkl_path.split(\"/\")[-1].split(\"__\")[0]\n",
    "    metadata_bc_pkl_path_dict[sample] = metadata_bc_pkl_path\n",
    "\n",
    "for sample in metadata_bc_pkl_path_dict.keys():\n",
    "    if os.path.exists(f\"selected_barcodes/{sample}_bc_passing_filters_otsu.pkl\"):\n",
    "        print(f\"{sample} bc passing filters exists, printing img and skipping\")\n",
    "        display(Image(f\"plots_qc/{sample}_qc_otsu.png\"))\n",
    "    else:\n",
    "        print(\n",
    "            f\"{sample} bc passing filters does not exist yet, generate using qc_plots.py script!\"\n",
    "        )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba735673",
   "metadata": {
    "papermill": {
     "duration": 0.007656,
     "end_time": "2022-09-16T16:14:57.669612",
     "exception": false,
     "start_time": "2022-09-16T16:14:57.661956",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Write raw barcode file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d29079d2",
   "metadata": {
    "papermill": {
     "duration": 0.007547,
     "end_time": "2022-09-16T16:14:57.684813",
     "exception": false,
     "start_time": "2022-09-16T16:14:57.677266",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "Freemuxlet reads barcodes in the bam file. These barcodes (tag: DB) do not contain the `sample` suffix, which is included in the `*_bc_passing_filters_otsu.txt` files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7bb30686",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-16T16:14:57.701086Z",
     "iopub.status.busy": "2022-09-16T16:14:57.700651Z",
     "iopub.status.idle": "2022-09-16T16:14:57.952708Z",
     "shell.execute_reply": "2022-09-16T16:14:57.951801Z"
    },
    "papermill": {
     "duration": 0.261706,
     "end_time": "2022-09-16T16:14:57.954284",
     "exception": false,
     "start_time": "2022-09-16T16:14:57.692578",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAGGTTCTTTCCTAAGAGACC_CCGCGATCACGCCAATTATTC_CTGCGCCTGTTTAGATTGCCT___BIO_ddseq_1.FULL\n",
      "GAGAGGTGACAATATGGTGAA___BIO_ddseq_1.FULL\n",
      "CGATTACGCTCCAAGAGCTAA_CTAGAGCACACGCGTACTTTC_GGCGTCCTAACGCCCGTAGCC___BIO_ddseq_1.FULL\n",
      "CTAGAGCGGTTAGTGTGGCGC_GCTTTATTTGGACTCGAGGTC_GTTGATGTCATACAGTGGTAG___BIO_ddseq_1.FULL\n",
      "CGCGCTTTCTTGTAAGCCGCC_CTTGTCCCGTGATATGATTGT___BIO_ddseq_1.FULL\n",
      "AGCCGCCGCACAGCCCATAAT_CGAGTGGTGGTGAATAGTGTT_GAATCAACGAGTGGCAGGCGG___BIO_ddseq_1.FULL\n",
      "CACATGATGCGAGCCTCATTT_CAGAGAGTGGTGAAACCTACC_GAATATGTGACCGCTGGTGAA_TCAAGACCGGATCTATTGCCT___BIO_ddseq_1.FULL\n",
      "CGAGTGGGAGGACAAGATAGG_CGTGATAATGAATATGCGAGC_TCAAGACGTGTGTCAACTCTT___BIO_ddseq_1.FULL\n",
      "CTCTTGCTGCGTTGCGGCTGC_TCGACAGCGGTACGAACGTAA___BIO_ddseq_1.FULL\n",
      "CCACTTGCACATGACTAACTC_TATGTGCACGAGAACCAAGCT___BIO_ddseq_1.FULL\n"
     ]
    }
   ],
   "source": [
    "!head selected_barcodes/BIO_ddseq_1.FIXEDCELLS_bc_passing_filters_otsu.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "430a8854",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-16T16:14:57.971384Z",
     "iopub.status.busy": "2022-09-16T16:14:57.971074Z",
     "iopub.status.idle": "2022-09-16T16:14:59.079107Z",
     "shell.execute_reply": "2022-09-16T16:14:59.078326Z"
    },
    "papermill": {
     "duration": 1.117432,
     "end_time": "2022-09-16T16:14:59.080167",
     "exception": false,
     "start_time": "2022-09-16T16:14:57.962735",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "selected_barcodes/MDC_mtscatac_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "MDC_mtscatac_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/STA_10xv11_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "STA_10xv11_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_10xv11_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_10xv11_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_10xmultiome_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_10xmultiome_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/SAN_10xmultiome_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "SAN_10xmultiome_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/VIB_10xmultiome_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "VIB_10xmultiome_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/STA_10xv11_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "STA_10xv11_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/VIB_hydrop_11.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "VIB_hydrop_11.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/MDC_mtscatac_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "MDC_mtscatac_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/EPF_hydrop_3.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "EPF_hydrop_3.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/BIO_ddseq_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "BIO_ddseq_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/OHS_s3atac_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "OHS_s3atac_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_10xv11_3.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_10xv11_3.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_mtscatac_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_mtscatac_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/OHS_s3atac_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "OHS_s3atac_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/VIB_10xv2_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "VIB_10xv2_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_10xmultiome_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_10xmultiome_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/VIB_10xv2_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "VIB_10xv2_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/TXG_10xv2_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "TXG_10xv2_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/HAR_ddseq_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "HAR_ddseq_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_10xv11_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_10xv11_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/TXG_10xv2_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "TXG_10xv2_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/VIB_hydrop_12.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "VIB_hydrop_12.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_ddseq_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_ddseq_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_10xv11_4.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_10xv11_4.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/VIB_10xv1_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "VIB_10xv1_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_mtscatac_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_mtscatac_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/BIO_ddseq_4.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "BIO_ddseq_4.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_10xv2_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_10xv2_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/SAN_10xmultiome_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "SAN_10xmultiome_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_ddseq_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_ddseq_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/VIB_mtscatac_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "VIB_mtscatac_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/VIB_mtscatac_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "VIB_mtscatac_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_10xv2_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_10xv2_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_hydrop_3.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_hydrop_3.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/BIO_ddseq_3.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "BIO_ddseq_3.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/BRO_mtscatac_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "BRO_mtscatac_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/EPF_hydrop_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "EPF_hydrop_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_hydrop_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_hydrop_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/EPF_hydrop_4.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "EPF_hydrop_4.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_hydrop_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_hydrop_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/CNA_10xv11_5.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "CNA_10xv11_5.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/VIB_10xmultiome_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "VIB_10xmultiome_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/TXG_10xv11_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "TXG_10xv11_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/EPF_hydrop_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "EPF_hydrop_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/HAR_ddseq_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "HAR_ddseq_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/BRO_mtscatac_2.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "BRO_mtscatac_2.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/VIB_hydrop_22.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "VIB_hydrop_22.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/VIB_10xv1_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "VIB_10xv1_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/BIO_ddseq_1.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "BIO_ddseq_1.FIXEDCELLS\n",
      "Done\n",
      "selected_barcodes/VIB_hydrop_21.FIXEDCELLS_bc_passing_filters_otsu.txt\n",
      "VIB_hydrop_21.FIXEDCELLS\n",
      "Done\n"
     ]
    }
   ],
   "source": [
    "for file in glob.glob(\"selected_barcodes/*_bc_passing_filters_otsu.txt\"):\n",
    "    print(file)\n",
    "    sample = file.split(\"/\")[-1].split(\"_bc_passing_filters_otsu.txt\")[0]\n",
    "    print(sample)\n",
    "    df = pd.read_csv(file, header=None, index_col=None)\n",
    "\n",
    "    with open(f\"selected_barcodes/{sample}_bc_passing_filters_otsu.RAW.txt\", \"w\") as fp:\n",
    "        for x in df[0]:\n",
    "            # write each item on a new line\n",
    "            fp.write(x.split(\"___\")[0] + \"\\n\")\n",
    "        print(\"Done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c1ed1b9d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-16T16:14:59.097768Z",
     "iopub.status.busy": "2022-09-16T16:14:59.097486Z",
     "iopub.status.idle": "2022-09-16T16:14:59.309378Z",
     "shell.execute_reply": "2022-09-16T16:14:59.308470Z"
    },
    "papermill": {
     "duration": 0.222072,
     "end_time": "2022-09-16T16:14:59.310760",
     "exception": false,
     "start_time": "2022-09-16T16:14:59.088688",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAGGTTCTTTCCTAAGAGACC_CCGCGATCACGCCAATTATTC_CTGCGCCTGTTTAGATTGCCT\n",
      "GAGAGGTGACAATATGGTGAA\n",
      "CGATTACGCTCCAAGAGCTAA_CTAGAGCACACGCGTACTTTC_GGCGTCCTAACGCCCGTAGCC\n",
      "CTAGAGCGGTTAGTGTGGCGC_GCTTTATTTGGACTCGAGGTC_GTTGATGTCATACAGTGGTAG\n",
      "CGCGCTTTCTTGTAAGCCGCC_CTTGTCCCGTGATATGATTGT\n",
      "AGCCGCCGCACAGCCCATAAT_CGAGTGGTGGTGAATAGTGTT_GAATCAACGAGTGGCAGGCGG\n",
      "CACATGATGCGAGCCTCATTT_CAGAGAGTGGTGAAACCTACC_GAATATGTGACCGCTGGTGAA_TCAAGACCGGATCTATTGCCT\n",
      "CGAGTGGGAGGACAAGATAGG_CGTGATAATGAATATGCGAGC_TCAAGACGTGTGTCAACTCTT\n",
      "CTCTTGCTGCGTTGCGGCTGC_TCGACAGCGGTACGAACGTAA\n",
      "CCACTTGCACATGACTAACTC_TATGTGCACGAGAACCAAGCT\n"
     ]
    }
   ],
   "source": [
    "!head selected_barcodes/BIO_ddseq_1.FIXEDCELLS_bc_passing_filters_otsu.RAW.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fe4aee4d",
   "metadata": {
    "papermill": {
     "duration": 0.00829,
     "end_time": "2022-09-16T16:14:59.327651",
     "exception": false,
     "start_time": "2022-09-16T16:14:59.319361",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "20221014_pycistopic.sif",
   "language": "python",
   "name": "cistopic_20221014"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 779.375713,
   "end_time": "2022-09-16T16:15:04.561214",
   "environment_variables": {},
   "exception": null,
   "input_path": "1_run_qc_consensus.ipynb",
   "output_path": "1_run_qc_consensus.OUT.ipynb",
   "parameters": {},
   "start_time": "2022-09-16T16:02:05.185501",
   "version": "2.3.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
