{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "871a4820",
   "metadata": {
    "papermill": {
     "duration": 0.007662,
     "end_time": "2022-09-30T13:26:35.233722",
     "exception": false,
     "start_time": "2022-09-30T13:26:35.226060",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%% md\n"
    },
    "tags": []
   },
   "source": [
    "# pycisTopic analysis\n",
    "\n",
    "Full dataset, using consensus peak regions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "00cf2587",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-21T16:01:52.296245Z",
     "iopub.status.busy": "2022-11-21T16:01:52.295515Z",
     "iopub.status.idle": "2022-11-21T16:01:53.402712Z",
     "shell.execute_reply": "2022-11-21T16:01:53.401770Z",
     "shell.execute_reply.started": "2022-11-21T16:01:52.296188Z"
    },
    "papermill": {
     "duration": 3.077325,
     "end_time": "2022-09-30T13:26:38.318357",
     "exception": false,
     "start_time": "2022-09-30T13:26:35.241032",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 3;\n",
       "                var nbb_unformatted_code = \"import pycisTopic\\n%load_ext nb_black\\nimport warnings\\nwarnings.filterwarnings('ignore')\\nwarnings.simplefilter('ignore')\\nimport pickle\\nimport pandas as pd\\nimport os\";\n",
       "                var nbb_formatted_code = \"import pycisTopic\\n\\n%load_ext nb_black\\nimport warnings\\n\\nwarnings.filterwarnings(\\\"ignore\\\")\\nwarnings.simplefilter(\\\"ignore\\\")\\nimport pickle\\nimport pandas as pd\\nimport os\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pycisTopic\n",
    "%load_ext nb_black\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "warnings.simplefilter('ignore')\n",
    "import pickle\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cf20faa7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T13:26:38.334071Z",
     "iopub.status.busy": "2022-09-30T13:26:38.333764Z",
     "iopub.status.idle": "2022-09-30T13:26:38.562608Z",
     "shell.execute_reply": "2022-09-30T13:26:38.562016Z"
    },
    "papermill": {
     "duration": 0.236999,
     "end_time": "2022-09-30T13:26:38.563611",
     "exception": false,
     "start_time": "2022-09-30T13:26:38.326612",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus\n"
     ]
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 2;\n",
       "                var nbb_unformatted_code = \"!pwd\";\n",
       "                var nbb_formatted_code = \"!pwd\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "!pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "baa6cf5a",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2022-09-30T13:26:38.578365Z",
     "iopub.status.busy": "2022-09-30T13:26:38.578106Z",
     "iopub.status.idle": "2022-09-30T13:26:38.583416Z",
     "shell.execute_reply": "2022-09-30T13:26:38.582926Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "papermill": {
     "duration": 0.013477,
     "end_time": "2022-09-30T13:26:38.584310",
     "exception": false,
     "start_time": "2022-09-30T13:26:38.570833",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 3;\n",
       "                var nbb_unformatted_code = \"wdir = '/dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus'\\nos.chdir( wdir )\";\n",
       "                var nbb_formatted_code = \"wdir = \\\"/dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus\\\"\\nos.chdir(wdir)\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "wdir = '/dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/fixedcells_3_cistopic_consensus'\n",
    "os.chdir( wdir )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "42185775",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T13:26:38.599099Z",
     "iopub.status.busy": "2022-09-30T13:26:38.598671Z",
     "iopub.status.idle": "2022-09-30T13:26:38.633542Z",
     "shell.execute_reply": "2022-09-30T13:26:38.633061Z"
    },
    "papermill": {
     "duration": 0.043287,
     "end_time": "2022-09-30T13:26:38.634575",
     "exception": false,
     "start_time": "2022-09-30T13:26:38.591288",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 4;\n",
       "                var nbb_unformatted_code = \"# create output directory:\\nf_final_dir = os.path.join(wdir, 'downstream_analysis')\\nif not os.path.exists(f_final_dir):\\n    os.makedirs(f_final_dir)\";\n",
       "                var nbb_formatted_code = \"# create output directory:\\nf_final_dir = os.path.join(wdir, \\\"downstream_analysis\\\")\\nif not os.path.exists(f_final_dir):\\n    os.makedirs(f_final_dir)\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# create output directory:\n",
    "f_final_dir = os.path.join(wdir, 'downstream_analysis')\n",
    "if not os.path.exists(f_final_dir):\n",
    "    os.makedirs(f_final_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1bedc812",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T13:26:38.649290Z",
     "iopub.status.busy": "2022-09-30T13:26:38.648893Z",
     "iopub.status.idle": "2022-09-30T13:26:38.652881Z",
     "shell.execute_reply": "2022-09-30T13:26:38.652374Z"
    },
    "papermill": {
     "duration": 0.012125,
     "end_time": "2022-09-30T13:26:38.653873",
     "exception": false,
     "start_time": "2022-09-30T13:26:38.641748",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 5;\n",
       "                var nbb_unformatted_code = \"import glob\";\n",
       "                var nbb_formatted_code = \"import glob\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2c0805b1",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T13:26:38.667870Z",
     "iopub.status.busy": "2022-09-30T13:26:38.667447Z",
     "iopub.status.idle": "2022-09-30T13:26:38.710509Z",
     "shell.execute_reply": "2022-09-30T13:26:38.710028Z"
    },
    "papermill": {
     "duration": 0.050981,
     "end_time": "2022-09-30T13:26:38.711436",
     "exception": false,
     "start_time": "2022-09-30T13:26:38.660455",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "      <th>Technology</th>\n",
       "      <th>Sample</th>\n",
       "      <th>Replicate</th>\n",
       "      <th>st</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>key</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Broad_1</th>\n",
       "      <td>Broad_1</td>\n",
       "      <td>BioRad</td>\n",
       "      <td>Harvard_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Harvard (BioRad)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Broad_2</th>\n",
       "      <td>Broad_2</td>\n",
       "      <td>BioRad</td>\n",
       "      <td>Harvard_2</td>\n",
       "      <td>2</td>\n",
       "      <td>Harvard (BioRad)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stanford_1</th>\n",
       "      <td>Stanford_1</td>\n",
       "      <td>10x v1.0</td>\n",
       "      <td>Stanford_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Stanford (10x v1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Stanford_2</th>\n",
       "      <td>Stanford_2</td>\n",
       "      <td>10x v1.0</td>\n",
       "      <td>Stanford_2</td>\n",
       "      <td>2</td>\n",
       "      <td>Stanford (10x v1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VIB_1</th>\n",
       "      <td>VIB_1</td>\n",
       "      <td>10x v1.0</td>\n",
       "      <td>VIB_1</td>\n",
       "      <td>1</td>\n",
       "      <td>VIB (10x v1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VIB_2</th>\n",
       "      <td>VIB_2</td>\n",
       "      <td>10x v1.0</td>\n",
       "      <td>VIB_2</td>\n",
       "      <td>2</td>\n",
       "      <td>VIB (10x v1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CNAG_1</th>\n",
       "      <td>CNAG_1</td>\n",
       "      <td>10x v1.1</td>\n",
       "      <td>CNAG_1</td>\n",
       "      <td>1</td>\n",
       "      <td>CNAG (10x v1.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CNAG_2</th>\n",
       "      <td>CNAG_2</td>\n",
       "      <td>10x v1.1</td>\n",
       "      <td>CNAG_2</td>\n",
       "      <td>2</td>\n",
       "      <td>CNAG (10x v1.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Broad_mito_1</th>\n",
       "      <td>Broad_mito_1</td>\n",
       "      <td>mito-scATAC-seq</td>\n",
       "      <td>Broad_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Broad (mito-scATAC-seq)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Broad_mito_2</th>\n",
       "      <td>Broad_mito_2</td>\n",
       "      <td>mito-scATAC-seq</td>\n",
       "      <td>Broad_2</td>\n",
       "      <td>2</td>\n",
       "      <td>Broad (mito-scATAC-seq)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sanger_1</th>\n",
       "      <td>Sanger_1</td>\n",
       "      <td>10x multiome</td>\n",
       "      <td>Sanger_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Sanger (10x multiome)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Sanger_2</th>\n",
       "      <td>Sanger_2</td>\n",
       "      <td>10x multiome</td>\n",
       "      <td>Sanger_2</td>\n",
       "      <td>2</td>\n",
       "      <td>Sanger (10x multiome)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s3atac</th>\n",
       "      <td>s3atac</td>\n",
       "      <td>s3-ATAC</td>\n",
       "      <td>OHSU</td>\n",
       "      <td>1</td>\n",
       "      <td>OHSU (s3-ATAC)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VIB_Hydrop_1</th>\n",
       "      <td>VIB_Hydrop_1</td>\n",
       "      <td>HyDrop</td>\n",
       "      <td>Hydrop_1</td>\n",
       "      <td>1</td>\n",
       "      <td>Hydrop (HyDrop)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>VIB_Hydrop_2</th>\n",
       "      <td>VIB_Hydrop_2</td>\n",
       "      <td>HyDrop</td>\n",
       "      <td>Hydrop_2</td>\n",
       "      <td>2</td>\n",
       "      <td>Hydrop (HyDrop)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       key       Technology      Sample Replicate  \\\n",
       "key                                                                 \n",
       "Broad_1            Broad_1           BioRad   Harvard_1         1   \n",
       "Broad_2            Broad_2           BioRad   Harvard_2         2   \n",
       "Stanford_1      Stanford_1         10x v1.0  Stanford_1         1   \n",
       "Stanford_2      Stanford_2         10x v1.0  Stanford_2         2   \n",
       "VIB_1                VIB_1         10x v1.0       VIB_1         1   \n",
       "VIB_2                VIB_2         10x v1.0       VIB_2         2   \n",
       "CNAG_1              CNAG_1         10x v1.1      CNAG_1         1   \n",
       "CNAG_2              CNAG_2         10x v1.1      CNAG_2         2   \n",
       "Broad_mito_1  Broad_mito_1  mito-scATAC-seq     Broad_1         1   \n",
       "Broad_mito_2  Broad_mito_2  mito-scATAC-seq     Broad_2         2   \n",
       "Sanger_1          Sanger_1     10x multiome    Sanger_1         1   \n",
       "Sanger_2          Sanger_2     10x multiome    Sanger_2         2   \n",
       "s3atac              s3atac          s3-ATAC        OHSU         1   \n",
       "VIB_Hydrop_1  VIB_Hydrop_1           HyDrop    Hydrop_1         1   \n",
       "VIB_Hydrop_2  VIB_Hydrop_2           HyDrop    Hydrop_2         2   \n",
       "\n",
       "                                   st  \n",
       "key                                    \n",
       "Broad_1              Harvard (BioRad)  \n",
       "Broad_2              Harvard (BioRad)  \n",
       "Stanford_1        Stanford (10x v1.0)  \n",
       "Stanford_2        Stanford (10x v1.0)  \n",
       "VIB_1                  VIB (10x v1.0)  \n",
       "VIB_2                  VIB (10x v1.0)  \n",
       "CNAG_1                CNAG (10x v1.1)  \n",
       "CNAG_2                CNAG (10x v1.1)  \n",
       "Broad_mito_1  Broad (mito-scATAC-seq)  \n",
       "Broad_mito_2  Broad (mito-scATAC-seq)  \n",
       "Sanger_1        Sanger (10x multiome)  \n",
       "Sanger_2        Sanger (10x multiome)  \n",
       "s3atac                 OHSU (s3-ATAC)  \n",
       "VIB_Hydrop_1          Hydrop (HyDrop)  \n",
       "VIB_Hydrop_2          Hydrop (HyDrop)  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 6;\n",
       "                var nbb_unformatted_code = \"import re\\n\\nsample_annot = pd.DataFrame(\\n    [\\n        ['Broad_1','BioRad'],\\n        ['Broad_2','BioRad'],\\n        ['Stanford_1','10x v1.0'],\\n        ['Stanford_2','10x v1.0'],\\n        ['VIB_1','10x v1.0'],\\n        ['VIB_2','10x v1.0'],\\n        ['CNAG_1','10x v1.1'],\\n        ['CNAG_2','10x v1.1'],\\n        ['Broad_mito_1','mito-scATAC-seq'],\\n        ['Broad_mito_2','mito-scATAC-seq'],\\n        ['Sanger_1','10x multiome'],\\n        ['Sanger_2','10x multiome'],\\n        ['s3atac','s3-ATAC'],\\n        ['VIB_Hydrop_1','HyDrop'],\\n        ['VIB_Hydrop_2','HyDrop'],\\n        \\n    ],\\n    columns=['key','Technology']\\n)\\nsample_annot['Sample'] = sample_annot['key']\\nsample_annot.set_index('key', drop=False, inplace=True)\\n# rename samples:\\nsample_annot.loc['Broad_1','Sample'] = 'Harvard_1'\\nsample_annot.loc['Broad_2','Sample'] = 'Harvard_2'\\nsample_annot.loc['Broad_mito_1','Sample'] = 'Broad_1'\\nsample_annot.loc['Broad_mito_2','Sample'] = 'Broad_2'\\nsample_annot.loc['s3atac','Sample'] = 'OHSU'\\nsample_annot.loc['VIB_Hydrop_1','Sample'] = 'Hydrop_1'\\nsample_annot.loc['VIB_Hydrop_2','Sample'] = 'Hydrop_2'\\n\\nsample_annot['Replicate'] = [ x.Sample.split('_')[-1] for i,x in sample_annot.iterrows() ]\\nsample_annot['st'] = [ re.sub('_[0-9]','',x.Sample)+' ('+x.Technology+')' for i,x in sample_annot.iterrows() ]\\n\\nsample_annot.loc[sample_annot['key']=='s3atac','Replicate'] = 1\\nsample_map = { x: y.Sample for x,y in sample_annot.iterrows() }\\nsample_annot\";\n",
       "                var nbb_formatted_code = \"import re\\n\\nsample_annot = pd.DataFrame(\\n    [\\n        [\\\"Broad_1\\\", \\\"BioRad\\\"],\\n        [\\\"Broad_2\\\", \\\"BioRad\\\"],\\n        [\\\"Stanford_1\\\", \\\"10x v1.0\\\"],\\n        [\\\"Stanford_2\\\", \\\"10x v1.0\\\"],\\n        [\\\"VIB_1\\\", \\\"10x v1.0\\\"],\\n        [\\\"VIB_2\\\", \\\"10x v1.0\\\"],\\n        [\\\"CNAG_1\\\", \\\"10x v1.1\\\"],\\n        [\\\"CNAG_2\\\", \\\"10x v1.1\\\"],\\n        [\\\"Broad_mito_1\\\", \\\"mito-scATAC-seq\\\"],\\n        [\\\"Broad_mito_2\\\", \\\"mito-scATAC-seq\\\"],\\n        [\\\"Sanger_1\\\", \\\"10x multiome\\\"],\\n        [\\\"Sanger_2\\\", \\\"10x multiome\\\"],\\n        [\\\"s3atac\\\", \\\"s3-ATAC\\\"],\\n        [\\\"VIB_Hydrop_1\\\", \\\"HyDrop\\\"],\\n        [\\\"VIB_Hydrop_2\\\", \\\"HyDrop\\\"],\\n    ],\\n    columns=[\\\"key\\\", \\\"Technology\\\"],\\n)\\nsample_annot[\\\"Sample\\\"] = sample_annot[\\\"key\\\"]\\nsample_annot.set_index(\\\"key\\\", drop=False, inplace=True)\\n# rename samples:\\nsample_annot.loc[\\\"Broad_1\\\", \\\"Sample\\\"] = \\\"Harvard_1\\\"\\nsample_annot.loc[\\\"Broad_2\\\", \\\"Sample\\\"] = \\\"Harvard_2\\\"\\nsample_annot.loc[\\\"Broad_mito_1\\\", \\\"Sample\\\"] = \\\"Broad_1\\\"\\nsample_annot.loc[\\\"Broad_mito_2\\\", \\\"Sample\\\"] = \\\"Broad_2\\\"\\nsample_annot.loc[\\\"s3atac\\\", \\\"Sample\\\"] = \\\"OHSU\\\"\\nsample_annot.loc[\\\"VIB_Hydrop_1\\\", \\\"Sample\\\"] = \\\"Hydrop_1\\\"\\nsample_annot.loc[\\\"VIB_Hydrop_2\\\", \\\"Sample\\\"] = \\\"Hydrop_2\\\"\\n\\nsample_annot[\\\"Replicate\\\"] = [\\n    x.Sample.split(\\\"_\\\")[-1] for i, x in sample_annot.iterrows()\\n]\\nsample_annot[\\\"st\\\"] = [\\n    re.sub(\\\"_[0-9]\\\", \\\"\\\", x.Sample) + \\\" (\\\" + x.Technology + \\\")\\\"\\n    for i, x in sample_annot.iterrows()\\n]\\n\\nsample_annot.loc[sample_annot[\\\"key\\\"] == \\\"s3atac\\\", \\\"Replicate\\\"] = 1\\nsample_map = {x: y.Sample for x, y in sample_annot.iterrows()}\\nsample_annot\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "sample_annot = pd.DataFrame(\n",
    "    [\n",
    "        ['Broad_1','BioRad'],\n",
    "        ['Broad_2','BioRad'],\n",
    "        ['Stanford_1','10x v1.0'],\n",
    "        ['Stanford_2','10x v1.0'],\n",
    "        ['VIB_1','10x v1.0'],\n",
    "        ['VIB_2','10x v1.0'],\n",
    "        ['CNAG_1','10x v1.1'],\n",
    "        ['CNAG_2','10x v1.1'],\n",
    "        ['Broad_mito_1','mito-scATAC-seq'],\n",
    "        ['Broad_mito_2','mito-scATAC-seq'],\n",
    "        ['Sanger_1','10x multiome'],\n",
    "        ['Sanger_2','10x multiome'],\n",
    "        ['s3atac','s3-ATAC'],\n",
    "        ['VIB_Hydrop_1','HyDrop'],\n",
    "        ['VIB_Hydrop_2','HyDrop'],\n",
    "        \n",
    "    ],\n",
    "    columns=['key','Technology']\n",
    ")\n",
    "sample_annot['Sample'] = sample_annot['key']\n",
    "sample_annot.set_index('key', drop=False, inplace=True)\n",
    "# rename samples:\n",
    "sample_annot.loc['Broad_1','Sample'] = 'Harvard_1'\n",
    "sample_annot.loc['Broad_2','Sample'] = 'Harvard_2'\n",
    "sample_annot.loc['Broad_mito_1','Sample'] = 'Broad_1'\n",
    "sample_annot.loc['Broad_mito_2','Sample'] = 'Broad_2'\n",
    "sample_annot.loc['s3atac','Sample'] = 'OHSU'\n",
    "sample_annot.loc['VIB_Hydrop_1','Sample'] = 'Hydrop_1'\n",
    "sample_annot.loc['VIB_Hydrop_2','Sample'] = 'Hydrop_2'\n",
    "\n",
    "sample_annot['Replicate'] = [ x.Sample.split('_')[-1] for i,x in sample_annot.iterrows() ]\n",
    "sample_annot['st'] = [ re.sub('_[0-9]','',x.Sample)+' ('+x.Technology+')' for i,x in sample_annot.iterrows() ]\n",
    "\n",
    "sample_annot.loc[sample_annot['key']=='s3atac','Replicate'] = 1\n",
    "sample_map = { x: y.Sample for x,y in sample_annot.iterrows() }\n",
    "sample_annot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b7ebfef",
   "metadata": {
    "papermill": {
     "duration": 0.006799,
     "end_time": "2022-09-30T13:26:38.725334",
     "exception": false,
     "start_time": "2022-09-30T13:26:38.718535",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%% md\n"
    },
    "tags": []
   },
   "source": [
    "## Save/load cisTopic objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ba4cd8c8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T13:26:38.739726Z",
     "iopub.status.busy": "2022-09-30T13:26:38.739357Z",
     "iopub.status.idle": "2022-09-30T13:26:38.815812Z",
     "shell.execute_reply": "2022-09-30T13:26:38.815336Z"
    },
    "papermill": {
     "duration": 0.084689,
     "end_time": "2022-09-30T13:26:38.816776",
     "exception": false,
     "start_time": "2022-09-30T13:26:38.732087",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': 'cistopic_objects/BIO_ddseq_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': 'cistopic_objects/BIO_ddseq_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_14topics.dimreduc.consensus.pkl',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': 'cistopic_objects/BIO_ddseq_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': 'cistopic_objects/BIO_ddseq_4.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': 'cistopic_objects/BRO_mtscatac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': 'cistopic_objects/BRO_mtscatac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_17topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': 'cistopic_objects/CNA_10xmultiome_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': 'cistopic_objects/CNA_10xmultiome_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_8topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': 'cistopic_objects/CNA_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': 'cistopic_objects/CNA_10xv11_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': 'cistopic_objects/CNA_10xv11_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': 'cistopic_objects/CNA_10xv11_4.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': 'cistopic_objects/CNA_10xv11_5.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': 'cistopic_objects/CNA_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': 'cistopic_objects/CNA_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': 'cistopic_objects/CNA_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': 'cistopic_objects/CNA_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': 'cistopic_objects/CNA_hydrop_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': 'cistopic_objects/CNA_mtscatac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': 'cistopic_objects/CNA_mtscatac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': 'cistopic_objects/EPF_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': 'cistopic_objects/EPF_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': 'cistopic_objects/EPF_hydrop_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_5topics.dimreduc.consensus.pkl',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': 'cistopic_objects/EPF_hydrop_4.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': 'cistopic_objects/HAR_ddseq_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': 'cistopic_objects/HAR_ddseq_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': 'cistopic_objects/MDC_mtscatac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_19topics.dimreduc.consensus.pkl',\n",
       " 'MDC_mtscatac_2.FIXEDCELLS': 'cistopic_objects/MDC_mtscatac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'OHS_s3atac_1.FIXEDCELLS': 'cistopic_objects/OHS_s3atac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'OHS_s3atac_2.FIXEDCELLS': 'cistopic_objects/OHS_s3atac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_7topics.dimreduc.consensus.pkl',\n",
       " 'SAN_10xmultiome_1.FIXEDCELLS': 'cistopic_objects/SAN_10xmultiome_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_7topics.dimreduc.consensus.pkl',\n",
       " 'SAN_10xmultiome_2.FIXEDCELLS': 'cistopic_objects/SAN_10xmultiome_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'STA_10xv11_1.FIXEDCELLS': 'cistopic_objects/STA_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl',\n",
       " 'STA_10xv11_2.FIXEDCELLS': 'cistopic_objects/STA_10xv11_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_7topics.dimreduc.consensus.pkl',\n",
       " 'TXG_10xv11_1.FIXEDCELLS': 'cistopic_objects/TXG_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_14topics.dimreduc.consensus.pkl',\n",
       " 'TXG_10xv2_1.FIXEDCELLS': 'cistopic_objects/TXG_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_21topics.dimreduc.consensus.pkl',\n",
       " 'TXG_10xv2_2.FIXEDCELLS': 'cistopic_objects/TXG_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_21topics.dimreduc.consensus.pkl',\n",
       " 'UCS_ddseq_1.FIXEDCELLS': 'cistopic_objects/UCS_ddseq_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'UCS_ddseq_2.FIXEDCELLS': 'cistopic_objects/UCS_ddseq_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xmultiome_1.FIXEDCELLS': 'cistopic_objects/VIB_10xmultiome_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xmultiome_2.FIXEDCELLS': 'cistopic_objects/VIB_10xmultiome_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xv1_1.FIXEDCELLS': 'cistopic_objects/VIB_10xv1_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xv1_2.FIXEDCELLS': 'cistopic_objects/VIB_10xv1_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_19topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xv2_1.FIXEDCELLS': 'cistopic_objects/VIB_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xv2_2.FIXEDCELLS': 'cistopic_objects/VIB_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'VIB_hydrop_1.FIXEDCELLS': 'cistopic_objects/VIB_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_17topics.dimreduc.consensus.pkl',\n",
       " 'VIB_hydrop_2.FIXEDCELLS': 'cistopic_objects/VIB_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_19topics.dimreduc.consensus.pkl'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 7;\n",
       "                var nbb_unformatted_code = \"cto_consensus_paths = sorted(glob.glob(f'cistopic_objects/*consensus.pkl'))\\ncistopic_obj_path_dict = {x.split('/')[-1].split(f'__')[0]:x for x in cto_consensus_paths}\\ncistopic_obj_path_dict\";\n",
       "                var nbb_formatted_code = \"cto_consensus_paths = sorted(glob.glob(f\\\"cistopic_objects/*consensus.pkl\\\"))\\ncistopic_obj_path_dict = {\\n    x.split(\\\"/\\\")[-1].split(f\\\"__\\\")[0]: x for x in cto_consensus_paths\\n}\\ncistopic_obj_path_dict\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "cto_consensus_paths = sorted(glob.glob(f'cistopic_objects/*consensus.pkl'))\n",
    "cistopic_obj_path_dict = {x.split('/')[-1].split(f'__')[0]:x for x in cto_consensus_paths}\n",
    "cistopic_obj_path_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e156b46",
   "metadata": {
    "papermill": {
     "duration": 0.007729,
     "end_time": "2022-09-30T13:26:38.832586",
     "exception": false,
     "start_time": "2022-09-30T13:26:38.824857",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%% md\n"
    },
    "tags": []
   },
   "source": [
    "## Topic binarization & QC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2e46ba58",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T13:26:38.847705Z",
     "iopub.status.busy": "2022-09-30T13:26:38.847330Z",
     "iopub.status.idle": "2022-09-30T13:26:58.269048Z",
     "shell.execute_reply": "2022-09-30T13:26:58.268469Z"
    },
    "papermill": {
     "duration": 19.457856,
     "end_time": "2022-09-30T13:26:58.297611",
     "exception": false,
     "start_time": "2022-09-30T13:26:38.839755",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 8;\n",
       "                var nbb_unformatted_code = \"from pycisTopic.topic_binarization import binarize_topics\\nfrom pycisTopic.topic_qc import compute_topic_metrics, plot_topic_qc\\nfrom pycisTopic.utils import fig2img\\nimport matplotlib.pyplot as plt\\nfrom pycisTopic.topic_qc import topic_annotation\\nfrom pycisTopic.diff_features import (\\n    impute_accessibility,\\n    normalize_scores,\\n    find_highly_variable_features,\\n    find_diff_features\\n)\\nfrom pycisTopic.clust_vis import plot_imputed_features\\nimport numpy as np\\nimport copy\";\n",
       "                var nbb_formatted_code = \"from pycisTopic.topic_binarization import binarize_topics\\nfrom pycisTopic.topic_qc import compute_topic_metrics, plot_topic_qc\\nfrom pycisTopic.utils import fig2img\\nimport matplotlib.pyplot as plt\\nfrom pycisTopic.topic_qc import topic_annotation\\nfrom pycisTopic.diff_features import (\\n    impute_accessibility,\\n    normalize_scores,\\n    find_highly_variable_features,\\n    find_diff_features,\\n)\\nfrom pycisTopic.clust_vis import plot_imputed_features\\nimport numpy as np\\nimport copy\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from pycisTopic.topic_binarization import binarize_topics\n",
    "from pycisTopic.topic_qc import compute_topic_metrics, plot_topic_qc\n",
    "from pycisTopic.utils import fig2img\n",
    "import matplotlib.pyplot as plt\n",
    "from pycisTopic.topic_qc import topic_annotation\n",
    "from pycisTopic.diff_features import (\n",
    "    impute_accessibility,\n",
    "    normalize_scores,\n",
    "    find_highly_variable_features,\n",
    "    find_diff_features\n",
    ")\n",
    "from pycisTopic.clust_vis import plot_imputed_features\n",
    "import numpy as np\n",
    "import copy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21999a24",
   "metadata": {
    "papermill": {
     "duration": 0.007856,
     "end_time": "2022-09-30T13:26:58.319080",
     "exception": false,
     "start_time": "2022-09-30T13:26:58.311224",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%% md\n"
    },
    "tags": []
   },
   "source": [
    "### Binarize the topic-region distributions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "13bd39fc",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T13:26:58.335762Z",
     "iopub.status.busy": "2022-09-30T13:26:58.335370Z",
     "iopub.status.idle": "2022-09-30T13:26:58.520395Z",
     "shell.execute_reply": "2022-09-30T13:26:58.519809Z"
    },
    "papermill": {
     "duration": 0.195041,
     "end_time": "2022-09-30T13:26:58.522266",
     "exception": false,
     "start_time": "2022-09-30T13:26:58.327225",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downstream_analysis/DARs/BIO_ddseq_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/BIO_ddseq_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/BIO_ddseq_3.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/BIO_ddseq_4.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/BRO_mtscatac_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/BRO_mtscatac_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/CNA_10xmultiome_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/CNA_10xmultiome_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/CNA_10xv11_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/CNA_10xv11_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/CNA_10xv11_3.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/CNA_10xv11_4.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/CNA_10xv11_5.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/CNA_10xv2_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/CNA_10xv2_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/CNA_hydrop_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/CNA_hydrop_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/CNA_hydrop_3.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/CNA_mtscatac_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/CNA_mtscatac_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/EPF_hydrop_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/EPF_hydrop_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/EPF_hydrop_3.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/EPF_hydrop_4.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/HAR_ddseq_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/HAR_ddseq_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/MDC_mtscatac_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/MDC_mtscatac_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/OHS_s3atac_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/OHS_s3atac_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/SAN_10xmultiome_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/SAN_10xmultiome_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/STA_10xv11_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/STA_10xv11_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/TXG_10xv11_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/TXG_10xv2_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/TXG_10xv2_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/UCS_ddseq_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/UCS_ddseq_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/VIB_10xmultiome_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/VIB_10xmultiome_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/VIB_10xv1_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/VIB_10xv1_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/VIB_10xv2_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/VIB_10xv2_2.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/VIB_hydrop_1.FIXEDCELLS already exists!\n",
      "downstream_analysis/DARs/VIB_hydrop_2.FIXEDCELLS already exists!\n"
     ]
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 9;\n",
       "                var nbb_unformatted_code = \"for sample in cistopic_obj_path_dict.keys():\\n# for sample  in ['SAN_10xmultiome_1.FIXEDCELLS']:\\n    path = cistopic_obj_path_dict[sample]\\n    if not os.path.exists(f'downstream_analysis/DARs/{sample}'):\\n        if os.path.isfile(path):\\n            print(f\\\"Loading {path}\\\")\\n            with open(path, 'rb') as f:\\n                cto = pickle.load(f)\\n                \\n            cto.cell_names = cto.cell_data.index\\n            if [x.split('-')[0] for x in cto.selected_model.cell_topic.columns] == [x.split('___')[0] for x in list(cto.cell_names)]:\\n                cto.selected_model.cell_topic.columns =  list(cto.cell_names)\\n                \\n            region_bin_topics = binarize_topics(\\n                cto,\\n                method='otsu',\\n                ntop=3000,\\n                plot=True,\\n                num_columns=6,\\n                save=f'plots_qc/{sample}__topic_region_binarized.png'\\n            )\\n\\n            with open(\\n                    f\\\"downstream_analysis/binarized_topics/{sample}__topic_region_binarized.pkl\\\",'wb'\\n            ) as f:\\n                pickle.dump(region_bin_topics, f, protocol=4)\\n\\n            binarized_cell_topics = binarize_topics(\\n                cto,\\n                target='cell',\\n                method='li',\\n                plot=True,\\n                num_columns=5,\\n                nbins=100,\\n                save=f'plots_qc/{sample}__cells_topic_binarized.png'\\n            )\\n\\n            with open(\\n                    f\\\"downstream_analysis/binarized_topics/{sample}__cells_topic_binarized.pkl\\\",'wb'\\n            ) as f:\\n                pickle.dump(region_bin_topics, f, protocol=4)\\n\\n            topic_qc_metrics = compute_topic_metrics(cto)\\n\\n            with open(\\n                    f\\\"downstream_analysis/binarized_topics/{sample}__topic_qc.pkl\\\",'wb'\\n            ) as f:\\n                pickle.dump(topic_qc_metrics, f, protocol=4)\\n\\n            fig_dict= {\\n                'CoherenceVSAssignments': plot_topic_qc(topic_qc_metrics, var_x='Coherence', var_y='Log10_Assignments',\\n                                                        var_color='Gini_index', plot=False, return_fig=True),\\n                'AssignmentsVSCells_in_bin': plot_topic_qc(topic_qc_metrics, var_x='Log10_Assignments',\\n                                                           var_y='Cells_in_binarized_topic', var_color='Gini_index',\\n                                                           plot=False, return_fig=True),\\n                'CoherenceVSCells_in_bin': plot_topic_qc(topic_qc_metrics, var_x='Coherence',\\n                                                         var_y='Cells_in_binarized_topic', var_color='Gini_index',\\n                                                         plot=False, return_fig=True),\\n                'CoherenceVSRegions_in_bin': plot_topic_qc(topic_qc_metrics, var_x='Coherence',\\n                                                           var_y='Regions_in_binarized_topic', var_color='Gini_index',\\n                                                           plot=False, return_fig=True),\\n                'CoherenceVSMarginal_dist': plot_topic_qc(topic_qc_metrics, var_x='Coherence',\\n                                                          var_y='Marginal_topic_dist', var_color='Gini_index',\\n                                                          plot=False, return_fig=True),\\n                'CoherenceVSGini_index': plot_topic_qc(topic_qc_metrics, var_x='Coherence', var_y='Gini_index',\\n                                                       var_color='Gini_index', plot=False, return_fig=True)}\\n\\n            fig=plt.figure(figsize=(40, 43))\\n            i = 1\\n            for fig_ in fig_dict.keys():\\n                plt.subplot(2, 3, i)\\n                img = fig2img(fig_dict[fig_]) #To convert figures to png to plot together, see .utils.py. This converts the figure to png.\\n                plt.imshow(img)\\n                plt.axis('off')\\n                i += 1\\n            plt.subplots_adjust(wspace=0, hspace=-0.70)\\n            #plt.tight_layout()\\n            plt.show()\\n            plt.savefig(f'plots_qc/{sample}__topic_qc_metrics.png', facecolor='white', dpi=150)\\n\\n            topic_annot = topic_annotation(\\n                cto,\\n                annot_var='consensus_cell_type',\\n                binarized_cell_topic=binarized_cell_topics,\\n                general_topic_thr = 0.2\\n            )\\n            topic_qc_metrics = pd.concat(\\n                [\\n                    topic_annot[['consensus_cell_type', 'Ratio_cells_in_topic', 'Ratio_group_in_population']],\\n                    topic_qc_metrics\\n                ],\\n                axis=1\\n            )\\n\\n            imputed_acc_obj = impute_accessibility(\\n                cto,\\n                selected_cells=None,\\n                selected_regions=None,\\n                scale_factor=10**6\\n            )\\n\\n            normalized_imputed_acc_obj = normalize_scores(\\n                imputed_acc_obj,\\n                scale_factor=10**4\\n            )\\n\\n            with open(\\n                f\\\"downstream_analysis/imputed_acc_objs/{sample}__imputed_acc_obs.pkl\\\",'wb'\\n            ) as f:\\n                pickle.dump(normalized_imputed_acc_obj, f, protocol=4)\\n\\n            variable_regions = find_highly_variable_features(\\n                normalized_imputed_acc_obj,\\n                min_disp = 0.05,\\n                min_mean = 0.0125,\\n                max_mean = 3,\\n                max_disp = np.inf,\\n                n_bins=20,\\n                n_top_features=None,\\n                plot=True,\\n                save=f'plots_qc/{sample}__HVR.png'\\n            )\\n\\n            print(f\\\"Found {len(variable_regions)} variable regions\\\")\\n            with open(f'downstream_analysis/HVRs/{sample}__HVRs.bed', 'w') as f:\\n                for line in [x.replace(':','\\\\t').replace('-','\\\\t') for x in variable_regions]:\\n                    f.write(f\\\"{line}\\\\n\\\")\\n                    \\n            with open(\\n                f'downstream_analysis/HVRs/{sample}__HVRs.pkl', 'w'\\n            ) as f:\\n                pickle.dump(variable_regions, f, protocol=4)\\n\\n            os.mkdir(f'downstream_analysis/DARs/{sample}')\\n\\n            markers_dict = find_diff_features(\\n                cto,\\n                normalized_imputed_acc_obj,\\n                variable='consensus_cell_type',\\n                var_features=variable_regions,\\n                contrasts=None,\\n                adjpval_thr=0.05,\\n                log2fc_thr=np.log2(1.5),\\n                n_cpu=5,\\n            )\\n            with open(\\n                    f\\\"downstream_analysis/DARs/{sample}/{sample}__DARs_dict.pkl\\\",'wb'\\n            ) as f:\\n                pickle.dump(markers_dict, f, protocol=4)\\n\\n            for cell_type in markers_dict.keys():\\n                markers = markers_dict[cell_type].index.tolist()\\n                df = pd.DataFrame(markers)\\n                if len(df) == 0:\\n                    print(f\\\"no DARs found for {cell_type} in {sample}\\\")\\n                else:\\n                    df[[0,1]] = df[0].str.split(':',expand=True)\\n                    df[[1,2]] = df[1].str.split('-',expand=True)\\n                    df[3] = markers_dict[cell_type].index.tolist()\\n                    df[3] = cell_type.replace(' ', '_')# + '_' + df[3].astype(str)\\n                    score = markers_dict[cell_type]['Log2FC']\\n                    score = score.reset_index(drop=True)\\n                    df[4] = score\\n                    df[5] = '.'\\n                    pval = markers_dict[cell_type]['Adjusted_pval']\\n                    pval = pval.reset_index(drop=True)\\n                    df[6] = pval\\n                    name = cell_type.replace(' ', '_')\\n                    # name = name.replace('/', '-')\\n                    if not os.path.exists(f'downstream_analysis/DARs/{sample}'):\\n                        os.mkdir(f'downstream_analysis/DARs/{sample}')\\n\\n                    df.to_csv(f'downstream_analysis/DARs/{sample}/{sample}__{cell_type.replace(\\\" \\\", \\\"_\\\")}__DARs.bed', sep='\\\\t', header=False, index=False)\\n                    df[0:2000].to_csv(f'downstream_analysis/DARs/{sample}/{sample}__{cell_type.replace(\\\" \\\", \\\"_\\\")}__DARs.TOP2k.bed', sep='\\\\t', header=False, index=False)\\n\\n        else:\\n            print(f'{path} does not exist!')\\n    else:\\n        print(f\\\"downstream_analysis/DARs/{sample} already exists!\\\")\";\n",
       "                var nbb_formatted_code = \"for sample in cistopic_obj_path_dict.keys():\\n    # for sample  in ['SAN_10xmultiome_1.FIXEDCELLS']:\\n    path = cistopic_obj_path_dict[sample]\\n    if not os.path.exists(f\\\"downstream_analysis/DARs/{sample}\\\"):\\n        if os.path.isfile(path):\\n            print(f\\\"Loading {path}\\\")\\n            with open(path, \\\"rb\\\") as f:\\n                cto = pickle.load(f)\\n\\n            cto.cell_names = cto.cell_data.index\\n            if [x.split(\\\"-\\\")[0] for x in cto.selected_model.cell_topic.columns] == [\\n                x.split(\\\"___\\\")[0] for x in list(cto.cell_names)\\n            ]:\\n                cto.selected_model.cell_topic.columns = list(cto.cell_names)\\n\\n            region_bin_topics = binarize_topics(\\n                cto,\\n                method=\\\"otsu\\\",\\n                ntop=3000,\\n                plot=True,\\n                num_columns=6,\\n                save=f\\\"plots_qc/{sample}__topic_region_binarized.png\\\",\\n            )\\n\\n            with open(\\n                f\\\"downstream_analysis/binarized_topics/{sample}__topic_region_binarized.pkl\\\",\\n                \\\"wb\\\",\\n            ) as f:\\n                pickle.dump(region_bin_topics, f, protocol=4)\\n\\n            binarized_cell_topics = binarize_topics(\\n                cto,\\n                target=\\\"cell\\\",\\n                method=\\\"li\\\",\\n                plot=True,\\n                num_columns=5,\\n                nbins=100,\\n                save=f\\\"plots_qc/{sample}__cells_topic_binarized.png\\\",\\n            )\\n\\n            with open(\\n                f\\\"downstream_analysis/binarized_topics/{sample}__cells_topic_binarized.pkl\\\",\\n                \\\"wb\\\",\\n            ) as f:\\n                pickle.dump(region_bin_topics, f, protocol=4)\\n\\n            topic_qc_metrics = compute_topic_metrics(cto)\\n\\n            with open(\\n                f\\\"downstream_analysis/binarized_topics/{sample}__topic_qc.pkl\\\", \\\"wb\\\"\\n            ) as f:\\n                pickle.dump(topic_qc_metrics, f, protocol=4)\\n\\n            fig_dict = {\\n                \\\"CoherenceVSAssignments\\\": plot_topic_qc(\\n                    topic_qc_metrics,\\n                    var_x=\\\"Coherence\\\",\\n                    var_y=\\\"Log10_Assignments\\\",\\n                    var_color=\\\"Gini_index\\\",\\n                    plot=False,\\n                    return_fig=True,\\n                ),\\n                \\\"AssignmentsVSCells_in_bin\\\": plot_topic_qc(\\n                    topic_qc_metrics,\\n                    var_x=\\\"Log10_Assignments\\\",\\n                    var_y=\\\"Cells_in_binarized_topic\\\",\\n                    var_color=\\\"Gini_index\\\",\\n                    plot=False,\\n                    return_fig=True,\\n                ),\\n                \\\"CoherenceVSCells_in_bin\\\": plot_topic_qc(\\n                    topic_qc_metrics,\\n                    var_x=\\\"Coherence\\\",\\n                    var_y=\\\"Cells_in_binarized_topic\\\",\\n                    var_color=\\\"Gini_index\\\",\\n                    plot=False,\\n                    return_fig=True,\\n                ),\\n                \\\"CoherenceVSRegions_in_bin\\\": plot_topic_qc(\\n                    topic_qc_metrics,\\n                    var_x=\\\"Coherence\\\",\\n                    var_y=\\\"Regions_in_binarized_topic\\\",\\n                    var_color=\\\"Gini_index\\\",\\n                    plot=False,\\n                    return_fig=True,\\n                ),\\n                \\\"CoherenceVSMarginal_dist\\\": plot_topic_qc(\\n                    topic_qc_metrics,\\n                    var_x=\\\"Coherence\\\",\\n                    var_y=\\\"Marginal_topic_dist\\\",\\n                    var_color=\\\"Gini_index\\\",\\n                    plot=False,\\n                    return_fig=True,\\n                ),\\n                \\\"CoherenceVSGini_index\\\": plot_topic_qc(\\n                    topic_qc_metrics,\\n                    var_x=\\\"Coherence\\\",\\n                    var_y=\\\"Gini_index\\\",\\n                    var_color=\\\"Gini_index\\\",\\n                    plot=False,\\n                    return_fig=True,\\n                ),\\n            }\\n\\n            fig = plt.figure(figsize=(40, 43))\\n            i = 1\\n            for fig_ in fig_dict.keys():\\n                plt.subplot(2, 3, i)\\n                img = fig2img(\\n                    fig_dict[fig_]\\n                )  # To convert figures to png to plot together, see .utils.py. This converts the figure to png.\\n                plt.imshow(img)\\n                plt.axis(\\\"off\\\")\\n                i += 1\\n            plt.subplots_adjust(wspace=0, hspace=-0.70)\\n            # plt.tight_layout()\\n            plt.show()\\n            plt.savefig(\\n                f\\\"plots_qc/{sample}__topic_qc_metrics.png\\\", facecolor=\\\"white\\\", dpi=150\\n            )\\n\\n            topic_annot = topic_annotation(\\n                cto,\\n                annot_var=\\\"consensus_cell_type\\\",\\n                binarized_cell_topic=binarized_cell_topics,\\n                general_topic_thr=0.2,\\n            )\\n            topic_qc_metrics = pd.concat(\\n                [\\n                    topic_annot[\\n                        [\\n                            \\\"consensus_cell_type\\\",\\n                            \\\"Ratio_cells_in_topic\\\",\\n                            \\\"Ratio_group_in_population\\\",\\n                        ]\\n                    ],\\n                    topic_qc_metrics,\\n                ],\\n                axis=1,\\n            )\\n\\n            imputed_acc_obj = impute_accessibility(\\n                cto, selected_cells=None, selected_regions=None, scale_factor=10**6\\n            )\\n\\n            normalized_imputed_acc_obj = normalize_scores(\\n                imputed_acc_obj, scale_factor=10**4\\n            )\\n\\n            with open(\\n                f\\\"downstream_analysis/imputed_acc_objs/{sample}__imputed_acc_obs.pkl\\\",\\n                \\\"wb\\\",\\n            ) as f:\\n                pickle.dump(normalized_imputed_acc_obj, f, protocol=4)\\n\\n            variable_regions = find_highly_variable_features(\\n                normalized_imputed_acc_obj,\\n                min_disp=0.05,\\n                min_mean=0.0125,\\n                max_mean=3,\\n                max_disp=np.inf,\\n                n_bins=20,\\n                n_top_features=None,\\n                plot=True,\\n                save=f\\\"plots_qc/{sample}__HVR.png\\\",\\n            )\\n\\n            print(f\\\"Found {len(variable_regions)} variable regions\\\")\\n            with open(f\\\"downstream_analysis/HVRs/{sample}__HVRs.bed\\\", \\\"w\\\") as f:\\n                for line in [\\n                    x.replace(\\\":\\\", \\\"\\\\t\\\").replace(\\\"-\\\", \\\"\\\\t\\\") for x in variable_regions\\n                ]:\\n                    f.write(f\\\"{line}\\\\n\\\")\\n\\n            with open(f\\\"downstream_analysis/HVRs/{sample}__HVRs.pkl\\\", \\\"w\\\") as f:\\n                pickle.dump(variable_regions, f, protocol=4)\\n\\n            os.mkdir(f\\\"downstream_analysis/DARs/{sample}\\\")\\n\\n            markers_dict = find_diff_features(\\n                cto,\\n                normalized_imputed_acc_obj,\\n                variable=\\\"consensus_cell_type\\\",\\n                var_features=variable_regions,\\n                contrasts=None,\\n                adjpval_thr=0.05,\\n                log2fc_thr=np.log2(1.5),\\n                n_cpu=5,\\n            )\\n            with open(\\n                f\\\"downstream_analysis/DARs/{sample}/{sample}__DARs_dict.pkl\\\", \\\"wb\\\"\\n            ) as f:\\n                pickle.dump(markers_dict, f, protocol=4)\\n\\n            for cell_type in markers_dict.keys():\\n                markers = markers_dict[cell_type].index.tolist()\\n                df = pd.DataFrame(markers)\\n                if len(df) == 0:\\n                    print(f\\\"no DARs found for {cell_type} in {sample}\\\")\\n                else:\\n                    df[[0, 1]] = df[0].str.split(\\\":\\\", expand=True)\\n                    df[[1, 2]] = df[1].str.split(\\\"-\\\", expand=True)\\n                    df[3] = markers_dict[cell_type].index.tolist()\\n                    df[3] = cell_type.replace(\\\" \\\", \\\"_\\\")  # + '_' + df[3].astype(str)\\n                    score = markers_dict[cell_type][\\\"Log2FC\\\"]\\n                    score = score.reset_index(drop=True)\\n                    df[4] = score\\n                    df[5] = \\\".\\\"\\n                    pval = markers_dict[cell_type][\\\"Adjusted_pval\\\"]\\n                    pval = pval.reset_index(drop=True)\\n                    df[6] = pval\\n                    name = cell_type.replace(\\\" \\\", \\\"_\\\")\\n                    # name = name.replace('/', '-')\\n                    if not os.path.exists(f\\\"downstream_analysis/DARs/{sample}\\\"):\\n                        os.mkdir(f\\\"downstream_analysis/DARs/{sample}\\\")\\n\\n                    df.to_csv(\\n                        f'downstream_analysis/DARs/{sample}/{sample}__{cell_type.replace(\\\" \\\", \\\"_\\\")}__DARs.bed',\\n                        sep=\\\"\\\\t\\\",\\n                        header=False,\\n                        index=False,\\n                    )\\n                    df[0:2000].to_csv(\\n                        f'downstream_analysis/DARs/{sample}/{sample}__{cell_type.replace(\\\" \\\", \\\"_\\\")}__DARs.TOP2k.bed',\\n                        sep=\\\"\\\\t\\\",\\n                        header=False,\\n                        index=False,\\n                    )\\n\\n        else:\\n            print(f\\\"{path} does not exist!\\\")\\n    else:\\n        print(f\\\"downstream_analysis/DARs/{sample} already exists!\\\")\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for sample in cistopic_obj_path_dict.keys():\n",
    "# for sample  in ['SAN_10xmultiome_1.FIXEDCELLS']:\n",
    "    path = cistopic_obj_path_dict[sample]\n",
    "    if not os.path.exists(f'downstream_analysis/DARs/{sample}'):\n",
    "        if os.path.isfile(path):\n",
    "            print(f\"Loading {path}\")\n",
    "            with open(path, 'rb') as f:\n",
    "                cto = pickle.load(f)\n",
    "                \n",
    "            cto.cell_names = cto.cell_data.index\n",
    "            if [x.split('-')[0] for x in cto.selected_model.cell_topic.columns] == [x.split('___')[0] for x in list(cto.cell_names)]:\n",
    "                cto.selected_model.cell_topic.columns =  list(cto.cell_names)\n",
    "                \n",
    "            region_bin_topics = binarize_topics(\n",
    "                cto,\n",
    "                method='otsu',\n",
    "                ntop=3000,\n",
    "                plot=True,\n",
    "                num_columns=6,\n",
    "                save=f'plots_qc/{sample}__topic_region_binarized.png'\n",
    "            )\n",
    "\n",
    "            with open(\n",
    "                    f\"downstream_analysis/binarized_topics/{sample}__topic_region_binarized.pkl\",'wb'\n",
    "            ) as f:\n",
    "                pickle.dump(region_bin_topics, f, protocol=4)\n",
    "\n",
    "            binarized_cell_topics = binarize_topics(\n",
    "                cto,\n",
    "                target='cell',\n",
    "                method='li',\n",
    "                plot=True,\n",
    "                num_columns=5,\n",
    "                nbins=100,\n",
    "                save=f'plots_qc/{sample}__cells_topic_binarized.png'\n",
    "            )\n",
    "\n",
    "            with open(\n",
    "                    f\"downstream_analysis/binarized_topics/{sample}__cells_topic_binarized.pkl\",'wb'\n",
    "            ) as f:\n",
    "                pickle.dump(region_bin_topics, f, protocol=4)\n",
    "\n",
    "            topic_qc_metrics = compute_topic_metrics(cto)\n",
    "\n",
    "            with open(\n",
    "                    f\"downstream_analysis/binarized_topics/{sample}__topic_qc.pkl\",'wb'\n",
    "            ) as f:\n",
    "                pickle.dump(topic_qc_metrics, f, protocol=4)\n",
    "\n",
    "            fig_dict= {\n",
    "                'CoherenceVSAssignments': plot_topic_qc(topic_qc_metrics, var_x='Coherence', var_y='Log10_Assignments',\n",
    "                                                        var_color='Gini_index', plot=False, return_fig=True),\n",
    "                'AssignmentsVSCells_in_bin': plot_topic_qc(topic_qc_metrics, var_x='Log10_Assignments',\n",
    "                                                           var_y='Cells_in_binarized_topic', var_color='Gini_index',\n",
    "                                                           plot=False, return_fig=True),\n",
    "                'CoherenceVSCells_in_bin': plot_topic_qc(topic_qc_metrics, var_x='Coherence',\n",
    "                                                         var_y='Cells_in_binarized_topic', var_color='Gini_index',\n",
    "                                                         plot=False, return_fig=True),\n",
    "                'CoherenceVSRegions_in_bin': plot_topic_qc(topic_qc_metrics, var_x='Coherence',\n",
    "                                                           var_y='Regions_in_binarized_topic', var_color='Gini_index',\n",
    "                                                           plot=False, return_fig=True),\n",
    "                'CoherenceVSMarginal_dist': plot_topic_qc(topic_qc_metrics, var_x='Coherence',\n",
    "                                                          var_y='Marginal_topic_dist', var_color='Gini_index',\n",
    "                                                          plot=False, return_fig=True),\n",
    "                'CoherenceVSGini_index': plot_topic_qc(topic_qc_metrics, var_x='Coherence', var_y='Gini_index',\n",
    "                                                       var_color='Gini_index', plot=False, return_fig=True)}\n",
    "\n",
    "            fig=plt.figure(figsize=(40, 43))\n",
    "            i = 1\n",
    "            for fig_ in fig_dict.keys():\n",
    "                plt.subplot(2, 3, i)\n",
    "                img = fig2img(fig_dict[fig_]) #To convert figures to png to plot together, see .utils.py. This converts the figure to png.\n",
    "                plt.imshow(img)\n",
    "                plt.axis('off')\n",
    "                i += 1\n",
    "            plt.subplots_adjust(wspace=0, hspace=-0.70)\n",
    "            #plt.tight_layout()\n",
    "            plt.show()\n",
    "            plt.savefig(f'plots_qc/{sample}__topic_qc_metrics.png', facecolor='white', dpi=150)\n",
    "\n",
    "            topic_annot = topic_annotation(\n",
    "                cto,\n",
    "                annot_var='consensus_cell_type',\n",
    "                binarized_cell_topic=binarized_cell_topics,\n",
    "                general_topic_thr = 0.2\n",
    "            )\n",
    "            topic_qc_metrics = pd.concat(\n",
    "                [\n",
    "                    topic_annot[['consensus_cell_type', 'Ratio_cells_in_topic', 'Ratio_group_in_population']],\n",
    "                    topic_qc_metrics\n",
    "                ],\n",
    "                axis=1\n",
    "            )\n",
    "\n",
    "            imputed_acc_obj = impute_accessibility(\n",
    "                cto,\n",
    "                selected_cells=None,\n",
    "                selected_regions=None,\n",
    "                scale_factor=10**6\n",
    "            )\n",
    "\n",
    "            normalized_imputed_acc_obj = normalize_scores(\n",
    "                imputed_acc_obj,\n",
    "                scale_factor=10**4\n",
    "            )\n",
    "\n",
    "            with open(\n",
    "                f\"downstream_analysis/imputed_acc_objs/{sample}__imputed_acc_obs.pkl\",'wb'\n",
    "            ) as f:\n",
    "                pickle.dump(normalized_imputed_acc_obj, f, protocol=4)\n",
    "\n",
    "            variable_regions = find_highly_variable_features(\n",
    "                normalized_imputed_acc_obj,\n",
    "                min_disp = 0.05,\n",
    "                min_mean = 0.0125,\n",
    "                max_mean = 3,\n",
    "                max_disp = np.inf,\n",
    "                n_bins=20,\n",
    "                n_top_features=None,\n",
    "                plot=True,\n",
    "                save=f'plots_qc/{sample}__HVR.png'\n",
    "            )\n",
    "\n",
    "            print(f\"Found {len(variable_regions)} variable regions\")\n",
    "            with open(f'downstream_analysis/HVRs/{sample}__HVRs.bed', 'w') as f:\n",
    "                for line in [x.replace(':','\\t').replace('-','\\t') for x in variable_regions]:\n",
    "                    f.write(f\"{line}\\n\")\n",
    "                    \n",
    "            with open(\n",
    "                f'downstream_analysis/HVRs/{sample}__HVRs.pkl', 'w'\n",
    "            ) as f:\n",
    "                pickle.dump(variable_regions, f, protocol=4)\n",
    "\n",
    "            os.mkdir(f'downstream_analysis/DARs/{sample}')\n",
    "\n",
    "            markers_dict = find_diff_features(\n",
    "                cto,\n",
    "                normalized_imputed_acc_obj,\n",
    "                variable='consensus_cell_type',\n",
    "                var_features=variable_regions,\n",
    "                contrasts=None,\n",
    "                adjpval_thr=0.05,\n",
    "                log2fc_thr=np.log2(1.5),\n",
    "                n_cpu=5,\n",
    "            )\n",
    "            with open(\n",
    "                    f\"downstream_analysis/DARs/{sample}/{sample}__DARs_dict.pkl\",'wb'\n",
    "            ) as f:\n",
    "                pickle.dump(markers_dict, f, protocol=4)\n",
    "\n",
    "            for cell_type in markers_dict.keys():\n",
    "                markers = markers_dict[cell_type].index.tolist()\n",
    "                df = pd.DataFrame(markers)\n",
    "                if len(df) == 0:\n",
    "                    print(f\"no DARs found for {cell_type} in {sample}\")\n",
    "                else:\n",
    "                    df[[0,1]] = df[0].str.split(':',expand=True)\n",
    "                    df[[1,2]] = df[1].str.split('-',expand=True)\n",
    "                    df[3] = markers_dict[cell_type].index.tolist()\n",
    "                    df[3] = cell_type.replace(' ', '_')# + '_' + df[3].astype(str)\n",
    "                    score = markers_dict[cell_type]['Log2FC']\n",
    "                    score = score.reset_index(drop=True)\n",
    "                    df[4] = score\n",
    "                    df[5] = '.'\n",
    "                    pval = markers_dict[cell_type]['Adjusted_pval']\n",
    "                    pval = pval.reset_index(drop=True)\n",
    "                    df[6] = pval\n",
    "                    name = cell_type.replace(' ', '_')\n",
    "                    # name = name.replace('/', '-')\n",
    "                    if not os.path.exists(f'downstream_analysis/DARs/{sample}'):\n",
    "                        os.mkdir(f'downstream_analysis/DARs/{sample}')\n",
    "\n",
    "                    df.to_csv(f'downstream_analysis/DARs/{sample}/{sample}__{cell_type.replace(\" \", \"_\")}__DARs.bed', sep='\\t', header=False, index=False)\n",
    "                    df[0:2000].to_csv(f'downstream_analysis/DARs/{sample}/{sample}__{cell_type.replace(\" \", \"_\")}__DARs.TOP2k.bed', sep='\\t', header=False, index=False)\n",
    "\n",
    "        else:\n",
    "            print(f'{path} does not exist!')\n",
    "    else:\n",
    "        print(f\"downstream_analysis/DARs/{sample} already exists!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebdc5a30",
   "metadata": {
    "papermill": {
     "duration": 0.008446,
     "end_time": "2022-09-30T13:26:58.539037",
     "exception": false,
     "start_time": "2022-09-30T13:26:58.530591",
     "status": "completed"
    },
    "tags": []
   },
   "source": [
    "# Calculate DARs between male and female"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "afb288ba",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T13:26:58.633954Z",
     "iopub.status.busy": "2022-09-30T13:26:58.633666Z",
     "iopub.status.idle": "2022-09-30T13:26:58.645095Z",
     "shell.execute_reply": "2022-09-30T13:26:58.644480Z"
    },
    "papermill": {
     "duration": 0.098407,
     "end_time": "2022-09-30T13:26:58.646062",
     "exception": false,
     "start_time": "2022-09-30T13:26:58.547655",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/BIO_ddseq_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/BIO_ddseq_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/BIO_ddseq_3.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/BIO_ddseq_4.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/BRO_mtscatac_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/BRO_mtscatac_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xmultiome_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xmultiome_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xv11_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xv11_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xv11_3.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xv11_4.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xv11_5.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xv2_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xv2_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_hydrop_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_hydrop_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_hydrop_3.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_mtscatac_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_mtscatac_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/EPF_hydrop_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/EPF_hydrop_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/EPF_hydrop_3.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/EPF_hydrop_4.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/HAR_ddseq_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/HAR_ddseq_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/MDC_mtscatac_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'MDC_mtscatac_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/MDC_mtscatac_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'OHS_s3atac_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/OHS_s3atac_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'OHS_s3atac_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/OHS_s3atac_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'SAN_10xmultiome_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/SAN_10xmultiome_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'SAN_10xmultiome_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/SAN_10xmultiome_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'STA_10xv11_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/STA_10xv11_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'STA_10xv11_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/STA_10xv11_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'TXG_10xv11_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/TXG_10xv11_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'TXG_10xv2_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/TXG_10xv2_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'TXG_10xv2_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/TXG_10xv2_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'UCS_ddseq_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/UCS_ddseq_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'UCS_ddseq_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/UCS_ddseq_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_10xmultiome_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_10xmultiome_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_10xmultiome_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_10xmultiome_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_10xv1_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_10xv1_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_10xv1_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_10xv1_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_10xv2_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_10xv2_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_10xv2_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_10xv2_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_hydrop_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_hydrop_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_hydrop_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_hydrop_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl'}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 10;\n",
       "                var nbb_unformatted_code = \"normalized_imputed_acc_obj_path_dict = {x.split('/')[-1].split(f'__')[0]:x for x in sorted(glob.glob(f'downstream_analysis/imputed_acc_objs/*normalized_imputed_acc_obs.pkl'))}\\nnormalized_imputed_acc_obj_path_dict\";\n",
       "                var nbb_formatted_code = \"normalized_imputed_acc_obj_path_dict = {\\n    x.split(\\\"/\\\")[-1].split(f\\\"__\\\")[0]: x\\n    for x in sorted(\\n        glob.glob(\\n            f\\\"downstream_analysis/imputed_acc_objs/*normalized_imputed_acc_obs.pkl\\\"\\n        )\\n    )\\n}\\nnormalized_imputed_acc_obj_path_dict\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "normalized_imputed_acc_obj_path_dict = {x.split('/')[-1].split(f'__')[0]:x for x in sorted(glob.glob(f'downstream_analysis/imputed_acc_objs/*normalized_imputed_acc_obs.pkl'))}\n",
    "normalized_imputed_acc_obj_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "8c5f9530",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T13:26:58.663154Z",
     "iopub.status.busy": "2022-09-30T13:26:58.662701Z",
     "iopub.status.idle": "2022-09-30T13:26:58.673011Z",
     "shell.execute_reply": "2022-09-30T13:26:58.672498Z"
    },
    "papermill": {
     "duration": 0.019586,
     "end_time": "2022-09-30T13:26:58.673912",
     "exception": false,
     "start_time": "2022-09-30T13:26:58.654326",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': 'downstream_analysis/HVRs/BIO_ddseq_1.FIXEDCELLS__HVRs.bed',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': 'downstream_analysis/HVRs/BIO_ddseq_2.FIXEDCELLS__HVRs.bed',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': 'downstream_analysis/HVRs/BIO_ddseq_3.FIXEDCELLS__HVRs.bed',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': 'downstream_analysis/HVRs/BIO_ddseq_4.FIXEDCELLS__HVRs.bed',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': 'downstream_analysis/HVRs/BRO_mtscatac_1.FIXEDCELLS__HVRs.bed',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': 'downstream_analysis/HVRs/BRO_mtscatac_2.FIXEDCELLS__HVRs.bed',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': 'downstream_analysis/HVRs/CNA_10xmultiome_1.FIXEDCELLS__HVRs.bed',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': 'downstream_analysis/HVRs/CNA_10xmultiome_2.FIXEDCELLS__HVRs.bed',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': 'downstream_analysis/HVRs/CNA_10xv11_1.FIXEDCELLS__HVRs.bed',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': 'downstream_analysis/HVRs/CNA_10xv11_2.FIXEDCELLS__HVRs.bed',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': 'downstream_analysis/HVRs/CNA_10xv11_3.FIXEDCELLS__HVRs.bed',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': 'downstream_analysis/HVRs/CNA_10xv11_4.FIXEDCELLS__HVRs.bed',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': 'downstream_analysis/HVRs/CNA_10xv11_5.FIXEDCELLS__HVRs.bed',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': 'downstream_analysis/HVRs/CNA_10xv2_1.FIXEDCELLS__HVRs.bed',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': 'downstream_analysis/HVRs/CNA_10xv2_2.FIXEDCELLS__HVRs.bed',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': 'downstream_analysis/HVRs/CNA_hydrop_1.FIXEDCELLS__HVRs.bed',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': 'downstream_analysis/HVRs/CNA_hydrop_2.FIXEDCELLS__HVRs.bed',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': 'downstream_analysis/HVRs/CNA_hydrop_3.FIXEDCELLS__HVRs.bed',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': 'downstream_analysis/HVRs/CNA_mtscatac_1.FIXEDCELLS__HVRs.bed',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': 'downstream_analysis/HVRs/CNA_mtscatac_2.FIXEDCELLS__HVRs.bed',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': 'downstream_analysis/HVRs/EPF_hydrop_1.FIXEDCELLS__HVRs.bed',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': 'downstream_analysis/HVRs/EPF_hydrop_2.FIXEDCELLS__HVRs.bed',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': 'downstream_analysis/HVRs/EPF_hydrop_3.FIXEDCELLS__HVRs.bed',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': 'downstream_analysis/HVRs/EPF_hydrop_4.FIXEDCELLS__HVRs.bed',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': 'downstream_analysis/HVRs/HAR_ddseq_1.FIXEDCELLS__HVRs.bed',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': 'downstream_analysis/HVRs/HAR_ddseq_2.FIXEDCELLS__HVRs.bed',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': 'downstream_analysis/HVRs/MDC_mtscatac_1.FIXEDCELLS__HVRs.bed',\n",
       " 'MDC_mtscatac_2.FIXEDCELLS': 'downstream_analysis/HVRs/MDC_mtscatac_2.FIXEDCELLS__HVRs.bed',\n",
       " 'OHS_s3atac_1.FIXEDCELLS': 'downstream_analysis/HVRs/OHS_s3atac_1.FIXEDCELLS__HVRs.bed',\n",
       " 'OHS_s3atac_2.FIXEDCELLS': 'downstream_analysis/HVRs/OHS_s3atac_2.FIXEDCELLS__HVRs.bed',\n",
       " 'SAN_10xmultiome_1.FIXEDCELLS': 'downstream_analysis/HVRs/SAN_10xmultiome_1.FIXEDCELLS__HVRs.bed',\n",
       " 'SAN_10xmultiome_2.FIXEDCELLS': 'downstream_analysis/HVRs/SAN_10xmultiome_2.FIXEDCELLS__HVRs.bed',\n",
       " 'STA_10xv11_1.FIXEDCELLS': 'downstream_analysis/HVRs/STA_10xv11_1.FIXEDCELLS__HVRs.bed',\n",
       " 'STA_10xv11_2.FIXEDCELLS': 'downstream_analysis/HVRs/STA_10xv11_2.FIXEDCELLS__HVRs.bed',\n",
       " 'TXG_10xv11_1.FIXEDCELLS': 'downstream_analysis/HVRs/TXG_10xv11_1.FIXEDCELLS__HVRs.bed',\n",
       " 'TXG_10xv2_1.FIXEDCELLS': 'downstream_analysis/HVRs/TXG_10xv2_1.FIXEDCELLS__HVRs.bed',\n",
       " 'TXG_10xv2_2.FIXEDCELLS': 'downstream_analysis/HVRs/TXG_10xv2_2.FIXEDCELLS__HVRs.bed',\n",
       " 'UCS_ddseq_1.FIXEDCELLS': 'downstream_analysis/HVRs/UCS_ddseq_1.FIXEDCELLS__HVRs.bed',\n",
       " 'UCS_ddseq_2.FIXEDCELLS': 'downstream_analysis/HVRs/UCS_ddseq_2.FIXEDCELLS__HVRs.bed',\n",
       " 'VIB_10xmultiome_1.FIXEDCELLS': 'downstream_analysis/HVRs/VIB_10xmultiome_1.FIXEDCELLS__HVRs.bed',\n",
       " 'VIB_10xmultiome_2.FIXEDCELLS': 'downstream_analysis/HVRs/VIB_10xmultiome_2.FIXEDCELLS__HVRs.bed',\n",
       " 'VIB_10xv1_1.FIXEDCELLS': 'downstream_analysis/HVRs/VIB_10xv1_1.FIXEDCELLS__HVRs.bed',\n",
       " 'VIB_10xv1_2.FIXEDCELLS': 'downstream_analysis/HVRs/VIB_10xv1_2.FIXEDCELLS__HVRs.bed',\n",
       " 'VIB_10xv2_1.FIXEDCELLS': 'downstream_analysis/HVRs/VIB_10xv2_1.FIXEDCELLS__HVRs.bed',\n",
       " 'VIB_10xv2_2.FIXEDCELLS': 'downstream_analysis/HVRs/VIB_10xv2_2.FIXEDCELLS__HVRs.bed',\n",
       " 'VIB_hydrop_1.FIXEDCELLS': 'downstream_analysis/HVRs/VIB_hydrop_1.FIXEDCELLS__HVRs.bed',\n",
       " 'VIB_hydrop_2.FIXEDCELLS': 'downstream_analysis/HVRs/VIB_hydrop_2.FIXEDCELLS__HVRs.bed'}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 11;\n",
       "                var nbb_unformatted_code = \"hvr_path_dict = {x.split('/')[-1].split(f'__')[0]:x for x in sorted(glob.glob(f'downstream_analysis/HVRs/*bed'))}\\nhvr_path_dict\";\n",
       "                var nbb_formatted_code = \"hvr_path_dict = {\\n    x.split(\\\"/\\\")[-1].split(f\\\"__\\\")[0]: x\\n    for x in sorted(glob.glob(f\\\"downstream_analysis/HVRs/*bed\\\"))\\n}\\nhvr_path_dict\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hvr_path_dict = {x.split('/')[-1].split(f'__')[0]:x for x in sorted(glob.glob(f'downstream_analysis/HVRs/*bed'))}\n",
    "hvr_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4bec77a9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T13:26:58.691015Z",
     "iopub.status.busy": "2022-09-30T13:26:58.690835Z",
     "iopub.status.idle": "2022-09-30T18:41:08.681010Z",
     "shell.execute_reply": "2022-09-30T18:41:08.680468Z"
    },
    "papermill": {
     "duration": 18850.001517,
     "end_time": "2022-09-30T18:41:08.683717",
     "exception": false,
     "start_time": "2022-09-30T13:26:58.682200",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading cistopic_objects/BIO_ddseq_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl\n",
      "Loading cistopic_objects/BIO_ddseq_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_14topics.dimreduc.consensus.pkl\n",
      "Loading cistopic_objects/BIO_ddseq_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl\n",
      "Loading cistopic_objects/BIO_ddseq_4.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl\n",
      "Loading cistopic_objects/BRO_mtscatac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/BRO_mtscatac_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/BRO_mtscatac_1.FIXEDCELLS__HVRs.pkl\n",
      "Dendritic cell\n",
      "CD14+ monocyte\n",
      "B cell\n",
      "CD4+ T cell\n",
      "Cytotoxic T cell\n",
      "Natural killer cell\n",
      "CD16+ monocyte\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 13:27:18,976\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4148889)\u001b[0m 2022-09-30 13:27:54,906 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148889)\u001b[0m 2022-09-30 13:27:55,931 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148888)\u001b[0m 2022-09-30 13:27:56,312 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148888)\u001b[0m 2022-09-30 13:27:57,196 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148886)\u001b[0m 2022-09-30 13:27:57,592 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148887)\u001b[0m 2022-09-30 13:27:59,134 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148885)\u001b[0m 2022-09-30 13:28:00,607 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148884)\u001b[0m 2022-09-30 13:28:02,006 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148885)\u001b[0m 2022-09-30 13:28:02,598 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148882)\u001b[0m 2022-09-30 13:28:03,600 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148884)\u001b[0m 2022-09-30 13:28:04,008 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148881)\u001b[0m 2022-09-30 13:28:04,905 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148886)\u001b[0m 2022-09-30 13:28:05,394 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148878)\u001b[0m 2022-09-30 13:28:06,340 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148887)\u001b[0m 2022-09-30 13:28:06,931 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148880)\u001b[0m 2022-09-30 13:28:07,581 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148878)\u001b[0m 2022-09-30 13:28:09,321 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148880)\u001b[0m 2022-09-30 13:28:10,556 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148882)\u001b[0m 2022-09-30 13:28:10,891 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148881)\u001b[0m 2022-09-30 13:28:12,278 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148889)\u001b[0m 2022-09-30 13:34:39,788 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148888)\u001b[0m 2022-09-30 13:34:46,153 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148885)\u001b[0m 2022-09-30 13:34:57,032 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148889)\u001b[0m 2022-09-30 13:34:59,622 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148889)\u001b[0m 2022-09-30 13:34:59,812 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148884)\u001b[0m 2022-09-30 13:35:00,749 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148889)\u001b[0m 2022-09-30 13:35:02,512 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148888)\u001b[0m 2022-09-30 13:35:06,022 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148888)\u001b[0m 2022-09-30 13:35:06,152 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148888)\u001b[0m 2022-09-30 13:35:08,757 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148880)\u001b[0m 2022-09-30 13:35:09,902 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148878)\u001b[0m 2022-09-30 13:35:15,109 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148885)\u001b[0m 2022-09-30 13:35:16,553 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148885)\u001b[0m 2022-09-30 13:35:16,721 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148885)\u001b[0m 2022-09-30 13:35:18,296 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148884)\u001b[0m 2022-09-30 13:35:20,563 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148884)\u001b[0m 2022-09-30 13:35:20,774 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148884)\u001b[0m 2022-09-30 13:35:22,447 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148880)\u001b[0m 2022-09-30 13:35:29,963 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148878)\u001b[0m 2022-09-30 13:35:34,692 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148882)\u001b[0m 2022-09-30 13:35:43,718 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148881)\u001b[0m 2022-09-30 13:35:45,200 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148886)\u001b[0m 2022-09-30 13:35:51,198 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148887)\u001b[0m 2022-09-30 13:35:54,426 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148882)\u001b[0m 2022-09-30 13:36:00,063 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148881)\u001b[0m 2022-09-30 13:36:01,373 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148886)\u001b[0m 2022-09-30 13:36:06,193 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148887)\u001b[0m 2022-09-30 13:36:08,204 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148889)\u001b[0m 2022-09-30 13:38:36,989 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148888)\u001b[0m 2022-09-30 13:38:41,150 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148885)\u001b[0m 2022-09-30 13:38:42,322 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148884)\u001b[0m 2022-09-30 13:38:45,632 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148889)\u001b[0m 2022-09-30 13:38:45,991 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148888)\u001b[0m 2022-09-30 13:38:49,280 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148885)\u001b[0m 2022-09-30 13:38:49,873 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4148884)\u001b[0m 2022-09-30 13:38:51,197 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in BRO_mtscatac_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in BRO_mtscatac_1.FIXEDCELLS\n",
      "no DARs found for sampleA_B cell_VS_sampleB_B cell in BRO_mtscatac_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in BRO_mtscatac_1.FIXEDCELLS\n",
      "Loading cistopic_objects/BRO_mtscatac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_17topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/BRO_mtscatac_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/BRO_mtscatac_2.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "Natural killer cell\n",
      "CD4+ T cell\n",
      "Cytotoxic T cell\n",
      "B cell\n",
      "CD16+ monocyte\n",
      "Dendritic cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 13:39:14,873\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4149868)\u001b[0m 2022-09-30 13:39:49,994 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149867)\u001b[0m 2022-09-30 13:39:51,163 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149866)\u001b[0m 2022-09-30 13:39:52,427 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149864)\u001b[0m 2022-09-30 13:39:53,767 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149866)\u001b[0m 2022-09-30 13:39:54,809 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149865)\u001b[0m 2022-09-30 13:39:55,005 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149864)\u001b[0m 2022-09-30 13:39:56,156 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149863)\u001b[0m 2022-09-30 13:39:56,353 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149868)\u001b[0m 2022-09-30 13:39:56,393 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149867)\u001b[0m 2022-09-30 13:39:57,721 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149862)\u001b[0m 2022-09-30 13:39:57,676 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149861)\u001b[0m 2022-09-30 13:39:58,995 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149859)\u001b[0m 2022-09-30 13:40:00,177 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149862)\u001b[0m 2022-09-30 13:40:00,861 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149860)\u001b[0m 2022-09-30 13:40:01,558 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149859)\u001b[0m 2022-09-30 13:40:01,647 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149861)\u001b[0m 2022-09-30 13:40:02,105 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149865)\u001b[0m 2022-09-30 13:40:02,166 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149860)\u001b[0m 2022-09-30 13:40:02,999 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149863)\u001b[0m 2022-09-30 13:40:03,449 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149859)\u001b[0m 2022-09-30 13:46:32,204 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149866)\u001b[0m 2022-09-30 13:46:35,524 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149864)\u001b[0m 2022-09-30 13:46:36,011 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149860)\u001b[0m 2022-09-30 13:46:37,770 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149861)\u001b[0m 2022-09-30 13:46:43,633 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149862)\u001b[0m 2022-09-30 13:46:44,371 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149859)\u001b[0m 2022-09-30 13:46:50,715 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149859)\u001b[0m 2022-09-30 13:46:50,849 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149859)\u001b[0m 2022-09-30 13:46:52,167 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149866)\u001b[0m 2022-09-30 13:46:54,409 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149866)\u001b[0m 2022-09-30 13:46:54,598 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149864)\u001b[0m 2022-09-30 13:46:54,979 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149864)\u001b[0m 2022-09-30 13:46:55,102 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149864)\u001b[0m 2022-09-30 13:46:55,718 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149866)\u001b[0m 2022-09-30 13:46:55,911 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149860)\u001b[0m 2022-09-30 13:46:56,329 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149860)\u001b[0m 2022-09-30 13:46:56,501 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149860)\u001b[0m 2022-09-30 13:46:57,042 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149861)\u001b[0m 2022-09-30 13:47:02,493 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149862)\u001b[0m 2022-09-30 13:47:03,368 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149867)\u001b[0m 2022-09-30 13:47:10,979 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149865)\u001b[0m 2022-09-30 13:47:11,792 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149863)\u001b[0m 2022-09-30 13:47:13,371 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149868)\u001b[0m 2022-09-30 13:47:16,691 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149867)\u001b[0m 2022-09-30 13:47:26,212 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149865)\u001b[0m 2022-09-30 13:47:26,919 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149863)\u001b[0m 2022-09-30 13:47:28,448 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149868)\u001b[0m 2022-09-30 13:47:30,721 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149864)\u001b[0m 2022-09-30 13:49:59,920 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149859)\u001b[0m 2022-09-30 13:50:00,934 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149860)\u001b[0m 2022-09-30 13:50:01,346 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149866)\u001b[0m 2022-09-30 13:50:04,322 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149864)\u001b[0m 2022-09-30 13:50:08,279 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149859)\u001b[0m 2022-09-30 13:50:09,066 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149860)\u001b[0m 2022-09-30 13:50:09,326 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4149866)\u001b[0m 2022-09-30 13:50:10,546 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in BRO_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleA_B cell_VS_sampleB_B cell in BRO_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in BRO_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in BRO_mtscatac_2.FIXEDCELLS\n",
      "Loading cistopic_objects/CNA_10xmultiome_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/CNA_10xmultiome_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/CNA_10xmultiome_1.FIXEDCELLS__HVRs.pkl\n",
      "Dendritic cell\n",
      "B cell\n",
      "CD14+ monocyte\n",
      "CD4+ T cell\n",
      "Natural killer cell\n",
      "Cytotoxic T cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 13:50:33,452\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4150928)\u001b[0m 2022-09-30 13:51:06,462 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150928)\u001b[0m 2022-09-30 13:51:07,300 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150927)\u001b[0m 2022-09-30 13:51:07,692 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150927)\u001b[0m 2022-09-30 13:51:08,389 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150925)\u001b[0m 2022-09-30 13:51:08,798 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150926)\u001b[0m 2022-09-30 13:51:09,913 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150924)\u001b[0m 2022-09-30 13:51:10,971 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150925)\u001b[0m 2022-09-30 13:51:11,087 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150922)\u001b[0m 2022-09-30 13:51:12,026 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150926)\u001b[0m 2022-09-30 13:51:12,232 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150923)\u001b[0m 2022-09-30 13:51:12,970 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150921)\u001b[0m 2022-09-30 13:51:14,201 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150920)\u001b[0m 2022-09-30 13:51:15,290 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150924)\u001b[0m 2022-09-30 13:51:15,514 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150919)\u001b[0m 2022-09-30 13:51:16,345 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150922)\u001b[0m 2022-09-30 13:51:16,606 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150920)\u001b[0m 2022-09-30 13:51:17,346 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150919)\u001b[0m 2022-09-30 13:51:18,360 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150923)\u001b[0m 2022-09-30 13:51:18,331 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150921)\u001b[0m 2022-09-30 13:51:19,584 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150928)\u001b[0m 2022-09-30 13:56:30,793 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150927)\u001b[0m 2022-09-30 13:56:34,462 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150928)\u001b[0m 2022-09-30 13:56:46,474 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150928)\u001b[0m 2022-09-30 13:56:46,596 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150925)\u001b[0m 2022-09-30 13:56:49,173 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150928)\u001b[0m 2022-09-30 13:56:50,009 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150927)\u001b[0m 2022-09-30 13:56:50,416 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150927)\u001b[0m 2022-09-30 13:56:50,550 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150926)\u001b[0m 2022-09-30 13:56:53,530 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150927)\u001b[0m 2022-09-30 13:56:53,867 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150919)\u001b[0m 2022-09-30 13:56:56,291 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150920)\u001b[0m 2022-09-30 13:56:56,466 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150925)\u001b[0m 2022-09-30 13:57:05,366 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150926)\u001b[0m 2022-09-30 13:57:09,401 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150919)\u001b[0m 2022-09-30 13:57:11,536 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150920)\u001b[0m 2022-09-30 13:57:11,625 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150922)\u001b[0m 2022-09-30 13:57:17,260 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150923)\u001b[0m 2022-09-30 13:57:19,734 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150921)\u001b[0m 2022-09-30 13:57:20,010 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150924)\u001b[0m 2022-09-30 13:57:21,578 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150922)\u001b[0m 2022-09-30 13:57:27,508 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150923)\u001b[0m 2022-09-30 13:57:29,778 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150921)\u001b[0m 2022-09-30 13:57:29,989 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150924)\u001b[0m 2022-09-30 13:57:30,996 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150928)\u001b[0m 2022-09-30 13:58:44,776 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150927)\u001b[0m 2022-09-30 13:58:46,581 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150928)\u001b[0m 2022-09-30 13:58:48,937 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4150927)\u001b[0m 2022-09-30 13:58:50,087 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in CNA_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in CNA_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleA_B cell_VS_sampleB_B cell in CNA_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in CNA_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell in CNA_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in CNA_10xmultiome_1.FIXEDCELLS\n",
      "Loading cistopic_objects/CNA_10xmultiome_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_8topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/CNA_10xmultiome_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/CNA_10xmultiome_2.FIXEDCELLS__HVRs.pkl\n",
      "Natural killer cell\n",
      "CD14+ monocyte\n",
      "B cell\n",
      "Cytotoxic T cell\n",
      "CD4+ T cell\n",
      "Dendritic cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 13:59:10,166\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4151909)\u001b[0m 2022-09-30 13:59:42,962 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151908)\u001b[0m 2022-09-30 13:59:44,037 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151909)\u001b[0m 2022-09-30 13:59:44,685 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151907)\u001b[0m 2022-09-30 13:59:45,130 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151908)\u001b[0m 2022-09-30 13:59:45,712 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151906)\u001b[0m 2022-09-30 13:59:46,042 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151904)\u001b[0m 2022-09-30 13:59:46,996 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151905)\u001b[0m 2022-09-30 13:59:48,181 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151904)\u001b[0m 2022-09-30 13:59:48,746 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151903)\u001b[0m 2022-09-30 13:59:49,074 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151907)\u001b[0m 2022-09-30 13:59:49,439 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151905)\u001b[0m 2022-09-30 13:59:49,901 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151902)\u001b[0m 2022-09-30 13:59:50,025 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151906)\u001b[0m 2022-09-30 13:59:50,318 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151901)\u001b[0m 2022-09-30 13:59:51,023 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151903)\u001b[0m 2022-09-30 13:59:51,681 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151900)\u001b[0m 2022-09-30 13:59:52,170 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151902)\u001b[0m 2022-09-30 13:59:52,687 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151901)\u001b[0m 2022-09-30 13:59:56,512 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151900)\u001b[0m 2022-09-30 13:59:57,549 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151909)\u001b[0m 2022-09-30 14:05:29,113 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151908)\u001b[0m 2022-09-30 14:05:29,497 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151904)\u001b[0m 2022-09-30 14:05:35,169 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151905)\u001b[0m 2022-09-30 14:05:38,250 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151902)\u001b[0m 2022-09-30 14:05:42,055 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151903)\u001b[0m 2022-09-30 14:05:43,727 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151909)\u001b[0m 2022-09-30 14:05:45,359 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151909)\u001b[0m 2022-09-30 14:05:45,485 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151908)\u001b[0m 2022-09-30 14:05:45,984 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151908)\u001b[0m 2022-09-30 14:05:46,113 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151909)\u001b[0m 2022-09-30 14:05:46,215 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151908)\u001b[0m 2022-09-30 14:05:46,789 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151904)\u001b[0m 2022-09-30 14:05:51,504 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151905)\u001b[0m 2022-09-30 14:05:54,534 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151902)\u001b[0m 2022-09-30 14:05:57,674 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151903)\u001b[0m 2022-09-30 14:05:58,896 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151907)\u001b[0m 2022-09-30 14:06:01,254 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151906)\u001b[0m 2022-09-30 14:06:02,554 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151901)\u001b[0m 2022-09-30 14:06:09,509 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151900)\u001b[0m 2022-09-30 14:06:11,273 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151907)\u001b[0m 2022-09-30 14:06:11,659 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151906)\u001b[0m 2022-09-30 14:06:12,775 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151901)\u001b[0m 2022-09-30 14:06:17,874 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151900)\u001b[0m 2022-09-30 14:06:18,965 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151908)\u001b[0m 2022-09-30 14:07:30,908 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151909)\u001b[0m 2022-09-30 14:07:31,346 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151908)\u001b[0m 2022-09-30 14:07:35,387 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4151909)\u001b[0m 2022-09-30 14:07:35,640 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "no DARs found for sampleA_Natural killer cell_VS_sampleB_Natural killer cell in CNA_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in CNA_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte in CNA_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte in CNA_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleA_B cell_VS_sampleB_B cell in CNA_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in CNA_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell in CNA_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in CNA_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in CNA_10xmultiome_2.FIXEDCELLS\n",
      "Loading cistopic_objects/CNA_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/CNA_10xv11_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/CNA_10xv11_1.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "CD16+ monocyte\n",
      "CD4+ T cell\n",
      "Cytotoxic T cell\n",
      "B cell\n",
      "Natural killer cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 14:07:55,234\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4152894)\u001b[0m 2022-09-30 14:08:27,995 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152893)\u001b[0m 2022-09-30 14:08:29,077 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152892)\u001b[0m 2022-09-30 14:08:30,004 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152892)\u001b[0m 2022-09-30 14:08:30,978 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152891)\u001b[0m 2022-09-30 14:08:31,017 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152891)\u001b[0m 2022-09-30 14:08:31,974 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152890)\u001b[0m 2022-09-30 14:08:31,952 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152894)\u001b[0m 2022-09-30 14:08:32,147 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152893)\u001b[0m 2022-09-30 14:08:33,089 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152889)\u001b[0m 2022-09-30 14:08:33,037 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152888)\u001b[0m 2022-09-30 14:08:33,912 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152886)\u001b[0m 2022-09-30 14:08:34,886 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152888)\u001b[0m 2022-09-30 14:08:35,690 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152885)\u001b[0m 2022-09-30 14:08:35,828 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152886)\u001b[0m 2022-09-30 14:08:36,653 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152890)\u001b[0m 2022-09-30 14:08:36,728 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152887)\u001b[0m 2022-09-30 14:08:36,948 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152889)\u001b[0m 2022-09-30 14:08:37,847 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152885)\u001b[0m 2022-09-30 14:08:38,050 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152887)\u001b[0m 2022-09-30 14:08:39,254 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152892)\u001b[0m 2022-09-30 14:14:44,443 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152891)\u001b[0m 2022-09-30 14:14:50,411 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152888)\u001b[0m 2022-09-30 14:14:59,090 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152886)\u001b[0m 2022-09-30 14:15:00,173 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152892)\u001b[0m 2022-09-30 14:15:02,878 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152892)\u001b[0m 2022-09-30 14:15:02,994 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152892)\u001b[0m 2022-09-30 14:15:04,829 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152885)\u001b[0m 2022-09-30 14:15:05,226 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152887)\u001b[0m 2022-09-30 14:15:07,986 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152891)\u001b[0m 2022-09-30 14:15:08,860 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152891)\u001b[0m 2022-09-30 14:15:09,059 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152891)\u001b[0m 2022-09-30 14:15:10,915 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152888)\u001b[0m 2022-09-30 14:15:17,190 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152886)\u001b[0m 2022-09-30 14:15:18,547 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152894)\u001b[0m 2022-09-30 14:15:18,714 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152893)\u001b[0m 2022-09-30 14:15:19,793 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152890)\u001b[0m 2022-09-30 14:15:21,363 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152885)\u001b[0m 2022-09-30 14:15:22,664 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152889)\u001b[0m 2022-09-30 14:15:24,121 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152887)\u001b[0m 2022-09-30 14:15:24,539 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152894)\u001b[0m 2022-09-30 14:15:31,417 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152893)\u001b[0m 2022-09-30 14:15:32,115 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152890)\u001b[0m 2022-09-30 14:15:32,928 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152889)\u001b[0m 2022-09-30 14:15:34,374 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152892)\u001b[0m 2022-09-30 14:17:02,837 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152891)\u001b[0m 2022-09-30 14:17:05,428 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4152892)\u001b[0m 2022-09-30 14:17:07,900 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "no DARs found for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte in CNA_10xv11_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte in CNA_10xv11_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in CNA_10xv11_1.FIXEDCELLS\n",
      "Loading cistopic_objects/CNA_10xv11_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/CNA_10xv11_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/CNA_10xv11_2.FIXEDCELLS__HVRs.pkl\n",
      "Cytotoxic T cell\n",
      "CD4+ T cell\n",
      "B cell\n",
      "Natural killer cell\n",
      "CD14+ monocyte\n",
      "Dendritic cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 14:17:30,327\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4153815)\u001b[0m 2022-09-30 14:18:03,015 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153813)\u001b[0m 2022-09-30 14:18:04,083 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153814)\u001b[0m 2022-09-30 14:18:05,150 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153815)\u001b[0m 2022-09-30 14:18:05,254 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153812)\u001b[0m 2022-09-30 14:18:06,061 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153813)\u001b[0m 2022-09-30 14:18:06,253 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153811)\u001b[0m 2022-09-30 14:18:07,044 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153810)\u001b[0m 2022-09-30 14:18:08,004 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153811)\u001b[0m 2022-09-30 14:18:08,923 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153808)\u001b[0m 2022-09-30 14:18:08,943 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153810)\u001b[0m 2022-09-30 14:18:09,904 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153814)\u001b[0m 2022-09-30 14:18:09,924 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153809)\u001b[0m 2022-09-30 14:18:09,961 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153812)\u001b[0m 2022-09-30 14:18:10,814 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153808)\u001b[0m 2022-09-30 14:18:10,732 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153807)\u001b[0m 2022-09-30 14:18:10,932 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153809)\u001b[0m 2022-09-30 14:18:11,701 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153806)\u001b[0m 2022-09-30 14:18:11,904 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153807)\u001b[0m 2022-09-30 14:18:15,643 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153806)\u001b[0m 2022-09-30 14:18:16,565 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153815)\u001b[0m 2022-09-30 14:24:15,839 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153813)\u001b[0m 2022-09-30 14:24:18,303 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153808)\u001b[0m 2022-09-30 14:24:20,207 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153811)\u001b[0m 2022-09-30 14:24:20,903 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153809)\u001b[0m 2022-09-30 14:24:22,541 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153810)\u001b[0m 2022-09-30 14:24:25,002 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153815)\u001b[0m 2022-09-30 14:24:33,244 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153815)\u001b[0m 2022-09-30 14:24:33,440 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153815)\u001b[0m 2022-09-30 14:24:33,879 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153813)\u001b[0m 2022-09-30 14:24:35,823 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153813)\u001b[0m 2022-09-30 14:24:36,079 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153813)\u001b[0m 2022-09-30 14:24:36,485 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153808)\u001b[0m 2022-09-30 14:24:37,768 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153811)\u001b[0m 2022-09-30 14:24:38,593 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153809)\u001b[0m 2022-09-30 14:24:39,793 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153810)\u001b[0m 2022-09-30 14:24:41,690 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153814)\u001b[0m 2022-09-30 14:24:44,976 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153812)\u001b[0m 2022-09-30 14:24:45,366 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153807)\u001b[0m 2022-09-30 14:24:50,636 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153806)\u001b[0m 2022-09-30 14:24:52,481 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153814)\u001b[0m 2022-09-30 14:24:56,316 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153812)\u001b[0m 2022-09-30 14:24:56,630 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153807)\u001b[0m 2022-09-30 14:25:00,342 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153806)\u001b[0m 2022-09-30 14:25:01,404 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153815)\u001b[0m 2022-09-30 14:26:19,682 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153813)\u001b[0m 2022-09-30 14:26:20,922 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4153815)\u001b[0m 2022-09-30 14:26:24,397 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in CNA_10xv11_2.FIXEDCELLS\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in CNA_10xv11_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in CNA_10xv11_2.FIXEDCELLS\n",
      "Loading cistopic_objects/CNA_10xv11_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/CNA_10xv11_3.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/CNA_10xv11_3.FIXEDCELLS__HVRs.pkl\n",
      "B cell\n",
      "CD14+ monocyte\n",
      "CD4+ T cell\n",
      "Natural killer cell\n",
      "Dendritic cell\n",
      "Cytotoxic T cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 14:26:53,393\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4154717)\u001b[0m 2022-09-30 14:27:27,902 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154716)\u001b[0m 2022-09-30 14:27:30,085 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154715)\u001b[0m 2022-09-30 14:27:31,749 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154717)\u001b[0m 2022-09-30 14:27:32,150 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154714)\u001b[0m 2022-09-30 14:27:33,456 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154716)\u001b[0m 2022-09-30 14:27:33,833 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154713)\u001b[0m 2022-09-30 14:27:35,053 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154712)\u001b[0m 2022-09-30 14:27:36,974 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154710)\u001b[0m 2022-09-30 14:27:38,799 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154711)\u001b[0m 2022-09-30 14:27:40,444 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154715)\u001b[0m 2022-09-30 14:27:41,151 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154708)\u001b[0m 2022-09-30 14:27:41,943 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154714)\u001b[0m 2022-09-30 14:27:42,728 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154708)\u001b[0m 2022-09-30 14:27:42,868 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154709)\u001b[0m 2022-09-30 14:27:43,692 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154710)\u001b[0m 2022-09-30 14:27:43,826 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154713)\u001b[0m 2022-09-30 14:27:44,462 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154709)\u001b[0m 2022-09-30 14:27:44,628 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154711)\u001b[0m 2022-09-30 14:27:45,485 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154712)\u001b[0m 2022-09-30 14:27:46,313 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154708)\u001b[0m 2022-09-30 14:34:20,815 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154709)\u001b[0m 2022-09-30 14:34:28,547 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154717)\u001b[0m 2022-09-30 14:34:31,149 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154716)\u001b[0m 2022-09-30 14:34:34,738 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154708)\u001b[0m 2022-09-30 14:34:40,017 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154708)\u001b[0m 2022-09-30 14:34:40,180 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154708)\u001b[0m 2022-09-30 14:34:44,247 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154709)\u001b[0m 2022-09-30 14:34:47,876 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154709)\u001b[0m 2022-09-30 14:34:48,052 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154717)\u001b[0m 2022-09-30 14:34:50,745 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154709)\u001b[0m 2022-09-30 14:34:52,049 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154716)\u001b[0m 2022-09-30 14:34:53,937 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154711)\u001b[0m 2022-09-30 14:34:54,238 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154710)\u001b[0m 2022-09-30 14:34:56,863 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154711)\u001b[0m 2022-09-30 14:35:10,202 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154710)\u001b[0m 2022-09-30 14:35:12,449 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154713)\u001b[0m 2022-09-30 14:35:15,280 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154712)\u001b[0m 2022-09-30 14:35:16,122 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154715)\u001b[0m 2022-09-30 14:35:23,567 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154714)\u001b[0m 2022-09-30 14:35:26,270 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154713)\u001b[0m 2022-09-30 14:35:27,938 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154712)\u001b[0m 2022-09-30 14:35:28,487 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154715)\u001b[0m 2022-09-30 14:35:33,792 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154714)\u001b[0m 2022-09-30 14:35:35,408 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154708)\u001b[0m 2022-09-30 14:37:03,527 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154709)\u001b[0m 2022-09-30 14:37:07,495 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154708)\u001b[0m 2022-09-30 14:37:08,878 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4154709)\u001b[0m 2022-09-30 14:37:11,407 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in CNA_10xv11_3.FIXEDCELLS\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in CNA_10xv11_3.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in CNA_10xv11_3.FIXEDCELLS\n",
      "Loading cistopic_objects/CNA_10xv11_4.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/CNA_10xv11_4.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/CNA_10xv11_4.FIXEDCELLS__HVRs.pkl\n",
      "CD4+ T cell\n",
      "Natural killer cell\n",
      "CD14+ monocyte\n",
      "Cytotoxic T cell\n",
      "B cell\n",
      "CD16+ monocyte\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 14:37:28,266\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4156106)\u001b[0m 2022-09-30 14:37:59,485 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156104)\u001b[0m 2022-09-30 14:37:59,854 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156105)\u001b[0m 2022-09-30 14:38:00,039 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156103)\u001b[0m 2022-09-30 14:38:00,520 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156100)\u001b[0m 2022-09-30 14:38:00,739 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156105)\u001b[0m 2022-09-30 14:38:00,808 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156106)\u001b[0m 2022-09-30 14:38:00,990 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156102)\u001b[0m 2022-09-30 14:38:01,083 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156104)\u001b[0m 2022-09-30 14:38:01,279 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156103)\u001b[0m 2022-09-30 14:38:01,297 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156101)\u001b[0m 2022-09-30 14:38:01,510 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156100)\u001b[0m 2022-09-30 14:38:01,763 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156099)\u001b[0m 2022-09-30 14:38:01,797 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156102)\u001b[0m 2022-09-30 14:38:02,096 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156101)\u001b[0m 2022-09-30 14:38:02,062 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156097)\u001b[0m 2022-09-30 14:38:02,255 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156099)\u001b[0m 2022-09-30 14:38:02,331 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156098)\u001b[0m 2022-09-30 14:38:02,504 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156097)\u001b[0m 2022-09-30 14:38:02,793 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156098)\u001b[0m 2022-09-30 14:38:02,968 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156099)\u001b[0m 2022-09-30 14:44:13,939 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156105)\u001b[0m 2022-09-30 14:44:15,434 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156103)\u001b[0m 2022-09-30 14:44:15,636 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156098)\u001b[0m 2022-09-30 14:44:16,363 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156097)\u001b[0m 2022-09-30 14:44:17,421 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156101)\u001b[0m 2022-09-30 14:44:17,521 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156102)\u001b[0m 2022-09-30 14:44:18,803 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156100)\u001b[0m 2022-09-30 14:44:22,789 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156104)\u001b[0m 2022-09-30 14:44:24,325 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156106)\u001b[0m 2022-09-30 14:44:27,178 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156099)\u001b[0m 2022-09-30 14:44:31,797 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156099)\u001b[0m 2022-09-30 14:44:31,933 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156099)\u001b[0m 2022-09-30 14:44:32,191 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156105)\u001b[0m 2022-09-30 14:44:33,562 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156103)\u001b[0m 2022-09-30 14:44:33,590 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156105)\u001b[0m 2022-09-30 14:44:33,684 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156105)\u001b[0m 2022-09-30 14:44:33,887 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156098)\u001b[0m 2022-09-30 14:44:34,231 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156097)\u001b[0m 2022-09-30 14:44:35,237 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156101)\u001b[0m 2022-09-30 14:44:35,430 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156102)\u001b[0m 2022-09-30 14:44:36,034 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156100)\u001b[0m 2022-09-30 14:44:38,218 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156104)\u001b[0m 2022-09-30 14:44:39,007 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156106)\u001b[0m 2022-09-30 14:44:40,163 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156099)\u001b[0m 2022-09-30 14:46:10,039 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156105)\u001b[0m 2022-09-30 14:46:11,421 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4156099)\u001b[0m 2022-09-30 14:46:15,022 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "no DARs found for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell in CNA_10xv11_4.FIXEDCELLS\n",
      "no DARs found for sampleA_Natural killer cell_VS_sampleB_Natural killer cell in CNA_10xv11_4.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in CNA_10xv11_4.FIXEDCELLS\n",
      "no DARs found for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte in CNA_10xv11_4.FIXEDCELLS\n",
      "no DARs found for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte in CNA_10xv11_4.FIXEDCELLS\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in CNA_10xv11_4.FIXEDCELLS\n",
      "no DARs found for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell in CNA_10xv11_4.FIXEDCELLS\n",
      "no DARs found for sampleA_B cell_VS_sampleB_B cell in CNA_10xv11_4.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in CNA_10xv11_4.FIXEDCELLS\n",
      "no DARs found for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte in CNA_10xv11_4.FIXEDCELLS\n",
      "no DARs found for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte in CNA_10xv11_4.FIXEDCELLS\n",
      "Loading cistopic_objects/CNA_10xv11_5.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/CNA_10xv11_5.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/CNA_10xv11_5.FIXEDCELLS__HVRs.pkl\n",
      "Cytotoxic T cell\n",
      "CD14+ monocyte\n",
      "Dendritic cell\n",
      "B cell\n",
      "Natural killer cell\n",
      "CD4+ T cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 14:46:28,968\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4157627)\u001b[0m 2022-09-30 14:47:00,518 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157626)\u001b[0m 2022-09-30 14:47:00,974 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157625)\u001b[0m 2022-09-30 14:47:01,505 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157627)\u001b[0m 2022-09-30 14:47:01,640 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157626)\u001b[0m 2022-09-30 14:47:01,994 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157624)\u001b[0m 2022-09-30 14:47:02,012 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157622)\u001b[0m 2022-09-30 14:47:02,532 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157622)\u001b[0m 2022-09-30 14:47:02,802 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157623)\u001b[0m 2022-09-30 14:47:02,925 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157623)\u001b[0m 2022-09-30 14:47:03,250 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157625)\u001b[0m 2022-09-30 14:47:03,376 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157621)\u001b[0m 2022-09-30 14:47:03,508 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157624)\u001b[0m 2022-09-30 14:47:03,835 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157620)\u001b[0m 2022-09-30 14:47:04,005 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157621)\u001b[0m 2022-09-30 14:47:04,495 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157618)\u001b[0m 2022-09-30 14:47:04,490 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157620)\u001b[0m 2022-09-30 14:47:04,934 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157619)\u001b[0m 2022-09-30 14:47:04,851 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157618)\u001b[0m 2022-09-30 14:47:05,340 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157619)\u001b[0m 2022-09-30 14:47:05,722 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157623)\u001b[0m 2022-09-30 14:53:52,663 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157622)\u001b[0m 2022-09-30 14:53:53,950 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157626)\u001b[0m 2022-09-30 14:54:00,543 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157619)\u001b[0m 2022-09-30 14:54:01,325 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157627)\u001b[0m 2022-09-30 14:54:04,869 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157621)\u001b[0m 2022-09-30 14:54:04,936 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157620)\u001b[0m 2022-09-30 14:54:07,095 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157618)\u001b[0m 2022-09-30 14:54:07,841 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157623)\u001b[0m 2022-09-30 14:54:12,573 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157623)\u001b[0m 2022-09-30 14:54:12,757 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157622)\u001b[0m 2022-09-30 14:54:14,266 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157622)\u001b[0m 2022-09-30 14:54:14,384 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157623)\u001b[0m 2022-09-30 14:54:14,685 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157625)\u001b[0m 2022-09-30 14:54:15,767 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157622)\u001b[0m 2022-09-30 14:54:16,297 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157624)\u001b[0m 2022-09-30 14:54:20,318 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157626)\u001b[0m 2022-09-30 14:54:20,734 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157619)\u001b[0m 2022-09-30 14:54:21,276 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157621)\u001b[0m 2022-09-30 14:54:24,315 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157627)\u001b[0m 2022-09-30 14:54:24,524 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157620)\u001b[0m 2022-09-30 14:54:25,737 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157618)\u001b[0m 2022-09-30 14:54:26,167 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157625)\u001b[0m 2022-09-30 14:54:29,666 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157624)\u001b[0m 2022-09-30 14:54:31,579 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157623)\u001b[0m 2022-09-30 14:56:13,258 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157622)\u001b[0m 2022-09-30 14:56:14,060 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4157623)\u001b[0m 2022-09-30 14:56:18,699 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in CNA_10xv11_5.FIXEDCELLS\n",
      "no DARs found for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell in CNA_10xv11_5.FIXEDCELLS\n",
      "no DARs found for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte in CNA_10xv11_5.FIXEDCELLS\n",
      "no DARs found for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte in CNA_10xv11_5.FIXEDCELLS\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in CNA_10xv11_5.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in CNA_10xv11_5.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in CNA_10xv11_5.FIXEDCELLS\n",
      "no DARs found for sampleA_Natural killer cell_VS_sampleB_Natural killer cell in CNA_10xv11_5.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in CNA_10xv11_5.FIXEDCELLS\n",
      "Loading cistopic_objects/CNA_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/CNA_10xv2_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/CNA_10xv2_1.FIXEDCELLS__HVRs.pkl\n",
      "B cell\n",
      "CD4+ T cell\n",
      "CD14+ monocyte\n",
      "Dendritic cell\n",
      "Cytotoxic T cell\n",
      "Natural killer cell\n",
      "CD16+ monocyte\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 14:56:46,398\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4158694)\u001b[0m 2022-09-30 14:57:21,842 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158692)\u001b[0m 2022-09-30 14:57:23,279 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158691)\u001b[0m 2022-09-30 14:57:24,728 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158694)\u001b[0m 2022-09-30 14:57:25,407 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158693)\u001b[0m 2022-09-30 14:57:26,214 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158692)\u001b[0m 2022-09-30 14:57:26,814 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158690)\u001b[0m 2022-09-30 14:57:27,513 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158688)\u001b[0m 2022-09-30 14:57:29,208 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158689)\u001b[0m 2022-09-30 14:57:30,453 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158689)\u001b[0m 2022-09-30 14:57:31,632 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158687)\u001b[0m 2022-09-30 14:57:31,920 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158690)\u001b[0m 2022-09-30 14:57:32,432 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158691)\u001b[0m 2022-09-30 14:57:32,540 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158687)\u001b[0m 2022-09-30 14:57:33,055 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158686)\u001b[0m 2022-09-30 14:57:33,255 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158693)\u001b[0m 2022-09-30 14:57:34,031 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158688)\u001b[0m 2022-09-30 14:57:34,046 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158685)\u001b[0m 2022-09-30 14:57:34,696 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158686)\u001b[0m 2022-09-30 14:57:37,693 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158685)\u001b[0m 2022-09-30 14:57:39,112 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158687)\u001b[0m 2022-09-30 15:03:39,517 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158689)\u001b[0m 2022-09-30 15:03:43,633 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158692)\u001b[0m 2022-09-30 15:03:56,164 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158694)\u001b[0m 2022-09-30 15:03:57,304 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158687)\u001b[0m 2022-09-30 15:03:57,347 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158687)\u001b[0m 2022-09-30 15:03:57,472 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158687)\u001b[0m 2022-09-30 15:04:00,293 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158689)\u001b[0m 2022-09-30 15:04:01,392 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158689)\u001b[0m 2022-09-30 15:04:01,503 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158689)\u001b[0m 2022-09-30 15:04:04,169 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158692)\u001b[0m 2022-09-30 15:04:14,051 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158692)\u001b[0m 2022-09-30 15:04:14,172 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158694)\u001b[0m 2022-09-30 15:04:15,180 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158694)\u001b[0m 2022-09-30 15:04:15,441 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158692)\u001b[0m 2022-09-30 15:04:15,652 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158685)\u001b[0m 2022-09-30 15:04:16,497 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158694)\u001b[0m 2022-09-30 15:04:16,884 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158690)\u001b[0m 2022-09-30 15:04:17,003 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158686)\u001b[0m 2022-09-30 15:04:17,030 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158688)\u001b[0m 2022-09-30 15:04:18,203 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158691)\u001b[0m 2022-09-30 15:04:33,616 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158685)\u001b[0m 2022-09-30 15:04:34,314 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158686)\u001b[0m 2022-09-30 15:04:34,882 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158690)\u001b[0m 2022-09-30 15:04:35,089 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158688)\u001b[0m 2022-09-30 15:04:36,170 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158693)\u001b[0m 2022-09-30 15:04:38,096 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158691)\u001b[0m 2022-09-30 15:04:45,797 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158693)\u001b[0m 2022-09-30 15:04:49,144 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158687)\u001b[0m 2022-09-30 15:07:11,221 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158689)\u001b[0m 2022-09-30 15:07:15,420 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158692)\u001b[0m 2022-09-30 15:07:15,584 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158694)\u001b[0m 2022-09-30 15:07:16,595 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158687)\u001b[0m 2022-09-30 15:07:19,419 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158692)\u001b[0m 2022-09-30 15:07:22,824 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158689)\u001b[0m 2022-09-30 15:07:22,761 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4158694)\u001b[0m 2022-09-30 15:07:23,243 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in CNA_10xv2_1.FIXEDCELLS\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in CNA_10xv2_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in CNA_10xv2_1.FIXEDCELLS\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in CNA_10xv2_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell in CNA_10xv2_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in CNA_10xv2_1.FIXEDCELLS\n",
      "no DARs found for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte in CNA_10xv2_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte in CNA_10xv2_1.FIXEDCELLS\n",
      "Loading cistopic_objects/CNA_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/CNA_10xv2_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/CNA_10xv2_2.FIXEDCELLS__HVRs.pkl\n",
      "CD16+ monocyte\n",
      "CD4+ T cell\n",
      "Natural killer cell\n",
      "CD14+ monocyte\n",
      "Cytotoxic T cell\n",
      "B cell\n",
      "Dendritic cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 15:08:08,216\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4161525)\u001b[0m 2022-09-30 15:08:47,571 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161524)\u001b[0m 2022-09-30 15:08:50,681 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161522)\u001b[0m 2022-09-30 15:08:55,526 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161525)\u001b[0m 2022-09-30 15:08:56,741 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161524)\u001b[0m 2022-09-30 15:08:59,305 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161523)\u001b[0m 2022-09-30 15:09:00,476 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161521)\u001b[0m 2022-09-30 15:09:04,504 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161520)\u001b[0m 2022-09-30 15:09:07,864 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161521)\u001b[0m 2022-09-30 15:09:10,308 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161517)\u001b[0m 2022-09-30 15:09:10,980 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161520)\u001b[0m 2022-09-30 15:09:13,141 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161519)\u001b[0m 2022-09-30 15:09:13,191 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161516)\u001b[0m 2022-09-30 15:09:15,160 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161518)\u001b[0m 2022-09-30 15:09:17,221 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161522)\u001b[0m 2022-09-30 15:09:18,657 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161523)\u001b[0m 2022-09-30 15:09:19,500 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161517)\u001b[0m 2022-09-30 15:09:21,956 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161519)\u001b[0m 2022-09-30 15:09:24,021 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161516)\u001b[0m 2022-09-30 15:09:24,113 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161518)\u001b[0m 2022-09-30 15:09:26,200 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161525)\u001b[0m 2022-09-30 15:15:21,855 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161524)\u001b[0m 2022-09-30 15:15:25,214 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161525)\u001b[0m 2022-09-30 15:15:40,028 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161525)\u001b[0m 2022-09-30 15:15:40,159 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161524)\u001b[0m 2022-09-30 15:15:43,358 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161524)\u001b[0m 2022-09-30 15:15:43,556 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161521)\u001b[0m 2022-09-30 15:15:44,582 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161520)\u001b[0m 2022-09-30 15:15:45,999 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161525)\u001b[0m 2022-09-30 15:15:46,150 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161524)\u001b[0m 2022-09-30 15:15:49,332 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161521)\u001b[0m 2022-09-30 15:16:02,646 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161521)\u001b[0m 2022-09-30 15:16:02,880 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161520)\u001b[0m 2022-09-30 15:16:04,253 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161520)\u001b[0m 2022-09-30 15:16:04,436 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161521)\u001b[0m 2022-09-30 15:16:04,641 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161520)\u001b[0m 2022-09-30 15:16:06,096 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161516)\u001b[0m 2022-09-30 15:16:16,246 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161518)\u001b[0m 2022-09-30 15:16:20,656 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161517)\u001b[0m 2022-09-30 15:16:30,641 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161519)\u001b[0m 2022-09-30 15:16:34,096 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161516)\u001b[0m 2022-09-30 15:16:34,616 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161518)\u001b[0m 2022-09-30 15:16:38,473 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161522)\u001b[0m 2022-09-30 15:16:46,685 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161517)\u001b[0m 2022-09-30 15:16:46,928 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161523)\u001b[0m 2022-09-30 15:16:46,979 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161519)\u001b[0m 2022-09-30 15:16:49,232 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161522)\u001b[0m 2022-09-30 15:16:58,861 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161523)\u001b[0m 2022-09-30 15:16:59,020 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161520)\u001b[0m 2022-09-30 15:19:14,809 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161521)\u001b[0m 2022-09-30 15:19:15,631 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161525)\u001b[0m 2022-09-30 15:19:17,684 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161524)\u001b[0m 2022-09-30 15:19:20,302 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161520)\u001b[0m 2022-09-30 15:19:22,995 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161521)\u001b[0m 2022-09-30 15:19:23,676 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4161525)\u001b[0m 2022-09-30 15:19:24,952 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "no DARs found for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte in CNA_10xv2_2.FIXEDCELLS\n",
      "no DARs found for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte in CNA_10xv2_2.FIXEDCELLS\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in CNA_10xv2_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in CNA_10xv2_2.FIXEDCELLS\n",
      "Loading cistopic_objects/CNA_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/CNA_hydrop_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/CNA_hydrop_1.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "Cytotoxic T cell\n",
      "CD4+ T cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 15:19:45,678\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4162813)\u001b[0m 2022-09-30 15:20:09,681 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162811)\u001b[0m 2022-09-30 15:20:09,982 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162812)\u001b[0m 2022-09-30 15:20:10,233 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162809)\u001b[0m 2022-09-30 15:20:10,521 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162813)\u001b[0m 2022-09-30 15:20:10,704 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162810)\u001b[0m 2022-09-30 15:20:10,720 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162810)\u001b[0m 2022-09-30 15:20:10,902 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162811)\u001b[0m 2022-09-30 15:20:10,960 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162808)\u001b[0m 2022-09-30 15:20:10,983 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162808)\u001b[0m 2022-09-30 15:20:11,175 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162812)\u001b[0m 2022-09-30 15:20:12,495 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162809)\u001b[0m 2022-09-30 15:20:12,789 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162808)\u001b[0m 2022-09-30 15:22:24,869 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162810)\u001b[0m 2022-09-30 15:22:25,544 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162808)\u001b[0m 2022-09-30 15:22:31,406 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162810)\u001b[0m 2022-09-30 15:22:32,102 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162813)\u001b[0m 2022-09-30 15:22:33,545 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162811)\u001b[0m 2022-09-30 15:22:33,736 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162813)\u001b[0m 2022-09-30 15:22:38,273 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162811)\u001b[0m 2022-09-30 15:22:38,397 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162812)\u001b[0m 2022-09-30 15:22:42,903 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162809)\u001b[0m 2022-09-30 15:22:43,327 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162812)\u001b[0m 2022-09-30 15:22:45,592 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4162809)\u001b[0m 2022-09-30 15:22:45,916 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "no DARs found for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte in CNA_hydrop_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte in CNA_hydrop_1.FIXEDCELLS\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in CNA_hydrop_1.FIXEDCELLS\n",
      "Loading cistopic_objects/CNA_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/CNA_hydrop_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/CNA_hydrop_2.FIXEDCELLS__HVRs.pkl\n",
      "Cytotoxic T cell\n",
      "CD14+ monocyte\n",
      "CD4+ T cell\n",
      "B cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 15:23:03,323\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4163807)\u001b[0m 2022-09-30 15:23:32,275 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163806)\u001b[0m 2022-09-30 15:23:33,041 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163805)\u001b[0m 2022-09-30 15:23:33,885 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163804)\u001b[0m 2022-09-30 15:23:34,701 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163807)\u001b[0m 2022-09-30 15:23:35,690 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163803)\u001b[0m 2022-09-30 15:23:35,648 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163802)\u001b[0m 2022-09-30 15:23:36,500 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163806)\u001b[0m 2022-09-30 15:23:36,552 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163800)\u001b[0m 2022-09-30 15:23:37,438 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163801)\u001b[0m 2022-09-30 15:23:38,160 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163805)\u001b[0m 2022-09-30 15:23:38,411 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163803)\u001b[0m 2022-09-30 15:23:38,587 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163804)\u001b[0m 2022-09-30 15:23:39,223 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163802)\u001b[0m 2022-09-30 15:23:39,312 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163800)\u001b[0m 2022-09-30 15:23:39,481 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163801)\u001b[0m 2022-09-30 15:23:40,131 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163801)\u001b[0m 2022-09-30 15:30:07,152 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163803)\u001b[0m 2022-09-30 15:30:08,439 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163802)\u001b[0m 2022-09-30 15:30:08,881 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163800)\u001b[0m 2022-09-30 15:30:13,381 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163807)\u001b[0m 2022-09-30 15:30:20,111 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163805)\u001b[0m 2022-09-30 15:30:23,000 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163806)\u001b[0m 2022-09-30 15:30:23,540 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163801)\u001b[0m 2022-09-30 15:30:26,175 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163803)\u001b[0m 2022-09-30 15:30:27,762 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163802)\u001b[0m 2022-09-30 15:30:28,224 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163804)\u001b[0m 2022-09-30 15:30:28,369 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163800)\u001b[0m 2022-09-30 15:30:31,228 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163807)\u001b[0m 2022-09-30 15:30:34,841 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163805)\u001b[0m 2022-09-30 15:30:36,495 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163806)\u001b[0m 2022-09-30 15:30:36,871 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4163804)\u001b[0m 2022-09-30 15:30:37,920 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "Loading cistopic_objects/CNA_hydrop_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/CNA_hydrop_3.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/CNA_hydrop_3.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "Cytotoxic T cell\n",
      "CD4+ T cell\n",
      "B cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 15:30:54,491\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4165266)\u001b[0m 2022-09-30 15:31:22,369 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165265)\u001b[0m 2022-09-30 15:31:22,555 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165264)\u001b[0m 2022-09-30 15:31:22,952 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165266)\u001b[0m 2022-09-30 15:31:23,114 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165263)\u001b[0m 2022-09-30 15:31:23,062 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165265)\u001b[0m 2022-09-30 15:31:23,292 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165262)\u001b[0m 2022-09-30 15:31:23,398 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165261)\u001b[0m 2022-09-30 15:31:23,739 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165260)\u001b[0m 2022-09-30 15:31:23,984 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165264)\u001b[0m 2022-09-30 15:31:24,157 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165259)\u001b[0m 2022-09-30 15:31:24,258 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165263)\u001b[0m 2022-09-30 15:31:24,299 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165260)\u001b[0m 2022-09-30 15:31:24,285 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165262)\u001b[0m 2022-09-30 15:31:24,528 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165259)\u001b[0m 2022-09-30 15:31:24,562 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165261)\u001b[0m 2022-09-30 15:31:24,846 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165259)\u001b[0m 2022-09-30 15:34:14,174 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165260)\u001b[0m 2022-09-30 15:34:14,977 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165266)\u001b[0m 2022-09-30 15:34:20,458 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165265)\u001b[0m 2022-09-30 15:34:20,954 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165259)\u001b[0m 2022-09-30 15:34:22,276 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165260)\u001b[0m 2022-09-30 15:34:22,934 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165262)\u001b[0m 2022-09-30 15:34:25,032 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165261)\u001b[0m 2022-09-30 15:34:26,086 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165264)\u001b[0m 2022-09-30 15:34:26,646 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165263)\u001b[0m 2022-09-30 15:34:26,903 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165266)\u001b[0m 2022-09-30 15:34:27,065 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165265)\u001b[0m 2022-09-30 15:34:27,527 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165262)\u001b[0m 2022-09-30 15:34:30,189 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165261)\u001b[0m 2022-09-30 15:34:30,799 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4165264)\u001b[0m 2022-09-30 15:34:31,032 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "no DARs found for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte in CNA_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte in CNA_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in CNA_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell in CNA_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell in CNA_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell in CNA_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleA_B cell_VS_sampleB_B cell in CNA_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in CNA_hydrop_3.FIXEDCELLS\n",
      "Loading cistopic_objects/CNA_mtscatac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/CNA_mtscatac_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/CNA_mtscatac_1.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "B cell\n",
      "CD4+ T cell\n",
      "Cytotoxic T cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 15:34:46,478\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4166070)\u001b[0m 2022-09-30 15:35:15,209 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166068)\u001b[0m 2022-09-30 15:35:15,917 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166070)\u001b[0m 2022-09-30 15:35:16,514 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166069)\u001b[0m 2022-09-30 15:35:16,569 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166068)\u001b[0m 2022-09-30 15:35:17,245 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166067)\u001b[0m 2022-09-30 15:35:17,366 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166069)\u001b[0m 2022-09-30 15:35:17,931 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166065)\u001b[0m 2022-09-30 15:35:18,046 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166067)\u001b[0m 2022-09-30 15:35:18,715 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166066)\u001b[0m 2022-09-30 15:35:18,708 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166063)\u001b[0m 2022-09-30 15:35:19,387 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166064)\u001b[0m 2022-09-30 15:35:20,425 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166065)\u001b[0m 2022-09-30 15:35:22,553 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166066)\u001b[0m 2022-09-30 15:35:23,167 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166063)\u001b[0m 2022-09-30 15:35:23,440 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166064)\u001b[0m 2022-09-30 15:35:24,240 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166069)\u001b[0m 2022-09-30 15:39:30,528 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166068)\u001b[0m 2022-09-30 15:39:30,924 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166070)\u001b[0m 2022-09-30 15:39:31,373 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166067)\u001b[0m 2022-09-30 15:39:33,004 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166069)\u001b[0m 2022-09-30 15:39:42,660 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166068)\u001b[0m 2022-09-30 15:39:43,025 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166070)\u001b[0m 2022-09-30 15:39:43,375 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166067)\u001b[0m 2022-09-30 15:39:44,348 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166064)\u001b[0m 2022-09-30 15:39:51,666 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166063)\u001b[0m 2022-09-30 15:39:51,923 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166066)\u001b[0m 2022-09-30 15:39:52,504 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166065)\u001b[0m 2022-09-30 15:39:54,122 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166064)\u001b[0m 2022-09-30 15:39:58,429 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166063)\u001b[0m 2022-09-30 15:39:58,718 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166066)\u001b[0m 2022-09-30 15:39:59,144 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166065)\u001b[0m 2022-09-30 15:39:59,895 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "no DARs found for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte in CNA_mtscatac_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte in CNA_mtscatac_1.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in CNA_mtscatac_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell in CNA_mtscatac_1.FIXEDCELLS\n",
      "Loading cistopic_objects/CNA_mtscatac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/CNA_mtscatac_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/CNA_mtscatac_2.FIXEDCELLS__HVRs.pkl\n",
      "Dendritic cell\n",
      "CD4+ T cell\n",
      "Cytotoxic T cell\n",
      "B cell\n",
      "CD14+ monocyte\n",
      "Natural killer cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 15:40:12,899\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4167003)\u001b[0m 2022-09-30 15:40:44,673 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167002)\u001b[0m 2022-09-30 15:40:44,896 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167003)\u001b[0m 2022-09-30 15:40:44,988 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167002)\u001b[0m 2022-09-30 15:40:45,263 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167000)\u001b[0m 2022-09-30 15:40:45,292 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167001)\u001b[0m 2022-09-30 15:40:45,950 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166999)\u001b[0m 2022-09-30 15:40:46,261 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166998)\u001b[0m 2022-09-30 15:40:46,649 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167000)\u001b[0m 2022-09-30 15:40:46,963 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166997)\u001b[0m 2022-09-30 15:40:47,031 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166996)\u001b[0m 2022-09-30 15:40:47,485 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167001)\u001b[0m 2022-09-30 15:40:47,608 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166999)\u001b[0m 2022-09-30 15:40:47,604 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166997)\u001b[0m 2022-09-30 15:40:47,656 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166995)\u001b[0m 2022-09-30 15:40:47,892 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166998)\u001b[0m 2022-09-30 15:40:47,980 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166996)\u001b[0m 2022-09-30 15:40:48,056 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166994)\u001b[0m 2022-09-30 15:40:48,273 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166995)\u001b[0m 2022-09-30 15:40:48,616 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166994)\u001b[0m 2022-09-30 15:40:48,964 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167002)\u001b[0m 2022-09-30 15:45:39,712 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167003)\u001b[0m 2022-09-30 15:45:39,975 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166997)\u001b[0m 2022-09-30 15:45:43,748 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166994)\u001b[0m 2022-09-30 15:45:47,283 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166995)\u001b[0m 2022-09-30 15:45:48,642 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166996)\u001b[0m 2022-09-30 15:45:50,400 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166999)\u001b[0m 2022-09-30 15:45:52,211 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167002)\u001b[0m 2022-09-30 15:45:53,869 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167002)\u001b[0m 2022-09-30 15:45:53,999 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167001)\u001b[0m 2022-09-30 15:45:54,051 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167003)\u001b[0m 2022-09-30 15:45:54,314 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167003)\u001b[0m 2022-09-30 15:45:54,479 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167002)\u001b[0m 2022-09-30 15:45:54,450 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166998)\u001b[0m 2022-09-30 15:45:54,811 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167003)\u001b[0m 2022-09-30 15:45:54,908 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167000)\u001b[0m 2022-09-30 15:45:56,105 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166997)\u001b[0m 2022-09-30 15:45:58,193 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166994)\u001b[0m 2022-09-30 15:46:01,187 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166995)\u001b[0m 2022-09-30 15:46:02,265 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166996)\u001b[0m 2022-09-30 15:46:03,895 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166999)\u001b[0m 2022-09-30 15:46:05,333 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167001)\u001b[0m 2022-09-30 15:46:06,308 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4166998)\u001b[0m 2022-09-30 15:46:06,645 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167000)\u001b[0m 2022-09-30 15:46:07,143 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167002)\u001b[0m 2022-09-30 15:47:18,223 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167003)\u001b[0m 2022-09-30 15:47:18,426 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167002)\u001b[0m 2022-09-30 15:47:21,919 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in CNA_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in CNA_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell in CNA_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell in CNA_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in CNA_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell in CNA_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleA_B cell_VS_sampleB_B cell in CNA_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in CNA_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte in CNA_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte in CNA_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleA_Natural killer cell_VS_sampleB_Natural killer cell in CNA_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in CNA_mtscatac_2.FIXEDCELLS\n",
      "Loading cistopic_objects/EPF_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/EPF_hydrop_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/EPF_hydrop_1.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "B cell\n",
      "Cytotoxic T cell\n",
      "CD4+ T cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 15:47:38,058\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4167982)\u001b[0m 2022-09-30 15:48:06,842 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167981)\u001b[0m 2022-09-30 15:48:07,551 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167980)\u001b[0m 2022-09-30 15:48:08,358 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167978)\u001b[0m 2022-09-30 15:48:09,149 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167980)\u001b[0m 2022-09-30 15:48:09,685 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167977)\u001b[0m 2022-09-30 15:48:09,945 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167978)\u001b[0m 2022-09-30 15:48:10,596 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167976)\u001b[0m 2022-09-30 15:48:10,840 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167975)\u001b[0m 2022-09-30 15:48:11,571 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167982)\u001b[0m 2022-09-30 15:48:11,851 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167973)\u001b[0m 2022-09-30 15:48:12,270 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167981)\u001b[0m 2022-09-30 15:48:12,554 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167977)\u001b[0m 2022-09-30 15:48:13,040 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167976)\u001b[0m 2022-09-30 15:48:13,777 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167975)\u001b[0m 2022-09-30 15:48:14,781 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167973)\u001b[0m 2022-09-30 15:48:15,397 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167980)\u001b[0m 2022-09-30 15:52:09,608 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167978)\u001b[0m 2022-09-30 15:52:10,606 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167980)\u001b[0m 2022-09-30 15:52:21,492 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167978)\u001b[0m 2022-09-30 15:52:22,396 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167977)\u001b[0m 2022-09-30 15:52:23,792 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167976)\u001b[0m 2022-09-30 15:52:24,331 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167973)\u001b[0m 2022-09-30 15:52:26,588 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167975)\u001b[0m 2022-09-30 15:52:28,481 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167977)\u001b[0m 2022-09-30 15:52:32,721 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167976)\u001b[0m 2022-09-30 15:52:33,094 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167973)\u001b[0m 2022-09-30 15:52:34,754 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167975)\u001b[0m 2022-09-30 15:52:35,912 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167982)\u001b[0m 2022-09-30 15:52:37,764 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167981)\u001b[0m 2022-09-30 15:52:37,988 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167982)\u001b[0m 2022-09-30 15:52:41,821 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4167981)\u001b[0m 2022-09-30 15:52:41,965 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "no DARs found for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte in EPF_hydrop_1.FIXEDCELLS\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in EPF_hydrop_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell in EPF_hydrop_1.FIXEDCELLS\n",
      "no DARs found for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell in EPF_hydrop_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell in EPF_hydrop_1.FIXEDCELLS\n",
      "Loading cistopic_objects/EPF_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/EPF_hydrop_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/EPF_hydrop_2.FIXEDCELLS__HVRs.pkl\n",
      "Cytotoxic T cell\n",
      "CD4+ T cell\n",
      "CD14+ monocyte\n",
      "Natural killer cell\n",
      "B cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 15:52:59,330\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4168794)\u001b[0m 2022-09-30 15:53:30,947 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168793)\u001b[0m 2022-09-30 15:53:31,731 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168792)\u001b[0m 2022-09-30 15:53:32,580 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168791)\u001b[0m 2022-09-30 15:53:33,330 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168794)\u001b[0m 2022-09-30 15:53:33,497 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168790)\u001b[0m 2022-09-30 15:53:34,033 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168793)\u001b[0m 2022-09-30 15:53:34,152 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168789)\u001b[0m 2022-09-30 15:53:34,819 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168788)\u001b[0m 2022-09-30 15:53:35,495 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168792)\u001b[0m 2022-09-30 15:53:36,175 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168787)\u001b[0m 2022-09-30 15:53:36,214 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168788)\u001b[0m 2022-09-30 15:53:36,652 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168791)\u001b[0m 2022-09-30 15:53:36,873 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168786)\u001b[0m 2022-09-30 15:53:36,963 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168787)\u001b[0m 2022-09-30 15:53:37,435 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168785)\u001b[0m 2022-09-30 15:53:37,745 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168790)\u001b[0m 2022-09-30 15:53:37,799 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168786)\u001b[0m 2022-09-30 15:53:37,997 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168789)\u001b[0m 2022-09-30 15:53:38,489 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168785)\u001b[0m 2022-09-30 15:53:38,699 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168787)\u001b[0m 2022-09-30 15:57:37,463 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168785)\u001b[0m 2022-09-30 15:57:38,365 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168786)\u001b[0m 2022-09-30 15:57:38,817 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168788)\u001b[0m 2022-09-30 15:57:39,446 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168793)\u001b[0m 2022-09-30 15:57:46,902 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168794)\u001b[0m 2022-09-30 15:57:48,765 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168787)\u001b[0m 2022-09-30 15:57:49,023 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168785)\u001b[0m 2022-09-30 15:57:49,918 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168786)\u001b[0m 2022-09-30 15:57:50,241 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168788)\u001b[0m 2022-09-30 15:57:50,822 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168792)\u001b[0m 2022-09-30 15:57:55,596 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168793)\u001b[0m 2022-09-30 15:57:55,788 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168794)\u001b[0m 2022-09-30 15:57:56,894 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168791)\u001b[0m 2022-09-30 15:57:56,951 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168789)\u001b[0m 2022-09-30 15:57:58,929 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168790)\u001b[0m 2022-09-30 15:57:59,282 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168792)\u001b[0m 2022-09-30 15:58:01,203 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168791)\u001b[0m 2022-09-30 15:58:02,196 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4168789)\u001b[0m 2022-09-30 15:58:03,366 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "no DARs found for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell in EPF_hydrop_2.FIXEDCELLS\n",
      "no DARs found for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte in EPF_hydrop_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in EPF_hydrop_2.FIXEDCELLS\n",
      "no DARs found for sampleA_B cell_VS_sampleB_B cell in EPF_hydrop_2.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in EPF_hydrop_2.FIXEDCELLS\n",
      "Loading cistopic_objects/EPF_hydrop_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_5topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/EPF_hydrop_3.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/EPF_hydrop_3.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "CD4+ T cell\n",
      "B cell\n",
      "Dendritic cell\n",
      "CD16+ monocyte\n",
      "Cytotoxic T cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 15:58:20,425\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4169600)\u001b[0m 2022-09-30 15:58:52,882 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169601)\u001b[0m 2022-09-30 15:58:53,854 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169599)\u001b[0m 2022-09-30 15:58:54,819 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169598)\u001b[0m 2022-09-30 15:58:55,645 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169596)\u001b[0m 2022-09-30 15:58:56,693 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169597)\u001b[0m 2022-09-30 15:58:57,477 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169600)\u001b[0m 2022-09-30 15:58:57,772 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169596)\u001b[0m 2022-09-30 15:58:58,043 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169595)\u001b[0m 2022-09-30 15:58:58,360 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169601)\u001b[0m 2022-09-30 15:58:58,585 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169597)\u001b[0m 2022-09-30 15:58:58,780 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169595)\u001b[0m 2022-09-30 15:58:58,983 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169593)\u001b[0m 2022-09-30 15:58:59,207 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169599)\u001b[0m 2022-09-30 15:58:59,505 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169593)\u001b[0m 2022-09-30 15:58:59,858 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169594)\u001b[0m 2022-09-30 15:59:00,077 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169598)\u001b[0m 2022-09-30 15:59:00,381 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169594)\u001b[0m 2022-09-30 15:59:00,617 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169592)\u001b[0m 2022-09-30 15:59:00,853 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169592)\u001b[0m 2022-09-30 15:59:01,418 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169594)\u001b[0m 2022-09-30 16:04:13,644 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169595)\u001b[0m 2022-09-30 16:04:15,480 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169592)\u001b[0m 2022-09-30 16:04:15,821 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169593)\u001b[0m 2022-09-30 16:04:16,275 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169597)\u001b[0m 2022-09-30 16:04:22,911 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169596)\u001b[0m 2022-09-30 16:04:24,546 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169594)\u001b[0m 2022-09-30 16:04:28,811 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169594)\u001b[0m 2022-09-30 16:04:28,962 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169595)\u001b[0m 2022-09-30 16:04:30,671 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169595)\u001b[0m 2022-09-30 16:04:30,774 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169592)\u001b[0m 2022-09-30 16:04:30,897 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169594)\u001b[0m 2022-09-30 16:04:31,031 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169593)\u001b[0m 2022-09-30 16:04:31,525 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169595)\u001b[0m 2022-09-30 16:04:32,600 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169597)\u001b[0m 2022-09-30 16:04:37,089 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169596)\u001b[0m 2022-09-30 16:04:38,171 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169599)\u001b[0m 2022-09-30 16:04:46,018 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169598)\u001b[0m 2022-09-30 16:04:47,533 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169600)\u001b[0m 2022-09-30 16:04:50,308 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169601)\u001b[0m 2022-09-30 16:04:50,617 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169599)\u001b[0m 2022-09-30 16:04:55,920 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169598)\u001b[0m 2022-09-30 16:04:57,221 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169600)\u001b[0m 2022-09-30 16:04:59,240 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169601)\u001b[0m 2022-09-30 16:04:59,379 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169594)\u001b[0m 2022-09-30 16:06:11,159 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169595)\u001b[0m 2022-09-30 16:06:11,474 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4169594)\u001b[0m 2022-09-30 16:06:15,427 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "no DARs found for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte in EPF_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell in EPF_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleA_B cell_VS_sampleB_B cell in EPF_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in EPF_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in EPF_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in EPF_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte in EPF_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte in EPF_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in EPF_hydrop_3.FIXEDCELLS\n",
      "no DARs found for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell in EPF_hydrop_3.FIXEDCELLS\n",
      "Loading cistopic_objects/EPF_hydrop_4.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/EPF_hydrop_4.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/EPF_hydrop_4.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "CD4+ T cell\n",
      "Cytotoxic T cell\n",
      "B cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 16:06:31,908\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4170501)\u001b[0m 2022-09-30 16:07:00,637 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170502)\u001b[0m 2022-09-30 16:07:01,115 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170500)\u001b[0m 2022-09-30 16:07:01,791 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170498)\u001b[0m 2022-09-30 16:07:02,493 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170499)\u001b[0m 2022-09-30 16:07:03,093 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170497)\u001b[0m 2022-09-30 16:07:03,855 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170501)\u001b[0m 2022-09-30 16:07:04,318 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170494)\u001b[0m 2022-09-30 16:07:04,580 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170502)\u001b[0m 2022-09-30 16:07:04,804 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170500)\u001b[0m 2022-09-30 16:07:05,133 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170495)\u001b[0m 2022-09-30 16:07:05,253 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170499)\u001b[0m 2022-09-30 16:07:05,446 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170494)\u001b[0m 2022-09-30 16:07:05,650 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170498)\u001b[0m 2022-09-30 16:07:05,840 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170497)\u001b[0m 2022-09-30 16:07:06,165 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170495)\u001b[0m 2022-09-30 16:07:06,286 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170494)\u001b[0m 2022-09-30 16:11:07,356 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170495)\u001b[0m 2022-09-30 16:11:07,811 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170499)\u001b[0m 2022-09-30 16:11:16,442 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170497)\u001b[0m 2022-09-30 16:11:17,542 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170494)\u001b[0m 2022-09-30 16:11:19,043 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170495)\u001b[0m 2022-09-30 16:11:19,358 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170498)\u001b[0m 2022-09-30 16:11:22,417 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170500)\u001b[0m 2022-09-30 16:11:24,232 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170499)\u001b[0m 2022-09-30 16:11:26,153 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170497)\u001b[0m 2022-09-30 16:11:27,026 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170498)\u001b[0m 2022-09-30 16:11:30,255 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170502)\u001b[0m 2022-09-30 16:11:31,220 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170500)\u001b[0m 2022-09-30 16:11:31,269 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170501)\u001b[0m 2022-09-30 16:11:31,409 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4170502)\u001b[0m 2022-09-30 16:11:35,344 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "no DARs found for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte in EPF_hydrop_4.FIXEDCELLS\n",
      "no DARs found for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell in EPF_hydrop_4.FIXEDCELLS\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in EPF_hydrop_4.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in EPF_hydrop_4.FIXEDCELLS\n",
      "Loading cistopic_objects/HAR_ddseq_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/HAR_ddseq_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/HAR_ddseq_1.FIXEDCELLS__HVRs.pkl\n",
      "CD4+ T cell\n",
      "Natural killer cell\n",
      "Cytotoxic T cell\n",
      "CD14+ monocyte\n",
      "CD16+ monocyte\n",
      "B cell\n",
      "Dendritic cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 16:11:59,390\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4171296)\u001b[0m 2022-09-30 16:12:33,110 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171295)\u001b[0m 2022-09-30 16:12:33,850 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171294)\u001b[0m 2022-09-30 16:12:34,777 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171293)\u001b[0m 2022-09-30 16:12:35,720 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171294)\u001b[0m 2022-09-30 16:12:36,638 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171292)\u001b[0m 2022-09-30 16:12:36,742 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171293)\u001b[0m 2022-09-30 16:12:37,569 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171290)\u001b[0m 2022-09-30 16:12:37,494 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171291)\u001b[0m 2022-09-30 16:12:38,473 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171292)\u001b[0m 2022-09-30 16:12:38,974 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171296)\u001b[0m 2022-09-30 16:12:39,283 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171289)\u001b[0m 2022-09-30 16:12:39,487 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171290)\u001b[0m 2022-09-30 16:12:39,805 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171295)\u001b[0m 2022-09-30 16:12:39,953 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171288)\u001b[0m 2022-09-30 16:12:40,223 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171291)\u001b[0m 2022-09-30 16:12:40,653 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171288)\u001b[0m 2022-09-30 16:12:40,943 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171287)\u001b[0m 2022-09-30 16:12:41,176 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171289)\u001b[0m 2022-09-30 16:12:41,705 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171287)\u001b[0m 2022-09-30 16:12:41,964 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171287)\u001b[0m 2022-09-30 16:16:05,794 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171288)\u001b[0m 2022-09-30 16:16:06,696 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171294)\u001b[0m 2022-09-30 16:16:10,154 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171293)\u001b[0m 2022-09-30 16:16:11,403 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171287)\u001b[0m 2022-09-30 16:16:15,290 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171287)\u001b[0m 2022-09-30 16:16:15,390 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171288)\u001b[0m 2022-09-30 16:16:16,376 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171288)\u001b[0m 2022-09-30 16:16:16,439 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171287)\u001b[0m 2022-09-30 16:16:17,431 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171290)\u001b[0m 2022-09-30 16:16:17,995 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171288)\u001b[0m 2022-09-30 16:16:18,451 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171294)\u001b[0m 2022-09-30 16:16:19,712 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171294)\u001b[0m 2022-09-30 16:16:19,912 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171294)\u001b[0m 2022-09-30 16:16:20,324 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171292)\u001b[0m 2022-09-30 16:16:20,627 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171293)\u001b[0m 2022-09-30 16:16:20,840 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171293)\u001b[0m 2022-09-30 16:16:20,971 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171293)\u001b[0m 2022-09-30 16:16:21,448 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171291)\u001b[0m 2022-09-30 16:16:21,425 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171289)\u001b[0m 2022-09-30 16:16:23,298 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171290)\u001b[0m 2022-09-30 16:16:27,498 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171292)\u001b[0m 2022-09-30 16:16:29,867 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171291)\u001b[0m 2022-09-30 16:16:30,610 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171289)\u001b[0m 2022-09-30 16:16:31,965 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171295)\u001b[0m 2022-09-30 16:16:44,274 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171296)\u001b[0m 2022-09-30 16:16:46,092 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171295)\u001b[0m 2022-09-30 16:16:50,268 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171296)\u001b[0m 2022-09-30 16:16:52,023 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171294)\u001b[0m 2022-09-30 16:17:59,217 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171293)\u001b[0m 2022-09-30 16:18:00,039 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171294)\u001b[0m 2022-09-30 16:18:03,546 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171293)\u001b[0m 2022-09-30 16:18:04,193 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171287)\u001b[0m 2022-09-30 16:18:04,550 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171288)\u001b[0m 2022-09-30 16:18:05,200 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4171287)\u001b[0m 2022-09-30 16:18:07,044 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in HAR_ddseq_1.FIXEDCELLS\n",
      "no DARs found for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte in HAR_ddseq_1.FIXEDCELLS\n",
      "no DARs found for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte in HAR_ddseq_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte in HAR_ddseq_1.FIXEDCELLS\n",
      "no DARs found for sampleA_B cell_VS_sampleB_B cell in HAR_ddseq_1.FIXEDCELLS\n",
      "Loading cistopic_objects/HAR_ddseq_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/HAR_ddseq_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/HAR_ddseq_2.FIXEDCELLS__HVRs.pkl\n",
      "CD4+ T cell\n",
      "B cell\n",
      "CD14+ monocyte\n",
      "Cytotoxic T cell\n",
      "Natural killer cell\n",
      "CD16+ monocyte\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 16:18:31,753\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4172137)\u001b[0m 2022-09-30 16:19:04,642 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172135)\u001b[0m 2022-09-30 16:19:05,655 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172136)\u001b[0m 2022-09-30 16:19:06,608 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172133)\u001b[0m 2022-09-30 16:19:07,539 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172134)\u001b[0m 2022-09-30 16:19:08,585 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172136)\u001b[0m 2022-09-30 16:19:08,871 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172132)\u001b[0m 2022-09-30 16:19:09,453 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172133)\u001b[0m 2022-09-30 16:19:09,742 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172131)\u001b[0m 2022-09-30 16:19:10,343 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172137)\u001b[0m 2022-09-30 16:19:11,215 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172129)\u001b[0m 2022-09-30 16:19:11,337 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172134)\u001b[0m 2022-09-30 16:19:11,560 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172135)\u001b[0m 2022-09-30 16:19:12,006 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172130)\u001b[0m 2022-09-30 16:19:12,243 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172132)\u001b[0m 2022-09-30 16:19:12,463 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172131)\u001b[0m 2022-09-30 16:19:12,861 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172128)\u001b[0m 2022-09-30 16:19:13,195 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172129)\u001b[0m 2022-09-30 16:19:13,943 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172130)\u001b[0m 2022-09-30 16:19:14,032 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172128)\u001b[0m 2022-09-30 16:19:14,989 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172128)\u001b[0m 2022-09-30 16:22:44,899 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172133)\u001b[0m 2022-09-30 16:22:45,767 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172136)\u001b[0m 2022-09-30 16:22:46,105 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172130)\u001b[0m 2022-09-30 16:22:46,129 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172131)\u001b[0m 2022-09-30 16:22:49,427 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172129)\u001b[0m 2022-09-30 16:22:52,755 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172128)\u001b[0m 2022-09-30 16:22:54,156 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172128)\u001b[0m 2022-09-30 16:22:54,267 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172128)\u001b[0m 2022-09-30 16:22:54,730 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172133)\u001b[0m 2022-09-30 16:22:55,218 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172133)\u001b[0m 2022-09-30 16:22:55,340 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172130)\u001b[0m 2022-09-30 16:22:55,369 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172136)\u001b[0m 2022-09-30 16:22:55,589 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172133)\u001b[0m 2022-09-30 16:22:55,783 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172134)\u001b[0m 2022-09-30 16:22:57,086 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172132)\u001b[0m 2022-09-30 16:22:57,985 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172131)\u001b[0m 2022-09-30 16:22:58,201 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172129)\u001b[0m 2022-09-30 16:23:00,699 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172134)\u001b[0m 2022-09-30 16:23:03,593 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172132)\u001b[0m 2022-09-30 16:23:04,374 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172137)\u001b[0m 2022-09-30 16:23:10,205 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172135)\u001b[0m 2022-09-30 16:23:11,253 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172137)\u001b[0m 2022-09-30 16:23:14,550 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172135)\u001b[0m 2022-09-30 16:23:15,528 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172128)\u001b[0m 2022-09-30 16:23:56,338 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172133)\u001b[0m 2022-09-30 16:23:56,923 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172128)\u001b[0m 2022-09-30 16:23:58,774 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172133)\u001b[0m 2022-09-30 16:23:59,173 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "no DARs found for sampleA_Natural killer cell_VS_sampleB_Natural killer cell in HAR_ddseq_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in HAR_ddseq_2.FIXEDCELLS\n",
      "no DARs found for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte in HAR_ddseq_2.FIXEDCELLS\n",
      "no DARs found for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte in HAR_ddseq_2.FIXEDCELLS\n",
      "Loading cistopic_objects/MDC_mtscatac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_19topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/MDC_mtscatac_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/MDC_mtscatac_1.FIXEDCELLS__HVRs.pkl\n",
      "Natural killer cell\n",
      "Cytotoxic T cell\n",
      "CD14+ monocyte\n",
      "CD4+ T cell\n",
      "Dendritic cell\n",
      "B cell\n",
      "CD16+ monocyte\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 16:24:47,184\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4172962)\u001b[0m 2022-09-30 16:25:21,645 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172960)\u001b[0m 2022-09-30 16:25:25,988 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172961)\u001b[0m 2022-09-30 16:25:30,246 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172962)\u001b[0m 2022-09-30 16:25:31,971 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172959)\u001b[0m 2022-09-30 16:25:35,489 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172960)\u001b[0m 2022-09-30 16:25:36,949 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172958)\u001b[0m 2022-09-30 16:25:42,479 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172956)\u001b[0m 2022-09-30 16:25:48,791 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172957)\u001b[0m 2022-09-30 16:25:51,718 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172955)\u001b[0m 2022-09-30 16:25:54,882 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172954)\u001b[0m 2022-09-30 16:25:57,466 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172953)\u001b[0m 2022-09-30 16:26:00,368 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172954)\u001b[0m 2022-09-30 16:26:00,833 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172961)\u001b[0m 2022-09-30 16:26:03,309 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172953)\u001b[0m 2022-09-30 16:26:03,613 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172959)\u001b[0m 2022-09-30 16:26:05,960 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172958)\u001b[0m 2022-09-30 16:26:14,144 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172956)\u001b[0m 2022-09-30 16:26:17,450 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172957)\u001b[0m 2022-09-30 16:26:37,076 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172955)\u001b[0m 2022-09-30 16:26:44,338 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172962)\u001b[0m 2022-09-30 16:32:19,039 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172960)\u001b[0m 2022-09-30 16:32:24,618 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172954)\u001b[0m 2022-09-30 16:32:35,927 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172962)\u001b[0m 2022-09-30 16:32:38,058 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172962)\u001b[0m 2022-09-30 16:32:38,709 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172953)\u001b[0m 2022-09-30 16:32:41,921 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172960)\u001b[0m 2022-09-30 16:32:43,653 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172960)\u001b[0m 2022-09-30 16:32:43,830 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172962)\u001b[0m 2022-09-30 16:32:51,369 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172954)\u001b[0m 2022-09-30 16:32:54,642 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172954)\u001b[0m 2022-09-30 16:32:54,831 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172960)\u001b[0m 2022-09-30 16:32:56,330 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172954)\u001b[0m 2022-09-30 16:32:56,943 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172953)\u001b[0m 2022-09-30 16:33:00,901 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172953)\u001b[0m 2022-09-30 16:33:01,094 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172953)\u001b[0m 2022-09-30 16:33:03,279 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172961)\u001b[0m 2022-09-30 16:33:39,634 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172959)\u001b[0m 2022-09-30 16:33:39,743 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172961)\u001b[0m 2022-09-30 16:33:58,956 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172959)\u001b[0m 2022-09-30 16:33:58,933 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172958)\u001b[0m 2022-09-30 16:34:13,966 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172956)\u001b[0m 2022-09-30 16:34:21,053 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172958)\u001b[0m 2022-09-30 16:34:29,747 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172956)\u001b[0m 2022-09-30 16:34:36,195 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172957)\u001b[0m 2022-09-30 16:34:55,415 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172955)\u001b[0m 2022-09-30 16:34:58,017 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172957)\u001b[0m 2022-09-30 16:35:07,301 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172955)\u001b[0m 2022-09-30 16:35:09,513 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172954)\u001b[0m 2022-09-30 16:36:46,693 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172953)\u001b[0m 2022-09-30 16:36:51,097 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172954)\u001b[0m 2022-09-30 16:36:55,120 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172962)\u001b[0m 2022-09-30 16:36:58,351 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172953)\u001b[0m 2022-09-30 16:36:58,594 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172960)\u001b[0m 2022-09-30 16:37:00,444 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4172962)\u001b[0m 2022-09-30 16:37:03,621 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in MDC_mtscatac_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in MDC_mtscatac_1.FIXEDCELLS\n",
      "Loading cistopic_objects/MDC_mtscatac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/MDC_mtscatac_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/MDC_mtscatac_2.FIXEDCELLS__HVRs.pkl\n",
      "CD4+ T cell\n",
      "Cytotoxic T cell\n",
      "Natural killer cell\n",
      "B cell\n",
      "CD14+ monocyte\n",
      "Dendritic cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 16:37:47,783\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4173990)\u001b[0m 2022-09-30 16:38:23,066 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173989)\u001b[0m 2022-09-30 16:38:27,345 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173987)\u001b[0m 2022-09-30 16:38:31,801 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173988)\u001b[0m 2022-09-30 16:38:34,816 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173984)\u001b[0m 2022-09-30 16:38:37,467 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173986)\u001b[0m 2022-09-30 16:38:39,879 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173984)\u001b[0m 2022-09-30 16:38:42,447 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173985)\u001b[0m 2022-09-30 16:38:43,506 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173986)\u001b[0m 2022-09-30 16:38:44,542 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173987)\u001b[0m 2022-09-30 16:38:44,694 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173983)\u001b[0m 2022-09-30 16:38:46,983 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173988)\u001b[0m 2022-09-30 16:38:47,340 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173985)\u001b[0m 2022-09-30 16:38:50,043 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173982)\u001b[0m 2022-09-30 16:38:50,562 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173981)\u001b[0m 2022-09-30 16:38:53,554 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173983)\u001b[0m 2022-09-30 16:38:53,966 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173990)\u001b[0m 2022-09-30 16:38:55,297 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173989)\u001b[0m 2022-09-30 16:38:57,451 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173982)\u001b[0m 2022-09-30 16:39:04,979 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173981)\u001b[0m 2022-09-30 16:39:07,523 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173986)\u001b[0m 2022-09-30 16:46:02,105 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173984)\u001b[0m 2022-09-30 16:46:02,576 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173985)\u001b[0m 2022-09-30 16:46:17,758 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173983)\u001b[0m 2022-09-30 16:46:20,826 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173986)\u001b[0m 2022-09-30 16:46:23,620 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173986)\u001b[0m 2022-09-30 16:46:23,798 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173984)\u001b[0m 2022-09-30 16:46:24,027 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173984)\u001b[0m 2022-09-30 16:46:25,025 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173986)\u001b[0m 2022-09-30 16:46:25,722 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173984)\u001b[0m 2022-09-30 16:46:26,943 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173988)\u001b[0m 2022-09-30 16:46:39,285 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173985)\u001b[0m 2022-09-30 16:46:39,336 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173987)\u001b[0m 2022-09-30 16:46:39,456 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173983)\u001b[0m 2022-09-30 16:46:42,284 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173987)\u001b[0m 2022-09-30 16:46:57,317 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173988)\u001b[0m 2022-09-30 16:46:57,383 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173981)\u001b[0m 2022-09-30 16:47:15,420 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173982)\u001b[0m 2022-09-30 16:47:18,352 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173981)\u001b[0m 2022-09-30 16:47:29,067 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173982)\u001b[0m 2022-09-30 16:47:31,882 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173989)\u001b[0m 2022-09-30 16:47:33,695 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173990)\u001b[0m 2022-09-30 16:47:35,820 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173989)\u001b[0m 2022-09-30 16:47:43,700 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173990)\u001b[0m 2022-09-30 16:47:45,586 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173986)\u001b[0m 2022-09-30 16:49:03,088 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173984)\u001b[0m 2022-09-30 16:49:04,071 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4173986)\u001b[0m 2022-09-30 16:49:08,440 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "no DARs found for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell in MDC_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in MDC_mtscatac_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in MDC_mtscatac_2.FIXEDCELLS\n",
      "Loading cistopic_objects/OHS_s3atac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/OHS_s3atac_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/OHS_s3atac_1.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "CD4+ T cell\n",
      "Natural killer cell\n",
      "Cytotoxic T cell\n",
      "B cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 16:49:35,459\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4174938)\u001b[0m 2022-09-30 16:50:08,931 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174937)\u001b[0m 2022-09-30 16:50:08,943 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174935)\u001b[0m 2022-09-30 16:50:10,155 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174936)\u001b[0m 2022-09-30 16:50:11,371 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174934)\u001b[0m 2022-09-30 16:50:12,622 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174938)\u001b[0m 2022-09-30 16:50:13,651 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174937)\u001b[0m 2022-09-30 16:50:13,695 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174933)\u001b[0m 2022-09-30 16:50:13,761 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174934)\u001b[0m 2022-09-30 16:50:14,214 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174930)\u001b[0m 2022-09-30 16:50:14,933 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174933)\u001b[0m 2022-09-30 16:50:15,386 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174932)\u001b[0m 2022-09-30 16:50:16,253 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174931)\u001b[0m 2022-09-30 16:50:17,449 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174930)\u001b[0m 2022-09-30 16:50:17,547 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174935)\u001b[0m 2022-09-30 16:50:18,519 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174929)\u001b[0m 2022-09-30 16:50:18,664 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174932)\u001b[0m 2022-09-30 16:50:18,924 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174936)\u001b[0m 2022-09-30 16:50:19,670 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174931)\u001b[0m 2022-09-30 16:50:20,552 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174929)\u001b[0m 2022-09-30 16:50:21,653 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174934)\u001b[0m 2022-09-30 16:56:48,866 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174933)\u001b[0m 2022-09-30 16:56:51,209 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174930)\u001b[0m 2022-09-30 16:57:05,056 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174932)\u001b[0m 2022-09-30 16:57:05,521 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174934)\u001b[0m 2022-09-30 16:57:07,373 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174933)\u001b[0m 2022-09-30 16:57:09,283 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174931)\u001b[0m 2022-09-30 16:57:09,275 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174929)\u001b[0m 2022-09-30 16:57:11,166 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174937)\u001b[0m 2022-09-30 16:57:20,722 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174930)\u001b[0m 2022-09-30 16:57:21,085 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174932)\u001b[0m 2022-09-30 16:57:21,449 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174938)\u001b[0m 2022-09-30 16:57:21,752 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174931)\u001b[0m 2022-09-30 16:57:23,753 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174929)\u001b[0m 2022-09-30 16:57:25,067 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174937)\u001b[0m 2022-09-30 16:57:30,579 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174938)\u001b[0m 2022-09-30 16:57:31,222 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174935)\u001b[0m 2022-09-30 16:57:34,986 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174936)\u001b[0m 2022-09-30 16:57:36,494 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4174935)\u001b[0m 2022-09-30 16:57:40,238 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "no DARs found for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte in OHS_s3atac_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte in OHS_s3atac_1.FIXEDCELLS\n",
      "no DARs found for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell in OHS_s3atac_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell in OHS_s3atac_1.FIXEDCELLS\n",
      "no DARs found for sampleA_Natural killer cell_VS_sampleB_Natural killer cell in OHS_s3atac_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in OHS_s3atac_1.FIXEDCELLS\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in OHS_s3atac_1.FIXEDCELLS\n",
      "no DARs found for sampleA_B cell_VS_sampleB_B cell in OHS_s3atac_1.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in OHS_s3atac_1.FIXEDCELLS\n",
      "Loading cistopic_objects/OHS_s3atac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_7topics.dimreduc.consensus.pkl\n",
      "Loading cistopic_objects/SAN_10xmultiome_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_7topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/SAN_10xmultiome_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/SAN_10xmultiome_1.FIXEDCELLS__HVRs.pkl\n",
      "CD4+ T cell\n",
      "CD14+ monocyte\n",
      "Cytotoxic T cell\n",
      "Natural killer cell\n",
      "B cell\n",
      "CD16+ monocyte\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 16:58:05,533\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4175798)\u001b[0m 2022-09-30 16:58:38,655 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175796)\u001b[0m 2022-09-30 16:58:39,919 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175797)\u001b[0m 2022-09-30 16:58:40,898 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175795)\u001b[0m 2022-09-30 16:58:42,104 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175794)\u001b[0m 2022-09-30 16:58:43,251 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175798)\u001b[0m 2022-09-30 16:58:43,829 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175793)\u001b[0m 2022-09-30 16:58:44,207 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175796)\u001b[0m 2022-09-30 16:58:44,915 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175791)\u001b[0m 2022-09-30 16:58:45,310 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175794)\u001b[0m 2022-09-30 16:58:45,652 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175797)\u001b[0m 2022-09-30 16:58:46,404 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175792)\u001b[0m 2022-09-30 16:58:46,347 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175793)\u001b[0m 2022-09-30 16:58:46,639 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175791)\u001b[0m 2022-09-30 16:58:47,265 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175789)\u001b[0m 2022-09-30 16:58:47,483 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175795)\u001b[0m 2022-09-30 16:58:47,700 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175792)\u001b[0m 2022-09-30 16:58:48,417 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175790)\u001b[0m 2022-09-30 16:58:48,703 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175789)\u001b[0m 2022-09-30 16:58:49,663 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175790)\u001b[0m 2022-09-30 16:58:50,843 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175792)\u001b[0m 2022-09-30 17:04:37,809 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175791)\u001b[0m 2022-09-30 17:04:38,911 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175793)\u001b[0m 2022-09-30 17:04:40,191 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175789)\u001b[0m 2022-09-30 17:04:43,327 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175794)\u001b[0m 2022-09-30 17:04:43,776 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175790)\u001b[0m 2022-09-30 17:04:44,025 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175792)\u001b[0m 2022-09-30 17:04:54,443 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175792)\u001b[0m 2022-09-30 17:04:54,558 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175791)\u001b[0m 2022-09-30 17:04:55,377 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175791)\u001b[0m 2022-09-30 17:04:55,494 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175792)\u001b[0m 2022-09-30 17:04:55,467 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175791)\u001b[0m 2022-09-30 17:04:56,444 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175793)\u001b[0m 2022-09-30 17:04:56,611 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175789)\u001b[0m 2022-09-30 17:04:59,636 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175798)\u001b[0m 2022-09-30 17:04:59,775 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175794)\u001b[0m 2022-09-30 17:05:00,112 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175790)\u001b[0m 2022-09-30 17:05:00,196 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175796)\u001b[0m 2022-09-30 17:05:01,752 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175795)\u001b[0m 2022-09-30 17:05:09,957 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175797)\u001b[0m 2022-09-30 17:05:10,339 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175798)\u001b[0m 2022-09-30 17:05:10,522 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175796)\u001b[0m 2022-09-30 17:05:12,044 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175795)\u001b[0m 2022-09-30 17:05:17,869 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175797)\u001b[0m 2022-09-30 17:05:18,176 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175792)\u001b[0m 2022-09-30 17:06:34,807 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175791)\u001b[0m 2022-09-30 17:06:35,399 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4175792)\u001b[0m 2022-09-30 17:06:39,147 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "no DARs found for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell in SAN_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in SAN_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell in SAN_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleA_Natural killer cell_VS_sampleB_Natural killer cell in SAN_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in SAN_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in SAN_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte in SAN_10xmultiome_1.FIXEDCELLS\n",
      "Loading cistopic_objects/SAN_10xmultiome_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/SAN_10xmultiome_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/SAN_10xmultiome_2.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "Cytotoxic T cell\n",
      "Natural killer cell\n",
      "B cell\n",
      "CD4+ T cell\n",
      "CD16+ monocyte\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 17:07:09,084\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4176731)\u001b[0m 2022-09-30 17:07:43,037 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176730)\u001b[0m 2022-09-30 17:07:44,546 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176729)\u001b[0m 2022-09-30 17:07:45,835 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176728)\u001b[0m 2022-09-30 17:07:47,084 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176727)\u001b[0m 2022-09-30 17:07:48,422 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176729)\u001b[0m 2022-09-30 17:07:49,438 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176726)\u001b[0m 2022-09-30 17:07:49,839 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176731)\u001b[0m 2022-09-30 17:07:50,243 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176728)\u001b[0m 2022-09-30 17:07:50,698 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176727)\u001b[0m 2022-09-30 17:07:51,169 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176725)\u001b[0m 2022-09-30 17:07:51,161 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176730)\u001b[0m 2022-09-30 17:07:51,665 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176724)\u001b[0m 2022-09-30 17:07:52,504 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176726)\u001b[0m 2022-09-30 17:07:52,580 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176723)\u001b[0m 2022-09-30 17:07:54,033 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176725)\u001b[0m 2022-09-30 17:07:54,517 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176722)\u001b[0m 2022-09-30 17:07:55,329 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176724)\u001b[0m 2022-09-30 17:07:55,680 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176723)\u001b[0m 2022-09-30 17:08:00,566 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176722)\u001b[0m 2022-09-30 17:08:01,776 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176726)\u001b[0m 2022-09-30 17:13:55,466 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176727)\u001b[0m 2022-09-30 17:13:55,604 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176729)\u001b[0m 2022-09-30 17:14:00,885 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176728)\u001b[0m 2022-09-30 17:14:01,826 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176724)\u001b[0m 2022-09-30 17:14:04,016 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176725)\u001b[0m 2022-09-30 17:14:05,639 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176726)\u001b[0m 2022-09-30 17:14:12,240 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176727)\u001b[0m 2022-09-30 17:14:12,380 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176726)\u001b[0m 2022-09-30 17:14:12,368 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176727)\u001b[0m 2022-09-30 17:14:12,574 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176726)\u001b[0m 2022-09-30 17:14:13,095 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176727)\u001b[0m 2022-09-30 17:14:13,298 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176729)\u001b[0m 2022-09-30 17:14:17,979 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176728)\u001b[0m 2022-09-30 17:14:19,189 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176724)\u001b[0m 2022-09-30 17:14:20,570 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176725)\u001b[0m 2022-09-30 17:14:21,925 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176723)\u001b[0m 2022-09-30 17:14:28,978 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176722)\u001b[0m 2022-09-30 17:14:29,119 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176731)\u001b[0m 2022-09-30 17:14:32,467 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176730)\u001b[0m 2022-09-30 17:14:34,327 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176723)\u001b[0m 2022-09-30 17:14:39,796 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176722)\u001b[0m 2022-09-30 17:14:39,841 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176731)\u001b[0m 2022-09-30 17:14:42,406 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176730)\u001b[0m 2022-09-30 17:14:43,532 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176726)\u001b[0m 2022-09-30 17:15:59,436 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176727)\u001b[0m 2022-09-30 17:16:00,020 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176726)\u001b[0m 2022-09-30 17:16:03,834 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4176727)\u001b[0m 2022-09-30 17:16:04,223 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in SAN_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in SAN_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte in SAN_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte in SAN_10xmultiome_2.FIXEDCELLS\n",
      "Loading cistopic_objects/STA_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/STA_10xv11_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/STA_10xv11_1.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "Cytotoxic T cell\n",
      "CD16+ monocyte\n",
      "CD4+ T cell\n",
      "Natural killer cell\n",
      "B cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 17:16:21,177\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4177637)\u001b[0m 2022-09-30 17:16:52,271 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177636)\u001b[0m 2022-09-30 17:16:52,690 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177635)\u001b[0m 2022-09-30 17:16:53,050 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177634)\u001b[0m 2022-09-30 17:16:53,498 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177635)\u001b[0m 2022-09-30 17:16:53,731 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177633)\u001b[0m 2022-09-30 17:16:53,923 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177637)\u001b[0m 2022-09-30 17:16:54,208 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177634)\u001b[0m 2022-09-30 17:16:54,191 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177633)\u001b[0m 2022-09-30 17:16:54,417 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177632)\u001b[0m 2022-09-30 17:16:54,368 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177636)\u001b[0m 2022-09-30 17:16:54,682 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177631)\u001b[0m 2022-09-30 17:16:54,717 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177632)\u001b[0m 2022-09-30 17:16:54,900 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177630)\u001b[0m 2022-09-30 17:16:55,212 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177629)\u001b[0m 2022-09-30 17:16:55,535 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177628)\u001b[0m 2022-09-30 17:16:55,998 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177631)\u001b[0m 2022-09-30 17:16:56,121 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177629)\u001b[0m 2022-09-30 17:16:56,126 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177630)\u001b[0m 2022-09-30 17:16:56,569 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177628)\u001b[0m 2022-09-30 17:16:56,589 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177632)\u001b[0m 2022-09-30 17:23:25,464 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177634)\u001b[0m 2022-09-30 17:23:29,016 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177628)\u001b[0m 2022-09-30 17:23:29,105 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177633)\u001b[0m 2022-09-30 17:23:29,626 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177629)\u001b[0m 2022-09-30 17:23:29,977 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177635)\u001b[0m 2022-09-30 17:23:30,586 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177631)\u001b[0m 2022-09-30 17:23:36,178 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177637)\u001b[0m 2022-09-30 17:23:39,820 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177630)\u001b[0m 2022-09-30 17:23:42,671 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177636)\u001b[0m 2022-09-30 17:23:44,131 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177632)\u001b[0m 2022-09-30 17:23:44,235 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177632)\u001b[0m 2022-09-30 17:23:44,370 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177632)\u001b[0m 2022-09-30 17:23:44,793 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177628)\u001b[0m 2022-09-30 17:23:48,017 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177628)\u001b[0m 2022-09-30 17:23:48,159 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177634)\u001b[0m 2022-09-30 17:23:48,225 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177628)\u001b[0m 2022-09-30 17:23:48,531 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177633)\u001b[0m 2022-09-30 17:23:48,605 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177629)\u001b[0m 2022-09-30 17:23:48,752 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177635)\u001b[0m 2022-09-30 17:23:49,443 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177631)\u001b[0m 2022-09-30 17:23:52,713 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177637)\u001b[0m 2022-09-30 17:23:54,933 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177630)\u001b[0m 2022-09-30 17:23:56,212 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177636)\u001b[0m 2022-09-30 17:23:56,801 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177632)\u001b[0m 2022-09-30 17:25:31,097 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177628)\u001b[0m 2022-09-30 17:25:31,602 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4177632)\u001b[0m 2022-09-30 17:25:36,276 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "no DARs found for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte in STA_10xv11_1.FIXEDCELLS\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in STA_10xv11_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell in STA_10xv11_1.FIXEDCELLS\n",
      "no DARs found for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte in STA_10xv11_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte in STA_10xv11_1.FIXEDCELLS\n",
      "no DARs found for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell in STA_10xv11_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell in STA_10xv11_1.FIXEDCELLS\n",
      "no DARs found for sampleA_Natural killer cell_VS_sampleB_Natural killer cell in STA_10xv11_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in STA_10xv11_1.FIXEDCELLS\n",
      "no DARs found for sampleA_B cell_VS_sampleB_B cell in STA_10xv11_1.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in STA_10xv11_1.FIXEDCELLS\n",
      "Loading cistopic_objects/STA_10xv11_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_7topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/STA_10xv11_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/STA_10xv11_2.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "CD4+ T cell\n",
      "B cell\n",
      "Natural killer cell\n",
      "CD16+ monocyte\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 17:25:50,907\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4178551)\u001b[0m 2022-09-30 17:26:21,687 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178550)\u001b[0m 2022-09-30 17:26:22,153 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178549)\u001b[0m 2022-09-30 17:26:23,026 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178548)\u001b[0m 2022-09-30 17:26:23,794 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178547)\u001b[0m 2022-09-30 17:26:24,497 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178546)\u001b[0m 2022-09-30 17:26:25,154 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178549)\u001b[0m 2022-09-30 17:26:25,290 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178547)\u001b[0m 2022-09-30 17:26:25,502 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178545)\u001b[0m 2022-09-30 17:26:25,813 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178548)\u001b[0m 2022-09-30 17:26:25,985 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178551)\u001b[0m 2022-09-30 17:26:26,175 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178546)\u001b[0m 2022-09-30 17:26:26,167 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178544)\u001b[0m 2022-09-30 17:26:26,415 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178550)\u001b[0m 2022-09-30 17:26:26,605 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178545)\u001b[0m 2022-09-30 17:26:27,037 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178543)\u001b[0m 2022-09-30 17:26:27,069 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178544)\u001b[0m 2022-09-30 17:26:27,538 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178543)\u001b[0m 2022-09-30 17:26:27,744 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178542)\u001b[0m 2022-09-30 17:26:27,704 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178542)\u001b[0m 2022-09-30 17:26:28,406 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178543)\u001b[0m 2022-09-30 17:33:16,376 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178542)\u001b[0m 2022-09-30 17:33:18,087 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178547)\u001b[0m 2022-09-30 17:33:20,114 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178545)\u001b[0m 2022-09-30 17:33:22,432 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178546)\u001b[0m 2022-09-30 17:33:24,805 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178544)\u001b[0m 2022-09-30 17:33:25,425 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178548)\u001b[0m 2022-09-30 17:33:27,928 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178549)\u001b[0m 2022-09-30 17:33:29,164 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178543)\u001b[0m 2022-09-30 17:33:35,935 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178542)\u001b[0m 2022-09-30 17:33:37,671 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178547)\u001b[0m 2022-09-30 17:33:39,236 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178545)\u001b[0m 2022-09-30 17:33:40,936 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178546)\u001b[0m 2022-09-30 17:33:42,381 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178544)\u001b[0m 2022-09-30 17:33:42,814 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178548)\u001b[0m 2022-09-30 17:33:44,163 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178549)\u001b[0m 2022-09-30 17:33:44,562 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178550)\u001b[0m 2022-09-30 17:33:45,783 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178551)\u001b[0m 2022-09-30 17:33:47,759 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4178550)\u001b[0m 2022-09-30 17:33:51,007 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "no DARs found for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte in STA_10xv11_2.FIXEDCELLS\n",
      "no DARs found for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell in STA_10xv11_2.FIXEDCELLS\n",
      "no DARs found for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell in STA_10xv11_2.FIXEDCELLS\n",
      "no DARs found for sampleA_B cell_VS_sampleB_B cell in STA_10xv11_2.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in STA_10xv11_2.FIXEDCELLS\n",
      "no DARs found for sampleA_Natural killer cell_VS_sampleB_Natural killer cell in STA_10xv11_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in STA_10xv11_2.FIXEDCELLS\n",
      "no DARs found for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte in STA_10xv11_2.FIXEDCELLS\n",
      "no DARs found for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte in STA_10xv11_2.FIXEDCELLS\n",
      "Loading cistopic_objects/TXG_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_14topics.dimreduc.consensus.pkl\n",
      "Loading cistopic_objects/TXG_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_21topics.dimreduc.consensus.pkl\n",
      "Loading cistopic_objects/TXG_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_21topics.dimreduc.consensus.pkl\n",
      "Loading cistopic_objects/UCS_ddseq_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl\n",
      "Loading cistopic_objects/UCS_ddseq_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl\n",
      "Loading cistopic_objects/VIB_10xmultiome_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/VIB_10xmultiome_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/VIB_10xmultiome_1.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "Cytotoxic T cell\n",
      "Natural killer cell\n",
      "CD4+ T cell\n",
      "B cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 17:34:14,569\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4179393)\u001b[0m 2022-09-30 17:34:45,409 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179392)\u001b[0m 2022-09-30 17:34:45,697 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179390)\u001b[0m 2022-09-30 17:34:46,091 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179391)\u001b[0m 2022-09-30 17:34:46,388 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179393)\u001b[0m 2022-09-30 17:34:46,799 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179389)\u001b[0m 2022-09-30 17:34:46,844 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179390)\u001b[0m 2022-09-30 17:34:46,946 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179392)\u001b[0m 2022-09-30 17:34:47,109 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179391)\u001b[0m 2022-09-30 17:34:47,234 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179388)\u001b[0m 2022-09-30 17:34:47,223 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179389)\u001b[0m 2022-09-30 17:34:47,346 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179386)\u001b[0m 2022-09-30 17:34:47,595 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179388)\u001b[0m 2022-09-30 17:34:47,719 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179387)\u001b[0m 2022-09-30 17:34:48,006 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179385)\u001b[0m 2022-09-30 17:34:48,390 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179384)\u001b[0m 2022-09-30 17:34:48,776 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179385)\u001b[0m 2022-09-30 17:34:49,057 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179386)\u001b[0m 2022-09-30 17:34:49,188 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179387)\u001b[0m 2022-09-30 17:34:49,503 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179384)\u001b[0m 2022-09-30 17:34:49,417 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179388)\u001b[0m 2022-09-30 17:39:09,434 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179389)\u001b[0m 2022-09-30 17:39:10,686 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179391)\u001b[0m 2022-09-30 17:39:16,634 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179384)\u001b[0m 2022-09-30 17:39:16,733 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179390)\u001b[0m 2022-09-30 17:39:16,989 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179385)\u001b[0m 2022-09-30 17:39:18,390 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179393)\u001b[0m 2022-09-30 17:39:21,428 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179388)\u001b[0m 2022-09-30 17:39:22,177 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179392)\u001b[0m 2022-09-30 17:39:23,121 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179389)\u001b[0m 2022-09-30 17:39:23,404 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179387)\u001b[0m 2022-09-30 17:39:25,168 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179386)\u001b[0m 2022-09-30 17:39:26,600 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179384)\u001b[0m 2022-09-30 17:39:28,535 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179391)\u001b[0m 2022-09-30 17:39:28,639 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179390)\u001b[0m 2022-09-30 17:39:28,918 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179385)\u001b[0m 2022-09-30 17:39:29,491 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179393)\u001b[0m 2022-09-30 17:39:30,899 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179392)\u001b[0m 2022-09-30 17:39:31,605 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179387)\u001b[0m 2022-09-30 17:39:32,241 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4179386)\u001b[0m 2022-09-30 17:39:32,569 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "no DARs found for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte in VIB_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in VIB_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleA_Natural killer cell_VS_sampleB_Natural killer cell in VIB_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in VIB_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleA_B cell_VS_sampleB_B cell in VIB_10xmultiome_1.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in VIB_10xmultiome_1.FIXEDCELLS\n",
      "Loading cistopic_objects/VIB_10xmultiome_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/VIB_10xmultiome_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/VIB_10xmultiome_2.FIXEDCELLS__HVRs.pkl\n",
      "CD4+ T cell\n",
      "Dendritic cell\n",
      "CD14+ monocyte\n",
      "Natural killer cell\n",
      "Cytotoxic T cell\n",
      "B cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 17:39:46,930\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4180187)\u001b[0m 2022-09-30 17:40:18,924 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180186)\u001b[0m 2022-09-30 17:40:19,572 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180185)\u001b[0m 2022-09-30 17:40:20,257 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180185)\u001b[0m 2022-09-30 17:40:20,631 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180183)\u001b[0m 2022-09-30 17:40:20,911 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180183)\u001b[0m 2022-09-30 17:40:21,269 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180184)\u001b[0m 2022-09-30 17:40:21,574 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180187)\u001b[0m 2022-09-30 17:40:22,077 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180180)\u001b[0m 2022-09-30 17:40:22,185 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180186)\u001b[0m 2022-09-30 17:40:22,690 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180182)\u001b[0m 2022-09-30 17:40:22,720 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180182)\u001b[0m 2022-09-30 17:40:23,287 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180181)\u001b[0m 2022-09-30 17:40:23,480 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180179)\u001b[0m 2022-09-30 17:40:23,948 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180181)\u001b[0m 2022-09-30 17:40:24,027 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180184)\u001b[0m 2022-09-30 17:40:24,174 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180178)\u001b[0m 2022-09-30 17:40:24,548 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180180)\u001b[0m 2022-09-30 17:40:24,794 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180179)\u001b[0m 2022-09-30 17:40:25,844 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180178)\u001b[0m 2022-09-30 17:40:26,532 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180185)\u001b[0m 2022-09-30 17:45:19,358 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180183)\u001b[0m 2022-09-30 17:45:23,026 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180182)\u001b[0m 2022-09-30 17:45:28,703 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180181)\u001b[0m 2022-09-30 17:45:31,098 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180185)\u001b[0m 2022-09-30 17:45:34,344 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180185)\u001b[0m 2022-09-30 17:45:34,478 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180185)\u001b[0m 2022-09-30 17:45:35,585 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180183)\u001b[0m 2022-09-30 17:45:37,867 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180183)\u001b[0m 2022-09-30 17:45:37,935 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180183)\u001b[0m 2022-09-30 17:45:39,104 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180182)\u001b[0m 2022-09-30 17:45:43,593 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180178)\u001b[0m 2022-09-30 17:45:45,235 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180181)\u001b[0m 2022-09-30 17:45:45,660 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180187)\u001b[0m 2022-09-30 17:45:47,178 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180179)\u001b[0m 2022-09-30 17:45:47,470 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180186)\u001b[0m 2022-09-30 17:45:48,139 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180184)\u001b[0m 2022-09-30 17:45:50,931 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180180)\u001b[0m 2022-09-30 17:45:53,219 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180178)\u001b[0m 2022-09-30 17:45:57,539 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180187)\u001b[0m 2022-09-30 17:45:59,270 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180179)\u001b[0m 2022-09-30 17:45:59,364 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180186)\u001b[0m 2022-09-30 17:46:00,102 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180184)\u001b[0m 2022-09-30 17:46:01,631 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180180)\u001b[0m 2022-09-30 17:46:02,660 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180185)\u001b[0m 2022-09-30 17:47:11,898 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180183)\u001b[0m 2022-09-30 17:47:13,232 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180185)\u001b[0m 2022-09-30 17:47:15,800 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4180183)\u001b[0m 2022-09-30 17:47:16,707 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in VIB_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in VIB_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte in VIB_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleA_Natural killer cell_VS_sampleB_Natural killer cell in VIB_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in VIB_10xmultiome_2.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in VIB_10xmultiome_2.FIXEDCELLS\n",
      "Loading cistopic_objects/VIB_10xv1_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/VIB_10xv1_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/VIB_10xv1_1.FIXEDCELLS__HVRs.pkl\n",
      "Dendritic cell\n",
      "CD16+ monocyte\n",
      "CD4+ T cell\n",
      "Cytotoxic T cell\n",
      "Natural killer cell\n",
      "CD14+ monocyte\n",
      "B cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 17:47:36,823\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4181041)\u001b[0m 2022-09-30 17:48:10,356 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181041)\u001b[0m 2022-09-30 17:48:10,978 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181040)\u001b[0m 2022-09-30 17:48:11,437 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181040)\u001b[0m 2022-09-30 17:48:11,918 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181038)\u001b[0m 2022-09-30 17:48:12,251 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181039)\u001b[0m 2022-09-30 17:48:13,113 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181038)\u001b[0m 2022-09-30 17:48:13,199 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181037)\u001b[0m 2022-09-30 17:48:14,080 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181039)\u001b[0m 2022-09-30 17:48:14,134 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181036)\u001b[0m 2022-09-30 17:48:14,850 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181035)\u001b[0m 2022-09-30 17:48:15,805 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181033)\u001b[0m 2022-09-30 17:48:16,632 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181034)\u001b[0m 2022-09-30 17:48:17,649 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181035)\u001b[0m 2022-09-30 17:48:17,828 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181037)\u001b[0m 2022-09-30 17:48:18,330 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181032)\u001b[0m 2022-09-30 17:48:18,498 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181033)\u001b[0m 2022-09-30 17:48:18,649 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181036)\u001b[0m 2022-09-30 17:48:19,193 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181034)\u001b[0m 2022-09-30 17:48:19,190 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181032)\u001b[0m 2022-09-30 17:48:20,028 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181040)\u001b[0m 2022-09-30 17:53:21,538 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181041)\u001b[0m 2022-09-30 17:53:21,814 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181039)\u001b[0m 2022-09-30 17:53:30,646 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181038)\u001b[0m 2022-09-30 17:53:32,138 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181041)\u001b[0m 2022-09-30 17:53:36,927 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181040)\u001b[0m 2022-09-30 17:53:36,801 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181040)\u001b[0m 2022-09-30 17:53:36,982 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181041)\u001b[0m 2022-09-30 17:53:37,054 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181034)\u001b[0m 2022-09-30 17:53:38,549 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181040)\u001b[0m 2022-09-30 17:53:40,731 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181041)\u001b[0m 2022-09-30 17:53:40,786 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181035)\u001b[0m 2022-09-30 17:53:42,335 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181032)\u001b[0m 2022-09-30 17:53:42,420 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181033)\u001b[0m 2022-09-30 17:53:43,768 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181039)\u001b[0m 2022-09-30 17:53:45,906 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181039)\u001b[0m 2022-09-30 17:53:46,026 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181038)\u001b[0m 2022-09-30 17:53:47,413 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181038)\u001b[0m 2022-09-30 17:53:47,531 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181039)\u001b[0m 2022-09-30 17:53:47,604 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181038)\u001b[0m 2022-09-30 17:53:49,130 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181034)\u001b[0m 2022-09-30 17:53:54,122 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181035)\u001b[0m 2022-09-30 17:53:57,453 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181032)\u001b[0m 2022-09-30 17:53:57,494 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181033)\u001b[0m 2022-09-30 17:53:58,431 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181037)\u001b[0m 2022-09-30 17:54:00,438 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181036)\u001b[0m 2022-09-30 17:54:01,066 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181037)\u001b[0m 2022-09-30 17:54:10,324 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181036)\u001b[0m 2022-09-30 17:54:10,903 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181039)\u001b[0m 2022-09-30 17:56:16,115 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181038)\u001b[0m 2022-09-30 17:56:17,146 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181040)\u001b[0m 2022-09-30 17:56:22,454 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181039)\u001b[0m 2022-09-30 17:56:23,236 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181041)\u001b[0m 2022-09-30 17:56:23,368 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181038)\u001b[0m 2022-09-30 17:56:24,100 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181040)\u001b[0m 2022-09-30 17:56:27,126 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181041)\u001b[0m 2022-09-30 17:56:27,454 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in VIB_10xv1_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in VIB_10xv1_1.FIXEDCELLS\n",
      "no DARs found for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte in VIB_10xv1_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte in VIB_10xv1_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in VIB_10xv1_1.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in VIB_10xv1_1.FIXEDCELLS\n",
      "Loading cistopic_objects/VIB_10xv1_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_19topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/VIB_10xv1_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/VIB_10xv1_2.FIXEDCELLS__HVRs.pkl\n",
      "Cytotoxic T cell\n",
      "CD4+ T cell\n",
      "Natural killer cell\n",
      "CD14+ monocyte\n",
      "CD16+ monocyte\n",
      "B cell\n",
      "Dendritic cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 17:57:10,957\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4181966)\u001b[0m 2022-09-30 17:57:45,381 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181964)\u001b[0m 2022-09-30 17:57:47,448 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181965)\u001b[0m 2022-09-30 17:57:51,061 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181963)\u001b[0m 2022-09-30 17:57:54,945 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181962)\u001b[0m 2022-09-30 17:57:58,982 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181966)\u001b[0m 2022-09-30 17:58:02,445 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181961)\u001b[0m 2022-09-30 17:58:03,182 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181964)\u001b[0m 2022-09-30 17:58:04,826 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181960)\u001b[0m 2022-09-30 17:58:08,009 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181959)\u001b[0m 2022-09-30 17:58:10,610 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181962)\u001b[0m 2022-09-30 17:58:13,074 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181957)\u001b[0m 2022-09-30 17:58:13,415 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181961)\u001b[0m 2022-09-30 17:58:16,230 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181958)\u001b[0m 2022-09-30 17:58:16,306 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181957)\u001b[0m 2022-09-30 17:58:18,508 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181958)\u001b[0m 2022-09-30 17:58:21,407 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181965)\u001b[0m 2022-09-30 17:58:30,070 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181963)\u001b[0m 2022-09-30 17:58:33,848 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181960)\u001b[0m 2022-09-30 17:58:39,546 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181959)\u001b[0m 2022-09-30 17:58:42,096 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181957)\u001b[0m 2022-09-30 18:04:23,245 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181958)\u001b[0m 2022-09-30 18:04:28,248 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181966)\u001b[0m 2022-09-30 18:04:33,720 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181964)\u001b[0m 2022-09-30 18:04:35,258 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181962)\u001b[0m 2022-09-30 18:04:38,539 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181957)\u001b[0m 2022-09-30 18:04:40,495 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181957)\u001b[0m 2022-09-30 18:04:40,830 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181961)\u001b[0m 2022-09-30 18:04:41,203 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181958)\u001b[0m 2022-09-30 18:04:45,381 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181958)\u001b[0m 2022-09-30 18:04:45,556 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181957)\u001b[0m 2022-09-30 18:04:50,967 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181966)\u001b[0m 2022-09-30 18:04:51,178 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181966)\u001b[0m 2022-09-30 18:04:51,489 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181964)\u001b[0m 2022-09-30 18:04:52,754 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181964)\u001b[0m 2022-09-30 18:04:52,990 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181966)\u001b[0m 2022-09-30 18:04:53,395 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181964)\u001b[0m 2022-09-30 18:04:55,016 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181958)\u001b[0m 2022-09-30 18:04:55,529 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181962)\u001b[0m 2022-09-30 18:04:56,541 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181961)\u001b[0m 2022-09-30 18:04:58,620 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181965)\u001b[0m 2022-09-30 18:05:53,190 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181963)\u001b[0m 2022-09-30 18:05:56,632 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181965)\u001b[0m 2022-09-30 18:06:08,032 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181963)\u001b[0m 2022-09-30 18:06:10,775 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181960)\u001b[0m 2022-09-30 18:06:14,887 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181959)\u001b[0m 2022-09-30 18:06:17,022 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181960)\u001b[0m 2022-09-30 18:06:26,235 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181959)\u001b[0m 2022-09-30 18:06:27,993 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181966)\u001b[0m 2022-09-30 18:08:08,866 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181964)\u001b[0m 2022-09-30 18:08:09,360 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181966)\u001b[0m 2022-09-30 18:08:16,657 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181964)\u001b[0m 2022-09-30 18:08:17,049 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181957)\u001b[0m 2022-09-30 18:08:17,955 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181958)\u001b[0m 2022-09-30 18:08:19,444 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4181957)\u001b[0m 2022-09-30 18:08:22,601 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "Loading cistopic_objects/VIB_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/VIB_10xv2_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/VIB_10xv2_1.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "Dendritic cell\n",
      "CD4+ T cell\n",
      "Cytotoxic T cell\n",
      "CD16+ monocyte\n",
      "Natural killer cell\n",
      "B cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 18:08:56,315\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4182982)\u001b[0m 2022-09-30 18:09:30,212 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182979)\u001b[0m 2022-09-30 18:09:31,316 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182981)\u001b[0m 2022-09-30 18:09:32,323 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182981)\u001b[0m 2022-09-30 18:09:32,912 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182978)\u001b[0m 2022-09-30 18:09:33,318 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182978)\u001b[0m 2022-09-30 18:09:33,933 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182982)\u001b[0m 2022-09-30 18:09:34,327 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182976)\u001b[0m 2022-09-30 18:09:34,340 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182979)\u001b[0m 2022-09-30 18:09:35,266 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182977)\u001b[0m 2022-09-30 18:09:35,490 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182974)\u001b[0m 2022-09-30 18:09:36,596 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182973)\u001b[0m 2022-09-30 18:09:37,665 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182975)\u001b[0m 2022-09-30 18:09:38,655 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182974)\u001b[0m 2022-09-30 18:09:38,760 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182972)\u001b[0m 2022-09-30 18:09:39,774 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182973)\u001b[0m 2022-09-30 18:09:39,888 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182975)\u001b[0m 2022-09-30 18:09:40,244 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182976)\u001b[0m 2022-09-30 18:09:40,349 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182972)\u001b[0m 2022-09-30 18:09:41,346 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182977)\u001b[0m 2022-09-30 18:09:41,501 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182981)\u001b[0m 2022-09-30 18:15:16,659 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182978)\u001b[0m 2022-09-30 18:15:17,256 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182972)\u001b[0m 2022-09-30 18:15:32,212 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182981)\u001b[0m 2022-09-30 18:15:33,549 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182981)\u001b[0m 2022-09-30 18:15:33,674 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182978)\u001b[0m 2022-09-30 18:15:34,099 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182978)\u001b[0m 2022-09-30 18:15:34,272 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182975)\u001b[0m 2022-09-30 18:15:35,014 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182973)\u001b[0m 2022-09-30 18:15:35,300 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182981)\u001b[0m 2022-09-30 18:15:35,402 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182974)\u001b[0m 2022-09-30 18:15:35,355 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182978)\u001b[0m 2022-09-30 18:15:36,018 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182972)\u001b[0m 2022-09-30 18:15:48,912 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182972)\u001b[0m 2022-09-30 18:15:49,081 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182972)\u001b[0m 2022-09-30 18:15:51,020 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182975)\u001b[0m 2022-09-30 18:15:51,928 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182973)\u001b[0m 2022-09-30 18:15:52,131 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182975)\u001b[0m 2022-09-30 18:15:52,082 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182974)\u001b[0m 2022-09-30 18:15:52,378 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182975)\u001b[0m 2022-09-30 18:15:53,598 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182982)\u001b[0m 2022-09-30 18:15:53,897 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182979)\u001b[0m 2022-09-30 18:15:55,628 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182976)\u001b[0m 2022-09-30 18:16:04,179 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182977)\u001b[0m 2022-09-30 18:16:05,683 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182982)\u001b[0m 2022-09-30 18:16:07,979 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182979)\u001b[0m 2022-09-30 18:16:09,695 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182976)\u001b[0m 2022-09-30 18:16:16,062 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182977)\u001b[0m 2022-09-30 18:16:17,058 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182981)\u001b[0m 2022-09-30 18:18:29,063 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182978)\u001b[0m 2022-09-30 18:18:29,187 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182972)\u001b[0m 2022-09-30 18:18:35,440 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182981)\u001b[0m 2022-09-30 18:18:36,226 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182978)\u001b[0m 2022-09-30 18:18:36,312 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182975)\u001b[0m 2022-09-30 18:18:37,013 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4182972)\u001b[0m 2022-09-30 18:18:40,213 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in VIB_10xv2_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in VIB_10xv2_1.FIXEDCELLS\n",
      "no DARs found for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte in VIB_10xv2_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in VIB_10xv2_1.FIXEDCELLS\n",
      "Loading cistopic_objects/VIB_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/VIB_10xv2_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/VIB_10xv2_2.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "CD16+ monocyte\n",
      "CD4+ T cell\n",
      "Dendritic cell\n",
      "B cell\n",
      "Cytotoxic T cell\n",
      "Natural killer cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 18:19:06,740\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4183933)\u001b[0m 2022-09-30 18:19:41,296 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183932)\u001b[0m 2022-09-30 18:19:42,339 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183931)\u001b[0m 2022-09-30 18:19:43,467 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183930)\u001b[0m 2022-09-30 18:19:44,476 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183931)\u001b[0m 2022-09-30 18:19:44,897 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183933)\u001b[0m 2022-09-30 18:19:45,503 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183929)\u001b[0m 2022-09-30 18:19:45,717 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183930)\u001b[0m 2022-09-30 18:19:45,909 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183932)\u001b[0m 2022-09-30 18:19:46,510 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183928)\u001b[0m 2022-09-30 18:19:46,736 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183927)\u001b[0m 2022-09-30 18:19:47,841 cisTopic     INFO     Formatting data for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183927)\u001b[0m 2022-09-30 18:19:48,469 cisTopic     INFO     Computing p-value for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183925)\u001b[0m 2022-09-30 18:19:49,080 cisTopic     INFO     Formatting data for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183925)\u001b[0m 2022-09-30 18:19:49,760 cisTopic     INFO     Computing p-value for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183926)\u001b[0m 2022-09-30 18:19:50,293 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183924)\u001b[0m 2022-09-30 18:19:51,317 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183929)\u001b[0m 2022-09-30 18:19:52,413 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183926)\u001b[0m 2022-09-30 18:19:52,437 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183928)\u001b[0m 2022-09-30 18:19:53,367 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183924)\u001b[0m 2022-09-30 18:19:53,313 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183927)\u001b[0m 2022-09-30 18:25:39,220 cisTopic     INFO     Computing log2FC for sampleA_Dendritic cell_VS_sampleB_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183925)\u001b[0m 2022-09-30 18:25:41,702 cisTopic     INFO     Computing log2FC for sampleB_Dendritic cell_VS_sampleA_Dendritic cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183931)\u001b[0m 2022-09-30 18:25:43,165 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183930)\u001b[0m 2022-09-30 18:25:43,968 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183926)\u001b[0m 2022-09-30 18:25:55,022 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183927)\u001b[0m 2022-09-30 18:25:56,199 cisTopic     INFO     sampleA_Dendritic cell_VS_sampleB_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183927)\u001b[0m 2022-09-30 18:25:56,327 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183924)\u001b[0m 2022-09-30 18:25:56,973 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183927)\u001b[0m 2022-09-30 18:25:58,329 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183925)\u001b[0m 2022-09-30 18:25:58,462 cisTopic     INFO     sampleB_Dendritic cell_VS_sampleA_Dendritic cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183925)\u001b[0m 2022-09-30 18:25:58,586 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183931)\u001b[0m 2022-09-30 18:26:00,215 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183931)\u001b[0m 2022-09-30 18:26:00,412 cisTopic     INFO     Formatting data for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183925)\u001b[0m 2022-09-30 18:26:00,695 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183930)\u001b[0m 2022-09-30 18:26:01,059 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183930)\u001b[0m 2022-09-30 18:26:01,239 cisTopic     INFO     Formatting data for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183931)\u001b[0m 2022-09-30 18:26:02,408 cisTopic     INFO     Computing p-value for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183930)\u001b[0m 2022-09-30 18:26:03,164 cisTopic     INFO     Computing p-value for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183932)\u001b[0m 2022-09-30 18:26:10,921 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183926)\u001b[0m 2022-09-30 18:26:12,448 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183933)\u001b[0m 2022-09-30 18:26:13,379 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183924)\u001b[0m 2022-09-30 18:26:14,121 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183932)\u001b[0m 2022-09-30 18:26:25,526 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183929)\u001b[0m 2022-09-30 18:26:26,655 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183928)\u001b[0m 2022-09-30 18:26:26,741 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183933)\u001b[0m 2022-09-30 18:26:27,385 cisTopic     INFO     sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183929)\u001b[0m 2022-09-30 18:26:37,847 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183928)\u001b[0m 2022-09-30 18:26:37,850 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183927)\u001b[0m 2022-09-30 18:28:54,286 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183925)\u001b[0m 2022-09-30 18:28:55,416 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183930)\u001b[0m 2022-09-30 18:28:56,150 cisTopic     INFO     Computing log2FC for sampleB_Natural killer cell_VS_sampleA_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183931)\u001b[0m 2022-09-30 18:28:56,241 cisTopic     INFO     Computing log2FC for sampleA_Natural killer cell_VS_sampleB_Natural killer cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183927)\u001b[0m 2022-09-30 18:29:02,032 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183925)\u001b[0m 2022-09-30 18:29:02,943 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183931)\u001b[0m 2022-09-30 18:29:03,447 cisTopic     INFO     sampleA_Natural killer cell_VS_sampleB_Natural killer cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4183930)\u001b[0m 2022-09-30 18:29:03,381 cisTopic     INFO     sampleB_Natural killer cell_VS_sampleA_Natural killer cell done!\n",
      "no DARs found for sampleA_Dendritic cell_VS_sampleB_Dendritic cell in VIB_10xv2_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Dendritic cell_VS_sampleA_Dendritic cell in VIB_10xv2_2.FIXEDCELLS\n",
      "no DARs found for sampleB_Natural killer cell_VS_sampleA_Natural killer cell in VIB_10xv2_2.FIXEDCELLS\n",
      "Loading cistopic_objects/VIB_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_17topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/VIB_hydrop_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/VIB_hydrop_1.FIXEDCELLS__HVRs.pkl\n",
      "CD14+ monocyte\n",
      "CD4+ T cell\n",
      "CD16+ monocyte\n",
      "Cytotoxic T cell\n",
      "B cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 18:29:20,779\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4184882)\u001b[0m 2022-09-30 18:29:52,090 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184881)\u001b[0m 2022-09-30 18:29:52,861 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184879)\u001b[0m 2022-09-30 18:29:53,651 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184880)\u001b[0m 2022-09-30 18:29:54,302 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184878)\u001b[0m 2022-09-30 18:29:55,085 cisTopic     INFO     Formatting data for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184882)\u001b[0m 2022-09-30 18:29:55,779 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184877)\u001b[0m 2022-09-30 18:29:55,785 cisTopic     INFO     Formatting data for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184878)\u001b[0m 2022-09-30 18:29:55,985 cisTopic     INFO     Computing p-value for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184879)\u001b[0m 2022-09-30 18:29:56,160 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184881)\u001b[0m 2022-09-30 18:29:56,517 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184877)\u001b[0m 2022-09-30 18:29:56,721 cisTopic     INFO     Computing p-value for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184876)\u001b[0m 2022-09-30 18:29:56,642 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184880)\u001b[0m 2022-09-30 18:29:56,775 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184875)\u001b[0m 2022-09-30 18:29:57,250 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184874)\u001b[0m 2022-09-30 18:29:57,899 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184873)\u001b[0m 2022-09-30 18:29:58,622 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184874)\u001b[0m 2022-09-30 18:29:59,056 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184876)\u001b[0m 2022-09-30 18:29:59,315 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184873)\u001b[0m 2022-09-30 18:29:59,794 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184875)\u001b[0m 2022-09-30 18:29:59,931 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184877)\u001b[0m 2022-09-30 18:34:46,997 cisTopic     INFO     Computing log2FC for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184878)\u001b[0m 2022-09-30 18:34:48,729 cisTopic     INFO     Computing log2FC for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184873)\u001b[0m 2022-09-30 18:34:49,834 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184874)\u001b[0m 2022-09-30 18:34:51,624 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184880)\u001b[0m 2022-09-30 18:35:00,161 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184879)\u001b[0m 2022-09-30 18:35:00,676 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184877)\u001b[0m 2022-09-30 18:35:00,721 cisTopic     INFO     sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184878)\u001b[0m 2022-09-30 18:35:02,459 cisTopic     INFO     sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184873)\u001b[0m 2022-09-30 18:35:03,374 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184876)\u001b[0m 2022-09-30 18:35:04,491 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184875)\u001b[0m 2022-09-30 18:35:04,800 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184874)\u001b[0m 2022-09-30 18:35:04,878 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184881)\u001b[0m 2022-09-30 18:35:10,031 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184880)\u001b[0m 2022-09-30 18:35:10,157 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184879)\u001b[0m 2022-09-30 18:35:10,578 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184882)\u001b[0m 2022-09-30 18:35:11,079 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184876)\u001b[0m 2022-09-30 18:35:12,495 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184875)\u001b[0m 2022-09-30 18:35:12,573 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4184881)\u001b[0m 2022-09-30 18:35:15,127 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "no DARs found for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell in VIB_hydrop_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell in VIB_hydrop_1.FIXEDCELLS\n",
      "no DARs found for sampleA_CD16+ monocyte_VS_sampleB_CD16+ monocyte in VIB_hydrop_1.FIXEDCELLS\n",
      "no DARs found for sampleB_CD16+ monocyte_VS_sampleA_CD16+ monocyte in VIB_hydrop_1.FIXEDCELLS\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in VIB_hydrop_1.FIXEDCELLS\n",
      "no DARs found for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell in VIB_hydrop_1.FIXEDCELLS\n",
      "no DARs found for sampleA_B cell_VS_sampleB_B cell in VIB_hydrop_1.FIXEDCELLS\n",
      "no DARs found for sampleB_B cell_VS_sampleA_B cell in VIB_hydrop_1.FIXEDCELLS\n",
      "Loading cistopic_objects/VIB_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_19topics.dimreduc.consensus.pkl\n",
      "Loading downstream_analysis/imputed_acc_objs/VIB_hydrop_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl\n",
      "Loading downstream_analysis/HVRs/VIB_hydrop_2.FIXEDCELLS__HVRs.pkl\n",
      "Cytotoxic T cell\n",
      "CD14+ monocyte\n",
      "B cell\n",
      "CD4+ T cell\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2022-09-30 18:35:32,307\tINFO services.py:1470 -- View the Ray dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265\u001b[39m\u001b[22m\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[2m\u001b[36m(markers_ray pid=4185678)\u001b[0m 2022-09-30 18:36:01,152 cisTopic     INFO     Formatting data for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185677)\u001b[0m 2022-09-30 18:36:01,861 cisTopic     INFO     Formatting data for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185676)\u001b[0m 2022-09-30 18:36:02,687 cisTopic     INFO     Formatting data for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185675)\u001b[0m 2022-09-30 18:36:03,477 cisTopic     INFO     Formatting data for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185673)\u001b[0m 2022-09-30 18:36:04,248 cisTopic     INFO     Formatting data for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185678)\u001b[0m 2022-09-30 18:36:05,112 cisTopic     INFO     Computing p-value for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185674)\u001b[0m 2022-09-30 18:36:05,220 cisTopic     INFO     Formatting data for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185677)\u001b[0m 2022-09-30 18:36:05,836 cisTopic     INFO     Computing p-value for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185673)\u001b[0m 2022-09-30 18:36:05,951 cisTopic     INFO     Computing p-value for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185671)\u001b[0m 2022-09-30 18:36:05,999 cisTopic     INFO     Formatting data for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185672)\u001b[0m 2022-09-30 18:36:06,648 cisTopic     INFO     Formatting data for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185674)\u001b[0m 2022-09-30 18:36:06,853 cisTopic     INFO     Computing p-value for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185676)\u001b[0m 2022-09-30 18:36:07,580 cisTopic     INFO     Computing p-value for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185671)\u001b[0m 2022-09-30 18:36:07,947 cisTopic     INFO     Computing p-value for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185675)\u001b[0m 2022-09-30 18:36:08,423 cisTopic     INFO     Computing p-value for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185672)\u001b[0m 2022-09-30 18:36:08,592 cisTopic     INFO     Computing p-value for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185673)\u001b[0m 2022-09-30 18:40:29,559 cisTopic     INFO     Computing log2FC for sampleA_B cell_VS_sampleB_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185674)\u001b[0m 2022-09-30 18:40:31,033 cisTopic     INFO     Computing log2FC for sampleB_B cell_VS_sampleA_B cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185671)\u001b[0m 2022-09-30 18:40:31,369 cisTopic     INFO     Computing log2FC for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185672)\u001b[0m 2022-09-30 18:40:35,009 cisTopic     INFO     Computing log2FC for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185673)\u001b[0m 2022-09-30 18:40:42,174 cisTopic     INFO     sampleA_B cell_VS_sampleB_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185674)\u001b[0m 2022-09-30 18:40:43,528 cisTopic     INFO     sampleB_B cell_VS_sampleA_B cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185671)\u001b[0m 2022-09-30 18:40:43,696 cisTopic     INFO     sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185672)\u001b[0m 2022-09-30 18:40:46,329 cisTopic     INFO     sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185677)\u001b[0m 2022-09-30 18:40:48,190 cisTopic     INFO     Computing log2FC for sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185678)\u001b[0m 2022-09-30 18:40:48,409 cisTopic     INFO     Computing log2FC for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185677)\u001b[0m 2022-09-30 18:40:55,461 cisTopic     INFO     sampleB_Cytotoxic T cell_VS_sampleA_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185678)\u001b[0m 2022-09-30 18:40:55,483 cisTopic     INFO     sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell done!\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185675)\u001b[0m 2022-09-30 18:40:56,751 cisTopic     INFO     Computing log2FC for sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185676)\u001b[0m 2022-09-30 18:40:57,627 cisTopic     INFO     Computing log2FC for sampleA_CD14+ monocyte_VS_sampleB_CD14+ monocyte\n",
      "\u001b[2m\u001b[36m(markers_ray pid=4185675)\u001b[0m 2022-09-30 18:41:00,886 cisTopic     INFO     sampleB_CD14+ monocyte_VS_sampleA_CD14+ monocyte done!\n",
      "no DARs found for sampleA_Cytotoxic T cell_VS_sampleB_Cytotoxic T cell in VIB_hydrop_2.FIXEDCELLS\n",
      "no DARs found for sampleA_CD4+ T cell_VS_sampleB_CD4+ T cell in VIB_hydrop_2.FIXEDCELLS\n",
      "no DARs found for sampleB_CD4+ T cell_VS_sampleA_CD4+ T cell in VIB_hydrop_2.FIXEDCELLS\n"
     ]
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 12;\n",
       "                var nbb_unformatted_code = \"for sample in cistopic_obj_path_dict.keys():\\n# for sample  in ['SAN_10xmultiome_1.FIXEDCELLS']:\\n    if not os.path.exists(f'male_female_DARs/{sample}'):\\n        cto_path = cistopic_obj_path_dict[sample]\\n        if os.path.isfile(cto_path):\\n            print(f\\\"Loading {cto_path}\\\")\\n            with open(cto_path, 'rb') as f:\\n                cto = pickle.load(f)\\n                \\n            cto.cell_names = cto.cell_data.index\\n            if [x.split('-')[0] for x in cto.selected_model.cell_topic.columns] == [x.split('___')[0] for x in list(cto.cell_names)]:\\n                cto.selected_model.cell_topic.columns =  list(cto.cell_names)\\n                \\n        if 'fmx_sample' in cto.cell_data.columns:\\n            acc_path = normalized_imputed_acc_obj_path_dict[sample]\\n\\n            if os.path.isfile(acc_path):\\n                print(f\\\"Loading {acc_path}\\\")\\n                with open(acc_path, 'rb') as f:\\n                    normalized_imputed_acc_obj = pickle.load(f)\\n\\n\\n            hvr_pkl_path = f'downstream_analysis/HVRs/{sample}__HVRs.pkl'\\n\\n            if os.path.exists(hvr_pkl_path):\\n                print(f\\\"Loading {hvr_pkl_path}\\\")\\n                with open(hvr_pkl_path, 'rb') as f:\\n                    variable_regions = pickle.load(f)\\n            else:\\n                print(\\\"HVRs do not exist!\\\")\\n\\n            cto.cell_data['fmx_sample_consensus_cell_type'] = cto.cell_data['fmx_sample'] + \\\"_\\\" + cto.cell_data[\\\"consensus_cell_type\\\"]\\n\\n            contrasts = []\\n            for cell_type in cto.cell_data['consensus_cell_type'].unique():\\n                print(cell_type)\\n                ct_a = 'sampleA_' + cell_type\\n                ct_b = 'sampleB_' + cell_type\\n                contrast = [[[ct_a], [ct_b]], [[ct_b], [ct_a]]]\\n                contrasts = contrasts + contrast\\n\\n            markers_dict = find_diff_features(\\n                cto,\\n                normalized_imputed_acc_obj,\\n                variable='fmx_sample_consensus_cell_type',\\n                var_features=variable_regions,\\n                contrasts=contrasts,\\n                adjpval_thr=0.05,\\n                log2fc_thr=np.log2(1.2),\\n                n_cpu=10,\\n            )\\n            \\n            if not os.path.exists(f'male_female_DARs/{sample}'):\\n                os.mkdir(f'male_female_DARs/{sample}')\\n                \\n            with open(\\n                f\\\"male_female_DARs/{sample}/{sample}__DARs_dict_1-2xfoldchange.pkl\\\",'wb'\\n            ) as f:\\n                pickle.dump(markers_dict, f, protocol=4)\\n                        \\n            for contrast in markers_dict.keys():\\n                markers = markers_dict[contrast].index.tolist()\\n                df = pd.DataFrame(markers)\\n                if len(df) == 0:\\n                    print(f\\\"no DARs found for {contrast} in {sample}\\\")\\n                else:\\n                    df[[0,1]] = df[0].str.split(':',expand=True)\\n                    df[[1,2]] = df[1].str.split('-',expand=True)\\n                    df[3] = markers_dict[contrast].index.tolist()\\n                    df[3] = contrast.replace(' ', '_')# + '_' + df[3].astype(str)\\n                    score = markers_dict[contrast]['Log2FC']\\n                    score = score.reset_index(drop=True)\\n                    df[4] = score\\n                    df[5] = '.'\\n                    pval = markers_dict[contrast]['Adjusted_pval']\\n                    pval = pval.reset_index(drop=True)\\n                    df[6] = pval\\n                    name = contrast.replace(' ', '_')\\n\\n                    df.to_csv(f'male_female_DARs/{sample}/{sample}__{contrast.replace(\\\" \\\", \\\"_\\\")}__1-2xfoldchange_DARs.bed', sep='\\\\t', header=False, index=False)\";\n",
       "                var nbb_formatted_code = \"for sample in cistopic_obj_path_dict.keys():\\n    # for sample  in ['SAN_10xmultiome_1.FIXEDCELLS']:\\n    if not os.path.exists(f\\\"male_female_DARs/{sample}\\\"):\\n        cto_path = cistopic_obj_path_dict[sample]\\n        if os.path.isfile(cto_path):\\n            print(f\\\"Loading {cto_path}\\\")\\n            with open(cto_path, \\\"rb\\\") as f:\\n                cto = pickle.load(f)\\n\\n            cto.cell_names = cto.cell_data.index\\n            if [x.split(\\\"-\\\")[0] for x in cto.selected_model.cell_topic.columns] == [\\n                x.split(\\\"___\\\")[0] for x in list(cto.cell_names)\\n            ]:\\n                cto.selected_model.cell_topic.columns = list(cto.cell_names)\\n\\n        if \\\"fmx_sample\\\" in cto.cell_data.columns:\\n            acc_path = normalized_imputed_acc_obj_path_dict[sample]\\n\\n            if os.path.isfile(acc_path):\\n                print(f\\\"Loading {acc_path}\\\")\\n                with open(acc_path, \\\"rb\\\") as f:\\n                    normalized_imputed_acc_obj = pickle.load(f)\\n\\n            hvr_pkl_path = f\\\"downstream_analysis/HVRs/{sample}__HVRs.pkl\\\"\\n\\n            if os.path.exists(hvr_pkl_path):\\n                print(f\\\"Loading {hvr_pkl_path}\\\")\\n                with open(hvr_pkl_path, \\\"rb\\\") as f:\\n                    variable_regions = pickle.load(f)\\n            else:\\n                print(\\\"HVRs do not exist!\\\")\\n\\n            cto.cell_data[\\\"fmx_sample_consensus_cell_type\\\"] = (\\n                cto.cell_data[\\\"fmx_sample\\\"] + \\\"_\\\" + cto.cell_data[\\\"consensus_cell_type\\\"]\\n            )\\n\\n            contrasts = []\\n            for cell_type in cto.cell_data[\\\"consensus_cell_type\\\"].unique():\\n                print(cell_type)\\n                ct_a = \\\"sampleA_\\\" + cell_type\\n                ct_b = \\\"sampleB_\\\" + cell_type\\n                contrast = [[[ct_a], [ct_b]], [[ct_b], [ct_a]]]\\n                contrasts = contrasts + contrast\\n\\n            markers_dict = find_diff_features(\\n                cto,\\n                normalized_imputed_acc_obj,\\n                variable=\\\"fmx_sample_consensus_cell_type\\\",\\n                var_features=variable_regions,\\n                contrasts=contrasts,\\n                adjpval_thr=0.05,\\n                log2fc_thr=np.log2(1.2),\\n                n_cpu=10,\\n            )\\n\\n            if not os.path.exists(f\\\"male_female_DARs/{sample}\\\"):\\n                os.mkdir(f\\\"male_female_DARs/{sample}\\\")\\n\\n            with open(\\n                f\\\"male_female_DARs/{sample}/{sample}__DARs_dict_1-2xfoldchange.pkl\\\",\\n                \\\"wb\\\",\\n            ) as f:\\n                pickle.dump(markers_dict, f, protocol=4)\\n\\n            for contrast in markers_dict.keys():\\n                markers = markers_dict[contrast].index.tolist()\\n                df = pd.DataFrame(markers)\\n                if len(df) == 0:\\n                    print(f\\\"no DARs found for {contrast} in {sample}\\\")\\n                else:\\n                    df[[0, 1]] = df[0].str.split(\\\":\\\", expand=True)\\n                    df[[1, 2]] = df[1].str.split(\\\"-\\\", expand=True)\\n                    df[3] = markers_dict[contrast].index.tolist()\\n                    df[3] = contrast.replace(\\\" \\\", \\\"_\\\")  # + '_' + df[3].astype(str)\\n                    score = markers_dict[contrast][\\\"Log2FC\\\"]\\n                    score = score.reset_index(drop=True)\\n                    df[4] = score\\n                    df[5] = \\\".\\\"\\n                    pval = markers_dict[contrast][\\\"Adjusted_pval\\\"]\\n                    pval = pval.reset_index(drop=True)\\n                    df[6] = pval\\n                    name = contrast.replace(\\\" \\\", \\\"_\\\")\\n\\n                    df.to_csv(\\n                        f'male_female_DARs/{sample}/{sample}__{contrast.replace(\\\" \\\", \\\"_\\\")}__1-2xfoldchange_DARs.bed',\\n                        sep=\\\"\\\\t\\\",\\n                        header=False,\\n                        index=False,\\n                    )\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for sample in cistopic_obj_path_dict.keys():\n",
    "# for sample  in ['SAN_10xmultiome_1.FIXEDCELLS']:\n",
    "    if not os.path.exists(f'male_female_DARs/{sample}'):\n",
    "        cto_path = cistopic_obj_path_dict[sample]\n",
    "        if os.path.isfile(cto_path):\n",
    "            print(f\"Loading {cto_path}\")\n",
    "            with open(cto_path, 'rb') as f:\n",
    "                cto = pickle.load(f)\n",
    "                \n",
    "            cto.cell_names = cto.cell_data.index\n",
    "            if [x.split('-')[0] for x in cto.selected_model.cell_topic.columns] == [x.split('___')[0] for x in list(cto.cell_names)]:\n",
    "                cto.selected_model.cell_topic.columns =  list(cto.cell_names)\n",
    "                \n",
    "        if 'fmx_sample' in cto.cell_data.columns:\n",
    "            acc_path = normalized_imputed_acc_obj_path_dict[sample]\n",
    "\n",
    "            if os.path.isfile(acc_path):\n",
    "                print(f\"Loading {acc_path}\")\n",
    "                with open(acc_path, 'rb') as f:\n",
    "                    normalized_imputed_acc_obj = pickle.load(f)\n",
    "\n",
    "\n",
    "            hvr_pkl_path = f'downstream_analysis/HVRs/{sample}__HVRs.pkl'\n",
    "\n",
    "            if os.path.exists(hvr_pkl_path):\n",
    "                print(f\"Loading {hvr_pkl_path}\")\n",
    "                with open(hvr_pkl_path, 'rb') as f:\n",
    "                    variable_regions = pickle.load(f)\n",
    "            else:\n",
    "                print(\"HVRs do not exist!\")\n",
    "\n",
    "            cto.cell_data['fmx_sample_consensus_cell_type'] = cto.cell_data['fmx_sample'] + \"_\" + cto.cell_data[\"consensus_cell_type\"]\n",
    "\n",
    "            contrasts = []\n",
    "            for cell_type in cto.cell_data['consensus_cell_type'].unique():\n",
    "                print(cell_type)\n",
    "                ct_a = 'sampleA_' + cell_type\n",
    "                ct_b = 'sampleB_' + cell_type\n",
    "                contrast = [[[ct_a], [ct_b]], [[ct_b], [ct_a]]]\n",
    "                contrasts = contrasts + contrast\n",
    "\n",
    "            markers_dict = find_diff_features(\n",
    "                cto,\n",
    "                normalized_imputed_acc_obj,\n",
    "                variable='fmx_sample_consensus_cell_type',\n",
    "                var_features=variable_regions,\n",
    "                contrasts=contrasts,\n",
    "                adjpval_thr=0.05,\n",
    "                log2fc_thr=np.log2(1.2),\n",
    "                n_cpu=10,\n",
    "            )\n",
    "            \n",
    "            if not os.path.exists(f'male_female_DARs/{sample}'):\n",
    "                os.mkdir(f'male_female_DARs/{sample}')\n",
    "                \n",
    "            with open(\n",
    "                f\"male_female_DARs/{sample}/{sample}__DARs_dict_1-2xfoldchange.pkl\",'wb'\n",
    "            ) as f:\n",
    "                pickle.dump(markers_dict, f, protocol=4)\n",
    "                        \n",
    "            for contrast in markers_dict.keys():\n",
    "                markers = markers_dict[contrast].index.tolist()\n",
    "                df = pd.DataFrame(markers)\n",
    "                if len(df) == 0:\n",
    "                    print(f\"no DARs found for {contrast} in {sample}\")\n",
    "                else:\n",
    "                    df[[0,1]] = df[0].str.split(':',expand=True)\n",
    "                    df[[1,2]] = df[1].str.split('-',expand=True)\n",
    "                    df[3] = markers_dict[contrast].index.tolist()\n",
    "                    df[3] = contrast.replace(' ', '_')# + '_' + df[3].astype(str)\n",
    "                    score = markers_dict[contrast]['Log2FC']\n",
    "                    score = score.reset_index(drop=True)\n",
    "                    df[4] = score\n",
    "                    df[5] = '.'\n",
    "                    pval = markers_dict[contrast]['Adjusted_pval']\n",
    "                    pval = pval.reset_index(drop=True)\n",
    "                    df[6] = pval\n",
    "                    name = contrast.replace(' ', '_')\n",
    "\n",
    "                    df.to_csv(f'male_female_DARs/{sample}/{sample}__{contrast.replace(\" \", \"_\")}__1-2xfoldchange_DARs.bed', sep='\\t', header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77faff59",
   "metadata": {
    "papermill": {
     "duration": 0.063755,
     "end_time": "2022-09-30T18:41:09.119808",
     "exception": false,
     "start_time": "2022-09-30T18:41:09.056053",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%% md\n"
    },
    "tags": []
   },
   "source": [
    "# visualise dars"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a45596bf",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2022-09-30T18:41:09.250859Z",
     "iopub.status.busy": "2022-09-30T18:41:09.250360Z",
     "iopub.status.idle": "2022-09-30T18:41:09.298573Z",
     "shell.execute_reply": "2022-09-30T18:41:09.298101Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "papermill": {
     "duration": 0.115018,
     "end_time": "2022-09-30T18:41:09.299730",
     "exception": false,
     "start_time": "2022-09-30T18:41:09.184712",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': 'downstream_analysis/DARs/BIO_ddseq_1.FIXEDCELLS/BIO_ddseq_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': 'downstream_analysis/DARs/BIO_ddseq_2.FIXEDCELLS/BIO_ddseq_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': 'downstream_analysis/DARs/BIO_ddseq_3.FIXEDCELLS/BIO_ddseq_3.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': 'downstream_analysis/DARs/BIO_ddseq_4.FIXEDCELLS/BIO_ddseq_4.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': 'downstream_analysis/DARs/BRO_mtscatac_1.FIXEDCELLS/BRO_mtscatac_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': 'downstream_analysis/DARs/BRO_mtscatac_2.FIXEDCELLS/BRO_mtscatac_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': 'downstream_analysis/DARs/CNA_10xmultiome_1.FIXEDCELLS/CNA_10xmultiome_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': 'downstream_analysis/DARs/CNA_10xmultiome_2.FIXEDCELLS/CNA_10xmultiome_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': 'downstream_analysis/DARs/CNA_10xv11_1.FIXEDCELLS/CNA_10xv11_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': 'downstream_analysis/DARs/CNA_10xv11_2.FIXEDCELLS/CNA_10xv11_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': 'downstream_analysis/DARs/CNA_10xv11_3.FIXEDCELLS/CNA_10xv11_3.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': 'downstream_analysis/DARs/CNA_10xv11_4.FIXEDCELLS/CNA_10xv11_4.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': 'downstream_analysis/DARs/CNA_10xv11_5.FIXEDCELLS/CNA_10xv11_5.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': 'downstream_analysis/DARs/CNA_10xv2_1.FIXEDCELLS/CNA_10xv2_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': 'downstream_analysis/DARs/CNA_10xv2_2.FIXEDCELLS/CNA_10xv2_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': 'downstream_analysis/DARs/CNA_hydrop_1.FIXEDCELLS/CNA_hydrop_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': 'downstream_analysis/DARs/CNA_hydrop_2.FIXEDCELLS/CNA_hydrop_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': 'downstream_analysis/DARs/CNA_hydrop_3.FIXEDCELLS/CNA_hydrop_3.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': 'downstream_analysis/DARs/CNA_mtscatac_1.FIXEDCELLS/CNA_mtscatac_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': 'downstream_analysis/DARs/CNA_mtscatac_2.FIXEDCELLS/CNA_mtscatac_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': 'downstream_analysis/DARs/EPF_hydrop_1.FIXEDCELLS/EPF_hydrop_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': 'downstream_analysis/DARs/EPF_hydrop_2.FIXEDCELLS/EPF_hydrop_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': 'downstream_analysis/DARs/EPF_hydrop_3.FIXEDCELLS/EPF_hydrop_3.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': 'downstream_analysis/DARs/EPF_hydrop_4.FIXEDCELLS/EPF_hydrop_4.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': 'downstream_analysis/DARs/HAR_ddseq_1.FIXEDCELLS/HAR_ddseq_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': 'downstream_analysis/DARs/HAR_ddseq_2.FIXEDCELLS/HAR_ddseq_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': 'downstream_analysis/DARs/MDC_mtscatac_1.FIXEDCELLS/MDC_mtscatac_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'MDC_mtscatac_2.FIXEDCELLS': 'downstream_analysis/DARs/MDC_mtscatac_2.FIXEDCELLS/MDC_mtscatac_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'OHS_s3atac_1.FIXEDCELLS': 'downstream_analysis/DARs/OHS_s3atac_1.FIXEDCELLS/OHS_s3atac_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'OHS_s3atac_2.FIXEDCELLS': 'downstream_analysis/DARs/OHS_s3atac_2.FIXEDCELLS/OHS_s3atac_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'SAN_10xmultiome_1.FIXEDCELLS': 'downstream_analysis/DARs/SAN_10xmultiome_1.FIXEDCELLS/SAN_10xmultiome_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'SAN_10xmultiome_2.FIXEDCELLS': 'downstream_analysis/DARs/SAN_10xmultiome_2.FIXEDCELLS/SAN_10xmultiome_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'STA_10xv11_1.FIXEDCELLS': 'downstream_analysis/DARs/STA_10xv11_1.FIXEDCELLS/STA_10xv11_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'STA_10xv11_2.FIXEDCELLS': 'downstream_analysis/DARs/STA_10xv11_2.FIXEDCELLS/STA_10xv11_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'TXG_10xv11_1.FIXEDCELLS': 'downstream_analysis/DARs/TXG_10xv11_1.FIXEDCELLS/TXG_10xv11_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'TXG_10xv2_1.FIXEDCELLS': 'downstream_analysis/DARs/TXG_10xv2_1.FIXEDCELLS/TXG_10xv2_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'TXG_10xv2_2.FIXEDCELLS': 'downstream_analysis/DARs/TXG_10xv2_2.FIXEDCELLS/TXG_10xv2_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'UCS_ddseq_1.FIXEDCELLS': 'downstream_analysis/DARs/UCS_ddseq_1.FIXEDCELLS/UCS_ddseq_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'UCS_ddseq_2.FIXEDCELLS': 'downstream_analysis/DARs/UCS_ddseq_2.FIXEDCELLS/UCS_ddseq_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'VIB_10xmultiome_1.FIXEDCELLS': 'downstream_analysis/DARs/VIB_10xmultiome_1.FIXEDCELLS/VIB_10xmultiome_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'VIB_10xmultiome_2.FIXEDCELLS': 'downstream_analysis/DARs/VIB_10xmultiome_2.FIXEDCELLS/VIB_10xmultiome_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'VIB_10xv1_1.FIXEDCELLS': 'downstream_analysis/DARs/VIB_10xv1_1.FIXEDCELLS/VIB_10xv1_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'VIB_10xv1_2.FIXEDCELLS': 'downstream_analysis/DARs/VIB_10xv1_2.FIXEDCELLS/VIB_10xv1_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'VIB_10xv2_1.FIXEDCELLS': 'downstream_analysis/DARs/VIB_10xv2_1.FIXEDCELLS/VIB_10xv2_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'VIB_10xv2_2.FIXEDCELLS': 'downstream_analysis/DARs/VIB_10xv2_2.FIXEDCELLS/VIB_10xv2_2.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'VIB_hydrop_1.FIXEDCELLS': 'downstream_analysis/DARs/VIB_hydrop_1.FIXEDCELLS/VIB_hydrop_1.FIXEDCELLS__DARs_dict.pkl',\n",
       " 'VIB_hydrop_2.FIXEDCELLS': 'downstream_analysis/DARs/VIB_hydrop_2.FIXEDCELLS/VIB_hydrop_2.FIXEDCELLS__DARs_dict.pkl'}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 14;\n",
       "                var nbb_unformatted_code = \"markers_path_dict = {x.split('/')[-1].split(f'__')[0]:x for x in sorted(glob.glob(f'downstream_analysis/DARs/*/*.pkl'))}\\nmarkers_path_dict\";\n",
       "                var nbb_formatted_code = \"markers_path_dict = {\\n    x.split(\\\"/\\\")[-1].split(f\\\"__\\\")[0]: x\\n    for x in sorted(glob.glob(f\\\"downstream_analysis/DARs/*/*.pkl\\\"))\\n}\\nmarkers_path_dict\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "markers_path_dict = {x.split('/')[-1].split(f'__')[0]:x for x in sorted(glob.glob(f'downstream_analysis/DARs/*/*.pkl'))}\n",
    "markers_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "57ce3fc6",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2022-09-30T18:41:09.432199Z",
     "iopub.status.busy": "2022-09-30T18:41:09.431717Z",
     "iopub.status.idle": "2022-09-30T18:41:09.443638Z",
     "shell.execute_reply": "2022-09-30T18:41:09.443167Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "papermill": {
     "duration": 0.078625,
     "end_time": "2022-09-30T18:41:09.444577",
     "exception": false,
     "start_time": "2022-09-30T18:41:09.365952",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/BIO_ddseq_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/BIO_ddseq_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/BIO_ddseq_3.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/BIO_ddseq_4.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/BRO_mtscatac_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/BRO_mtscatac_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xmultiome_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xmultiome_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xv11_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xv11_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xv11_3.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xv11_4.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xv11_5.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xv2_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_10xv2_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_hydrop_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_hydrop_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_hydrop_3.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_mtscatac_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/CNA_mtscatac_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/EPF_hydrop_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/EPF_hydrop_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/EPF_hydrop_3.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/EPF_hydrop_4.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/HAR_ddseq_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/HAR_ddseq_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/MDC_mtscatac_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'MDC_mtscatac_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/MDC_mtscatac_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'OHS_s3atac_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/OHS_s3atac_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'OHS_s3atac_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/OHS_s3atac_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'SAN_10xmultiome_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/SAN_10xmultiome_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'SAN_10xmultiome_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/SAN_10xmultiome_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'STA_10xv11_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/STA_10xv11_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'STA_10xv11_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/STA_10xv11_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'TXG_10xv11_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/TXG_10xv11_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'TXG_10xv2_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/TXG_10xv2_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'TXG_10xv2_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/TXG_10xv2_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'UCS_ddseq_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/UCS_ddseq_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'UCS_ddseq_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/UCS_ddseq_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_10xmultiome_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_10xmultiome_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_10xmultiome_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_10xmultiome_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_10xv1_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_10xv1_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_10xv1_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_10xv1_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_10xv2_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_10xv2_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_10xv2_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_10xv2_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_hydrop_1.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_hydrop_1.FIXEDCELLS__normalized_imputed_acc_obs.pkl',\n",
       " 'VIB_hydrop_2.FIXEDCELLS': 'downstream_analysis/imputed_acc_objs/VIB_hydrop_2.FIXEDCELLS__normalized_imputed_acc_obs.pkl'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 15;\n",
       "                var nbb_unformatted_code = \"imputed_acc_obj_dict = {x.split('/')[-1].split(f'__')[0]:x for x in sorted(glob.glob(f'downstream_analysis/imputed_acc_objs/*.pkl'))}\\nimputed_acc_obj_dict\";\n",
       "                var nbb_formatted_code = \"imputed_acc_obj_dict = {\\n    x.split(\\\"/\\\")[-1].split(f\\\"__\\\")[0]: x\\n    for x in sorted(glob.glob(f\\\"downstream_analysis/imputed_acc_objs/*.pkl\\\"))\\n}\\nimputed_acc_obj_dict\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "imputed_acc_obj_dict = {x.split('/')[-1].split(f'__')[0]:x for x in sorted(glob.glob(f'downstream_analysis/imputed_acc_objs/*.pkl'))}\n",
    "imputed_acc_obj_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "59c22bf7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T18:41:09.577277Z",
     "iopub.status.busy": "2022-09-30T18:41:09.576866Z",
     "iopub.status.idle": "2022-09-30T18:41:09.623684Z",
     "shell.execute_reply": "2022-09-30T18:41:09.623206Z"
    },
    "papermill": {
     "duration": 0.118727,
     "end_time": "2022-09-30T18:41:09.629739",
     "exception": false,
     "start_time": "2022-09-30T18:41:09.511012",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BIO_ddseq_1.FIXEDCELLS\n",
      "\tplots_qc/BIO_ddseq_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "BIO_ddseq_2.FIXEDCELLS\n",
      "\tplots_qc/BIO_ddseq_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "BIO_ddseq_3.FIXEDCELLS\n",
      "\tplots_qc/BIO_ddseq_3.FIXEDCELLS_DAR_umap.png already exists\n",
      "BIO_ddseq_4.FIXEDCELLS\n",
      "\tplots_qc/BIO_ddseq_4.FIXEDCELLS_DAR_umap.png already exists\n",
      "BRO_mtscatac_1.FIXEDCELLS\n",
      "\tplots_qc/BRO_mtscatac_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "BRO_mtscatac_2.FIXEDCELLS\n",
      "\tplots_qc/BRO_mtscatac_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "CNA_10xmultiome_1.FIXEDCELLS\n",
      "\tplots_qc/CNA_10xmultiome_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "CNA_10xmultiome_2.FIXEDCELLS\n",
      "\tplots_qc/CNA_10xmultiome_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "CNA_10xv11_1.FIXEDCELLS\n",
      "\tplots_qc/CNA_10xv11_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "CNA_10xv11_2.FIXEDCELLS\n",
      "\tplots_qc/CNA_10xv11_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "CNA_10xv11_3.FIXEDCELLS\n",
      "\tplots_qc/CNA_10xv11_3.FIXEDCELLS_DAR_umap.png already exists\n",
      "CNA_10xv11_4.FIXEDCELLS\n",
      "\tplots_qc/CNA_10xv11_4.FIXEDCELLS_DAR_umap.png already exists\n",
      "CNA_10xv11_5.FIXEDCELLS\n",
      "\tplots_qc/CNA_10xv11_5.FIXEDCELLS_DAR_umap.png already exists\n",
      "CNA_10xv2_1.FIXEDCELLS\n",
      "\tplots_qc/CNA_10xv2_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "CNA_10xv2_2.FIXEDCELLS\n",
      "\tplots_qc/CNA_10xv2_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "CNA_hydrop_1.FIXEDCELLS\n",
      "\tplots_qc/CNA_hydrop_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "CNA_hydrop_2.FIXEDCELLS\n",
      "\tplots_qc/CNA_hydrop_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "CNA_hydrop_3.FIXEDCELLS\n",
      "\tplots_qc/CNA_hydrop_3.FIXEDCELLS_DAR_umap.png already exists\n",
      "CNA_mtscatac_1.FIXEDCELLS\n",
      "\tplots_qc/CNA_mtscatac_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "CNA_mtscatac_2.FIXEDCELLS\n",
      "\tplots_qc/CNA_mtscatac_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "EPF_hydrop_1.FIXEDCELLS\n",
      "\tplots_qc/EPF_hydrop_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "EPF_hydrop_2.FIXEDCELLS\n",
      "\tplots_qc/EPF_hydrop_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "EPF_hydrop_3.FIXEDCELLS\n",
      "\tplots_qc/EPF_hydrop_3.FIXEDCELLS_DAR_umap.png already exists\n",
      "EPF_hydrop_4.FIXEDCELLS\n",
      "\tplots_qc/EPF_hydrop_4.FIXEDCELLS_DAR_umap.png already exists\n",
      "HAR_ddseq_1.FIXEDCELLS\n",
      "\tplots_qc/HAR_ddseq_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "HAR_ddseq_2.FIXEDCELLS\n",
      "\tplots_qc/HAR_ddseq_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "MDC_mtscatac_1.FIXEDCELLS\n",
      "\tplots_qc/MDC_mtscatac_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "MDC_mtscatac_2.FIXEDCELLS\n",
      "\tplots_qc/MDC_mtscatac_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "OHS_s3atac_1.FIXEDCELLS\n",
      "\tplots_qc/OHS_s3atac_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "OHS_s3atac_2.FIXEDCELLS\n",
      "\tplots_qc/OHS_s3atac_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "SAN_10xmultiome_1.FIXEDCELLS\n",
      "\tplots_qc/SAN_10xmultiome_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "SAN_10xmultiome_2.FIXEDCELLS\n",
      "\tplots_qc/SAN_10xmultiome_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "STA_10xv11_1.FIXEDCELLS\n",
      "\tplots_qc/STA_10xv11_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "STA_10xv11_2.FIXEDCELLS\n",
      "\tplots_qc/STA_10xv11_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "TXG_10xv11_1.FIXEDCELLS\n",
      "\tplots_qc/TXG_10xv11_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "TXG_10xv2_1.FIXEDCELLS\n",
      "\tplots_qc/TXG_10xv2_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "TXG_10xv2_2.FIXEDCELLS\n",
      "\tplots_qc/TXG_10xv2_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "UCS_ddseq_1.FIXEDCELLS\n",
      "\tplots_qc/UCS_ddseq_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "UCS_ddseq_2.FIXEDCELLS\n",
      "\tplots_qc/UCS_ddseq_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "VIB_10xmultiome_1.FIXEDCELLS\n",
      "\tplots_qc/VIB_10xmultiome_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "VIB_10xmultiome_2.FIXEDCELLS\n",
      "\tplots_qc/VIB_10xmultiome_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "VIB_10xv1_1.FIXEDCELLS\n",
      "\tplots_qc/VIB_10xv1_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "VIB_10xv1_2.FIXEDCELLS\n",
      "\tplots_qc/VIB_10xv1_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "VIB_10xv2_1.FIXEDCELLS\n",
      "\tplots_qc/VIB_10xv2_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "VIB_10xv2_2.FIXEDCELLS\n",
      "\tplots_qc/VIB_10xv2_2.FIXEDCELLS_DAR_umap.png already exists\n",
      "VIB_hydrop_1.FIXEDCELLS\n",
      "\tplots_qc/VIB_hydrop_1.FIXEDCELLS_DAR_umap.png already exists\n",
      "VIB_hydrop_2.FIXEDCELLS\n",
      "\tplots_qc/VIB_hydrop_2.FIXEDCELLS_DAR_umap.png already exists\n"
     ]
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 16;\n",
       "                var nbb_unformatted_code = \"for sample in markers_path_dict.keys():\\n    print(sample)\\n    fig_path = f'plots_qc/{sample}_DAR_umap.png'\\n    if not os.path.exists(fig_path):\\n        with open(cistopic_obj_path_dict[sample], 'rb') as f:\\n            cto = pickle.load(f)\\n\\n        with open(imputed_acc_obj_dict[sample], 'rb') as f:\\n            imputed_acc_obj = pickle.load(f)\\n\\n        with open(markers_path_dict[sample], 'rb') as f:\\n            markers_dict = pickle.load(f)\\n\\n        [ print(x + ': '+ str(len(markers_dict[x]))) for x in markers_dict.keys() ]\\n        plot_imputed_features(\\n            cto,\\n            reduction_name='UMAP',\\n            imputed_data=imputed_acc_obj,\\n            features=[markers_dict[x].index.tolist()[0] for x in markers_dict.keys() if len(markers_dict[x])> 0],\\n            scale=False,\\n            num_columns=3,\\n            selected_cells = cto.projections['cell']['UMAP'].index.tolist(),\\n            save=fig_path\\n        )\\n    else:\\n        print(f\\\"\\\\t{fig_path} already exists\\\")\";\n",
       "                var nbb_formatted_code = \"for sample in markers_path_dict.keys():\\n    print(sample)\\n    fig_path = f\\\"plots_qc/{sample}_DAR_umap.png\\\"\\n    if not os.path.exists(fig_path):\\n        with open(cistopic_obj_path_dict[sample], \\\"rb\\\") as f:\\n            cto = pickle.load(f)\\n\\n        with open(imputed_acc_obj_dict[sample], \\\"rb\\\") as f:\\n            imputed_acc_obj = pickle.load(f)\\n\\n        with open(markers_path_dict[sample], \\\"rb\\\") as f:\\n            markers_dict = pickle.load(f)\\n\\n        [print(x + \\\": \\\" + str(len(markers_dict[x]))) for x in markers_dict.keys()]\\n        plot_imputed_features(\\n            cto,\\n            reduction_name=\\\"UMAP\\\",\\n            imputed_data=imputed_acc_obj,\\n            features=[\\n                markers_dict[x].index.tolist()[0]\\n                for x in markers_dict.keys()\\n                if len(markers_dict[x]) > 0\\n            ],\\n            scale=False,\\n            num_columns=3,\\n            selected_cells=cto.projections[\\\"cell\\\"][\\\"UMAP\\\"].index.tolist(),\\n            save=fig_path,\\n        )\\n    else:\\n        print(f\\\"\\\\t{fig_path} already exists\\\")\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for sample in markers_path_dict.keys():\n",
    "    print(sample)\n",
    "    fig_path = f'plots_qc/{sample}_DAR_umap.png'\n",
    "    if not os.path.exists(fig_path):\n",
    "        with open(cistopic_obj_path_dict[sample], 'rb') as f:\n",
    "            cto = pickle.load(f)\n",
    "\n",
    "        with open(imputed_acc_obj_dict[sample], 'rb') as f:\n",
    "            imputed_acc_obj = pickle.load(f)\n",
    "\n",
    "        with open(markers_path_dict[sample], 'rb') as f:\n",
    "            markers_dict = pickle.load(f)\n",
    "\n",
    "        [ print(x + ': '+ str(len(markers_dict[x]))) for x in markers_dict.keys() ]\n",
    "        plot_imputed_features(\n",
    "            cto,\n",
    "            reduction_name='UMAP',\n",
    "            imputed_data=imputed_acc_obj,\n",
    "            features=[markers_dict[x].index.tolist()[0] for x in markers_dict.keys() if len(markers_dict[x])> 0],\n",
    "            scale=False,\n",
    "            num_columns=3,\n",
    "            selected_cells = cto.projections['cell']['UMAP'].index.tolist(),\n",
    "            save=fig_path\n",
    "        )\n",
    "    else:\n",
    "        print(f\"\\t{fig_path} already exists\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1eeeb89d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-21T16:01:11.019261Z",
     "iopub.status.busy": "2022-11-21T16:01:11.018173Z",
     "iopub.status.idle": "2022-11-21T16:01:11.026085Z",
     "shell.execute_reply": "2022-11-21T16:01:11.025244Z",
     "shell.execute_reply.started": "2022-11-21T16:01:11.019071Z"
    }
   },
   "source": [
    "# Check chromosome distribution of DARs:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1f4b4a21",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-21T16:02:18.639022Z",
     "iopub.status.busy": "2022-11-21T16:02:18.638294Z",
     "iopub.status.idle": "2022-11-21T16:02:18.647139Z",
     "shell.execute_reply": "2022-11-21T16:02:18.646382Z",
     "shell.execute_reply.started": "2022-11-21T16:02:18.638958Z"
    }
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 5;\n",
       "                var nbb_unformatted_code = \"import glob\";\n",
       "                var nbb_formatted_code = \"import glob\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7f36bc45",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-21T16:05:44.538848Z",
     "iopub.status.busy": "2022-11-21T16:05:44.537890Z",
     "iopub.status.idle": "2022-11-21T16:05:44.558311Z",
     "shell.execute_reply": "2022-11-21T16:05:44.557724Z",
     "shell.execute_reply.started": "2022-11-21T16:05:44.538780Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 16;\n",
       "                var nbb_unformatted_code = \"dar_path_dict = {x.split('/')[-1].split('__1-2xfoldchange_DARs.ENCODE3_overlap.bed')[0]: x for x in glob.glob('male_female_DARs/*/*DARs.bed')}\";\n",
       "                var nbb_formatted_code = \"dar_path_dict = {\\n    x.split(\\\"/\\\")[-1].split(\\\"__1-2xfoldchange_DARs.ENCODE3_overlap.bed\\\")[0]: x\\n    for x in glob.glob(\\\"male_female_DARs/*/*DARs.bed\\\")\\n}\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "dar_path_dict = {x.split('/')[-1].split('__1-2xfoldchange_DARs.ENCODE3_overlap.bed')[0]: x for x in glob.glob('male_female_DARs/*/*DARs.bed')}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "92f637fc",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-21T16:06:09.801056Z",
     "iopub.status.busy": "2022-11-21T16:06:09.800055Z",
     "iopub.status.idle": "2022-11-21T16:06:09.819063Z",
     "shell.execute_reply": "2022-11-21T16:06:09.818299Z",
     "shell.execute_reply.started": "2022-11-21T16:06:09.800987Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr9</td>\n",
       "      <td>137591036</td>\n",
       "      <td>137591536</td>\n",
       "      <td>sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell</td>\n",
       "      <td>0.362751</td>\n",
       "      <td>.</td>\n",
       "      <td>0.032575</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>108589182</td>\n",
       "      <td>108589682</td>\n",
       "      <td>sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell</td>\n",
       "      <td>0.325670</td>\n",
       "      <td>.</td>\n",
       "      <td>0.044384</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>36105672</td>\n",
       "      <td>36106172</td>\n",
       "      <td>sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell</td>\n",
       "      <td>0.272849</td>\n",
       "      <td>.</td>\n",
       "      <td>0.009967</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0          1          2                                           3  \\\n",
       "0   chr9  137591036  137591536  sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell   \n",
       "1   chr6  108589182  108589682  sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell   \n",
       "2  chr17   36105672   36106172  sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell   \n",
       "\n",
       "          4  5         6  \n",
       "0  0.362751  .  0.032575  \n",
       "1  0.325670  .  0.044384  \n",
       "2  0.272849  .  0.009967  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 22;\n",
       "                var nbb_unformatted_code = \"peaks_df\";\n",
       "                var nbb_formatted_code = \"peaks_df\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "peaks_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "0de9c384",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-21T16:06:33.149958Z",
     "iopub.status.busy": "2022-11-21T16:06:33.149236Z",
     "iopub.status.idle": "2022-11-21T16:06:33.160829Z",
     "shell.execute_reply": "2022-11-21T16:06:33.160081Z",
     "shell.execute_reply.started": "2022-11-21T16:06:33.149895Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chr17', 'chr6', 'chr9']"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 24;\n",
       "                var nbb_unformatted_code = \"chroms_in_df\";\n",
       "                var nbb_formatted_code = \"chroms_in_df\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "chroms_in_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0d599e24",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-21T16:08:14.880958Z",
     "iopub.status.busy": "2022-11-21T16:08:14.880578Z",
     "iopub.status.idle": "2022-11-21T16:08:14.892359Z",
     "shell.execute_reply": "2022-11-21T16:08:14.891436Z",
     "shell.execute_reply.started": "2022-11-21T16:08:14.880931Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['chr1',\n",
       " 'chr2',\n",
       " 'chr3',\n",
       " 'chr4',\n",
       " 'chr5',\n",
       " 'chr6',\n",
       " 'chr7',\n",
       " 'chr8',\n",
       " 'chr9',\n",
       " 'chr10',\n",
       " 'chr11',\n",
       " 'chr12',\n",
       " 'chr13',\n",
       " 'chr14',\n",
       " 'chr15',\n",
       " 'chr16',\n",
       " 'chr17',\n",
       " 'chr18',\n",
       " 'chr19',\n",
       " 'chr20',\n",
       " 'chr21',\n",
       " 'chr22',\n",
       " 'chrX']"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 30;\n",
       "                var nbb_unformatted_code = \"chroms_standard\";\n",
       "                var nbb_formatted_code = \"chroms_standard\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "chroms_standard"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "05a0a53c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-21T16:09:16.292391Z",
     "iopub.status.busy": "2022-11-21T16:09:16.291410Z",
     "iopub.status.idle": "2022-11-21T16:09:16.302553Z",
     "shell.execute_reply": "2022-11-21T16:09:16.301845Z",
     "shell.execute_reply.started": "2022-11-21T16:09:16.292323Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 35;\n",
       "                var nbb_unformatted_code = \"chroms_standard = [\\\"chr\\\" + str(x + 1) for x in range(22)] + ['chrX']\";\n",
       "                var nbb_formatted_code = \"chroms_standard = [\\\"chr\\\" + str(x + 1) for x in range(22)] + [\\\"chrX\\\"]\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "chroms_standard = [\"chr\" + str(x + 1) for x in range(22)] + ['chrX']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "a5fa28f0",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-21T16:09:18.266347Z",
     "iopub.status.busy": "2022-11-21T16:09:18.265396Z",
     "iopub.status.idle": "2022-11-21T16:09:18.344972Z",
     "shell.execute_reply": "2022-11-21T16:09:18.343602Z",
     "shell.execute_reply.started": "2022-11-21T16:09:18.266319Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr7', 'chr8', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chrX'] not in index\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [36], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mpeaks_df\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m]\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvalue_counts\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m[\u001b[49m\u001b[43mchroms_standard\u001b[49m\u001b[43m]\u001b[49m\n",
      "File \u001b[0;32m/opt/venv/lib/python3.8/site-packages/pandas/core/series.py:1008\u001b[0m, in \u001b[0;36mSeries.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1005\u001b[0m     key \u001b[38;5;241m=\u001b[39m np\u001b[38;5;241m.\u001b[39masarray(key, dtype\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mbool\u001b[39m)\n\u001b[1;32m   1006\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get_values(key)\n\u001b[0;32m-> 1008\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_with\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/venv/lib/python3.8/site-packages/pandas/core/series.py:1048\u001b[0m, in \u001b[0;36mSeries._get_with\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1045\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39miloc[key]\n\u001b[1;32m   1047\u001b[0m \u001b[38;5;66;03m# handle the dup indexing case GH#4246\u001b[39;00m\n\u001b[0;32m-> 1048\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m]\u001b[49m\n",
      "File \u001b[0;32m/opt/venv/lib/python3.8/site-packages/pandas/core/indexing.py:1074\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1071\u001b[0m axis \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maxis \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;241m0\u001b[39m\n\u001b[1;32m   1073\u001b[0m maybe_callable \u001b[38;5;241m=\u001b[39m com\u001b[38;5;241m.\u001b[39mapply_if_callable(key, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj)\n\u001b[0;32m-> 1074\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmaybe_callable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/venv/lib/python3.8/site-packages/pandas/core/indexing.py:1302\u001b[0m, in \u001b[0;36m_LocIndexer._getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1299\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(key, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mndim\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m key\u001b[38;5;241m.\u001b[39mndim \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   1300\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot index with multidimensional key\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m-> 1302\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_iterable\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1304\u001b[0m \u001b[38;5;66;03m# nested tuple slicing\u001b[39;00m\n\u001b[1;32m   1305\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_nested_tuple(key, labels):\n",
      "File \u001b[0;32m/opt/venv/lib/python3.8/site-packages/pandas/core/indexing.py:1240\u001b[0m, in \u001b[0;36m_LocIndexer._getitem_iterable\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1237\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_key(key, axis)\n\u001b[1;32m   1239\u001b[0m \u001b[38;5;66;03m# A collection of keys\u001b[39;00m\n\u001b[0;32m-> 1240\u001b[0m keyarr, indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_listlike_indexer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1241\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_reindex_with_indexers(\n\u001b[1;32m   1242\u001b[0m     {axis: [keyarr, indexer]}, copy\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m, allow_dups\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m\n\u001b[1;32m   1243\u001b[0m )\n",
      "File \u001b[0;32m/opt/venv/lib/python3.8/site-packages/pandas/core/indexing.py:1433\u001b[0m, in \u001b[0;36m_LocIndexer._get_listlike_indexer\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1430\u001b[0m ax \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_get_axis(axis)\n\u001b[1;32m   1431\u001b[0m axis_name \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj\u001b[38;5;241m.\u001b[39m_get_axis_name(axis)\n\u001b[0;32m-> 1433\u001b[0m keyarr, indexer \u001b[38;5;241m=\u001b[39m \u001b[43max\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_indexer_strict\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis_name\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1435\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m keyarr, indexer\n",
      "File \u001b[0;32m/opt/venv/lib/python3.8/site-packages/pandas/core/indexes/base.py:6108\u001b[0m, in \u001b[0;36mIndex._get_indexer_strict\u001b[0;34m(self, key, axis_name)\u001b[0m\n\u001b[1;32m   6105\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   6106\u001b[0m     keyarr, indexer, new_indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reindex_non_unique(keyarr)\n\u001b[0;32m-> 6108\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_raise_if_missing\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkeyarr\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mindexer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis_name\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   6110\u001b[0m keyarr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtake(indexer)\n\u001b[1;32m   6111\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(key, Index):\n\u001b[1;32m   6112\u001b[0m     \u001b[38;5;66;03m# GH 42790 - Preserve name from an Index\u001b[39;00m\n",
      "File \u001b[0;32m/opt/venv/lib/python3.8/site-packages/pandas/core/indexes/base.py:6171\u001b[0m, in \u001b[0;36mIndex._raise_if_missing\u001b[0;34m(self, key, indexer, axis_name)\u001b[0m\n\u001b[1;32m   6168\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mNone of [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m] are in the [\u001b[39m\u001b[38;5;132;01m{\u001b[39;00maxis_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m]\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m   6170\u001b[0m not_found \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(ensure_index(key)[missing_mask\u001b[38;5;241m.\u001b[39mnonzero()[\u001b[38;5;241m0\u001b[39m]]\u001b[38;5;241m.\u001b[39munique())\n\u001b[0;32m-> 6171\u001b[0m \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mnot_found\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m not in index\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr7', 'chr8', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15', 'chr16', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22', 'chrX'] not in index\""
     ]
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 36;\n",
       "                var nbb_unformatted_code = \"peaks_df[0].value_counts()[chroms_standard]\";\n",
       "                var nbb_formatted_code = \"peaks_df[0].value_counts()[chroms_standard]\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "peaks_df[0].value_counts()[chroms_standard]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "0bd83f76",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-21T16:10:43.911374Z",
     "iopub.status.busy": "2022-11-21T16:10:43.910644Z",
     "iopub.status.idle": "2022-11-21T16:10:43.919701Z",
     "shell.execute_reply": "2022-11-21T16:10:43.918987Z",
     "shell.execute_reply.started": "2022-11-21T16:10:43.911311Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chr17', 'chr6', 'chr9'}"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 39;\n",
       "                var nbb_unformatted_code = \"set(chroms_in_df).intersection(set(chroms_standard))\";\n",
       "                var nbb_formatted_code = \"set(chroms_in_df).intersection(set(chroms_standard))\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "bf067b0b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-11-21T16:12:07.257397Z",
     "iopub.status.busy": "2022-11-21T16:12:07.256810Z",
     "iopub.status.idle": "2022-11-21T16:12:09.009986Z",
     "shell.execute_reply": "2022-11-21T16:12:09.009124Z",
     "shell.execute_reply.started": "2022-11-21T16:12:07.257357Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EPF_hydrop_2.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 3\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "EPF_hydrop_2.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 27347\n",
      "\tdars on contigs: 60\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.21892217316743898%\n",
      "\tdars on chrY: 2\n",
      "EPF_hydrop_2.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 7233\n",
      "\tdars on contigs: 6\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.08288437629506838%\n",
      "\tdars on chrY: 6\n",
      "EPF_hydrop_2.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 260\n",
      "\tdars on contigs: 9\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 3.3457249070631967%\n",
      "\tdars on chrY: 1\n",
      "EPF_hydrop_2.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 2\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_3.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 361\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.2762430939226519%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_3.FIXEDCELLS__sampleB_B_cell_VS_sampleA_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 20\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_3.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 630\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.15847860538827258%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_3.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 176\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_3.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 3370\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.02966478789676654%\n",
      "\tdars on chrY: 1\n",
      "CNA_10xv11_3.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 692\n",
      "\tdars on contigs: 4\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.5747126436781609%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_3.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 4164\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.024009603841536616%\n",
      "\tdars on chrY: 1\n",
      "CNA_10xv11_3.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 133\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_3.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 209\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_2.FIXEDCELLS__sampleB_Natural_killer_cell_VS_sampleA_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 15\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_2.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 105\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_2.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 2529\n",
      "\tdars on contigs: 4\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.15791551519936833%\n",
      "\tdars on chrY: 3\n",
      "CNA_10xv11_2.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1243\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_2.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 211\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.4716981132075472%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_2.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1080\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_2.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 8879\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.01126126126126126%\n",
      "\tdars on chrY: 1\n",
      "CNA_10xv11_2.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 419\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_2.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 4\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv2_2.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv2_2.FIXEDCELLS__sampleA_CD16+_monocyte_VS_sampleB_CD16+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 237\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.42016806722689076%\n",
      "\tdars on chrY: 1\n",
      "VIB_10xv2_2.FIXEDCELLS__sampleB_CD16+_monocyte_VS_sampleA_CD16+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 58\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv2_2.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 9\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv2_2.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 47\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv2_2.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 6921\n",
      "\tdars on contigs: 3\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.04332755632582322%\n",
      "\tdars on chrY: 2\n",
      "VIB_10xv2_2.FIXEDCELLS__sampleB_B_cell_VS_sampleA_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 615\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv2_2.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1165\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv2_2.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 5005\n",
      "\tdars on contigs: 2\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.039944078290393446%\n",
      "\tdars on chrY: 2\n",
      "VIB_10xv2_2.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 4522\n",
      "\tdars on contigs: 7\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.1545595054095827%\n",
      "\tdars on chrY: 4\n",
      "VIB_10xv2_2.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 58\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "MDC_mtscatac_2.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 252\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "MDC_mtscatac_2.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 38\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "MDC_mtscatac_2.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 518\n",
      "\tdars on contigs: 3\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.5758157389635317%\n",
      "\tdars on chrY: 1\n",
      "MDC_mtscatac_2.FIXEDCELLS__sampleB_B_cell_VS_sampleA_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1640\n",
      "\tdars on contigs: 56\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 3.30188679245283%\n",
      "\tdars on chrY: 0\n",
      "MDC_mtscatac_2.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 5800\n",
      "\tdars on contigs: 27\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.4633602196670671%\n",
      "\tdars on chrY: 2\n",
      "MDC_mtscatac_2.FIXEDCELLS__sampleB_Natural_killer_cell_VS_sampleA_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 52\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "MDC_mtscatac_2.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 4912\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.02035416242621616%\n",
      "\tdars on chrY: 0\n",
      "MDC_mtscatac_2.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 29349\n",
      "\tdars on contigs: 10\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.03406110562348854%\n",
      "\tdars on chrY: 4\n",
      "MDC_mtscatac_2.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 15223\n",
      "\tdars on contigs: 54\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.35347254042023957%\n",
      "\tdars on chrY: 2\n",
      "VIB_10xv1_1.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 9\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv1_1.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 4083\n",
      "\tdars on contigs: 5\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.1223091976516634%\n",
      "\tdars on chrY: 2\n",
      "VIB_10xv1_1.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 29\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv1_1.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 10\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv1_1.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 82\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv1_1.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 59\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv1_1.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 7366\n",
      "\tdars on contigs: 3\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.04071108698602253%\n",
      "\tdars on chrY: 3\n",
      "VIB_10xv1_1.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1924\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "HAR_ddseq_2.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 338\n",
      "\tdars on contigs: 2\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.5882352941176471%\n",
      "\tdars on chrY: 0\n",
      "HAR_ddseq_2.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 317\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "HAR_ddseq_2.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 8621\n",
      "\tdars on contigs: 10\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.11586142973004288%\n",
      "\tdars on chrY: 3\n",
      "HAR_ddseq_2.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 11639\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.00859106529209622%\n",
      "\tdars on chrY: 1\n",
      "HAR_ddseq_2.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1331\n",
      "\tdars on contigs: 4\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.299625468164794%\n",
      "\tdars on chrY: 2\n",
      "HAR_ddseq_2.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 10\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "HAR_ddseq_2.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 8888\n",
      "\tdars on contigs: 15\n",
      "\tdars on chrM: 20\n",
      "\t% dars non standard chromosomes: 0.392244760730696%\n",
      "\tdars on chrY: 3\n",
      "HAR_ddseq_2.FIXEDCELLS__sampleB_B_cell_VS_sampleA_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 669\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_5.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 303\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.3289473684210526%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_5.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 2\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_5.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 56\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_hydrop_2.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 12771\n",
      "\tdars on contigs: 51\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.39775386055217593%\n",
      "\tdars on chrY: 9\n",
      "CNA_hydrop_2.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 20534\n",
      "\tdars on contigs: 69\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.3349026840751347%\n",
      "\tdars on chrY: 6\n",
      "CNA_hydrop_2.FIXEDCELLS__sampleB_B_cell_VS_sampleA_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 39073\n",
      "\tdars on contigs: 102\n",
      "\tdars on chrM: 1\n",
      "\t% dars non standard chromosomes: 0.2629160710639167%\n",
      "\tdars on chrY: 17\n",
      "CNA_hydrop_2.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 15425\n",
      "\tdars on contigs: 53\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.3424215014859801%\n",
      "\tdars on chrY: 11\n",
      "CNA_hydrop_2.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 4530\n",
      "\tdars on contigs: 19\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.4176742141129918%\n",
      "\tdars on chrY: 3\n",
      "CNA_hydrop_2.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 885\n",
      "\tdars on contigs: 13\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 1.447661469933185%\n",
      "\tdars on chrY: 0\n",
      "CNA_hydrop_2.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 13323\n",
      "\tdars on contigs: 41\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.3067943729422329%\n",
      "\tdars on chrY: 1\n",
      "CNA_hydrop_2.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 26600\n",
      "\tdars on contigs: 76\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.2849002849002849%\n",
      "\tdars on chrY: 15\n",
      "CNA_10xmultiome_1.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 4431\n",
      "\tdars on contigs: 5\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.1127141568981064%\n",
      "\tdars on chrY: 2\n",
      "CNA_10xmultiome_1.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1087\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xmultiome_1.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 5975\n",
      "\tdars on contigs: 4\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0669008195350393%\n",
      "\tdars on chrY: 4\n",
      "CNA_10xmultiome_1.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1347\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xmultiome_1.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 3569\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.028011204481792715%\n",
      "\tdars on chrY: 1\n",
      "CNA_10xmultiome_1.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 3\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_hydrop_2.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1240\n",
      "\tdars on contigs: 2\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.1610305958132045%\n",
      "\tdars on chrY: 2\n",
      "VIB_hydrop_2.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 3209\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_hydrop_2.FIXEDCELLS__sampleB_B_cell_VS_sampleA_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 738\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.13531799729364005%\n",
      "\tdars on chrY: 0\n",
      "VIB_hydrop_2.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 29308\n",
      "\tdars on contigs: 6\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.020468035750835778%\n",
      "\tdars on chrY: 4\n",
      "VIB_hydrop_2.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 9154\n",
      "\tdars on contigs: 3\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.03276182155727859%\n",
      "\tdars on chrY: 2\n",
      "CNA_10xv2_2.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 6962\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.014361625736033319%\n",
      "\tdars on chrY: 1\n",
      "CNA_10xv2_2.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 5\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv2_2.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 328\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv2_2.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 3898\n",
      "\tdars on contigs: 9\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.2303557716918352%\n",
      "\tdars on chrY: 6\n",
      "CNA_10xv2_2.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 4511\n",
      "\tdars on contigs: 2\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.044316419233325945%\n",
      "\tdars on chrY: 2\n",
      "CNA_10xv2_2.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 632\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv2_2.FIXEDCELLS__sampleB_Natural_killer_cell_VS_sampleA_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv2_2.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 7885\n",
      "\tdars on contigs: 6\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.07603599036877455%\n",
      "\tdars on chrY: 4\n",
      "CNA_10xv2_2.FIXEDCELLS__sampleB_B_cell_VS_sampleA_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 4\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv2_2.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 682\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_hydrop_1.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 5522\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_hydrop_1.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 37\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "EPF_hydrop_3.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 2844\n",
      "\tdars on contigs: 25\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.871383757406762%\n",
      "\tdars on chrY: 0\n",
      "EPF_hydrop_3.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 297\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.33557046979865773%\n",
      "\tdars on chrY: 0\n",
      "OHS_s3atac_1.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 6010\n",
      "\tdars on contigs: 8\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.13293452974410103%\n",
      "\tdars on chrY: 0\n",
      "SAN_10xmultiome_2.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1175\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "SAN_10xmultiome_2.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 27204\n",
      "\tdars on contigs: 11\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.040418886643395185%\n",
      "\tdars on chrY: 2\n",
      "SAN_10xmultiome_2.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 659\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "SAN_10xmultiome_2.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 66\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 1.4925373134328357%\n",
      "\tdars on chrY: 0\n",
      "SAN_10xmultiome_2.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 400\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "SAN_10xmultiome_2.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 21\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "SAN_10xmultiome_2.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 5\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "SAN_10xmultiome_2.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 2027\n",
      "\tdars on contigs: 3\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.1477832512315271%\n",
      "\tdars on chrY: 1\n",
      "STA_10xv11_1.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 291\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.3424657534246575%\n",
      "\tdars on chrY: 1\n",
      "EPF_hydrop_4.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 2306\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "EPF_hydrop_4.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 6270\n",
      "\tdars on contigs: 11\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.17513134851138354%\n",
      "\tdars on chrY: 2\n",
      "EPF_hydrop_4.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 46\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "EPF_hydrop_4.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 13539\n",
      "\tdars on contigs: 21\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.15486725663716813%\n",
      "\tdars on chrY: 1\n",
      "CNA_10xv11_1.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 482\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_1.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 14\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_1.FIXEDCELLS__sampleB_B_cell_VS_sampleA_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 836\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_1.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 8438\n",
      "\tdars on contigs: 6\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.07105637138796779%\n",
      "\tdars on chrY: 6\n",
      "CNA_10xv11_1.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 661\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.1510574018126888%\n",
      "\tdars on chrY: 1\n",
      "CNA_10xv11_1.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 3361\n",
      "\tdars on contigs: 6\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.1782001782001782%\n",
      "\tdars on chrY: 4\n",
      "CNA_10xv11_1.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 58\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_1.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 4115\n",
      "\tdars on contigs: 2\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.048579062424095217%\n",
      "\tdars on chrY: 2\n",
      "CNA_10xv11_1.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 15\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_mtscatac_1.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 35\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_mtscatac_1.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 7143\n",
      "\tdars on contigs: 4\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.05596753882748006%\n",
      "\tdars on chrY: 3\n",
      "CNA_mtscatac_1.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_mtscatac_1.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 27\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_hydrop_1.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 19352\n",
      "\tdars on contigs: 96\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.49362402303578773%\n",
      "\tdars on chrY: 13\n",
      "CNA_hydrop_1.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 7837\n",
      "\tdars on contigs: 24\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.3053046686172243%\n",
      "\tdars on chrY: 6\n",
      "CNA_hydrop_1.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 21089\n",
      "\tdars on contigs: 83\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.3920272057434347%\n",
      "\tdars on chrY: 8\n",
      "VIB_10xmultiome_1.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 736\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xmultiome_1.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 45\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xmultiome_1.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 3562\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.028066236317709797%\n",
      "\tdars on chrY: 1\n",
      "VIB_10xmultiome_1.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 29\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_1.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 2011\n",
      "\tdars on contigs: 7\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.3468780971258672%\n",
      "\tdars on chrY: 3\n",
      "BRO_mtscatac_1.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 234\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_1.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 425\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_1.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 3713\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_1.FIXEDCELLS__sampleB_CD16+_monocyte_VS_sampleA_CD16+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 192\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_1.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 43\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_1.FIXEDCELLS__sampleB_B_cell_VS_sampleA_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1543\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_1.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 2\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_1.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1233\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_1.FIXEDCELLS__sampleA_CD16+_monocyte_VS_sampleB_CD16+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1166\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv11_4.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 9574\n",
      "\tdars on contigs: 9\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.09391631013252635%\n",
      "\tdars on chrY: 4\n",
      "MDC_mtscatac_1.FIXEDCELLS__sampleA_CD16+_monocyte_VS_sampleB_CD16+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "MDC_mtscatac_1.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 2474\n",
      "\tdars on contigs: 19\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.762133975130365%\n",
      "\tdars on chrY: 0\n",
      "MDC_mtscatac_1.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 665\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.15015015015015015%\n",
      "\tdars on chrY: 0\n",
      "MDC_mtscatac_1.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1832\n",
      "\tdars on contigs: 6\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.3264417845484222%\n",
      "\tdars on chrY: 5\n",
      "MDC_mtscatac_1.FIXEDCELLS__sampleB_CD16+_monocyte_VS_sampleA_CD16+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 2\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "MDC_mtscatac_1.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 8140\n",
      "\tdars on contigs: 31\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.37939052747521723%\n",
      "\tdars on chrY: 0\n",
      "MDC_mtscatac_1.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 811\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.12315270935960591%\n",
      "\tdars on chrY: 1\n",
      "MDC_mtscatac_1.FIXEDCELLS__sampleB_B_cell_VS_sampleA_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 3495\n",
      "\tdars on contigs: 25\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.7102272727272727%\n",
      "\tdars on chrY: 1\n",
      "MDC_mtscatac_1.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 506\n",
      "\tdars on contigs: 3\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.5893909626719057%\n",
      "\tdars on chrY: 3\n",
      "MDC_mtscatac_1.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1487\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "MDC_mtscatac_1.FIXEDCELLS__sampleB_Natural_killer_cell_VS_sampleA_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 89\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "MDC_mtscatac_1.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 5170\n",
      "\tdars on contigs: 13\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.25081998842369285%\n",
      "\tdars on chrY: 10\n",
      "VIB_10xmultiome_2.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 251\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xmultiome_2.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 5037\n",
      "\tdars on contigs: 4\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0793493354493156%\n",
      "\tdars on chrY: 4\n",
      "VIB_10xmultiome_2.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 2869\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.03484320557491289%\n",
      "\tdars on chrY: 1\n",
      "VIB_10xmultiome_2.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 69\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xmultiome_2.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 4229\n",
      "\tdars on contigs: 5\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.11809163911195086%\n",
      "\tdars on chrY: 5\n",
      "VIB_10xmultiome_2.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 568\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_2.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 906\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.11025358324145534%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_2.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 145\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.684931506849315%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_2.FIXEDCELLS__sampleB_CD16+_monocyte_VS_sampleA_CD16+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 205\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_2.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 530\n",
      "\tdars on contigs: 2\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.37593984962406013%\n",
      "\tdars on chrY: 2\n",
      "BRO_mtscatac_2.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 73\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_2.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 73\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_2.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 85\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_2.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 49\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 2.0%\n",
      "\tdars on chrY: 1\n",
      "BRO_mtscatac_2.FIXEDCELLS__sampleB_B_cell_VS_sampleA_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 310\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "BRO_mtscatac_2.FIXEDCELLS__sampleA_CD16+_monocyte_VS_sampleB_CD16+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv2_1.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 45\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv2_1.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 10\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv2_1.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 2455\n",
      "\tdars on contigs: 3\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.12205044751830757%\n",
      "\tdars on chrY: 2\n",
      "VIB_10xv2_1.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 874\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv2_1.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 91\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv2_1.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 484\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv2_1.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 5562\n",
      "\tdars on contigs: 2\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.03594536304816679%\n",
      "\tdars on chrY: 2\n",
      "VIB_10xv2_1.FIXEDCELLS__sampleB_CD16+_monocyte_VS_sampleA_CD16+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 28\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv2_1.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 603\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv2_1.FIXEDCELLS__sampleB_B_cell_VS_sampleA_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 46\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv1_2.FIXEDCELLS__sampleB_B_cell_VS_sampleA_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1213\n",
      "\tdars on contigs: 2\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.1646090534979424%\n",
      "\tdars on chrY: 1\n",
      "VIB_10xv1_2.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1360\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv1_2.FIXEDCELLS__sampleB_Dendritic_cell_VS_sampleA_Dendritic_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 153\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv1_2.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 6234\n",
      "\tdars on contigs: 6\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.09615384615384616%\n",
      "\tdars on chrY: 1\n",
      "VIB_10xv1_2.FIXEDCELLS__sampleA_Dendritic_cell_VS_sampleB_Dendritic_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 316\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv1_2.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 7216\n",
      "\tdars on contigs: 7\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.09691264017721168%\n",
      "\tdars on chrY: 4\n",
      "VIB_10xv1_2.FIXEDCELLS__sampleB_Natural_killer_cell_VS_sampleA_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 280\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv1_2.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 207\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv1_2.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 8927\n",
      "\tdars on contigs: 3\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0335946248600224%\n",
      "\tdars on chrY: 1\n",
      "VIB_10xv1_2.FIXEDCELLS__sampleB_CD16+_monocyte_VS_sampleA_CD16+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 20\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv1_2.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 3091\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.03234152652005175%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv1_2.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 9601\n",
      "\tdars on contigs: 5\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.052050801582344375%\n",
      "\tdars on chrY: 2\n",
      "VIB_10xv1_2.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 139\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "VIB_10xv1_2.FIXEDCELLS__sampleA_CD16+_monocyte_VS_sampleB_CD16+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 3312\n",
      "\tdars on contigs: 4\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.12062726176115801%\n",
      "\tdars on chrY: 2\n",
      "EPF_hydrop_1.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 22915\n",
      "\tdars on contigs: 15\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.06541648495420846%\n",
      "\tdars on chrY: 9\n",
      "EPF_hydrop_1.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 5186\n",
      "\tdars on contigs: 77\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 1.4630438913167394%\n",
      "\tdars on chrY: 0\n",
      "EPF_hydrop_1.FIXEDCELLS__sampleB_B_cell_VS_sampleA_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 11030\n",
      "\tdars on contigs: 79\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.7111351156719777%\n",
      "\tdars on chrY: 2\n",
      "CNA_10xmultiome_2.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xmultiome_2.FIXEDCELLS__sampleA_Cytotoxic_T_cell_VS_sampleB_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 870\n",
      "\tdars on contigs: 3\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.3436426116838488%\n",
      "\tdars on chrY: 2\n",
      "CNA_10xmultiome_2.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 26\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "STA_10xv11_2.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 4346\n",
      "\tdars on contigs: 7\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.16080863772111187%\n",
      "\tdars on chrY: 3\n",
      "SAN_10xmultiome_1.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 4289\n",
      "\tdars on contigs: 4\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.09317493594223154%\n",
      "\tdars on chrY: 2\n",
      "SAN_10xmultiome_1.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 64\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "SAN_10xmultiome_1.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 686\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.1455604075691412%\n",
      "\tdars on chrY: 1\n",
      "SAN_10xmultiome_1.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "SAN_10xmultiome_1.FIXEDCELLS__sampleA_CD16+_monocyte_VS_sampleB_CD16+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 995\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv2_1.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 82\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv2_1.FIXEDCELLS__sampleA_B_cell_VS_sampleB_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1664\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv2_1.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 10\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv2_1.FIXEDCELLS__sampleA_CD14+_monocyte_VS_sampleB_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1501\n",
      "\tdars on contigs: 2\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.1330671989354624%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv2_1.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 20\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "CNA_10xv2_1.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 3307\n",
      "\tdars on contigs: 4\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.12080942313500453%\n",
      "\tdars on chrY: 1\n",
      "HAR_ddseq_1.FIXEDCELLS__sampleA_Natural_killer_cell_VS_sampleB_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1036\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "HAR_ddseq_1.FIXEDCELLS__sampleB_Dendritic_cell_VS_sampleA_Dendritic_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 424\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "HAR_ddseq_1.FIXEDCELLS__sampleB_CD4+_T_cell_VS_sampleA_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 126\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "HAR_ddseq_1.FIXEDCELLS__sampleB_CD14+_monocyte_VS_sampleA_CD14+_monocyte__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 5\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "HAR_ddseq_1.FIXEDCELLS__sampleA_Dendritic_cell_VS_sampleB_Dendritic_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1979\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "HAR_ddseq_1.FIXEDCELLS__sampleB_Natural_killer_cell_VS_sampleA_Natural_killer_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 4\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "HAR_ddseq_1.FIXEDCELLS__sampleA_CD4+_T_cell_VS_sampleB_CD4+_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 1118\n",
      "\tdars on contigs: 1\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.08936550491510277%\n",
      "\tdars on chrY: 0\n",
      "HAR_ddseq_1.FIXEDCELLS__sampleB_Cytotoxic_T_cell_VS_sampleA_Cytotoxic_T_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 5\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n",
      "HAR_ddseq_1.FIXEDCELLS__sampleB_B_cell_VS_sampleA_B_cell__1-2xfoldchange_DARs.bed\n",
      "\tdars on standard chromosomes: 72\n",
      "\tdars on contigs: 0\n",
      "\tdars on chrM: 0\n",
      "\t% dars non standard chromosomes: 0.0%\n",
      "\tdars on chrY: 0\n"
     ]
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 43;\n",
       "                var nbb_unformatted_code = \"for sample, path in dar_path_dict.items():\\n    print(sample)\\n    dars_df = pd.read_csv(path, sep='\\\\t', header=None)\\n    \\n    chroms_in_df = list(sorted(dars_df[0].unique()))\\n    chroms_standard = [\\\"chr\\\" + str(x + 1) for x in range(22)] + ['chrX']\\n    chroms_nonstandard = list(set(chroms_in_df ) - set(chroms_standard) -  set(['chrM']))\\n    \\n    chroms_standard_in_sample = set(chroms_in_df).intersection(set(chroms_standard))\\n    n_standard = dars_df[0].value_counts()[chroms_standard_in_sample].sum()\\n    \\n    \\n    n_contigs = dars_df[0].value_counts()[chroms_nonstandard].sum()\\n    if 'chrM' in dars_df[0].value_counts().index:\\n        n_chrm = dars_df[0].value_counts()['chrM'].sum()\\n    else:\\n        n_chrm = 0\\n    pct_nonstandard = (n_contigs + n_chrm)/len(dars_df)*100\\n        \\n    print(f\\\"\\\\tdars on standard chromosomes: {n_standard}\\\")\\n    print(f\\\"\\\\tdars on contigs: {n_contigs}\\\")\\n    print(f\\\"\\\\tdars on chrM: {n_chrm}\\\")\\n    print(f\\\"\\\\t% dars non standard chromosomes: {pct_nonstandard}%\\\")\\n    if 'chrY' in chroms_in_df:\\n        n_chrY = dars_df[0].value_counts()['chrY'].sum()\\n        print(f\\\"\\\\tdars on chrY: {n_chrY}\\\")\\n    else:\\n        print(f\\\"\\\\tdars on chrY: 0\\\")\";\n",
       "                var nbb_formatted_code = \"for sample, path in dar_path_dict.items():\\n    print(sample)\\n    dars_df = pd.read_csv(path, sep=\\\"\\\\t\\\", header=None)\\n\\n    chroms_in_df = list(sorted(dars_df[0].unique()))\\n    chroms_standard = [\\\"chr\\\" + str(x + 1) for x in range(22)] + [\\\"chrX\\\"]\\n    chroms_nonstandard = list(set(chroms_in_df) - set(chroms_standard) - set([\\\"chrM\\\"]))\\n\\n    chroms_standard_in_sample = set(chroms_in_df).intersection(set(chroms_standard))\\n    n_standard = dars_df[0].value_counts()[chroms_standard_in_sample].sum()\\n\\n    n_contigs = dars_df[0].value_counts()[chroms_nonstandard].sum()\\n    if \\\"chrM\\\" in dars_df[0].value_counts().index:\\n        n_chrm = dars_df[0].value_counts()[\\\"chrM\\\"].sum()\\n    else:\\n        n_chrm = 0\\n    pct_nonstandard = (n_contigs + n_chrm) / len(dars_df) * 100\\n\\n    print(f\\\"\\\\tdars on standard chromosomes: {n_standard}\\\")\\n    print(f\\\"\\\\tdars on contigs: {n_contigs}\\\")\\n    print(f\\\"\\\\tdars on chrM: {n_chrm}\\\")\\n    print(f\\\"\\\\t% dars non standard chromosomes: {pct_nonstandard}%\\\")\\n    if \\\"chrY\\\" in chroms_in_df:\\n        n_chrY = dars_df[0].value_counts()[\\\"chrY\\\"].sum()\\n        print(f\\\"\\\\tdars on chrY: {n_chrY}\\\")\\n    else:\\n        print(f\\\"\\\\tdars on chrY: 0\\\")\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for sample, path in dar_path_dict.items():\n",
    "    print(sample)\n",
    "    dars_df = pd.read_csv(path, sep='\\t', header=None)\n",
    "    \n",
    "    chroms_in_df = list(sorted(dars_df[0].unique()))\n",
    "    chroms_standard = [\"chr\" + str(x + 1) for x in range(22)] + ['chrX']\n",
    "    chroms_nonstandard = list(set(chroms_in_df ) - set(chroms_standard) -  set(['chrM']))\n",
    "    \n",
    "    chroms_standard_in_sample = set(chroms_in_df).intersection(set(chroms_standard))\n",
    "    n_standard = dars_df[0].value_counts()[chroms_standard_in_sample].sum()\n",
    "    \n",
    "    \n",
    "    n_contigs = dars_df[0].value_counts()[chroms_nonstandard].sum()\n",
    "    if 'chrM' in dars_df[0].value_counts().index:\n",
    "        n_chrm = dars_df[0].value_counts()['chrM'].sum()\n",
    "    else:\n",
    "        n_chrm = 0\n",
    "    pct_nonstandard = (n_contigs + n_chrm)/len(dars_df)*100\n",
    "        \n",
    "    print(f\"\\tdars on standard chromosomes: {n_standard}\")\n",
    "    print(f\"\\tdars on contigs: {n_contigs}\")\n",
    "    print(f\"\\tdars on chrM: {n_chrm}\")\n",
    "    print(f\"\\t% dars non standard chromosomes: {pct_nonstandard}%\")\n",
    "    if 'chrY' in chroms_in_df:\n",
    "        n_chrY = dars_df[0].value_counts()['chrY'].sum()\n",
    "        print(f\"\\tdars on chrY: {n_chrY}\")\n",
    "    else:\n",
    "        print(f\"\\tdars on chrY: 0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bd4b8a1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "876865b9",
   "metadata": {
    "papermill": {
     "duration": 0.065009,
     "end_time": "2022-09-30T18:41:09.761191",
     "exception": false,
     "start_time": "2022-09-30T18:41:09.696182",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%% md\n"
    },
    "tags": []
   },
   "source": [
    "## Gene activity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "466755ba",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T18:41:09.893739Z",
     "iopub.status.busy": "2022-09-30T18:41:09.893408Z",
     "iopub.status.idle": "2022-09-30T18:41:10.241770Z",
     "shell.execute_reply": "2022-09-30T18:41:10.241279Z"
    },
    "papermill": {
     "duration": 0.415758,
     "end_time": "2022-09-30T18:41:10.242698",
     "exception": false,
     "start_time": "2022-09-30T18:41:09.826940",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 17;\n",
       "                var nbb_unformatted_code = \"import pyranges as pr\\nimport requests\\nimport pybiomart as pbm\";\n",
       "                var nbb_formatted_code = \"import pyranges as pr\\nimport requests\\nimport pybiomart as pbm\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pyranges as pr\n",
    "import requests\n",
    "import pybiomart as pbm"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e395a0e",
   "metadata": {
    "papermill": {
     "duration": 0.06587,
     "end_time": "2022-09-30T18:41:10.375139",
     "exception": false,
     "start_time": "2022-09-30T18:41:10.309269",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%% md\n"
    },
    "tags": []
   },
   "source": [
    "### Get gene annotations and chromosome sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6b92a450",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T18:41:10.507666Z",
     "iopub.status.busy": "2022-09-30T18:41:10.507368Z",
     "iopub.status.idle": "2022-09-30T18:41:17.698832Z",
     "shell.execute_reply": "2022-09-30T18:41:17.698345Z"
    },
    "papermill": {
     "duration": 7.258952,
     "end_time": "2022-09-30T18:41:17.699750",
     "exception": false,
     "start_time": "2022-09-30T18:41:10.440798",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Transcription_Start_Site</th>\n",
       "      <th>Transcript_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>1471765</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>1471784</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>+</td>\n",
       "      <td>PRDM16</td>\n",
       "      <td>3069168</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>+</td>\n",
       "      <td>PRDM16</td>\n",
       "      <td>3069197</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>+</td>\n",
       "      <td>PRDM16</td>\n",
       "      <td>3069211</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85681</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6865918</td>\n",
       "      <td>6911752</td>\n",
       "      <td>-</td>\n",
       "      <td>AMELY</td>\n",
       "      <td>6911752</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85682</th>\n",
       "      <td>chrY</td>\n",
       "      <td>6865918</td>\n",
       "      <td>6911752</td>\n",
       "      <td>-</td>\n",
       "      <td>AMELY</td>\n",
       "      <td>6872608</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85683</th>\n",
       "      <td>chrY</td>\n",
       "      <td>21903618</td>\n",
       "      <td>21918042</td>\n",
       "      <td>-</td>\n",
       "      <td>RBMY1E</td>\n",
       "      <td>21918032</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85684</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24045229</td>\n",
       "      <td>24048019</td>\n",
       "      <td>-</td>\n",
       "      <td>CDY1B</td>\n",
       "      <td>24047969</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85685</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24045229</td>\n",
       "      <td>24048019</td>\n",
       "      <td>-</td>\n",
       "      <td>CDY1B</td>\n",
       "      <td>24048019</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>85686 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+--------------+-----------+-----------+--------------+------------+-------+\n",
       "| Chromosome   | Start     | End       | Strand       | Gene       | +2    |\n",
       "| (category)   | (int32)   | (int32)   | (category)   | (object)   | ...   |\n",
       "|--------------+-----------+-----------+--------------+------------+-------|\n",
       "| chr1         | 1471765   | 1497848   | +            | ATAD3B     | ...   |\n",
       "| chr1         | 1471765   | 1497848   | +            | ATAD3B     | ...   |\n",
       "| chr1         | 3069168   | 3438621   | +            | PRDM16     | ...   |\n",
       "| chr1         | 3069168   | 3438621   | +            | PRDM16     | ...   |\n",
       "| ...          | ...       | ...       | ...          | ...        | ...   |\n",
       "| chrY         | 6865918   | 6911752   | -            | AMELY      | ...   |\n",
       "| chrY         | 21903618  | 21918042  | -            | RBMY1E     | ...   |\n",
       "| chrY         | 24045229  | 24048019  | -            | CDY1B      | ...   |\n",
       "| chrY         | 24045229  | 24048019  | -            | CDY1B      | ...   |\n",
       "+--------------+-----------+-----------+--------------+------------+-------+\n",
       "Stranded PyRanges object has 85,686 rows and 7 columns from 330 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome and Strand.\n",
       "2 hidden columns: Transcription_Start_Site, Transcript_type"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 18;\n",
       "                var nbb_unformatted_code = \"# For human\\ndataset = pbm.Dataset(name='hsapiens_gene_ensembl',  host='http://www.ensembl.org')\\nannot = dataset.query(attributes=['chromosome_name', 'start_position', 'end_position', 'strand', 'external_gene_name', 'transcription_start_site', 'transcript_biotype'])\\nannot['Chromosome/scaffold name'] = 'chr' + annot['Chromosome/scaffold name'].astype(str)\\nannot.columns=['Chromosome', 'Start', 'End', 'Strand', 'Gene','Transcription_Start_Site', 'Transcript_type']\\nannot = annot[annot.Transcript_type == 'protein_coding']\\nannot.Strand[annot.Strand == 1] = '+'\\nannot.Strand[annot.Strand == -1] = '-'\\npr_annotation = pr.PyRanges(annot.dropna(axis = 0))\\npr_annotation\";\n",
       "                var nbb_formatted_code = \"# For human\\ndataset = pbm.Dataset(name=\\\"hsapiens_gene_ensembl\\\", host=\\\"http://www.ensembl.org\\\")\\nannot = dataset.query(\\n    attributes=[\\n        \\\"chromosome_name\\\",\\n        \\\"start_position\\\",\\n        \\\"end_position\\\",\\n        \\\"strand\\\",\\n        \\\"external_gene_name\\\",\\n        \\\"transcription_start_site\\\",\\n        \\\"transcript_biotype\\\",\\n    ]\\n)\\nannot[\\\"Chromosome/scaffold name\\\"] = \\\"chr\\\" + annot[\\\"Chromosome/scaffold name\\\"].astype(\\n    str\\n)\\nannot.columns = [\\n    \\\"Chromosome\\\",\\n    \\\"Start\\\",\\n    \\\"End\\\",\\n    \\\"Strand\\\",\\n    \\\"Gene\\\",\\n    \\\"Transcription_Start_Site\\\",\\n    \\\"Transcript_type\\\",\\n]\\nannot = annot[annot.Transcript_type == \\\"protein_coding\\\"]\\nannot.Strand[annot.Strand == 1] = \\\"+\\\"\\nannot.Strand[annot.Strand == -1] = \\\"-\\\"\\npr_annotation = pr.PyRanges(annot.dropna(axis=0))\\npr_annotation\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# For human\n",
    "dataset = pbm.Dataset(name='hsapiens_gene_ensembl',  host='http://www.ensembl.org')\n",
    "annot = dataset.query(attributes=['chromosome_name', 'start_position', 'end_position', 'strand', 'external_gene_name', 'transcription_start_site', 'transcript_biotype'])\n",
    "annot['Chromosome/scaffold name'] = 'chr' + annot['Chromosome/scaffold name'].astype(str)\n",
    "annot.columns=['Chromosome', 'Start', 'End', 'Strand', 'Gene','Transcription_Start_Site', 'Transcript_type']\n",
    "annot = annot[annot.Transcript_type == 'protein_coding']\n",
    "annot.Strand[annot.Strand == 1] = '+'\n",
    "annot.Strand[annot.Strand == -1] = '-'\n",
    "pr_annotation = pr.PyRanges(annot.dropna(axis = 0))\n",
    "pr_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "77a6ea53",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T18:41:17.838030Z",
     "iopub.status.busy": "2022-09-30T18:41:17.837465Z",
     "iopub.status.idle": "2022-09-30T18:41:21.216428Z",
     "shell.execute_reply": "2022-09-30T18:41:21.215925Z"
    },
    "papermill": {
     "duration": 3.687825,
     "end_time": "2022-09-30T18:41:21.457583",
     "exception": false,
     "start_time": "2022-09-30T18:41:17.769758",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>0</td>\n",
       "      <td>248956422</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_GL383518v1_alt</td>\n",
       "      <td>0</td>\n",
       "      <td>182439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_GL383519v1_alt</td>\n",
       "      <td>0</td>\n",
       "      <td>110268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_GL383520v2_alt</td>\n",
       "      <td>0</td>\n",
       "      <td>366580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_KI270706v1_random</td>\n",
       "      <td>0</td>\n",
       "      <td>175055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450</th>\n",
       "      <td>chrX_KI270880v1_alt</td>\n",
       "      <td>0</td>\n",
       "      <td>284869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>chrX_KI270881v1_alt</td>\n",
       "      <td>0</td>\n",
       "      <td>144206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td>chrX_KI270913v1_alt</td>\n",
       "      <td>0</td>\n",
       "      <td>274009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td>chrY</td>\n",
       "      <td>0</td>\n",
       "      <td>57227415</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td>chrY_KI270740v1_random</td>\n",
       "      <td>0</td>\n",
       "      <td>37240</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>455 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "+------------------------+-----------+-----------+\n",
       "| Chromosome             | Start     | End       |\n",
       "| (category)             | (int32)   | (int32)   |\n",
       "|------------------------+-----------+-----------|\n",
       "| chr1                   | 0         | 248956422 |\n",
       "| chr1_GL383518v1_alt    | 0         | 182439    |\n",
       "| chr1_GL383519v1_alt    | 0         | 110268    |\n",
       "| chr1_GL383520v2_alt    | 0         | 366580    |\n",
       "| ...                    | ...       | ...       |\n",
       "| chrX_KI270881v1_alt    | 0         | 144206    |\n",
       "| chrX_KI270913v1_alt    | 0         | 274009    |\n",
       "| chrY                   | 0         | 57227415  |\n",
       "| chrY_KI270740v1_random | 0         | 37240     |\n",
       "+------------------------+-----------+-----------+\n",
       "Unstranded PyRanges object has 455 rows and 3 columns from 455 chromosomes.\n",
       "For printing, the PyRanges was sorted on Chromosome."
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 19;\n",
       "                var nbb_unformatted_code = \"# get chromosome sizes (hg38)\\ntarget_url = 'http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes'\\nchromsizes = pd.read_csv(target_url, sep='\\\\t', header=None)\\nchromsizes.columns = ['Chromosome', 'End']\\nchromsizes['Start'] = [0]*chromsizes.shape[0]\\nchromsizes = chromsizes.loc[:,['Chromosome', 'Start', 'End']]\\nchromsizes = pr.PyRanges(chromsizes)\\nchromsizes\";\n",
       "                var nbb_formatted_code = \"# get chromosome sizes (hg38)\\ntarget_url = \\\"http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes\\\"\\nchromsizes = pd.read_csv(target_url, sep=\\\"\\\\t\\\", header=None)\\nchromsizes.columns = [\\\"Chromosome\\\", \\\"End\\\"]\\nchromsizes[\\\"Start\\\"] = [0] * chromsizes.shape[0]\\nchromsizes = chromsizes.loc[:, [\\\"Chromosome\\\", \\\"Start\\\", \\\"End\\\"]]\\nchromsizes = pr.PyRanges(chromsizes)\\nchromsizes\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# get chromosome sizes (hg38)\n",
    "target_url = 'http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes'\n",
    "chromsizes = pd.read_csv(target_url, sep='\\t', header=None)\n",
    "chromsizes.columns = ['Chromosome', 'End']\n",
    "chromsizes['Start'] = [0]*chromsizes.shape[0]\n",
    "chromsizes = chromsizes.loc[:,['Chromosome', 'Start', 'End']]\n",
    "chromsizes = pr.PyRanges(chromsizes)\n",
    "chromsizes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b04ef7d7",
   "metadata": {
    "papermill": {
     "duration": 0.065696,
     "end_time": "2022-09-30T18:41:21.595726",
     "exception": false,
     "start_time": "2022-09-30T18:41:21.530030",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%% md\n"
    },
    "tags": []
   },
   "source": [
    "### Infer gene activity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "79be7503",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T18:41:21.730271Z",
     "iopub.status.busy": "2022-09-30T18:41:21.729816Z",
     "iopub.status.idle": "2022-09-30T18:41:21.735861Z",
     "shell.execute_reply": "2022-09-30T18:41:21.735411Z"
    },
    "papermill": {
     "duration": 0.073489,
     "end_time": "2022-09-30T18:41:21.736773",
     "exception": false,
     "start_time": "2022-09-30T18:41:21.663284",
     "status": "completed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 20;\n",
       "                var nbb_unformatted_code = \"from pycisTopic.gene_activity import get_gene_activity\\nfrom pycisTopic.diff_features import find_diff_features\";\n",
       "                var nbb_formatted_code = \"from pycisTopic.gene_activity import get_gene_activity\\nfrom pycisTopic.diff_features import find_diff_features\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from pycisTopic.gene_activity import get_gene_activity\n",
    "from pycisTopic.diff_features import find_diff_features"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "892da777",
   "metadata": {
    "tags": [
     "papermill-error-cell-tag"
    ]
   },
   "source": [
    "<span id=\"papermill-error-cell\" style=\"color:red; font-family:Helvetica Neue, Helvetica, Arial, sans-serif; font-size:2em;\">Execution using papermill encountered an exception here and stopped:</span>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ccdc461f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-30T18:41:21.872758Z",
     "iopub.status.busy": "2022-09-30T18:41:21.872288Z",
     "iopub.status.idle": "2022-09-30T18:41:23.260373Z",
     "shell.execute_reply": "2022-09-30T18:41:23.259693Z"
    },
    "papermill": {
     "duration": 1.456393,
     "end_time": "2022-09-30T18:41:23.261074",
     "exception": true,
     "start_time": "2022-09-30T18:41:21.804681",
     "status": "failed"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'cistopic_obj_dict' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[0;32mIn [21]\u001b[0m, in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m gene_act_dict \u001b[38;5;241m=\u001b[39m {}\n\u001b[0;32m----> 3\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m key \u001b[38;5;129;01min\u001b[39;00m \u001b[43mcistopic_obj_dict\u001b[49m\u001b[38;5;241m.\u001b[39mkeys():\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;28mprint\u001b[39m(key)\n\u001b[1;32m      5\u001b[0m     gene_act, weights \u001b[38;5;241m=\u001b[39m get_gene_activity(\n\u001b[1;32m      6\u001b[0m         imputed_acc_obj_dict[key], \u001b[38;5;66;03m# Region-cell probabilities\u001b[39;00m\n\u001b[1;32m      7\u001b[0m         pr_annotation, \u001b[38;5;66;03m# Gene annotation\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     27\u001b[0m         return_weights\u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m, \u001b[38;5;66;03m# Whether to return the final weights\u001b[39;00m\n\u001b[1;32m     28\u001b[0m         project\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mGene_activity\u001b[39m\u001b[38;5;124m'\u001b[39m) \u001b[38;5;66;03m# Project name for the gene activity object\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'cistopic_obj_dict' is not defined"
     ]
    },
    {
     "data": {
      "application/javascript": [
       "\n",
       "            setTimeout(function() {\n",
       "                var nbb_cell_id = 21;\n",
       "                var nbb_unformatted_code = \"gene_act_dict = {}\\n\\nfor key in cistopic_obj_dict.keys():\\n    print(key)\\n    gene_act, weights = get_gene_activity(\\n        imputed_acc_obj_dict[key], # Region-cell probabilities\\n        pr_annotation, # Gene annotation\\n        chromsizes, # Chromosome size\\n        use_gene_boundaries=True, # Whether to use the whole search space or stop when encountering another gene\\n        upstream=[1000, 100000], # Search space upstream. The minimum means that even if there is a gene right next to it \\n                                 #these bp will be taken (1kbp here)\\n        downstream=[1000,100000], # Search space downstream\\n        distance_weight=True, # Whether to add a distance weight (an exponential function, the weight will decrease with distance)\\n        decay_rate=1, # Exponent for the distance exponential funciton (the higher the faster will be the decrease)\\n        extend_gene_body_upstream=10000, # Number of bp upstream immune to the distance weight (their value will be maximum for \\n                                         #this weight)\\n        extend_gene_body_downstream=500, # Number of bp downstream immune to the distance weight\\n        gene_size_weight=False, # Whether to add a weights based on the length of the gene\\n        gene_size_scale_factor='median', # Dividend to calculate the gene size weigth. Default is the median value of all genes\\n                                         #in the genome\\n        remove_promoters=False, # Whether to remove promoters when computing gene activity scores\\n        average_scores=True, # Whether to divide by the total number of region assigned to a gene when calculating the gene \\n                             # activity score\\n        scale_factor=1, # Value to multiply for the final gene activity matrix\\n        extend_tss=[10,10], # Space to consider a promoter\\n        gini_weight = True, # Whether to add a gini index weigth. The more unique the region is, the higher this weight will be\\n        return_weights= True, # Whether to return the final weights\\n        project='Gene_activity') # Project name for the gene activity object\\n    gene_act_dict[key] = copy.copy(gene_act)\";\n",
       "                var nbb_formatted_code = \"gene_act_dict = {}\\n\\nfor key in cistopic_obj_dict.keys():\\n    print(key)\\n    gene_act, weights = get_gene_activity(\\n        imputed_acc_obj_dict[key],  # Region-cell probabilities\\n        pr_annotation,  # Gene annotation\\n        chromsizes,  # Chromosome size\\n        use_gene_boundaries=True,  # Whether to use the whole search space or stop when encountering another gene\\n        upstream=[\\n            1000,\\n            100000,\\n        ],  # Search space upstream. The minimum means that even if there is a gene right next to it\\n        # these bp will be taken (1kbp here)\\n        downstream=[1000, 100000],  # Search space downstream\\n        distance_weight=True,  # Whether to add a distance weight (an exponential function, the weight will decrease with distance)\\n        decay_rate=1,  # Exponent for the distance exponential funciton (the higher the faster will be the decrease)\\n        extend_gene_body_upstream=10000,  # Number of bp upstream immune to the distance weight (their value will be maximum for\\n        # this weight)\\n        extend_gene_body_downstream=500,  # Number of bp downstream immune to the distance weight\\n        gene_size_weight=False,  # Whether to add a weights based on the length of the gene\\n        gene_size_scale_factor=\\\"median\\\",  # Dividend to calculate the gene size weigth. Default is the median value of all genes\\n        # in the genome\\n        remove_promoters=False,  # Whether to remove promoters when computing gene activity scores\\n        average_scores=True,  # Whether to divide by the total number of region assigned to a gene when calculating the gene\\n        # activity score\\n        scale_factor=1,  # Value to multiply for the final gene activity matrix\\n        extend_tss=[10, 10],  # Space to consider a promoter\\n        gini_weight=True,  # Whether to add a gini index weigth. The more unique the region is, the higher this weight will be\\n        return_weights=True,  # Whether to return the final weights\\n        project=\\\"Gene_activity\\\",\\n    )  # Project name for the gene activity object\\n    gene_act_dict[key] = copy.copy(gene_act)\";\n",
       "                var nbb_cells = Jupyter.notebook.get_cells();\n",
       "                for (var i = 0; i < nbb_cells.length; ++i) {\n",
       "                    if (nbb_cells[i].input_prompt_number == nbb_cell_id) {\n",
       "                        if (nbb_cells[i].get_text() == nbb_unformatted_code) {\n",
       "                             nbb_cells[i].set_text(nbb_formatted_code);\n",
       "                        }\n",
       "                        break;\n",
       "                    }\n",
       "                }\n",
       "            }, 500);\n",
       "            "
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "gene_act_dict = {}\n",
    "\n",
    "for key in cistopic_obj_dict.keys():\n",
    "    print(key)\n",
    "    gene_act, weights = get_gene_activity(\n",
    "        imputed_acc_obj_dict[key], # Region-cell probabilities\n",
    "        pr_annotation, # Gene annotation\n",
    "        chromsizes, # Chromosome size\n",
    "        use_gene_boundaries=True, # Whether to use the whole search space or stop when encountering another gene\n",
    "        upstream=[1000, 100000], # Search space upstream. The minimum means that even if there is a gene right next to it \n",
    "                                 #these bp will be taken (1kbp here)\n",
    "        downstream=[1000,100000], # Search space downstream\n",
    "        distance_weight=True, # Whether to add a distance weight (an exponential function, the weight will decrease with distance)\n",
    "        decay_rate=1, # Exponent for the distance exponential funciton (the higher the faster will be the decrease)\n",
    "        extend_gene_body_upstream=10000, # Number of bp upstream immune to the distance weight (their value will be maximum for \n",
    "                                         #this weight)\n",
    "        extend_gene_body_downstream=500, # Number of bp downstream immune to the distance weight\n",
    "        gene_size_weight=False, # Whether to add a weights based on the length of the gene\n",
    "        gene_size_scale_factor='median', # Dividend to calculate the gene size weigth. Default is the median value of all genes\n",
    "                                         #in the genome\n",
    "        remove_promoters=False, # Whether to remove promoters when computing gene activity scores\n",
    "        average_scores=True, # Whether to divide by the total number of region assigned to a gene when calculating the gene \n",
    "                             # activity score\n",
    "        scale_factor=1, # Value to multiply for the final gene activity matrix\n",
    "        extend_tss=[10,10], # Space to consider a promoter\n",
    "        gini_weight = True, # Whether to add a gini index weigth. The more unique the region is, the higher this weight will be\n",
    "        return_weights= True, # Whether to return the final weights\n",
    "        project='Gene_activity') # Project name for the gene activity object\n",
    "    gene_act_dict[key] = copy.copy(gene_act)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "298be380",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "pycharm": {
     "name": "#%% md\n"
    },
    "tags": []
   },
   "source": [
    "### Infer the Differentially Accessible Genes (DAGs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2dd51a1",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pycisTopic.clust_vis import plot_imputed_features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12fe82e7",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "dag_markers_dict = {}\n",
    "for key in cistopic_obj_dict.keys():\n",
    "    print(key)\n",
    "    dag_markers_dict[key] = find_diff_features(cistopic_obj_dict[key],\n",
    "                      gene_act_dict[key],\n",
    "                      variable='consensus_cell_type',\n",
    "                      var_features=None,\n",
    "                      contrasts=None,\n",
    "                      adjpval_thr=0.05,\n",
    "                      log2fc_thr=np.log2(1.5),\n",
    "                      n_cpu=5,\n",
    "                      )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8ecf962",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "for key in cistopic_obj_dict.keys():\n",
    "    print(key)\n",
    "    plot_imputed_features(cistopic_obj_dict[key],\n",
    "                    reduction_name='UMAP',\n",
    "                    imputed_data=gene_act,\n",
    "                    features=['CD34', 'THY1', # CD34+\n",
    "                              'NCAM1', 'FCGR3A', # NK\n",
    "                              'CD14', 'CD68', # Monocytes\n",
    "                              'MS4A1', 'CD79A', # B cells\n",
    "                              'CD8A', 'CD8B', # Cytotoxic CD8 cells\n",
    "                              'CD4', # CD4 T\n",
    "                              'CD3D', # T\n",
    "                              'IL3RA', 'CD1C', 'BATF3', # Dendritic\n",
    "                             ],\n",
    "                    scale=True,\n",
    "                    num_columns=5,\n",
    "                    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "095e7301",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "pycharm": {
     "name": "#%% md\n"
    },
    "tags": []
   },
   "source": [
    "## Export to loom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7e17df5",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# cistopic_obj_dict # ok\n",
    "# imputed_acc_obj_dict # ok\n",
    "# region_bin_topics #(region_bin_topics_dict) ok\n",
    "# binarized_cell_topic # ok\n",
    "# DARs_dict #(markers_dict_dict) ok"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c825d13",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "f_region_loom_dir = os.path.join(f_final_dir, 'region_acc_loom')\n",
    "if not os.path.exists(f_region_loom_dir):\n",
    "    os.makedirs(f_region_loom_dir)\n",
    "    \n",
    "f_gene_loom_dir = os.path.join(f_final_dir, 'gene_act_loom')\n",
    "if not os.path.exists(f_gene_loom_dir):\n",
    "    os.makedirs(f_gene_loom_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18216414",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "pycharm": {
     "name": "#%% md\n"
    },
    "tags": []
   },
   "source": [
    "### Region accessibility"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5224fcc",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pycisTopic.loom import (\n",
    "    export_region_accessibility_to_loom,\n",
    "    export_gene_activity_to_loom\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db8633c7",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "for key in cistopic_obj_dict.keys():\n",
    "    print(key)\n",
    "    s = 'merged' if key=='merged' else sample_annot.loc[key,'Sample']\n",
    "    f_out = os.path.join(f_region_loom_dir, s + '__libDS_region_accessibility.loom')\n",
    "    if os.path.exists(f_out):\n",
    "        print(f\"Skipping {f_out}: already exists.\")\n",
    "        continue\n",
    "    # Subset regions, we will use only regions in topics and DARs here to make it faster\n",
    "    regions_in_topics = list(set(sum([region_bin_topics_dict[key][i].index.tolist() for i in region_bin_topics_dict[key].keys()],[])))\n",
    "    regions_in_DARs = list(set(sum([markers_dict_dict[key][i].index.tolist() for i in markers_dict_dict[key].keys()],[])))\n",
    "    # make sure we only take regions that actually exist in the accessibility matrix:\n",
    "    selected_regions = list(set(regions_in_topics + regions_in_DARs).intersection(set(imputed_acc_obj_dict[key].feature_names)))\n",
    "\n",
    "    # Export to loom\n",
    "    export_region_accessibility_to_loom(\n",
    "        accessibility_matrix = imputed_acc_obj_dict[key],\n",
    "        cistopic_obj = cistopic_obj_dict[key], \n",
    "        binarized_topic_region = region_bin_topics_dict[key],\n",
    "        binarized_cell_topic = binarized_cell_topic_dict[key],\n",
    "        out_fname = f_out,\n",
    "        selected_regions = selected_regions ,\n",
    "        selected_cells = [ x.split('-')[0] + '-' + x.split('-')[1]  for x in cistopic_obj_dict[key].cell_names ], # this leaves a cell barcode of the format type 'TGCATGTCGCCGTTCCAAGA-21'\n",
    "        # selected_cells = cistopic_obj_dict[key].projections['cell']['UMAP'].index.tolist(), # cflerin original\n",
    "        cluster_annotation = ['consensus_cell_type'],\n",
    "        cluster_markers = {'consensus_cell_type': markers_dict_dict[key]},\n",
    "        tree_structure = ('scATAC-seq_Benchmark', 'ATAC_library_downsampled', 'Region_accessibility'),\n",
    "        title = s + ' - Region accessibility all',\n",
    "        nomenclature = \"hg38\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3230fa6b",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "len([ x.split('-')[0] + '-' + x.split('-')[1]  for x in cistopic_obj_dict[key].cell_names ])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "738e7900",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "b36c1b35",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "pycharm": {
     "name": "#%% md\n"
    },
    "tags": []
   },
   "source": [
    "### Gene activity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "861c8852",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from ctxcore.genesig import Regulon\n",
    "\n",
    "# generate a dummy regulon (required for export_gene_activity_to_loom):\n",
    "phreg = Regulon(\n",
    "        name='placeholder regulon',\n",
    "        gene2weight={'phreg': 1.0},\n",
    "        transcription_factor=\"phreg\",\n",
    "        gene2occurrence={\"phreg\": 1},\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f89ec89",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "pycharm": {
     "name": "#%%\n"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "for key in cistopic_obj_dict.keys():\n",
    "    print(key)\n",
    "    s = 'merged' if key=='merged' else sample_annot.loc[key,'Sample']\n",
    "    f_out = os.path.join(f_gene_loom_dir, s + '__libDS_gene_activity.loom')\n",
    "    if os.path.exists(f_out):\n",
    "        print(f\"Skipping {f_out}: already exists.\")\n",
    "        continue\n",
    "    export_gene_activity_to_loom(\n",
    "        gene_activity_matrix = gene_act_dict[key],\n",
    "        cistopic_obj = cistopic_obj_dict[key], \n",
    "        regulons = [phreg],\n",
    "        selected_cells = [ x.split('-')[0] + '-' + x.split('-')[1]  for x in cistopic_obj_dict[key].cell_names ], # this leaves a cell barcode of the format type 'TGCATGTCGCCGTTCCAAGA-21'\n",
    "        # selected_cells = cistopic_obj_dict[key].projections['cell']['UMAP'].index.tolist(), # cflerin original        out_fname = f_out,\n",
    "        cluster_annotation = ['consensus_cell_type'],\n",
    "        cluster_markers = {'consensus_cell_type': markers_dict_dict[key]},\n",
    "        tree_structure = ('scATAC-seq_Benchmark', 'ATAC_library_downsampled', 'Gene_activity'),\n",
    "        title = s + ' - Gene activity',\n",
    "        nomenclature = \"hg38\"\n",
    "    )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "20221014_pycistopic.sif",
   "language": "python",
   "name": "cistopic_20221014"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 18901.281753,
   "end_time": "2022-09-30T18:41:26.947373",
   "environment_variables": {},
   "exception": true,
   "input_path": "7b_downstream_analysis.ipynb",
   "output_path": "7b_downstream_analysis.OUT.ipynb",
   "parameters": {},
   "start_time": "2022-09-30T13:26:25.665620",
   "version": "2.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
