{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a7f24fc6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-03T14:58:32.765289Z",
     "iopub.status.busy": "2022-12-03T14:58:32.764287Z",
     "iopub.status.idle": "2022-12-03T14:58:48.865414Z",
     "shell.execute_reply": "2022-12-03T14:58:48.863749Z",
     "shell.execute_reply.started": "2022-12-03T14:58:32.765122Z"
    },
    "papermill": {
     "duration": 9.431443,
     "end_time": "2022-09-17T13:41:45.940549",
     "exception": false,
     "start_time": "2022-09-17T13:41:36.509106",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pycisTopic.pseudobulk_peak_calling import export_pseudobulk, peak_calling\n",
    "import pyranges as pr\n",
    "import requests\n",
    "import os\n",
    "import pandas as pd\n",
    "import glob\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "97251182",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-03T14:58:48.869598Z",
     "iopub.status.busy": "2022-12-03T14:58:48.868481Z",
     "iopub.status.idle": "2022-12-03T14:58:49.121030Z",
     "shell.execute_reply": "2022-12-03T14:58:49.119528Z",
     "shell.execute_reply.started": "2022-12-03T14:58:48.869536Z"
    },
    "papermill": {
     "duration": 0.278601,
     "end_time": "2022-09-17T13:41:46.248189",
     "exception": false,
     "start_time": "2022-09-17T13:41:45.969588",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%load_ext lab_black"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c1352160",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-03T14:58:49.122780Z",
     "iopub.status.busy": "2022-12-03T14:58:49.122308Z",
     "iopub.status.idle": "2022-12-03T14:58:49.208866Z",
     "shell.execute_reply": "2022-12-03T14:58:49.207404Z",
     "shell.execute_reply.started": "2022-12-03T14:58:49.122752Z"
    },
    "papermill": {
     "duration": 1.952478,
     "end_time": "2022-09-17T13:41:48.205405",
     "exception": false,
     "start_time": "2022-09-17T13:41:46.252927",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# get chromosome sizes (hg38)\n",
    "if not os.path.exists(\"chromsizes.txt\"):\n",
    "    target_url = (\n",
    "        \"http://hgdownload.cse.ucsc.edu/goldenPath/hg38/bigZips/hg38.chrom.sizes\"\n",
    "    )\n",
    "    chromsizes = pd.read_csv(target_url, sep=\"\\t\", header=None)\n",
    "    chromsizes.columns = [\"Chromosome\", \"End\"]\n",
    "    chromsizes[\"Start\"] = [0] * chromsizes.shape[0]\n",
    "    chromsizes = chromsizes.loc[:, [\"Chromosome\", \"Start\", \"End\"]]\n",
    "    chromsizes = pr.PyRanges(chromsizes)\n",
    "    chromsizes.to_csv(\"chromsizes.txt\")\n",
    "    chromsizes\n",
    "else:\n",
    "    chromsizes = pd.read_csv(\"chromsizes.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "4a5dca7a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-03T14:58:49.211890Z",
     "iopub.status.busy": "2022-12-03T14:58:49.211291Z",
     "iopub.status.idle": "2022-12-03T14:58:49.235979Z",
     "shell.execute_reply": "2022-12-03T14:58:49.234980Z",
     "shell.execute_reply.started": "2022-12-03T14:58:49.211849Z"
    },
    "papermill": {
     "duration": 0.033073,
     "end_time": "2022-09-17T13:41:48.242908",
     "exception": false,
     "start_time": "2022-09-17T13:41:48.209835",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': 'cistopic_objects/BIO_ddseq_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': 'cistopic_objects/BIO_ddseq_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_14topics.dimreduc.consensus.pkl',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': 'cistopic_objects/BIO_ddseq_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': 'cistopic_objects/BIO_ddseq_4.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': 'cistopic_objects/BRO_mtscatac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': 'cistopic_objects/BRO_mtscatac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_17topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': 'cistopic_objects/CNA_10xmultiome_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': 'cistopic_objects/CNA_10xmultiome_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_8topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': 'cistopic_objects/CNA_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': 'cistopic_objects/CNA_10xv11_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': 'cistopic_objects/CNA_10xv11_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': 'cistopic_objects/CNA_10xv11_4.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': 'cistopic_objects/CNA_10xv11_5.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': 'cistopic_objects/CNA_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': 'cistopic_objects/CNA_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': 'cistopic_objects/CNA_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': 'cistopic_objects/CNA_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': 'cistopic_objects/CNA_hydrop_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': 'cistopic_objects/CNA_mtscatac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': 'cistopic_objects/CNA_mtscatac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': 'cistopic_objects/EPF_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': 'cistopic_objects/EPF_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': 'cistopic_objects/EPF_hydrop_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_5topics.dimreduc.consensus.pkl',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': 'cistopic_objects/EPF_hydrop_4.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': 'cistopic_objects/HAR_ddseq_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': 'cistopic_objects/HAR_ddseq_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': 'cistopic_objects/MDC_mtscatac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_19topics.dimreduc.consensus.pkl',\n",
       " 'MDC_mtscatac_2.FIXEDCELLS': 'cistopic_objects/MDC_mtscatac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.pkl',\n",
       " 'OHS_s3atac_1.FIXEDCELLS': 'cistopic_objects/OHS_s3atac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'OHS_s3atac_2.FIXEDCELLS': 'cistopic_objects/OHS_s3atac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_7topics.dimreduc.consensus.pkl',\n",
       " 'SAN_10xmultiome_1.FIXEDCELLS': 'cistopic_objects/SAN_10xmultiome_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_7topics.dimreduc.consensus.pkl',\n",
       " 'SAN_10xmultiome_2.FIXEDCELLS': 'cistopic_objects/SAN_10xmultiome_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'STA_10xv11_1.FIXEDCELLS': 'cistopic_objects/STA_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl',\n",
       " 'STA_10xv11_2.FIXEDCELLS': 'cistopic_objects/STA_10xv11_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_7topics.dimreduc.consensus.pkl',\n",
       " 'TXG_10xv11_1.FIXEDCELLS': 'cistopic_objects/TXG_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_14topics.dimreduc.consensus.pkl',\n",
       " 'TXG_10xv2_1.FIXEDCELLS': 'cistopic_objects/TXG_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_21topics.dimreduc.consensus.pkl',\n",
       " 'TXG_10xv2_2.FIXEDCELLS': 'cistopic_objects/TXG_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_21topics.dimreduc.consensus.pkl',\n",
       " 'UCS_ddseq_1.FIXEDCELLS': 'cistopic_objects/UCS_ddseq_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'UCS_ddseq_2.FIXEDCELLS': 'cistopic_objects/UCS_ddseq_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xmultiome_1.FIXEDCELLS': 'cistopic_objects/VIB_10xmultiome_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xmultiome_2.FIXEDCELLS': 'cistopic_objects/VIB_10xmultiome_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xv1_1.FIXEDCELLS': 'cistopic_objects/VIB_10xv1_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xv1_2.FIXEDCELLS': 'cistopic_objects/VIB_10xv1_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_19topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xv2_1.FIXEDCELLS': 'cistopic_objects/VIB_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'VIB_10xv2_2.FIXEDCELLS': 'cistopic_objects/VIB_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.pkl',\n",
       " 'VIB_hydrop_1.FIXEDCELLS': 'cistopic_objects/VIB_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_17topics.dimreduc.consensus.pkl',\n",
       " 'VIB_hydrop_2.FIXEDCELLS': 'cistopic_objects/VIB_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_19topics.dimreduc.consensus.pkl'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cto_consensus_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"cistopic_objects/*singlets.model*consensus.pkl\"))\n",
    "}\n",
    "cto_consensus_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "810ade15",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-03T15:00:44.188077Z",
     "iopub.status.busy": "2022-12-03T15:00:44.187172Z",
     "iopub.status.idle": "2022-12-03T15:00:44.209965Z",
     "shell.execute_reply": "2022-12-03T15:00:44.208611Z",
     "shell.execute_reply.started": "2022-12-03T15:00:44.188014Z"
    },
    "papermill": {
     "duration": 0.014187,
     "end_time": "2022-09-17T13:41:48.261268",
     "exception": false,
     "start_time": "2022-09-17T13:41:48.247081",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BIO_ddseq_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BIO_ddseq_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BIO_ddseq_3.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BIO_ddseq_4.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BRO_mtscatac_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BRO_mtscatac_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xmultiome_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xmultiome_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv11_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv11_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv11_3.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv11_4.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv11_5.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv2_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv2_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_hydrop_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_hydrop_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_hydrop_3.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_mtscatac_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_mtscatac_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/EPF_hydrop_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/EPF_hydrop_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/EPF_hydrop_3.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/EPF_hydrop_4.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/HAR_ddseq_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/HAR_ddseq_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/MDC_mtscatac_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/MDC_mtscatac_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'OHS_s3atac_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/OHS_s3atac_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'OHS_s3atac_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/OHS_s3atac_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/SAN_10xmultiome_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/SAN_10xmultiome_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'STA_10xv11_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/STA_10xv11_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'STA_10xv11_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/STA_10xv11_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'TXG_10xv11_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/TXG_10xv11_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'TXG_10xv2_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/TXG_10xv2_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'TXG_10xv2_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/TXG_10xv2_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'UCS_ddseq_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/UCS_ddseq_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'UCS_ddseq_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/UCS_ddseq_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xmultiome_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xmultiome_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xv1_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xv1_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xv1_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xv1_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xv2_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xv2_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xv2_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xv2_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_11.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_11.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_12.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_12.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_21.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_21.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_22.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_22.FIXEDCELLS.fragments.tsv.gz'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fragments_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\".fragments.tsv.gz\")[0]: x\n",
    "    for x in sorted(glob.glob(\"../1_data_repository/fixedcells_fragments/*.tsv.gz\"))\n",
    "}\n",
    "fragments_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d40de736",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-03T15:00:46.546305Z",
     "iopub.status.busy": "2022-12-03T15:00:46.545504Z",
     "iopub.status.idle": "2022-12-03T15:00:46.647486Z",
     "shell.execute_reply": "2022-12-03T15:00:46.645915Z",
     "shell.execute_reply.started": "2022-12-03T15:00:46.546246Z"
    },
    "papermill": {
     "duration": 0.083163,
     "end_time": "2022-09-17T13:41:48.348632",
     "exception": false,
     "start_time": "2022-09-17T13:41:48.265469",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import gc\n",
    "import logging\n",
    "import os\n",
    "import re\n",
    "import subprocess\n",
    "import sys\n",
    "from typing import Dict, List, Optional, Union\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyBigWig\n",
    "import pyranges as pr\n",
    "import ray\n",
    "\n",
    "from pycisTopic.cistopic_class import *\n",
    "from pycisTopic.utils import *\n",
    "\n",
    "\n",
    "def export_pseudobulk(\n",
    "    input_data: Union[\"CistopicObject\", pd.DataFrame, Dict[str, pd.DataFrame]],\n",
    "    variable: str,\n",
    "    chromsizes: Union[pd.DataFrame, pr.PyRanges],\n",
    "    bed_path: str,\n",
    "    bigwig_path: str,\n",
    "    path_to_fragments: Optional[Dict[str, str]] = None,\n",
    "    sample_id_col: Optional[str] = \"sample_id\",\n",
    "    n_cpu: Optional[int] = 1,\n",
    "    normalize_bigwig: Optional[bool] = True,\n",
    "    remove_duplicates: Optional[bool] = True,\n",
    "    split_pattern: Optional[str] = \"___\",\n",
    "    use_polars: Optional[bool] = True,\n",
    "    **kwargs\n",
    "):\n",
    "    \"\"\"\n",
    "    Create pseudobulks as bed and bigwig from single cell fragments file given a barcode annotation.\n",
    "    Parameters\n",
    "    ---------\n",
    "    input_data: CistopicObject or pd.DataFrame\n",
    "            A :class:`CistopicObject` containing the specified `variable` as a column in :class:`CistopicObject.cell_data` or a cell metadata\n",
    "            :class:`pd.DataFrame` containing barcode as rows, containing the specified `variable` as a column (additional columns are\n",
    "            possible) and a `sample_id` column. Index names must contain the BARCODE (e.g. ATGTCGTC-1), additional tags are possible separating with -\n",
    "            (e.g. ATGCTGTGCG-1-Sample_1). The levels in the sample_id column must agree with the keys in the path_to_fragments dictionary.\n",
    "            Alternatively, if the cell metadata contains a column named barcode it will be used instead of the index names.\n",
    "    variable: str\n",
    "            A character string indicating the column that will be used to create the different group pseudobulk. It must be included in\n",
    "            the cell metadata provided as input_data.\n",
    "    chromsizes: pd.DataFrame or pr.PyRanges\n",
    "            A data frame or :class:`pr.PyRanges` containing size of each chromosome, containing 'Chromosome', 'Start' and 'End' columns.\n",
    "    bed_path: str\n",
    "            Path to folder where the fragments bed files per group will be saved. If None, files will not be generated.\n",
    "    bigwig_path: str\n",
    "            Path to folder where the bigwig files per group will be saved. If None, files will not be generated.\n",
    "    path_to_fragments: str or dict, optional\n",
    "            A dictionary of character strings, with sample name as names indicating the path to the fragments file/s from which pseudobulk profiles have to\n",
    "            be created. If a :class:`CistopicObject` is provided as input it will be ignored, but if a cell metadata :class:`pd.DataFrame` is provided it\n",
    "            is necessary to provide it. The keys of the dictionary need to match with the sample_id tag added to the index names of the input data frame.\n",
    "    sample_id_col: str, optional\n",
    "            Name of the column containing the sample name per barcode in the input :class:`CistopicObject.cell_data` or class:`pd.DataFrame`. Default: 'sample_id'.\n",
    "    n_cpu: int, optional\n",
    "            Number of cores to use. Default: 1.\n",
    "    normalize_bigwig: bool, optional\n",
    "            Whether bigwig files should be CPM normalized. Default: True.\n",
    "    remove_duplicates: bool, optional\n",
    "            Whether duplicates should be removed before converting the data to bigwig.\n",
    "    split_pattern: str, optional\n",
    "            Pattern to split cell barcode from sample id. Default: ___ .\n",
    "    use_polars: bool, optional\n",
    "            Whether to use polars to read fragments files. Default: True.\n",
    "    **kwargs\n",
    "            Additional parameters for ray.init()\n",
    "    Return\n",
    "    ------\n",
    "    dict\n",
    "            A dictionary containing the paths to the newly created bed fragments files per group a dictionary containing the paths to the\n",
    "            newly created bigwig files per group.\n",
    "    \"\"\"\n",
    "    # Create logger\n",
    "    level = logging.INFO\n",
    "    log_format = \"%(asctime)s %(name)-12s %(levelname)-8s %(message)s\"\n",
    "    handlers = [logging.StreamHandler(stream=sys.stdout)]\n",
    "    logging.basicConfig(level=level, format=log_format, handlers=handlers)\n",
    "    log = logging.getLogger(\"cisTopic\")\n",
    "\n",
    "    # Get fragments file\n",
    "    if isinstance(input_data, CistopicObject):\n",
    "        path_to_fragments = input_data.path_to_fragments\n",
    "        if path_to_fragments is None:\n",
    "            log.error(\"No path_to_fragments in this cisTopic object.\")\n",
    "        cell_data = input_data.cell_data\n",
    "    elif isinstance(input_data, pd.DataFrame):\n",
    "        if path_to_fragments is None:\n",
    "            log.error(\"Please, provide path_to_fragments.\")\n",
    "        cell_data = input_data\n",
    "    # Check for sample_id column\n",
    "    try:\n",
    "        sample_ids = list(set(cell_data[sample_id_col]))\n",
    "    except ValueError:\n",
    "        print(\n",
    "            'Please, include a sample identification column (e.g. \"sample_id\") in your cell metadata!'\n",
    "        )\n",
    "\n",
    "    # Get fragments\n",
    "    fragments_df_dict = {}\n",
    "    for sample_id in path_to_fragments.keys():\n",
    "        if sample_id not in sample_ids:\n",
    "            log.info(\n",
    "                \"The following path_to_fragments entry is not found in the cell metadata sample_id_col: \",\n",
    "                sample_id,\n",
    "                \". It will be ignored.\",\n",
    "            )\n",
    "        else:\n",
    "            log.info(\"Reading fragments from \" + path_to_fragments[sample_id])\n",
    "            fragments_df = read_fragments_from_file(\n",
    "                path_to_fragments[sample_id], use_polars=use_polars\n",
    "            ).df\n",
    "            # Convert to int32 for memory efficiency\n",
    "            fragments_df.Start = np.int32(fragments_df.Start)\n",
    "            fragments_df.End = np.int32(fragments_df.End)\n",
    "            if \"Score\" in fragments_df:\n",
    "                fragments_df.Score = np.int32(fragments_df.Score)\n",
    "            if \"barcode\" in cell_data:\n",
    "                fragments_df = fragments_df.loc[\n",
    "                    fragments_df[\"Name\"].isin(cell_data[\"barcode\"].tolist())\n",
    "                ]\n",
    "            else:\n",
    "                fragments_df = fragments_df.loc[\n",
    "                    fragments_df[\"Name\"].isin(\n",
    "                        prepare_tag_cells(cell_data.index.tolist(), split_pattern)\n",
    "                    )\n",
    "                ]\n",
    "            fragments_df_dict[sample_id] = fragments_df\n",
    "\n",
    "    # Set groups\n",
    "    if \"barcode\" in cell_data:\n",
    "        cell_data = cell_data.loc[:, [variable, sample_id_col, \"barcode\"]]\n",
    "    else:\n",
    "        cell_data = cell_data.loc[:, [variable, sample_id_col]]\n",
    "    cell_data[variable] = cell_data[variable].replace(\" \", \"\", regex=True)\n",
    "    cell_data[variable] = cell_data[variable].replace(\"[^A-Za-z0-9]+\", \"_\", regex=True)\n",
    "    groups = sorted(list(set(cell_data[variable])))\n",
    "    # Check chromosome sizes\n",
    "    if isinstance(chromsizes, pd.DataFrame):\n",
    "        chromsizes = chromsizes.loc[:, [\"Chromosome\", \"Start\", \"End\"]]\n",
    "        chromsizes = pr.PyRanges(chromsizes)\n",
    "    # Check that output dir exist and generate output paths\n",
    "    if isinstance(bed_path, str):\n",
    "        if not os.path.exists(bed_path):\n",
    "            os.makedirs(bed_path)\n",
    "        bed_paths = {\n",
    "            group: os.path.join(bed_path, str(group) + \".bed.gz\") for group in groups\n",
    "        }\n",
    "    else:\n",
    "        bed_paths = {}\n",
    "    if isinstance(bigwig_path, str):\n",
    "        if not os.path.exists(bigwig_path):\n",
    "            os.makedirs(bigwig_path)\n",
    "        bw_paths = {\n",
    "            group: os.path.join(bigwig_path, str(group) + \".bw\") for group in groups\n",
    "        }\n",
    "    else:\n",
    "        bw_paths = {}\n",
    "    # Create pseudobulks\n",
    "    if n_cpu > 1:\n",
    "        ray.init(num_cpus=n_cpu, **kwargs)\n",
    "        ray_handle = ray.wait(\n",
    "            [\n",
    "                export_pseudobulk_ray.remote(\n",
    "                    cell_data,\n",
    "                    group,\n",
    "                    fragments_df_dict,\n",
    "                    chromsizes,\n",
    "                    bigwig_path,\n",
    "                    bed_path,\n",
    "                    sample_id_col,\n",
    "                    normalize_bigwig,\n",
    "                    remove_duplicates,\n",
    "                    split_pattern,\n",
    "                )\n",
    "                for group in groups\n",
    "            ],\n",
    "            num_returns=len(groups),\n",
    "        )\n",
    "        ray.shutdown()\n",
    "    else:\n",
    "        [\n",
    "            export_pseudobulk_one_sample(\n",
    "                cell_data,\n",
    "                group,\n",
    "                fragments_df_dict,\n",
    "                chromsizes,\n",
    "                bigwig_path,\n",
    "                bed_path,\n",
    "                sample_id_col,\n",
    "                normalize_bigwig,\n",
    "                remove_duplicates,\n",
    "                split_pattern,\n",
    "            )\n",
    "            for group in groups\n",
    "        ]\n",
    "\n",
    "    return bw_paths, bed_paths\n",
    "\n",
    "\n",
    "def export_pseudobulk_one_sample(\n",
    "    cell_data: pd.DataFrame,\n",
    "    group: str,\n",
    "    fragments_df_dict: Dict[str, pd.DataFrame],\n",
    "    chromsizes: pr.PyRanges,\n",
    "    bigwig_path: str,\n",
    "    bed_path: str,\n",
    "    sample_id_col: Optional[str] = \"sample_id\",\n",
    "    normalize_bigwig: Optional[bool] = True,\n",
    "    remove_duplicates: Optional[bool] = True,\n",
    "    split_pattern: Optional[str] = \"___\",\n",
    "):\n",
    "    \"\"\"\n",
    "    Create pseudobulk as bed and bigwig from single cell fragments file given a barcode annotation and a group.\n",
    "    Parameters\n",
    "    ---------\n",
    "    cell_data: pd.DataFrame\n",
    "            A cell metadata :class:`pd.Dataframe` containing barcodes, their annotation and their sample of origin.\n",
    "    group: str\n",
    "            A character string indicating the group for which pseudobulks will be created.\n",
    "    fragments_df_dict: dict\n",
    "            A dictionary containing data frames as values with 'Chromosome', 'Start', 'End', 'Name', and 'Score' as columns; and sample label\n",
    "            as keys. 'Score' indicates the number of times that a fragments is found assigned to that barcode.\n",
    "    chromsizes: pr.PyRanges\n",
    "            A :class:`pr.PyRanges` containing size of each column, containing 'Chromosome', 'Start' and 'End' columns.\n",
    "    bigwig_path: str\n",
    "            Path to folder where the bigwig file will be saved.\n",
    "    bed_path: str\n",
    "            Path to folder where the fragments bed file will be saved.\n",
    "    sample_id_col: str, optional\n",
    "            Name of the column containing the sample name per barcode in the input :class:`CistopicObject.cell_data` or class:`pd.DataFrame`. Default: 'sample_id'.\n",
    "    normalize_bigwig: bool, optional\n",
    "            Whether bigwig files should be CPM normalized. Default: True.\n",
    "    remove_duplicates: bool, optional\n",
    "            Whether duplicates should be removed before converting the data to bigwig.\n",
    "    split_pattern: str\n",
    "            Pattern to split cell barcode from sample id. Default: ___ .\n",
    "    \"\"\"\n",
    "    # Create logger\n",
    "    level = logging.INFO\n",
    "    log_format = \"%(asctime)s %(name)-12s %(levelname)-8s %(message)s\"\n",
    "    handlers = [logging.StreamHandler(stream=sys.stdout)]\n",
    "    logging.basicConfig(level=level, format=log_format, handlers=handlers)\n",
    "    log = logging.getLogger(\"cisTopic\")\n",
    "\n",
    "    log.info(\"Creating pseudobulk for \" + str(group))\n",
    "    group_fragments_list = []\n",
    "    group_fragments_dict = {}\n",
    "    for sample_id in fragments_df_dict:\n",
    "        sample_data = cell_data[cell_data.loc[:, sample_id_col].isin([sample_id])]\n",
    "        if \"barcode\" in sample_data:\n",
    "            sample_data.index = sample_data[\"barcode\"].tolist()\n",
    "        else:\n",
    "            sample_data.index = prepare_tag_cells(\n",
    "                sample_data.index.tolist(), split_pattern\n",
    "            )\n",
    "        group_var = sample_data.iloc[:, 0]\n",
    "        barcodes = group_var[group_var.isin([group])].index.tolist()\n",
    "        fragments_df = fragments_df_dict[sample_id]\n",
    "        group_fragments = fragments_df.loc[fragments_df[\"Name\"].isin(barcodes)]\n",
    "        if len(fragments_df_dict) > 1:\n",
    "            group_fragments_dict[sample_id] = group_fragments\n",
    "\n",
    "    if len(fragments_df_dict) > 1:\n",
    "        group_fragments_list = [\n",
    "            group_fragments_dict[list(group_fragments_dict.keys())[x]]\n",
    "            for x in range(len(fragments_df_dict))\n",
    "        ]\n",
    "        group_fragments = group_fragments_list[0].append(group_fragments_list[1:])\n",
    "\n",
    "    group_fragments = group_fragments[\n",
    "        group_fragments[\"Chromosome\"].isin(chromsizes.Chromosome)\n",
    "    ]\n",
    "\n",
    "    del group_fragments_dict\n",
    "    del group_fragments_list\n",
    "    del fragments_df\n",
    "    gc.collect()\n",
    "\n",
    "    group_pr = pr.PyRanges(group_fragments)\n",
    "    if isinstance(bigwig_path, str):\n",
    "        bigwig_path_group = os.path.join(bigwig_path, str(group) + \".bw\")\n",
    "        if remove_duplicates:\n",
    "            group_pr.to_bigwig(\n",
    "                path=bigwig_path_group,\n",
    "                chromosome_sizes=chromsizes,\n",
    "                rpm=normalize_bigwig,\n",
    "            )\n",
    "        else:\n",
    "            group_pr.to_bigwig(\n",
    "                path=bigwig_path_group,\n",
    "                chromosome_sizes=chromsizes,\n",
    "                rpm=normalize_bigwig,\n",
    "                value_col=\"Score\",\n",
    "            )\n",
    "    if isinstance(bed_path, str):\n",
    "        bed_path_group = os.path.join(bed_path, str(group) + \".bed.gz\")\n",
    "        group_pr.to_bed(\n",
    "            path=bed_path_group, keep=False, compression=\"infer\", chain=False\n",
    "        )\n",
    "\n",
    "    log.info(str(group) + \" done!\")\n",
    "\n",
    "\n",
    "@ray.remote\n",
    "def export_pseudobulk_ray(\n",
    "    cell_data: pd.DataFrame,\n",
    "    group: str,\n",
    "    fragments_df_dict: Dict[str, pd.DataFrame],\n",
    "    chromsizes: pr.PyRanges,\n",
    "    bigwig_path: str,\n",
    "    bed_path: str,\n",
    "    sample_id_col: Optional[str] = \"sample_id\",\n",
    "    normalize_bigwig: Optional[bool] = True,\n",
    "    remove_duplicates: Optional[bool] = True,\n",
    "    split_pattern: Optional[str] = \"___\",\n",
    "):\n",
    "    \"\"\"\n",
    "    Create pseudobulk as bed and bigwig from single cell fragments file given a barcode annotation and a group.\n",
    "    Parameters\n",
    "    ---------\n",
    "    cell_data: pd.DataFrame\n",
    "            A cell metadata :class:`pd.Dataframe` containing barcodes, their annotation and their sample of origin.\n",
    "    group: str\n",
    "            A character string indicating the group for which pseudobulks will be created.\n",
    "    fragments_df_dict: dict\n",
    "            A dictionary containing data frames as values with 'Chromosome', 'Start', 'End', 'Name', and 'Score' as columns; and sample label\n",
    "            as keys. 'Score' indicates the number of times that a fragments is found assigned to that barcode.\n",
    "    chromsizes: pr.PyRanges\n",
    "            A :class:`pr.PyRanges` containing size of each column, containing 'Chromosome', 'Start' and 'End' columns.\n",
    "    bed_path: str\n",
    "            Path to folder where the fragments bed file will be saved.\n",
    "    bigwig_path: str\n",
    "            Path to folder where the bigwig file will be saved.\n",
    "    sample_id_col: str, optional\n",
    "            Name of the column containing the sample name per barcode in the input :class:`CistopicObject.cell_data` or class:`pd.DataFrame`. Default: 'sample_id'.\n",
    "    normalize_bigwig: bool, optional\n",
    "            Whether bigwig files should be CPM normalized. Default: True.\n",
    "    remove_duplicates: bool, optional\n",
    "            Whether duplicates should be removed before converting the data to bigwig.\n",
    "    split_pattern: str\n",
    "            Pattern to split cell barcode from sample id. Default: ___ .\n",
    "    \"\"\"\n",
    "    export_pseudobulk_one_sample(\n",
    "        cell_data,\n",
    "        group,\n",
    "        fragments_df_dict,\n",
    "        chromsizes,\n",
    "        bigwig_path,\n",
    "        bed_path,\n",
    "        sample_id_col,\n",
    "        normalize_bigwig,\n",
    "        remove_duplicates,\n",
    "        split_pattern,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2770e26d",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "source": [
    "# consensus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6d0fd585",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-03T15:00:49.482753Z",
     "iopub.status.busy": "2022-12-03T15:00:49.481914Z",
     "iopub.status.idle": "2022-12-03T15:00:49.504419Z",
     "shell.execute_reply": "2022-12-03T15:00:49.503259Z",
     "shell.execute_reply.started": "2022-12-03T15:00:49.482679Z"
    },
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': 'cistopic_objects/BIO_ddseq_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': 'cistopic_objects/BIO_ddseq_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_14topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': 'cistopic_objects/BIO_ddseq_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': 'cistopic_objects/BIO_ddseq_4.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': 'cistopic_objects/BRO_mtscatac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': 'cistopic_objects/BRO_mtscatac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_17topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': 'cistopic_objects/CNA_10xmultiome_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': 'cistopic_objects/CNA_10xmultiome_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_8topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': 'cistopic_objects/CNA_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': 'cistopic_objects/CNA_10xv11_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': 'cistopic_objects/CNA_10xv11_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': 'cistopic_objects/CNA_10xv11_4.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': 'cistopic_objects/CNA_10xv11_5.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': 'cistopic_objects/CNA_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': 'cistopic_objects/CNA_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': 'cistopic_objects/CNA_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': 'cistopic_objects/CNA_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': 'cistopic_objects/CNA_hydrop_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': 'cistopic_objects/CNA_mtscatac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': 'cistopic_objects/CNA_mtscatac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': 'cistopic_objects/EPF_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': 'cistopic_objects/EPF_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': 'cistopic_objects/EPF_hydrop_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_5topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': 'cistopic_objects/EPF_hydrop_4.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': 'cistopic_objects/HAR_ddseq_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': 'cistopic_objects/HAR_ddseq_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': 'cistopic_objects/MDC_mtscatac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_19topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'MDC_mtscatac_2.FIXEDCELLS': 'cistopic_objects/MDC_mtscatac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_11topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'OHS_s3atac_1.FIXEDCELLS': 'cistopic_objects/OHS_s3atac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'OHS_s3atac_2.FIXEDCELLS': 'cistopic_objects/OHS_s3atac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_7topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'SAN_10xmultiome_1.FIXEDCELLS': 'cistopic_objects/SAN_10xmultiome_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_7topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'SAN_10xmultiome_2.FIXEDCELLS': 'cistopic_objects/SAN_10xmultiome_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'STA_10xv11_1.FIXEDCELLS': 'cistopic_objects/STA_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'STA_10xv11_2.FIXEDCELLS': 'cistopic_objects/STA_10xv11_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_7topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'TXG_10xv11_1.FIXEDCELLS': 'cistopic_objects/TXG_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_14topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'TXG_10xv2_1.FIXEDCELLS': 'cistopic_objects/TXG_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_21topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'TXG_10xv2_2.FIXEDCELLS': 'cistopic_objects/TXG_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_21topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'UCS_ddseq_1.FIXEDCELLS': 'cistopic_objects/UCS_ddseq_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'UCS_ddseq_2.FIXEDCELLS': 'cistopic_objects/UCS_ddseq_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_10xmultiome_1.FIXEDCELLS': 'cistopic_objects/VIB_10xmultiome_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_9topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_10xmultiome_2.FIXEDCELLS': 'cistopic_objects/VIB_10xmultiome_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_10topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_10xv1_1.FIXEDCELLS': 'cistopic_objects/VIB_10xv1_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_12topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_10xv1_2.FIXEDCELLS': 'cistopic_objects/VIB_10xv1_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_19topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_10xv2_1.FIXEDCELLS': 'cistopic_objects/VIB_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_10xv2_2.FIXEDCELLS': 'cistopic_objects/VIB_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_13topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_hydrop_1.FIXEDCELLS': 'cistopic_objects/VIB_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_17topics.dimreduc.consensus.cell_data.tsv',\n",
       " 'VIB_hydrop_2.FIXEDCELLS': 'cistopic_objects/VIB_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.model_19topics.dimreduc.consensus.cell_data.tsv'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_data_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"cistopic_objects/*consensus.cell_data.tsv\"))\n",
    "}\n",
    "cell_data_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "baa38988",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-03T15:00:53.214474Z",
     "iopub.status.busy": "2022-12-03T15:00:53.213612Z",
     "iopub.status.idle": "2022-12-03T15:00:53.235441Z",
     "shell.execute_reply": "2022-12-03T15:00:53.234096Z",
     "shell.execute_reply.started": "2022-12-03T15:00:53.214414Z"
    },
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_3.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_4.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': 'final_consensus_peaks/BRO_mtscatac_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': 'final_consensus_peaks/BRO_mtscatac_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': 'final_consensus_peaks/CNA_10xmultiome_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': 'final_consensus_peaks/CNA_10xmultiome_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_3.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_4.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_5.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv2_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv2_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': 'final_consensus_peaks/CNA_hydrop_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': 'final_consensus_peaks/CNA_hydrop_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': 'final_consensus_peaks/CNA_hydrop_3.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': 'final_consensus_peaks/CNA_mtscatac_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': 'final_consensus_peaks/CNA_mtscatac_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_3.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_4.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': 'final_consensus_peaks/HAR_ddseq_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': 'final_consensus_peaks/HAR_ddseq_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': 'final_consensus_peaks/MDC_mtscatac_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'MDC_mtscatac_2.FIXEDCELLS': 'final_consensus_peaks/MDC_mtscatac_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'OHS_s3atac_1.FIXEDCELLS': 'final_consensus_peaks/OHS_s3atac_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'OHS_s3atac_2.FIXEDCELLS': 'final_consensus_peaks/OHS_s3atac_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'SAN_10xmultiome_1.FIXEDCELLS': 'final_consensus_peaks/SAN_10xmultiome_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'SAN_10xmultiome_2.FIXEDCELLS': 'final_consensus_peaks/SAN_10xmultiome_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'STA_10xv11_1.FIXEDCELLS': 'final_consensus_peaks/STA_10xv11_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'STA_10xv11_2.FIXEDCELLS': 'final_consensus_peaks/STA_10xv11_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'TXG_10xv11_1.FIXEDCELLS': 'final_consensus_peaks/TXG_10xv11_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'TXG_10xv2_1.FIXEDCELLS': 'final_consensus_peaks/TXG_10xv2_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'TXG_10xv2_2.FIXEDCELLS': 'final_consensus_peaks/TXG_10xv2_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'UCS_ddseq_1.FIXEDCELLS': 'final_consensus_peaks/UCS_ddseq_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'UCS_ddseq_2.FIXEDCELLS': 'final_consensus_peaks/UCS_ddseq_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_10xmultiome_1.FIXEDCELLS': 'final_consensus_peaks/VIB_10xmultiome_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_10xmultiome_2.FIXEDCELLS': 'final_consensus_peaks/VIB_10xmultiome_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_10xv1_1.FIXEDCELLS': 'final_consensus_peaks/VIB_10xv1_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_10xv1_2.FIXEDCELLS': 'final_consensus_peaks/VIB_10xv1_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_10xv2_1.FIXEDCELLS': 'final_consensus_peaks/VIB_10xv2_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_10xv2_2.FIXEDCELLS': 'final_consensus_peaks/VIB_10xv2_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_hydrop_1.FIXEDCELLS': 'final_consensus_peaks/VIB_hydrop_1.FIXEDCELLS__SCREEN_pseudobulk_bw_files',\n",
       " 'VIB_hydrop_2.FIXEDCELLS': 'final_consensus_peaks/VIB_hydrop_2.FIXEDCELLS__SCREEN_pseudobulk_bw_files'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bw_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"final_consensus_peaks/*_pseudobulk_bw_files\"))\n",
    "}\n",
    "bw_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c37d0454",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-03T15:00:55.658518Z",
     "iopub.status.busy": "2022-12-03T15:00:55.657729Z",
     "iopub.status.idle": "2022-12-03T15:00:55.678924Z",
     "shell.execute_reply": "2022-12-03T15:00:55.677599Z",
     "shell.execute_reply.started": "2022-12-03T15:00:55.658460Z"
    },
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_3.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_4.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': 'final_consensus_peaks/BRO_mtscatac_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': 'final_consensus_peaks/BRO_mtscatac_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': 'final_consensus_peaks/CNA_10xmultiome_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': 'final_consensus_peaks/CNA_10xmultiome_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_3.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_4.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_5.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv2_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv2_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': 'final_consensus_peaks/CNA_hydrop_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': 'final_consensus_peaks/CNA_hydrop_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': 'final_consensus_peaks/CNA_hydrop_3.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': 'final_consensus_peaks/CNA_mtscatac_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': 'final_consensus_peaks/CNA_mtscatac_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_3.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_4.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': 'final_consensus_peaks/HAR_ddseq_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': 'final_consensus_peaks/HAR_ddseq_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': 'final_consensus_peaks/MDC_mtscatac_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'MDC_mtscatac_2.FIXEDCELLS': 'final_consensus_peaks/MDC_mtscatac_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'OHS_s3atac_1.FIXEDCELLS': 'final_consensus_peaks/OHS_s3atac_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'OHS_s3atac_2.FIXEDCELLS': 'final_consensus_peaks/OHS_s3atac_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'SAN_10xmultiome_1.FIXEDCELLS': 'final_consensus_peaks/SAN_10xmultiome_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'SAN_10xmultiome_2.FIXEDCELLS': 'final_consensus_peaks/SAN_10xmultiome_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'STA_10xv11_1.FIXEDCELLS': 'final_consensus_peaks/STA_10xv11_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'STA_10xv11_2.FIXEDCELLS': 'final_consensus_peaks/STA_10xv11_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'TXG_10xv11_1.FIXEDCELLS': 'final_consensus_peaks/TXG_10xv11_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'TXG_10xv2_1.FIXEDCELLS': 'final_consensus_peaks/TXG_10xv2_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'TXG_10xv2_2.FIXEDCELLS': 'final_consensus_peaks/TXG_10xv2_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'UCS_ddseq_1.FIXEDCELLS': 'final_consensus_peaks/UCS_ddseq_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'UCS_ddseq_2.FIXEDCELLS': 'final_consensus_peaks/UCS_ddseq_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_10xmultiome_1.FIXEDCELLS': 'final_consensus_peaks/VIB_10xmultiome_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_10xmultiome_2.FIXEDCELLS': 'final_consensus_peaks/VIB_10xmultiome_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_10xv1_1.FIXEDCELLS': 'final_consensus_peaks/VIB_10xv1_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_10xv1_2.FIXEDCELLS': 'final_consensus_peaks/VIB_10xv1_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_10xv2_1.FIXEDCELLS': 'final_consensus_peaks/VIB_10xv2_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_10xv2_2.FIXEDCELLS': 'final_consensus_peaks/VIB_10xv2_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_hydrop_1.FIXEDCELLS': 'final_consensus_peaks/VIB_hydrop_1.FIXEDCELLS__SCREEN_pseudobulk_bed_files',\n",
       " 'VIB_hydrop_2.FIXEDCELLS': 'final_consensus_peaks/VIB_hydrop_2.FIXEDCELLS__SCREEN_pseudobulk_bed_files'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bed_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"final_consensus_peaks/*_pseudobulk_bed_files\"))\n",
    "}\n",
    "bed_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "56b6489a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-03T15:00:57.919302Z",
     "iopub.status.busy": "2022-12-03T15:00:57.918501Z",
     "iopub.status.idle": "2022-12-03T15:00:57.949769Z",
     "shell.execute_reply": "2022-12-03T15:00:57.948418Z",
     "shell.execute_reply.started": "2022-12-03T15:00:57.919244Z"
    },
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "for sample in bed_path_dict.keys():\n",
    "    bed_paths = {\n",
    "        x.split(\"/\")[-1].split(\"__\")[0].split(\".bed.gz\")[0]: x\n",
    "        for x in glob.glob(bed_path_dict[sample] + \"/*\")\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "1c214f18",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-03T15:00:58.131747Z",
     "iopub.status.busy": "2022-12-03T15:00:58.130954Z",
     "iopub.status.idle": "2022-12-03T15:00:58.139990Z",
     "shell.execute_reply": "2022-12-03T15:00:58.138526Z",
     "shell.execute_reply.started": "2022-12-03T15:00:58.131672Z"
    },
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pycisTopic.pseudobulk_peak_calling import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f8e4b28a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-03T15:00:58.551947Z",
     "iopub.status.busy": "2022-12-03T15:00:58.551151Z",
     "iopub.status.idle": "2022-12-03T15:00:58.559418Z",
     "shell.execute_reply": "2022-12-03T15:00:58.557995Z",
     "shell.execute_reply.started": "2022-12-03T15:00:58.551890Z"
    },
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import ray"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9eb34ce",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "source": [
    "# call consensus peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "527a07d4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-03T15:01:00.898593Z",
     "iopub.status.busy": "2022-12-03T15:01:00.897735Z",
     "iopub.status.idle": "2022-12-03T15:01:00.910971Z",
     "shell.execute_reply": "2022-12-03T15:01:00.909446Z",
     "shell.execute_reply.started": "2022-12-03T15:01:00.898533Z"
    },
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pycisTopic.iterative_peak_calling import get_consensus_peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9e5204aa",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-03T15:01:01.450640Z",
     "iopub.status.busy": "2022-12-03T15:01:01.449855Z",
     "iopub.status.idle": "2022-12-03T15:01:01.472052Z",
     "shell.execute_reply": "2022-12-03T15:01:01.470729Z",
     "shell.execute_reply.started": "2022-12-03T15:01:01.450582Z"
    },
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_3.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_4.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': 'final_consensus_peaks/BRO_mtscatac_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': 'final_consensus_peaks/BRO_mtscatac_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': 'final_consensus_peaks/CNA_10xmultiome_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': 'final_consensus_peaks/CNA_10xmultiome_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_3.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_4.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_5.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv2_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv2_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': 'final_consensus_peaks/CNA_hydrop_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': 'final_consensus_peaks/CNA_hydrop_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': 'final_consensus_peaks/CNA_hydrop_3.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': 'final_consensus_peaks/CNA_mtscatac_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': 'final_consensus_peaks/CNA_mtscatac_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_3.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_4.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': 'final_consensus_peaks/HAR_ddseq_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': 'final_consensus_peaks/HAR_ddseq_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': 'final_consensus_peaks/MDC_mtscatac_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'MDC_mtscatac_2.FIXEDCELLS': 'final_consensus_peaks/MDC_mtscatac_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'OHS_s3atac_1.FIXEDCELLS': 'final_consensus_peaks/OHS_s3atac_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'OHS_s3atac_2.FIXEDCELLS': 'final_consensus_peaks/OHS_s3atac_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'SAN_10xmultiome_1.FIXEDCELLS': 'final_consensus_peaks/SAN_10xmultiome_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'SAN_10xmultiome_2.FIXEDCELLS': 'final_consensus_peaks/SAN_10xmultiome_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'STA_10xv11_1.FIXEDCELLS': 'final_consensus_peaks/STA_10xv11_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'STA_10xv11_2.FIXEDCELLS': 'final_consensus_peaks/STA_10xv11_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'TXG_10xv11_1.FIXEDCELLS': 'final_consensus_peaks/TXG_10xv11_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'TXG_10xv2_1.FIXEDCELLS': 'final_consensus_peaks/TXG_10xv2_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'TXG_10xv2_2.FIXEDCELLS': 'final_consensus_peaks/TXG_10xv2_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'UCS_ddseq_1.FIXEDCELLS': 'final_consensus_peaks/UCS_ddseq_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'UCS_ddseq_2.FIXEDCELLS': 'final_consensus_peaks/UCS_ddseq_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_10xmultiome_1.FIXEDCELLS': 'final_consensus_peaks/VIB_10xmultiome_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_10xmultiome_2.FIXEDCELLS': 'final_consensus_peaks/VIB_10xmultiome_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_10xv1_1.FIXEDCELLS': 'final_consensus_peaks/VIB_10xv1_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_10xv1_2.FIXEDCELLS': 'final_consensus_peaks/VIB_10xv1_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_10xv2_1.FIXEDCELLS': 'final_consensus_peaks/VIB_10xv2_1.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_10xv2_2.FIXEDCELLS': 'final_consensus_peaks/VIB_10xv2_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl',\n",
       " 'VIB_hydrop_2.FIXEDCELLS': 'final_consensus_peaks/VIB_hydrop_2.FIXEDCELLS__SCREEN_narrow_peaks_dict.pkl'}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "narrow_peaks_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"final_consensus_peaks/*_narrow_peaks_dict.pkl\"))\n",
    "}\n",
    "narrow_peaks_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "7a51bd15",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "path_to_blacklist = \"../0_resources/regions/hg38-blacklist.v2.bed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "2adbdafd",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-03T15:01:03.684305Z",
     "iopub.status.busy": "2022-12-03T15:01:03.683497Z",
     "iopub.status.idle": "2022-12-03T15:01:03.721408Z",
     "shell.execute_reply": "2022-12-03T15:01:03.720356Z",
     "shell.execute_reply.started": "2022-12-03T15:01:03.684247Z"
    },
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BIO_ddseq_1.FIXEDCELLS\n",
      "final_consensus_peaks/BIO_ddseq_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "BIO_ddseq_2.FIXEDCELLS\n",
      "final_consensus_peaks/BIO_ddseq_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "BIO_ddseq_3.FIXEDCELLS\n",
      "final_consensus_peaks/BIO_ddseq_3.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "BIO_ddseq_4.FIXEDCELLS\n",
      "final_consensus_peaks/BIO_ddseq_4.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "BRO_mtscatac_1.FIXEDCELLS\n",
      "final_consensus_peaks/BRO_mtscatac_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "BRO_mtscatac_2.FIXEDCELLS\n",
      "final_consensus_peaks/BRO_mtscatac_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xmultiome_1.FIXEDCELLS\n",
      "final_consensus_peaks/CNA_10xmultiome_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xmultiome_2.FIXEDCELLS\n",
      "final_consensus_peaks/CNA_10xmultiome_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xv11_1.FIXEDCELLS\n",
      "final_consensus_peaks/CNA_10xv11_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xv11_2.FIXEDCELLS\n",
      "final_consensus_peaks/CNA_10xv11_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xv11_3.FIXEDCELLS\n",
      "final_consensus_peaks/CNA_10xv11_3.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xv11_4.FIXEDCELLS\n",
      "final_consensus_peaks/CNA_10xv11_4.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xv11_5.FIXEDCELLS\n",
      "final_consensus_peaks/CNA_10xv11_5.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xv2_1.FIXEDCELLS\n",
      "final_consensus_peaks/CNA_10xv2_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_10xv2_2.FIXEDCELLS\n",
      "final_consensus_peaks/CNA_10xv2_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_hydrop_1.FIXEDCELLS\n",
      "final_consensus_peaks/CNA_hydrop_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_hydrop_2.FIXEDCELLS\n",
      "final_consensus_peaks/CNA_hydrop_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_hydrop_3.FIXEDCELLS\n",
      "final_consensus_peaks/CNA_hydrop_3.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_mtscatac_1.FIXEDCELLS\n",
      "final_consensus_peaks/CNA_mtscatac_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "CNA_mtscatac_2.FIXEDCELLS\n",
      "final_consensus_peaks/CNA_mtscatac_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "EPF_hydrop_1.FIXEDCELLS\n",
      "final_consensus_peaks/EPF_hydrop_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "EPF_hydrop_2.FIXEDCELLS\n",
      "final_consensus_peaks/EPF_hydrop_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "EPF_hydrop_3.FIXEDCELLS\n",
      "final_consensus_peaks/EPF_hydrop_3.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "EPF_hydrop_4.FIXEDCELLS\n",
      "final_consensus_peaks/EPF_hydrop_4.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "HAR_ddseq_1.FIXEDCELLS\n",
      "final_consensus_peaks/HAR_ddseq_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "HAR_ddseq_2.FIXEDCELLS\n",
      "final_consensus_peaks/HAR_ddseq_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "MDC_mtscatac_1.FIXEDCELLS\n",
      "final_consensus_peaks/MDC_mtscatac_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "MDC_mtscatac_2.FIXEDCELLS\n",
      "final_consensus_peaks/MDC_mtscatac_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "OHS_s3atac_1.FIXEDCELLS\n",
      "final_consensus_peaks/OHS_s3atac_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "OHS_s3atac_2.FIXEDCELLS\n",
      "final_consensus_peaks/OHS_s3atac_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "SAN_10xmultiome_1.FIXEDCELLS\n",
      "final_consensus_peaks/SAN_10xmultiome_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "SAN_10xmultiome_2.FIXEDCELLS\n",
      "final_consensus_peaks/SAN_10xmultiome_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "STA_10xv11_1.FIXEDCELLS\n",
      "final_consensus_peaks/STA_10xv11_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "STA_10xv11_2.FIXEDCELLS\n",
      "final_consensus_peaks/STA_10xv11_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "TXG_10xv11_1.FIXEDCELLS\n",
      "final_consensus_peaks/TXG_10xv11_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "TXG_10xv2_1.FIXEDCELLS\n",
      "final_consensus_peaks/TXG_10xv2_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "TXG_10xv2_2.FIXEDCELLS\n",
      "final_consensus_peaks/TXG_10xv2_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "UCS_ddseq_1.FIXEDCELLS\n",
      "final_consensus_peaks/UCS_ddseq_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "UCS_ddseq_2.FIXEDCELLS\n",
      "final_consensus_peaks/UCS_ddseq_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_10xmultiome_1.FIXEDCELLS\n",
      "final_consensus_peaks/VIB_10xmultiome_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_10xmultiome_2.FIXEDCELLS\n",
      "final_consensus_peaks/VIB_10xmultiome_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_10xv1_1.FIXEDCELLS\n",
      "final_consensus_peaks/VIB_10xv1_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_10xv1_2.FIXEDCELLS\n",
      "final_consensus_peaks/VIB_10xv1_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_10xv2_1.FIXEDCELLS\n",
      "final_consensus_peaks/VIB_10xv2_1.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_10xv2_2.FIXEDCELLS\n",
      "final_consensus_peaks/VIB_10xv2_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n",
      "VIB_hydrop_2.FIXEDCELLS\n",
      "final_consensus_peaks/VIB_hydrop_2.FIXEDCELLS__SCREEN_consensus_peaks.bed already exists, skipping...\n"
     ]
    }
   ],
   "source": [
    "peak_half_width = 250\n",
    "\n",
    "# Get consensus peaks\n",
    "consensus_peaks_dict = {}\n",
    "for sample in narrow_peaks_path_dict.keys():\n",
    "    print(sample)\n",
    "    consensus_out_path = narrow_peaks_path_dict[sample].replace(\n",
    "        \"_narrow_peaks_dict.pkl\", \"_consensus_peaks.bed\"\n",
    "    )\n",
    "    if not os.path.exists(consensus_out_path):\n",
    "        cell_data = pd.read_csv(cell_data_path_dict[sample])\n",
    "        cto_celltypes = set(\n",
    "            [\n",
    "                x.replace(\" \", \"\").replace(\"+\", \"\").replace(\"_\", \"\")\n",
    "                for x in set(cell_data[\"consensus_cell_type\"].unique())\n",
    "            ]\n",
    "        )\n",
    "\n",
    "        with open(narrow_peaks_path_dict[sample], \"rb\") as f:\n",
    "            narrow_peaks_dict = pickle.load(f)\n",
    "        peaks_celltypes = set(\n",
    "            [\n",
    "                x.replace(\" \", \"\").replace(\"+\", \"\").replace(\"_\", \"\")\n",
    "                for x in set(narrow_peaks_dict.keys())\n",
    "            ]\n",
    "        )\n",
    "\n",
    "        if cto_celltypes == peaks_celltypes:\n",
    "            consensus_peaks = get_consensus_peaks(\n",
    "                narrow_peaks_dict,\n",
    "                peak_half_width,\n",
    "                chromsizes=chromsizes,\n",
    "                path_to_blacklist=path_to_blacklist,\n",
    "            )\n",
    "\n",
    "            consensus_peaks.to_bed(\n",
    "                path=consensus_out_path, keep=True, compression=\"infer\", chain=False\n",
    "            )\n",
    "        else:\n",
    "            print(\"CELL TYPE SETS NOT MATCHING! Rerun peak calling.\")\n",
    "            print(peaks_celltypes - cto_celltypes)\n",
    "            print(cto_celltypes - peaks_celltypes)\n",
    "    else:\n",
    "        print(f\"{consensus_out_path} already exists, skipping...\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4a840ce",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "source": [
    "# Check % chrM in consensus peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "cadfcdd4",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_3.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': 'final_consensus_peaks/BIO_ddseq_4.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': 'final_consensus_peaks/BRO_mtscatac_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': 'final_consensus_peaks/BRO_mtscatac_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': 'final_consensus_peaks/CNA_10xmultiome_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': 'final_consensus_peaks/CNA_10xmultiome_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_3.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_4.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv11_5.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv2_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': 'final_consensus_peaks/CNA_10xv2_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': 'final_consensus_peaks/CNA_hydrop_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': 'final_consensus_peaks/CNA_hydrop_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': 'final_consensus_peaks/CNA_hydrop_3.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': 'final_consensus_peaks/CNA_mtscatac_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': 'final_consensus_peaks/CNA_mtscatac_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_3.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': 'final_consensus_peaks/EPF_hydrop_4.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': 'final_consensus_peaks/HAR_ddseq_1.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': 'final_consensus_peaks/HAR_ddseq_2.FIXEDCELLS__SCREEN_consensus_peaks.bed',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': 'final_consensus_peaks/MDC_mtscatac_1.FIXEDCELLS__SCREEN_consensus_peaks.bed'}"
      ]
     },
     "execution_count": 138,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "consensus_peaks_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\"__\")[0]: x\n",
    "    for x in sorted(glob.glob(\"final_consensus_peaks/*consensus_peaks.bed\"))\n",
    "}\n",
    "consensus_peaks_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "id": "f4e4c1c9",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BIO_ddseq_1.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 128186\n",
      "\tpeaks on contigs: 115\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.10754028864436893%\n",
      "\tpeaks on chrY: 0\n",
      "BIO_ddseq_2.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 126797\n",
      "\tpeaks on contigs: 55\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.060690133518293735%\n",
      "\tpeaks on chrY: 0\n",
      "BIO_ddseq_3.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 167832\n",
      "\tpeaks on contigs: 242\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.15764707282104976%\n",
      "\tpeaks on chrY: 170\n",
      "BIO_ddseq_4.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 154291\n",
      "\tpeaks on contigs: 245\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.1733965670067741%\n",
      "\tpeaks on chrY: 176\n",
      "BRO_mtscatac_1.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 202745\n",
      "\tpeaks on contigs: 254\n",
      "\tpeaks on chrM: 21\n",
      "\t% peaks non standard chromosomes: 0.13545463501132893%\n",
      "\tpeaks on chrY: 147\n",
      "BRO_mtscatac_2.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 185765\n",
      "\tpeaks on contigs: 224\n",
      "\tpeaks on chrM: 21\n",
      "\t% peaks non standard chromosomes: 0.13171334874469115%\n",
      "\tpeaks on chrY: 132\n",
      "CNA_10xmultiome_1.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 181498\n",
      "\tpeaks on contigs: 229\n",
      "\tpeaks on chrM: 21\n",
      "\t% peaks non standard chromosomes: 0.13755309549486103%\n",
      "\tpeaks on chrY: 105\n",
      "CNA_10xmultiome_2.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 172549\n",
      "\tpeaks on contigs: 319\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.19781249457750835%\n",
      "\tpeaks on chrY: 99\n",
      "CNA_10xv11_1.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 174810\n",
      "\tpeaks on contigs: 293\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.18044151068373626%\n",
      "\tpeaks on chrY: 125\n",
      "CNA_10xv11_2.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 167220\n",
      "\tpeaks on contigs: 216\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.14212518959978027%\n",
      "\tpeaks on chrY: 103\n",
      "CNA_10xv11_3.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 224321\n",
      "\tpeaks on contigs: 455\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.21263439783984805%\n",
      "\tpeaks on chrY: 140\n",
      "CNA_10xv11_4.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 160128\n",
      "\tpeaks on contigs: 176\n",
      "\tpeaks on chrM: 24\n",
      "\t% peaks non standard chromosomes: 0.12474427423781248%\n",
      "\tpeaks on chrY: 79\n",
      "CNA_10xv11_5.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 200550\n",
      "\tpeaks on contigs: 324\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.17272532690881398%\n",
      "\tpeaks on chrY: 99\n",
      "CNA_10xv2_1.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 243541\n",
      "\tpeaks on contigs: 270\n",
      "\tpeaks on chrM: 24\n",
      "\t% peaks non standard chromosomes: 0.12057333852810302%\n",
      "\tpeaks on chrY: 180\n",
      "CNA_10xv2_2.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 277961\n",
      "\tpeaks on contigs: 341\n",
      "\tpeaks on chrM: 24\n",
      "\t% peaks non standard chromosomes: 0.13114117976761064%\n",
      "\tpeaks on chrY: 204\n",
      "CNA_hydrop_1.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 91705\n",
      "\tpeaks on contigs: 275\n",
      "\tpeaks on chrM: 21\n",
      "\t% peaks non standard chromosomes: 0.3217356333083336%\n",
      "\tpeaks on chrY: 58\n",
      "CNA_hydrop_2.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 154168\n",
      "\tpeaks on contigs: 427\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.291039853057212%\n",
      "\tpeaks on chrY: 118\n",
      "CNA_hydrop_3.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 87897\n",
      "\tpeaks on contigs: 219\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.27343484081780844%\n",
      "\tpeaks on chrY: 57\n",
      "CNA_mtscatac_1.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 108214\n",
      "\tpeaks on contigs: 153\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.16237660300765755%\n",
      "\tpeaks on chrY: 66\n",
      "CNA_mtscatac_2.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 133139\n",
      "\tpeaks on contigs: 156\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.13426544052566045%\n",
      "\tpeaks on chrY: 56\n",
      "EPF_hydrop_1.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 133302\n",
      "\tpeaks on contigs: 286\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.23052166753985479%\n",
      "\tpeaks on chrY: 61\n",
      "EPF_hydrop_2.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 123022\n",
      "\tpeaks on contigs: 271\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.2376028869156226%\n",
      "\tpeaks on chrY: 42\n",
      "EPF_hydrop_3.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 123157\n",
      "\tpeaks on contigs: 252\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.22198637295330997%\n",
      "\tpeaks on chrY: 75\n",
      "EPF_hydrop_4.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 108773\n",
      "\tpeaks on contigs: 204\n",
      "\tpeaks on chrM: 22\n",
      "\t% peaks non standard chromosomes: 0.2073413517555207%\n",
      "\tpeaks on chrY: 51\n",
      "HAR_ddseq_1.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 163972\n",
      "\tpeaks on contigs: 199\n",
      "\tpeaks on chrM: 24\n",
      "\t% peaks non standard chromosomes: 0.135814123450775%\n",
      "\tpeaks on chrY: 97\n",
      "HAR_ddseq_2.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 164991\n",
      "\tpeaks on contigs: 199\n",
      "\tpeaks on chrM: 23\n",
      "\t% peaks non standard chromosomes: 0.1343719925187485%\n",
      "\tpeaks on chrY: 93\n",
      "MDC_mtscatac_1.FIXEDCELLS\n",
      "\tpeaks on standard chromosomes: 234413\n",
      "\tpeaks on contigs: 535\n",
      "\tpeaks on chrM: 21\n",
      "\t% peaks non standard chromosomes: 0.23662695930101418%\n",
      "\tpeaks on chrY: 137\n"
     ]
    }
   ],
   "source": [
    "for sample, path in consensus_peaks_path_dict.items():\n",
    "    print(sample)\n",
    "    peaks_df = pd.read_csv(path, sep=\"\\t\", header=None)\n",
    "\n",
    "    chroms_in_df = list(sorted(peaks_df[0].unique()))\n",
    "    chroms_standard = [\"chr\" + str(x + 1) for x in range(22)] + [\"chrX\"]\n",
    "    chroms_nonstandard = list(set(chroms_in_df) - set(chroms_standard) - set([\"chrM\"]))\n",
    "\n",
    "    n_standard = peaks_df[0].value_counts()[chroms_standard].sum()\n",
    "    n_contigs = peaks_df[0].value_counts()[chroms_nonstandard].sum()\n",
    "    n_chrm = peaks_df[0].value_counts()[\"chrM\"].sum()\n",
    "    pct_nonstandard = (n_contigs + n_chrm) / len(peaks_df) * 100\n",
    "\n",
    "    print(f\"\\tpeaks on standard chromosomes: {n_standard}\")\n",
    "    print(f\"\\tpeaks on contigs: {n_contigs}\")\n",
    "    print(f\"\\tpeaks on chrM: {n_chrm}\")\n",
    "    print(f\"\\t% peaks non standard chromosomes: {pct_nonstandard}%\")\n",
    "    if \"chrY\" in chroms_in_df:\n",
    "        n_chrY = peaks_df[0].value_counts()[\"chrY\"].sum()\n",
    "        print(f\"\\tpeaks on chrY: {n_chrY}\")\n",
    "    else:\n",
    "        print(f\"\\tpeaks on chrY: 0\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4b81468",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8f81d58",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "69683456",
   "metadata": {
    "papermill": {
     "duration": null,
     "end_time": null,
     "exception": null,
     "start_time": null,
     "status": "pending"
    },
    "tags": []
   },
   "source": [
    "# run seurat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3f53dbf7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frags_path_dict = {\n",
    "    x.split(\"/\")[-1].split(f\".fragments.tsv.gz\")[0]: x\n",
    "    for x in sorted(\n",
    "        glob.glob(f\"*k/*k_preprocessing_out/data/fragments/*fragments.tsv.gz\")\n",
    "    )\n",
    "}\n",
    "frags_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a0a8a538",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-17T13:22:20.942499Z",
     "iopub.status.busy": "2022-09-17T13:22:20.941596Z",
     "iopub.status.idle": "2022-09-17T13:22:20.955100Z",
     "shell.execute_reply": "2022-09-17T13:22:20.953479Z",
     "shell.execute_reply.started": "2022-09-17T13:22:20.942436Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BIO_ddseq_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BIO_ddseq_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BIO_ddseq_3.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BIO_ddseq_4.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BRO_mtscatac_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/BRO_mtscatac_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xmultiome_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xmultiome_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv11_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv11_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv11_3.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv11_4.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv11_5.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv2_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_10xv2_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_hydrop_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_hydrop_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_hydrop_3.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_mtscatac_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/CNA_mtscatac_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/EPF_hydrop_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/EPF_hydrop_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/EPF_hydrop_3.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/EPF_hydrop_4.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/HAR_ddseq_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/HAR_ddseq_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/MDC_mtscatac_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/MDC_mtscatac_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'OHS_s3atac_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/OHS_s3atac_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'OHS_s3atac_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/OHS_s3atac_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/SAN_10xmultiome_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/SAN_10xmultiome_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'STA_10xv11_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/STA_10xv11_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'STA_10xv11_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/STA_10xv11_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'TXG_10xv11_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/TXG_10xv11_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'TXG_10xv2_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/TXG_10xv2_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'TXG_10xv2_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/TXG_10xv2_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'UCS_ddseq_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/UCS_ddseq_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'UCS_ddseq_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/UCS_ddseq_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xmultiome_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xmultiome_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xv1_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xv1_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xv1_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xv1_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xv2_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xv2_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_10xv2_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_10xv2_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_1.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_1.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_11.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_11.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_12.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_12.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_2.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_2.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_21.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_21.FIXEDCELLS.fragments.tsv.gz',\n",
       " 'VIB_hydrop_22.FIXEDCELLS': '../1_data_repository/fixedcells_fragments/VIB_hydrop_22.FIXEDCELLS.fragments.tsv.gz'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scrub_name_suffix = \"0-4\"\n",
    "frags_path_dict  = {x.split('/')[-1].split(f'.fragments.tsv.gz')[0] :x for x in sorted(glob.glob(f'../1_data_repository/fixedcells_fragments/*.fragments.tsv.gz'))}\n",
    "frags_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f85e6982",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-17T13:22:24.988121Z",
     "iopub.status.busy": "2022-09-17T13:22:24.987742Z",
     "iopub.status.idle": "2022-09-17T13:22:24.998155Z",
     "shell.execute_reply": "2022-09-17T13:22:24.997470Z",
     "shell.execute_reply.started": "2022-09-17T13:22:24.988091Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.FIXEDCELLS': 'cell_region_looms/BIO_ddseq_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'BIO_ddseq_2.FIXEDCELLS': 'cell_region_looms/BIO_ddseq_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'BIO_ddseq_3.FIXEDCELLS': 'cell_region_looms/BIO_ddseq_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'BIO_ddseq_4.FIXEDCELLS': 'cell_region_looms/BIO_ddseq_4.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'BRO_mtscatac_1.FIXEDCELLS': 'cell_region_looms/BRO_mtscatac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'BRO_mtscatac_2.FIXEDCELLS': 'cell_region_looms/BRO_mtscatac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xmultiome_1.FIXEDCELLS': 'cell_region_looms/CNA_10xmultiome_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xmultiome_2.FIXEDCELLS': 'cell_region_looms/CNA_10xmultiome_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xv11_1.FIXEDCELLS': 'cell_region_looms/CNA_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xv11_2.FIXEDCELLS': 'cell_region_looms/CNA_10xv11_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xv11_3.FIXEDCELLS': 'cell_region_looms/CNA_10xv11_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xv11_4.FIXEDCELLS': 'cell_region_looms/CNA_10xv11_4.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xv11_5.FIXEDCELLS': 'cell_region_looms/CNA_10xv11_5.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xv2_1.FIXEDCELLS': 'cell_region_looms/CNA_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_10xv2_2.FIXEDCELLS': 'cell_region_looms/CNA_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_hydrop_1.FIXEDCELLS': 'cell_region_looms/CNA_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_hydrop_2.FIXEDCELLS': 'cell_region_looms/CNA_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_hydrop_3.FIXEDCELLS': 'cell_region_looms/CNA_hydrop_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_mtscatac_1.FIXEDCELLS': 'cell_region_looms/CNA_mtscatac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'CNA_mtscatac_2.FIXEDCELLS': 'cell_region_looms/CNA_mtscatac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'EPF_hydrop_1.FIXEDCELLS': 'cell_region_looms/EPF_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'EPF_hydrop_2.FIXEDCELLS': 'cell_region_looms/EPF_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'EPF_hydrop_3.FIXEDCELLS': 'cell_region_looms/EPF_hydrop_3.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'EPF_hydrop_4.FIXEDCELLS': 'cell_region_looms/EPF_hydrop_4.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'HAR_ddseq_1.FIXEDCELLS': 'cell_region_looms/HAR_ddseq_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'HAR_ddseq_2.FIXEDCELLS': 'cell_region_looms/HAR_ddseq_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'MDC_mtscatac_1.FIXEDCELLS': 'cell_region_looms/MDC_mtscatac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'MDC_mtscatac_2.FIXEDCELLS': 'cell_region_looms/MDC_mtscatac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'OHS_s3atac_1.FIXEDCELLS': 'cell_region_looms/OHS_s3atac_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'OHS_s3atac_2.FIXEDCELLS': 'cell_region_looms/OHS_s3atac_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'SAN_10xmultiome_1.FIXEDCELLS': 'cell_region_looms/SAN_10xmultiome_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'SAN_10xmultiome_2.FIXEDCELLS': 'cell_region_looms/SAN_10xmultiome_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'STA_10xv11_1.FIXEDCELLS': 'cell_region_looms/STA_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'STA_10xv11_2.FIXEDCELLS': 'cell_region_looms/STA_10xv11_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'TXG_10xv11_1.FIXEDCELLS': 'cell_region_looms/TXG_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'TXG_10xv2_1.FIXEDCELLS': 'cell_region_looms/TXG_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'TXG_10xv2_2.FIXEDCELLS': 'cell_region_looms/TXG_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'UCS_ddseq_1.FIXEDCELLS': 'cell_region_looms/UCS_ddseq_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'UCS_ddseq_2.FIXEDCELLS': 'cell_region_looms/UCS_ddseq_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_10xmultiome_1.FIXEDCELLS': 'cell_region_looms/VIB_10xmultiome_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_10xmultiome_2.FIXEDCELLS': 'cell_region_looms/VIB_10xmultiome_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_10xv1_1.FIXEDCELLS': 'cell_region_looms/VIB_10xv1_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_10xv1_2.FIXEDCELLS': 'cell_region_looms/VIB_10xv1_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_10xv2_1.FIXEDCELLS': 'cell_region_looms/VIB_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_10xv2_2.FIXEDCELLS': 'cell_region_looms/VIB_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_hydrop_1.FIXEDCELLS': 'cell_region_looms/VIB_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_hydrop_11.FIXEDCELLS': 'cell_region_looms/VIB_hydrop_11.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_hydrop_12.FIXEDCELLS': 'cell_region_looms/VIB_hydrop_12.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_hydrop_2.FIXEDCELLS': 'cell_region_looms/VIB_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_hydrop_21.FIXEDCELLS': 'cell_region_looms/VIB_hydrop_21.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom',\n",
       " 'VIB_hydrop_22.FIXEDCELLS': 'cell_region_looms/VIB_hydrop_22.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scrub_name_suffix = \"0-4\"\n",
    "loom_path_dict  = {x.split('/')[-1].split(f'__')[0]:x for x in sorted(glob.glob(f'cell_region_looms/*singlets.loom'))}\n",
    "loom_path_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "8ca7a963",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#!/usr/bin/env Rscript\n",
      "library(Seurat)\n",
      "library(SeuratDisk)\n",
      "library(Signac)\n",
      "# library(EnsDb.Hsapiens.v86)\n",
      "library(ggplot2)\n",
      "library(stringr)\n",
      "\n",
      "args = commandArgs(trailingOnly=TRUE)\n",
      "sample_id = args[1]\n",
      "f_loom = args[2]\n",
      "f_frag = args[3]\n",
      "f_reference = args[4]\n",
      "f_annotation = args[5]\n",
      "f_out = args[6]\n",
      "\n",
      "print(paste0(\"Processing sample \", args[1]))\n",
      "\n",
      "# load pbmc object\n",
      "pbmc.rna <- readRDS(\"../0_resources/seurat_references/pbmc_integrated.RDS\")\n",
      "#pbmc.rna <- readRDS('/lustre1/project/stg_00090/scatac_benchmark/0_resources/seurat_references/pbmc_ssc_mat__integrated.rds')\n",
      "\n",
      "################################################################################\n",
      "# ATAC\n",
      "################################################################################\n",
      "\n",
      "### get data from loom:\n",
      "atacloomcon <- Connect(filename = f_loom, mode = \"r\")\n",
      "atacloomcon\n",
      "atac_tmp <- as.Seurat(atacloomcon, assay='ATAC')\n",
      "atacloomcon$close_all()\n",
      "\n",
      "# subset by removing contig chromosomes\n",
      "rawregions = rownames(GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC'))\n",
      "standard_regions = rawregions[!grepl('random|KI|Un', rawregions)]\n",
      "\n",
      "atac_tmp_data_subset = GetAssayData(atac_tmp[standard_regions], slot = \"counts\", assay='ATAC')\n",
      "atac_tmp <- CreateSeuratObject(atac_tmp_data_subset, assay='ATAC')\n",
      "\n",
      "\n",
      "# subset by removing contig chromosomes\n",
      "rawregions = rownames(GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC'))\n",
      "# standard_regions = rawregions[!grepl('random|KIe\n",
      "pbmc.rna <- readRDS('../0_resources/seurat_references/pbmc_ssc_mat__integrated.rds')\n",
      "\n",
      "\n",
      "\n",
      "# correctly parse regions (default delims are '-','-')\n",
      "regions = StringToGRanges(\n",
      "    rownames(GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC')),\n",
      "    sep=c(':','-')\n",
      "    )\n",
      "\n",
      "# create chromatin assay\n",
      "chromatinassay = CreateChromatinAssay(\n",
      "    counts=GetAssayData(atac_tmp, slot = \"counts\", assay='ATAC'),\n",
      "    # genome='hg38',\n",
      "    fragments = f_frag,\n",
      "    ranges=regions\n",
      "    )\n",
      "    #annotation=annotation)\n",
      "\n",
      "atac <- CreateSeuratObject(counts = chromatinassay, assay='ATAC')\n",
      "\n",
      "annotations <- readRDS(f_annotation)\n",
      "Annotation(atac) <- annotations\n",
      "\n",
      "# We exclude the first dimension as this is typically correlated with sequencing depth\n",
      "atac <- RunTFIDF(atac)\n",
      "atac <- FindTopFeatures(atac, min.cutoff = \"q0\")\n",
      "atac <- RunSVD(atac)\n",
      "atac <- RunUMAP(atac, reduction = \"lsi\", dims = 2:30, reduction.name = \"umap.atac\", reduction.key = \"atacUMAP_\")\n",
      "\n",
      "##################################################\n",
      "# Identify RNA-ATAC anchors\n",
      "\n",
      "# quantify gene activity\n",
      "gene.activities <- GeneActivity(atac, features = VariableFeatures(pbmc.rna))\n",
      "# gene.activities <- gene.activities[rownames(gene.activities)!=\"\",]\n",
      "\n",
      "# add gene activities as a new assay\n",
      "atac[[\"ACTIVITY\"]] <- CreateAssayObject(counts = gene.activities)\n",
      "\n",
      "# normalize gene activities\n",
      "DefaultAssay(atac) <- \"ACTIVITY\"\n",
      "atac <- NormalizeData(atac)\n",
      "atac <- ScaleData(atac, features = rownames(atac))\n",
      "\n",
      "# Identify anchors\n",
      "transfer.anchors <- FindTransferAnchors(reference = pbmc.rna,\n",
      "    query = atac,\n",
      "    features = VariableFeatures(object = pbmc.rna),\n",
      "    reference.assay = \"RNA\",\n",
      "    query.assay = \"ACTIVITY\",\n",
      "    reduction = \"cca\")\n",
      "\n",
      "# predict celltype\n",
      "celltype.predictions <- TransferData(\n",
      "    anchorset = transfer.anchors,\n",
      "    refdata = pbmc.rna$CellType,\n",
      "    weight.reduction = atac[[\"lsi\"]],\n",
      "    dims = 2:30)\n",
      "\n",
      "pbmc.atac <- AddMetaData(atac, metadata = celltype.predictions)\n",
      "\n",
      "md = celltype.predictions\n",
      "pred_thr = 0.7\n",
      "\n",
      "tmp = data.frame(\n",
      "      composite_sample_id = paste0(rownames(md),'-',sample_id),\n",
      "      barcode = rownames(md),\n",
      "      sample_id = sample_id,\n",
      "      cell_type = md$predicted.id,\n",
      "      cell_type_pred_score = md$prediction.score.max\n",
      "      )\n",
      "\n",
      "tmp$cell_type_hiconf_70 = tmp$cell_type\n",
      "tmp$cell_type_hiconf_70[tmp$cell_type_pred_score<pred_thr] = 'Unknown'\n",
      "\n",
      "table(tmp$cell_type)\n",
      "table(tmp$cell_type_hiconf_70)\n",
      "\n",
      "write.table(tmp, file=f_out, sep='\\t', row.names=FALSE, quote=FALSE)\n",
      "print(paste0('Wrote file ', f_out))\n"
     ]
    }
   ],
   "source": [
    "!cat ../0_resources/scripts/seurat_label_transfer_consensus.R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "774780c6",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-17T13:22:27.997150Z",
     "iopub.status.busy": "2022-09-17T13:22:27.996449Z",
     "iopub.status.idle": "2022-09-17T13:22:28.011574Z",
     "shell.execute_reply": "2022-09-17T13:22:28.010799Z",
     "shell.execute_reply.started": "2022-09-17T13:22:27.997090Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#cell_type_classification/BIO_ddseq_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_3.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_4.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BRO_mtscatac_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BRO_mtscatac_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xmultiome_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xmultiome_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_3.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_4.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_5.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv2_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv2_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_3.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_mtscatac_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_mtscatac_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_3.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_4.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/HAR_ddseq_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/HAR_ddseq_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/MDC_mtscatac_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/MDC_mtscatac_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/OHS_s3atac_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/OHS_s3atac_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/SAN_10xmultiome_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/SAN_10xmultiome_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/STA_10xv11_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/STA_10xv11_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/TXG_10xv11_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/TXG_10xv2_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/TXG_10xv2_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/UCS_ddseq_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/UCS_ddseq_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xmultiome_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xmultiome_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv1_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv1_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv2_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv2_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "img_path=../0_resources/vsn_cache/cflerin-seurat-4.0.3-plus.sif && singularity exec --cleanenv -H $PWD -B /dodrio,/readonly/dodrio,/tmp $img_path Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R VIB_hydrop_1.FIXEDCELLS cell_region_looms/VIB_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/fixedcells_fragments/VIB_hydrop_1.FIXEDCELLS.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/VIB_hydrop_1.FIXEDCELLS__cell_type_seurat.txt\n",
      "#cell_type_classification/VIB_hydrop_11.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "img_path=../0_resources/vsn_cache/cflerin-seurat-4.0.3-plus.sif && singularity exec --cleanenv -H $PWD -B /dodrio,/readonly/dodrio,/tmp $img_path Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R VIB_hydrop_12.FIXEDCELLS cell_region_looms/VIB_hydrop_12.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/fixedcells_fragments/VIB_hydrop_12.FIXEDCELLS.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/VIB_hydrop_12.FIXEDCELLS__cell_type_seurat.txt\n",
      "img_path=../0_resources/vsn_cache/cflerin-seurat-4.0.3-plus.sif && singularity exec --cleanenv -H $PWD -B /dodrio,/readonly/dodrio,/tmp $img_path Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R VIB_hydrop_2.FIXEDCELLS cell_region_looms/VIB_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/fixedcells_fragments/VIB_hydrop_2.FIXEDCELLS.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/VIB_hydrop_2.FIXEDCELLS__cell_type_seurat.txt\n",
      "img_path=../0_resources/vsn_cache/cflerin-seurat-4.0.3-plus.sif && singularity exec --cleanenv -H $PWD -B /dodrio,/readonly/dodrio,/tmp $img_path Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R VIB_hydrop_21.FIXEDCELLS cell_region_looms/VIB_hydrop_21.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/fixedcells_fragments/VIB_hydrop_21.FIXEDCELLS.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/VIB_hydrop_21.FIXEDCELLS__cell_type_seurat.txt\n",
      "img_path=../0_resources/vsn_cache/cflerin-seurat-4.0.3-plus.sif && singularity exec --cleanenv -H $PWD -B /dodrio,/readonly/dodrio,/tmp $img_path Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R VIB_hydrop_22.FIXEDCELLS cell_region_looms/VIB_hydrop_22.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/fixedcells_fragments/VIB_hydrop_22.FIXEDCELLS.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/VIB_hydrop_22.FIXEDCELLS__cell_type_seurat.txt\n"
     ]
    }
   ],
   "source": [
    "parallel_filename = \"seurat_label_transfer.parallel\"\n",
    "script_path = \"../0_resources/scripts/seurat_label_transfer_consensus.R\"\n",
    "img_path = \"../0_resources/vsn_cache/cflerin-seurat-4.0.3-plus.sif\"\n",
    "reference_path = \"../0_resources/seurat_references/pbmc_ref.rds\"\n",
    "annotation_path = \"../0_resources/seurat_references/granges_annotation.rds\"\n",
    "bind_mounts = \"/dodrio,/readonly/dodrio,/tmp\"\n",
    "#bind_mounts = \"/lustre1,/staging,${VSC_SCRATCH}/tmp:/tmp\"\n",
    "\n",
    "with open(parallel_filename, 'w') as f:\n",
    "    for sample in loom_path_dict.keys():\n",
    "        outfile = f\"cell_type_classification/{sample}__cell_type_seurat.txt\"\n",
    "        if not os.path.exists(outfile):\n",
    "            loomfile = loom_path_dict[sample]\n",
    "            fragfile = frags_path_dict[sample]\n",
    "            command=f\"img_path={img_path} && singularity exec --cleanenv -H $PWD -B {bind_mounts} $img_path Rscript {script_path} {sample} {loomfile} {fragfile} {reference_path} {annotation_path} {outfile}\"\n",
    "            f.write(f\"{command}\\n\")\n",
    "            print(command)\n",
    "\n",
    "        else:\n",
    "            f.write(f\"#{outfile} already exists!\\n\")\n",
    "            print(f\"#{outfile} already exists!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9ea46faa",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-09-17T13:22:38.757113Z",
     "iopub.status.busy": "2022-09-17T13:22:38.756200Z",
     "iopub.status.idle": "2022-09-17T13:22:38.775216Z",
     "shell.execute_reply": "2022-09-17T13:22:38.774674Z",
     "shell.execute_reply.started": "2022-09-17T13:22:38.757051Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#cell_type_classification/BIO_ddseq_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_3.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_4.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BRO_mtscatac_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BRO_mtscatac_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xmultiome_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xmultiome_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_3.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_4.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_5.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv2_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv2_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_3.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_mtscatac_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_mtscatac_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_3.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_4.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/HAR_ddseq_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/HAR_ddseq_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/MDC_mtscatac_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/MDC_mtscatac_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/OHS_s3atac_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/OHS_s3atac_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/SAN_10xmultiome_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/SAN_10xmultiome_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/STA_10xv11_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/STA_10xv11_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/TXG_10xv11_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/TXG_10xv2_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/TXG_10xv2_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/UCS_ddseq_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/UCS_ddseq_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xmultiome_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xmultiome_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv1_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv1_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv2_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv2_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R VIB_hydrop_1.FIXEDCELLS cell_region_looms/VIB_hydrop_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/fixedcells_fragments/VIB_hydrop_1.FIXEDCELLS.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/VIB_hydrop_1.FIXEDCELLS__cell_type_seurat.txt\n",
      "#cell_type_classification/VIB_hydrop_11.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R VIB_hydrop_12.FIXEDCELLS cell_region_looms/VIB_hydrop_12.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/fixedcells_fragments/VIB_hydrop_12.FIXEDCELLS.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/VIB_hydrop_12.FIXEDCELLS__cell_type_seurat.txt\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R VIB_hydrop_2.FIXEDCELLS cell_region_looms/VIB_hydrop_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/fixedcells_fragments/VIB_hydrop_2.FIXEDCELLS.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/VIB_hydrop_2.FIXEDCELLS__cell_type_seurat.txt\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R VIB_hydrop_21.FIXEDCELLS cell_region_looms/VIB_hydrop_21.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/fixedcells_fragments/VIB_hydrop_21.FIXEDCELLS.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/VIB_hydrop_21.FIXEDCELLS__cell_type_seurat.txt\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R VIB_hydrop_22.FIXEDCELLS cell_region_looms/VIB_hydrop_22.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/fixedcells_fragments/VIB_hydrop_22.FIXEDCELLS.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/VIB_hydrop_22.FIXEDCELLS__cell_type_seurat.txt\n"
     ]
    }
   ],
   "source": [
    "parallel_filename = \"seurat_label_transfer.parallel\"\n",
    "script_path = \"../0_resources/scripts/seurat_label_transfer_consensus.R\"\n",
    "img_path = \"../0_resources/vsn_cache/cflerin-seurat-4.0.3-plus.sif\"\n",
    "reference_path = \"../0_resources/seurat_references/pbmc_ref.rds\"\n",
    "annotation_path = \"../0_resources/seurat_references/granges_annotation.rds\"\n",
    "bind_mounts = \"/dodrio,/readonly/dodrio,/tmp\"\n",
    "#bind_mounts = \"/lustre1,/staging,${VSC_SCRATCH}/tmp:/tmp\"\n",
    "\n",
    "with open(parallel_filename, 'w') as f:\n",
    "    for sample in loom_path_dict.keys():\n",
    "        outfile = f\"cell_type_classification/{sample}__cell_type_seurat.txt\"\n",
    "        if not os.path.exists(outfile):\n",
    "            loomfile = loom_path_dict[sample]\n",
    "            fragfile = frags_path_dict[sample]\n",
    "            command=f\"Rscript {script_path} {sample} {loomfile} {fragfile} {reference_path} {annotation_path} {outfile}\"\n",
    "            f.write(f\"{command}\\n\")\n",
    "            print(command)\n",
    "\n",
    "        else:\n",
    "            f.write(f\"#{outfile} already exists!\\n\")\n",
    "            print(f\"#{outfile} already exists!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "34402990",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#cell_type_classification/BIO_ddseq_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_3.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BIO_ddseq_4.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BRO_mtscatac_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/BRO_mtscatac_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xmultiome_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xmultiome_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_3.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_4.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv11_5.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv2_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_10xv2_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_hydrop_3.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_mtscatac_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/CNA_mtscatac_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_3.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/EPF_hydrop_4.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/HAR_ddseq_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/HAR_ddseq_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/MDC_mtscatac_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/MDC_mtscatac_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/OHS_s3atac_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/OHS_s3atac_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/SAN_10xmultiome_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/SAN_10xmultiome_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/STA_10xv11_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/STA_10xv11_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R TXG_10xv11_1.FIXEDCELLS cell_region_looms/TXG_10xv11_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/fixedcells_fragments/TXG_10xv11_1.FIXEDCELLS.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/TXG_10xv11_1.FIXEDCELLS__cell_type_seurat.txt\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R TXG_10xv2_1.FIXEDCELLS cell_region_looms/TXG_10xv2_1.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/fixedcells_fragments/TXG_10xv2_1.FIXEDCELLS.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/TXG_10xv2_1.FIXEDCELLS__cell_type_seurat.txt\n",
      "Rscript ../0_resources/scripts/seurat_label_transfer_consensus.R TXG_10xv2_2.FIXEDCELLS cell_region_looms/TXG_10xv2_2.FIXEDCELLS__cto.scrublet0-4.fmx.singlets.loom ../1_data_repository/fixedcells_fragments/TXG_10xv2_2.FIXEDCELLS.fragments.tsv.gz ../0_resources/seurat_references/pbmc_ref.rds ../0_resources/seurat_references/granges_annotation.rds cell_type_classification/TXG_10xv2_2.FIXEDCELLS__cell_type_seurat.txt\n",
      "#cell_type_classification/UCS_ddseq_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/UCS_ddseq_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xmultiome_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xmultiome_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv1_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv1_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv2_1.FIXEDCELLS__cell_type_seurat.txt already exists!\n",
      "#cell_type_classification/VIB_10xv2_2.FIXEDCELLS__cell_type_seurat.txt already exists!\n"
     ]
    }
   ],
   "source": [
    "parallel_filename = \"seurat_label_transfer.parallel\"\n",
    "script_path = \"../0_resources/scripts/seurat_label_transfer_consensus.R\"\n",
    "img_path = \"../0_resources/vsn_cache/cflerin-seurat-4.0.3-plus.sif\"\n",
    "reference_path = \"../0_resources/seurat_references/pbmc_ref.rds\"\n",
    "annotation_path = \"../0_resources/seurat_references/granges_annotation.rds\"\n",
    "bind_mounts = \"/dodrio,/readonly/dodrio,/tmp\"\n",
    "#bind_mounts = \"/lustre1,/staging,${VSC_SCRATCH}/tmp:/tmp\"\n",
    "\n",
    "with open(parallel_filename, 'w') as f:\n",
    "    for sample in loom_path_dict.keys():\n",
    "        outfile = f\"cell_type_classification/{sample}__cell_type_seurat.txt\"\n",
    "        if not os.path.exists(outfile):\n",
    "            loomfile = loom_path_dict[sample]\n",
    "            fragfile = frags_path_dict[sample]\n",
    "            command=f\"Rscript {script_path} {sample} {loomfile} {fragfile} {reference_path} {annotation_path} {outfile}\"\n",
    "            f.write(f\"{command}\\n\")\n",
    "            print(command)\n",
    "\n",
    "        else:\n",
    "            f.write(f\"#{outfile} already exists!\\n\")\n",
    "            print(f\"#{outfile} already exists!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "516121d7",
   "metadata": {},
   "source": [
    "# Run the Rscript"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d716c6d8",
   "metadata": {},
   "source": [
    "```\n",
    "mkdir cell_type_classification\n",
    "export OMP_THREAD_LIMIT=80\n",
    "cat seurat_label_transfer.parallel | parallel -j 2 --progress\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ba15c60b",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (4107442686.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  Input \u001b[0;32mIn [20]\u001b[0;36m\u001b[0m\n\u001b[0;31m    kill $(jobs -p)\u001b[0m\n\u001b[0m         ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "kill $(jobs -p)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "20221014_pycistopic.sif",
   "language": "python",
   "name": "cistopic_20221014"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "papermill": {
   "default_parameters": {},
   "duration": null,
   "end_time": null,
   "environment_variables": {},
   "exception": null,
   "input_path": "8_consensus_peaks.ipynb",
   "output_path": "8_consensus_peaks.OUT.ipynb",
   "parameters": {},
   "start_time": "2022-09-17T13:41:25.502802",
   "version": "2.3.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
