{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "92d4dbfc",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-08T11:06:57.513809Z",
     "iopub.status.busy": "2022-12-08T11:06:57.513312Z",
     "iopub.status.idle": "2022-12-08T11:06:58.740980Z",
     "shell.execute_reply": "2022-12-08T11:06:58.740226Z",
     "shell.execute_reply.started": "2022-12-08T11:06:57.513780Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import gc\n",
    "import gzip\n",
    "import logging\n",
    "import math\n",
    "import re\n",
    "import sys\n",
    "\n",
    "import matplotlib.backends.backend_pdf\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "from PIL import Image\n",
    "from scipy import sparse\n",
    "\n",
    "import collections as cl\n",
    "import gc\n",
    "import logging\n",
    "import sys\n",
    "from typing import Dict, List, Optional, Tuple, Union\n",
    "\n",
    "import matplotlib.backends.backend_pdf\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "import ray\n",
    "import seaborn as sns\n",
    "from scipy.stats import gaussian_kde, norm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0534eb4e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-08T11:07:01.691575Z",
     "iopub.status.busy": "2022-12-08T11:07:01.690888Z",
     "iopub.status.idle": "2022-12-08T11:07:01.727017Z",
     "shell.execute_reply": "2022-12-08T11:07:01.725980Z",
     "shell.execute_reply.started": "2022-12-08T11:07:01.691543Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "def read_fragments_from_file(\n",
    "    fragments_bed_filename, use_polars: bool = True\n",
    ") -> pr.PyRanges:\n",
    "    \"\"\"\n",
    "    Read fragments BED file to PyRanges object.\n",
    "    Parameters\n",
    "    ----------\n",
    "    fragments_bed_filename: Fragments BED filename.\n",
    "    use_polars: Use polars instead of pandas for reading the fragments BED file.\n",
    "    Returns\n",
    "    -------\n",
    "    PyRanges object of fragments.\n",
    "    \"\"\"\n",
    "\n",
    "    bed_column_names = (\n",
    "        \"Chromosome\",\n",
    "        \"Start\",\n",
    "        \"End\",\n",
    "        \"Name\",\n",
    "        \"Score\",\n",
    "        \"Strand\",\n",
    "        \"ThickStart\",\n",
    "        \"ThickEnd\",\n",
    "        \"ItemRGB\",\n",
    "        \"BlockCount\",\n",
    "        \"BlockSizes\",\n",
    "        \"BlockStarts\",\n",
    "    )\n",
    "\n",
    "    # Set the correct open function depending if the fragments BED file is gzip compressed or not.\n",
    "    open_fn = gzip.open if fragments_bed_filename.endswith(\".gz\") else open\n",
    "\n",
    "    skip_rows = 0\n",
    "    nbr_columns = 0\n",
    "    with open_fn(fragments_bed_filename, \"rt\") as fragments_bed_fh:\n",
    "        for line in fragments_bed_fh:\n",
    "            # Remove newlines and spaces.\n",
    "            line = line.strip()\n",
    "\n",
    "            if not line or line.startswith(\"#\"):\n",
    "                # Count number of empty lines and lines which start with a comment before the actual data.\n",
    "                skip_rows += 1\n",
    "            else:\n",
    "                # Get number of columns from the first real BED entry.\n",
    "                nbr_columns = len(line.split(\"\\t\"))\n",
    "\n",
    "                # Stop reading the BED file.\n",
    "                break\n",
    "\n",
    "    if nbr_columns < 4:\n",
    "        raise ValueError(\n",
    "            f'Fragments BED file needs to have at least 4 columns. \"{fragments_bed_filename}\" contains only '\n",
    "            f\"{nbr_columns} columns.\"\n",
    "        )\n",
    "\n",
    "    if use_polars:\n",
    "        import polars as pl\n",
    "\n",
    "        # Read fragments BED file with polars.\n",
    "        df = (\n",
    "            pl.read_csv(\n",
    "                fragments_bed_filename,\n",
    "                has_headers=False,\n",
    "                skip_rows=skip_rows,\n",
    "                sep=\"\\t\",\n",
    "                use_pyarrow=True,\n",
    "                new_columns=bed_column_names[:nbr_columns],\n",
    "            )\n",
    "            .with_columns(\n",
    "                [\n",
    "                    pl.col(\"Chromosome\").cast(pl.Utf8),\n",
    "                    pl.col(\"Start\").cast(pl.Int32),\n",
    "                    pl.col(\"End\").cast(pl.Int32),\n",
    "                    pl.col(\"Name\").cast(pl.Utf8),\n",
    "                ]\n",
    "            )\n",
    "            .to_pandas()\n",
    "        )\n",
    "\n",
    "        # Convert \"Name\" column to pd.Categorical as groupby operations will be done on it later.\n",
    "        df[\"Name\"] = df[\"Name\"].astype(\"category\")\n",
    "    else:\n",
    "        # Read fragments BED file with pandas.\n",
    "        df = pd.read_table(\n",
    "            fragments_bed_filename,\n",
    "            sep=\"\\t\",\n",
    "            skiprows=skip_rows,\n",
    "            header=None,\n",
    "            names=bed_column_names[:nbr_columns],\n",
    "            doublequote=False,\n",
    "            engine=\"c\",\n",
    "            dtype={\n",
    "                \"Chromosome\": str,\n",
    "                \"Start'\": np.int32,\n",
    "                \"End\": np.int32,\n",
    "                \"Name\": \"category\",\n",
    "                \"Strand\": str,\n",
    "            },\n",
    "        )\n",
    "\n",
    "    # Convert pandas dataframe to PyRanges dataframe.\n",
    "    # This will convert \"Chromosome\" and \"Strand\" columns to pd.Categorical.\n",
    "    return pr.PyRanges(df)\n",
    "\n",
    "def profile_tss(\n",
    "    fragments: Union[str, pd.DataFrame],\n",
    "    annotation: Union[pd.DataFrame, pr.PyRanges],\n",
    "    valid_bc: Optional[List[str]] = None,\n",
    "    plot: Optional[bool] = True,\n",
    "    plot_data: Optional[pd.DataFrame] = None,\n",
    "    n_cpu: Optional[int] = 1,\n",
    "    partition: Optional[int] = 5,\n",
    "    flank_window: Optional[int] = 1000,\n",
    "    tss_window: Optional[int] = 50,\n",
    "    minimum_signal_window: Optional[int] = 100,\n",
    "    rolling_window: Optional[int] = 10,\n",
    "    min_norm: Optional[int] = 0.2,\n",
    "    color: Optional[str] = None,\n",
    "    save: Optional[str] = None,\n",
    "    return_TSS_enrichment_per_barcode: Optional[bool] = False,\n",
    "    return_TSS_coverage_matrix_per_barcode: Optional[bool] = False,\n",
    "    return_plot_data: Optional[bool] = False,\n",
    "    use_polars: Optional[bool] = True\n",
    "):\n",
    "    \"\"\"\n",
    "    Plot the Transcription Start Site (TSS) profile. It is computed as the summed accessibility signal (sample-level), or the number of cut sites per base (barcode-level), in a space around the full set of annotated TSSs and is normalized by the minimum signal in the window. This profile is helpful to assess the signal-to-noise ratio of the library, as it is well known that TSSs and the promoter regions around them have, on average, a high degree of chromatin accessibility compared to the intergenic and intronic regions of the genome.\n",
    "    Parameters\n",
    "    ---------\n",
    "    fragments: str or pd.DataFrame\n",
    "            The path to the fragments file containing chromosome, start, end and assigned barcode for each read (e.g. from CellRanger ATAC (/outs/fragments.tsv.gz)) or a data frame\n",
    "            containing 'Chromosome', 'Start', 'End', 'Name', and 'Score', which indicates the number of times that a fragments is found assigned to that barcode. The fragments data\n",
    "            frame can be obtained using PyRanges:\n",
    "                    import pyranges as pr\n",
    "                    fragments = pr.read_bed(fragments_file, as_df=True))\n",
    "    annotation: pd.DataFrame or pyRanges\n",
    "            A data frame or pyRanges containing transcription start sites for each gene, with 'Chromosome', 'Start' and 'Strand' as columns (additional columns will be ignored). This data frame can be easily obtained via pybiomart:\n",
    "                    # Get TSS annotations\n",
    "                    import pybiomart as pbm\n",
    "                    # For mouse\n",
    "                    dataset = pbm.Dataset(name='mmusculus_gene_ensembl',  host='http://www.ensembl.org')\n",
    "                    # For human\n",
    "                    dataset = pbm.Dataset(name='hsapiens_gene_ensembl',  host='http://www.ensembl.org')\n",
    "                    # For fly\n",
    "                    dataset = pbm.Dataset(name='dmelanogaster_gene_ensembl',  host='http://www.ensembl.org')\n",
    "                    # Query TSS list and format\n",
    "                    annot = dataset.query(attributes=['chromosome_name', 'transcription_start_site', 'strand', 'external_gene_name', 'transcript_biotype'])\n",
    "                    filter = annot['Chromosome/scaffold name'].str.contains('CHR|GL|JH|MT')\n",
    "                    annot = annot[~filter]\n",
    "                    annot['Chromosome/scaffold name'] = annot['Chromosome/scaffold name'].str.replace(r'(\\b\\\\S)', r'chr\\1')\n",
    "                    annot.columns=['Chromosome', 'Start', 'Strand', 'Gene', 'Transcript_type']\n",
    "                    # Select TSSs of protein coding genes\n",
    "                    annot = annot[annot.Transcript_type == 'protein_coding']\n",
    "    valid_bc: list, optional\n",
    "            A list containing selected barcodes. Default: None.\n",
    "    plot: bool, optional\n",
    "            Whether to return the plot to the console. Default: True.\n",
    "    plot_data: pd.DataFrame, optional\n",
    "            Data frame containing precomputed plot data. Default: None.\n",
    "    flank_window: int, optional\n",
    "            Flanking window around the TSS. Default: 1000 (+/- 1000 bp).\n",
    "    tss_window: int, optional\n",
    "            Window around the TSS used to count fragments in the TSS when calculating the TSS enrichment per barcode. Default: 50 (+/- 50 bp).\n",
    "    minimum_signal_window: int, optional\n",
    "            Tail window use to normalize the TSS enrichment. Default: 100 (average signal in the 100bp in the extremes of the TSS window).\n",
    "    rolling_window: int, optional\n",
    "            Rolling window used to smooth signal. Default: 10.\n",
    "    min_norm: int, optional\n",
    "            Minimum normalization score. If the average minimum signal value is below this value, this number is used to normalize the TSS signal. This approach penalizes cells with fewer reads.\n",
    "    color: str, optional\n",
    "            Line color for the plot. Default: None.\n",
    "    save: str, optional\n",
    "            Output file to save plot. Default: None.\n",
    "    remove_duplicates: bool, optional\n",
    "            Whether to remove duplicates. Default: True.\n",
    "    return_TSS_enrichment_per_barcode: bool, optional\n",
    "            Whether to return a data frame containing the normalized enrichment score on the TSS for each barcode. Default: False.\n",
    "    return_TSS_coverage_matrix_per_barcode: bool, optional\n",
    "            Whether to return a matrix containing the normalized enrichment in each position in the window for each barcode, with positions as columns and barcodes as rows. Default: False.\n",
    "    return_plot_data: bool, optional\n",
    "            Whether to return the TSS profile plot data. Default: False.\n",
    "    Return\n",
    "    ------\n",
    "    dict\n",
    "            A dictionary containing a :class:`pd.DataFrame` with the normalized enrichment score on the TSS for each barcode, a :class:`pd.DataFrame` with the normalized enrichment scores in each position for each barcode and/or a :class:`pd.DataFrame` with the TSS profile plot data.\n",
    "    \"\"\"\n",
    "    # Create logger\n",
    "    level = logging.INFO\n",
    "    log_format = \"%(asctime)s %(name)-12s %(levelname)-8s %(message)s\"\n",
    "    handlers = [logging.StreamHandler(stream=sys.stdout)]\n",
    "    logging.basicConfig(level=level, format=log_format, handlers=handlers)\n",
    "    log = logging.getLogger(\"cisTopic\")\n",
    "\n",
    "    if isinstance(plot_data, pd.DataFrame):\n",
    "        log.info(\"Using plot_data. TSS enrichment per barcode will not be computed\")\n",
    "        fig, ax = plt.subplots()\n",
    "        ax.plot(plot_data.Position, plot_data.TSSEnrichment)\n",
    "        plt.xlim(-space_TSS, space_TSS)\n",
    "        plt.xlabel(\"Position from TSS\", fontsize=10)\n",
    "        plt.ylabel(\"Normalized enrichment\", fontsize=10)\n",
    "    else:\n",
    "        if isinstance(fragments, str):\n",
    "            log.info(\"Reading fragments file\")\n",
    "            fragments = read_fragments_from_file(fragments, use_polars=use_polars)\n",
    "        else:\n",
    "            if isinstance(fragments, pd.DataFrame):\n",
    "                fragments = pr.PyRanges(fragments)\n",
    "\n",
    "        if valid_bc is not None:\n",
    "            log.info(\"Using provided valid barcodes\")\n",
    "            fragments = fragments[fragments.Name.isin(set(valid_bc))]\n",
    "        else:\n",
    "            valid_bc = list(set(fragments.Name.tolist()))\n",
    "\n",
    "        log.info(\"Formatting annnotation\")\n",
    "        if isinstance(annotation, pr.PyRanges):\n",
    "            annotation = annotation.df\n",
    "        tss_space_annotation = annotation[[\"Chromosome\", \"Start\", \"Strand\"]]\n",
    "        tss_space_annotation[\"End\"] = tss_space_annotation[\"Start\"] + flank_window\n",
    "        tss_space_annotation[\"Start\"] = tss_space_annotation[\"Start\"] - flank_window\n",
    "        tss_space_annotation = tss_space_annotation[\n",
    "            [\"Chromosome\", \"Start\", \"End\", \"Strand\"]\n",
    "        ]\n",
    "        tss_space_annotation = pr.PyRanges(tss_space_annotation)\n",
    "\n",
    "        log.info(\"Creating coverage matrix\")\n",
    "        if partition > 1:\n",
    "            barcode_list = np.array_split(valid_bc, partition)\n",
    "            TSS_matrix = pd.concat(\n",
    "                [\n",
    "                    get_tss_matrix(\n",
    "                        fragments[fragments.Name.isin(set(barcode_list[x]))],\n",
    "                        flank_window,\n",
    "                        tss_space_annotation,\n",
    "                    ).fillna(0)\n",
    "                    for x in range(partition)\n",
    "                ]\n",
    "            )\n",
    "        else:\n",
    "            TSS_matrix = get_tss_matrix(fragments, flank_window, tss_space_annotation)\n",
    "        log.info(\"Coverage matrix done\")\n",
    "        if not TSS_matrix.columns.tolist() == list(range(2 * flank_window + 1)):\n",
    "            missing_values = list(\n",
    "                set(TSS_matrix.columns.tolist()).symmetric_difference(\n",
    "                    list(range(2 * flank_window + 1))\n",
    "                )\n",
    "            )\n",
    "            for x in missing_values:\n",
    "                TSS_matrix[x] = 0\n",
    "\n",
    "            TSS_matrix = TSS_matrix.reindex(sorted(TSS_matrix.columns), axis=1)\n",
    "\n",
    "        if rolling_window is not None:\n",
    "            TSS_matrix = TSS_matrix.rolling(\n",
    "                window=rolling_window, min_periods=0, axis=1\n",
    "            ).mean()\n",
    "\n",
    "        TSS_counts = TSS_matrix.values.sum(axis=0)\n",
    "        div = max(\n",
    "            (\n",
    "                np.mean(TSS_counts[-minimum_signal_window:])\n",
    "                + np.mean(TSS_counts[0:minimum_signal_window])\n",
    "            )\n",
    "            / 2,\n",
    "            min_norm,\n",
    "        )\n",
    "        if plot is True or save is not None:\n",
    "            fig, ax = plt.subplots()\n",
    "            ax.plot(\n",
    "                range(-flank_window - 1, flank_window), TSS_counts / div, color=color\n",
    "            )\n",
    "            plt.xlim(-flank_window, flank_window)\n",
    "            plt.xlabel(\"Position from TSS\", fontsize=10)\n",
    "            plt.ylabel(\"Normalized enrichment\", fontsize=10)\n",
    "            if save is not None:\n",
    "                fig.savefig(save)\n",
    "            if plot:\n",
    "                log.info(\"Plotting normalized sample TSS enrichment\")\n",
    "                plt.show()\n",
    "            else:\n",
    "                plt.close(fig)\n",
    "\n",
    "    output = {}\n",
    "    flag = False\n",
    "    if return_TSS_enrichment_per_barcode:\n",
    "        TSS_enrich = TSS_matrix.apply(\n",
    "            lambda x: x\n",
    "            / max(\n",
    "                [\n",
    "                    (\n",
    "                        (\n",
    "                            np.mean(x[-minimum_signal_window:])\n",
    "                            + np.mean(x[0:minimum_signal_window])\n",
    "                        )\n",
    "                        / 2\n",
    "                    ),\n",
    "                    min_norm,\n",
    "                ]\n",
    "            ),\n",
    "            axis=1,\n",
    "        )\n",
    "        TSS_enrich = pd.DataFrame(\n",
    "            TSS_enrich.iloc[\n",
    "                :, range(flank_window - tss_window, flank_window + tss_window)\n",
    "            ].mean(axis=1)\n",
    "        )\n",
    "        TSS_enrich.columns = [\"TSS_enrichment\"]\n",
    "        output.update({\"TSS_enrichment\": TSS_enrich})\n",
    "        flag = True\n",
    "    if return_TSS_coverage_matrix_per_barcode:\n",
    "        log.info(\"Returning normalized TSS coverage matrix per barcode\")\n",
    "        TSS_mat = TSS_matrix.apply(\n",
    "            lambda x: x\n",
    "            / max(\n",
    "                [\n",
    "                    (\n",
    "                        (\n",
    "                            np.mean(x[-minimum_signal_window:])\n",
    "                            + np.mean(x[0:minimum_signal_window])\n",
    "                        )\n",
    "                        / 2\n",
    "                    ),\n",
    "                    min_norm,\n",
    "                ]\n",
    "            ),\n",
    "            axis=1,\n",
    "        )\n",
    "        output.update({\"TSS_coverage_mat\": TSS_mat})\n",
    "        flag = True\n",
    "    if return_plot_data:\n",
    "        log.info(\"Returning normalized sample TSS enrichment data\")\n",
    "        output.update(\n",
    "            {\n",
    "                \"TSS_plot_data\": pd.DataFrame(\n",
    "                    {\n",
    "                        \"Position\": range(-flank_window - 1, flank_window),\n",
    "                        \"TSS_enrichment\": TSS_counts / div,\n",
    "                    }\n",
    "                )\n",
    "            }\n",
    "        )\n",
    "        flag = True\n",
    "    del TSS_matrix\n",
    "    if flag:\n",
    "        return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1f993bdf",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-08T12:47:20.360800Z",
     "iopub.status.busy": "2022-12-08T12:47:20.360400Z",
     "iopub.status.idle": "2022-12-08T12:47:20.485497Z",
     "shell.execute_reply": "2022-12-08T12:47:20.484607Z",
     "shell.execute_reply.started": "2022-12-08T12:47:20.360773Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading cached genome annotation...\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "genome = \"hg38\"\n",
    "\n",
    "pbm_genome_name_dict = {\n",
    "    \"hg38\": \"hsapiens_gene_ensembl\",\n",
    "    \"hg37\": \"hsapiens_gene_ensembl\",\n",
    "    \"mm10\": \"mmusculus_gene_ensembl\",\n",
    "    \"dm6\": \"dmelanogaster_gene_ensembl\",\n",
    "}\n",
    "\n",
    "pbm_host_dict = {\n",
    "    \"hg38\": \"http://www.ensembl.org\",\n",
    "    \"hg37\": \"http://grch37.ensembl.org/\",\n",
    "    \"mm10\": \"http://nov2020.archive.ensembl.org/\",\n",
    "    \"dm6\": \"http://www.ensembl.org\",\n",
    "}\n",
    "\n",
    "if os.path.exists(f\"annotation.tsv\"):\n",
    "    print(f\"Loading cached genome annotation...\")\n",
    "    annotation = pd.read_csv(\"annotation.tsv\", sep=\"\\t\", header=0, index_col=0)\n",
    "else:\n",
    "    dataset = pbm.Dataset(name=pbm_genome_name_dict[genome], host=pbm_host_dict[genome])\n",
    "\n",
    "    annotation = dataset.query(\n",
    "        attributes=[\n",
    "            \"chromosome_name\",\n",
    "            \"transcription_start_site\",\n",
    "            \"strand\",\n",
    "            \"external_gene_name\",\n",
    "            \"transcript_biotype\",\n",
    "        ]\n",
    "    )\n",
    "    filter = annotation[\"Chromosome/scaffold name\"].str.contains(\"CHR|GL|JH|MT\")\n",
    "    annotation = annotation[~filter]\n",
    "    annotation[\"Chromosome/scaffold name\"] = annotation[\n",
    "        \"Chromosome/scaffold name\"\n",
    "    ].str.replace(r\"(\\b\\S)\", r\"chr\\1\")\n",
    "    annotation.columns = [\"Chromosome\", \"Start\", \"Strand\", \"Gene\", \"Transcript_type\"]\n",
    "    annotation = annotation[annotation.Transcript_type == \"protein_coding\"]\n",
    "    annotation.to_csv(\"annotation.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "8f82e7ee",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-08T12:49:04.707749Z",
     "iopub.status.busy": "2022-12-08T12:49:04.706761Z",
     "iopub.status.idle": "2022-12-08T12:49:07.116144Z",
     "shell.execute_reply": "2022-12-08T12:49:07.113090Z",
     "shell.execute_reply.started": "2022-12-08T12:49:04.707682Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pycisTopic.cistopic_class import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "095ea11c",
   "metadata": {
    "collapsed": true,
    "execution": {
     "iopub.execute_input": "2022-12-08T11:07:54.078100Z",
     "iopub.status.busy": "2022-12-08T11:07:54.077620Z",
     "iopub.status.idle": "2022-12-08T11:07:54.100029Z",
     "shell.execute_reply": "2022-12-08T11:07:54.099259Z",
     "shell.execute_reply.started": "2022-12-08T11:07:54.078070Z"
    },
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'BIO_ddseq_1.10k': '10k/10k_preprocessing_out/data/fragments/BIO_ddseq_1.10k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_2.10k': '10k/10k_preprocessing_out/data/fragments/BIO_ddseq_2.10k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_3.10k': '10k/10k_preprocessing_out/data/fragments/BIO_ddseq_3.10k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_4.10k': '10k/10k_preprocessing_out/data/fragments/BIO_ddseq_4.10k.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_1.10k': '10k/10k_preprocessing_out/data/fragments/BRO_mtscatac_1.10k.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_2.10k': '10k/10k_preprocessing_out/data/fragments/BRO_mtscatac_2.10k.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_1.10k': '10k/10k_preprocessing_out/data/fragments/CNA_10xmultiome_1.10k.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_2.10k': '10k/10k_preprocessing_out/data/fragments/CNA_10xmultiome_2.10k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_1.10k': '10k/10k_preprocessing_out/data/fragments/CNA_10xv11_1.10k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_2.10k': '10k/10k_preprocessing_out/data/fragments/CNA_10xv11_2.10k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_3.10k': '10k/10k_preprocessing_out/data/fragments/CNA_10xv11_3.10k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_4.10k': '10k/10k_preprocessing_out/data/fragments/CNA_10xv11_4.10k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_5.10k': '10k/10k_preprocessing_out/data/fragments/CNA_10xv11_5.10k.fragments.tsv.gz',\n",
       " 'CNA_10xv2_1.10k': '10k/10k_preprocessing_out/data/fragments/CNA_10xv2_1.10k.fragments.tsv.gz',\n",
       " 'CNA_10xv2_2.10k': '10k/10k_preprocessing_out/data/fragments/CNA_10xv2_2.10k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_1.10k': '10k/10k_preprocessing_out/data/fragments/CNA_hydrop_1.10k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_2.10k': '10k/10k_preprocessing_out/data/fragments/CNA_hydrop_2.10k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_3.10k': '10k/10k_preprocessing_out/data/fragments/CNA_hydrop_3.10k.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_1.10k': '10k/10k_preprocessing_out/data/fragments/CNA_mtscatac_1.10k.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_2.10k': '10k/10k_preprocessing_out/data/fragments/CNA_mtscatac_2.10k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_1.10k': '10k/10k_preprocessing_out/data/fragments/EPF_hydrop_1.10k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_2.10k': '10k/10k_preprocessing_out/data/fragments/EPF_hydrop_2.10k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_3.10k': '10k/10k_preprocessing_out/data/fragments/EPF_hydrop_3.10k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_4.10k': '10k/10k_preprocessing_out/data/fragments/EPF_hydrop_4.10k.fragments.tsv.gz',\n",
       " 'HAR_ddseq_1.10k': '10k/10k_preprocessing_out/data/fragments/HAR_ddseq_1.10k.fragments.tsv.gz',\n",
       " 'HAR_ddseq_2.10k': '10k/10k_preprocessing_out/data/fragments/HAR_ddseq_2.10k.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_1.10k': '10k/10k_preprocessing_out/data/fragments/MDC_mtscatac_1.10k.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_2.10k': '10k/10k_preprocessing_out/data/fragments/MDC_mtscatac_2.10k.fragments.tsv.gz',\n",
       " 'OHS_s3atac_1.10k': '10k/10k_preprocessing_out/data/fragments/OHS_s3atac_1.10k.fragments.tsv.gz',\n",
       " 'OHS_s3atac_2.10k': '10k/10k_preprocessing_out/data/fragments/OHS_s3atac_2.10k.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_1.10k': '10k/10k_preprocessing_out/data/fragments/SAN_10xmultiome_1.10k.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_2.10k': '10k/10k_preprocessing_out/data/fragments/SAN_10xmultiome_2.10k.fragments.tsv.gz',\n",
       " 'STA_10xv11_1.10k': '10k/10k_preprocessing_out/data/fragments/STA_10xv11_1.10k.fragments.tsv.gz',\n",
       " 'STA_10xv11_2.10k': '10k/10k_preprocessing_out/data/fragments/STA_10xv11_2.10k.fragments.tsv.gz',\n",
       " 'TXG_10xv11_1.10k': '10k/10k_preprocessing_out/data/fragments/TXG_10xv11_1.10k.fragments.tsv.gz',\n",
       " 'TXG_10xv2_1.10k': '10k/10k_preprocessing_out/data/fragments/TXG_10xv2_1.10k.fragments.tsv.gz',\n",
       " 'TXG_10xv2_2.10k': '10k/10k_preprocessing_out/data/fragments/TXG_10xv2_2.10k.fragments.tsv.gz',\n",
       " 'UCS_ddseq_1.10k': '10k/10k_preprocessing_out/data/fragments/UCS_ddseq_1.10k.fragments.tsv.gz',\n",
       " 'UCS_ddseq_2.10k': '10k/10k_preprocessing_out/data/fragments/UCS_ddseq_2.10k.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_1.10k': '10k/10k_preprocessing_out/data/fragments/VIB_10xmultiome_1.10k.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_2.10k': '10k/10k_preprocessing_out/data/fragments/VIB_10xmultiome_2.10k.fragments.tsv.gz',\n",
       " 'VIB_10xv1_1.10k': '10k/10k_preprocessing_out/data/fragments/VIB_10xv1_1.10k.fragments.tsv.gz',\n",
       " 'VIB_10xv1_2.10k': '10k/10k_preprocessing_out/data/fragments/VIB_10xv1_2.10k.fragments.tsv.gz',\n",
       " 'VIB_10xv2_1.10k': '10k/10k_preprocessing_out/data/fragments/VIB_10xv2_1.10k.fragments.tsv.gz',\n",
       " 'VIB_10xv2_2.10k': '10k/10k_preprocessing_out/data/fragments/VIB_10xv2_2.10k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_1.10k': '10k/10k_preprocessing_out/data/fragments/VIB_hydrop_1.10k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_11.10k': '10k/10k_preprocessing_out/data/fragments/VIB_hydrop_11.10k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_12.10k': '10k/10k_preprocessing_out/data/fragments/VIB_hydrop_12.10k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_2.10k': '10k/10k_preprocessing_out/data/fragments/VIB_hydrop_2.10k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_21.10k': '10k/10k_preprocessing_out/data/fragments/VIB_hydrop_21.10k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_22.10k': '10k/10k_preprocessing_out/data/fragments/VIB_hydrop_22.10k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_1.15k': '15k/15k_preprocessing_out/data/fragments/BIO_ddseq_1.15k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_2.15k': '15k/15k_preprocessing_out/data/fragments/BIO_ddseq_2.15k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_3.15k': '15k/15k_preprocessing_out/data/fragments/BIO_ddseq_3.15k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_4.15k': '15k/15k_preprocessing_out/data/fragments/BIO_ddseq_4.15k.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_1.15k': '15k/15k_preprocessing_out/data/fragments/BRO_mtscatac_1.15k.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_2.15k': '15k/15k_preprocessing_out/data/fragments/BRO_mtscatac_2.15k.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_1.15k': '15k/15k_preprocessing_out/data/fragments/CNA_10xmultiome_1.15k.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_2.15k': '15k/15k_preprocessing_out/data/fragments/CNA_10xmultiome_2.15k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_1.15k': '15k/15k_preprocessing_out/data/fragments/CNA_10xv11_1.15k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_2.15k': '15k/15k_preprocessing_out/data/fragments/CNA_10xv11_2.15k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_3.15k': '15k/15k_preprocessing_out/data/fragments/CNA_10xv11_3.15k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_4.15k': '15k/15k_preprocessing_out/data/fragments/CNA_10xv11_4.15k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_5.15k': '15k/15k_preprocessing_out/data/fragments/CNA_10xv11_5.15k.fragments.tsv.gz',\n",
       " 'CNA_10xv2_1.15k': '15k/15k_preprocessing_out/data/fragments/CNA_10xv2_1.15k.fragments.tsv.gz',\n",
       " 'CNA_10xv2_2.15k': '15k/15k_preprocessing_out/data/fragments/CNA_10xv2_2.15k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_1.15k': '15k/15k_preprocessing_out/data/fragments/CNA_hydrop_1.15k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_2.15k': '15k/15k_preprocessing_out/data/fragments/CNA_hydrop_2.15k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_3.15k': '15k/15k_preprocessing_out/data/fragments/CNA_hydrop_3.15k.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_1.15k': '15k/15k_preprocessing_out/data/fragments/CNA_mtscatac_1.15k.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_2.15k': '15k/15k_preprocessing_out/data/fragments/CNA_mtscatac_2.15k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_1.15k': '15k/15k_preprocessing_out/data/fragments/EPF_hydrop_1.15k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_2.15k': '15k/15k_preprocessing_out/data/fragments/EPF_hydrop_2.15k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_3.15k': '15k/15k_preprocessing_out/data/fragments/EPF_hydrop_3.15k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_4.15k': '15k/15k_preprocessing_out/data/fragments/EPF_hydrop_4.15k.fragments.tsv.gz',\n",
       " 'HAR_ddseq_1.15k': '15k/15k_preprocessing_out/data/fragments/HAR_ddseq_1.15k.fragments.tsv.gz',\n",
       " 'HAR_ddseq_2.15k': '15k/15k_preprocessing_out/data/fragments/HAR_ddseq_2.15k.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_1.15k': '15k/15k_preprocessing_out/data/fragments/MDC_mtscatac_1.15k.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_2.15k': '15k/15k_preprocessing_out/data/fragments/MDC_mtscatac_2.15k.fragments.tsv.gz',\n",
       " 'OHS_s3atac_1.15k': '15k/15k_preprocessing_out/data/fragments/OHS_s3atac_1.15k.fragments.tsv.gz',\n",
       " 'OHS_s3atac_2.15k': '15k/15k_preprocessing_out/data/fragments/OHS_s3atac_2.15k.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_1.15k': '15k/15k_preprocessing_out/data/fragments/SAN_10xmultiome_1.15k.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_2.15k': '15k/15k_preprocessing_out/data/fragments/SAN_10xmultiome_2.15k.fragments.tsv.gz',\n",
       " 'STA_10xv11_1.15k': '15k/15k_preprocessing_out/data/fragments/STA_10xv11_1.15k.fragments.tsv.gz',\n",
       " 'STA_10xv11_2.15k': '15k/15k_preprocessing_out/data/fragments/STA_10xv11_2.15k.fragments.tsv.gz',\n",
       " 'TXG_10xv11_1.15k': '15k/15k_preprocessing_out/data/fragments/TXG_10xv11_1.15k.fragments.tsv.gz',\n",
       " 'TXG_10xv2_1.15k': '15k/15k_preprocessing_out/data/fragments/TXG_10xv2_1.15k.fragments.tsv.gz',\n",
       " 'TXG_10xv2_2.15k': '15k/15k_preprocessing_out/data/fragments/TXG_10xv2_2.15k.fragments.tsv.gz',\n",
       " 'UCS_ddseq_1.15k': '15k/15k_preprocessing_out/data/fragments/UCS_ddseq_1.15k.fragments.tsv.gz',\n",
       " 'UCS_ddseq_2.15k': '15k/15k_preprocessing_out/data/fragments/UCS_ddseq_2.15k.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_1.15k': '15k/15k_preprocessing_out/data/fragments/VIB_10xmultiome_1.15k.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_2.15k': '15k/15k_preprocessing_out/data/fragments/VIB_10xmultiome_2.15k.fragments.tsv.gz',\n",
       " 'VIB_10xv1_1.15k': '15k/15k_preprocessing_out/data/fragments/VIB_10xv1_1.15k.fragments.tsv.gz',\n",
       " 'VIB_10xv1_2.15k': '15k/15k_preprocessing_out/data/fragments/VIB_10xv1_2.15k.fragments.tsv.gz',\n",
       " 'VIB_10xv2_1.15k': '15k/15k_preprocessing_out/data/fragments/VIB_10xv2_1.15k.fragments.tsv.gz',\n",
       " 'VIB_10xv2_2.15k': '15k/15k_preprocessing_out/data/fragments/VIB_10xv2_2.15k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_1.15k': '15k/15k_preprocessing_out/data/fragments/VIB_hydrop_1.15k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_11.15k': '15k/15k_preprocessing_out/data/fragments/VIB_hydrop_11.15k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_12.15k': '15k/15k_preprocessing_out/data/fragments/VIB_hydrop_12.15k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_2.15k': '15k/15k_preprocessing_out/data/fragments/VIB_hydrop_2.15k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_21.15k': '15k/15k_preprocessing_out/data/fragments/VIB_hydrop_21.15k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_22.15k': '15k/15k_preprocessing_out/data/fragments/VIB_hydrop_22.15k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_1.20k': '20k/20k_preprocessing_out/data/fragments/BIO_ddseq_1.20k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_2.20k': '20k/20k_preprocessing_out/data/fragments/BIO_ddseq_2.20k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_3.20k': '20k/20k_preprocessing_out/data/fragments/BIO_ddseq_3.20k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_4.20k': '20k/20k_preprocessing_out/data/fragments/BIO_ddseq_4.20k.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_1.20k': '20k/20k_preprocessing_out/data/fragments/BRO_mtscatac_1.20k.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_2.20k': '20k/20k_preprocessing_out/data/fragments/BRO_mtscatac_2.20k.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_1.20k': '20k/20k_preprocessing_out/data/fragments/CNA_10xmultiome_1.20k.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_2.20k': '20k/20k_preprocessing_out/data/fragments/CNA_10xmultiome_2.20k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_1.20k': '20k/20k_preprocessing_out/data/fragments/CNA_10xv11_1.20k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_2.20k': '20k/20k_preprocessing_out/data/fragments/CNA_10xv11_2.20k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_3.20k': '20k/20k_preprocessing_out/data/fragments/CNA_10xv11_3.20k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_4.20k': '20k/20k_preprocessing_out/data/fragments/CNA_10xv11_4.20k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_5.20k': '20k/20k_preprocessing_out/data/fragments/CNA_10xv11_5.20k.fragments.tsv.gz',\n",
       " 'CNA_10xv2_1.20k': '20k/20k_preprocessing_out/data/fragments/CNA_10xv2_1.20k.fragments.tsv.gz',\n",
       " 'CNA_10xv2_2.20k': '20k/20k_preprocessing_out/data/fragments/CNA_10xv2_2.20k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_1.20k': '20k/20k_preprocessing_out/data/fragments/CNA_hydrop_1.20k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_2.20k': '20k/20k_preprocessing_out/data/fragments/CNA_hydrop_2.20k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_3.20k': '20k/20k_preprocessing_out/data/fragments/CNA_hydrop_3.20k.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_1.20k': '20k/20k_preprocessing_out/data/fragments/CNA_mtscatac_1.20k.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_2.20k': '20k/20k_preprocessing_out/data/fragments/CNA_mtscatac_2.20k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_1.20k': '20k/20k_preprocessing_out/data/fragments/EPF_hydrop_1.20k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_2.20k': '20k/20k_preprocessing_out/data/fragments/EPF_hydrop_2.20k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_3.20k': '20k/20k_preprocessing_out/data/fragments/EPF_hydrop_3.20k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_4.20k': '20k/20k_preprocessing_out/data/fragments/EPF_hydrop_4.20k.fragments.tsv.gz',\n",
       " 'HAR_ddseq_1.20k': '20k/20k_preprocessing_out/data/fragments/HAR_ddseq_1.20k.fragments.tsv.gz',\n",
       " 'HAR_ddseq_2.20k': '20k/20k_preprocessing_out/data/fragments/HAR_ddseq_2.20k.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_1.20k': '20k/20k_preprocessing_out/data/fragments/MDC_mtscatac_1.20k.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_2.20k': '20k/20k_preprocessing_out/data/fragments/MDC_mtscatac_2.20k.fragments.tsv.gz',\n",
       " 'OHS_s3atac_1.20k': '20k/20k_preprocessing_out/data/fragments/OHS_s3atac_1.20k.fragments.tsv.gz',\n",
       " 'OHS_s3atac_2.20k': '20k/20k_preprocessing_out/data/fragments/OHS_s3atac_2.20k.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_1.20k': '20k/20k_preprocessing_out/data/fragments/SAN_10xmultiome_1.20k.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_2.20k': '20k/20k_preprocessing_out/data/fragments/SAN_10xmultiome_2.20k.fragments.tsv.gz',\n",
       " 'STA_10xv11_1.20k': '20k/20k_preprocessing_out/data/fragments/STA_10xv11_1.20k.fragments.tsv.gz',\n",
       " 'STA_10xv11_2.20k': '20k/20k_preprocessing_out/data/fragments/STA_10xv11_2.20k.fragments.tsv.gz',\n",
       " 'TXG_10xv11_1.20k': '20k/20k_preprocessing_out/data/fragments/TXG_10xv11_1.20k.fragments.tsv.gz',\n",
       " 'TXG_10xv2_1.20k': '20k/20k_preprocessing_out/data/fragments/TXG_10xv2_1.20k.fragments.tsv.gz',\n",
       " 'TXG_10xv2_2.20k': '20k/20k_preprocessing_out/data/fragments/TXG_10xv2_2.20k.fragments.tsv.gz',\n",
       " 'UCS_ddseq_1.20k': '20k/20k_preprocessing_out/data/fragments/UCS_ddseq_1.20k.fragments.tsv.gz',\n",
       " 'UCS_ddseq_2.20k': '20k/20k_preprocessing_out/data/fragments/UCS_ddseq_2.20k.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_1.20k': '20k/20k_preprocessing_out/data/fragments/VIB_10xmultiome_1.20k.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_2.20k': '20k/20k_preprocessing_out/data/fragments/VIB_10xmultiome_2.20k.fragments.tsv.gz',\n",
       " 'VIB_10xv1_1.20k': '20k/20k_preprocessing_out/data/fragments/VIB_10xv1_1.20k.fragments.tsv.gz',\n",
       " 'VIB_10xv1_2.20k': '20k/20k_preprocessing_out/data/fragments/VIB_10xv1_2.20k.fragments.tsv.gz',\n",
       " 'VIB_10xv2_1.20k': '20k/20k_preprocessing_out/data/fragments/VIB_10xv2_1.20k.fragments.tsv.gz',\n",
       " 'VIB_10xv2_2.20k': '20k/20k_preprocessing_out/data/fragments/VIB_10xv2_2.20k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_1.20k': '20k/20k_preprocessing_out/data/fragments/VIB_hydrop_1.20k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_11.20k': '20k/20k_preprocessing_out/data/fragments/VIB_hydrop_11.20k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_12.20k': '20k/20k_preprocessing_out/data/fragments/VIB_hydrop_12.20k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_2.20k': '20k/20k_preprocessing_out/data/fragments/VIB_hydrop_2.20k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_21.20k': '20k/20k_preprocessing_out/data/fragments/VIB_hydrop_21.20k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_22.20k': '20k/20k_preprocessing_out/data/fragments/VIB_hydrop_22.20k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_1.25k': '25k/25k_preprocessing_out/data/fragments/BIO_ddseq_1.25k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_2.25k': '25k/25k_preprocessing_out/data/fragments/BIO_ddseq_2.25k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_3.25k': '25k/25k_preprocessing_out/data/fragments/BIO_ddseq_3.25k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_4.25k': '25k/25k_preprocessing_out/data/fragments/BIO_ddseq_4.25k.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_1.25k': '25k/25k_preprocessing_out/data/fragments/BRO_mtscatac_1.25k.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_2.25k': '25k/25k_preprocessing_out/data/fragments/BRO_mtscatac_2.25k.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_1.25k': '25k/25k_preprocessing_out/data/fragments/CNA_10xmultiome_1.25k.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_2.25k': '25k/25k_preprocessing_out/data/fragments/CNA_10xmultiome_2.25k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_1.25k': '25k/25k_preprocessing_out/data/fragments/CNA_10xv11_1.25k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_2.25k': '25k/25k_preprocessing_out/data/fragments/CNA_10xv11_2.25k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_3.25k': '25k/25k_preprocessing_out/data/fragments/CNA_10xv11_3.25k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_4.25k': '25k/25k_preprocessing_out/data/fragments/CNA_10xv11_4.25k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_5.25k': '25k/25k_preprocessing_out/data/fragments/CNA_10xv11_5.25k.fragments.tsv.gz',\n",
       " 'CNA_10xv2_1.25k': '25k/25k_preprocessing_out/data/fragments/CNA_10xv2_1.25k.fragments.tsv.gz',\n",
       " 'CNA_10xv2_2.25k': '25k/25k_preprocessing_out/data/fragments/CNA_10xv2_2.25k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_1.25k': '25k/25k_preprocessing_out/data/fragments/CNA_hydrop_1.25k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_2.25k': '25k/25k_preprocessing_out/data/fragments/CNA_hydrop_2.25k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_3.25k': '25k/25k_preprocessing_out/data/fragments/CNA_hydrop_3.25k.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_1.25k': '25k/25k_preprocessing_out/data/fragments/CNA_mtscatac_1.25k.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_2.25k': '25k/25k_preprocessing_out/data/fragments/CNA_mtscatac_2.25k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_1.25k': '25k/25k_preprocessing_out/data/fragments/EPF_hydrop_1.25k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_2.25k': '25k/25k_preprocessing_out/data/fragments/EPF_hydrop_2.25k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_3.25k': '25k/25k_preprocessing_out/data/fragments/EPF_hydrop_3.25k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_4.25k': '25k/25k_preprocessing_out/data/fragments/EPF_hydrop_4.25k.fragments.tsv.gz',\n",
       " 'HAR_ddseq_1.25k': '25k/25k_preprocessing_out/data/fragments/HAR_ddseq_1.25k.fragments.tsv.gz',\n",
       " 'HAR_ddseq_2.25k': '25k/25k_preprocessing_out/data/fragments/HAR_ddseq_2.25k.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_1.25k': '25k/25k_preprocessing_out/data/fragments/MDC_mtscatac_1.25k.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_2.25k': '25k/25k_preprocessing_out/data/fragments/MDC_mtscatac_2.25k.fragments.tsv.gz',\n",
       " 'OHS_s3atac_1.25k': '25k/25k_preprocessing_out/data/fragments/OHS_s3atac_1.25k.fragments.tsv.gz',\n",
       " 'OHS_s3atac_2.25k': '25k/25k_preprocessing_out/data/fragments/OHS_s3atac_2.25k.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_1.25k': '25k/25k_preprocessing_out/data/fragments/SAN_10xmultiome_1.25k.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_2.25k': '25k/25k_preprocessing_out/data/fragments/SAN_10xmultiome_2.25k.fragments.tsv.gz',\n",
       " 'STA_10xv11_1.25k': '25k/25k_preprocessing_out/data/fragments/STA_10xv11_1.25k.fragments.tsv.gz',\n",
       " 'STA_10xv11_2.25k': '25k/25k_preprocessing_out/data/fragments/STA_10xv11_2.25k.fragments.tsv.gz',\n",
       " 'TXG_10xv11_1.25k': '25k/25k_preprocessing_out/data/fragments/TXG_10xv11_1.25k.fragments.tsv.gz',\n",
       " 'TXG_10xv2_1.25k': '25k/25k_preprocessing_out/data/fragments/TXG_10xv2_1.25k.fragments.tsv.gz',\n",
       " 'TXG_10xv2_2.25k': '25k/25k_preprocessing_out/data/fragments/TXG_10xv2_2.25k.fragments.tsv.gz',\n",
       " 'UCS_ddseq_1.25k': '25k/25k_preprocessing_out/data/fragments/UCS_ddseq_1.25k.fragments.tsv.gz',\n",
       " 'UCS_ddseq_2.25k': '25k/25k_preprocessing_out/data/fragments/UCS_ddseq_2.25k.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_1.25k': '25k/25k_preprocessing_out/data/fragments/VIB_10xmultiome_1.25k.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_2.25k': '25k/25k_preprocessing_out/data/fragments/VIB_10xmultiome_2.25k.fragments.tsv.gz',\n",
       " 'VIB_10xv1_1.25k': '25k/25k_preprocessing_out/data/fragments/VIB_10xv1_1.25k.fragments.tsv.gz',\n",
       " 'VIB_10xv1_2.25k': '25k/25k_preprocessing_out/data/fragments/VIB_10xv1_2.25k.fragments.tsv.gz',\n",
       " 'VIB_10xv2_1.25k': '25k/25k_preprocessing_out/data/fragments/VIB_10xv2_1.25k.fragments.tsv.gz',\n",
       " 'VIB_10xv2_2.25k': '25k/25k_preprocessing_out/data/fragments/VIB_10xv2_2.25k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_1.25k': '25k/25k_preprocessing_out/data/fragments/VIB_hydrop_1.25k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_11.25k': '25k/25k_preprocessing_out/data/fragments/VIB_hydrop_11.25k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_12.25k': '25k/25k_preprocessing_out/data/fragments/VIB_hydrop_12.25k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_2.25k': '25k/25k_preprocessing_out/data/fragments/VIB_hydrop_2.25k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_21.25k': '25k/25k_preprocessing_out/data/fragments/VIB_hydrop_21.25k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_22.25k': '25k/25k_preprocessing_out/data/fragments/VIB_hydrop_22.25k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_1.30k': '30k/30k_preprocessing_out/data/fragments/BIO_ddseq_1.30k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_2.30k': '30k/30k_preprocessing_out/data/fragments/BIO_ddseq_2.30k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_3.30k': '30k/30k_preprocessing_out/data/fragments/BIO_ddseq_3.30k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_4.30k': '30k/30k_preprocessing_out/data/fragments/BIO_ddseq_4.30k.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_1.30k': '30k/30k_preprocessing_out/data/fragments/BRO_mtscatac_1.30k.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_2.30k': '30k/30k_preprocessing_out/data/fragments/BRO_mtscatac_2.30k.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_1.30k': '30k/30k_preprocessing_out/data/fragments/CNA_10xmultiome_1.30k.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_2.30k': '30k/30k_preprocessing_out/data/fragments/CNA_10xmultiome_2.30k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_1.30k': '30k/30k_preprocessing_out/data/fragments/CNA_10xv11_1.30k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_2.30k': '30k/30k_preprocessing_out/data/fragments/CNA_10xv11_2.30k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_3.30k': '30k/30k_preprocessing_out/data/fragments/CNA_10xv11_3.30k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_4.30k': '30k/30k_preprocessing_out/data/fragments/CNA_10xv11_4.30k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_5.30k': '30k/30k_preprocessing_out/data/fragments/CNA_10xv11_5.30k.fragments.tsv.gz',\n",
       " 'CNA_10xv2_1.30k': '30k/30k_preprocessing_out/data/fragments/CNA_10xv2_1.30k.fragments.tsv.gz',\n",
       " 'CNA_10xv2_2.30k': '30k/30k_preprocessing_out/data/fragments/CNA_10xv2_2.30k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_1.30k': '30k/30k_preprocessing_out/data/fragments/CNA_hydrop_1.30k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_2.30k': '30k/30k_preprocessing_out/data/fragments/CNA_hydrop_2.30k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_3.30k': '30k/30k_preprocessing_out/data/fragments/CNA_hydrop_3.30k.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_1.30k': '30k/30k_preprocessing_out/data/fragments/CNA_mtscatac_1.30k.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_2.30k': '30k/30k_preprocessing_out/data/fragments/CNA_mtscatac_2.30k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_1.30k': '30k/30k_preprocessing_out/data/fragments/EPF_hydrop_1.30k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_2.30k': '30k/30k_preprocessing_out/data/fragments/EPF_hydrop_2.30k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_3.30k': '30k/30k_preprocessing_out/data/fragments/EPF_hydrop_3.30k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_4.30k': '30k/30k_preprocessing_out/data/fragments/EPF_hydrop_4.30k.fragments.tsv.gz',\n",
       " 'HAR_ddseq_1.30k': '30k/30k_preprocessing_out/data/fragments/HAR_ddseq_1.30k.fragments.tsv.gz',\n",
       " 'HAR_ddseq_2.30k': '30k/30k_preprocessing_out/data/fragments/HAR_ddseq_2.30k.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_1.30k': '30k/30k_preprocessing_out/data/fragments/MDC_mtscatac_1.30k.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_2.30k': '30k/30k_preprocessing_out/data/fragments/MDC_mtscatac_2.30k.fragments.tsv.gz',\n",
       " 'OHS_s3atac_1.30k': '30k/30k_preprocessing_out/data/fragments/OHS_s3atac_1.30k.fragments.tsv.gz',\n",
       " 'OHS_s3atac_2.30k': '30k/30k_preprocessing_out/data/fragments/OHS_s3atac_2.30k.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_1.30k': '30k/30k_preprocessing_out/data/fragments/SAN_10xmultiome_1.30k.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_2.30k': '30k/30k_preprocessing_out/data/fragments/SAN_10xmultiome_2.30k.fragments.tsv.gz',\n",
       " 'STA_10xv11_1.30k': '30k/30k_preprocessing_out/data/fragments/STA_10xv11_1.30k.fragments.tsv.gz',\n",
       " 'STA_10xv11_2.30k': '30k/30k_preprocessing_out/data/fragments/STA_10xv11_2.30k.fragments.tsv.gz',\n",
       " 'TXG_10xv11_1.30k': '30k/30k_preprocessing_out/data/fragments/TXG_10xv11_1.30k.fragments.tsv.gz',\n",
       " 'TXG_10xv2_1.30k': '30k/30k_preprocessing_out/data/fragments/TXG_10xv2_1.30k.fragments.tsv.gz',\n",
       " 'TXG_10xv2_2.30k': '30k/30k_preprocessing_out/data/fragments/TXG_10xv2_2.30k.fragments.tsv.gz',\n",
       " 'UCS_ddseq_1.30k': '30k/30k_preprocessing_out/data/fragments/UCS_ddseq_1.30k.fragments.tsv.gz',\n",
       " 'UCS_ddseq_2.30k': '30k/30k_preprocessing_out/data/fragments/UCS_ddseq_2.30k.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_1.30k': '30k/30k_preprocessing_out/data/fragments/VIB_10xmultiome_1.30k.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_2.30k': '30k/30k_preprocessing_out/data/fragments/VIB_10xmultiome_2.30k.fragments.tsv.gz',\n",
       " 'VIB_10xv1_1.30k': '30k/30k_preprocessing_out/data/fragments/VIB_10xv1_1.30k.fragments.tsv.gz',\n",
       " 'VIB_10xv1_2.30k': '30k/30k_preprocessing_out/data/fragments/VIB_10xv1_2.30k.fragments.tsv.gz',\n",
       " 'VIB_10xv2_1.30k': '30k/30k_preprocessing_out/data/fragments/VIB_10xv2_1.30k.fragments.tsv.gz',\n",
       " 'VIB_10xv2_2.30k': '30k/30k_preprocessing_out/data/fragments/VIB_10xv2_2.30k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_1.30k': '30k/30k_preprocessing_out/data/fragments/VIB_hydrop_1.30k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_11.30k': '30k/30k_preprocessing_out/data/fragments/VIB_hydrop_11.30k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_12.30k': '30k/30k_preprocessing_out/data/fragments/VIB_hydrop_12.30k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_2.30k': '30k/30k_preprocessing_out/data/fragments/VIB_hydrop_2.30k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_21.30k': '30k/30k_preprocessing_out/data/fragments/VIB_hydrop_21.30k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_22.30k': '30k/30k_preprocessing_out/data/fragments/VIB_hydrop_22.30k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_1.35k': '35k/35k_preprocessing_out/data/fragments/BIO_ddseq_1.35k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_2.35k': '35k/35k_preprocessing_out/data/fragments/BIO_ddseq_2.35k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_3.35k': '35k/35k_preprocessing_out/data/fragments/BIO_ddseq_3.35k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_4.35k': '35k/35k_preprocessing_out/data/fragments/BIO_ddseq_4.35k.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_1.35k': '35k/35k_preprocessing_out/data/fragments/BRO_mtscatac_1.35k.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_2.35k': '35k/35k_preprocessing_out/data/fragments/BRO_mtscatac_2.35k.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_1.35k': '35k/35k_preprocessing_out/data/fragments/CNA_10xmultiome_1.35k.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_2.35k': '35k/35k_preprocessing_out/data/fragments/CNA_10xmultiome_2.35k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_1.35k': '35k/35k_preprocessing_out/data/fragments/CNA_10xv11_1.35k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_2.35k': '35k/35k_preprocessing_out/data/fragments/CNA_10xv11_2.35k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_3.35k': '35k/35k_preprocessing_out/data/fragments/CNA_10xv11_3.35k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_4.35k': '35k/35k_preprocessing_out/data/fragments/CNA_10xv11_4.35k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_5.35k': '35k/35k_preprocessing_out/data/fragments/CNA_10xv11_5.35k.fragments.tsv.gz',\n",
       " 'CNA_10xv2_1.35k': '35k/35k_preprocessing_out/data/fragments/CNA_10xv2_1.35k.fragments.tsv.gz',\n",
       " 'CNA_10xv2_2.35k': '35k/35k_preprocessing_out/data/fragments/CNA_10xv2_2.35k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_1.35k': '35k/35k_preprocessing_out/data/fragments/CNA_hydrop_1.35k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_2.35k': '35k/35k_preprocessing_out/data/fragments/CNA_hydrop_2.35k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_3.35k': '35k/35k_preprocessing_out/data/fragments/CNA_hydrop_3.35k.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_1.35k': '35k/35k_preprocessing_out/data/fragments/CNA_mtscatac_1.35k.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_2.35k': '35k/35k_preprocessing_out/data/fragments/CNA_mtscatac_2.35k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_1.35k': '35k/35k_preprocessing_out/data/fragments/EPF_hydrop_1.35k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_2.35k': '35k/35k_preprocessing_out/data/fragments/EPF_hydrop_2.35k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_3.35k': '35k/35k_preprocessing_out/data/fragments/EPF_hydrop_3.35k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_4.35k': '35k/35k_preprocessing_out/data/fragments/EPF_hydrop_4.35k.fragments.tsv.gz',\n",
       " 'HAR_ddseq_1.35k': '35k/35k_preprocessing_out/data/fragments/HAR_ddseq_1.35k.fragments.tsv.gz',\n",
       " 'HAR_ddseq_2.35k': '35k/35k_preprocessing_out/data/fragments/HAR_ddseq_2.35k.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_1.35k': '35k/35k_preprocessing_out/data/fragments/MDC_mtscatac_1.35k.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_2.35k': '35k/35k_preprocessing_out/data/fragments/MDC_mtscatac_2.35k.fragments.tsv.gz',\n",
       " 'OHS_s3atac_1.35k': '35k/35k_preprocessing_out/data/fragments/OHS_s3atac_1.35k.fragments.tsv.gz',\n",
       " 'OHS_s3atac_2.35k': '35k/35k_preprocessing_out/data/fragments/OHS_s3atac_2.35k.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_1.35k': '35k/35k_preprocessing_out/data/fragments/SAN_10xmultiome_1.35k.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_2.35k': '35k/35k_preprocessing_out/data/fragments/SAN_10xmultiome_2.35k.fragments.tsv.gz',\n",
       " 'STA_10xv11_1.35k': '35k/35k_preprocessing_out/data/fragments/STA_10xv11_1.35k.fragments.tsv.gz',\n",
       " 'STA_10xv11_2.35k': '35k/35k_preprocessing_out/data/fragments/STA_10xv11_2.35k.fragments.tsv.gz',\n",
       " 'TXG_10xv11_1.35k': '35k/35k_preprocessing_out/data/fragments/TXG_10xv11_1.35k.fragments.tsv.gz',\n",
       " 'TXG_10xv2_1.35k': '35k/35k_preprocessing_out/data/fragments/TXG_10xv2_1.35k.fragments.tsv.gz',\n",
       " 'TXG_10xv2_2.35k': '35k/35k_preprocessing_out/data/fragments/TXG_10xv2_2.35k.fragments.tsv.gz',\n",
       " 'UCS_ddseq_1.35k': '35k/35k_preprocessing_out/data/fragments/UCS_ddseq_1.35k.fragments.tsv.gz',\n",
       " 'UCS_ddseq_2.35k': '35k/35k_preprocessing_out/data/fragments/UCS_ddseq_2.35k.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_1.35k': '35k/35k_preprocessing_out/data/fragments/VIB_10xmultiome_1.35k.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_2.35k': '35k/35k_preprocessing_out/data/fragments/VIB_10xmultiome_2.35k.fragments.tsv.gz',\n",
       " 'VIB_10xv1_1.35k': '35k/35k_preprocessing_out/data/fragments/VIB_10xv1_1.35k.fragments.tsv.gz',\n",
       " 'VIB_10xv1_2.35k': '35k/35k_preprocessing_out/data/fragments/VIB_10xv1_2.35k.fragments.tsv.gz',\n",
       " 'VIB_10xv2_1.35k': '35k/35k_preprocessing_out/data/fragments/VIB_10xv2_1.35k.fragments.tsv.gz',\n",
       " 'VIB_10xv2_2.35k': '35k/35k_preprocessing_out/data/fragments/VIB_10xv2_2.35k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_1.35k': '35k/35k_preprocessing_out/data/fragments/VIB_hydrop_1.35k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_11.35k': '35k/35k_preprocessing_out/data/fragments/VIB_hydrop_11.35k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_12.35k': '35k/35k_preprocessing_out/data/fragments/VIB_hydrop_12.35k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_2.35k': '35k/35k_preprocessing_out/data/fragments/VIB_hydrop_2.35k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_21.35k': '35k/35k_preprocessing_out/data/fragments/VIB_hydrop_21.35k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_22.35k': '35k/35k_preprocessing_out/data/fragments/VIB_hydrop_22.35k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_1.5k': '5k/5k_preprocessing_out/data/fragments/BIO_ddseq_1.5k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_2.5k': '5k/5k_preprocessing_out/data/fragments/BIO_ddseq_2.5k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_3.5k': '5k/5k_preprocessing_out/data/fragments/BIO_ddseq_3.5k.fragments.tsv.gz',\n",
       " 'BIO_ddseq_4.5k': '5k/5k_preprocessing_out/data/fragments/BIO_ddseq_4.5k.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_1.5k': '5k/5k_preprocessing_out/data/fragments/BRO_mtscatac_1.5k.fragments.tsv.gz',\n",
       " 'BRO_mtscatac_2.5k': '5k/5k_preprocessing_out/data/fragments/BRO_mtscatac_2.5k.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_1.5k': '5k/5k_preprocessing_out/data/fragments/CNA_10xmultiome_1.5k.fragments.tsv.gz',\n",
       " 'CNA_10xmultiome_2.5k': '5k/5k_preprocessing_out/data/fragments/CNA_10xmultiome_2.5k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_1.5k': '5k/5k_preprocessing_out/data/fragments/CNA_10xv11_1.5k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_2.5k': '5k/5k_preprocessing_out/data/fragments/CNA_10xv11_2.5k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_3.5k': '5k/5k_preprocessing_out/data/fragments/CNA_10xv11_3.5k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_4.5k': '5k/5k_preprocessing_out/data/fragments/CNA_10xv11_4.5k.fragments.tsv.gz',\n",
       " 'CNA_10xv11_5.5k': '5k/5k_preprocessing_out/data/fragments/CNA_10xv11_5.5k.fragments.tsv.gz',\n",
       " 'CNA_10xv2_1.5k': '5k/5k_preprocessing_out/data/fragments/CNA_10xv2_1.5k.fragments.tsv.gz',\n",
       " 'CNA_10xv2_2.5k': '5k/5k_preprocessing_out/data/fragments/CNA_10xv2_2.5k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_1.5k': '5k/5k_preprocessing_out/data/fragments/CNA_hydrop_1.5k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_2.5k': '5k/5k_preprocessing_out/data/fragments/CNA_hydrop_2.5k.fragments.tsv.gz',\n",
       " 'CNA_hydrop_3.5k': '5k/5k_preprocessing_out/data/fragments/CNA_hydrop_3.5k.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_1.5k': '5k/5k_preprocessing_out/data/fragments/CNA_mtscatac_1.5k.fragments.tsv.gz',\n",
       " 'CNA_mtscatac_2.5k': '5k/5k_preprocessing_out/data/fragments/CNA_mtscatac_2.5k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_1.5k': '5k/5k_preprocessing_out/data/fragments/EPF_hydrop_1.5k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_2.5k': '5k/5k_preprocessing_out/data/fragments/EPF_hydrop_2.5k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_3.5k': '5k/5k_preprocessing_out/data/fragments/EPF_hydrop_3.5k.fragments.tsv.gz',\n",
       " 'EPF_hydrop_4.5k': '5k/5k_preprocessing_out/data/fragments/EPF_hydrop_4.5k.fragments.tsv.gz',\n",
       " 'HAR_ddseq_1.5k': '5k/5k_preprocessing_out/data/fragments/HAR_ddseq_1.5k.fragments.tsv.gz',\n",
       " 'HAR_ddseq_2.5k': '5k/5k_preprocessing_out/data/fragments/HAR_ddseq_2.5k.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_1.5k': '5k/5k_preprocessing_out/data/fragments/MDC_mtscatac_1.5k.fragments.tsv.gz',\n",
       " 'MDC_mtscatac_2.5k': '5k/5k_preprocessing_out/data/fragments/MDC_mtscatac_2.5k.fragments.tsv.gz',\n",
       " 'OHS_s3atac_1.5k': '5k/5k_preprocessing_out/data/fragments/OHS_s3atac_1.5k.fragments.tsv.gz',\n",
       " 'OHS_s3atac_2.5k': '5k/5k_preprocessing_out/data/fragments/OHS_s3atac_2.5k.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_1.5k': '5k/5k_preprocessing_out/data/fragments/SAN_10xmultiome_1.5k.fragments.tsv.gz',\n",
       " 'SAN_10xmultiome_2.5k': '5k/5k_preprocessing_out/data/fragments/SAN_10xmultiome_2.5k.fragments.tsv.gz',\n",
       " 'STA_10xv11_1.5k': '5k/5k_preprocessing_out/data/fragments/STA_10xv11_1.5k.fragments.tsv.gz',\n",
       " 'STA_10xv11_2.5k': '5k/5k_preprocessing_out/data/fragments/STA_10xv11_2.5k.fragments.tsv.gz',\n",
       " 'TXG_10xv11_1.5k': '5k/5k_preprocessing_out/data/fragments/TXG_10xv11_1.5k.fragments.tsv.gz',\n",
       " 'TXG_10xv2_1.5k': '5k/5k_preprocessing_out/data/fragments/TXG_10xv2_1.5k.fragments.tsv.gz',\n",
       " 'TXG_10xv2_2.5k': '5k/5k_preprocessing_out/data/fragments/TXG_10xv2_2.5k.fragments.tsv.gz',\n",
       " 'UCS_ddseq_1.5k': '5k/5k_preprocessing_out/data/fragments/UCS_ddseq_1.5k.fragments.tsv.gz',\n",
       " 'UCS_ddseq_2.5k': '5k/5k_preprocessing_out/data/fragments/UCS_ddseq_2.5k.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_1.5k': '5k/5k_preprocessing_out/data/fragments/VIB_10xmultiome_1.5k.fragments.tsv.gz',\n",
       " 'VIB_10xmultiome_2.5k': '5k/5k_preprocessing_out/data/fragments/VIB_10xmultiome_2.5k.fragments.tsv.gz',\n",
       " 'VIB_10xv1_1.5k': '5k/5k_preprocessing_out/data/fragments/VIB_10xv1_1.5k.fragments.tsv.gz',\n",
       " 'VIB_10xv1_2.5k': '5k/5k_preprocessing_out/data/fragments/VIB_10xv1_2.5k.fragments.tsv.gz',\n",
       " 'VIB_10xv2_1.5k': '5k/5k_preprocessing_out/data/fragments/VIB_10xv2_1.5k.fragments.tsv.gz',\n",
       " 'VIB_10xv2_2.5k': '5k/5k_preprocessing_out/data/fragments/VIB_10xv2_2.5k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_1.5k': '5k/5k_preprocessing_out/data/fragments/VIB_hydrop_1.5k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_11.5k': '5k/5k_preprocessing_out/data/fragments/VIB_hydrop_11.5k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_12.5k': '5k/5k_preprocessing_out/data/fragments/VIB_hydrop_12.5k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_2.5k': '5k/5k_preprocessing_out/data/fragments/VIB_hydrop_2.5k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_21.5k': '5k/5k_preprocessing_out/data/fragments/VIB_hydrop_21.5k.fragments.tsv.gz',\n",
       " 'VIB_hydrop_22.5k': '5k/5k_preprocessing_out/data/fragments/VIB_hydrop_22.5k.fragments.tsv.gz'}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "fragments_list = sorted(\n",
    "    glob.glob(f\"*k/*_preprocessing_out/data/fragments/*fragments.tsv.gz\")\n",
    ")\n",
    "fragments_dict = {}\n",
    "for fragments_file in fragments_list:\n",
    "    sample = fragments_file.split(\"/\")[-1].split(\".fragments.tsv.gz\")[0]\n",
    "    fragments_dict[sample] = fragments_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "9893d46a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-08T12:44:48.110222Z",
     "iopub.status.busy": "2022-12-08T12:44:48.109609Z",
     "iopub.status.idle": "2022-12-08T12:46:12.504246Z",
     "shell.execute_reply": "2022-12-08T12:46:12.502506Z",
     "shell.execute_reply.started": "2022-12-08T12:44:48.110181Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_27578/2738452975.py:61: DeprecationWarning: `has_headers` is deprecated as an argument to `read_csv`; use `has_header` instead.\n",
      "  pl.read_csv(\n"
     ]
    }
   ],
   "source": [
    "fragments_df = read_fragments_from_file(fragments_dict['BIO_ddseq_1.10k'], use_polars=True).df\n",
    "fragments_df[\"Score\"] = 1\n",
    "fragments_df = pr.PyRanges(fragments_df)\n",
    "\n",
    "# d_fragments_from_file(fragments_dict['BIO_ddseq_1.10k'])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "20221014_pycistopic.sif",
   "language": "python",
   "name": "cistopic_20221014"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
