{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "4a01f27a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35k/35k_preprocessing_out/data/bam/VIB_hydrop_22.35k___VIB_hydrop_22__R1.bwa.out.possorted.bam\n"
     ]
    }
   ],
   "source": [
    "echo $file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "cf6c0cda",
   "metadata": {},
   "outputs": [],
   "source": [
    "singularity exec --cleanenv -H $PWD -B /dodrio,/readonly/dodrio,/tmp /dodrio/scratch/projects/starting_2022_023/benchmark/scatac_benchmark/0_resources/vsn_cache/vibsinglecellnf-singlecelltoolkit-2022-07-07-0638c1d.img bash"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "efcdcc6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping_summary () {\n",
    "    mapping_stats=\"${1}\";\n",
    "    bam=\"${2}\";\n",
    "    sampleId=\"${3}\";\n",
    "\n",
    "    # Get mapping statistics from BAM file:\n",
    "    #   - Read BAM file and write uncompressed BAM.\n",
    "    #   - Uncompressed BAM file is written to each samtools command with tee (writes to each specified file and stdout).\n",
    "    #   - samtools commands:\n",
    "    #       - Get samtools statistics with:\n",
    "    #           samtools stat \"${bam}\" > \"${sampleId}.stat\"\n",
    "    #       - Uniquely mapped reads (BWA):\n",
    "    #           samtools view -c -F 0x4 -F 0x100 -F 0x800 -e '! [XA] && ! [SA]' \"${bam}\"\n",
    "    #       - Fraction of total read pairs mapped confidently to genome (>30 mapq):\n",
    "    #           samtools view -c -F 0x4 -F 0x100 -F 0x800 -q 30 \"${bam}\"\n",
    "    #   - Only use threads for \"samtools stat\". Using it with any of the other samtools commands\n",
    "    #     makes everything slower than not using any threads at all.\n",
    "    samtools view -u \"${bam}\" \\\n",
    "      | tee \\\n",
    "            >(samtools view -c -F 0x4 -F 0x100 -F 0x800 -e '! [XA] && ! [SA]' - > \"${sampleId}.uniquely_mapped_reads.txt\") \\\n",
    "            >(samtools view -c -F 0x4 -F 0x100 -F 0x800 -q 30 - > \"${sampleId}.fraction_total_read_pairs.txt\") \\\n",
    "      | samtools stat -@ 2 - > \"${sampleId}.stat\"\n",
    "\n",
    "\n",
    "    # Output file:\n",
    "    printf \"\\t${sampleId}\\n\" > \"${mapping_stats}\";\n",
    "\n",
    "    grep '^SN' \"${sampleId}.stat\" | cut -f 2,3 >> \"${mapping_stats}\";\n",
    "\n",
    "    printf \"Uniquely mapped reads:\\t\" >> \"${mapping_stats}\";\n",
    "    cat \"${sampleId}.uniquely_mapped_reads.txt\" >> \"${mapping_stats}\";\n",
    "\n",
    "    printf \"Reads mapped with MAPQ>30:\\t\" >> \"${mapping_stats}\";\n",
    "    cat \"${sampleId}.fraction_total_read_pairs.txt\" >> \"${mapping_stats}\";\n",
    "\n",
    "    rm \"${sampleId}.uniquely_mapped_reads.txt\" \"${sampleId}.fraction_total_read_pairs.txt\";\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "83f3778c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35k\n",
      "35k/35k_preprocessing_out/data/bam/BIO_ddseq_1.35k___BIO_ddseq_1__R1.bwa.out.possorted.bam\n",
      "\tBIO_ddseq_1.35k\n",
      "\tBIO_ddseq_1\n",
      "35k/35k_preprocessing_out/data/bam/BIO_ddseq_2.35k___BIO_ddseq_2__R1.bwa.out.possorted.bam\n",
      "\tBIO_ddseq_2.35k\n",
      "\tBIO_ddseq_2\n",
      "35k/35k_preprocessing_out/data/bam/BIO_ddseq_3.35k___BIO_ddseq_3__R1.bwa.out.possorted.bam\n",
      "\tBIO_ddseq_3.35k\n",
      "\tBIO_ddseq_3\n",
      "35k/35k_preprocessing_out/data/bam/BIO_ddseq_4.35k___BIO_ddseq_4__R1.bwa.out.possorted.bam\n",
      "\tBIO_ddseq_4.35k\n",
      "\tBIO_ddseq_4\n",
      "35k/35k_preprocessing_out/data/bam/BRO_mtscatac_1.35k___BRO_mtscatac_1__R1.bwa.out.possorted.bam\n",
      "\tBRO_mtscatac_1.35k\n",
      "\tBRO_mtscatac_1\n",
      "35k/35k_preprocessing_out/data/bam/BRO_mtscatac_2.35k___BRO_mtscatac_2__R1.bwa.out.possorted.bam\n",
      "\tBRO_mtscatac_2.35k\n",
      "\tBRO_mtscatac_2\n",
      "35k/35k_preprocessing_out/data/bam/CNA_10xmultiome_1.35k___CNA_10xmultiome_1__R1.bwa.out.possorted.bam\n",
      "\tCNA_10xmultiome_1.35k\n",
      "\tCNA_10xmultiome_1\n",
      "35k/35k_preprocessing_out/data/bam/CNA_10xmultiome_2.35k___CNA_10xmultiome_2__R1.bwa.out.possorted.bam\n",
      "\tCNA_10xmultiome_2.35k\n",
      "\tCNA_10xmultiome_2\n",
      "35k/35k_preprocessing_out/data/bam/CNA_10xv11_1.35k___CNA_10xv11_1__R1.bwa.out.possorted.bam\n",
      "\tCNA_10xv11_1.35k\n",
      "\tCNA_10xv11_1\n",
      "35k/35k_preprocessing_out/data/bam/CNA_10xv11_2.35k___CNA_10xv11_2__R1.bwa.out.possorted.bam\n",
      "\tCNA_10xv11_2.35k\n",
      "\tCNA_10xv11_2\n",
      "35k/35k_preprocessing_out/data/bam/CNA_10xv11_3.35k___CNA_10xv11_3__R1.bwa.out.possorted.bam\n",
      "\tCNA_10xv11_3.35k\n",
      "\tCNA_10xv11_3\n",
      "35k/35k_preprocessing_out/data/bam/CNA_10xv11_4.35k___CNA_10xv11_4__R1.bwa.out.possorted.bam\n",
      "\tCNA_10xv11_4.35k\n",
      "\tCNA_10xv11_4\n",
      "35k/35k_preprocessing_out/data/bam/CNA_10xv11_5.35k___CNA_10xv11_5__R1.bwa.out.possorted.bam\n",
      "\tCNA_10xv11_5.35k\n",
      "\tCNA_10xv11_5\n",
      "35k/35k_preprocessing_out/data/bam/CNA_10xv2_1.35k___CNA_10xv2_1__R1.bwa.out.possorted.bam\n",
      "\tCNA_10xv2_1.35k\n",
      "\tCNA_10xv2_1\n",
      "35k/35k_preprocessing_out/data/bam/CNA_10xv2_2.35k___CNA_10xv2_2__R1.bwa.out.possorted.bam\n",
      "\tCNA_10xv2_2.35k\n",
      "\tCNA_10xv2_2\n",
      "35k/35k_preprocessing_out/data/bam/CNA_hydrop_1.35k___CNA_hydrop_1__R1.bwa.out.possorted.bam\n",
      "\tCNA_hydrop_1.35k\n",
      "\tCNA_hydrop_1\n",
      "35k/35k_preprocessing_out/data/bam/CNA_hydrop_2.35k___CNA_hydrop_2__R1.bwa.out.possorted.bam\n",
      "\tCNA_hydrop_2.35k\n",
      "\tCNA_hydrop_2\n",
      "35k/35k_preprocessing_out/data/bam/CNA_hydrop_3.35k___CNA_hydrop_3__R1.bwa.out.possorted.bam\n",
      "\tCNA_hydrop_3.35k\n",
      "\tCNA_hydrop_3\n",
      "35k/35k_preprocessing_out/data/bam/CNA_mtscatac_1.35k___CNA_mtscatac_1__R1.bwa.out.possorted.bam\n",
      "\tCNA_mtscatac_1.35k\n",
      "\tCNA_mtscatac_1\n",
      "35k/35k_preprocessing_out/data/bam/CNA_mtscatac_2.35k___CNA_mtscatac_2__R1.bwa.out.possorted.bam\n",
      "\tCNA_mtscatac_2.35k\n",
      "\tCNA_mtscatac_2\n",
      "35k/35k_preprocessing_out/data/bam/EPF_hydrop_1.35k___EPF_hydrop_1__R1.bwa.out.possorted.bam\n",
      "\tEPF_hydrop_1.35k\n",
      "\tEPF_hydrop_1\n",
      "35k/35k_preprocessing_out/data/bam/EPF_hydrop_2.35k___EPF_hydrop_2__R1.bwa.out.possorted.bam\n",
      "\tEPF_hydrop_2.35k\n",
      "\tEPF_hydrop_2\n",
      "35k/35k_preprocessing_out/data/bam/EPF_hydrop_3.35k___EPF_hydrop_3__R1.bwa.out.possorted.bam\n",
      "\tEPF_hydrop_3.35k\n",
      "\tEPF_hydrop_3\n",
      "35k/35k_preprocessing_out/data/bam/EPF_hydrop_4.35k___EPF_hydrop_4__R1.bwa.out.possorted.bam\n",
      "\tEPF_hydrop_4.35k\n",
      "\tEPF_hydrop_4\n",
      "35k/35k_preprocessing_out/data/bam/HAR_ddseq_1.35k___HAR_ddseq_1__R1.bwa.out.possorted.bam\n",
      "\tHAR_ddseq_1.35k\n",
      "\tHAR_ddseq_1\n",
      "35k/35k_preprocessing_out/data/bam/HAR_ddseq_2.35k___HAR_ddseq_2__R1.bwa.out.possorted.bam\n",
      "\tHAR_ddseq_2.35k\n",
      "\tHAR_ddseq_2\n",
      "35k/35k_preprocessing_out/data/bam/MDC_mtscatac_1.35k___MDC_mtscatac_1__R1.bwa.out.possorted.bam\n",
      "\tMDC_mtscatac_1.35k\n",
      "\tMDC_mtscatac_1\n",
      "\t35k/35k_preprocessing_out/data/reports/mapping_stats/MDC_mtscatac_1.35k___MDC_mtscatac_1__R1.mapping_stats.tsv does not exist!\n",
      "35k/35k_preprocessing_out/data/bam/MDC_mtscatac_2.35k___MDC_mtscatac_2__R1.bwa.out.possorted.bam\n",
      "\tMDC_mtscatac_2.35k\n",
      "\tMDC_mtscatac_2\n",
      "35k/35k_preprocessing_out/data/bam/OHS_s3atac_1.35k___OHS_s3atac_1__R1.bwa.out.possorted.bam\n",
      "\tOHS_s3atac_1.35k\n",
      "\tOHS_s3atac_1\n",
      "35k/35k_preprocessing_out/data/bam/OHS_s3atac_2.35k___OHS_s3atac_2__R1.bwa.out.possorted.bam\n",
      "\tOHS_s3atac_2.35k\n",
      "\tOHS_s3atac_2\n",
      "35k/35k_preprocessing_out/data/bam/SAN_10xmultiome_1.35k___SAN_10xmultiome_1__R1.bwa.out.possorted.bam\n",
      "\tSAN_10xmultiome_1.35k\n",
      "\tSAN_10xmultiome_1\n",
      "35k/35k_preprocessing_out/data/bam/SAN_10xmultiome_2.35k___SAN_10xmultiome_2__R1.bwa.out.possorted.bam\n",
      "\tSAN_10xmultiome_2.35k\n",
      "\tSAN_10xmultiome_2\n",
      "35k/35k_preprocessing_out/data/bam/STA_10xv11_1.35k___STA_10xv11_1__R1.bwa.out.possorted.bam\n",
      "\tSTA_10xv11_1.35k\n",
      "\tSTA_10xv11_1\n",
      "35k/35k_preprocessing_out/data/bam/STA_10xv11_2.35k___STA_10xv11_2__R1.bwa.out.possorted.bam\n",
      "\tSTA_10xv11_2.35k\n",
      "\tSTA_10xv11_2\n",
      "35k/35k_preprocessing_out/data/bam/TXG_10xv11_1.35k___TXG_10xv11_1__R1.bwa.out.possorted.bam\n",
      "\tTXG_10xv11_1.35k\n",
      "\tTXG_10xv11_1\n",
      "\t35k/35k_preprocessing_out/data/reports/mapping_stats/TXG_10xv11_1.35k___TXG_10xv11_1__R1.mapping_stats.tsv does not exist!\n",
      "35k/35k_preprocessing_out/data/bam/TXG_10xv2_1.35k___TXG_10xv2_1__R1.bwa.out.possorted.bam\n",
      "\tTXG_10xv2_1.35k\n",
      "\tTXG_10xv2_1\n",
      "35k/35k_preprocessing_out/data/bam/TXG_10xv2_2.35k___TXG_10xv2_2__R1.bwa.out.possorted.bam\n",
      "\tTXG_10xv2_2.35k\n",
      "\tTXG_10xv2_2\n",
      "35k/35k_preprocessing_out/data/bam/VIB_10xmultiome_1.35k___VIB_10xmultiome_1__R1.bwa.out.possorted.bam\n",
      "\tVIB_10xmultiome_1.35k\n",
      "\tVIB_10xmultiome_1\n",
      "35k/35k_preprocessing_out/data/bam/VIB_10xmultiome_2.35k___VIB_10xmultiome_2__R1.bwa.out.possorted.bam\n",
      "\tVIB_10xmultiome_2.35k\n",
      "\tVIB_10xmultiome_2\n",
      "35k/35k_preprocessing_out/data/bam/VIB_10xv1_1.35k___VIB_10xv1_1__R1.bwa.out.possorted.bam\n",
      "\tVIB_10xv1_1.35k\n",
      "\tVIB_10xv1_1\n",
      "35k/35k_preprocessing_out/data/bam/VIB_10xv1_2.35k___VIB_10xv1_2__R1.bwa.out.possorted.bam\n",
      "\tVIB_10xv1_2.35k\n",
      "\tVIB_10xv1_2\n",
      "35k/35k_preprocessing_out/data/bam/VIB_10xv2_1.35k___VIB_10xv2_1__R1.bwa.out.possorted.bam\n",
      "\tVIB_10xv2_1.35k\n",
      "\tVIB_10xv2_1\n",
      "35k/35k_preprocessing_out/data/bam/VIB_10xv2_2.35k___VIB_10xv2_2__R1.bwa.out.possorted.bam\n",
      "\tVIB_10xv2_2.35k\n",
      "\tVIB_10xv2_2\n",
      "35k/35k_preprocessing_out/data/bam/VIB_hydrop_11.35k___VIB_hydrop_11__R1.bwa.out.possorted.bam\n",
      "\tVIB_hydrop_11.35k\n",
      "\tVIB_hydrop_11\n",
      "35k/35k_preprocessing_out/data/bam/VIB_hydrop_12.35k___VIB_hydrop_12__R1.bwa.out.possorted.bam\n",
      "\tVIB_hydrop_12.35k\n",
      "\tVIB_hydrop_12\n",
      "35k/35k_preprocessing_out/data/bam/VIB_hydrop_21.35k___VIB_hydrop_21__R1.bwa.out.possorted.bam\n",
      "\tVIB_hydrop_21.35k\n",
      "\tVIB_hydrop_21\n",
      "35k/35k_preprocessing_out/data/bam/VIB_hydrop_22.35k___VIB_hydrop_22__R1.bwa.out.possorted.bam\n",
      "\tVIB_hydrop_22.35k\n",
      "\tVIB_hydrop_22\n"
     ]
    }
   ],
   "source": [
    "dirs=(35k 30k 25k)\n",
    "for dir in $dirs\n",
    "do\n",
    "    echo $dir\n",
    "    for file in $(ls $dir/${dir}_preprocessing_out/data/bam/*.bam)\n",
    "    do\n",
    "        echo $file\n",
    "        sample=`basename $file`\n",
    "        sample=(${sample//___/ })\n",
    "        sample=${sample[0]}\n",
    "        sample_raw=(${sample//\\./ })\n",
    "        echo -e \"\\t\"$sample\n",
    "        echo -e \"\\t\"$sample_raw\n",
    "        mapping_stats=$dir/${dir}_preprocessing_out/data/reports/mapping_stats/${sample}___${sample_raw}__R1.mapping_stats.tsv\n",
    "        if [ ! -f \"$mapping_stats\" ]; then\n",
    "            echo -e \"\\t\"$mapping_stats does not exist!\n",
    "            echo -e \"\\t\\t\"$file\n",
    "            echo -e \"\\t\\t\"$mapping_stats\n",
    "            # Get mapping statistics from BAM file:\n",
    "            #   - Read BAM file and write uncompressed BAM.\n",
    "            #   - Uncompressed BAM file is written to each samtools command with tee (writes to each specified file and stdout).\n",
    "            #   - samtools commands:\n",
    "            #       - Get samtools statistics with:\n",
    "            #           samtools stat \"${bam}\" > \"${sampleId}.stat\"\n",
    "            #       - Uniquely mapped reads (BWA):\n",
    "            #           samtools view -c -F 0x4 -F 0x100 -F 0x800 -e '! [XA] && ! [SA]' \"${bam}\"\n",
    "            #       - Fraction of total read pairs mapped confidently to genome (>30 mapq):\n",
    "            #           samtools view -c -F 0x4 -F 0x100 -F 0x800 -q 30 \"${bam}\"\n",
    "            #   - Only use threads for \"samtools stat\". Using it with any of the other samtools commands\n",
    "            #     makes everything slower than not using any threads at all.\n",
    "            samtools view -u \"${file}\" \\\n",
    "              | tee \\\n",
    "                    >(samtools view -c -F 0x4 -F 0x100 -F 0x800 -e '! [XA] && ! [SA]' - > \"${sample}.uniquely_mapped_reads.txt\") \\\n",
    "                    >(samtools view -c -F 0x4 -F 0x100 -F 0x800 -q 30 - > \"${sample}.fraction_total_read_pairs.txt\") \\\n",
    "              | samtools stat -@ 2 - > \"${sample}.stat\"\n",
    "\n",
    "            printf \"\\t${sample}\\n\" > \"${mapping_stats}\";\n",
    "\n",
    "            grep '^SN' \"${sample}.stat\" | cut -f 2,3 >> \"${mapping_stats}\";\n",
    "\n",
    "            printf \"Uniquely mapped reads:\\t\" >> \"${mapping_stats}\";\n",
    "            cat \"${sample}.uniquely_mapped_reads.txt\" >> \"${mapping_stats}\";\n",
    "\n",
    "            printf \"Reads mapped with MAPQ>30:\\t\" >> \"${mapping_stats}\";\n",
    "            cat \"${sample}.fraction_total_read_pairs.txt\" >> \"${mapping_stats}\";\n",
    "\n",
    "            rm \"${sample}.uniquely_mapped_reads.txt\" \"${sample}.fraction_total_read_pairs.txt\";\n",
    "        fi\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c958174d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
